// codetocloudmcp/out/toolsImplementations/azureQuotaCheck/azureQuotaChecker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PostgreSQLQuotaChecker = exports.SQLQuotaChecker = exports.ResourceProvider = void 0;
exports.GetAzureQuota = GetAzureQuota;
const skuParser_1 = require("../shared/skuParser");
// For simplicity, we currently apply a single rule for all Azure resource providers:
//   - Any resource provider not listed in the enum is treated as having no quota limitations.
// Ideally, we’d differentiate between the following cases:
//   1. The resource provider has no quota limitations.
//   2. The resource provider has quota limitations but does not expose a quota API.
//   3. The resource provider exposes a quota API, but it's not yet supported by the checker.
var ResourceProvider;
(function (ResourceProvider) {
    ResourceProvider["CognitiveServices"] = "Microsoft.CognitiveServices";
    ResourceProvider["Compute"] = "Microsoft.Compute";
    ResourceProvider["Storage"] = "Microsoft.Storage";
    ResourceProvider["ContainerApp"] = "Microsoft.App";
    ResourceProvider["Network"] = "Microsoft.Network";
    ResourceProvider["MachineLearning"] = "Microsoft.MachineLearningServices";
    ResourceProvider["PostgreSQL"] = "Microsoft.DBforPostgreSQL";
    ResourceProvider["HDInsight"] = "Microsoft.HDInsight";
    ResourceProvider["Search"] = "Microsoft.Search";
    ResourceProvider["ContainerInstance"] = "Microsoft.ContainerInstance";
    ResourceProvider["SQL"] = "Microsoft.Sql";
    ResourceProvider["AppServiceManagedInstance"] = "appservicemanagedinstance";
})(ResourceProvider || (exports.ResourceProvider = ResourceProvider = {}));
const emptyQuotaInfo = [{
        name: "NoSKUAvaliable",
        limit: 0,
        used: 0
    }];
// Abstract base class for checking Azure quotas
class AzureQuotaChecker {
    azureClient;
    constructor(azureClient) {
        this.azureClient = azureClient;
    }
}
// Compute Quota Checker Implementation
class ComputeQuotaChecker extends AzureQuotaChecker {
    constructor(azureClient) {
        super(azureClient);
    }
    async getUsagesForLocation(location) {
        const usages = await this.azureClient.getComputeUsageForLocation(location);
        const usageMap = new Map();
        for (const usage of usages) {
            if (usage.name) {
                const normalizedName = usage.name.replace(/\s+/g, '').toLowerCase();
                usageMap.set(normalizedName, {
                    limit: usage.limit,
                    used: usage.used
                });
            }
        }
        return usageMap;
    }
    async getQuotaForLocation(location) {
        try {
            const skus = await this.azureClient.getComputeSkus(location);
            const usageMap = await this.getUsagesForLocation(location);
            // Check VM quota availability
            const vmUsage = usageMap.get('virtualmachines');
            if (!vmUsage) {
                console.log(`No virtualMachines usage info found for location: ${location}`);
                return emptyQuotaInfo;
            }
            if (vmUsage.limit - vmUsage.used <= 0) {
                console.log(`No virtualMachines quota available for location: ${location}`);
                return emptyQuotaInfo;
            }
            // Check available cores
            const coresUsage = usageMap.get('cores');
            const availableCores = coresUsage ? coresUsage.limit - coresUsage.used : 0;
            // Collect, sort and select top VM SKUs
            const skusBySeries = (0, skuParser_1.collectVmSkusBySeries)(skus, usageMap, availableCores);
            const result = (0, skuParser_1.sortAndSelectTopSkus)(skusBySeries);
            if (result.length === 0) {
                return emptyQuotaInfo;
            }
            return result;
        }
        catch (error) {
            console.error(`Error fetching VM quotas: ${error}`);
            return [];
        }
    }
}
// Microsoft.CognitiveServices Quota Checker Implementation
class CognitiveServicesQuotaChecker extends AzureQuotaChecker {
    constructor(azureClient) {
        super(azureClient);
    }
    async getQuotaForLocation(location) {
        try {
            return await this.azureClient.getCognitiveServicesUsageForLocation(location);
        }
        catch (error) {
            console.error(`Error fetching cognitive services quotas: ${error}`);
            return [];
        }
    }
}
class SQLQuotaChecker extends AzureQuotaChecker {
    constructor(azureClient) {
        super(azureClient);
    }
    getSkuDetail(name, remainingQuota) {
        if (name.toLowerCase() === "serverquota") {
            return `${name} (must be checked for all models, remaining quota: ${remainingQuota})`;
        }
        if (name.toLowerCase() === "regionalvcorequotaforsqldbanddw") {
            return `${name} (must be checked if choose vCore model, remaining quota: ${remainingQuota})`;
        }
        if (name.toLowerCase() === "subscriptionfreedatabasecount") {
            return `${name} (must be checked if choose free tier, remaining quota: ${remainingQuota})`;
        }
        return `${name} (remaining quota: ${remainingQuota})`;
    }
    async getQuotaForLocation(location) {
        try {
            const apiVersion = "2015-05-01-preview";
            const requestUrl = `https://management.azure.com/subscriptions/${this.azureClient.getCurrentSubscriptionId()}/providers/Microsoft.Sql/locations/${location}/usages?api-version=${apiVersion}`;
            console.log(`requestUrl: ${requestUrl}`);
            const rawResponse = await this.azureClient.queryManagementPlane(requestUrl);
            if (!rawResponse || !rawResponse.value) {
                return [];
            }
            const skuNameList = ["ServerQuota", "RegionalVCoreQuotaForSQLDBAndDW", "SubscriptionFreeDatabaseCount", "SubnetQuota"];
            if (rawResponse.value.some((q) => q.name === "ServerQuota" && q.properties?.limit - q.properties?.currentValue <= 0)) {
                return emptyQuotaInfo;
            }
            return rawResponse.value
                .filter((item) => skuNameList.includes(item.name || ''))
                .map((item) => ({
                name: this.getSkuDetail(item.name || '', item.properties?.limit - item.properties?.currentValue),
                limit: item.properties?.limit || 0,
                used: item.properties?.currentValue || 0,
                unit: item.properties?.unit
            }));
        }
        catch (error) {
            console.error(`Error fetching SQL quotas: ${error}`);
            return [];
        }
    }
}
exports.SQLQuotaChecker = SQLQuotaChecker;
// Storage Quota Checker Implementation
class StorageQuotaChecker extends AzureQuotaChecker {
    constructor(azureClient) {
        super(azureClient);
    }
    async getQuotaForLocation(location) {
        try {
            return await this.azureClient.getStorageUsageForLocation(location);
        }
        catch (error) {
            console.error(`Error fetching storage quotas: ${error}`);
            return [];
        }
    }
}
// Container Apps Quota Checker Implementation
class ContainerAppQuotaChecker extends AzureQuotaChecker {
    constructor(azureClient) {
        super(azureClient);
    }
    async getQuotaForLocation(location) {
        try {
            return await this.azureClient.getContainerAppUsageForLocation(location);
        }
        catch (error) {
            console.error(`Error fetching Container Apps quotas: ${error}`);
            return [];
        }
    }
}
class AppServiceManagedInstanceQuotaChecker extends AzureQuotaChecker {
    constructor(azureClient) {
        super(azureClient);
    }
    async getQuotaForLocation(location) {
        try {
            return await this.azureClient.getAppServiceManagedInstanceUsageForLocation(location);
        }
        catch (error) {
            console.error(`Error fetching App Service Managed Instance quotas: ${error}`);
            return [];
        }
    }
}
// Network Quota Checker Implementation
class NetworkQuotaChecker extends AzureQuotaChecker {
    constructor(azureClient) {
        super(azureClient);
    }
    async getQuotaForLocation(location) {
        try {
            return await this.azureClient.getNetworkUsageForLocation(location);
        }
        catch (error) {
            console.error(`Error fetching network quotas: ${error}`);
            return [];
        }
    }
}
// Machine Learning Services Quota Checker Implementation
class MachineLearningQuotaChecker extends AzureQuotaChecker {
    constructor(azureClient) {
        super(azureClient);
    }
    async getQuotaForLocation(location) {
        try {
            return await this.azureClient.getAMLUsageForLocation(location);
        }
        catch (error) {
            console.error(`Error fetching Machine Learning Services quotas: ${error}`);
            return [];
        }
    }
}
// PostgreSQL Quota Checker Implementation
class PostgreSQLQuotaChecker extends AzureQuotaChecker {
    constructor(azureClient) {
        super(azureClient);
    }
    getSkuDetail(name, remainingQuota) {
        // Microsoft.DBforPostgreSQL/flexibleServers: cores, standardBSFamily, standardDADSv5Family, standardDDSv4Family, standardDDSv5Family, standardDSv3Family, standardEADSv5Family, standardEDSv4Family, standardEDSv5Family, standardESv3Family
        if (name.startsWith("standardB")) {
            return `${name} (Burstable tier)`;
        }
        if (name.startsWith("standardD")) {
            let skuParts = name.slice(9).split('Family')[0];
            return `${name} (GeneralPurpose tier, e.g. Standard_D2${skuParts.slice(0, skuParts.length - 2).toLowerCase()}_${skuParts.slice(-2)})`;
        }
        if (name.startsWith("standardE")) {
            let skuParts = name.slice(9).split('Family')[0];
            return `${name} (MemoryOptimized tier, e.g. Standard_E2${skuParts.slice(0, skuParts.length - 2).toLowerCase()}_${skuParts.slice(-2)})`;
        }
        if (name === "cores") {
            return `${name} (remaining quota: ${remainingQuota})`;
        }
        return `${name}`;
    }
    async getQuotaForLocation(location) {
        try {
            const requestUrl = `https://management.azure.com/subscriptions/${this.azureClient.getCurrentSubscriptionId()}/providers/Microsoft.DBforPostgreSQL/locations/${location}/resourceType/flexibleServers/usages?api-version=2023-06-01-preview`;
            const rawResponse = await this.azureClient.queryManagementPlane(requestUrl);
            if (!rawResponse || !rawResponse.value) {
                return emptyQuotaInfo;
            }
            if (rawResponse.value.some((q) => q.name.value === "cores" && q.limit - q.currentValue < 2)) {
                return emptyQuotaInfo;
            }
            return rawResponse.value.map((item) => ({
                name: this.getSkuDetail(item.name?.value || '', item.limit - item.currentValue),
                limit: item.limit || 0,
                used: item.currentValue || 0,
                unit: item.unit
            }));
        }
        catch (error) {
            console.error(`Error fetching PostgreSQL quotas: ${error}`);
            return [];
        }
    }
}
exports.PostgreSQLQuotaChecker = PostgreSQLQuotaChecker;
// HDInsight Quota Checker Implementation
class HDInsightQuotaChecker extends AzureQuotaChecker {
    constructor(azureClient) {
        super(azureClient);
    }
    async getQuotaForLocation(location) {
        try {
            const requestUrl = `https://management.azure.com/subscriptions/${this.azureClient.getCurrentSubscriptionId()}/providers/Microsoft.HDInsight/locations/${location}/usages?api-version=2018-06-01-preview`;
            const rawResponse = await this.azureClient.queryManagementPlane(requestUrl);
            if (!rawResponse || !rawResponse.value) {
                return [];
            }
            return rawResponse.value.map((item) => ({
                name: item.name?.value || '',
                limit: item.limit || 0,
                used: item.currentValue || 0,
                unit: item.unit
            }));
        }
        catch (error) {
            console.error(`Error fetching HDInsight quotas: ${error}`);
            return [];
        }
    }
}
// Search Quota Checker Implementation
class SearchQuotaChecker extends AzureQuotaChecker {
    constructor(azureClient) {
        super(azureClient);
    }
    async getQuotaForLocation(location) {
        try {
            const requestUrl = `https://management.azure.com/subscriptions/${this.azureClient.getCurrentSubscriptionId()}/providers/Microsoft.Search/locations/${location}/usages?api-version=2023-11-01`;
            const rawResponse = await this.azureClient.queryManagementPlane(requestUrl);
            if (!rawResponse || !rawResponse.value) {
                return [];
            }
            return rawResponse.value.map((item) => ({
                name: item.name?.value || '',
                limit: item.limit || 0,
                used: item.currentValue || 0,
                unit: item.unit
            }));
        }
        catch (error) {
            console.error(`Error fetching Search quotas: ${error}`);
            return [];
        }
    }
}
// Container Instance Quota Checker Implementation
class ContainerInstanceQuotaChecker extends AzureQuotaChecker {
    constructor(azureClient) {
        super(azureClient);
    }
    async getQuotaForLocation(location) {
        try {
            const requestUrl = `https://management.azure.com/subscriptions/${this.azureClient.getCurrentSubscriptionId()}/providers/Microsoft.ContainerInstance/locations/${location}/usages?api-version=2023-05-01`;
            const rawResponse = await this.azureClient.queryManagementPlane(requestUrl);
            if (!rawResponse || !rawResponse.value) {
                return [];
            }
            return rawResponse.value.map((item) => ({
                name: item.name?.value || '',
                limit: item.limit || 0,
                used: item.currentValue || 0,
                unit: item.unit
            }));
        }
        catch (error) {
            console.error(`Error fetching Container Instance quotas: ${error}`);
            return [];
        }
    }
}
// Factory function to create quota checkers
function createQuotaChecker(provider, azureClient) {
    const values = Object.values(ResourceProvider);
    const matchingProvider = values.find(enumValue => enumValue === provider);
    if (!matchingProvider) {
        throw new Error(`Unsupported resource provider: ${provider}`);
    }
    switch (matchingProvider) {
        case ResourceProvider.Compute:
            return new ComputeQuotaChecker(azureClient);
        case ResourceProvider.CognitiveServices:
            return new CognitiveServicesQuotaChecker(azureClient);
        case ResourceProvider.Storage:
            return new StorageQuotaChecker(azureClient);
        case ResourceProvider.ContainerApp:
            return new ContainerAppQuotaChecker(azureClient);
        case ResourceProvider.Network:
            return new NetworkQuotaChecker(azureClient);
        case ResourceProvider.MachineLearning:
            return new MachineLearningQuotaChecker(azureClient);
        case ResourceProvider.PostgreSQL:
            return new PostgreSQLQuotaChecker(azureClient);
        case ResourceProvider.HDInsight:
            return new HDInsightQuotaChecker(azureClient);
        case ResourceProvider.Search:
            return new SearchQuotaChecker(azureClient);
        case ResourceProvider.ContainerInstance:
            return new ContainerInstanceQuotaChecker(azureClient);
        case ResourceProvider.SQL:
            console.log("Checking for SQL provider");
            return new SQLQuotaChecker(azureClient);
        case ResourceProvider.AppServiceManagedInstance:
            return new AppServiceManagedInstanceQuotaChecker(azureClient);
        default:
            throw new Error(`No implementation for provider: ${provider}`);
    }
}
// Function to get Azure quota for a list of resource types
async function GetAzureQuota(resourceTypes, azureClient, location) {
    const result = {};
    const processedProviders = new Set();
    const quotaPromises = resourceTypes.map(resourceType => new Promise(async (resolve) => {
        try {
            const provider = resourceType.split('/')[0];
            if (processedProviders.has(provider)) {
                resolve();
                return;
            }
            processedProviders.add(provider);
            const quotaChecker = createQuotaChecker(provider, azureClient);
            console.log(`Checking for quota ${location}`);
            const quotaInfo = await quotaChecker.getQuotaForLocation(location);
            resourceTypes
                .filter(rt => rt.split('/')[0] === provider)
                .forEach(rt => {
                result[rt] = quotaInfo;
            });
        }
        catch (error) {
            result[resourceType] = [];
        }
        resolve();
    }));
    await Promise.all(quotaPromises);
    return result;
}
