// containerization-assist-mcp/dist/src/config/policy-rego.js
/**
 * Rego Policy Evaluation Module
 * Hybrid approach: WASM (fast, no deps) + OPA CLI fallback
 *
 * This module provides integration with Open Policy Agent (OPA) for policy evaluation.
 * It supports loading and evaluating Rego policies against Dockerfile and Kubernetes content.
 *
 * **Evaluation Strategy:**
 * 1. Pre-compiled WASM bundle (policies/compiled/policies.wasm) - Fast, zero-dependency
 * 2. OPA CLI fallback for custom .rego files - Requires OPA installation
 */
import { readFile, writeFile } from 'node:fs/promises';
import { existsSync } from 'node:fs';
import { execFile } from 'node:child_process';
import { promisify } from 'node:util';
import { join, dirname, basename } from 'node:path';
import tmp from 'tmp';
import { loadPolicy as loadWasmPolicy } from '@open-policy-agent/opa-wasm';
import { Success, Failure } from '../types/index.js';
import { ERROR_MESSAGES } from '../lib/errors.js';
const execFileAsync = promisify(execFile);
/**
 * Default OPA policy namespace
 *
 * All containerization policies should export results under this namespace.
 * Used in OPA queries: data.containerization.security.result
 */
const POLICY_NAMESPACE = 'containerization';
/**
 * Built-in policy mappings
 * Maps policy filenames to their WASM bundle entrypoint names
 */
const BUILT_IN_POLICY_MODULES = {
    'security-baseline.rego': `${POLICY_NAMESPACE}/security/result`,
    'base-images.rego': `${POLICY_NAMESPACE}/base_images/result`,
    'container-best-practices.rego': `${POLICY_NAMESPACE}/best_practices/result`,
};
/**
 * Cached project root to avoid blocking the event loop on repeated calls
 */
let cachedProjectRoot;
/**
 * Get project root directory
 * Works in both ESM and CJS builds
 *
 * Note: This searches from process.cwd() upward, which works for:
 * - Direct package usage (policies/ in package root)
 * - Monorepos (policies/ in package directory)
 *
 * For packages installed in node_modules, the WASM bundle is shipped
 * in the package and resolved relative to the built dist/ directory.
 */
function getProjectRoot() {
    if (cachedProjectRoot) {
        return cachedProjectRoot;
    }
    // Try to find policies directory by walking up from current working directory
    let currentDir = process.cwd();
    while (currentDir !== dirname(currentDir)) {
        const policiesDir = join(currentDir, 'policies');
        if (existsSync(policiesDir)) {
            cachedProjectRoot = currentDir;
            return currentDir;
        }
        currentDir = dirname(currentDir);
    }
    // Fall back to cwd (works for npm packages since policies/ is at package root)
    cachedProjectRoot = process.cwd();
    return cachedProjectRoot;
}
const projectRoot = getProjectRoot();
/**
 * Helper function to implement evaluatePolicy wrapper around evaluate
 * This provides a convenience method for tests that wraps evaluate with Result handling
 */
function createEvaluatePolicyWrapper(evaluate) {
    return async (result, _packageName) => {
        if (!result.ok) {
            return result;
        }
        try {
            const policyResult = await evaluate(result.value);
            return Success({
                blocking: policyResult.violations,
                warnings: policyResult.warnings,
                suggestions: policyResult.suggestions,
            });
        }
        catch (error) {
            const message = error instanceof Error ? error.message : String(error);
            return Failure(`Policy evaluation failed: ${message}`);
        }
    };
}
/**
 * Get path to OPA binary
 */
function getOpaBinaryPath() {
    // Determine platform-specific OPA binary name
    const opaBinaryName = process.platform === 'win32' ? 'opa.exe' : 'opa';
    // First try project's node_modules
    const localOpa = join(process.cwd(), 'node_modules', '.bin', opaBinaryName);
    if (existsSync(localOpa)) {
        return localOpa;
    }
    // Fall back to system OPA
    return opaBinaryName;
}
/**
 * Create an error response for missing policy entrypoint mapping
 */
function createPolicyMappingError(policyFileName) {
    return {
        allow: false,
        violations: [
            {
                rule: 'policy-mapping-error',
                category: 'system',
                message: `No WASM entrypoint mapping for policy: ${policyFileName}`,
                severity: 'block',
            },
        ],
        warnings: [],
        suggestions: [],
    };
}
/**
 * Evaluate a specific WASM policy module (for single policy file loads)
 */
async function evaluateWasmPolicyModule(wasmPolicy, input, entrypointName, logger) {
    try {
        const inputData = typeof input === 'string' ? { content: input } : input;
        logger.debug({ entrypoint: entrypointName }, 'Evaluating WASM policy module');
        const resultSet = wasmPolicy.evaluate(inputData, entrypointName);
        if (!resultSet || resultSet.length === 0 || !resultSet[0]?.result) {
            logger.warn({ entrypoint: entrypointName }, 'WASM policy module returned no results');
            return {
                allow: true,
                violations: [],
                warnings: [],
                suggestions: [],
            };
        }
        const moduleResult = resultSet[0].result;
        logger.info({
            allow: moduleResult.allow ?? true,
            violations: moduleResult.violations?.length ?? 0,
            warnings: moduleResult.warnings?.length ?? 0,
            suggestions: moduleResult.suggestions?.length ?? 0,
        }, 'WASM policy module evaluation completed');
        return {
            allow: moduleResult.allow ?? true,
            violations: moduleResult.violations ?? [],
            warnings: moduleResult.warnings ?? [],
            suggestions: moduleResult.suggestions ?? [],
            summary: {
                total_violations: moduleResult.violations?.length ?? 0,
                total_warnings: moduleResult.warnings?.length ?? 0,
                total_suggestions: moduleResult.suggestions?.length ?? 0,
            },
        };
    }
    catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        logger.error({ error: message }, 'WASM policy module evaluation failed');
        return {
            allow: false,
            violations: [
                {
                    rule: 'policy-evaluation-error',
                    category: 'system',
                    message: `WASM policy evaluation failed: ${message}`,
                    severity: 'block',
                },
            ],
            warnings: [],
            suggestions: [],
        };
    }
}
/**
 * Evaluate all WASM policy modules and merge their results
 *
 * This function is shared by both evaluate() and evaluatePolicy()
 * to avoid code duplication.
 *
 * @param wasmPolicy - Loaded WASM policy instance
 * @param inputData - Input data for policy evaluation (must be object)
 * @param policyPaths - Array of policy file paths to evaluate
 * @param logger - Logger instance for diagnostics
 * @returns Merged policy evaluation result
 */
function evaluateAllWasmPolicies(wasmPolicy, inputData, policyPaths, logger) {
    const allViolations = [];
    const allWarnings = [];
    const allSuggestions = [];
    let overallAllow = true;
    logger.debug({ policyCount: policyPaths.length }, 'Evaluating all WASM policy modules');
    for (const policyPath of policyPaths) {
        // Use path.basename() instead of regex splitting
        const fileName = basename(policyPath);
        const policyModule = BUILT_IN_POLICY_MODULES[fileName];
        if (!policyModule) {
            logger.debug({ fileName }, 'Policy file has no WASM entrypoint mapping, skipping');
            continue;
        }
        try {
            const wasmResult = wasmPolicy.evaluate(inputData, policyModule);
            if (!wasmResult?.[0]?.result) {
                logger.debug({ policyModule }, 'WASM policy returned no result');
                continue;
            }
            const result = wasmResult[0].result;
            // Merge violations
            if (result.violations) {
                allViolations.push(...result.violations);
            }
            // Merge warnings
            if (result.warnings) {
                allWarnings.push(...result.warnings);
            }
            // Merge suggestions
            if (result.suggestions) {
                allSuggestions.push(...result.suggestions);
            }
            // Merge allow (false if any policy blocks)
            if (result.allow === false) {
                overallAllow = false;
            }
            logger.debug({
                policy: fileName,
                violations: result.violations?.length || 0,
                warnings: result.warnings?.length || 0,
                suggestions: result.suggestions?.length || 0,
                allow: result.allow ?? true,
            }, 'WASM policy evaluated');
        }
        catch (error) {
            logger.warn({ error, policyModule, fileName }, 'WASM policy evaluation failed, skipping');
            // Continue with other policies even if one fails
        }
    }
    logger.info({
        totalViolations: allViolations.length,
        totalWarnings: allWarnings.length,
        totalSuggestions: allSuggestions.length,
        allow: overallAllow,
    }, 'All WASM policies evaluated and merged');
    return {
        allow: overallAllow,
        violations: allViolations,
        warnings: allWarnings,
        suggestions: allSuggestions,
        summary: {
            total_violations: allViolations.length,
            total_warnings: allWarnings.length,
            total_suggestions: allSuggestions.length,
        },
    };
}
/**
 * Load and compile a Rego policy from file
 *
 * Hybrid approach:
 * 1. Check for pre-compiled WASM (fast, no OPA required)
 * 2. Fall back to OPA binary for .rego files (requires OPA installed)
 *
 * @param policyPath - Path to .rego policy file
 * @param logger - Logger instance for diagnostics
 * @returns Result containing RegoEvaluator or error
 *
 * @example
 * ```typescript
 * const result = await loadRegoPolicy('policies/security.rego', logger);
 * if (result.ok) {
 *   const evalResult = await result.value.evaluate(dockerfileContent);
 *   if (!evalResult.allow) {
 *     console.log('Violations:', evalResult.violations);
 *   }
 * }
 * ```
 */
export async function loadRegoPolicy(policyPath, logger) {
    try {
        // Validate file exists
        if (!existsSync(policyPath)) {
            return Failure(`Policy file not found: ${policyPath}`, {
                message: 'Rego policy file does not exist',
                hint: `Attempted to load: ${policyPath}\nCurrent directory: ${process.cwd()}`,
                resolution: 'Ensure the policy file path is correct and the file exists',
            });
        }
        // Validate .rego extension
        if (!policyPath.endsWith('.rego')) {
            return Failure('Only .rego policy files are supported', {
                message: 'Invalid policy file format',
                hint: `File: ${policyPath}`,
                resolution: 'Provide a .rego policy file',
            });
        }
        // Check for pre-compiled WASM bundle (fast path)
        // Only use WASM when loading one of the built-in policy files
        // This ensures tests that load specific policies get only those policies
        const policyFileName = basename(policyPath);
        const isBuiltInPolicy = policyFileName in BUILT_IN_POLICY_MODULES;
        const wasmPath = join(projectRoot, 'policies', 'compiled', 'policies.wasm');
        if (isBuiltInPolicy && existsSync(wasmPath)) {
            try {
                logger.info({ wasmPath, policyFile: policyFileName }, 'Loading pre-compiled WASM policy bundle (fast path)');
                const wasmBytes = await readFile(wasmPath);
                const wasmPolicy = await loadWasmPolicy(wasmBytes);
                const evaluator = {
                    policyPaths: [policyPath],
                    evaluate: async (input) => {
                        // Get the WASM entrypoint for the requested policy module
                        const entrypointName = BUILT_IN_POLICY_MODULES[policyFileName];
                        // This should always be defined since isBuiltInPolicy is true
                        // but TypeScript can't verify this statically
                        if (!entrypointName) {
                            logger.error({ policyFileName }, 'No entrypoint mapping found for built-in policy');
                            return createPolicyMappingError(policyFileName);
                        }
                        return evaluateWasmPolicyModule(wasmPolicy, input, entrypointName, logger);
                    },
                    evaluatePolicy: createEvaluatePolicyWrapper(async (input) => {
                        const entrypointName = BUILT_IN_POLICY_MODULES[policyFileName];
                        if (!entrypointName) {
                            return createPolicyMappingError(policyFileName);
                        }
                        return evaluateWasmPolicyModule(wasmPolicy, input, entrypointName, logger);
                    }),
                    queryConfig: async (packageName, input) => {
                        try {
                            logger.debug({ packageName, wasmMode: true }, 'Querying WASM policy for config');
                            const inputData = input;
                            // Query the WASM policy for the package
                            const resultSet = wasmPolicy.evaluate(inputData, packageName);
                            if (!resultSet || resultSet.length === 0 || !resultSet[0]?.result) {
                                logger.debug({ packageName }, 'WASM policy returned no config data');
                                return null;
                            }
                            logger.debug({ packageName, hasValue: true }, 'WASM policy config query successful');
                            return resultSet[0].result;
                        }
                        catch (error) {
                            const message = error instanceof Error ? error.message : String(error);
                            logger.warn({ error: message, packageName }, 'WASM policy config query failed, returning null');
                            return null;
                        }
                    },
                    close: () => {
                        logger.debug({ policyPath }, 'Cleaning up WASM policy resources');
                    },
                };
                logger.info({ policyModule: policyFileName }, 'WASM policy module loaded successfully');
                return Success(evaluator);
            }
            catch (wasmError) {
                logger.warn({
                    error: wasmError,
                    wasmPath,
                    policyFile: policyFileName,
                    reason: 'WASM bundle not found or policy not in bundle',
                }, 'Failed to load WASM bundle, falling back to OPA binary');
                // Add hint about building WASM
                logger.info('To use WASM mode (faster, no OPA required): npm run build:policies');
                // Continue to OPA binary fallback
            }
        }
        // Fall back to OPA binary evaluation
        logger.info({ policyPath }, 'Using OPA binary for policy evaluation (requires OPA installed)');
        // Read the policy file to validate it
        const policyContent = await readFile(policyPath, 'utf-8');
        logger.info({ policyPath, size: policyContent.length }, 'Loading Rego policy');
        // Test that OPA binary is available
        const opaBinary = getOpaBinaryPath();
        try {
            await execFileAsync(opaBinary, ['version']);
        }
        catch {
            return Failure('OPA binary not found', {
                message: 'Open Policy Agent (OPA) CLI is not installed',
                hint: `Tried to execute: ${opaBinary}\nPath: ${process.env.PATH}`,
                resolution: 'Install OPA from: https://www.openpolicyagent.org/docs/latest/#running-opa\n' +
                    'Or build WASM bundle: npm run build:policies',
            });
        }
        logger.info({ policyPath }, 'Rego policy loaded successfully via OPA binary');
        // Create evaluator
        const evaluator = {
            policyPaths: [policyPath],
            evaluate: async (input) => {
                return evaluateRegoPolicy(policyPath, input, logger);
            },
            evaluatePolicy: createEvaluatePolicyWrapper(async (input) => {
                return evaluateRegoPolicy(policyPath, input, logger);
            }),
            queryConfig: async (packageName, input) => {
                return queryRegoConfig(policyPath, packageName, input, logger);
            },
            close: () => {
                logger.debug({ policyPath }, 'Cleaning up Rego policy resources');
            },
        };
        return Success(evaluator);
    }
    catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        return Failure(ERROR_MESSAGES.POLICY_LOAD_FAILED(message), {
            message: 'Failed to load Rego policy',
            hint: `Policy: ${policyPath}\nError: ${message}`,
            resolution: existsSync(policyPath)
                ? 'Check policy file syntax with: opa check <policy-file>'
                : 'Ensure the policy file path is correct',
        });
    }
}
/**
 * Query Rego policy for configuration data using OPA CLI
 *
 * @param policyPaths - Path(s) to the Rego policy file(s)
 * @param packageName - Package name to query (e.g., 'containerization.generation_config')
 * @param input - Input data for the query
 * @param logger - Logger instance
 * @returns Configuration data or null if not found
 */
async function queryRegoConfig(policyPaths, packageName, input, logger) {
    try {
        logger.debug({ packageName, inputKeys: Object.keys(input) }, 'Querying Rego policy for config');
        // Create a secure temporary file for input using tmp package
        const tmpFile = tmp.fileSync({ prefix: 'opa-input-', postfix: '.json' });
        await writeFile(tmpFile.name, JSON.stringify(input));
        try {
            const opaBinary = getOpaBinaryPath();
            // Build args with multiple -d flags for each policy file
            const paths = Array.isArray(policyPaths) ? policyPaths : [policyPaths];
            const policyArgs = paths.flatMap(p => ['-d', p]);
            // Query the specific package for configuration
            const query = `data.${packageName}`;
            const { stdout, stderr } = await execFileAsync(opaBinary, [
                'eval',
                ...policyArgs,
                '-i', tmpFile.name,
                '-f', 'json',
                query,
            ], {
                maxBuffer: 10 * 1024 * 1024, // 10MB buffer
            });
            if (stderr) {
                logger.debug({ stderr }, 'OPA eval stderr output');
            }
            // Parse the OPA output
            const output = JSON.parse(stdout);
            const opaOutput = output;
            // OPA JSON format: { result: [{ expressions: [{ value: ... }] }] }
            if (opaOutput?.result && Array.isArray(opaOutput.result) && opaOutput.result.length > 0) {
                const firstResult = opaOutput.result[0];
                if (firstResult?.expressions &&
                    Array.isArray(firstResult.expressions) &&
                    firstResult.expressions.length > 0) {
                    const value = firstResult.expressions[0]?.value;
                    if (value !== undefined && value !== null) {
                        logger.debug({ packageName, hasValue: true }, 'Policy config query successful');
                        return value;
                    }
                }
            }
            logger.debug({ packageName }, 'Policy config query returned no data');
            return null;
        }
        finally {
            // Clean up temp file
            try {
                tmpFile.removeCallback();
            }
            catch {
                // Ignore cleanup errors
            }
        }
    }
    catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        logger.warn({ error: message, packageName }, 'Policy config query failed, returning null');
        return null;
    }
}
/**
 * Evaluate Rego policy against input using OPA CLI
 *
 * @param policyPaths - Path(s) to the Rego policy file(s)
 * @param input - Content to evaluate
 * @param logger - Logger instance
 * @returns Policy evaluation result
 */
async function evaluateRegoPolicy(policyPaths, input, logger) {
    try {
        // Convert input to the format expected by the policy
        const inputData = typeof input === 'string' ? { content: input } : input;
        logger.debug({ inputType: typeof input }, 'Evaluating Rego policy');
        // Create a secure temporary file for input using tmp package
        const tmpFile = tmp.fileSync({ prefix: 'opa-input-', postfix: '.json' });
        await writeFile(tmpFile.name, JSON.stringify(inputData));
        try {
            const opaBinary = getOpaBinaryPath();
            // Build args with multiple -d flags for each policy file (OPA will merge them)
            const paths = Array.isArray(policyPaths) ? policyPaths : [policyPaths];
            const policyArgs = paths.flatMap(p => ['-d', p]);
            // Run OPA eval command to evaluate the policy
            // Use -f json for JSON output and query data.containerization to get all results
            const { stdout, stderr } = await execFileAsync(opaBinary, [
                'eval',
                ...policyArgs,
                '-i', tmpFile.name,
                '-f', 'json',
                `data.${POLICY_NAMESPACE}`,
            ], {
                maxBuffer: 10 * 1024 * 1024, // 10MB buffer
            });
            if (stderr) {
                logger.debug({ stderr }, 'OPA eval stderr output');
            }
            // Parse the OPA output
            const combinedResult = {
                allow: true,
                violations: [],
                warnings: [],
                suggestions: [],
            };
            try {
                const output = JSON.parse(stdout);
                const opaOutput = output;
                // OPA JSON format: { result: [{ expressions: [{ value: ... }] }] }
                if (opaOutput?.result && Array.isArray(opaOutput.result) && opaOutput.result.length > 0) {
                    const firstResult = opaOutput.result[0];
                    if (firstResult?.expressions &&
                        Array.isArray(firstResult.expressions) &&
                        firstResult.expressions.length > 0) {
                        const containerization = firstResult.expressions[0]?.value;
                        if (containerization) {
                            // Dynamically merge results from ALL policy namespaces
                            // This allows custom policies in any namespace (e.g., platform, compliance, etc.)
                            for (const [, nsObj] of Object.entries(containerization)) {
                                // Only process keys where value is an object with a 'result' property
                                if (nsObj &&
                                    typeof nsObj === 'object' &&
                                    'result' in nsObj &&
                                    nsObj.result &&
                                    typeof nsObj.result === 'object') {
                                    const nsResult = nsObj.result;
                                    // Merge allow (false if any policy blocks)
                                    if (nsResult.allow === false) {
                                        combinedResult.allow = false;
                                    }
                                    // Merge violations, warnings, suggestions
                                    if (nsResult.violations) {
                                        const violations = Array.isArray(nsResult.violations)
                                            ? nsResult.violations
                                            : Object.values(nsResult.violations);
                                        combinedResult.violations.push(...violations);
                                    }
                                    if (nsResult.warnings) {
                                        const warnings = Array.isArray(nsResult.warnings)
                                            ? nsResult.warnings
                                            : Object.values(nsResult.warnings);
                                        combinedResult.warnings.push(...warnings);
                                    }
                                    if (nsResult.suggestions) {
                                        const suggestions = Array.isArray(nsResult.suggestions)
                                            ? nsResult.suggestions
                                            : Object.values(nsResult.suggestions);
                                        combinedResult.suggestions.push(...suggestions);
                                    }
                                }
                            }
                        }
                    }
                }
            }
            catch (parseError) {
                logger.warn({ stdout, parseError }, 'Failed to parse OPA output');
            }
            const violations = combinedResult.violations;
            const warnings = combinedResult.warnings;
            const suggestions = combinedResult.suggestions;
            const allow = Boolean(combinedResult.allow);
            logger.info({
                allow,
                violations: violations.length,
                warnings: warnings.length,
                suggestions: suggestions.length,
                violationDetails: violations.map(v => ({ rule: v.rule, message: v.message })),
            }, 'Rego policy evaluation completed');
            return {
                allow,
                violations,
                warnings,
                suggestions,
                summary: {
                    total_violations: violations.length,
                    total_warnings: warnings.length,
                    total_suggestions: suggestions.length,
                },
            };
        }
        finally {
            // Clean up temp file using tmp's cleanup mechanism
            try {
                tmpFile.removeCallback();
            }
            catch {
                // Ignore cleanup errors
            }
        }
    }
    catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        logger.error({ error: message }, 'Rego policy evaluation failed');
        // Return a safe default on error
        return {
            allow: false,
            violations: [
                {
                    rule: 'policy-evaluation-error',
                    category: 'system',
                    message: `Policy evaluation failed: ${message}`,
                    severity: 'block',
                },
            ],
            warnings: [],
            suggestions: [],
        };
    }
}
/**
 * Load and merge multiple Rego policy files
 *
 * OPA CLI automatically merges multiple policy files when passed with multiple -d flags.
 * All policies will be evaluated together and their results combined.
 *
 * @param policyPaths - Array of .rego policy file paths
 * @param logger - Logger instance
 * @returns Result containing RegoEvaluator or error
 */
export async function loadAndMergeRegoPolicies(policyPaths, logger) {
    if (policyPaths.length === 0) {
        return Failure('No policy paths provided');
    }
    // Validate all policy files exist
    for (const policyPath of policyPaths) {
        if (!existsSync(policyPath)) {
            return Failure(`Policy file not found: ${policyPath}`, {
                message: 'Rego policy file does not exist',
                hint: `Attempted to load: ${policyPath}`,
                resolution: 'Ensure all policy file paths are correct',
            });
        }
        if (!policyPath.endsWith('.rego')) {
            return Failure(`Invalid policy file format: ${policyPath}`, {
                message: 'All policy files must be .rego format',
                hint: `File: ${policyPath}`,
                resolution: 'Provide only .rego policy files',
            });
        }
    }
    logger.info({ policyCount: policyPaths.length, policies: policyPaths }, 'Loading and merging Rego policies');
    // Check if all policies are built-in (can use WASM bundle)
    const allBuiltIn = policyPaths.every(path => {
        const fileName = basename(path);
        return fileName in BUILT_IN_POLICY_MODULES;
    });
    // Fast path: Use pre-compiled WASM bundle for built-in policies (no OPA needed)
    // Try multiple locations for the WASM bundle (development, packaged, installed)
    let wasmPath;
    const wasmSearchPaths = [
        join(projectRoot, 'policies', 'compiled', 'policies.wasm'),
    ];
    // If policies are in the package, try relative to the first policy path
    if (policyPaths.length > 0 && policyPaths[0]) {
        wasmSearchPaths.push(join(dirname(policyPaths[0]), 'compiled', 'policies.wasm'));
    }
    for (const path of wasmSearchPaths) {
        if (existsSync(path)) {
            wasmPath = path;
            break;
        }
    }
    if (allBuiltIn && wasmPath) {
        try {
            logger.info({ wasmPath, policyCount: policyPaths.length }, 'Loading pre-compiled WASM bundle for built-in policies');
            const wasmBytes = await readFile(wasmPath);
            const wasmPolicy = await loadWasmPolicy(wasmBytes);
            // Create evaluator that queries all requested policy modules
            const evaluator = {
                policyPaths,
                evaluate: async (input) => {
                    // Wrap string input in content object (Dockerfile text, etc.)
                    const inputData = typeof input === 'string' ? { content: input } : input;
                    return evaluateAllWasmPolicies(wasmPolicy, inputData, policyPaths, logger);
                },
                evaluatePolicy: createEvaluatePolicyWrapper(async (input) => {
                    const inputData = typeof input === 'string' ? { content: input } : input;
                    return evaluateAllWasmPolicies(wasmPolicy, inputData, policyPaths, logger);
                }),
                queryConfig: async (packageName, input) => {
                    try {
                        const wasmResult = wasmPolicy.evaluate(input, packageName);
                        if (wasmResult?.[0]?.result) {
                            return wasmResult[0].result;
                        }
                        return null;
                    }
                    catch (error) {
                        logger.debug({ error, packageName }, 'WASM config query returned no result');
                        return null;
                    }
                },
                close: () => {
                    logger.debug({ policyPaths }, 'Cleaning up WASM policy resources');
                },
            };
            logger.info({ policyPaths }, 'Built-in policies loaded via WASM bundle (zero-dependency mode)');
            return Success(evaluator);
        }
        catch (wasmError) {
            logger.warn({
                error: wasmError,
                wasmPath,
                policyPaths,
                reason: 'WASM bundle loading failed or policy not in bundle',
            }, 'Failed to load WASM bundle, falling back to OPA binary');
            // Continue to OPA binary fallback
        }
    }
    // Fallback path: Use OPA binary for custom policies or if WASM failed
    const opaBinary = getOpaBinaryPath();
    try {
        await execFileAsync(opaBinary, ['version']);
    }
    catch {
        return Failure('OPA binary not found', {
            message: 'Open Policy Agent (OPA) CLI is not installed',
            hint: allBuiltIn
                ? 'WASM bundle failed to load and OPA binary is not available. Run: npm run build:policies'
                : 'OPA is required for custom policy evaluation',
            resolution: 'Install OPA: https://www.openpolicyagent.org/docs/latest/#running-opa',
        });
    }
    // Create evaluator that will evaluate all policies together
    const evaluator = {
        policyPaths,
        evaluate: async (input) => {
            return evaluateRegoPolicy(policyPaths, input, logger);
        },
        evaluatePolicy: createEvaluatePolicyWrapper(async (input) => {
            return evaluateRegoPolicy(policyPaths, input, logger);
        }),
        queryConfig: async (packageName, input) => {
            return queryRegoConfig(policyPaths, packageName, input, logger);
        },
        close: () => {
            logger.debug({ policyPaths }, 'Cleaning up merged Rego policy resources');
        },
    };
    logger.info({ policyPaths }, 'Rego policies loaded and ready for merged evaluation');
    return Success(evaluator);
}
/**
 * Simplified policy loading API for tests
 * Wrapper around loadAndMergeRegoPolicies for convenience
 *
 * @param options - Loading options
 * @param options.policiesPath - Directory containing policies
 * @param options.filePattern - Glob pattern for policy files (e.g., '*.rego', 'validation.rego')
 * @param logger - Optional logger instance
 * @returns Policy evaluator or failure
 */
export async function loadPolicies(options, logger) {
    const log = logger || (await import('../lib/logger.js')).createLogger({ name: 'policy-loader', level: 'info' });
    // Use glob to find matching policy files
    const { glob } = await import('glob');
    const pattern = options.filePattern || '*.rego';
    const fullPattern = join(options.policiesPath, pattern);
    const policyFiles = await glob(fullPattern);
    if (policyFiles.length === 0) {
        return Failure(`No policy files found matching pattern: ${fullPattern}`, {
            message: 'Policy files not found',
            hint: `No .rego files matched pattern: ${pattern}`,
            resolution: 'Check the policiesPath and filePattern parameters',
        });
    }
    log.debug({ policyFiles, pattern: fullPattern }, 'Found policy files');
    // Load and merge all found policies
    return loadAndMergeRegoPolicies(policyFiles, log);
}
