// codetocloudmcp/out/common/getAppLogs/appLogRetriever.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppLogRetriever = void 0;
exports.getAppConsoleLogs = getAppConsoleLogs;
const models_1 = require("./models");
class AppLogRetriever {
    azureClient;
    resourceGroupName;
    limit;
    logAnalyticsWorkspaceIds = [];
    constructor(options) {
        this.azureClient = options.azureClient;
        this.resourceGroupName = options.resourceGroupName;
        this.limit = options.limit || 200;
    }
    async initialize() {
        await this.validateResourceGroup();
        await this.discoverLogAnalyticsWorkspaceIds();
    }
    async validateResourceGroup() {
        const resourceGroups = await this.azureClient.getResourceGroupsInSubscription();
        const found = resourceGroups.find(rg => rg.name === this.resourceGroupName);
        if (!found) {
            throw new Error(`Resource group '${this.resourceGroupName}' not found in the subscription.`);
        }
    }
    async discoverLogAnalyticsWorkspaceIds() {
        if (this.logAnalyticsWorkspaceIds.length > 0) {
            return;
        }
        const filter = `resourceType eq 'Microsoft.OperationalInsights/workspaces'`;
        const resources = await this.azureClient.getResourcesInResourceGroupByFilter(this.resourceGroupName, filter);
        for (const resource of resources) {
            if (resource.id) {
                this.logAnalyticsWorkspaceIds.push(resource.id);
            }
        }
        if (this.logAnalyticsWorkspaceIds.length === 0) {
            throw new Error(`No log analytics workspaces found for resource group ${this.resourceGroupName}. Logs cannot be retrieved using this tool.`);
        }
    }
    getResourceGroupNameValue() {
        return this.resourceGroupName;
    }
    availableTablesQuery() {
        return `search * | summarize by $table`;
    }
    getConsoleLogTableName(resourceType) {
        switch (resourceType) {
            case models_1.SupportedResourceType.ContainerApp:
                return "ContainerAppConsoleLogs_CL";
            case models_1.SupportedResourceType.AppService:
                return "AppServiceConsoleLogs";
            case models_1.SupportedResourceType.FunctionApp:
                return "AppTraces";
        }
    }
    containerAppLogsQuery(containerAppName, limit) {
        return `${this.getConsoleLogTableName(models_1.SupportedResourceType.ContainerApp)} | where ContainerAppName_s == '${containerAppName}' | order by _timestamp_d desc | project TimeGenerated, Log_s | take ${limit}`;
    }
    appServiceLogsQuery(appServiceResourceId, limit) {
        return `${this.getConsoleLogTableName(models_1.SupportedResourceType.AppService)} | where _ResourceId == '${appServiceResourceId.toLocaleLowerCase()}' | order by TimeGenerated desc | project TimeGenerated, ResultDescription | take ${limit}`;
    }
    functionAppLogsQuery(functionAppName, limit) {
        return `${this.getConsoleLogTableName(models_1.SupportedResourceType.FunctionApp)} | where AppRoleName == '${functionAppName}' | order by TimeGenerated desc | project TimeGenerated, Message | take ${limit}`;
    }
    async getAppByName(resourceType, resourceName) {
        const [type, kind] = resourceType.split("|");
        const filter = `resourceType eq '${type}'`;
        const resources = await this.azureClient.getResourcesInResourceGroupByFilter(this.resourceGroupName, filter);
        const app = resources.find(r => r.name === resourceName &&
            (!kind || r.kind?.startsWith(kind)));
        if (!app) {
            throw new Error(`Resource '${resourceName}' of type '${resourceType}' not found in resource group '${this.resourceGroupName}'.`);
        }
        return app;
    }
    async queryAppLogsByName(resourceType, resourceName) {
        const app = await this.getAppByName(resourceType, resourceName);
        return this.queryAppLogsInternal(resourceType, app, resourceName);
    }
    async queryAppLogsInternal(resourceType, app, displayName) {
        // Check if logs have been ingested
        const tableCheckResult = await this.azureClient.queryLogsFromWorkspace(this.logAnalyticsWorkspaceIds[0], this.availableTablesQuery(), new Date(new Date().getTime() - 4 * 60 * 60 * 1000), new Date());
        const consoleLogTableName = this.getConsoleLogTableName(resourceType);
        const tableExists = tableCheckResult.status === "Success"
            && tableCheckResult.tables && tableCheckResult.tables.length > 0
            && tableCheckResult.tables[0].rows.map((row) => row[0]).includes(consoleLogTableName);
        if (!tableExists) {
            throw new Error(`Log table ${consoleLogTableName} not found in the log analytics workspace. Logs might not be ingested yet. Try again later.`);
        }
        const getLogErrors = [];
        let appLogs = "";
        let getLogSuccess = false;
        let logSearchQuery;
        let lastDeploymentTime;
        if (resourceType === models_1.SupportedResourceType.ContainerApp) {
            logSearchQuery = this.containerAppLogsQuery(app.name, this.limit);
            const lastRevision = await this.azureClient.getContainerAppsLatestActiveRevision(this.resourceGroupName, app.name);
            if (lastRevision) {
                lastDeploymentTime = lastRevision.createdTime;
            }
        }
        else if (resourceType === models_1.SupportedResourceType.AppService) {
            logSearchQuery = this.appServiceLogsQuery(app.id, this.limit);
            const lastDeployment = await this.azureClient.getWebSiteLatestDeployment(this.resourceGroupName, app.name);
            if (lastDeployment) {
                lastDeploymentTime = lastDeployment.startTime;
            }
        }
        else if (resourceType === models_1.SupportedResourceType.FunctionApp) {
            logSearchQuery = this.functionAppLogsQuery(app.name, this.limit);
            const lastDeployment = await this.azureClient.getWebSiteLatestDeployment(this.resourceGroupName, app.name);
            if (lastDeployment) {
                lastDeploymentTime = lastDeployment.startTime;
            }
        }
        else {
            throw new Error(`Unsupported resource type: ${resourceType}`);
        }
        // Use override times if provided, otherwise use deployment time or default to 4 hours ago
        const endTime = new Date();
        const startTime = lastDeploymentTime ?? new Date(endTime.getTime() - 4 * 60 * 60 * 1000);
        for (const logAnalyticsId of this.logAnalyticsWorkspaceIds) {
            try {
                const logSearchResult = await this.azureClient.queryLogsFromWorkspace(logAnalyticsId, logSearchQuery, startTime, endTime);
                if (logSearchResult.status === "Success") {
                    getLogSuccess = true;
                    if (!logSearchResult.tables || logSearchResult.tables.length === 0 || logSearchResult.tables[0].rows.length === 0) {
                        continue;
                    }
                    for (const table of logSearchResult.tables) {
                        for (const row of table.rows) {
                            appLogs += `[${row[0]}] ${row[1]}\n`;
                        }
                    }
                    break; // break if logs are found
                }
            }
            catch (error) {
                getLogErrors.push(`Error retrieving logs for ${app.name} from ${logAnalyticsId}: ${error.message}`);
            }
        }
        if (!getLogSuccess) {
            throw new Error(getLogErrors.join("\n"));
        }
        if (appLogs.length === 0) {
            return "No available logs found in the log analytics workspaces. The compute resources might not be configured to send logs.";
        }
        return `Console Logs for ${displayName} with resource ID ${app.id} from ${startTime.toISOString()} to ${endTime.toISOString()}:\n${appLogs}`;
    }
}
exports.AppLogRetriever = AppLogRetriever;
async function getAppConsoleLogs(options) {
    const toolErrorLogs = [];
    const consoleLogs = [];
    try {
        const retriever = new AppLogRetriever({
            azureClient: options.azureClient,
            resourceGroupName: options.resourceGroupName,
            limit: options.limit,
        });
        await retriever.initialize();
        for (const resource of options.resources) {
            try {
                consoleLogs.push(await retriever.queryAppLogsByName(resource.type, resource.name));
            }
            catch (error) {
                toolErrorLogs.push(`Error finding app logs for resource ${resource.name}: ${error.message}`);
            }
        }
    }
    catch (error) {
        toolErrorLogs.push(`${error.message}`);
    }
    let finalLogOutput = "";
    if (consoleLogs.length > 0) {
        finalLogOutput += `App Console Logs retrieved: \n ${consoleLogs.join('\n\n')}\n`;
    }
    if (toolErrorLogs.length > 0) {
        finalLogOutput += `Error during retrieval of app logs:  \n ${toolErrorLogs.join('\n')}`;
    }
    return finalLogOutput;
}
