// codetocloudmcp/out/common/getAppLogs/models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LLMSupportedResourceType = exports.SupportedResourceType = void 0;
exports.toSupportedResourceType = toSupportedResourceType;
exports.toLLMResourceForLogsArray = toLLMResourceForLogsArray;
var SupportedResourceType;
(function (SupportedResourceType) {
    SupportedResourceType["ContainerApp"] = "Microsoft.App/containerApps";
    SupportedResourceType["AppService"] = "Microsoft.Web/sites|app";
    SupportedResourceType["FunctionApp"] = "Microsoft.Web/sites|functionapp";
})(SupportedResourceType || (exports.SupportedResourceType = SupportedResourceType = {}));
/**
 * LLM-friendly enum with simple, easy-to-remember values.
 * Used in schemas exposed to LLMs, then converted to SupportedResourceType internally.
 */
var LLMSupportedResourceType;
(function (LLMSupportedResourceType) {
    LLMSupportedResourceType["ContainerApp"] = "containerapp";
    LLMSupportedResourceType["AppService"] = "appservice";
    LLMSupportedResourceType["FunctionApp"] = "functionapp";
})(LLMSupportedResourceType || (exports.LLMSupportedResourceType = LLMSupportedResourceType = {}));
const llmToSupportedResourceTypeMap = {
    [LLMSupportedResourceType.ContainerApp]: SupportedResourceType.ContainerApp,
    [LLMSupportedResourceType.AppService]: SupportedResourceType.AppService,
    [LLMSupportedResourceType.FunctionApp]: SupportedResourceType.FunctionApp,
};
function toSupportedResourceType(llmType) {
    const result = llmToSupportedResourceTypeMap[llmType];
    if (!result) {
        throw new Error(`Unsupported LLM resource type: '${llmType}'. Valid values: ${Object.values(LLMSupportedResourceType).join(', ')}`);
    }
    return result;
}
function toLLMResourceForLogsArray(llmResources) {
    return llmResources.map(r => ({
        name: r.name,
        type: toSupportedResourceType(r.type),
    }));
}
