// containerization-assist-mcp/dist-cjs/src/config/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.config = exports.applyPolicy = exports.clearPolicyCache = exports.loadAndMergePolicies = exports.loadPolicy = void 0;
exports.logConfigSummaryIfDev = logConfigSummaryIfDev;
/**
 * Consolidated Configuration - Main Config
 *
 * Single source of configuration replacing multiple separate config files.
 * Simple, focused configuration without complex validation overhead.
 */
const socket_validation_1 = require("../infra/docker/socket-validation");
const env_utils_1 = require("./env-utils");
// Export consolidated constants (includes environment schema and defaults)
__exportStar(require("./constants"), exports);
var policy_io_1 = require("./policy-io");
Object.defineProperty(exports, "loadPolicy", { enumerable: true, get: function () { return policy_io_1.loadPolicy; } });
Object.defineProperty(exports, "loadAndMergePolicies", { enumerable: true, get: function () { return policy_io_1.loadAndMergePolicies; } });
Object.defineProperty(exports, "clearPolicyCache", { enumerable: true, get: function () { return policy_io_1.clearPolicyCache; } });
var policy_eval_1 = require("./policy-eval");
Object.defineProperty(exports, "applyPolicy", { enumerable: true, get: function () { return policy_eval_1.applyPolicy; } });
exports.config = {
    server: {
        logLevel: (0, env_utils_1.parseStringEnv)('LOG_LEVEL', 'info'),
        port: (0, env_utils_1.parseIntEnv)('PORT', 3000),
    },
    workspace: {
        workspaceDir: (0, env_utils_1.parseStringEnv)('WORKSPACE_DIR', process.cwd()),
        maxFileSize: (0, env_utils_1.parseIntEnv)('MAX_FILE_SIZE', 10485760),
    },
    docker: {
        socketPath: (0, env_utils_1.parseStringEnv)('DOCKER_SOCKET', (0, socket_validation_1.autoDetectDockerSocket)()),
        timeout: (0, env_utils_1.parseIntEnv)('DOCKER_TIMEOUT', 60000),
    },
    toolLogging: {
        dirPath: (0, env_utils_1.parseStringEnv)('CONTAINERIZATION_ASSIST_TOOL_LOGS_DIR_PATH', ''),
        get enabled() {
            return this.dirPath.trim().length > 0;
        },
    },
};
/**
 * Configuration utilities
 */
function logConfigSummaryIfDev(logger) {
    if (process.env.NODE_ENV === 'development') {
        const configData = {
            server: {
                logLevel: exports.config.server.logLevel,
                port: exports.config.server.port,
            },
            workspace: exports.config.workspace.workspaceDir,
            docker: exports.config.docker.socketPath,
            toolLogging: {
                enabled: exports.config.toolLogging.enabled,
                dirPath: exports.config.toolLogging.dirPath || 'not configured',
            },
        };
        if (logger) {
            logger.info('Configuration loaded', configData);
        }
    }
}
