// containerization-assist-mcp/dist-cjs/src/config/policy-io.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadPolicy = loadPolicy;
exports.loadAndMergePolicies = loadAndMergePolicies;
exports.clearPolicyCache = clearPolicyCache;
/**
 * Policy IO - Rego Policy Loading
 *
 * This module provides loading and caching for Rego policies.
 * YAML policy support has been removed in favor of industry-standard OPA Rego.
 */
const path = __importStar(require("node:path"));
const types_1 = require("../types");
const errors_1 = require("../lib/errors");
const logger_1 = require("../lib/logger");
const policy_rego_1 = require("./policy-rego");
const log = (0, logger_1.createLogger)().child({ module: 'policy-io' });
// Simple cache without TTL - single-user CLI tool loads policies once
const regoCache = new Map();
/**
 * Load Rego policy from file
 *
 * @param file - Path to .rego policy file
 * @returns Promise<Result<RegoEvaluator>>
 */
async function loadPolicy(file) {
    try {
        // Validate .rego extension
        if (!file.endsWith('.rego')) {
            return (0, types_1.Failure)('Only .rego policy files are supported', {
                message: 'Invalid policy file format',
                hint: `File: ${file}`,
                resolution: 'Provide a .rego policy file. YAML policies are no longer supported.',
            });
        }
        // Check cache first
        const cached = regoCache.get(path.resolve(file));
        if (cached) {
            log.debug({ file }, 'Using cached Rego policy');
            return (0, types_1.Success)(cached);
        }
        // Load and compile Rego policy
        const result = await (0, policy_rego_1.loadRegoPolicy)(file, log);
        if (result.ok) {
            regoCache.set(path.resolve(file), result.value);
        }
        return result;
    }
    catch (err) {
        return (0, types_1.Failure)(errors_1.ERROR_MESSAGES.POLICY_LOAD_FAILED((0, errors_1.extractErrorMessage)(err)));
    }
}
/**
 * Load and merge multiple Rego policy files
 *
 * @param policyPaths - Array of .rego policy file paths
 * @returns Promise<Result<RegoEvaluator>>
 */
async function loadAndMergePolicies(policyPaths) {
    // Validate all paths are .rego files
    const nonRegoPaths = policyPaths.filter((p) => !p.endsWith('.rego'));
    if (nonRegoPaths.length > 0) {
        return (0, types_1.Failure)('All policy files must be .rego format', {
            message: 'Invalid policy file formats detected',
            hint: `Non-Rego files: ${nonRegoPaths.join(', ')}`,
            resolution: 'Convert YAML policies to Rego format. YAML is no longer supported.',
        });
    }
    return (0, policy_rego_1.loadAndMergeRegoPolicies)(policyPaths, log);
}
/**
 * Clear the policy cache
 * Useful for testing or when policies are updated
 */
function clearPolicyCache() {
    regoCache.clear();
    log.debug('Policy cache cleared');
}
