// containerization-assist-mcp/dist-cjs/src/config/policy-generation-config.js
"use strict";
/**
 * Policy Generation Configuration Types and Schemas
 *
 * Defines TypeScript interfaces and Zod schemas for policy-driven
 * generation configuration. These types are used by tools to query
 * policies for pre-generation defaults and organizational standards.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.configQueryInputSchema = exports.generationConfigSchema = exports.dynamicDefaultsSchema = exports.hpaConfigSchema = exports.healthCheckConfigSchema = exports.templateAdditionsSchema = exports.k8sTemplateComponentSchema = exports.dockerfileTemplateSnippetSchema = exports.k8sGenerationConfigSchema = exports.featureTogglesSchema = exports.orgStandardsSchema = exports.resourceDefaultsSchema = exports.dockerfileGenerationConfigSchema = void 0;
const zod_1 = require("zod");
// ===== DOCKERFILE GENERATION CONFIG =====
/**
 * Dockerfile generation configuration from policy
 */
exports.dockerfileGenerationConfigSchema = zod_1.z.object({
    /**
     * Build strategy override (multi-stage, single-stage, distroless)
     */
    buildStrategy: zod_1.z.enum(['multi-stage', 'single-stage', 'distroless']).optional(),
    /**
     * Base image category preference (official, distroless, alpine, minimal)
     */
    baseImageCategory: zod_1.z.enum(['official', 'distroless', 'alpine', 'minimal']).optional(),
    /**
     * Optimization priority (speed, size, security)
     */
    optimizationPriority: zod_1.z.enum(['speed', 'size', 'security', 'balanced']).optional(),
    /**
     * Security features to enable
     */
    securityFeatures: zod_1.z.object({
        nonRootUser: zod_1.z.boolean().optional(),
        readOnlyRootFS: zod_1.z.boolean().optional(),
        noNewPrivileges: zod_1.z.boolean().optional(),
        dropCapabilities: zod_1.z.boolean().optional(),
    }).optional(),
    /**
     * Build features to enable
     */
    buildFeatures: zod_1.z.object({
        buildCache: zod_1.z.boolean().optional(),
        layerOptimization: zod_1.z.boolean().optional(),
        healthcheck: zod_1.z.boolean().optional(),
    }).optional(),
}).strict();
// ===== KUBERNETES MANIFEST GENERATION CONFIG =====
/**
 * Resource defaults for different languages and environments
 */
exports.resourceDefaultsSchema = zod_1.z.object({
    /**
     * CPU request (e.g., "100m", "500m", "1")
     */
    cpuRequest: zod_1.z.string().optional(),
    /**
     * CPU limit (e.g., "500m", "1", "2")
     */
    cpuLimit: zod_1.z.string().optional(),
    /**
     * Memory request (e.g., "128Mi", "256Mi", "512Mi")
     */
    memoryRequest: zod_1.z.string().optional(),
    /**
     * Memory limit (e.g., "256Mi", "512Mi", "1Gi")
     */
    memoryLimit: zod_1.z.string().optional(),
}).strict();
/**
 * Organizational standards for Kubernetes resources
 */
exports.orgStandardsSchema = zod_1.z.object({
    /**
     * Required labels to apply to all resources
     */
    requiredLabels: zod_1.z.record(zod_1.z.string()).optional(),
    /**
     * Namespace naming pattern or override
     */
    namespace: zod_1.z.string().optional(),
    /**
     * Allowed container registries by environment
     */
    allowedRegistries: zod_1.z.array(zod_1.z.string()).optional(),
    /**
     * Service account name to use
     */
    serviceAccount: zod_1.z.string().optional(),
    /**
     * Image pull policy (Always, IfNotPresent, Never)
     */
    imagePullPolicy: zod_1.z.enum(['Always', 'IfNotPresent', 'Never']).optional(),
}).strict();
/**
 * Feature toggles for Kubernetes manifest generation
 */
exports.featureTogglesSchema = zod_1.z.object({
    /**
     * Enable health checks (liveness/readiness probes)
     */
    healthChecks: zod_1.z.boolean().optional(),
    /**
     * Enable horizontal pod autoscaling
     */
    autoscaling: zod_1.z.boolean().optional(),
    /**
     * Enable resource quotas
     */
    resourceQuotas: zod_1.z.boolean().optional(),
    /**
     * Enable network policies
     */
    networkPolicies: zod_1.z.boolean().optional(),
    /**
     * Enable pod security policies
     */
    podSecurityPolicies: zod_1.z.boolean().optional(),
    /**
     * Enable ingress
     */
    ingress: zod_1.z.boolean().optional(),
}).strict();
/**
 * Kubernetes manifest generation configuration from policy
 */
exports.k8sGenerationConfigSchema = zod_1.z.object({
    /**
     * Resource defaults based on language and environment
     */
    resourceDefaults: exports.resourceDefaultsSchema.optional(),
    /**
     * Organizational standards to apply
     */
    orgStandards: exports.orgStandardsSchema.optional(),
    /**
     * Feature toggles for manifest generation
     */
    features: exports.featureTogglesSchema.optional(),
    /**
     * Replica count default
     */
    replicas: zod_1.z.number().int().positive().optional(),
    /**
     * Deployment strategy (RollingUpdate, Recreate)
     */
    deploymentStrategy: zod_1.z.enum(['RollingUpdate', 'Recreate']).optional(),
}).strict();
// ===== TEMPLATE INJECTION TYPES =====
/**
 * Dockerfile template snippet
 * Represents a snippet of Dockerfile code to be injected into generated Dockerfiles
 */
exports.dockerfileTemplateSnippetSchema = zod_1.z.object({
    /**
     * Unique identifier for this template
     */
    id: zod_1.z.string(),
    /**
     * Section where this template should be injected
     * - setup: Early in the build process (after FROM, before main build steps)
     * - security: Security-related configurations (CA certs, user setup)
     * - observability: Observability agents and instrumentation
     * - runtime: Runtime configurations (before CMD/ENTRYPOINT)
     */
    section: zod_1.z.enum(['setup', 'security', 'observability', 'runtime']),
    /**
     * The Dockerfile code snippet to inject
     */
    content: zod_1.z.string(),
    /**
     * Human-readable description of what this template does
     */
    description: zod_1.z.string(),
    /**
     * Priority for ordering within the same section (higher = earlier)
     */
    priority: zod_1.z.number().int().optional(),
    /**
     * Conditions under which this template should be applied
     */
    conditions: zod_1.z.object({
        languages: zod_1.z.array(zod_1.z.string()).optional(),
        environments: zod_1.z.array(zod_1.z.string()).optional(),
        frameworks: zod_1.z.array(zod_1.z.string()).optional(),
    }).optional(),
}).strict();
/**
 * Kubernetes template component
 * Represents a component to be injected into Kubernetes manifests
 */
exports.k8sTemplateComponentSchema = zod_1.z.object({
    /**
     * Unique identifier for this template
     */
    id: zod_1.z.string(),
    /**
     * Component type
     */
    type: zod_1.z.enum(['sidecar', 'initContainer', 'volume', 'volumeMount', 'envVar', 'configMap', 'secret']),
    /**
     * Human-readable description of what this component does
     */
    description: zod_1.z.string(),
    /**
     * The component specification (varies by type)
     */
    spec: zod_1.z.record(zod_1.z.unknown()),
    /**
     * Priority for ordering (higher = higher priority)
     */
    priority: zod_1.z.number().int().optional(),
    /**
     * Conditions under which this template should be applied
     */
    conditions: zod_1.z.object({
        languages: zod_1.z.array(zod_1.z.string()).optional(),
        environments: zod_1.z.array(zod_1.z.string()).optional(),
        frameworks: zod_1.z.array(zod_1.z.string()).optional(),
    }).optional(),
}).strict();
/**
 * Template additions returned from policy
 */
exports.templateAdditionsSchema = zod_1.z.object({
    /**
     * Dockerfile template snippets
     */
    dockerfile: zod_1.z.array(exports.dockerfileTemplateSnippetSchema).optional(),
    /**
     * Kubernetes template components
     */
    kubernetes: zod_1.z.array(exports.k8sTemplateComponentSchema).optional(),
}).strict();
// ===== DYNAMIC DEFAULTS TYPES =====
/**
 * Health check configuration calculated from context
 */
exports.healthCheckConfigSchema = zod_1.z.object({
    /**
     * Initial delay before first check (seconds)
     */
    initialDelaySeconds: zod_1.z.number().int().positive(),
    /**
     * Interval between checks (seconds)
     */
    periodSeconds: zod_1.z.number().int().positive(),
    /**
     * Timeout for each check (seconds)
     */
    timeoutSeconds: zod_1.z.number().int().positive(),
    /**
     * Failure threshold before marking unhealthy
     */
    failureThreshold: zod_1.z.number().int().positive(),
    /**
     * Success threshold before marking healthy
     */
    successThreshold: zod_1.z.number().int().positive(),
}).strict();
/**
 * HPA (Horizontal Pod Autoscaler) configuration
 */
exports.hpaConfigSchema = zod_1.z.object({
    /**
     * Minimum number of replicas
     */
    minReplicas: zod_1.z.number().int().positive(),
    /**
     * Maximum number of replicas
     */
    maxReplicas: zod_1.z.number().int().positive(),
    /**
     * Target CPU utilization percentage
     */
    targetCPUUtilization: zod_1.z.number().int().min(1).max(100).optional(),
    /**
     * Target memory utilization percentage
     */
    targetMemoryUtilization: zod_1.z.number().int().min(1).max(100).optional(),
}).strict();
/**
 * Dynamic defaults calculated from context
 */
exports.dynamicDefaultsSchema = zod_1.z.object({
    /**
     * Replica count based on environment and criticality
     */
    replicas: zod_1.z.number().int().positive().optional(),
    /**
     * Health check configuration based on language and startup characteristics
     */
    healthChecks: exports.healthCheckConfigSchema.optional(),
    /**
     * HPA configuration based on workload characteristics
     */
    autoscaling: exports.hpaConfigSchema.optional(),
}).strict();
// ===== COMPLETE GENERATION CONFIG =====
/**
 * Complete generation configuration from policy
 * Combines Dockerfile and Kubernetes manifest configuration, templates, and dynamic defaults
 */
exports.generationConfigSchema = zod_1.z.object({
    /**
     * Dockerfile generation configuration
     */
    dockerfile: exports.dockerfileGenerationConfigSchema.optional(),
    /**
     * Kubernetes manifest generation configuration
     */
    kubernetes: exports.k8sGenerationConfigSchema.optional(),
    /**
     * Template additions to inject
     */
    templates: exports.templateAdditionsSchema.optional(),
    /**
     * Dynamic defaults calculated from context
     */
    dynamicDefaults: exports.dynamicDefaultsSchema.optional(),
}).strict();
// ===== QUERY INPUT TYPES =====
/**
 * Input for querying generation configuration
 */
exports.configQueryInputSchema = zod_1.z.object({
    /**
     * Detected language (node, python, go, java, etc.)
     */
    language: zod_1.z.string().optional(),
    /**
     * Detected framework (express, fastapi, gin, spring, etc.)
     */
    framework: zod_1.z.string().optional(),
    /**
     * Environment (dev, staging, prod)
     */
    environment: zod_1.z.string().optional(),
    /**
     * Application name
     */
    appName: zod_1.z.string().optional(),
    /**
     * Additional context
     */
    context: zod_1.z.record(zod_1.z.unknown()).optional(),
}).strict();
