// codetocloudmcp/out/toolsImplementations/getAppLogs/tool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetAppLogsTool = void 0;
const appLogRetriever_1 = require("../../common/getAppLogs/appLogRetriever");
const models_1 = require("../../common/getAppLogs/models");
const codeToCloudTool_1 = require("../shared/codeToCloudTool");
const toolType_1 = require("../shared/toolType");
class GetAppLogsTool extends codeToCloudTool_1.CodeToCloudTool {
    constructor(stateManager) {
        super(toolType_1.ToolType.GetAppLogs, stateManager);
    }
    async innerInvoke(params) {
        const logs = await (0, appLogRetriever_1.getAppConsoleLogs)({
            azureClient: await this.stateManager.getAzureClient(params.subscriptionId),
            resourceGroupName: params.resourceGroupName,
            resources: (0, models_1.toLLMResourceForLogsArray)(params.resources),
            limit: params.limit
        });
        return logs || "No logs found for the specified parameters.";
    }
}
exports.GetAppLogsTool = GetAppLogsTool;
