// codetocloudmcp/out/toolsImplementations/diagnoseExistingResources/keyVault/aca/acaKeyVaultDiagnosticHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AcaKeyVaultDiagnosticHandler = void 0;
const acaKeyVaultAuthChecker_1 = require("./acaKeyVaultAuthChecker");
const acaKeyVaultFirewallChecker_1 = require("./acaKeyVaultFirewallChecker");
const CONTAINER_APP_ID_REGEX = /^\/subscriptions\/([^\/]+)\/resourceGroups\/([^\/]+)\/providers\/Microsoft\.App\/containerApps\/([^\/]+)$/i;
class AcaKeyVaultDiagnosticHandler {
    authChecker;
    stateManager;
    constructor(stateManager) {
        this.stateManager = stateManager;
        this.authChecker = new acaKeyVaultAuthChecker_1.AcaKeyVaultAuthChecker(stateManager);
    }
    async diagnoseResource(containerAppId, keyVault) {
        const appMatch = containerAppId.match(CONTAINER_APP_ID_REGEX);
        if (!appMatch) {
            return {
                success: false,
                message: `Invalid containerAppId format: ${containerAppId}. Expected format: /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.App/containerApps/{containerAppName}`,
            };
        }
        const [, subscriptionId, resourceGroupName, containerAppName] = appMatch;
        let containerApp;
        try {
            const azureClient = await this.stateManager
                .getAzureClient(subscriptionId);
            containerApp = await azureClient.getContainerApp(resourceGroupName, containerAppName);
        }
        catch (error) {
            return {
                success: false,
                message: `Failed to access Container App ${containerAppId}: ${error}`,
            };
        }
        // Check all ACA identities for Key Vault access
        const authResult = await this.authChecker.checkAllIdentitiesKeyVaultAccess(containerApp, keyVault.id, subscriptionId, resourceGroupName, containerAppName, keyVault.name);
        if (!authResult.success) {
            return authResult;
        }
        // Check firewall configuration
        const firewallResult = await (0, acaKeyVaultFirewallChecker_1.checkAcaKeyVaultFirewall)(this.stateManager, containerApp, keyVault.id, subscriptionId, resourceGroupName, containerAppName, keyVault.name);
        if (!firewallResult.success) {
            return firewallResult;
        }
        return {
            success: true,
            message: [authResult.message, firewallResult.message]
                .filter(Boolean)
                .join(" "),
            recommendations: [
                ...(authResult.recommendations ?? []),
                ...(firewallResult.recommendations ?? []),
            ],
        };
    }
}
exports.AcaKeyVaultDiagnosticHandler = AcaKeyVaultDiagnosticHandler;
