// codetocloudmcp/out/toolsImplementations/diagnoseExistingResources/keyVault/aca/acaKeyVaultAuthChecker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AcaKeyVaultAuthChecker = void 0;
const KEY_VAULT_SECRETS_USER_ROLE_ID = "4633458b-17de-408a-b874-0445c86b69e6";
function parseKeyVaultId(keyVaultId) {
    const match = keyVaultId.match(/\/resourceGroups\/([^/]+)\/providers\/Microsoft\.KeyVault\/vaults\/([^/]+)/i);
    if (!match) {
        return null;
    }
    return { resourceGroupName: match[1], vaultName: match[2] };
}
class AcaKeyVaultAuthChecker {
    stateManager;
    constructor(stateManager) {
        this.stateManager = stateManager;
    }
    async checkAllIdentitiesKeyVaultAccess(containerApp, keyVaultId, subscriptionId, resourceGroupName, containerAppName, keyVaultName) {
        const identities = [];
        if (containerApp.identity?.principalId) {
            identities.push({
                principalId: containerApp.identity.principalId,
                label: "system-assigned identity",
            });
        }
        const userAssigned = containerApp.identity?.userAssignedIdentities;
        if (userAssigned) {
            for (const [resourceId, value] of Object.entries(userAssigned)) {
                if (value?.principalId) {
                    const identityName = resourceId.split("/").pop() ?? resourceId;
                    identities.push({
                        principalId: value.principalId,
                        label: `user-assigned identity ${identityName}`,
                        resourceId: resourceId,
                    });
                }
            }
        }
        if (identities.length === 0) {
            return {
                success: false,
                message: `Container App ${containerAppName} does not have any managed identities configured.`,
                recommendations: [
                    "Container Apps require managed identity to access Key Vault secrets.",
                    `Enable system-assigned identity: az containerapp identity assign -g ${resourceGroupName} -n ${containerAppName} --system-assigned`,
                    "Or create and assign a user-assigned identity for more flexibility.",
                ],
            };
        }
        const accessibleIdentities = [];
        const failedIdentities = [];
        for (const identity of identities) {
            const result = await this.checkIdentityKeyVaultAccess(identity.principalId, keyVaultId, subscriptionId, containerAppName, keyVaultName, identity.label);
            if (result.success) {
                accessibleIdentities.push(identity);
            }
            else {
                failedIdentities.push({
                    identity,
                    reason: result.message ?? "Unknown error",
                });
            }
        }
        if (accessibleIdentities.length > 0) {
            const accessLabels = accessibleIdentities.map((i) => i.label).join(", ");
            return {
                success: true,
                message: `Container App ${containerAppName} can access Key Vault ${keyVaultName} via: ${accessLabels}.`,
                recommendations: failedIdentities.length > 0
                    ? [
                        `Note: ${failedIdentities.length} identity/identities do not have access: ${failedIdentities.map((f) => f.identity.label).join(", ")}. Consider granting access if needed.`,
                    ]
                    : undefined,
            };
        }
        // No identity has access - provide fix recommendations
        const recommendations = [];
        const parsed = parseKeyVaultId(keyVaultId);
        let isRbacMode = true;
        if (parsed) {
            try {
                const azureClient = await this.stateManager
                    .getAzureClient(subscriptionId);
                const vault = await azureClient.getKeyVault(parsed.resourceGroupName, parsed.vaultName);
                isRbacMode = vault.properties?.enableRbacAuthorization === true;
            }
            catch {
                // Keep default assumption
            }
        }
        if (isRbacMode) {
            recommendations.push(`Key Vault ${keyVaultName} uses RBAC authorization. Grant "Key Vault Secrets User" role to an identity:`);
            for (const identity of identities) {
                recommendations.push(`  az role assignment create --assignee-object-id ${identity.principalId} --role "Key Vault Secrets User" --scope ${keyVaultId}`);
            }
        }
        else {
            recommendations.push(`Key Vault ${keyVaultName} uses Access Policy authorization. Grant secret permissions to an identity:`);
            for (const identity of identities) {
                recommendations.push(`  az keyvault set-policy --name ${keyVaultName} --object-id ${identity.principalId} --secret-permissions get list`);
            }
        }
        return {
            success: false,
            message: `Container App ${containerAppName} has ${identities.length} identity/identities, but none have access to Key Vault ${keyVaultName}.`,
            recommendations,
        };
    }
    async checkIdentityKeyVaultAccess(principalId, keyVaultId, subscriptionId, containerAppName, keyVaultName, identityLabel) {
        const parsed = parseKeyVaultId(keyVaultId);
        if (!parsed) {
            return {
                success: false,
                message: `Invalid Key Vault resource ID format: ${keyVaultId}`,
            };
        }
        let vault;
        try {
            const azureClient = await this.stateManager
                .getAzureClient(subscriptionId);
            vault = await azureClient.getKeyVault(parsed.resourceGroupName, parsed.vaultName);
        }
        catch {
            return {
                success: false,
                message: `Failed to retrieve Key Vault ${keyVaultName} configuration.`,
                recommendations: [
                    "Verify the Key Vault exists and you have permission to access it.",
                ],
            };
        }
        const isRbacMode = vault.properties?.enableRbacAuthorization === true;
        const identityLabelCapitalized = identityLabel.charAt(0).toUpperCase() + identityLabel.slice(1);
        if (isRbacMode) {
            // Key Vault uses RBAC authorization
            const hasPermission = await this.checkKeyVaultRbacPermission(keyVaultId, subscriptionId, principalId);
            if (!hasPermission) {
                return {
                    success: false,
                    message: `Key Vault ${keyVaultName} uses RBAC authorization mode. ${identityLabelCapitalized} does not have the required RBAC role.`,
                    recommendations: [
                        `Grant the "Key Vault Secrets User" role: az role assignment create --assignee-object-id ${principalId} --role "Key Vault Secrets User" --scope ${keyVaultId}`,
                    ],
                };
            }
            return {
                success: true,
                message: `Container App ${containerAppName} is correctly configured to access Key Vault ${keyVaultName} using ${identityLabel} (RBAC mode).`,
            };
        }
        else {
            // Key Vault uses Access Policy authorization
            const hasPermission = this.checkAccessPolicyPermission(vault, principalId);
            if (!hasPermission) {
                return {
                    success: false,
                    message: `Key Vault ${keyVaultName} uses Access Policy authorization mode. ${identityLabelCapitalized} does not have Secret Get permission in the access policies.`,
                    recommendations: [
                        `Grant access policy: az keyvault set-policy --name ${keyVaultName} --object-id ${principalId} --secret-permissions get list`,
                    ],
                };
            }
            return {
                success: true,
                message: `Container App ${containerAppName} is correctly configured to access Key Vault ${keyVaultName} using ${identityLabel} (Access Policy mode).`,
            };
        }
    }
    async checkKeyVaultRbacPermission(keyVaultId, subscriptionId, principalId) {
        try {
            const azureClient = await this.stateManager
                .getAzureClient(subscriptionId);
            const roles = await azureClient.listRoleAssignmentsForScope(keyVaultId, `principalId eq '${principalId}'`);
            for await (const role of roles) {
                if (role.roleDefinitionId
                    ?.toLowerCase()
                    .endsWith(`/roledefinitions/${KEY_VAULT_SECRETS_USER_ROLE_ID}`)) {
                    return true;
                }
            }
        }
        catch {
            // RBAC check failed
        }
        return false;
    }
    checkAccessPolicyPermission(vault, principalId) {
        const policies = vault.properties?.accessPolicies ?? [];
        const secretGet = "get";
        for (const policy of policies) {
            if (policy.objectId?.toLowerCase() === principalId.toLowerCase() &&
                Array.isArray(policy.permissions?.secrets) &&
                policy.permissions.secrets
                    .map((p) => p?.toLowerCase())
                    .includes(secretGet)) {
                return true;
            }
        }
        return false;
    }
}
exports.AcaKeyVaultAuthChecker = AcaKeyVaultAuthChecker;
