// containerization-assist-mcp/dist-cjs/src/tools/ops/tool.js
"use strict";
/**
 * Ops Tool - MCP Server Diagnostics
 *
 * Provides health monitoring and connectivity testing for the MCP server itself.
 * This tool operates on the server infrastructure, not on user applications.
 *
 * **Use Cases:**
 * - Health checks for monitoring systems
 * - Connectivity testing during troubleshooting
 * - Resource usage monitoring (memory, CPU)
 * - Server diagnostics and metadata
 *
 * **NOT for:**
 * - Application containerization (use build-image-context, etc.)
 * - Docker operations (use Docker tools)
 * - Kubernetes operations (use K8s tools)
 *
 * @packageDocumentation
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ping = ping;
exports.serverStatus = serverStatus;
const os = __importStar(require("os"));
const errors_1 = require("../../lib/errors");
const tool_context_helpers_1 = require("../../lib/tool-context-helpers");
const types_1 = require("../../types");
const summary_helpers_1 = require("../../lib/summary-helpers");
const types_2 = require("./types");
/**
 * Ping operation - test server connectivity
 * @public
 */
async function ping(config, context) {
    const { logger, timer } = (0, tool_context_helpers_1.setupToolContext)(context, 'ops-ping');
    try {
        const { message = 'ping' } = config;
        logger.info({ message }, 'Processing ping request');
        const timestamp = new Date().toISOString();
        const summary = `✅ Server is responsive. Ping successful at ${(0, summary_helpers_1.formatTimestamp)(timestamp)}.`;
        const result = {
            kind: 'ping',
            summary,
            success: true,
            message: `pong: ${message}`,
            timestamp,
            server: {
                name: 'containerization-assist-mcp',
                version: '2.0.0',
                uptime: process.uptime(),
                pid: process.pid,
            },
            capabilities: {
                tools: true,
                progress: true,
            },
        };
        timer.end();
        return (0, types_1.Success)(result);
    }
    catch (error) {
        timer.error(error);
        logger.error({ error }, 'Ping failed');
        return (0, types_1.Failure)((0, errors_1.extractErrorMessage)(error), {
            message: (0, errors_1.extractErrorMessage)(error),
            hint: 'An unexpected error occurred during the ping operation',
            resolution: 'Check the server logs for details. This is typically a server-side issue',
        });
    }
}
/**
 * Get server status
 * @public
 */
async function serverStatus(config, context) {
    const { logger, timer } = (0, tool_context_helpers_1.setupToolContext)(context, 'ops-server-status');
    try {
        const { details = false } = config;
        logger.info({ details }, 'Server status requested');
        const uptime = Math.floor(process.uptime());
        const version = '2.0.0';
        const totalMem = os.totalmem();
        const freeMem = os.freemem();
        const usedMem = totalMem - freeMem;
        const memPercentage = Math.round((usedMem / totalMem) * 100);
        const cpus = os.cpus();
        const loadAverage = os.loadavg();
        const migratedToolCount = 12;
        // Generate summary
        const uptimeStr = (0, summary_helpers_1.formatDuration)(uptime);
        const summary = `✅ Server healthy. Running for ${uptimeStr}. Memory: ${memPercentage}% used, CPU: ${cpus.length} cores.`;
        const status = {
            kind: 'status',
            summary,
            success: true,
            version,
            uptime,
            memory: {
                used: usedMem,
                total: totalMem,
                free: freeMem,
                percentage: memPercentage,
            },
            cpu: {
                model: cpus[0]?.model ?? 'unknown',
                cores: cpus.length,
                loadAverage,
            },
            system: {
                platform: os.platform(),
                release: os.release(),
                hostname: os.hostname(),
            },
            tools: {
                count: 14,
                migrated: migratedToolCount,
            },
        };
        logger.info({
            uptime,
            memoryUsed: usedMem,
            memoryPercentage: memPercentage,
            toolsMigrated: migratedToolCount,
        }, 'Server status compiled');
        timer.end();
        return (0, types_1.Success)(status);
    }
    catch (error) {
        timer.error(error);
        logger.error({ error }, 'Error collecting server status');
        return (0, types_1.Failure)((0, errors_1.extractErrorMessage)(error), {
            message: (0, errors_1.extractErrorMessage)(error),
            hint: 'An unexpected error occurred while collecting server status',
            resolution: 'Check the server logs for details. This is typically a server-side issue',
        });
    }
}
/**
 * Main ops implementation
 */
async function handleOps(input, context) {
    const { operation } = input;
    switch (operation) {
        case 'ping':
            return ping({ ...(input.message !== undefined && { message: input.message }) }, context);
        case 'status':
            return serverStatus({ ...(input.details !== undefined && { details: input.details }) }, context);
        default:
            return (0, types_1.Failure)(`Unknown operation: ${input.operation}`, {
                message: `Unknown operation: ${input.operation}`,
                hint: 'The requested operation is not supported',
                resolution: 'Use one of the supported operations: "ping" for connectivity testing or "status" for server information',
            });
    }
}
/**
 * Ops tool conforming to Tool interface
 */
const tool_1 = require("../../types/tool");
exports.default = (0, tool_1.tool)({
    ...types_2.opsToolDefinition,
    handler: handleOps,
});
