// codetocloudmcp/out/mcpServers/codeToCloudServer.js
#!/usr/bin/env node
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setupCodeToCloudMcpServer = setupCodeToCloudMcpServer;
const mcp_js_1 = require("@modelcontextprotocol/sdk/server/mcp.js");
const stdio_js_1 = require("@modelcontextprotocol/sdk/server/stdio.js");
const types_1 = require("containerization-assist-mcp/tools/types");
const models_1 = require("../common/models");
const copilotDefinitions_1 = require("../toolsImplementations/shared/copilotDefinitions");
const toolType_1 = require("../toolsImplementations/shared/toolType");
const containerAssistMCPServerTool_1 = require("./containerAssistMCPServerTool");
const mcpToolStateManager_1 = require("./mcpToolStateManager");
const mcpAzureGetRegionsForDeploymentTool_1 = require("./tools/mcpAzureGetRegionsForDeploymentTool");
const mcpAzurePricingTool_1 = require("./tools/mcpAzurePricingTool");
const mcpAzureQuotaCheckTool_1 = require("./tools/mcpAzureQuotaCheckTool");
const mcpAzureRegionCheckTool_1 = require("./tools/mcpAzureRegionCheckTool");
const mcpDebugInBrowserTool_1 = require("./tools/mcpDebugInBrowserTool");
const mcpDeployPipelineTool_1 = require("./tools/mcpDeployPipelineTool");
const mcpDiagnoseExistingResourcesTool_1 = require("./tools/mcpDiagnoseExistingResourcesTool");
const mcpGenerateArchitectureDiagramTool_1 = require("./tools/mcpGenerateArchitectureDiagramTool");
const mcpGetALZPlanTool_1 = require("./tools/mcpGetALZPlanTool");
const mcpGetAppLogsTool_1 = require("./tools/mcpGetAppLogsTool");
const mcpGetAzureResourceRulesTool_1 = require("./tools/mcpGetAzureResourceRulesTool");
const mcpGetContainerizationPlanTool_1 = require("./tools/mcpGetContainerizationPlanTool");
const mcpGetPlanTool_1 = require("./tools/mcpGetPlanTool");
const mcpGetWAFRulesTool_1 = require("./tools/mcpGetWAFRulesTool");
const mcpRunInTerminalTool_1 = require("./tools/mcpRunInTerminalTool");
const mcpSummarizeResultTool_1 = require("./tools/mcpSummarizeResultTool");
function setupCodeToCloudMcpServer(settings) {
    const customerChannel = settings.customerChannel ?? models_1.CustomerChannel.CODETOCLOUD;
    const version = settings.version ?? "0.0.1";
    const stateOptions = {
        disableTelemetry: settings.disableTelemetry ?? true,
        getCustomerChannel: settings.getCustomerChannel ?? (() => customerChannel),
        isVsCode: settings.isForVsCode ?? true,
        version: version
    };
    const stateManager = new mcpToolStateManager_1.McpToolStateManager(stateOptions);
    if (settings.useSubAgent === true) {
        stateManager.registerTool(toolType_1.ToolType.RunSubagent, copilotDefinitions_1.ToolNames.RunSubagent);
    }
    const useTestOnlyTools = settings.useTestOnlyTools ?? false;
    const serverName = settings.serverName ?? "azure-deploy-server";
    const defaultDeployWorkflowName = settings.defaultDeployWorkflowName ?? "codetocloud-workflow";
    const defaultContainerizationWorkflowName = settings.defaultContainerizationWorkflowName ?? "codetocloud-containerization-workflow";
    const defaultDeployToExistingWorkflowName = settings.defaultDeployToExistingWorkflowName ?? "codetocloud-deploy-to-existing-workflow";
    const defaultCICDPipelineSetupWorkflowName = "codetocloud-cicd-pipeline-setup-workflow";
    const defaultProvisionOnlyWorkflowName = "codetocloud-generate-iac-workflow";
    const server = settings.outerServer ?? new mcp_js_1.McpServer({
        name: serverName,
        version: version
    }, {
        capabilities: {
            logging: {},
        }
    });
    function registerContainerAssistTools(stateManager) {
        stateManager.registerTool(toolType_1.ToolType.AnalyzeRepository, copilotDefinitions_1.ToolNames.AnalyzeRepository);
        stateManager.registerTool(toolType_1.ToolType.PlanGenerateDockerfile, copilotDefinitions_1.ToolNames.PlanGenerateDockerfile);
        stateManager.registerTool(toolType_1.ToolType.GenerateKubernetesManifest, copilotDefinitions_1.ToolNames.GenerateKubernetesManifest);
        stateManager.registerTool(toolType_1.ToolType.BuildDockerImage, copilotDefinitions_1.ToolNames.BuildDockerImage);
        const tools = [
            new containerAssistMCPServerTool_1.ContainerAssistMCPServerTool(types_1.analyzeRepoToolDefinition.name, toolType_1.ToolType.AnalyzeRepository, types_1.analyzeRepoToolDefinition.schema.shape, types_1.analyzeRepoToolDefinition.description, stateManager),
            new containerAssistMCPServerTool_1.ContainerAssistMCPServerTool(types_1.generateDockerfileToolDefinition.name, toolType_1.ToolType.PlanGenerateDockerfile, types_1.generateDockerfileToolDefinition.schema.shape, types_1.generateDockerfileToolDefinition.description, stateManager),
            new containerAssistMCPServerTool_1.ContainerAssistMCPServerTool(types_1.generateK8sManifestsToolDefinition.name, toolType_1.ToolType.GenerateKubernetesManifest, types_1.generateK8sManifestsToolDefinition.schema._def.schema.shape, types_1.generateK8sManifestsToolDefinition.description, stateManager),
            new containerAssistMCPServerTool_1.ContainerAssistMCPServerTool(types_1.buildImageContextToolDefinition.name, toolType_1.ToolType.BuildDockerImage, types_1.buildImageContextToolDefinition.schema.shape, types_1.buildImageContextToolDefinition.description, stateManager),
        ];
        // Register each tool with telemetry using createToolHandler
        // This automatically handles ToolContext creation and formats output correctly
        for (const tool of tools) {
            server.registerTool(tool.getName(), {
                title: tool.getDescription(),
                description: tool.getDescription(),
                inputSchema: tool.getInputSchema(),
                annotations: { readOnlyHint: true }
            }, async (params, extra) => {
                return await tool.handler(params, extra);
            });
        }
        console.log(`✅ Registered ${tools.length} Container Assist tools with telemetry`);
    }
    ;
    registerContainerAssistTools(stateManager);
    // tools
    const azureQuotaCheckTool = new mcpAzureQuotaCheckTool_1.McpAzureQuotaCheckTool(stateManager);
    const azurePricingTool = new mcpAzurePricingTool_1.McpAzurePricingTool(stateManager);
    const azureRegionCheckTool = new mcpAzureRegionCheckTool_1.McpAzureRegionCheckTool(stateManager);
    const getRegionsForDeploymentTool = new mcpAzureGetRegionsForDeploymentTool_1.McpAzureGetRegionsForDeploymentTool(stateManager);
    const diagnoseExistingResourcesTool = new mcpDiagnoseExistingResourcesTool_1.McpDiagnoseExistingResourcesTool(stateManager);
    const deployPipelineTool = new mcpDeployPipelineTool_1.McpDeployPipelineTool(stateManager);
    const generateArchitectureDiagramTool = new mcpGenerateArchitectureDiagramTool_1.McpGenerateArchitectureDiagramTool(stateManager);
    const getAzdAppLogsTool = new mcpGetAppLogsTool_1.McpGetAppLogsTool(stateManager);
    const getAzureResourceRulesTool = new mcpGetAzureResourceRulesTool_1.McpGetAzureResourceRulesTool(stateManager);
    const getPlanTool = new mcpGetPlanTool_1.McpGetPlanTool(stateManager);
    const getContainerizationPlanTool = new mcpGetContainerizationPlanTool_1.McpGetContainerizationPlanTool(stateManager);
    const summarizeResultTool = new mcpSummarizeResultTool_1.McpSummarizeResultTool(stateManager);
    const runInTerminalTool = new mcpRunInTerminalTool_1.McpRunInTerminalTool(stateManager);
    const getALZPlanTool = new mcpGetALZPlanTool_1.McpGetALZPlanTool(stateManager);
    const getWAFRulesTool = new mcpGetWAFRulesTool_1.McpGetWAFRulesTool(stateManager);
    // registration
    registerTool(copilotDefinitions_1.ToolNames.AzureQuotaCheck, azureQuotaCheckTool, stateManager);
    registerTool(copilotDefinitions_1.ToolNames.AzurePricing, azurePricingTool, stateManager);
    registerTool(copilotDefinitions_1.ToolNames.AzureRegionCheck, azureRegionCheckTool, stateManager);
    registerTool(copilotDefinitions_1.ToolNames.GetRegionsForDeployment, getRegionsForDeploymentTool, stateManager);
    registerTool(copilotDefinitions_1.ToolNames.DeployPipeline, deployPipelineTool, stateManager);
    registerTool(copilotDefinitions_1.ToolNames.GenerateArchitectureDiagram, generateArchitectureDiagramTool, stateManager);
    registerTool(copilotDefinitions_1.ToolNames.GetAppLogs, getAzdAppLogsTool, stateManager);
    registerTool(copilotDefinitions_1.ToolNames.GetAzureResourceRules, getAzureResourceRulesTool, stateManager);
    registerTool(copilotDefinitions_1.ToolNames.GetPlan, getPlanTool, stateManager);
    registerTool(copilotDefinitions_1.ToolNames.GetContainerizationPlan, getContainerizationPlanTool, stateManager);
    registerTool(copilotDefinitions_1.ToolNames.SummarizeResult, summarizeResultTool, stateManager);
    registerTool(copilotDefinitions_1.ToolNames.DiagnoseExistingResources, diagnoseExistingResourcesTool, stateManager);
    registerTool(copilotDefinitions_1.ToolNames.GetALZPlan, getALZPlanTool, stateManager);
    registerTool(copilotDefinitions_1.ToolNames.GetWAFRules, getWAFRulesTool, stateManager);
    // test tool
    if (useTestOnlyTools) {
        const debugInBrowserTool = new mcpDebugInBrowserTool_1.McpDebugInBrowserTool(stateManager);
        registerTool(copilotDefinitions_1.ToolNames.DebugInBrowser, debugInBrowserTool, stateManager);
        registerTool(copilotDefinitions_1.ToolNames.RunInTerminal, runInTerminalTool, stateManager, false);
    }
    const provisionAndDeploy = "provision-and-deploy";
    const deployOnly = "deploy-only";
    let promptPrefix = [
        `Scan my project carefully to identify all Azure-relevant resources, programming languages, frameworks, dependencies, and configuration files needed for deployment.`,
        `Develop a `
    ].join(' ');
    let hostingServiceCheck = "<check with the user if they want to use: Azure App Service, Azure App Service Managed Instance, Azure Container Apps, Azure Functions, or Azure Kubernetes Service (AKS use Terraform by default).>";
    let promptSuffix = ` plan WITH TOOL #${stateManager.getToolName(toolType_1.ToolType.GetPlan)} and execute it to deploy my project to Azure. `;
    let deployToExistingSuffix = [
        "You must use current subscription id and resource group as per AZ CLI context for deployment. Ask for this from the user if not sure.",
        "Then use AzCli command to get resources information in the resource group.",
        "When deploying the application, use the existing resources and don't create new resources.",
        "If the project has any dependencies such as database that doesn't exist, ask user to provide more information."
    ].join(' ');
    const provisionAndDeployPrompt = `${promptPrefix}${provisionAndDeploy}${promptSuffix}deployTool: azcli, IacType: Bicep, hosting service: ${hostingServiceCheck}.`;
    const deployOnlyPrompt = `${promptPrefix}${deployOnly}${promptSuffix}${deployToExistingSuffix}`;
    const provisionOnlyPrompt = `${promptPrefix} plan WITH TOOL #${stateManager.getToolName(toolType_1.ToolType.GetPlan)} and follow it to generate IaC for my project. deployTool: azcli, deployOption: provision-only, IacType: Bicep, hosting service: ${hostingServiceCheck}.`;
    // prompt
    server.registerPrompt(defaultDeployWorkflowName, { description: "Generate a workflow for the App Modernization deployment process" }, async (_args) => {
        return {
            messages: [
                {
                    role: "user",
                    content: {
                        type: "text",
                        text: provisionAndDeployPrompt
                    }
                }
            ]
        };
    });
    server.registerPrompt(defaultContainerizationWorkflowName, { description: "Generate a workflow for the App Modernization containerization process" }, async (_args) => {
        return {
            messages: [
                {
                    role: "user",
                    content: {
                        type: "text",
                        text: `Scan my project and help me plan how to containerize my application using the #${stateManager.getToolName(toolType_1.ToolType.GetContainerizationPlan)} tool. Execute the plan. The end goal is to have Dockerfiles that are able to be built.`
                    }
                }
            ]
        };
    });
    server.registerPrompt(defaultDeployToExistingWorkflowName, { description: "Generate a workflow for deploying to existing Azure resources" }, async (_args) => {
        return {
            messages: [
                {
                    role: "user",
                    content: {
                        type: "text",
                        text: deployOnlyPrompt
                    }
                }
            ]
        };
    });
    server.registerPrompt(defaultProvisionOnlyWorkflowName, { description: "Create a workflow to generate IaC files and provisioning Azure resources" }, async (_args) => {
        return {
            messages: [
                {
                    role: "user",
                    content: {
                        type: "text",
                        text: provisionOnlyPrompt
                    }
                }
            ]
        };
    });
    const cicdPipelineSetupPrompt = `Help me set up a CI/CD pipeline with ${stateManager.getToolName(toolType_1.ToolType.DeployPipeline)} to deploy to Azure. Before tool calling, you should check with me about 1. I am using Github Actions or ADO pipeline or GitLab CI/CD 2. whether I have existing resources. You need to know about the subscription ID, resource groups and hosting service type for existing resources.`;
    server.registerPrompt(defaultCICDPipelineSetupWorkflowName, { description: "Set up a CI/CD pipeline for the project" }, async (_args) => {
        return {
            messages: [
                {
                    role: "user",
                    content: {
                        type: "text",
                        text: cicdPipelineSetupPrompt
                    }
                }
            ]
        };
    });
    let previousOnInitialized;
    if (server.server.oninitialized) {
        previousOnInitialized = server.server.oninitialized;
    }
    server.server.oninitialized = () => {
        if (previousOnInitialized) {
            previousOnInitialized();
        }
        const clientInfo = server.server.getClientVersion();
        settings.clientInfo = clientInfo;
        stateManager.setClientInfo(clientInfo);
        console.info(`MCP server initialized with client info: ${JSON.stringify(clientInfo)}`);
    };
    if (!settings.outerServer) {
        const transport = new stdio_js_1.StdioServerTransport();
        server
            .connect(transport)
            .then(async () => {
            console.info('MCP server connected successfully.');
        })
            .catch((error) => {
            console.error('Error occurred while connect to MCP server:', error);
        })
            .finally(() => {
            console.info('MCP server is running. Press Ctrl+C to exit.');
        });
    }
    function registerTool(toolName, tool, stateManager, isReadOnly = true) {
        const toolType = tool.getCodeToCloudTool().getToolType();
        stateManager.registerTool(toolType, toolName);
        if (isReadOnly) {
            server.registerTool(toolName, {
                title: tool.getDescription(),
                description: tool.getDescription(),
                inputSchema: tool.getInputSchema(),
                annotations: { readOnlyHint: true }
            }, async (args, extra) => {
                try {
                    const response = await tool.handler(args);
                    return formatToolResult(response);
                }
                catch (error) {
                    console.error('Error occurred while handling tool request:', error);
                    return formatToolResult(`MCP Server Tool failed: ${error}`, true);
                }
            });
        }
        else {
            server.registerTool(toolName, {
                title: tool.getDescription(),
                description: tool.getDescription(),
                inputSchema: tool.getInputSchema()
            }, async (args, extra) => {
                try {
                    const response = await tool.handler(args);
                    return formatToolResult(response);
                }
                catch (error) {
                    console.error('Error occurred while handling tool request:', error);
                    return formatToolResult(`MCP Server Tool failed: ${error}`, true);
                }
            });
        }
    }
    function formatToolResult(toolResult, isError) {
        return {
            isError: isError || false,
            content: [
                {
                    type: "text",
                    text: toolResult
                }
            ]
        };
    }
}
