// codetocloudmcp/out/mcpServers/codeToCloudMCPServerTool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodeToCloudMCPServerTool = void 0;
const telemetryUtil_1 = require("./common/telemetry/telemetryUtil");
class CodeToCloudMCPServerTool {
    schema;
    stateManager;
    constructor(schema, stateManager) {
        this.schema = schema;
        this.stateManager = stateManager;
    }
    getTelemetryClient() {
        return this.stateManager.getTelemetryClient();
    }
    getToolType() {
        return this.getCodeToCloudTool().getToolType();
    }
    getDescription() {
        return this.getCodeToCloudTool().getToolCopilotDescription();
    }
    getInputSchema() {
        return this.schema.shape;
    }
    // To convert Zod schema input to tool parameters, just served as a type conversion
    convertInputParameters(input) {
        return input;
    }
    updateTelemetryPropertiesWithInput(params, telemetryProperties) {
        telemetryProperties.responseLatency = Date.now() - (new Date(telemetryProperties.time)).getTime();
        telemetryProperties.resultCode = telemetryUtil_1.ResultCode.InProgress;
    }
    updateTelemetryPropertiesWithToolResponse(output, telemetryProperties) {
        telemetryProperties.responseLatency = Date.now() - (new Date(telemetryProperties.time)).getTime();
        telemetryProperties.resultCode = telemetryUtil_1.ResultCode.Success;
    }
    updateTelemetryPropertiesWithError(error, telemetryProperties) {
        telemetryProperties.responseLatency = Date.now() - (new Date(telemetryProperties.time)).getTime();
        telemetryProperties.error = `${error}`;
        telemetryProperties.resultCode = telemetryUtil_1.ResultCode.Failure;
    }
    async handler(request) {
        const telemetryClient = this.getTelemetryClient();
        var telemetryProperties = await telemetryClient.initTelemetryProperties(this.stateManager.getVersion(), this.getToolType(), this.stateManager.getOrCreateWorkspaceId(request.workspaceFolder ?? undefined));
        telemetryClient.sendEvent(this.getToolType(), telemetryProperties);
        try {
            var params = this.convertInputParameters(request);
            this.updateTelemetryPropertiesWithInput(params, telemetryProperties);
            telemetryClient.sendEvent(this.getToolType(), telemetryProperties);
            const response = await this.getCodeToCloudTool().invokeTool(params);
            this.updateTelemetryPropertiesWithToolResponse(response, telemetryProperties);
            telemetryClient.sendEvent(this.getToolType(), telemetryProperties);
            return response;
        }
        catch (error) {
            this.updateTelemetryPropertiesWithError(error, telemetryProperties);
            telemetryClient.sendEvent(this.getToolType(), telemetryProperties);
            throw new Error(`MCP Server Tool failed: ${error}`);
        }
    }
}
exports.CodeToCloudMCPServerTool = CodeToCloudMCPServerTool;
