// codetocloudmcp/out/mcpServers/containerAssistMCPServerTool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContainerAssistMCPServerTool = void 0;
const helpers_1 = require("../common/helpers");
const telemetryUtil_1 = require("../common/telemetry/telemetryUtil");
class ContainerAssistMCPServerTool {
    containerAssistToolName;
    overrideToolName;
    toolType;
    mcpToolDescription;
    schema;
    stateManager;
    constructor(containerAssistToolName, toolType, inputSchema, description, stateManager) {
        this.containerAssistToolName = containerAssistToolName;
        this.toolType = toolType;
        this.overrideToolName = stateManager.getToolName(toolType);
        this.schema = inputSchema;
        this.mcpToolDescription = description;
        this.stateManager = stateManager;
    }
    getContainerAssistToolName() {
        return this.containerAssistToolName;
    }
    getName() {
        return this.overrideToolName;
    }
    getDescription() {
        return this.mcpToolDescription;
    }
    getInputSchema() {
        return this.schema;
    }
    getToolType() {
        return this.toolType;
    }
    getTelemetryClient() {
        return this.stateManager.getTelemetryClient();
    }
    updateTelemetryPropertiesWithInput(params, telemetryProperties) {
        const obfuscatedParams = params ? (0, helpers_1.obfuscateContainerAssistParams)(params) : undefined;
        telemetryProperties.serializedInput = obfuscatedParams ? JSON.stringify(obfuscatedParams) : '';
        telemetryProperties.responseLatency = Date.now() - (new Date(telemetryProperties.time)).getTime();
        telemetryProperties.resultCode = telemetryUtil_1.ResultCode.InProgress;
    }
    updateTelemetryPropertiesWithToolResponse(output, telemetryProperties) {
        telemetryProperties.serializedOutput = "Tool call succeeded. Output obfuscated due to unknown customer data.";
        telemetryProperties.responseLatency = Date.now() - (new Date(telemetryProperties.time)).getTime();
        telemetryProperties.resultCode = telemetryUtil_1.ResultCode.Success;
    }
    updateTelemetryPropertiesWithError(error, telemetryProperties) {
        const errorMessage = error instanceof Error ? error.message : String(error);
        telemetryProperties.responseLatency = Date.now() - (new Date(telemetryProperties.time)).getTime();
        telemetryProperties.error = errorMessage;
        telemetryProperties.resultCode = telemetryUtil_1.ResultCode.Failure;
    }
    async handler(params, extra) {
        const telemetryClient = this.getTelemetryClient();
        var telemetryProperties = await telemetryClient.initTelemetryProperties(this.stateManager.getVersion(), this.getToolType(), this.stateManager.getOrCreateWorkspaceId(params.workspaceFolder ?? undefined));
        this.updateTelemetryPropertiesWithInput(params, telemetryProperties);
        telemetryClient.sendEvent(this.getToolType(), telemetryProperties);
        const telemetryOptions = {
            transport: 'app-modernization-mcp',
            // Note: The toolName parameter from createToolHandler will be the Container Assist
            // tool name ('analyze-repo'), but telemetryProperties.toolName contains the MCP
            // override name ('appmod-analyze-repository') for proper telemetry tracking
            onSuccess: (_result, _toolName, _params) => {
                const resultString = _result !== null ? JSON.stringify(_result) : '';
                this.updateTelemetryPropertiesWithToolResponse(resultString, telemetryProperties);
                telemetryClient.sendEvent(this.getToolType(), telemetryProperties);
            },
            // Track errors
            onError: (error, _toolName, _params) => {
                const errorMessage = error instanceof Error ? error.message : String(error);
                this.updateTelemetryPropertiesWithError(errorMessage, telemetryProperties);
                telemetryClient.sendEvent(this.getToolType(), telemetryProperties);
            },
        };
        return await this.stateManager.handleContainerAssistToolCall(params, extra, this.getContainerAssistToolName(), telemetryOptions);
    }
}
exports.ContainerAssistMCPServerTool = ContainerAssistMCPServerTool;
