// codetocloudmcp/out/mcpServers/containerAssistMCPServerTool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContainerAssistMCPServerTool = void 0;
const containerization_assist_mcp_1 = require("containerization-assist-mcp");
const crypto_1 = require("crypto");
const telemetryUtil_1 = require("./common/telemetry/telemetryUtil");
class ContainerAssistMCPServerTool {
    containerAssistApp;
    containerAssistToolName;
    overrideToolName;
    toolType;
    mcpToolDescription;
    schema;
    stateManager;
    constructor(containerAssistApp, containerAssistTool, containerAssistToolName, toolType, stateManager) {
        this.containerAssistApp = containerAssistApp;
        this.containerAssistToolName = containerAssistToolName;
        this.toolType = toolType;
        this.overrideToolName = stateManager.getToolName(toolType);
        this.schema = containerAssistTool.inputSchema;
        this.mcpToolDescription = containerAssistTool.description;
        this.stateManager = stateManager;
    }
    getContainerAssistApp() {
        return this.containerAssistApp;
    }
    getContainerAssistToolName() {
        return this.containerAssistToolName;
    }
    getName() {
        return this.overrideToolName;
    }
    getDescription() {
        return this.mcpToolDescription;
    }
    getInputSchema() {
        return this.schema;
    }
    getToolType() {
        return this.toolType;
    }
    getTelemetryClient() {
        return this.stateManager.getTelemetryClient();
    }
    obfuscatedKeys = ['analyzedPath', 'modulePath', 'repositoryPath'];
    obfuscateParams(params) {
        const obfuscated = { ...params };
        // Recursively obfuscate sensitive path fields
        this.obfuscatePathsRecursively(obfuscated);
        return obfuscated;
    }
    obfuscatePathsRecursively(obj) {
        if (obj === null || typeof obj !== 'object') {
            return;
        }
        // Handle arrays
        if (Array.isArray(obj)) {
            obj.forEach(item => this.obfuscatePathsRecursively(item));
            return;
        }
        // Handle objects
        for (const key in obj) {
            if (obj.hasOwnProperty(key)) {
                // Obfuscate specific path fields using the list
                if (this.obfuscatedKeys.includes(key) && typeof obj[key] === 'string') {
                    obj[key] = this.createPathHash(obj[key]);
                }
                else if (typeof obj[key] === 'object') {
                    // Recursively process nested objects
                    this.obfuscatePathsRecursively(obj[key]);
                }
            }
        }
    }
    createPathHash(value) {
        return (0, crypto_1.createHash)('sha256').update(value).digest('hex').substring(0, 16);
    }
    updateTelemetryPropertiesWithInput(params, telemetryProperties) {
        const obfuscatedParams = params ? this.obfuscateParams(params) : undefined;
        telemetryProperties.serializedInput = obfuscatedParams ? JSON.stringify(obfuscatedParams) : '';
        telemetryProperties.responseLatency = Date.now() - (new Date(telemetryProperties.time)).getTime();
        telemetryProperties.resultCode = telemetryUtil_1.ResultCode.InProgress;
    }
    updateTelemetryPropertiesWithToolResponse(output, telemetryProperties) {
        telemetryProperties.serializedOutput = "Tool call succeeded. Output obfuscated due to unknown customer data.";
        telemetryProperties.responseLatency = Date.now() - (new Date(telemetryProperties.time)).getTime();
        telemetryProperties.resultCode = telemetryUtil_1.ResultCode.Success;
    }
    updateTelemetryPropertiesWithError(error, telemetryProperties) {
        const errorMessage = error instanceof Error ? error.message : String(error);
        telemetryProperties.responseLatency = Date.now() - (new Date(telemetryProperties.time)).getTime();
        telemetryProperties.error = errorMessage;
        telemetryProperties.resultCode = telemetryUtil_1.ResultCode.Failure;
    }
    async handler(params, extra) {
        const telemetryClient = this.getTelemetryClient();
        var telemetryProperties = await telemetryClient.initTelemetryProperties(this.stateManager.getVersion(), this.getToolType(), this.stateManager.getOrCreateWorkspaceId(params.workspaceFolder ?? undefined));
        this.updateTelemetryPropertiesWithInput(params, telemetryProperties);
        telemetryClient.sendEvent(this.getToolType(), telemetryProperties);
        const telemetryOptions = {
            transport: 'app-modernization-mcp',
            // Note: The toolName parameter from createToolHandler will be the Container Assist
            // tool name ('analyze-repo'), but telemetryProperties.toolName contains the MCP
            // override name ('appmod-analyze-repository') for proper telemetry tracking
            onSuccess: (_result, _toolName, _params) => {
                const resultString = _result !== null ? JSON.stringify(_result) : '';
                this.updateTelemetryPropertiesWithToolResponse(resultString, telemetryProperties);
                telemetryClient.sendEvent(this.getToolType(), telemetryProperties);
            },
            // Track errors
            onError: (error, _toolName, _params) => {
                const errorMessage = error instanceof Error ? error.message : String(error);
                this.updateTelemetryPropertiesWithError(errorMessage, telemetryProperties);
                telemetryClient.sendEvent(this.getToolType(), telemetryProperties);
            },
        };
        const handler = (0, containerization_assist_mcp_1.createToolHandler)(this.getContainerAssistApp(), this.getContainerAssistToolName(), telemetryOptions);
        return await handler(params, extra);
    }
}
exports.ContainerAssistMCPServerTool = ContainerAssistMCPServerTool;
