// codetocloudmcp/out/mcpServers/mcpToolStateManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.McpToolStateManager = void 0;
const identity_1 = require("@azure/identity");
const uuid_1 = require("uuid");
const models_1 = require("./common/models");
const telemetryClient_1 = require("./common/telemetry/telemetryClient");
const azureClient_1 = require("./toolsImplementations/shared/azureClient");
class McpToolStateManager {
    isForVsCode;
    workspaceIdMap;
    registeredTools;
    getCustomerChannel;
    telemetryClient;
    version;
    constructor(options) {
        this.isForVsCode = options.isVsCode;
        this.workspaceIdMap = new Map();
        this.registeredTools = new Map();
        this.getCustomerChannel = options?.getCustomerChannel || (() => models_1.CustomerChannel.COPILOTCLI);
        this.telemetryClient = new telemetryClient_1.TelemetryClient(this.getCustomerChannel, options.disableTelemetry);
        this.version = options.version;
    }
    toolTypeExists(toolType) {
        return this.registeredTools.has(toolType);
    }
    getOrCreateWorkspaceId(workspaceFolder) {
        if (!workspaceFolder) {
            return "";
        }
        if (this.workspaceIdMap.has(workspaceFolder)) {
            return this.workspaceIdMap.get(workspaceFolder);
        }
        const newWorkspaceId = (0, uuid_1.v4)();
        this.workspaceIdMap.set(workspaceFolder, newWorkspaceId);
        return newWorkspaceId;
    }
    registerTool(toolType, toolName) {
        if (this.registeredTools.has(toolType)) {
            throw new Error(`ToolType ${toolType} is already registered with name ${this.registeredTools.get(toolType)}`);
        }
        this.registeredTools.set(toolType, toolName);
    }
    getToolName(toolType) {
        return this.registeredTools.get(toolType) ?? toolType;
    }
    getTelemetryClient() {
        return this.telemetryClient;
    }
    getIsForVsCode() {
        return this.isForVsCode;
    }
    getCopilotMarkdownExtension() {
        return this.getIsForVsCode() ? "copilotmd" : "copilot.md";
    }
    getAzureClient(subscriptionId) {
        return new azureClient_1.AzureClient(subscriptionId, new identity_1.DefaultAzureCredential());
    }
    getCustomerChannelValue() {
        return this.getCustomerChannel();
    }
    getVersion() {
        return this.version;
    }
}
exports.McpToolStateManager = McpToolStateManager;
