// codetocloudmcp/out/mcpServers/mcpToolStateManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.McpToolStateManager = void 0;
const abstractToolStateManager_1 = require("../common/abstractToolStateManager");
const models_1 = require("../common/models");
const telemetryClient_1 = require("../common/telemetry/telemetryClient");
class McpToolStateManager extends abstractToolStateManager_1.AbstractToolStateManager {
    isForVsCode;
    _containerAssistAppRuntimeConfig; // Lazy-loaded container assist app runtime config
    constructor(options) {
        const getCustomerChannel = options?.getCustomerChannel || (() => models_1.CustomerChannel.COPILOTCLI);
        const telemetryClient = options.telemetryClient ?? new telemetryClient_1.TelemetryClient(getCustomerChannel, options.disableTelemetry ?? true);
        super({
            telemetryClient,
            getCustomerChannel,
            version: options.version,
        });
        this.isForVsCode = options.isVsCode;
    }
    getIsForVsCode() {
        return this.isForVsCode;
    }
    getCopilotMarkdownExtension() {
        return this.getIsForVsCode() ? "copilotmd" : "copilot.md";
    }
    async getContainerAssistAppRuntime() {
        const { createApp } = await import("containerization-assist-mcp");
        const { OUTPUTFORMAT } = await import("containerization-assist-mcp/server");
        const { analyzeRepoTool, generateDockerfileTool, buildImageContextTool, generateK8sManifestsTool } = await import("containerization-assist-mcp/tools");
        if (!this._containerAssistAppRuntimeConfig) {
            this._containerAssistAppRuntimeConfig = createApp({
                // Register only the tools we need
                tools: [
                    analyzeRepoTool,
                    generateDockerfileTool,
                    buildImageContextTool,
                    generateK8sManifestsTool
                ],
                // Disable chain hints (automatic next-step suggestions)
                chainHintsMode: "disabled",
                // Use natural language format for rich, user-friendly output
                outputFormat: OUTPUTFORMAT.NATURAL_LANGUAGE
            });
        }
        return this._containerAssistAppRuntimeConfig;
    }
    async handleContainerAssistToolCall(params, extra, toolName, telemetryOptions) {
        const { createToolHandler } = await import("containerization-assist-mcp");
        const containerAssistRuntime = await this.getContainerAssistAppRuntime();
        const handler = createToolHandler(containerAssistRuntime, toolName, telemetryOptions);
        return await handler(params, extra);
    }
}
exports.McpToolStateManager = McpToolStateManager;
