// codetocloudmcp/out/toolsImplementations/getPlan/executionSteps/serviceStepProviders/AKSStepProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AKSStepProvider = void 0;
const models_1 = require("../../../../common/models");
const planStep_1 = require("../../../../common/planGeneration/planStep");
const project_1 = require("../../../../common/project");
const BaseStepProvider_1 = require("./BaseStepProvider");
class AKSStepProvider extends BaseStepProvider_1.BaseStepProvider {
    getServiceTypeName() {
        return project_1.ComputeResourceType.AZURE_KUBERNETES_SERVICE;
    }
    getCliName() {
        return "aks";
    }
    getRegionCheckPrompts() {
        return "You must check Microsoft.Compute/virtualMachines quota is available for AKS nodes.";
    }
    getProvisionPrompts(deployOption, deployTool) {
        if (deployTool === models_1.DeploymentTool.Azd) {
            return "Run `azd provision --no-prompt` to provision the resources and confirm the result.";
        }
        return super.getProvisionPrompts(deployOption, deployTool);
    }
    getDeploymentStep(deployOption, deployTool) {
        var aksSteps = new planStep_1.PlanStep("Azure Kubernetes Service Deployment", planStep_1.PlanStepBulletType.Numbered);
        var filesExistSteps = new planStep_1.PlanStep("If Kubernetes manifests/Helm charts exist", planStep_1.PlanStepBulletType.Bulleted);
        filesExistSteps.addSteps([
            `Existing file paths: <filePaths>`,
            `Check if the existing files match what we required in the plan and list the necessary changes, e.g. using Azure Container Registry for image`,
        ]);
        var filesNotExistSteps = new planStep_1.PlanStep(`If expected files do not exist`, planStep_1.PlanStepBulletType.Bulleted);
        filesNotExistSteps.addStep(`Check if the project prefer to use Kubernetes Manifests(default) or Helm charts, and create the files accordingly.`);
        aksSteps.addSteps([filesExistSteps, filesNotExistSteps]);
        aksSteps.addSteps([
            `Prepare the deployment script (build + push image to ACR, deploy to AKS with kubectl/helm).`,
            `Deploy with the files and verify the output. If errors occur, **fix the files until it works**`,
            `Output: Kubernetes YAML files/Helm Charts, deployment script`
        ]);
        return aksSteps;
    }
    getExistingResourcesSteps() {
        var existingResourceStep = new planStep_1.PlanStep(this.getServiceTypeName(), planStep_1.PlanStepBulletType.Bulleted);
        existingResourceStep.addStep(`AKS cluster name: <>, resource group: <>, subscription: <>, provisioningState: Succeeded, node count: 2, vmSize: Standard_D4ds_v5, osType: Linux. Check with 'az aks show'`);
        const dependencyChecking = new planStep_1.PlanStep("Check dependencies existence", planStep_1.PlanStepBulletType.Numbered);
        this.service?.dependencies.forEach(dep => {
            if ([models_1.AzureServiceNames.AzureAppService, models_1.AzureServiceNames.AzureFunctionApp, models_1.AzureServiceNames.AzureContainerApp, models_1.AzureServiceNames.AzureKubernetesService].includes(dep.serviceType)) {
                return;
            }
            let stepStr = `${dep.serviceType}: `;
            if ([models_1.AzureServiceNames.AzureMySQL, models_1.AzureServiceNames.AzurePostgresSQL, models_1.AzureServiceNames.AzureSQLDB, models_1.AzureServiceNames.AzureRedisCache].includes(dep.serviceType)) {
                stepStr += `server: <>, database: <>, connection string: <>. If not exists, ask if user want to 1) create a built-in resource in Kubernetes cluster (default). 2) have an Azure resource and provide its id, then get the resource information with Az CLI command`;
            }
            else {
                stepStr += `name: <>, resource group: <>.`;
            }
            dependencyChecking.addStep(stepStr);
        });
        if ((this.service?.dependencies) && this.service?.dependencies.length > 0) {
            existingResourceStep.addStep(dependencyChecking);
        }
        return [existingResourceStep];
    }
    getValidationStep(stateManager, deployTool) {
        var validationStep = new planStep_1.PlanStep("Deployment Validation", planStep_1.PlanStepBulletType.Numbered);
        validationStep.addStep("Check the AKS cluster is up and running, and the pods are ready.");
        validationStep.addStep("Check the deployed application is running.");
        return validationStep;
    }
}
exports.AKSStepProvider = AKSStepProvider;
