// codetocloudmcp/out/toolsImplementations/getPlan/executionSteps/serviceStepProviders/BaseStepProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseStepProvider = void 0;
const models_1 = require("../../../../common/models");
const planStep_1 = require("../../../../common/planGeneration/planStep");
const project_1 = require("../../../../common/project");
const toolType_1 = require("../../../shared/toolType");
class BaseStepProvider {
    targetServiceType;
    projectName;
    service;
    constructor(targetServiceName, projectName, service) {
        this.targetServiceType = targetServiceName;
        this.projectName = projectName;
        this.service = service;
    }
    getRegionCheckPrompts() {
        return '';
    }
    getProvisionPrompts(deployOption, deployTool) {
        if (deployTool === models_1.DeploymentTool.Azd) {
            if (deployOption === models_1.DeployOption.provisionOnly) {
                return "Run `azd provision --no-prompt` to provision the resources and confirm the result.";
            }
            return "Run `azd provision --preview --no-prompt` to dry run your infrastructure and confirm everything is well-formed.";
        }
        else {
            return "";
        }
    }
    getDeploymentStep(deployOption, deployTool) {
        if (deployTool === models_1.DeploymentTool.AzCLI) {
            var cliSteps = new planStep_1.PlanStep(this.getServiceTypeName() + " Deployment", planStep_1.PlanStepBulletType.Numbered);
            cliSteps.addStep(`Create deploy script (build + push image to ACR, deploy to ${this.getServiceTypeName()}).`);
            cliSteps.addStep(`Run the script and fix it until it works.`);
            cliSteps.addStep(`Output: Azure CLI scripts`);
            return cliSteps;
        }
        var azdDeploySteps = new planStep_1.PlanStep("Azd App Deployment", planStep_1.PlanStepBulletType.Numbered);
        if (deployOption === models_1.DeployOption.deployOnly) {
            azdDeploySteps.addStep("Check if IaC files exist. If not, suggest users to use Azure CLI for deployment. Otherwise, proceed with `azd deploy --no-prompt`.");
        }
        else {
            azdDeploySteps.addStep("Run `azd up --no-prompt`. Iterate on any errors, and retry.");
        }
        return azdDeploySteps;
    }
    getValidationStep(stateManager, deployTool) {
        var validationStep = new planStep_1.PlanStep("Deployment Validation", planStep_1.PlanStepBulletType.Numbered);
        validationStep.addStep(`Call tool \`${stateManager.getToolName(toolType_1.ToolType.GetAppLogs)}\` to check application logs and ensure the services are running.`);
        return validationStep;
    }
    getExistingResourcesSteps() {
        const resourceTypeName = this.getServiceTypeName();
        const cliName = this.getCliName();
        var existingResourceStep = new planStep_1.PlanStep(resourceTypeName + " for app " + this.service?.name, planStep_1.PlanStepBulletType.Bulleted);
        existingResourceStep.addSteps([
            `name: <>, resource group: <>, subscription: <>, provisioningState: Succeeded, runningStatus: Running${this.targetServiceType === models_1.TargetAppServiceType.AzureAppService ? ', kind: "app,linux,container"' : ''}. Check with 'az ${cliName} show -o json'`,
        ]);
        const depChecking = new planStep_1.PlanStep("Check dependencies existence", planStep_1.PlanStepBulletType.Numbered);
        this.service?.dependencies.forEach(dep => {
            let depStr = dep.serviceType + ": ";
            if ([models_1.AzureServiceNames.AzureAppService, models_1.AzureServiceNames.AzureFunctionApp, models_1.AzureServiceNames.AzureContainerApp, models_1.AzureServiceNames.AzureKubernetesService].includes(dep.serviceType)) {
                return;
            }
            if ([models_1.AzureServiceNames.AzureMySQL, models_1.AzureServiceNames.AzurePostgresSQL, models_1.AzureServiceNames.AzureSQLDB].includes(dep.serviceType) && (dep.connectionType === project_1.ConnectionType.Secret)) {
                depStr += "server: <>, database: <>, connection string: <>.";
            }
            else {
                depStr += `name: <>, resource group: <>.`;
            }
            depChecking.addStep(depStr);
        });
        if ((this.service?.dependencies) && this.service?.dependencies.length > 0) {
            existingResourceStep.addStep(depChecking);
        }
        return [existingResourceStep];
    }
}
exports.BaseStepProvider = BaseStepProvider;
