// @azure/arm-machinelearning/dist-esm/src/operations/features.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
import { __asyncDelegator, __asyncGenerator, __asyncValues, __await } from "tslib";
import { setContinuationToken } from "../pagingHelper";
import * as coreClient from "@azure/core-client";
import * as Mappers from "../models/mappers";
import * as Parameters from "../models/parameters";
/// <reference lib="esnext.asynciterable" />
/** Class containing Features operations. */
export class FeaturesImpl {
    /**
     * Initialize a new instance of the class Features class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * List Features.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param workspaceName Name of Azure Machine Learning workspace.
     * @param featuresetName Featureset name. This is case-sensitive.
     * @param featuresetVersion Featureset Version identifier. This is case-sensitive.
     * @param options The options parameters.
     */
    list(resourceGroupName, workspaceName, featuresetName, featuresetVersion, options) {
        const iter = this.listPagingAll(resourceGroupName, workspaceName, featuresetName, featuresetVersion, options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings === null || settings === void 0 ? void 0 : settings.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.listPagingPage(resourceGroupName, workspaceName, featuresetName, featuresetVersion, options, settings);
            },
        };
    }
    listPagingPage(resourceGroupName, workspaceName, featuresetName, featuresetVersion, options, settings) {
        return __asyncGenerator(this, arguments, function* listPagingPage_1() {
            let result;
            let continuationToken = settings === null || settings === void 0 ? void 0 : settings.continuationToken;
            if (!continuationToken) {
                result = yield __await(this._list(resourceGroupName, workspaceName, featuresetName, featuresetVersion, options));
                let page = result.value || [];
                continuationToken = result.nextLink;
                setContinuationToken(page, continuationToken);
                yield yield __await(page);
            }
            while (continuationToken) {
                result = yield __await(this._listNext(resourceGroupName, workspaceName, featuresetName, featuresetVersion, continuationToken, options));
                continuationToken = result.nextLink;
                let page = result.value || [];
                setContinuationToken(page, continuationToken);
                yield yield __await(page);
            }
        });
    }
    listPagingAll(resourceGroupName, workspaceName, featuresetName, featuresetVersion, options) {
        return __asyncGenerator(this, arguments, function* listPagingAll_1() {
            var _a, e_1, _b, _c;
            try {
                for (var _d = true, _e = __asyncValues(this.listPagingPage(resourceGroupName, workspaceName, featuresetName, featuresetVersion, options)), _f; _f = yield __await(_e.next()), _a = _f.done, !_a; _d = true) {
                    _c = _f.value;
                    _d = false;
                    const page = _c;
                    yield __await(yield* __asyncDelegator(__asyncValues(page)));
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (!_d && !_a && (_b = _e.return)) yield __await(_b.call(_e));
                }
                finally { if (e_1) throw e_1.error; }
            }
        });
    }
    /**
     * List Features.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param workspaceName Name of Azure Machine Learning workspace.
     * @param featuresetName Featureset name. This is case-sensitive.
     * @param featuresetVersion Featureset Version identifier. This is case-sensitive.
     * @param options The options parameters.
     */
    _list(resourceGroupName, workspaceName, featuresetName, featuresetVersion, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            workspaceName,
            featuresetName,
            featuresetVersion,
            options,
        }, listOperationSpec);
    }
    /**
     * Get feature.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param workspaceName Name of Azure Machine Learning workspace.
     * @param featuresetName Feature set name. This is case-sensitive.
     * @param featuresetVersion Feature set version identifier. This is case-sensitive.
     * @param featureName Feature Name. This is case-sensitive.
     * @param options The options parameters.
     */
    get(resourceGroupName, workspaceName, featuresetName, featuresetVersion, featureName, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            workspaceName,
            featuresetName,
            featuresetVersion,
            featureName,
            options,
        }, getOperationSpec);
    }
    /**
     * ListNext
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param workspaceName Name of Azure Machine Learning workspace.
     * @param featuresetName Featureset name. This is case-sensitive.
     * @param featuresetVersion Featureset Version identifier. This is case-sensitive.
     * @param nextLink The nextLink from the previous successful call to the List method.
     * @param options The options parameters.
     */
    _listNext(resourceGroupName, workspaceName, featuresetName, featuresetVersion, nextLink, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            workspaceName,
            featuresetName,
            featuresetVersion,
            nextLink,
            options,
        }, listNextOperationSpec);
    }
}
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const listOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.MachineLearningServices/workspaces/{workspaceName}/featuresets/{featuresetName}/versions/{featuresetVersion}/features",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.FeatureResourceArmPaginatedResult,
        },
        default: {
            bodyMapper: Mappers.ErrorResponse,
        },
    },
    queryParameters: [
        Parameters.apiVersion,
        Parameters.skip,
        Parameters.listViewType,
        Parameters.description,
        Parameters.tags1,
        Parameters.featureName,
        Parameters.pageSize1,
    ],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.workspaceName,
        Parameters.featuresetName,
        Parameters.featuresetVersion,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const getOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.MachineLearningServices/workspaces/{workspaceName}/featuresets/{featuresetName}/versions/{featuresetVersion}/features/{featureName}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.Feature,
        },
        default: {
            bodyMapper: Mappers.ErrorResponse,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.workspaceName,
        Parameters.featuresetName,
        Parameters.featuresetVersion,
        Parameters.featureName1,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listNextOperationSpec = {
    path: "{nextLink}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.FeatureResourceArmPaginatedResult,
        },
        default: {
            bodyMapper: Mappers.ErrorResponse,
        },
    },
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.workspaceName,
        Parameters.nextLink,
        Parameters.featuresetName,
        Parameters.featuresetVersion,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
