// @azure/arm-msi/dist/react-native/operations/federatedIdentityCredentials.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
import { setContinuationToken } from "../pagingHelper.js";
import * as coreClient from "@azure/core-client";
import * as Mappers from "../models/mappers.js";
import * as Parameters from "../models/parameters.js";
/// <reference lib="esnext.asynciterable" />
/** Class containing FederatedIdentityCredentials operations. */
export class FederatedIdentityCredentialsImpl {
    client;
    /**
     * Initialize a new instance of the class FederatedIdentityCredentials class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * Lists all the federated identity credentials under the specified user assigned identity.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param resourceName The name of the identity resource.
     * @param options The options parameters.
     */
    list(resourceGroupName, resourceName, options) {
        const iter = this.listPagingAll(resourceGroupName, resourceName, options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings?.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.listPagingPage(resourceGroupName, resourceName, options, settings);
            },
        };
    }
    async *listPagingPage(resourceGroupName, resourceName, options, settings) {
        let result;
        let continuationToken = settings?.continuationToken;
        if (!continuationToken) {
            result = await this._list(resourceGroupName, resourceName, options);
            let page = result.value || [];
            continuationToken = result.nextLink;
            setContinuationToken(page, continuationToken);
            yield page;
        }
        while (continuationToken) {
            result = await this._listNext(resourceGroupName, resourceName, continuationToken, options);
            continuationToken = result.nextLink;
            let page = result.value || [];
            setContinuationToken(page, continuationToken);
            yield page;
        }
    }
    async *listPagingAll(resourceGroupName, resourceName, options) {
        for await (const page of this.listPagingPage(resourceGroupName, resourceName, options)) {
            yield* page;
        }
    }
    /**
     * Lists all the federated identity credentials under the specified user assigned identity.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param resourceName The name of the identity resource.
     * @param options The options parameters.
     */
    _list(resourceGroupName, resourceName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, resourceName, options }, listOperationSpec);
    }
    /**
     * Create or update a federated identity credential under the specified user assigned identity.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param resourceName The name of the identity resource.
     * @param federatedIdentityCredentialResourceName The name of the federated identity credential
     *                                                resource.
     * @param parameters Parameters to create or update the federated identity credential.
     * @param options The options parameters.
     */
    createOrUpdate(resourceGroupName, resourceName, federatedIdentityCredentialResourceName, parameters, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            resourceName,
            federatedIdentityCredentialResourceName,
            parameters,
            options,
        }, createOrUpdateOperationSpec);
    }
    /**
     * Gets the federated identity credential.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param resourceName The name of the identity resource.
     * @param federatedIdentityCredentialResourceName The name of the federated identity credential
     *                                                resource.
     * @param options The options parameters.
     */
    get(resourceGroupName, resourceName, federatedIdentityCredentialResourceName, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            resourceName,
            federatedIdentityCredentialResourceName,
            options,
        }, getOperationSpec);
    }
    /**
     * Deletes the federated identity credential.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param resourceName The name of the identity resource.
     * @param federatedIdentityCredentialResourceName The name of the federated identity credential
     *                                                resource.
     * @param options The options parameters.
     */
    delete(resourceGroupName, resourceName, federatedIdentityCredentialResourceName, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            resourceName,
            federatedIdentityCredentialResourceName,
            options,
        }, deleteOperationSpec);
    }
    /**
     * ListNext
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param resourceName The name of the identity resource.
     * @param nextLink The nextLink from the previous successful call to the List method.
     * @param options The options parameters.
     */
    _listNext(resourceGroupName, resourceName, nextLink, options) {
        return this.client.sendOperationRequest({ resourceGroupName, resourceName, nextLink, options }, listNextOperationSpec);
    }
}
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const listOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ManagedIdentity/userAssignedIdentities/{resourceName}/federatedIdentityCredentials",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.FederatedIdentityCredentialsListResult,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [
        Parameters.apiVersion,
        Parameters.top,
        Parameters.skiptoken,
    ],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceName,
        Parameters.resourceGroupName1,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const createOrUpdateOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ManagedIdentity/userAssignedIdentities/{resourceName}/federatedIdentityCredentials/{federatedIdentityCredentialResourceName}",
    httpMethod: "PUT",
    responses: {
        200: {
            bodyMapper: Mappers.FederatedIdentityCredential,
        },
        201: {
            bodyMapper: Mappers.FederatedIdentityCredential,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    requestBody: Parameters.parameters2,
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceName,
        Parameters.resourceGroupName1,
        Parameters.federatedIdentityCredentialResourceName,
    ],
    headerParameters: [Parameters.accept, Parameters.contentType],
    mediaType: "json",
    serializer,
};
const getOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ManagedIdentity/userAssignedIdentities/{resourceName}/federatedIdentityCredentials/{federatedIdentityCredentialResourceName}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.FederatedIdentityCredential,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceName,
        Parameters.resourceGroupName1,
        Parameters.federatedIdentityCredentialResourceName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const deleteOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ManagedIdentity/userAssignedIdentities/{resourceName}/federatedIdentityCredentials/{federatedIdentityCredentialResourceName}",
    httpMethod: "DELETE",
    responses: {
        200: {},
        204: {},
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceName,
        Parameters.resourceGroupName1,
        Parameters.federatedIdentityCredentialResourceName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listNextOperationSpec = {
    path: "{nextLink}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.FederatedIdentityCredentialsListResult,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    urlParameters: [
        Parameters.$host,
        Parameters.nextLink,
        Parameters.subscriptionId,
        Parameters.resourceName,
        Parameters.resourceGroupName1,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
