// codetocloudmcp/out/common/abstractToolStateManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractToolStateManager = void 0;
const identity_1 = require("@azure/identity");
const uuid_1 = require("uuid");
class AbstractToolStateManager {
    workspaceIdMap;
    registeredTools;
    getCustomerChannelFunc;
    telemetryClient;
    version;
    clientInfo;
    _azureCredential; // Lazy-loaded credential
    constructor(options) {
        this.workspaceIdMap = new Map();
        this.registeredTools = new Map();
        this.getCustomerChannelFunc = options.getCustomerChannel;
        if (!options.telemetryClient) {
            throw new Error("telemetryClient is required to initialize AbstractToolStateManager");
        }
        this.telemetryClient = options.telemetryClient;
        this.version = options.version;
    }
    toolTypeExists(toolType) {
        return this.registeredTools.has(toolType);
    }
    getOrCreateWorkspaceId(workspaceFolder) {
        if (!workspaceFolder) {
            return "";
        }
        if (this.workspaceIdMap.has(workspaceFolder)) {
            return this.workspaceIdMap.get(workspaceFolder);
        }
        const newWorkspaceId = (0, uuid_1.v4)();
        this.workspaceIdMap.set(workspaceFolder, newWorkspaceId);
        return newWorkspaceId;
    }
    registerTool(toolType, toolName) {
        if (this.registeredTools.has(toolType)) {
            throw new Error(`ToolType ${toolType} is already registered with name ${this.registeredTools.get(toolType)}`);
        }
        this.registeredTools.set(toolType, toolName);
    }
    getToolName(toolType) {
        return this.registeredTools.get(toolType) ?? toolType;
    }
    getTelemetryClient() {
        return this.telemetryClient;
    }
    getCopilotMarkdownExtension() {
        return this.getIsForVsCode() ? "copilotmd" : "copilot.md";
    }
    async getAzureClient(subscriptionId) {
        // Lazy-load DefaultAzureCredential only when actually needed
        if (!this._azureCredential) {
            this._azureCredential = new identity_1.DefaultAzureCredential();
        }
        // Dynamically import AzureClient to defer all Azure SDK imports until first use
        const { AzureClient } = await import("../toolsImplementations/shared/azureClient.js");
        return new AzureClient(subscriptionId, this._azureCredential);
    }
    getCustomerChannelValue() {
        return this.getCustomerChannelFunc();
    }
    getVersion() {
        return this.version;
    }
    getClientInfo() {
        return this.clientInfo;
    }
    setClientInfo(clientInfo) {
        this.clientInfo = clientInfo;
    }
}
exports.AbstractToolStateManager = AbstractToolStateManager;
