// codetocloudmcp/out/common/commands.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.runRecommendFirst = exports.followUpPrompts = exports.updateCommandName = exports.recommendScratchCommandName = exports.recommendProjectCommandName = exports.generateCommandName = void 0;
exports.generateCommandName = "generate";
exports.recommendProjectCommandName = "recommend(from project)";
exports.recommendScratchCommandName = "recommend(from scratch)";
exports.updateCommandName = "update";
// The sample prompt for the codetocloud slash commands.
exports.followUpPrompts = new Map([
    [exports.recommendProjectCommandName, ["Recommend Azure services based on my project."]],
    [exports.recommendScratchCommandName, ["Recommend Azure services for an AI based chat app."]],
    [exports.updateCommandName, ["Replace container app with app service.", "Add an environment variable KEY=val to my project.", "Update the environment variable KEY to \"val\"."]],
    [exports.generateCommandName, ["Generate bicep files based on the recommendations."]],
]);
exports.runRecommendFirst = `To initiate the code-to-cloud process, start by recommending Azure services based on a project scan or an infrastructure description. For example, you might try:\n
- ${exports.followUpPrompts.get(exports.recommendProjectCommandName)?.at(0) || ""} \n
- ${exports.followUpPrompts.get(exports.recommendScratchCommandName)?.at(0) || ""} \n`;
