// codetocloudmcp/out/common/helpers.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.dumpProjectScanResultImpl = dumpProjectScanResultImpl;
exports.findFileIteratively = findFileIteratively;
exports.detectPackageJson = detectPackageJson;
exports.obfuscateContainerAssistParams = obfuscateContainerAssistParams;
exports.normalizeToolNameForVsCode = normalizeToolNameForVsCode;
const crypto_1 = require("crypto");
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const constants_1 = require("./constants");
function dumpProjectScanResultImpl(projectScanResult, workspaceFolder) {
    if (!workspaceFolder) {
        throw new Error("No workspace folder is opened.");
    }
    const filePath = path.join(workspaceFolder, constants_1.projectScanResultRelativePath);
    const dirPath = path.dirname(filePath);
    // Ensure the directory exists
    if (!fs.existsSync(dirPath)) {
        fs.mkdirSync(dirPath, { recursive: true });
    }
    fs.writeFileSync(filePath, JSON.stringify(projectScanResult.toJson(), null, 2));
}
function findFileIteratively(currentPath, fileName) {
    const normalizedPath = path.resolve(currentPath);
    const parsed = path.parse(normalizedPath);
    // Stop at filesystem root
    if (normalizedPath === parsed.root) {
        return undefined;
    }
    const filePath = path.join(normalizedPath, fileName);
    if (fs.existsSync(filePath)) {
        return filePath;
    }
    // Only recurse up one level
    return findFileIteratively(parsed.dir, fileName);
}
function detectPackageJson(currentPath) {
    const packageJsonFile = findFileIteratively(currentPath, 'package.json');
    if (!packageJsonFile) {
        return undefined;
    }
    return JSON.parse(fs.readFileSync(packageJsonFile, 'utf8'));
}
const obfuscatedKeys = ['analyzedPath', 'modulePath', 'repositoryPath', 'path', 'dockerfilePath',];
function obfuscateContainerAssistParams(params) {
    const obfuscated = { ...params };
    // Recursively obfuscate sensitive path fields
    obfuscatePathsRecursively(obfuscated);
    return obfuscated;
}
function obfuscatePathsRecursively(obj) {
    if (obj === null || typeof obj !== 'object') {
        return;
    }
    // Handle arrays
    if (Array.isArray(obj)) {
        obj.forEach(item => obfuscatePathsRecursively(item));
        return;
    }
    // Handle objects
    for (const key in obj) {
        if (obj.hasOwnProperty(key)) {
            // Obfuscate specific path fields using the list
            if (obfuscatedKeys.includes(key) && typeof obj[key] === 'string') {
                obj[key] = createPathHash(obj[key]);
            }
            else if (typeof obj[key] === 'object') {
                // Recursively process nested objects
                obfuscatePathsRecursively(obj[key]);
            }
        }
    }
}
function createPathHash(value) {
    return (0, crypto_1.createHash)('sha256').update(value).digest('hex').substring(0, 16);
}
function normalizeToolNameForVsCode(toolName) {
    // convert everything to lowercase, replace spaces with underscores, and hyphens with underscores
    return toolName.toLowerCase().replace(/[\s-]+/g, '_');
}
