// codetocloudmcp/out/common/project.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.TargetEnvDefaultKeys = exports.EnvType = exports.ProjectScanResult = exports.SettingEnv = exports.Setting = exports.Service = exports.Docker = exports.Binding = exports.ResourceInstance = exports.Resource = exports.BindingResourceType = exports.ComputeResourceType = exports.ConnectionType = exports.ProjectType = void 0;
exports.flattenResourceType = flattenResourceType;
exports.standardizeResourceType = standardizeResourceType;
exports.getResourceReference = getResourceReference;
exports.standardizeConnectionType = standardizeConnectionType;
var ProjectType;
(function (ProjectType) {
    ProjectType["Azd"] = "azd";
    ProjectType["Raw"] = "raw";
})(ProjectType || (exports.ProjectType = ProjectType = {}));
var ConnectionType;
(function (ConnectionType) {
    ConnectionType["Secret"] = "secret";
    ConnectionType["SystemIdentity"] = "system-identity";
    ConnectionType["UserIdentity"] = "user-identity";
    ConnectionType["Http"] = "http";
    ConnectionType["BotConnection"] = "bot-connection";
})(ConnectionType || (exports.ConnectionType = ConnectionType = {}));
var ComputeResourceType;
(function (ComputeResourceType) {
    ComputeResourceType["AZURE_APP_SERVICE"] = "Azure App Service";
    ComputeResourceType["AZURE_STATIC_WEB_APPS"] = "Azure Static Web Apps";
    ComputeResourceType["AZURE_CONTAINER_APP"] = "Azure Container App";
    ComputeResourceType["AZURE_FUNCTIONS"] = "Azure Functions";
    ComputeResourceType["AZURE_KUBERNETES_SERVICE"] = "Azure Kubernetes Service";
    ComputeResourceType["AZURE_APP_SERVICE_MANAGED_INSTANCE"] = "Azure App Service Managed Instance";
})(ComputeResourceType || (exports.ComputeResourceType = ComputeResourceType = {}));
var BindingResourceType;
(function (BindingResourceType) {
    BindingResourceType["AZURE_AI_SEARCH"] = "Azure AI Search";
    BindingResourceType["AZURE_AI_SERVICES"] = "Azure AI Services";
    BindingResourceType["AZURE_APPLICATION_INSIGHTS"] = "Azure Application Insights";
    BindingResourceType["AZURE_BOT_SERVICE"] = "Azure Bot Service";
    BindingResourceType["AZURE_CACHE_FOR_REDIS"] = "Azure Cache for Redis";
    BindingResourceType["AZURE_COSMOS_DB"] = "Azure Cosmos DB";
    BindingResourceType["AZURE_KEY_VAULT"] = "Azure Key Vault";
    BindingResourceType["AZURE_DATABASE_FOR_MYSQL"] = "Azure Database for MySQL";
    BindingResourceType["AZURE_DATABASE_FOR_POSTGRESQL"] = "Azure Database for PostgreSQL";
    BindingResourceType["AZURE_OPEN_AI"] = "Azure Open AI";
    BindingResourceType["AZURE_SQL_DATABASE"] = "Azure SQL Database";
    BindingResourceType["AZURE_STORAGE_ACCOUNT"] = "Azure Storage Account";
    BindingResourceType["AZURE_SERVICE_BUS"] = "Azure Service Bus";
    BindingResourceType["AZURE_WEB_PUBSUB"] = "Azure Web PubSub";
    BindingResourceType["AZURE_SIGNALR_SERVICE"] = "Azure SignalR Service";
})(BindingResourceType || (exports.BindingResourceType = BindingResourceType = {}));
// Flatten the resource type
function flattenResourceType(resourceType) {
    return resourceType.replace(/\s/g, '').toLowerCase();
}
// Standardize the resource type
function standardizeResourceType(resourceType) {
    const validResourceTypes = [...Object.values(ComputeResourceType), ...Object.values(BindingResourceType)];
    for (const type of validResourceTypes) {
        if (flattenResourceType(type) === resourceType) {
            return type;
        }
    }
    return undefined;
}
// Get resource reference
function getResourceReference(resourceType, resourceName) {
    return '$' + `{${flattenResourceType(resourceType)}.${resourceName}}`;
}
// Standardize the connection type
function standardizeConnectionType(connectionType) {
    for (const type of Object.values(ConnectionType)) {
        if (type === connectionType) {
            return type;
        }
    }
    return undefined;
}
// Define the Resource class
class Resource {
    type;
    instances;
    constructor(json) {
        if (!json.type) {
            throw new Error("Type is required to construct a Resource");
        }
        this.type = standardizeResourceType(json.type);
        this.instances = [];
        if (json.instances && json.instances.length > 0) {
            for (const instance of json.instances) {
                this.addInstance(instance.name);
            }
        }
        else {
            // add a default instance if not provided
            this.addInstance();
        }
    }
    toJson() {
        return {
            type: flattenResourceType(this.type),
            instances: this.instances.map(instance => instance.toJson())
        };
    }
    getType() {
        return standardizeResourceType(this.type);
    }
    // Add a new instance to the resource
    addInstance(name) {
        const length = this.instances.length;
        // generate a default name if not provided: <resourceType>-<instanceNumber>
        const instanceName = name || `${this.type.split(' ').pop()?.toLowerCase() || ''}${length}`;
        this.instances.push(new ResourceInstance({ name: instanceName }));
    }
}
exports.Resource = Resource;
// Define the ResourceInstance class
class ResourceInstance {
    name;
    constructor(json) {
        if (!json.name) {
            throw new Error("Name is required to construct a ResourceInstance");
        }
        this.name = json.name;
    }
    toJson() {
        return {
            name: this.name
        };
    }
}
exports.ResourceInstance = ResourceInstance;
// Define the Binding class
class Binding {
    source;
    target;
    connection;
    store;
    customKeys;
    constructor(json) {
        if (!json.source) {
            throw new Error("Source resource is required to construct a Binding");
        }
        this.source = json.source;
        this.target = json.target;
        this.connection = json.connection;
        this.store = json.store || '';
        this.customKeys = { ...json.customKeys };
    }
    toJson() {
        return {
            source: this.source,
            target: this.target,
            connection: this.connection,
            store: this.store,
            customKeys: this.customKeys
        };
    }
    // get binging source type
    getSourceType() {
        const source = this.source.slice(2, -1);
        return standardizeResourceType(source.split('.')[0]);
    }
    // get binding source name
    getSourceName() {
        const source = this.source.slice(2, -1);
        return source.split('.')[1];
    }
    // get binging target type
    getTargetType() {
        const target = this.target.slice(2, -1);
        return standardizeResourceType(target.split('.')[0]);
    }
    // get binding target name
    getTargetName() {
        const target = this.target.slice(2, -1);
        return target.split('.')[1];
    }
    getConnectionType() {
        return this.connection ?
            standardizeConnectionType(this.connection) : '';
    }
}
exports.Binding = Binding;
// Define the Docker class
class Docker {
    path;
    context;
    constructor(json) {
        this.path = json.path;
        this.context = json.context;
    }
    toJson() {
        return {
            path: this.path,
            context: this.context
        };
    }
}
exports.Docker = Docker;
// Define the Service class
class Service {
    host;
    language;
    project;
    port;
    docker;
    constructor(json) {
        if (!json.language || !json.project) {
            throw new Error("Language, and project are required");
        }
        this.host = json.host;
        this.language = json.language;
        this.project = json.project.endsWith('/') ?
            json.project : json.project + '/';
        this.port = json.port || '80';
        this.docker = json.docker ? new Docker(json.docker) : undefined;
    }
    toJson() {
        return {
            host: this.host,
            language: this.language,
            project: this.project,
            port: this.port,
            docker: this.docker ? this.docker.toJson() : undefined
        };
    }
    setHost(resourceType, resourceName) {
        this.host = getResourceReference(resourceType, resourceName);
    }
    getHostType() {
        const host = this.host.slice(2, -1);
        return standardizeResourceType(host.split('.')[0]);
    }
    getHostName() {
        const host = this.host.slice(2, -1);
        return host.split('.')[1];
    }
    getHostFriendlyString() {
        return `${this.getHostType()}.${this.getHostName()}`;
    }
}
exports.Service = Service;
// Define the Setting class
class Setting {
    resource;
    envs;
    constructor(json) {
        this.resource = json.resource || '';
        this.envs = json.envs && json.envs.length > 0
            ? json.envs.map((env) => new SettingEnv(env))
            : [];
    }
    toJson() {
        return {
            resource: this.resource,
            envs: this.envs.map(env => env.toJson())
        };
    }
    // Add a new environment variable to the setting
    addEnv(env) {
        this.envs.push(env);
    }
    // Remove an environment variable
    removeEnv(envName) {
        this.envs = this.envs.filter(env => env.name !== envName);
    }
    // Get host type
    getHostType() {
        const host = this.resource.slice(2, -1);
        return standardizeResourceType(host.split('.')[0]);
    }
}
exports.Setting = Setting;
// Define the SettingEnv class
class SettingEnv {
    name;
    value;
    constructor(json) {
        if (json.name === undefined) {
            throw new Error("Name is required to construct a SettingEnv");
        }
        this.name = json.name;
        this.value = json.value || '';
    }
    toJson() {
        return {
            name: this.name,
            value: this.value
        };
    }
}
exports.SettingEnv = SettingEnv;
// Define the ProjectScanResult class
// The ProjectScanResult class is the main model for the project scan result.
// It could be deserialize from the JSON object and serialize back to JSON object.
class ProjectScanResult {
    projectType;
    iacTypes;
    resources;
    bindingCreatedResources;
    bindings;
    services;
    settings;
    constructor(json = undefined) {
        // set defaults
        this.projectType = json?.projectType;
        this.iacTypes = new Array();
        this.resources = new Array();
        this.bindingCreatedResources = new Array();
        this.bindings = new Array();
        this.services = new Array();
        this.settings = new Array();
        if (json === undefined) {
            return;
        }
        if (!json.resources) {
            throw new Error("Resources are required to construct a Project");
        }
        this.projectType = json.projectType;
        if (json.iacTypes && json.iacTypes.length > 0) {
            this.iacTypes = json.iacTypes.map((iacType) => iacType);
        }
        if (json.resources && json.resources.length > 0) {
            this.resources = json.resources.map((resource) => new Resource(resource));
        }
        if (json.bindings && json.bindings.length > 0) {
            this.bindings = json.bindings.map((binding) => new Binding(binding));
        }
        if (json.services && json.services.length) {
            this.services = json.services.map((service) => new Service(service));
        }
        if (json.settings && json.settings.length > 0) {
            this.settings = json.settings.map((setting) => new Setting(setting));
        }
    }
    toJson() {
        return {
            projectType: this.projectType,
            iacTypes: this.iacTypes,
            resources: this.resources.map(resource => resource.toJson()),
            bindings: this.bindings.map(binding => binding.toJson()),
            services: this.services.map(service => service.toJson()),
            settings: this.settings.map(setting => setting.toJson())
        };
    }
    // Add a new resource instance to the project
    addResource(resourceType, resourceName) {
        const resource = this.resources.find(resource => resource.type === resourceType);
        if (resource) {
            resource.addInstance(resourceName);
        }
        else {
            this.resources.push(new Resource({
                type: flattenResourceType(resourceType),
                instances: [{ name: resourceName }]
            }));
        }
    }
    // Remove a resource instance from the project
    removeResource(resourceType, resourceName) {
        const resource = this.resources.find(resource => resource.type === resourceType);
        if (resource) {
            resource.instances = resource.instances.filter(instance => instance.name !== resourceName);
        }
        // remove the resource type if no instance left
        if (resource && resource.instances.length === 0) {
            this.resources = this.resources.filter(resource => resource.type !== resourceType);
        }
    }
    addBindingCreatedResource(resource) {
        this.resources.push(resource);
        this.bindingCreatedResources.push(resource);
    }
    // Add a new service to the project
    addService(serviceJson) {
        this.services.push(new Service(serviceJson));
    }
    // Remove a service from the project
    removeService(targetService) {
        if (targetService) {
            // remove the service
            this.services = this.services.filter(service => service.project !== targetService.project);
            // remove the bindings related to the service
            this.bindings = this.bindings.filter(binding => binding.source !== targetService.host && binding.target !== targetService.host);
            // remove the settings related to the service
            this.settings = this.settings.filter(setting => setting.resource !== targetService.host);
            // remove the resource instance related to the service
            this.removeResource(targetService.getHostType(), targetService.getHostName());
        }
    }
    // Add a linked resource to the project
    addLinkedResource(service, resourceType, resourceName, reuseResource = true) {
        // add the resource instance if it doesn't exist
        const resource = this.resources.find(resource => resource.type === resourceType);
        if (!resource || !reuseResource) {
            this.addResource(resourceType);
        }
        // get the name of the resource instance
        const instances = this.resources.find(resource => resource.type === resourceType)?.instances;
        resourceName = instances ? instances[instances.length - 1].name : '';
        // add a binding if not exists
        const source = service.host;
        const target = getResourceReference(resourceType, resourceName);
        const binding = this.bindings.find(binding => binding.source === source && binding.target === target);
        if (!binding) {
            this.bindings.push(new Binding({ source: source, target: target }));
        }
    }
    // Add a new setting to a service in the project
    addServiceSetting(service, name, value) {
        // check if the service setting exists
        let setting = this.settings.find(setting => setting.resource === service.host);
        if (!setting) {
            setting = new Setting({ resource: service.host });
            this.settings.push(setting);
        }
        setting.addEnv(new SettingEnv({ name: name, value: value || '' }));
    }
    // Remove a setting from a service in the project
    removeServiceSetting(service, envName) {
        const setting = this.settings.find(setting => setting.resource === service.host);
        if (setting) {
            setting.removeEnv(envName);
        }
    }
    getMarkdownForResource(resource) {
        let markdownString = '';
        for (const instance of resource.instances) {
            const sourceRef = getResourceReference(resource.type, instance.name);
            const service = this.services.find(service => service.host === sourceRef);
            markdownString += service ?
                `- ${resource.type} (\`${service.getHostName()}\`)\n` :
                `- ${resource.type}\n`;
        }
        return markdownString;
    }
    exportMarkdownForRecommend() {
        let markdownString = '';
        if (this.services.length >= 0) {
            markdownString += '### Detected services: \n';
            for (const service of this.services) {
                markdownString += `Service \`${service.getHostName()}\`: \n`;
                markdownString += `- Path: \`${service.project}\`\n`;
                markdownString += `- Language: \`${service.language}\`\n`;
                markdownString += `- Port: \`${service.port}\`\n\n`;
            }
        }
        markdownString += '### Recommended resources: \n';
        for (const resource of this.resources) {
            markdownString += this.getMarkdownForResource(resource);
        }
        return markdownString;
    }
    // Export result of the bind command, which is the rest part of the
    // project scan result, for faster response time.
    exportMarkdownForBind() {
        let markdownString = '';
        if (this.bindingCreatedResources.length > 0) {
            markdownString += '### Additional recommended resources: \n';
            for (const resource of this.bindingCreatedResources) {
                markdownString += this.getMarkdownForResource(resource);
            }
        }
        markdownString += '### Resource bindings: \n';
        if (this.bindings.length === 0) {
            markdownString += 'No resource bindings found.\n';
        }
        else {
            for (const binding of this.bindings) {
                const sourceType = binding.getSourceType();
                const targetType = binding.getTargetType();
                const sourceService = this.services.find(service => service.host === binding.source);
                const targetService = this.services.find(service => service.host === binding.target);
                const bindingSource = sourceService ?
                    `${sourceType} (\`${sourceService.getHostName()}\`)` :
                    sourceType;
                const bindingTarget = targetService ?
                    `${targetType} (\`${targetService.getHostName()}\`)` :
                    targetType;
                // binding [source -- target]
                markdownString += `- ${bindingSource} -- ${bindingTarget}\n\n`;
                // binding custom keys
                if (binding.customKeys && Object.keys(binding.customKeys).length > 0) {
                    const sortedCustomKeys = Object.entries(binding.customKeys)
                        .sort(([a], [b]) => a.localeCompare(b))
                        .reduce((items, [key, value]) => {
                        items[key] = value;
                        return items;
                    }, {});
                    markdownString += `| Binding Env Name | Description |\n`;
                    markdownString += `| ----------- | ---------- |\n`;
                    for (const key in sortedCustomKeys) {
                        let description = exports.TargetEnvDefaultKeys.find((item) => `AZURE_${item.defaultKey}` === key)?.description
                            || this.convertKeyToDescriptionString(key);
                        markdownString += `| ${sortedCustomKeys[key]} |  ${description} | \n`;
                    }
                }
            }
        }
        markdownString += '### Environment variables: \n';
        if (this.settings.length === 0) {
            markdownString += 'No other environment variables found.\n\n';
        }
        else {
            for (const setting of this.settings) {
                const service = this.services.find(service => service.host === setting.resource);
                const settingSource = service ?
                    `${service.getHostType()} (\`${service.getHostName()}\`)` :
                    setting.getHostType();
                // source of the setting
                markdownString += `- ${settingSource}\n\n`;
                // environment variables
                if (setting.envs && setting.envs.length > 0) {
                    setting.envs.sort((a, b) => a.name.localeCompare(b.name));
                    markdownString += `| Name | Value |\n`;
                    markdownString += `| ---- | ----- |\n`;
                    for (const env of setting.envs) {
                        markdownString += `| ${env.name} | ${env.value ? env.value : '<Need user to provide>'} |\n`;
                    }
                }
            }
        }
        return markdownString;
    }
    // Export the project to a markdown string
    exportMarkdownForUpdate() {
        return this.exportMarkdownForRecommend() + this.exportMarkdownForBind();
    }
    // return description string for the key, eg. SERVICE_URL to service url.
    convertKeyToDescriptionString(key) {
        return key.split('_').map((word) => word.toLowerCase()).join(' ');
    }
}
exports.ProjectScanResult = ProjectScanResult;
var EnvType;
(function (EnvType) {
    EnvType["Endpoint"] = "endpoint";
    EnvType["Config"] = "config";
    // AzureBotService
    EnvType["BotId"] = "BOT_ID";
    EnvType["BotPassword"] = "BOT_PASSWORD";
    EnvType["BotDomain"] = "BOT_DOMAIN";
    // AzureMysql
    EnvType["MysqlHost"] = "MYSQL_HOST";
    EnvType["MysqlUser"] = "MYSQL_USER";
    EnvType["MysqlPassword"] = "MYSQL_PASSWORD";
    EnvType["MysqlDatabase"] = "MYSQL_DATABASE";
    EnvType["MysqlPort"] = "MYSQL_PORT";
    EnvType["MysqlSsl"] = "MYSQL_SSL";
    EnvType["MysqlConnectionString"] = "MYSQL_CONNECTIONSTRING";
    // AzurePostgreSql
    EnvType["PostgreSqlHost"] = "POSTGRESQL_HOST";
    EnvType["PostgreSqlUser"] = "POSTGRESQL_USERNAME";
    EnvType["PostgreSqlPassword"] = "POSTGRESQL_PASSWORD";
    EnvType["PostgreSqlDatabase"] = "POSTGRESQL_DATABASE";
    EnvType["PostgreSqlPort"] = "POSTGRESQL_PORT";
    EnvType["PostgreSqlConnectionString"] = "POSTGRESQL_CONNECTIONSTRING";
    // AzureCosmosDB
    EnvType["CosmosConnectionstring"] = "COSMOS_CONNECTIONSTRING";
    EnvType["CosmosResourceEndpoint"] = "COSMOS_RESOURCEENDPOINT";
    // AzureKeyVault
    EnvType["KeyvaultResourceendpoint"] = "KEYVAULT_RESOURCEENDPOINT";
    // AzureApplicaitonInsights
    EnvType["AppinsightsConnectionString"] = "APPINSIGHTS_CONNECTION_STRING";
    // AzureOpenAI
    EnvType["OpenaiBase"] = "OPENAI_BASE";
    EnvType["OpenaiName"] = "OPENAI_NAME";
    EnvType["OpenaiKey"] = "OPENAI_KEY";
    EnvType["OpenaiDeployment"] = "OPENAI_DEPLOYMENT";
    // AzureBlobStorage
    EnvType["StorageblobConnectionstring"] = "STORAGEBLOB_CONNECTIONSTRING";
    EnvType["StorageblobResourceendpoint"] = "STORAGEBLOB_RESOURCEENDPOINT";
    EnvType["StorageblobName"] = "STORAGEBLOB_NAME";
    // StorageAccount
    EnvType["StorageBlobendpoint"] = "STORAGE_BLOBENDPOINT";
    EnvType["StorageTableendpoint"] = "STORAGE_TABLEENDPOINT";
    EnvType["StorageQueueendpoint"] = "STORAGE_QUEUEENDPOINT";
    EnvType["StorageFileendpoint"] = "STORAGE_FILEENDPOINT";
    EnvType["StorageConnectionString"] = "STORAGE_CONNECTIONSTRING";
    // AzureAISearch
    EnvType["AisearchEndpoint"] = "AISEARCH_ENDPOINT";
    EnvType["AisearchName"] = "AISEARCH_NAME";
    EnvType["AisearchKey"] = "AISEARCH_KEY";
    // AzureAIServices
    EnvType["AiservicesOpenaiBase"] = "AISERVICES_OPENAI_BASE";
    EnvType["AiservicesSpeechEndpoint"] = "AISERVICES_SPEECH_ENDPOINT";
    EnvType["AiservicesCognitiveservicesEndpoint"] = "AISERVICES_COGNITIVESERVICES_ENDPOINT";
    EnvType["AiservicesKey"] = "AISERVICES_KEY";
    EnvType["AiservicesOpenaiDeployment"] = "AISERVICES_OPENAI_DEPLOYMENT";
    // SignalR
    EnvType["SignalRConnectionString"] = "SIGNALR_CONNECTIONSTRING";
    // Redis
    EnvType["RedisHost"] = "REDIS_HOST";
    EnvType["RedisPort"] = "REDIS_PORT";
    EnvType["RedisPassword"] = "REDIS_PASSWORD";
    EnvType["RedisSsl"] = "REDIS_SSL";
    EnvType["RedisDatabase"] = "REDIS_DATABASE";
    EnvType["RedisKey"] = "REDIS_KEY";
    EnvType["RedisConnectionString"] = "REDIS_CONNECTIONSTRING";
    // ServiceBus
    EnvType["ServiceBusConnectionString"] = "SERVICEBUS_CONNECTIONSTRING";
    EnvType["ServiceBusNameSpace"] = "SERVICEBUS_FULLYQUALIFIEDNAMESPACE";
    // Sql
    EnvType["SqlServer"] = "SQL_SERVER";
    EnvType["SqlDatabase"] = "SQL_DATABASE";
    EnvType["SqlUsername"] = "SQL_USERNAME";
    EnvType["SqlPassword"] = "SQL_PASSWORD";
    EnvType["SqlPort"] = "SQL_PORT";
    EnvType["SqlConnectionString"] = "SQL_CONNECTIONSTRING";
    // WebPubSub
    EnvType["WebPubSubConnectionString"] = "WEBPUBSUB_CONNECTIONSTRING";
    EnvType["WebPubSubHost"] = "WEBPUBSUB_HOST";
})(EnvType || (exports.EnvType = EnvType = {}));
;
exports.TargetEnvDefaultKeys = [
    {
        targetType: BindingResourceType.AZURE_AI_SEARCH,
        defaultKey: EnvType.AisearchEndpoint,
        description: "The endpoint of AI Search service",
    },
    {
        targetType: BindingResourceType.AZURE_AI_SEARCH,
        defaultKey: EnvType.AisearchKey,
        description: "The API key to connect to AI Search service",
    },
    {
        targetType: BindingResourceType.AZURE_AI_SERVICES,
        defaultKey: EnvType.AiservicesOpenaiBase,
        description: "The base URL or endpoint of Azure OpenAI in AI Services",
    },
    {
        targetType: BindingResourceType.AZURE_AI_SERVICES,
        defaultKey: EnvType.AiservicesSpeechEndpoint,
        description: "The endpoint of Azure Speech service in AI Services",
    },
    {
        targetType: BindingResourceType.AZURE_AI_SERVICES,
        defaultKey: EnvType.AiservicesCognitiveservicesEndpoint,
        description: "The endpoint of Azure Cognitive Services in AI Services",
    },
    {
        targetType: BindingResourceType.AZURE_AI_SERVICES,
        defaultKey: EnvType.AiservicesKey,
        description: "The API key to connect to AI Services",
    },
    {
        targetType: BindingResourceType.AZURE_AI_SERVICES,
        defaultKey: EnvType.AiservicesOpenaiDeployment,
        description: "The deployment ID of Azure OpenAI in AI Services",
    },
    {
        targetType: BindingResourceType.AZURE_APPLICATION_INSIGHTS,
        defaultKey: EnvType.AppinsightsConnectionString,
        description: "The connection string of Application Insights/Monitor",
    },
    {
        targetType: BindingResourceType.AZURE_BOT_SERVICE,
        defaultKey: EnvType.BotId,
        description: "The ID of the Azure Bot Service",
    },
    {
        targetType: BindingResourceType.AZURE_BOT_SERVICE,
        defaultKey: EnvType.BotPassword,
        description: "The password of the Azure Bot Service",
    },
    {
        targetType: BindingResourceType.AZURE_BOT_SERVICE,
        defaultKey: EnvType.BotDomain,
        description: "The domain of the Azure Bot Service",
    },
    {
        targetType: BindingResourceType.AZURE_CACHE_FOR_REDIS,
        defaultKey: EnvType.RedisConnectionString,
        description: "The connection string of Redis Cache",
    },
    {
        targetType: BindingResourceType.AZURE_CACHE_FOR_REDIS,
        defaultKey: EnvType.RedisHost,
        description: "The host/endpoint of Redis Cache",
    },
    {
        targetType: BindingResourceType.AZURE_CACHE_FOR_REDIS,
        defaultKey: EnvType.RedisPort,
        description: "The port of Redis Cache",
    },
    {
        targetType: BindingResourceType.AZURE_CACHE_FOR_REDIS,
        defaultKey: EnvType.RedisPassword,
        description: "The password of Redis Cache",
    },
    {
        targetType: BindingResourceType.AZURE_CACHE_FOR_REDIS,
        defaultKey: EnvType.RedisSsl,
        description: "The SSL setting of Redis Cache",
    },
    {
        targetType: BindingResourceType.AZURE_CACHE_FOR_REDIS,
        defaultKey: EnvType.RedisDatabase,
        description: "The database of Redis Cache",
    },
    {
        targetType: BindingResourceType.AZURE_CACHE_FOR_REDIS,
        defaultKey: EnvType.RedisKey,
        description: "The key of Redis Cache",
    },
    {
        targetType: BindingResourceType.AZURE_COSMOS_DB,
        defaultKey: EnvType.CosmosConnectionstring,
        description: "The connection string of Cosmos DB/MongoDB",
    },
    {
        targetType: BindingResourceType.AZURE_COSMOS_DB,
        defaultKey: EnvType.CosmosResourceEndpoint,
        description: "The endpoint of Cosmos DB/MongoDB",
    },
    {
        targetType: BindingResourceType.AZURE_KEY_VAULT,
        defaultKey: EnvType.KeyvaultResourceendpoint,
        description: "The endpoint of Key Vault",
    },
    {
        targetType: BindingResourceType.AZURE_DATABASE_FOR_MYSQL,
        defaultKey: EnvType.MysqlHost,
        description: "The host/endpoint of MySQL Database",
    },
    {
        targetType: BindingResourceType.AZURE_DATABASE_FOR_MYSQL,
        defaultKey: EnvType.MysqlUser,
        description: "The user name of MySQL Database",
    },
    {
        targetType: BindingResourceType.AZURE_DATABASE_FOR_MYSQL,
        defaultKey: EnvType.MysqlPassword,
        description: "The password of MySQL Database",
    },
    {
        targetType: BindingResourceType.AZURE_DATABASE_FOR_MYSQL,
        defaultKey: EnvType.MysqlDatabase,
        description: "The database name of MySQL Database",
    },
    {
        targetType: BindingResourceType.AZURE_DATABASE_FOR_MYSQL,
        defaultKey: EnvType.MysqlPort,
        description: "The port of MySQL Database",
    },
    {
        targetType: BindingResourceType.AZURE_DATABASE_FOR_MYSQL,
        defaultKey: EnvType.MysqlSsl,
        description: "The SSL setting of MySQL Database",
    },
    {
        targetType: BindingResourceType.AZURE_DATABASE_FOR_MYSQL,
        defaultKey: EnvType.MysqlConnectionString,
        description: "The connection string of MySQL Database",
    },
    {
        targetType: BindingResourceType.AZURE_DATABASE_FOR_POSTGRESQL,
        defaultKey: EnvType.PostgreSqlHost,
        description: "The host/endpoint of PostgreSQL Database/DB for PostgreSQL",
    },
    {
        targetType: BindingResourceType.AZURE_DATABASE_FOR_POSTGRESQL,
        defaultKey: EnvType.PostgreSqlUser,
        description: "The user name ofPostgreSQL Database/DB for PostgreSQL",
    },
    {
        targetType: BindingResourceType.AZURE_DATABASE_FOR_POSTGRESQL,
        defaultKey: EnvType.PostgreSqlPassword,
        description: "The password of PostgreSQL Database/DB for PostgreSQL",
    },
    {
        targetType: BindingResourceType.AZURE_DATABASE_FOR_POSTGRESQL,
        defaultKey: EnvType.PostgreSqlDatabase,
        description: "The database name of PostgreSQL Database/DB for PostgreSQL",
    },
    {
        targetType: BindingResourceType.AZURE_DATABASE_FOR_POSTGRESQL,
        defaultKey: EnvType.PostgreSqlPort,
        description: "The port of PostgreSQL Database/DB for PostgreSQL",
    },
    {
        targetType: BindingResourceType.AZURE_DATABASE_FOR_POSTGRESQL,
        defaultKey: EnvType.PostgreSqlConnectionString,
        description: "The connection string of PostgreSQL Database/DB for PostgreSQL",
    },
    {
        targetType: BindingResourceType.AZURE_OPEN_AI,
        defaultKey: EnvType.OpenaiBase,
        description: "The base URL or endpoint of OpenAI service",
    },
    {
        targetType: BindingResourceType.AZURE_OPEN_AI,
        defaultKey: EnvType.OpenaiKey,
        description: "The API key to connect to OpenAI service",
    },
    {
        targetType: BindingResourceType.AZURE_OPEN_AI,
        defaultKey: EnvType.OpenaiDeployment,
        description: "The deployment ID of OpenAI service",
    },
    {
        targetType: BindingResourceType.AZURE_SQL_DATABASE,
        defaultKey: EnvType.SqlServer,
        description: "The server name of SQL Database",
    },
    {
        targetType: BindingResourceType.AZURE_SQL_DATABASE,
        defaultKey: EnvType.SqlDatabase,
        description: "The database name of SQL Database",
    },
    {
        targetType: BindingResourceType.AZURE_SQL_DATABASE,
        defaultKey: EnvType.SqlUsername,
        description: "The user name of SQL Database",
    },
    {
        targetType: BindingResourceType.AZURE_SQL_DATABASE,
        defaultKey: EnvType.SqlPassword,
        description: "The password of SQL Database",
    },
    {
        targetType: BindingResourceType.AZURE_SQL_DATABASE,
        defaultKey: EnvType.SqlPort,
        description: "The port of SQL Database",
    },
    {
        targetType: BindingResourceType.AZURE_SQL_DATABASE,
        defaultKey: EnvType.SqlConnectionString,
        description: "The connection string of SQL Database",
    },
    {
        targetType: BindingResourceType.AZURE_STORAGE_ACCOUNT,
        defaultKey: EnvType.StorageBlobendpoint,
        description: "The blob endpoint of Storage Account",
    },
    {
        targetType: BindingResourceType.AZURE_STORAGE_ACCOUNT,
        defaultKey: EnvType.StorageTableendpoint,
        description: "The table endpoint of Storage Account",
    },
    {
        targetType: BindingResourceType.AZURE_STORAGE_ACCOUNT,
        defaultKey: EnvType.StorageQueueendpoint,
        description: "The queue endpoint of Storage Account",
    },
    {
        targetType: BindingResourceType.AZURE_STORAGE_ACCOUNT,
        defaultKey: EnvType.StorageFileendpoint,
        description: "The file endpoint of Storage Account",
    },
    {
        targetType: BindingResourceType.AZURE_STORAGE_ACCOUNT,
        defaultKey: EnvType.StorageConnectionString,
        description: "The connection string of Storage Account",
    },
    {
        targetType: BindingResourceType.AZURE_SERVICE_BUS,
        defaultKey: EnvType.ServiceBusConnectionString,
        description: "The connection string of Service Bus",
    },
    {
        targetType: BindingResourceType.AZURE_SERVICE_BUS,
        defaultKey: EnvType.ServiceBusNameSpace,
        description: "The fully qualified namespace of Service Bus",
    },
    {
        targetType: BindingResourceType.AZURE_WEB_PUBSUB,
        defaultKey: EnvType.WebPubSubHost,
        description: "The host/endpoint of Web PubSub service",
    },
    {
        targetType: BindingResourceType.AZURE_WEB_PUBSUB,
        defaultKey: EnvType.WebPubSubConnectionString,
        description: "The connection string of Web PubSub service",
    },
    {
        targetType: BindingResourceType.AZURE_SIGNALR_SERVICE,
        defaultKey: EnvType.SignalRConnectionString,
        description: "The connection string of SignalR service",
    },
];
