// codetocloudmcp/out/common/stringHelpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRandomStr = getRandomStr;
exports.getHashStr = getHashStr;
exports.getKebab = getKebab;
exports.getPascal = getPascal;
exports.getCamel = getCamel;
exports.checkGUIDFormat = checkGUIDFormat;
exports.isValidResourceType = isValidResourceType;
exports.checkValidIaCType = checkValidIaCType;
const params_1 = require("../toolsImplementations/getAzureResourceRules/params");
const models_1 = require("./models");
function getRandomStr(length) {
    const alphabet = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
    return Array.from({ length }, () => alphabet.charAt(Math.floor(Math.random() * alphabet.length))).join('').toLowerCase();
}
function getHashStr(str, length) {
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
        hash = (hash << 5) - hash + str.charCodeAt(i);
        hash |= 0; // Convert to 32bit integer
    }
    const hashStr = Math.abs(hash).toString(36).substring(0, length);
    return hashStr;
}
function getKebab(name) {
    return name.replace('_', '-').toLowerCase();
}
function getPascal(name) {
    return name.split('-').map(word => word.charAt(0).toUpperCase() + word.slice(1)).join('');
}
function getCamel(s) {
    const words = s.split('-');
    return words[0].toLowerCase() + words.slice(1).map(word => word.charAt(0).toUpperCase() + word.slice(1)).join('');
}
function checkGUIDFormat(guid) {
    const guidRegex = /^[{(]?[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}[)}]?$/i;
    return guidRegex.test(guid);
}
function isValidResourceType(resourceType) {
    if (Object.values(params_1.RulesSupportedResourceType).includes(resourceType)) {
        return true;
    }
    if (Object.values(models_1.WafResourceType).includes(resourceType)) {
        return true;
    }
    if (resourceType === 'appservicemanagedinstance') {
        return true;
    }
    if (/^Microsoft\.[^/]+\/[^/]+$/.test(resourceType)) {
        return true;
    }
    return false;
}
function checkValidIaCType(iacType) {
    const validIaCTypes = ['bicep', 'terraform'];
    return validIaCTypes.includes(iacType.toLowerCase());
}
