// codetocloudmcp/out/toolsImplementations/getPlan/executionSteps/azCliExecutionSteps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzCliExecutionStepGenerator = void 0;
const models_1 = require("../../../common/models");
const planStep_1 = require("../../../common/planGeneration/planStep");
const toolType_1 = require("../../shared/toolType");
const baseExecutionSteps_1 = require("./baseExecutionSteps");
const AKSStepProvider_1 = require("./serviceStepProviders/AKSStepProvider");
class AzCliExecutionStepGenerator extends baseExecutionSteps_1.BaseExecutionSteps {
    constructor(parameters, stateManager) {
        super(parameters, stateManager);
    }
    generateExecutionSteps() {
        this.requiredToolsInPlan = [];
        var planSteps = new planStep_1.PlanStep("Execution Steps", planStep_1.PlanStepBulletType.Numbered);
        if (this.shouldIncludeContainerization()) {
            planSteps.addStep(this.getContainerizationStep());
        }
        if (this.shouldIncludeProvision()) {
            planSteps.addStep(this.getCreateIaCFilesStep());
        }
        planSteps.addStep(this.getEnvironmentSetupStep());
        if (this.shouldIncludeProvision()) {
            planSteps.addStep(this.getProvisionStep());
        }
        if (this.deployOption === models_1.DeployOption.deployOnly) {
            planSteps.addStep(this.getExistingResourceStep());
        }
        if (this.shouldIncludeDeployment()) {
            planSteps.addStep(this.getDeploymentStep());
        }
        if (this.shouldIncludeCICDPipeline()) {
            planSteps.addStep(this.getCICDPipelineStep());
        }
        planSteps.addStep(this.getSummarizationStep());
        return {
            planSteps: planSteps,
            requiredTools: this.requiredToolsInPlan
        };
    }
    getProvisionStep() {
        if (!this.shouldIncludeProvision()) {
            return '';
        }
        let provisionSteps = new planStep_1.PlanStep(`Provision Azure Infrastructure with azcli`, planStep_1.PlanStepBulletType.Numbered);
        provisionSteps.addStep(`If region and SKU validation is skipped in previous steps, use ${this.stateManager.getToolName(toolType_1.ToolType.GetRegionsForDeployment)} to fill in the placeholders.`);
        if (this.iacOptions === models_1.IacType.Bicep) {
            provisionSteps.addStep("Provision resources with Az CLI `az deployment sub create`(targetScope = 'subscription') or `az deployment group create`(targetScope = 'resourceGroup').");
        }
        else {
            provisionSteps.addStep("Run the generated Terraform script to provision the resources and confirm the result.");
        }
        let postStep = this.getPostProvisionStep();
        if (postStep) {
            provisionSteps.addStep(postStep);
        }
        return provisionSteps;
    }
    getCreateIaCFilesStep() {
        if (!this.shouldIncludeProvision()) {
            return '';
        }
        var createIaCSteps = new planStep_1.PlanStep(`Create ${this.iacOptions} Files (**IMPORTANT: SKIP az login even if terminal shows login errors in this step**)`, planStep_1.PlanStepBulletType.Numbered);
        createIaCSteps.addStep(`Provisioning tool: azcli. Expected files: ${this.iacOptions}.`);
        createIaCSteps.addStep(this.getRegionsPrompt(this.hasAKS()));
        var filesExistSteps = new planStep_1.PlanStep(`If expected files exist, please follow the below steps`, planStep_1.PlanStepBulletType.Bulleted);
        filesExistSteps.addSteps([
            `Existing file paths: <filePaths>`,
            `Check if the Azure resources in the existing files match what we required in the plan. If not, add steps to update the existing files with missing Azure resource.`,
        ]);
        var filesNotExistSteps = new planStep_1.PlanStep(`If expected files do not exist, please follow the below steps`, planStep_1.PlanStepBulletType.Bulleted);
        const getResourceRulesToolName = this.stateManager.getToolName(toolType_1.ToolType.GetAzureResourceRules);
        filesNotExistSteps.addSteps([
            `Generate the ${this.iacOptions} files in infra/ folder. Call tool \`${getResourceRulesToolName}\` to get IaC rules`,
        ]);
        if (this.iacOptions === models_1.IacType.Terraform) {
            filesNotExistSteps.addSteps([
                "Add a provisioning script to provision resources with `terraform` command.",
                "Ensure to run `terraform init` and `terraform validate` before `terraform apply` to initialize the working directory."
            ]);
        }
        this.requiredToolsInPlan.push(getResourceRulesToolName);
        createIaCSteps.addSteps([filesExistSteps, filesNotExistSteps]);
        let iacValidationStep = new planStep_1.PlanStep(`Validate all generated files to ensure they are runnable and free of syntax errors`, planStep_1.PlanStepBulletType.Bulleted);
        iacValidationStep.addStep(`Call 'get_errors' on all generated files and iterate until all errors are resolved.`);
        if (this.iacOptions === models_1.IacType.Terraform) {
            iacValidationStep.addStep(`Call 'terraform validate' and fix any errors.`);
        }
        createIaCSteps.addStep(iacValidationStep);
        return createIaCSteps;
    }
    getEnvironmentSetupStep() {
        var envSetupStep = new planStep_1.PlanStep(`Env setup for AzCLI`, planStep_1.PlanStepBulletType.Numbered);
        envSetupStep.addSteps([`Install AZ CLI if not installed.`,
            `Ensure there is a default subscription set. If provided, override the default subscription with the provided subscription ID.`,
            `Subscription ID: ${this.subscriptionId ? this.subscriptionId : "Use default subscription"}`,]);
        if (this.includePasswordless()) {
            envSetupStep.addStep(`Install Service Connector AzCLI extension: az extension add --name serviceconnector-passwordless --upgrade`);
        }
        if (this.targetAppServices.some(service => service.azureComputeHost === models_1.TargetAppServiceType.AzureAppServiceManagedInstance) && !this.iacOptions) {
            envSetupStep.addStep(`AZ CLI supports Azure App Service Managed Instance in version >=2.80.0`);
        }
        return envSetupStep;
    }
    getDeploymentStep() {
        if (!this.shouldIncludeDeployment()) {
            return '';
        }
        let deploymentSteps = new planStep_1.PlanStep("Deployment", planStep_1.PlanStepBulletType.Numbered);
        let deploySet = {};
        this.serviceProviders.forEach(provider => {
            const providerKey = provider.constructor.name;
            if (deploySet[providerKey]) {
                return;
            }
            deploySet[providerKey] = true;
            deploymentSteps.addStep(provider.getDeploymentStep(this.deployOption, models_1.DeploymentTool.AzCLI));
        });
        deploymentSteps.addStep(this.getValidationStep());
        return deploymentSteps;
    }
    getValidationStep() {
        if (!this.shouldIncludeValidation()) {
            return undefined;
        }
        const lastProvider = this.serviceProviders.at(-1);
        if (!lastProvider || lastProvider instanceof AKSStepProvider_1.AKSStepProvider) {
            const validationStep = new planStep_1.PlanStep("Deployment Validation", planStep_1.PlanStepBulletType.Numbered);
            validationStep.addStep("Check the deployed application is running.");
            return validationStep;
        }
        return lastProvider.getValidationStep(this.stateManager, models_1.DeploymentTool.AzCLI);
    }
}
exports.AzCliExecutionStepGenerator = AzCliExecutionStepGenerator;
