// codetocloudmcp/out/toolsImplementations/getPlan/executionSteps/baseExecutionSteps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseExecutionSteps = void 0;
const models_1 = require("../../../common/models");
const planStep_1 = require("../../../common/planGeneration/planStep");
const project_1 = require("../../../common/project");
const toolType_1 = require("../../shared/toolType");
const IaCPromptProviderFactory_1 = require("./IaCPromptProviders/IaCPromptProviderFactory");
const ServiceStepProviderFactory_1 = require("./serviceStepProviders/ServiceStepProviderFactory");
class BaseExecutionSteps {
    deployOption;
    deploymentTool;
    targetAppServices;
    subscriptionId;
    resourceGroupName;
    iacOptions;
    serviceProviders;
    iaCPromptProvider;
    stateManager;
    requiredToolsInPlan = [];
    projectName = "";
    constructor(parameters, stateManager) {
        this.projectName = parameters.projectName;
        this.deployOption = parameters.deployOption;
        this.targetAppServices = parameters.services;
        this.subscriptionId = parameters.subscriptionId;
        this.resourceGroupName = parameters.resourceGroupName;
        this.deploymentTool = parameters.provisioningTool;
        this.iacOptions = parameters.iacOptions || '';
        this.serviceProviders = ServiceStepProviderFactory_1.ServiceStepProviderFactory.createProviders(parameters.projectName, parameters.services, parameters.deployOption, this.iacOptions);
        this.iaCPromptProvider = IaCPromptProviderFactory_1.IaCPromptProviderFactory.createIaCProviders(this.iacOptions);
        this.stateManager = stateManager;
    }
    // Common methods with default implementations
    shouldIncludeProvision() {
        return this.deployOption !== models_1.DeployOption.deployOnly && this.iacOptions !== '';
    }
    shouldIncludeContainerization() {
        return this.deployOption !== models_1.DeployOption.provisionOnly &&
            (this.targetAppServices.some(i => i.azureComputeHost === models_1.TargetAppServiceType.AzureKubernetesService || i.azureComputeHost === models_1.TargetAppServiceType.AzureContainerApp));
    }
    shouldIncludeDeployment() {
        return this.deployOption !== models_1.DeployOption.provisionOnly;
    }
    shouldIncludeValidation() {
        return this.deployOption !== models_1.DeployOption.provisionOnly;
    }
    shouldIncludeCICDPipeline() {
        return this.isCodingAgent();
    }
    isCodingAgent() {
        const clientInfo = this.stateManager.getClientInfo();
        return clientInfo?.name === "github-copilot-developer";
    }
    getPostProvisionStep(postProvisionStep) {
        postProvisionStep = postProvisionStep ?? new planStep_1.PlanStep(`Post-provisioning steps`, planStep_1.PlanStepBulletType.Bulleted);
        if (this.includePasswordless()) {
            postProvisionStep.addStep(`Build connection between the database and ${this.targetAppServices.map(s => s.azureComputeHost).join(", ")} with Managed Identity.`);
            // Remove duplicate CLI names so each service type is listed only once in the composed az command.
            let cliName = Array.from(new Set(this.serviceProviders.map(provider => provider.getCliName()))).join("/");
            postProvisionStep.addStep(`Create the connection using Service Connector: az ${cliName} connection create mysql-flexible/postgres-flexible/sql --connection connectionname --user-identity client-id=XX subs-id=XX --source-id <azure resource id> --tg <db resource group> --server servername --database dbname -y.`);
            postProvisionStep.addStep(`Check the connection with 'az ${cliName} connection show -g WebAppRG -n MyWebApp --connection MyConnection -o json'.`);
            if (this.hasAKS()) {
                postProvisionStep.addStep(`'kubernetesResourceName.secret' property in the output has a secret object created by the connection. Check its name and mount it to Deployment spec in envFrom.secretRef. 'configurations.name' property in the output has env variable names that will be mounted with secret object.`);
            }
            else {
                postProvisionStep.addStep(`Check the 'configurations' property in the output. Each name means an env variable name and you should use it in your application to connect to the database.`);
            }
            return postProvisionStep;
        }
    }
    getExistingResourceStep() {
        if (this.deployOption !== models_1.DeployOption.deployOnly) {
            return undefined;
        }
        let steps = new planStep_1.PlanStep("Check Azure resources existence", planStep_1.PlanStepBulletType.Numbered);
        this.serviceProviders.forEach(provider => {
            steps.addSteps(provider.getExistingResourcesSteps());
        });
        if (this.shouldIncludeContainerization() && !this.targetAppServices.some(service => service.dependencies.map(dep => dep.serviceType).includes(models_1.AzureServiceNames.AzureContainerRegistry))) {
            const acrCheckingStep = new planStep_1.PlanStep("Azure Container Registry", planStep_1.PlanStepBulletType.Bulleted);
            acrCheckingStep.addStep("login server: <>. Check with 'az acr show -o json'");
            steps.addStep(acrCheckingStep);
        }
        let creatingMissingResourcesStep = new planStep_1.PlanStep("Create missing resources", planStep_1.PlanStepBulletType.Bulleted);
        creatingMissingResourcesStep.addSteps([
            "If any resource is missing, ask user to provide the resource id or create a new one, then get the resource information with Az CLI command",
            "If user want to create new resources, generate a script to do so using Azure CLI command. Run the script and confirms all resources are ready."
        ]);
        steps.addStep(creatingMissingResourcesStep);
        return steps;
    }
    getBuildImageInstruction(buildDockerImageToolName) {
        return this.deployOption === models_1.DeployOption.deployOnly
            ? `Build each image: use 'az acr build' if ACR is available, otherwise use ${buildDockerImageToolName}.`
            : `Build each dockerfile using ${buildDockerImageToolName}.`;
    }
    getContainerizationStep() {
        if (this.shouldIncludeContainerization()) {
            var containerizationSteps = new planStep_1.PlanStep("Containerization", planStep_1.PlanStepBulletType.Bulleted);
            containerizationSteps.addSteps([
                "If Dockerfiles exist, list their paths."
            ]);
            const analyzeRepositoryToolName = this.stateManager.getToolName(toolType_1.ToolType.AnalyzeRepository);
            const planGenerateDockerfileToolName = this.stateManager.getToolName(toolType_1.ToolType.PlanGenerateDockerfile);
            const buildDockerImageToolName = this.stateManager.getToolName(toolType_1.ToolType.BuildDockerImage);
            const generateKubernetesManifestToolName = this.stateManager.getToolName(toolType_1.ToolType.GenerateKubernetesManifest);
            if (this.stateManager.toolTypeExists(toolType_1.ToolType.RunSubagent)) {
                const subagentToolName = this.stateManager.getToolName(toolType_1.ToolType.RunSubagent);
                let subAgentPrompt = `Analyze my repo using tool ${analyzeRepositoryToolName}. Then, generate Dockerfiles based on the analysis with ${planGenerateDockerfileToolName}.`;
                subAgentPrompt += ` Then, ${this.getBuildImageInstruction(buildDockerImageToolName)}`;
                if (this.hasAKS()) {
                    subAgentPrompt += ` Then, generate Kubernetes manifests using '${generateKubernetesManifestToolName}'. Use this tool to generate Helm charts instead if requested by the user.`;
                }
                subAgentPrompt += ` Return each dockerfile${this.hasAKS() ? " and Kubernetes manifests" : ""} created, and its required docker build context path, and an instruction that containerization is complete and we can continue with deployment.`;
                this.requiredToolsInPlan.push(`${subagentToolName} for containerization`);
                containerizationSteps.addSteps([
                    `If missing: call tool #${subagentToolName} with the following prompt:`,
                    `<prompt>${subAgentPrompt}</prompt>`,
                    `Agent must create Dockerfile first before deployment!`
                ]);
            }
            else {
                this.requiredToolsInPlan.push(analyzeRepositoryToolName);
                this.requiredToolsInPlan.push(planGenerateDockerfileToolName);
                this.requiredToolsInPlan.push(buildDockerImageToolName);
                containerizationSteps.addSteps([
                    `If missing: Ensure the repo is analyzed using ${analyzeRepositoryToolName}.`,
                    `Based on that response, use ${planGenerateDockerfileToolName} to generate the required Dockerfiles.`,
                    this.getBuildImageInstruction(buildDockerImageToolName)
                ]);
                if (this.hasAKS()) {
                    this.requiredToolsInPlan.push(generateKubernetesManifestToolName);
                    containerizationSteps.addSteps([
                        `Additionally, generate Kubernetes manifests using '${generateKubernetesManifestToolName}'. Use this tool to generate Helm charts instead if requested by the user.`
                    ]);
                }
                containerizationSteps.addStep(`Keep track of each dockerfile${this.hasAKS() ? " and Kubernetes manifest" : ""} created, and its required docker build context path. Agent must create Dockerfile first before deployment!`);
            }
            containerizationSteps.addStep("Output: Docker artifacts");
            return containerizationSteps;
        }
    }
    getSummarizationStep() {
        const summarizeToolName = this.stateManager.getToolName(toolType_1.ToolType.SummarizeResult);
        this.requiredToolsInPlan.push(summarizeToolName);
        let summarizeSteps = new planStep_1.PlanStep("Summarize Result", planStep_1.PlanStepBulletType.Numbered);
        var summary = `Use \`${summarizeToolName}\` tool to summarize the deployment result.
    2. Generating files: .azure/summary.${this.stateManager.getCopilotMarkdownExtension()}`;
        summarizeSteps.addStep(summary);
        return summarizeSteps;
    }
    hasAKS() {
        return this.targetAppServices.some(service => service.azureComputeHost === models_1.TargetAppServiceType.AzureKubernetesService);
    }
    includePasswordless() {
        return this.targetAppServices.some(service => service.dependencies?.filter(dep => (dep.connectionType === project_1.ConnectionType.UserIdentity || dep.connectionType === project_1.ConnectionType.SystemIdentity) && (dep.serviceType === models_1.AzureServiceNames.AzureSQLDB || dep.serviceType === models_1.AzureServiceNames.AzureMySQL || dep.serviceType === models_1.AzureServiceNames.AzurePostgresSQL)).length > 0);
    }
    getRegionsPrompt(isAKS) {
        const getRegionsToolName = this.stateManager.getToolName(toolType_1.ToolType.GetRegionsForDeployment);
        let prompt = '';
        if (this.stateManager.toolTypeExists(toolType_1.ToolType.RunSubagent)) {
            const subagentToolName = this.stateManager.getToolName(toolType_1.ToolType.RunSubagent);
            let subagentPrompt = `Grab current subscriptionID using 'az account show --query id -o tsv', call tool \`${getRegionsToolName}\` to get available region and SKUs for all needed Azure resource types. Check for <REQUIRED_RESOURCES>. Return with ONLY the available regions for each resource, and their available SKUs. For SKUs, choose one for each region-resource combination. Do not return the full list of SKUs. Also add instructions to continue the deployment. If subscription ID is not available, use 'placeholder' for region/location and SKU in ${this.iacOptions} files.`;
            this.requiredToolsInPlan.push(`${subagentToolName} for region check`);
            prompt = `Use tool #${subagentToolName} with the following prompt. Populate <REQUIRED_RESOURCES> based on your context. <prompt>${subagentPrompt}</prompt>`;
        }
        else {
            this.requiredToolsInPlan.push(getRegionsToolName);
            prompt = `**NEVER do azure login in this step, even if terminal shows login errors.**. Call tool \`${getRegionsToolName}\` to get available region and SKUs for all needed Azure resource types. If subscription ID is not available, skip calling the tool and use 'placeholder' for region/location and SKU in ${this.iacOptions} files.`;
            prompt += this.serviceProviders.map(provider => provider.getRegionCheckPrompts()).join(" ");
        }
        return prompt;
    }
    getCICDPipelineStep() {
        const deployPipelineToolName = this.stateManager.getToolName(toolType_1.ToolType.DeployPipeline);
        this.requiredToolsInPlan.push(deployPipelineToolName);
        let cicdStep = new planStep_1.PlanStep("Set up CI/CD Pipeline", planStep_1.PlanStepBulletType.Numbered);
        cicdStep.addSteps([
            `Use tool \`${deployPipelineToolName}\` and set pipelinePlatform to be GithubActions to set up a CI/CD pipeline to deploy to Azure.`,
            `Generate the pipeline file based on the deployment configuration from previtous steps.`
        ]);
        return cicdStep;
    }
}
exports.BaseExecutionSteps = BaseExecutionSteps;
