// containerization-assist-mcp/dist-cjs/src/app/orchestrator.js
"use strict";
/**
 * Tool Orchestrator
 * Tool execution with optional dependency resolution
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.discoverBuiltInPolicies = discoverBuiltInPolicies;
exports.discoverUserPolicies = discoverUserPolicies;
exports.discoverCustomPolicies = discoverCustomPolicies;
exports.discoverPolicies = discoverPolicies;
exports.createOrchestrator = createOrchestrator;
const index_1 = require("../types/index");
const logger_1 = require("../lib/logger");
const module_url_1 = require("../lib/module-url");
const module_path_resolver_1 = require("../lib/module-path-resolver");
const context_1 = require("../mcp/context");
const errors_1 = require("../lib/errors");
const orchestrator_types_1 = require("./orchestrator-types");
const tool_helpers_1 = require("../lib/tool-helpers");
const tool_logger_1 = require("../lib/tool-logger");
const policy_rego_1 = require("../config/policy-rego");
const node_fs_1 = require("node:fs");
const node_path_1 = require("node:path");
const constants_1 = require("../config/constants");
// Capture import.meta.url at module scope (only available in ESM builds)
// This will be undefined in CJS builds, which is expected
const MODULE_URL = (0, module_url_1.getModuleUrl)();
// ===== Types =====
/**
 * Discover built-in policy files from the policies directory
 * Returns paths to all .rego files (excluding test files)
 *
 * Uses shared module path resolver with symlink support.
 * Works in both ESM (dist/) and CJS (dist-cjs/) builds, and when installed via npm.
 */
function discoverBuiltInPolicies(logger) {
    try {
        // Use shared path resolver utility
        const searchPaths = (0, module_path_resolver_1.resolveModulePaths)({
            relativePath: 'policies',
            logger,
            ...(MODULE_URL && { moduleUrl: MODULE_URL }),
        });
        // Try each search path until we find one that exists
        for (const policiesDir of searchPaths) {
            if ((0, node_fs_1.existsSync)(policiesDir)) {
                // Find all .rego files except test files
                const files = (0, node_fs_1.readdirSync)(policiesDir)
                    .filter((file) => file.endsWith('.rego') && !file.endsWith('_test.rego'))
                    .map((file) => (0, node_path_1.resolve)((0, node_path_1.join)(policiesDir, file)));
                if (files.length > 0) {
                    logger.debug({ count: files.length, dir: policiesDir }, 'Discovered built-in policies');
                    return files;
                }
            }
        }
        return [];
    }
    catch (error) {
        logger.warn({ error }, 'Failed to discover built-in policies');
        return [];
    }
}
/**
 * Discover policies in policies.user/ directory (source installation)
 * Returns paths to all .rego files (excluding test files)
 */
function discoverUserPolicies(logger) {
    try {
        // Search upward for policies.user directory (similar to built-in search)
        let currentDir = process.cwd();
        let policiesUserDir = (0, node_path_1.join)(currentDir, 'policies.user');
        let attempts = 0;
        const maxAttempts = 5;
        while (!(0, node_fs_1.existsSync)(policiesUserDir) && attempts < maxAttempts) {
            const parentDir = (0, node_path_1.dirname)(currentDir);
            if (parentDir === currentDir)
                break;
            currentDir = parentDir;
            policiesUserDir = (0, node_path_1.join)(currentDir, 'policies.user');
            attempts++;
        }
        if (!(0, node_fs_1.existsSync)(policiesUserDir)) {
            return [];
        }
        const files = (0, node_fs_1.readdirSync)(policiesUserDir)
            .filter((file) => file.endsWith('.rego') && !file.endsWith('_test.rego'))
            .map((file) => (0, node_path_1.resolve)((0, node_path_1.join)(policiesUserDir, file)));
        if (files.length > 0) {
            logger.info({ policiesUserDir, count: files.length }, 'Discovered user policies from policies.user/');
        }
        return files;
    }
    catch (error) {
        logger.warn({ error }, 'Failed to discover policies.user/ policies');
        return [];
    }
}
/**
 * Discover policies in custom directory (NPM installation)
 * Returns paths to all .rego files (excluding test files)
 */
function discoverCustomPolicies(customPath, logger) {
    try {
        const resolvedPath = (0, node_path_1.resolve)(customPath);
        if (!(0, node_fs_1.existsSync)(resolvedPath)) {
            logger.warn({ path: resolvedPath }, 'Custom policy path does not exist');
            return [];
        }
        const stats = (0, node_fs_1.statSync)(resolvedPath);
        // If it's a file, return it directly
        if (stats.isFile()) {
            if (resolvedPath.endsWith('.rego')) {
                return [resolvedPath];
            }
            logger.warn({ path: resolvedPath }, 'Custom policy path is not a .rego file');
            return [];
        }
        // If it's a directory, discover all .rego files
        if (stats.isDirectory()) {
            const files = (0, node_fs_1.readdirSync)(resolvedPath)
                .filter((file) => file.endsWith('.rego') && !file.endsWith('_test.rego'))
                .map((file) => (0, node_path_1.resolve)((0, node_path_1.join)(resolvedPath, file)));
            return files;
        }
        return [];
    }
    catch (error) {
        logger.warn({ error, path: customPath }, 'Failed to discover custom policies');
        return [];
    }
}
/**
 * Discover policy files with priority-ordered search paths:
 * 1. Built-in policies/ directory (lowest priority)
 * 2. policies.user/ directory (middle priority - source installation users)
 * 3. Custom directory via CUSTOM_POLICY_PATH (highest priority - NPM users)
 *
 * Returns array of policy paths, with higher priority policies later
 * (later policies override earlier ones during merging)
 */
function discoverPolicies(logger) {
    const allPolicies = [];
    // Priority 3 (lowest): Built-in policies
    const builtInPolicies = discoverBuiltInPolicies(logger);
    allPolicies.push(...builtInPolicies);
    // Priority 2: policies.user/ directory (source installation)
    const userPolicies = discoverUserPolicies(logger);
    allPolicies.push(...userPolicies);
    // Priority 1 (highest): Custom directory via env var
    const customPath = process.env[constants_1.ENV_VARS.CUSTOM_POLICY_PATH];
    if (customPath) {
        const customPolicies = discoverCustomPolicies(customPath, logger);
        if (customPolicies.length > 0) {
            logger.info({ path: customPath, count: customPolicies.length }, 'Discovered custom policies from CUSTOM_POLICY_PATH');
            allPolicies.push(...customPolicies);
        }
    }
    return allPolicies;
}
/**
 * Create a child logger with additional bindings
 * Assumes Pino logger (fail fast if not)
 */
function childLogger(logger, bindings) {
    return logger.child(bindings);
}
/**
 * Create a ToolContext for the given request
 * Delegates to the canonical createToolContext from @mcp/context
 */
function createContextForTool(request, logger, policy) {
    const metadata = request.metadata;
    logger.debug({
        hasPolicy: !!policy,
        toolName: request.toolName,
    }, 'Creating tool context');
    return (0, context_1.createToolContext)(logger, {
        ...(metadata?.signal && { signal: metadata.signal }),
        ...(metadata?.progress !== undefined && { progress: metadata.progress }),
        ...(metadata?.sendNotification && { sendNotification: metadata.sendNotification }),
        ...(policy && { policy }),
    });
}
/**
 * Create a tool orchestrator
 */
function createOrchestrator(options) {
    const { registry, server, config = { chainHintsMode: orchestrator_types_1.CHAINHINTSMODE.ENABLED } } = options;
    const logger = options.logger || (0, logger_1.createLogger)({ name: 'orchestrator' });
    logger.info('createOrchestrator called - policy loading will happen on first tool execution');
    // Cache the loaded policy to avoid reloading on every execution
    let policyCache;
    let policyLoadPromise;
    async function execute(request) {
        const { toolName } = request;
        const tool = registry.get(toolName);
        if (!tool) {
            return (0, index_1.Failure)(errors_1.ERROR_MESSAGES.TOOL_NOT_FOUND(toolName));
        }
        const contextualLogger = childLogger(logger, {
            tool: tool.name,
            ...(request.metadata?.loggerContext ?? {}),
        });
        logger.info({ toolName, hasPolicyPromise: !!policyLoadPromise }, 'Orchestrator execute called');
        // Load policies once (with Promise-based guard to prevent race conditions)
        if (!policyLoadPromise) {
            policyLoadPromise = (async () => {
                try {
                    // Use new priority-ordered discovery (includes built-in, user, and custom policies)
                    const policyPaths = discoverPolicies(logger);
                    if (policyPaths.length === 0) {
                        logger.warn({ cwd: process.cwd() }, 'No policies discovered - tools will run without policy constraints');
                        return; // Not an error, just no policies available
                    }
                    logger.info({ count: policyPaths.length, paths: policyPaths }, 'Discovered policies, loading...');
                    const policyResult = await (0, policy_rego_1.loadAndMergeRegoPolicies)(policyPaths, logger);
                    if (policyResult.ok) {
                        policyCache = policyResult.value;
                        logger.info({
                            total: policyPaths.length,
                            policyPaths,
                        }, 'Policies loaded and merged successfully');
                    }
                    else {
                        logger.error({ error: policyResult.error }, 'Failed to load policies - tools will run without policy constraints');
                        // Note: We don't throw here, tools can still run without policies
                        // But we log clearly that policies are not available
                    }
                }
                catch (error) {
                    // Reset promise to allow retry on next execution
                    policyLoadPromise = undefined;
                    logger.error({ error }, 'Failed to load policies - will retry on next tool execution');
                    // Re-throw to signal failure
                    throw error;
                }
            })();
        }
        // Wait for policy loading to complete if in progress
        try {
            await policyLoadPromise;
        }
        catch {
            // Policy loading errors are logged above, continue without policies
            logger.warn('Policy loading failed, continuing tool execution without policies');
        }
        return await executeWithOrchestration(tool, request, {
            registry,
            logger: contextualLogger,
            config,
            ...(server && { server }),
        }, policyCache);
    }
    function close() {
        // Cleanup policy resources if loaded
        if (policyCache) {
            policyCache.close();
        }
    }
    return { execute, close };
}
/**
 * Execute with full orchestration (dependencies, policies)
 */
async function executeWithOrchestration(tool, request, env, policy) {
    const { params } = request;
    const { logger } = env;
    // Validate parameters using Zod safeParse
    const validation = validateParams(params, tool.schema);
    if (!validation.ok)
        return validation;
    const validatedParams = validation.value;
    const toolContext = createContextForTool(request, logger, policy);
    const tracker = (0, tool_helpers_1.createStandardizedToolTracker)(tool.name, {}, logger);
    const startTime = Date.now();
    const logEntry = (0, tool_logger_1.createToolLogEntry)(tool.name, validatedParams);
    // Execute tool directly (single attempt)
    try {
        const result = await tool.handler(validatedParams, toolContext);
        const durationMs = Date.now() - startTime;
        logEntry.output = result.ok ? result.value : { error: result.error };
        logEntry.success = result.ok;
        logEntry.durationMs = durationMs;
        if (!result.ok) {
            logEntry.error = result.error;
            if (result.guidance) {
                logEntry.errorGuidance = result.guidance;
            }
        }
        await (0, tool_logger_1.logToolExecution)(logEntry, logger);
        // Add metadata to successful results
        if (result.ok) {
            let valueWithMessages = result.value;
            if (env.config.chainHintsMode === orchestrator_types_1.CHAINHINTSMODE.ENABLED && tool.chainHints) {
                valueWithMessages = {
                    ...valueWithMessages,
                    nextSteps: tool.chainHints.success,
                };
            }
            result.value = valueWithMessages;
        }
        else if (result.guidance && tool.chainHints) {
            // Add failure hint to error guidance
            result.guidance.hint = tool.chainHints.failure;
        }
        tracker.complete({});
        return result;
    }
    catch (error) {
        const durationMs = Date.now() - startTime;
        const errorMessage = error.message || 'Unknown error';
        logEntry.output = { error: errorMessage };
        logEntry.success = false;
        logEntry.durationMs = durationMs;
        logEntry.error = errorMessage;
        await (0, tool_logger_1.logToolExecution)(logEntry, logger);
        logger.error({ error: errorMessage }, 'Tool execution failed');
        tracker.fail(error);
        return (0, index_1.Failure)(errorMessage);
    }
}
/**
 * Validate parameters against schema using safeParse
 */
function validateParams(params, schema) {
    const parsed = schema.safeParse(params);
    if (!parsed.success) {
        const issues = parsed.error.issues.map((i) => `${i.path.join('.')}: ${i.message}`).join(', ');
        return (0, index_1.Failure)(errors_1.ERROR_MESSAGES.VALIDATION_FAILED(issues));
    }
    return (0, index_1.Success)(parsed.data);
}
