// containerization-assist-mcp/dist-cjs/src/tools/scan-image/schema.js
"use strict";
/**
 * Schema definition for scan tool
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.scanImageSchema = void 0;
const zod_1 = require("zod");
exports.scanImageSchema = zod_1.z.object({
    imageId: zod_1.z.string().describe('Docker image ID or name to scan'),
    severity: zod_1.z
        .union([
        zod_1.z.enum(['LOW', 'MEDIUM', 'HIGH', 'CRITICAL']),
        zod_1.z.enum(['low', 'medium', 'high', 'critical']),
    ])
        .optional()
        .describe('Minimum severity to report'),
    scanType: zod_1.z
        .enum(['vulnerability', 'config', 'all'])
        .default('vulnerability') // Added default
        .describe('Type of scan to perform'),
    scanner: zod_1.z
        .enum(['trivy', 'snyk', 'grype'])
        .default('trivy') // Added default
        .describe('Scanner to use for vulnerability detection'),
    enableAISuggestions: zod_1.z
        .boolean()
        .default(true)
        .describe('Enable AI-powered suggestions for vulnerability remediation'),
    aiEnhancementOptions: zod_1.z
        .object({
        mode: zod_1.z.enum(['suggestions', 'fixes', 'analysis']).default('suggestions'),
        focus: zod_1.z.enum(['security', 'performance', 'best-practices', 'all']).default('security'),
        confidence: zod_1.z.number().min(0).max(1).default(0.8),
        maxSuggestions: zod_1.z.number().min(1).max(10).default(5),
        includeExamples: zod_1.z.boolean().default(true),
    })
        .optional()
        .describe('Configuration for AI-powered enhancement'),
});
