// containerization-assist-mcp/dist-cjs/src/tools/scan-image/tool.js
"use strict";
/**
 * Scan Image Tool - Standardized Implementation
 *
 * Scans Docker images for security vulnerabilities
 * Uses standardized helpers for consistency
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.scanImage = void 0;
const tool_context_helpers_1 = require("../../lib/tool-context-helpers");
const scanner_1 = require("../../infra/security/scanner");
const types_1 = require("../../types");
const index_1 = require("../../knowledge/index");
const summary_helpers_1 = require("../../lib/summary-helpers");
const types_2 = require("./types");
/**
 * Analyze vulnerabilities and generate actionable fix recommendations
 * Groups vulnerabilities by package for cleaner output
 */
function analyzeFixActions(vulnerabilities) {
    const fixable = vulnerabilities.filter((v) => v.fixedVersion !== undefined);
    if (fixable.length === 0)
        return [];
    const normalizeVersion = (value) => {
        if (value === undefined)
            return 'unknown';
        return value.trim() === '' ? 'unknown' : value;
    };
    const byPackageVersion = new Map();
    for (const vuln of fixable) {
        const currentVersion = normalizeVersion(vuln.version);
        const fixedVersion = normalizeVersion(vuln.fixedVersion);
        const key = `${vuln.package}::${currentVersion}::${fixedVersion}`;
        const grouped = byPackageVersion.get(key) || [];
        grouped.push(vuln);
        byPackageVersion.set(key, grouped);
    }
    const actions = Array.from(byPackageVersion.entries()).map(([, vulns]) => {
        const severityCounts = {
            critical: 0,
            high: 0,
            medium: 0,
            low: 0,
            negligible: 0,
            unknown: 0,
        };
        for (const vuln of vulns) {
            switch (vuln.severity) {
                case 'CRITICAL':
                    severityCounts.critical += 1;
                    break;
                case 'HIGH':
                    severityCounts.high += 1;
                    break;
                case 'MEDIUM':
                    severityCounts.medium += 1;
                    break;
                case 'LOW':
                    severityCounts.low += 1;
                    break;
                case 'NEGLIGIBLE':
                    severityCounts.negligible += 1;
                    break;
                case 'UNKNOWN':
                    severityCounts.unknown += 1;
                    break;
            }
        }
        const vulnerabilityIds = [...new Set(vulns.map((v) => v.id))].slice(0, 5);
        const packageName = vulns[0]?.package ?? 'unknown';
        const currentVersion = normalizeVersion(vulns[0]?.version);
        const fixedVersion = normalizeVersion(vulns[0]?.fixedVersion);
        return {
            type: 'UPGRADE_PACKAGE',
            action: `Upgrade ${packageName}`,
            current: `${packageName}: ${currentVersion}`,
            recommended: `${packageName}: ${fixedVersion}`,
            package: packageName,
            vulnerabilitiesFixed: vulns.length,
            severityCounts,
            vulnerabilityIds,
        };
    });
    const severityOrder = [
        'critical',
        'high',
        'medium',
        'low',
        'negligible',
        'unknown',
    ];
    actions.sort((a, b) => {
        for (const severity of severityOrder) {
            if (b.severityCounts[severity] !== a.severityCounts[severity]) {
                return b.severityCounts[severity] - a.severityCounts[severity];
            }
        }
        return b.vulnerabilitiesFixed - a.vulnerabilitiesFixed;
    });
    return actions.slice(0, 5);
}
/**
 * Scan image handler - direct execution without wrapper
 */
async function handleScanImage(params, context) {
    if (!params || typeof params !== 'object') {
        return (0, types_1.Failure)('Invalid parameters provided', {
            message: 'Parameters must be a valid object',
            hint: 'Tool received invalid or missing parameters',
            resolution: 'Ensure parameters are provided as a JSON object',
        });
    }
    const { logger, timer } = (0, tool_context_helpers_1.setupToolContext)(context, 'scan-image');
    const { scanner = 'osv', severity, scanType = 'vulnerability', enableAISuggestions = true, } = params;
    if (scanType !== 'vulnerability') {
        return (0, types_1.Failure)(`Scan type '${scanType}' is not supported`, {
            message: 'Only vulnerability scans are currently supported',
            hint: 'Use scanType="vulnerability" for image vulnerability checks',
            resolution: 'Update scanType to "vulnerability" and retry',
        });
    }
    // Map severity parameter to threshold
    const finalSeverityThreshold = severity
        ? severity.toLowerCase()
        : 'high';
    try {
        logger.info({ scanner, severityThreshold: finalSeverityThreshold, scanType }, 'Starting image security scan');
        const securityScanner = (0, scanner_1.createSecurityScanner)(logger, scanner);
        const imageId = params.imageId;
        if (!imageId) {
            return (0, types_1.Failure)('No image specified. Provide imageId parameter.', {
                message: 'Missing required parameter: imageId',
                hint: 'Image ID or name must be specified to scan',
                resolution: 'Add imageId parameter with the Docker image ID or name to scan',
            });
        }
        logger.info({ imageId, scanner, scanType }, 'Scanning image for vulnerabilities');
        // Scan image using security scanner
        const scanResultWrapper = await securityScanner.scanImage(imageId);
        if (!scanResultWrapper.ok) {
            return (0, types_1.Failure)(`Failed to scan image: ${scanResultWrapper.error ?? 'Unknown error'}`, scanResultWrapper.guidance);
        }
        const scanResult = scanResultWrapper.value;
        // Convert BasicScanResult to DockerScanResult
        const dockerScanResult = {
            vulnerabilities: scanResult.vulnerabilities.map((v) => ({
                id: v.id,
                severity: v.severity,
                package: v.package,
                version: v.version,
                description: v.description,
                ...(v.fixedVersion !== undefined && { fixedVersion: v.fixedVersion }),
            })),
            summary: {
                critical: scanResult.criticalCount,
                high: scanResult.highCount,
                medium: scanResult.mediumCount,
                low: scanResult.lowCount,
                negligible: scanResult.negligibleCount,
                unknown: scanResult.unknownCount,
                total: scanResult.totalVulnerabilities,
            },
            scanTime: scanResult.scanDate.toISOString(),
            metadata: {
                image: imageId,
            },
        };
        // Determine if scan passed based on threshold
        const thresholdMap = {
            critical: ['critical'],
            high: ['critical', 'high'],
            medium: ['critical', 'high', 'medium'],
            low: ['critical', 'high', 'medium', 'low'],
        };
        const failingSeverities = thresholdMap[finalSeverityThreshold] || thresholdMap['high'];
        let vulnerabilityCount = 0;
        for (const severity of failingSeverities) {
            if (severity === 'critical') {
                vulnerabilityCount += scanResult.criticalCount;
            }
            else if (severity === 'high') {
                vulnerabilityCount += scanResult.highCount;
            }
            else if (severity === 'medium') {
                vulnerabilityCount += scanResult.mediumCount;
            }
            else if (severity === 'low') {
                vulnerabilityCount += scanResult.lowCount;
            }
        }
        const passed = vulnerabilityCount === 0;
        // Get knowledge-based remediation guidance for vulnerabilities
        let remediationGuidance = [];
        if (enableAISuggestions &&
            dockerScanResult.vulnerabilities &&
            dockerScanResult.vulnerabilities.length > 0) {
            try {
                // Create a summary of vulnerabilities for knowledge query
                const vulnSummary = dockerScanResult.vulnerabilities
                    .slice(0, 10) // Limit to top 10 for performance
                    .map((v) => `${v.package}:${v.version} (${v.severity})`)
                    .join(', ');
                const securityKnowledge = await (0, index_1.getKnowledgeForCategory)('security', vulnSummary);
                // Add general security recommendations
                const generalKnowledge = await (0, index_1.getKnowledgeForCategory)('security', undefined);
                remediationGuidance = [
                    ...securityKnowledge.map((match) => ({
                        vulnerability: 'General',
                        recommendation: match.entry.recommendation,
                        ...(match.entry.severity && { severity: match.entry.severity }),
                        ...(match.entry.example && { example: match.entry.example }),
                    })),
                    ...generalKnowledge.map((match) => ({
                        vulnerability: 'Best Practice',
                        recommendation: match.entry.recommendation,
                        ...(match.entry.severity && { severity: match.entry.severity }),
                        ...(match.entry.example && { example: match.entry.example }),
                    })),
                ];
                logger.info({ guidanceCount: remediationGuidance.length }, 'Added knowledge-based remediation guidance');
            }
            catch (error) {
                logger.debug({ error }, 'Failed to get remediation guidance, continuing without');
            }
        }
        // Generate summary
        const vulnSummary = (0, summary_helpers_1.formatVulnerabilities)({
            critical: scanResult.criticalCount,
            high: scanResult.highCount,
            medium: scanResult.mediumCount,
            low: scanResult.lowCount,
            total: scanResult.totalVulnerabilities,
        });
        const remediationText = remediationGuidance.length > 0
            ? ` ${(0, summary_helpers_1.pluralize)(remediationGuidance.length, 'remediation')} available.`
            : '';
        const summary = (0, summary_helpers_1.buildStatusSummary)(passed, `🔒 Security scan passed (${scanner}). ${vulnSummary}.${remediationText}`, `🔒 Security scan failed (${scanner}). ${vulnSummary}.${remediationText}`);
        const vulnerabilityDetails = dockerScanResult.vulnerabilities && dockerScanResult.vulnerabilities.length > 0
            ? dockerScanResult.vulnerabilities.map((v) => ({
                id: v.id ?? 'UNKNOWN',
                severity: v.severity,
                package: v.package ?? 'unknown',
                version: v.version ?? 'unknown',
                description: v.description ?? 'No description available',
                ...(v.fixedVersion !== undefined && { fixedVersion: v.fixedVersion }),
            }))
            : undefined;
        const recommendedActions = vulnerabilityDetails && vulnerabilityDetails.length > 0
            ? analyzeFixActions(vulnerabilityDetails)
            : undefined;
        const result = {
            summary,
            success: true,
            scanner,
            ...(recommendedActions && recommendedActions.length > 0 && { recommendedActions }),
            ...(remediationGuidance.length > 0 && { remediationGuidance }),
            vulnerabilities: {
                critical: scanResult.criticalCount,
                high: scanResult.highCount,
                medium: scanResult.mediumCount,
                low: scanResult.lowCount,
                negligible: scanResult.negligibleCount,
                unknown: scanResult.unknownCount,
                total: scanResult.totalVulnerabilities,
            },
            ...(vulnerabilityDetails && { vulnerabilityDetails }),
            scanTime: dockerScanResult.scanTime ?? new Date().toISOString(),
            passed,
        };
        timer.end({
            vulnerabilities: scanResult.totalVulnerabilities,
            critical: scanResult.criticalCount,
            high: scanResult.highCount,
            passed,
        });
        logger.info({
            imageId,
            vulnerabilities: scanResult.totalVulnerabilities,
            passed,
        }, 'Image scan completed');
        return (0, types_1.Success)(result);
    }
    catch (error) {
        timer.error(error);
        logger.error({ error }, 'Image scan failed');
        const errorMessage = error instanceof Error ? error.message : String(error);
        return (0, types_1.Failure)(errorMessage, {
            message: errorMessage,
            hint: 'An unexpected error occurred during the security scan',
            resolution: 'Verify that the scanner is installed and accessible, the image exists, and you have proper permissions',
        });
    }
}
/**
 * Scan image tool
 */
exports.scanImage = handleScanImage;
const tool_1 = require("../../types/tool");
exports.default = (0, tool_1.tool)({
    ...types_2.scanImageToolDefinition,
    handler: handleScanImage,
});
