// containerization-assist-mcp/dist-cjs/src/tools/scan-image/tool.js
"use strict";
/**
 * Scan Image Tool - Standardized Implementation
 *
 * Scans Docker images for security vulnerabilities
 * Uses standardized helpers for consistency
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.scanImage = void 0;
const tool_context_helpers_1 = require("../../lib/tool-context-helpers");
const scanner_1 = require("../../infra/security/scanner");
const types_1 = require("../../types");
const index_1 = require("../../knowledge/index");
const schema_1 = require("./schema");
const summary_helpers_1 = require("../../lib/summary-helpers");
/**
 * Scan image handler - direct execution without wrapper
 */
async function handleScanImage(params, context) {
    if (!params || typeof params !== 'object') {
        return (0, types_1.Failure)('Invalid parameters provided', {
            message: 'Parameters must be a valid object',
            hint: 'Tool received invalid or missing parameters',
            resolution: 'Ensure parameters are provided as a JSON object',
        });
    }
    const { logger, timer } = (0, tool_context_helpers_1.setupToolContext)(context, 'scan-image');
    const { scanner = 'trivy', severity } = params;
    // Map severity parameter to threshold
    const finalSeverityThreshold = severity
        ? severity.toLowerCase()
        : 'high';
    try {
        logger.info({ scanner, severityThreshold: finalSeverityThreshold }, 'Starting image security scan');
        const securityScanner = (0, scanner_1.createSecurityScanner)(logger, scanner);
        const imageId = params.imageId;
        if (!imageId) {
            return (0, types_1.Failure)('No image specified. Provide imageId parameter.', {
                message: 'Missing required parameter: imageId',
                hint: 'Image ID or name must be specified to scan',
                resolution: 'Add imageId parameter with the Docker image ID or name to scan',
            });
        }
        logger.info({ imageId, scanner }, 'Scanning image for vulnerabilities');
        // Scan image using security scanner
        const scanResultWrapper = await securityScanner.scanImage(imageId);
        if (!scanResultWrapper.ok) {
            return (0, types_1.Failure)(`Failed to scan image: ${scanResultWrapper.error ?? 'Unknown error'}`, scanResultWrapper.guidance);
        }
        const scanResult = scanResultWrapper.value;
        // Convert BasicScanResult to DockerScanResult
        const dockerScanResult = {
            vulnerabilities: scanResult.vulnerabilities.map((v) => ({
                id: v.id,
                severity: v.severity,
                package: v.package,
                version: v.version,
                description: v.description,
                ...(v.fixedVersion !== undefined && { fixedVersion: v.fixedVersion }),
            })),
            summary: {
                critical: scanResult.criticalCount,
                high: scanResult.highCount,
                medium: scanResult.mediumCount,
                low: scanResult.lowCount,
                negligible: scanResult.negligibleCount,
                unknown: scanResult.unknownCount,
                total: scanResult.totalVulnerabilities,
            },
            scanTime: scanResult.scanDate.toISOString(),
            metadata: {
                image: imageId,
            },
        };
        // Determine if scan passed based on threshold
        const thresholdMap = {
            critical: ['critical'],
            high: ['critical', 'high'],
            medium: ['critical', 'high', 'medium'],
            low: ['critical', 'high', 'medium', 'low'],
        };
        const failingSeverities = thresholdMap[finalSeverityThreshold] || thresholdMap['high'];
        let vulnerabilityCount = 0;
        for (const severity of failingSeverities) {
            if (severity === 'critical') {
                vulnerabilityCount += scanResult.criticalCount;
            }
            else if (severity === 'high') {
                vulnerabilityCount += scanResult.highCount;
            }
            else if (severity === 'medium') {
                vulnerabilityCount += scanResult.mediumCount;
            }
            else if (severity === 'low') {
                vulnerabilityCount += scanResult.lowCount;
            }
        }
        const passed = vulnerabilityCount === 0;
        // Get knowledge-based remediation guidance for vulnerabilities
        let remediationGuidance = [];
        if (dockerScanResult.vulnerabilities && dockerScanResult.vulnerabilities.length > 0) {
            try {
                // Create a summary of vulnerabilities for knowledge query
                const vulnSummary = dockerScanResult.vulnerabilities
                    .slice(0, 10) // Limit to top 10 for performance
                    .map((v) => `${v.package}:${v.version} (${v.severity})`)
                    .join(', ');
                const securityKnowledge = await (0, index_1.getKnowledgeForCategory)('security', vulnSummary);
                // Add general security recommendations
                const generalKnowledge = await (0, index_1.getKnowledgeForCategory)('security', undefined);
                remediationGuidance = [
                    ...securityKnowledge.map((match) => ({
                        vulnerability: 'General',
                        recommendation: match.entry.recommendation,
                        ...(match.entry.severity && { severity: match.entry.severity }),
                        ...(match.entry.example && { example: match.entry.example }),
                    })),
                    ...generalKnowledge.map((match) => ({
                        vulnerability: 'Best Practice',
                        recommendation: match.entry.recommendation,
                        ...(match.entry.severity && { severity: match.entry.severity }),
                        ...(match.entry.example && { example: match.entry.example }),
                    })),
                ];
                logger.info({ guidanceCount: remediationGuidance.length }, 'Added knowledge-based remediation guidance');
            }
            catch (error) {
                logger.debug({ error }, 'Failed to get remediation guidance, continuing without');
            }
        }
        // Generate summary
        const vulnSummary = (0, summary_helpers_1.formatVulnerabilities)({
            critical: scanResult.criticalCount,
            high: scanResult.highCount,
            medium: scanResult.mediumCount,
            low: scanResult.lowCount,
            total: scanResult.totalVulnerabilities,
        });
        const remediationText = remediationGuidance.length > 0
            ? ` ${(0, summary_helpers_1.pluralize)(remediationGuidance.length, 'remediation')} available.`
            : '';
        const summary = (0, summary_helpers_1.buildStatusSummary)(passed, `🔒 Security scan passed. ${vulnSummary}.${remediationText}`, `🔒 Security scan failed. ${vulnSummary}.${remediationText}`);
        // Prepare the result
        const result = {
            summary,
            success: true,
            ...(remediationGuidance.length > 0 && { remediationGuidance }),
            vulnerabilities: {
                critical: scanResult.criticalCount,
                high: scanResult.highCount,
                medium: scanResult.mediumCount,
                low: scanResult.lowCount,
                negligible: scanResult.negligibleCount,
                unknown: scanResult.unknownCount,
                total: scanResult.totalVulnerabilities,
            },
            scanTime: dockerScanResult.scanTime ?? new Date().toISOString(),
            passed,
        };
        timer.end({
            vulnerabilities: scanResult.totalVulnerabilities,
            critical: scanResult.criticalCount,
            high: scanResult.highCount,
            passed,
        });
        logger.info({
            imageId,
            vulnerabilities: scanResult.totalVulnerabilities,
            passed,
        }, 'Image scan completed');
        return (0, types_1.Success)(result);
    }
    catch (error) {
        timer.error(error);
        logger.error({ error }, 'Image scan failed');
        const errorMessage = error instanceof Error ? error.message : String(error);
        return (0, types_1.Failure)(errorMessage, {
            message: errorMessage,
            hint: 'An unexpected error occurred during the security scan',
            resolution: 'Verify that the scanner (Trivy) is installed and accessible, the image exists, and you have proper permissions',
        });
    }
}
/**
 * Scan image tool
 */
exports.scanImage = handleScanImage;
const tool_1 = require("../../types/tool");
exports.default = (0, tool_1.tool)({
    name: 'scan-image',
    description: 'Scan Docker images for security vulnerabilities with knowledge-based remediation guidance',
    category: 'security',
    version: '2.0.0',
    schema: schema_1.scanImageSchema,
    metadata: {
        knowledgeEnhanced: true,
    },
    chainHints: {
        success: 'Security scan passed! Proceed with push-image to push to a registry, or continue with deployment preparation.',
        failure: 'Security scan found vulnerabilities. Use fix-dockerfile to address security issues in your base images and dependencies.',
    },
    handler: handleScanImage,
});
