// @azure/arm-cognitiveservices/dist/commonjs/models/index.js
"use strict";
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.KnownCapabilityHostProvisioningState = exports.KnownCapabilityHostKind = exports.KnownManagedPEStatus = exports.KnownManagedPERequirement = exports.KnownConnectionGroup = exports.KnownConnectionCategory = exports.KnownConnectionAuthType = exports.KnownDefenderForAISettingState = exports.KnownNspAccessRuleDirection = exports.KnownRaiPolicyContentSource = exports.KnownContentLevel = exports.KnownRaiPolicyMode = exports.KnownRaiPolicyType = exports.KnownEncryptionScopeState = exports.KnownEncryptionScopeProvisioningState = exports.KnownCommitmentPlanProvisioningState = exports.KnownDeploymentModelVersionUpgradeOption = exports.KnownDeploymentScaleType = exports.KnownDeploymentProvisioningState = exports.KnownHostingModel = exports.KnownActionType = exports.KnownOrigin = exports.KnownModelLifecycleStatus = exports.KnownQuotaUsageStatus = exports.KnownUnitType = exports.KnownResourceSkuRestrictionsReasonCode = exports.KnownScenarioType = exports.KnownAbusePenaltyAction = exports.KnownRoutingMethods = exports.KnownPublicNetworkAccess = exports.KnownPrivateEndpointConnectionProvisioningState = exports.KnownPrivateEndpointServiceConnectionStatus = exports.KnownKeySource = exports.KnownByPassSelection = exports.KnownNetworkRuleAction = exports.KnownProvisioningState = exports.KnownCreatedByType = exports.KnownSkuTier = void 0;
/** Known values of {@link SkuTier} that the service accepts. */
var KnownSkuTier;
(function (KnownSkuTier) {
    /** Free */
    KnownSkuTier["Free"] = "Free";
    /** Basic */
    KnownSkuTier["Basic"] = "Basic";
    /** Standard */
    KnownSkuTier["Standard"] = "Standard";
    /** Premium */
    KnownSkuTier["Premium"] = "Premium";
    /** Enterprise */
    KnownSkuTier["Enterprise"] = "Enterprise";
})(KnownSkuTier || (exports.KnownSkuTier = KnownSkuTier = {}));
/** Known values of {@link CreatedByType} that the service accepts. */
var KnownCreatedByType;
(function (KnownCreatedByType) {
    /** User */
    KnownCreatedByType["User"] = "User";
    /** Application */
    KnownCreatedByType["Application"] = "Application";
    /** ManagedIdentity */
    KnownCreatedByType["ManagedIdentity"] = "ManagedIdentity";
    /** Key */
    KnownCreatedByType["Key"] = "Key";
})(KnownCreatedByType || (exports.KnownCreatedByType = KnownCreatedByType = {}));
/** Known values of {@link ProvisioningState} that the service accepts. */
var KnownProvisioningState;
(function (KnownProvisioningState) {
    /** Accepted */
    KnownProvisioningState["Accepted"] = "Accepted";
    /** Creating */
    KnownProvisioningState["Creating"] = "Creating";
    /** Deleting */
    KnownProvisioningState["Deleting"] = "Deleting";
    /** Moving */
    KnownProvisioningState["Moving"] = "Moving";
    /** Failed */
    KnownProvisioningState["Failed"] = "Failed";
    /** Succeeded */
    KnownProvisioningState["Succeeded"] = "Succeeded";
    /** ResolvingDNS */
    KnownProvisioningState["ResolvingDNS"] = "ResolvingDNS";
    /** Canceled */
    KnownProvisioningState["Canceled"] = "Canceled";
})(KnownProvisioningState || (exports.KnownProvisioningState = KnownProvisioningState = {}));
/** Known values of {@link NetworkRuleAction} that the service accepts. */
var KnownNetworkRuleAction;
(function (KnownNetworkRuleAction) {
    /** Allow */
    KnownNetworkRuleAction["Allow"] = "Allow";
    /** Deny */
    KnownNetworkRuleAction["Deny"] = "Deny";
})(KnownNetworkRuleAction || (exports.KnownNetworkRuleAction = KnownNetworkRuleAction = {}));
/** Known values of {@link ByPassSelection} that the service accepts. */
var KnownByPassSelection;
(function (KnownByPassSelection) {
    /** None */
    KnownByPassSelection["None"] = "None";
    /** AzureServices */
    KnownByPassSelection["AzureServices"] = "AzureServices";
})(KnownByPassSelection || (exports.KnownByPassSelection = KnownByPassSelection = {}));
/** Known values of {@link KeySource} that the service accepts. */
var KnownKeySource;
(function (KnownKeySource) {
    /** MicrosoftCognitiveServices */
    KnownKeySource["MicrosoftCognitiveServices"] = "Microsoft.CognitiveServices";
    /** MicrosoftKeyVault */
    KnownKeySource["MicrosoftKeyVault"] = "Microsoft.KeyVault";
})(KnownKeySource || (exports.KnownKeySource = KnownKeySource = {}));
/** Known values of {@link PrivateEndpointServiceConnectionStatus} that the service accepts. */
var KnownPrivateEndpointServiceConnectionStatus;
(function (KnownPrivateEndpointServiceConnectionStatus) {
    /** Pending */
    KnownPrivateEndpointServiceConnectionStatus["Pending"] = "Pending";
    /** Approved */
    KnownPrivateEndpointServiceConnectionStatus["Approved"] = "Approved";
    /** Rejected */
    KnownPrivateEndpointServiceConnectionStatus["Rejected"] = "Rejected";
})(KnownPrivateEndpointServiceConnectionStatus || (exports.KnownPrivateEndpointServiceConnectionStatus = KnownPrivateEndpointServiceConnectionStatus = {}));
/** Known values of {@link PrivateEndpointConnectionProvisioningState} that the service accepts. */
var KnownPrivateEndpointConnectionProvisioningState;
(function (KnownPrivateEndpointConnectionProvisioningState) {
    /** Succeeded */
    KnownPrivateEndpointConnectionProvisioningState["Succeeded"] = "Succeeded";
    /** Creating */
    KnownPrivateEndpointConnectionProvisioningState["Creating"] = "Creating";
    /** Deleting */
    KnownPrivateEndpointConnectionProvisioningState["Deleting"] = "Deleting";
    /** Failed */
    KnownPrivateEndpointConnectionProvisioningState["Failed"] = "Failed";
})(KnownPrivateEndpointConnectionProvisioningState || (exports.KnownPrivateEndpointConnectionProvisioningState = KnownPrivateEndpointConnectionProvisioningState = {}));
/** Known values of {@link PublicNetworkAccess} that the service accepts. */
var KnownPublicNetworkAccess;
(function (KnownPublicNetworkAccess) {
    /** Enabled */
    KnownPublicNetworkAccess["Enabled"] = "Enabled";
    /** Disabled */
    KnownPublicNetworkAccess["Disabled"] = "Disabled";
})(KnownPublicNetworkAccess || (exports.KnownPublicNetworkAccess = KnownPublicNetworkAccess = {}));
/** Known values of {@link RoutingMethods} that the service accepts. */
var KnownRoutingMethods;
(function (KnownRoutingMethods) {
    /** Priority */
    KnownRoutingMethods["Priority"] = "Priority";
    /** Weighted */
    KnownRoutingMethods["Weighted"] = "Weighted";
    /** Performance */
    KnownRoutingMethods["Performance"] = "Performance";
})(KnownRoutingMethods || (exports.KnownRoutingMethods = KnownRoutingMethods = {}));
/** Known values of {@link AbusePenaltyAction} that the service accepts. */
var KnownAbusePenaltyAction;
(function (KnownAbusePenaltyAction) {
    /** Throttle */
    KnownAbusePenaltyAction["Throttle"] = "Throttle";
    /** Block */
    KnownAbusePenaltyAction["Block"] = "Block";
})(KnownAbusePenaltyAction || (exports.KnownAbusePenaltyAction = KnownAbusePenaltyAction = {}));
/** Known values of {@link ScenarioType} that the service accepts. */
var KnownScenarioType;
(function (KnownScenarioType) {
    /** None */
    KnownScenarioType["None"] = "none";
    /** Agent */
    KnownScenarioType["Agent"] = "agent";
})(KnownScenarioType || (exports.KnownScenarioType = KnownScenarioType = {}));
/** Known values of {@link ResourceSkuRestrictionsReasonCode} that the service accepts. */
var KnownResourceSkuRestrictionsReasonCode;
(function (KnownResourceSkuRestrictionsReasonCode) {
    /** QuotaId */
    KnownResourceSkuRestrictionsReasonCode["QuotaId"] = "QuotaId";
    /** NotAvailableForSubscription */
    KnownResourceSkuRestrictionsReasonCode["NotAvailableForSubscription"] = "NotAvailableForSubscription";
})(KnownResourceSkuRestrictionsReasonCode || (exports.KnownResourceSkuRestrictionsReasonCode = KnownResourceSkuRestrictionsReasonCode = {}));
/** Known values of {@link UnitType} that the service accepts. */
var KnownUnitType;
(function (KnownUnitType) {
    /** Count */
    KnownUnitType["Count"] = "Count";
    /** Bytes */
    KnownUnitType["Bytes"] = "Bytes";
    /** Seconds */
    KnownUnitType["Seconds"] = "Seconds";
    /** Percent */
    KnownUnitType["Percent"] = "Percent";
    /** CountPerSecond */
    KnownUnitType["CountPerSecond"] = "CountPerSecond";
    /** BytesPerSecond */
    KnownUnitType["BytesPerSecond"] = "BytesPerSecond";
    /** Milliseconds */
    KnownUnitType["Milliseconds"] = "Milliseconds";
})(KnownUnitType || (exports.KnownUnitType = KnownUnitType = {}));
/** Known values of {@link QuotaUsageStatus} that the service accepts. */
var KnownQuotaUsageStatus;
(function (KnownQuotaUsageStatus) {
    /** Included */
    KnownQuotaUsageStatus["Included"] = "Included";
    /** Blocked */
    KnownQuotaUsageStatus["Blocked"] = "Blocked";
    /** InOverage */
    KnownQuotaUsageStatus["InOverage"] = "InOverage";
    /** Unknown */
    KnownQuotaUsageStatus["Unknown"] = "Unknown";
})(KnownQuotaUsageStatus || (exports.KnownQuotaUsageStatus = KnownQuotaUsageStatus = {}));
/** Known values of {@link ModelLifecycleStatus} that the service accepts. */
var KnownModelLifecycleStatus;
(function (KnownModelLifecycleStatus) {
    /** Stable */
    KnownModelLifecycleStatus["Stable"] = "Stable";
    /** Preview */
    KnownModelLifecycleStatus["Preview"] = "Preview";
    /** GenerallyAvailable */
    KnownModelLifecycleStatus["GenerallyAvailable"] = "GenerallyAvailable";
    /** Deprecating */
    KnownModelLifecycleStatus["Deprecating"] = "Deprecating";
    /** Deprecated */
    KnownModelLifecycleStatus["Deprecated"] = "Deprecated";
})(KnownModelLifecycleStatus || (exports.KnownModelLifecycleStatus = KnownModelLifecycleStatus = {}));
/** Known values of {@link Origin} that the service accepts. */
var KnownOrigin;
(function (KnownOrigin) {
    /** User */
    KnownOrigin["User"] = "user";
    /** System */
    KnownOrigin["System"] = "system";
    /** UserSystem */
    KnownOrigin["UserSystem"] = "user,system";
})(KnownOrigin || (exports.KnownOrigin = KnownOrigin = {}));
/** Known values of {@link ActionType} that the service accepts. */
var KnownActionType;
(function (KnownActionType) {
    /** Internal */
    KnownActionType["Internal"] = "Internal";
})(KnownActionType || (exports.KnownActionType = KnownActionType = {}));
/** Known values of {@link HostingModel} that the service accepts. */
var KnownHostingModel;
(function (KnownHostingModel) {
    /** Web */
    KnownHostingModel["Web"] = "Web";
    /** ConnectedContainer */
    KnownHostingModel["ConnectedContainer"] = "ConnectedContainer";
    /** DisconnectedContainer */
    KnownHostingModel["DisconnectedContainer"] = "DisconnectedContainer";
    /** ProvisionedWeb */
    KnownHostingModel["ProvisionedWeb"] = "ProvisionedWeb";
})(KnownHostingModel || (exports.KnownHostingModel = KnownHostingModel = {}));
/** Known values of {@link DeploymentProvisioningState} that the service accepts. */
var KnownDeploymentProvisioningState;
(function (KnownDeploymentProvisioningState) {
    /** Accepted */
    KnownDeploymentProvisioningState["Accepted"] = "Accepted";
    /** Creating */
    KnownDeploymentProvisioningState["Creating"] = "Creating";
    /** Deleting */
    KnownDeploymentProvisioningState["Deleting"] = "Deleting";
    /** Moving */
    KnownDeploymentProvisioningState["Moving"] = "Moving";
    /** Failed */
    KnownDeploymentProvisioningState["Failed"] = "Failed";
    /** Succeeded */
    KnownDeploymentProvisioningState["Succeeded"] = "Succeeded";
    /** Disabled */
    KnownDeploymentProvisioningState["Disabled"] = "Disabled";
    /** Canceled */
    KnownDeploymentProvisioningState["Canceled"] = "Canceled";
})(KnownDeploymentProvisioningState || (exports.KnownDeploymentProvisioningState = KnownDeploymentProvisioningState = {}));
/** Known values of {@link DeploymentScaleType} that the service accepts. */
var KnownDeploymentScaleType;
(function (KnownDeploymentScaleType) {
    /** Standard */
    KnownDeploymentScaleType["Standard"] = "Standard";
    /** Manual */
    KnownDeploymentScaleType["Manual"] = "Manual";
})(KnownDeploymentScaleType || (exports.KnownDeploymentScaleType = KnownDeploymentScaleType = {}));
/** Known values of {@link DeploymentModelVersionUpgradeOption} that the service accepts. */
var KnownDeploymentModelVersionUpgradeOption;
(function (KnownDeploymentModelVersionUpgradeOption) {
    /** OnceNewDefaultVersionAvailable */
    KnownDeploymentModelVersionUpgradeOption["OnceNewDefaultVersionAvailable"] = "OnceNewDefaultVersionAvailable";
    /** OnceCurrentVersionExpired */
    KnownDeploymentModelVersionUpgradeOption["OnceCurrentVersionExpired"] = "OnceCurrentVersionExpired";
    /** NoAutoUpgrade */
    KnownDeploymentModelVersionUpgradeOption["NoAutoUpgrade"] = "NoAutoUpgrade";
})(KnownDeploymentModelVersionUpgradeOption || (exports.KnownDeploymentModelVersionUpgradeOption = KnownDeploymentModelVersionUpgradeOption = {}));
/** Known values of {@link CommitmentPlanProvisioningState} that the service accepts. */
var KnownCommitmentPlanProvisioningState;
(function (KnownCommitmentPlanProvisioningState) {
    /** Accepted */
    KnownCommitmentPlanProvisioningState["Accepted"] = "Accepted";
    /** Creating */
    KnownCommitmentPlanProvisioningState["Creating"] = "Creating";
    /** Deleting */
    KnownCommitmentPlanProvisioningState["Deleting"] = "Deleting";
    /** Moving */
    KnownCommitmentPlanProvisioningState["Moving"] = "Moving";
    /** Failed */
    KnownCommitmentPlanProvisioningState["Failed"] = "Failed";
    /** Succeeded */
    KnownCommitmentPlanProvisioningState["Succeeded"] = "Succeeded";
    /** Canceled */
    KnownCommitmentPlanProvisioningState["Canceled"] = "Canceled";
})(KnownCommitmentPlanProvisioningState || (exports.KnownCommitmentPlanProvisioningState = KnownCommitmentPlanProvisioningState = {}));
/** Known values of {@link EncryptionScopeProvisioningState} that the service accepts. */
var KnownEncryptionScopeProvisioningState;
(function (KnownEncryptionScopeProvisioningState) {
    /** Accepted */
    KnownEncryptionScopeProvisioningState["Accepted"] = "Accepted";
    /** Creating */
    KnownEncryptionScopeProvisioningState["Creating"] = "Creating";
    /** Deleting */
    KnownEncryptionScopeProvisioningState["Deleting"] = "Deleting";
    /** Moving */
    KnownEncryptionScopeProvisioningState["Moving"] = "Moving";
    /** Failed */
    KnownEncryptionScopeProvisioningState["Failed"] = "Failed";
    /** Succeeded */
    KnownEncryptionScopeProvisioningState["Succeeded"] = "Succeeded";
    /** Canceled */
    KnownEncryptionScopeProvisioningState["Canceled"] = "Canceled";
})(KnownEncryptionScopeProvisioningState || (exports.KnownEncryptionScopeProvisioningState = KnownEncryptionScopeProvisioningState = {}));
/** Known values of {@link EncryptionScopeState} that the service accepts. */
var KnownEncryptionScopeState;
(function (KnownEncryptionScopeState) {
    /** Disabled */
    KnownEncryptionScopeState["Disabled"] = "Disabled";
    /** Enabled */
    KnownEncryptionScopeState["Enabled"] = "Enabled";
})(KnownEncryptionScopeState || (exports.KnownEncryptionScopeState = KnownEncryptionScopeState = {}));
/** Known values of {@link RaiPolicyType} that the service accepts. */
var KnownRaiPolicyType;
(function (KnownRaiPolicyType) {
    /** UserManaged */
    KnownRaiPolicyType["UserManaged"] = "UserManaged";
    /** SystemManaged */
    KnownRaiPolicyType["SystemManaged"] = "SystemManaged";
})(KnownRaiPolicyType || (exports.KnownRaiPolicyType = KnownRaiPolicyType = {}));
/** Known values of {@link RaiPolicyMode} that the service accepts. */
var KnownRaiPolicyMode;
(function (KnownRaiPolicyMode) {
    /** Default */
    KnownRaiPolicyMode["Default"] = "Default";
    /** Deferred */
    KnownRaiPolicyMode["Deferred"] = "Deferred";
    /** Blocking */
    KnownRaiPolicyMode["Blocking"] = "Blocking";
    /** AsynchronousFilter */
    KnownRaiPolicyMode["AsynchronousFilter"] = "Asynchronous_filter";
})(KnownRaiPolicyMode || (exports.KnownRaiPolicyMode = KnownRaiPolicyMode = {}));
/** Known values of {@link ContentLevel} that the service accepts. */
var KnownContentLevel;
(function (KnownContentLevel) {
    /** Low */
    KnownContentLevel["Low"] = "Low";
    /** Medium */
    KnownContentLevel["Medium"] = "Medium";
    /** High */
    KnownContentLevel["High"] = "High";
})(KnownContentLevel || (exports.KnownContentLevel = KnownContentLevel = {}));
/** Known values of {@link RaiPolicyContentSource} that the service accepts. */
var KnownRaiPolicyContentSource;
(function (KnownRaiPolicyContentSource) {
    /** Prompt */
    KnownRaiPolicyContentSource["Prompt"] = "Prompt";
    /** Completion */
    KnownRaiPolicyContentSource["Completion"] = "Completion";
})(KnownRaiPolicyContentSource || (exports.KnownRaiPolicyContentSource = KnownRaiPolicyContentSource = {}));
/** Known values of {@link NspAccessRuleDirection} that the service accepts. */
var KnownNspAccessRuleDirection;
(function (KnownNspAccessRuleDirection) {
    /** Inbound */
    KnownNspAccessRuleDirection["Inbound"] = "Inbound";
    /** Outbound */
    KnownNspAccessRuleDirection["Outbound"] = "Outbound";
})(KnownNspAccessRuleDirection || (exports.KnownNspAccessRuleDirection = KnownNspAccessRuleDirection = {}));
/** Known values of {@link DefenderForAISettingState} that the service accepts. */
var KnownDefenderForAISettingState;
(function (KnownDefenderForAISettingState) {
    /** Disabled */
    KnownDefenderForAISettingState["Disabled"] = "Disabled";
    /** Enabled */
    KnownDefenderForAISettingState["Enabled"] = "Enabled";
})(KnownDefenderForAISettingState || (exports.KnownDefenderForAISettingState = KnownDefenderForAISettingState = {}));
/** Known values of {@link ConnectionAuthType} that the service accepts. */
var KnownConnectionAuthType;
(function (KnownConnectionAuthType) {
    /** PAT */
    KnownConnectionAuthType["PAT"] = "PAT";
    /** ManagedIdentity */
    KnownConnectionAuthType["ManagedIdentity"] = "ManagedIdentity";
    /** UsernamePassword */
    KnownConnectionAuthType["UsernamePassword"] = "UsernamePassword";
    /** None */
    KnownConnectionAuthType["None"] = "None";
    /** SAS */
    KnownConnectionAuthType["SAS"] = "SAS";
    /** AccountKey */
    KnownConnectionAuthType["AccountKey"] = "AccountKey";
    /** ServicePrincipal */
    KnownConnectionAuthType["ServicePrincipal"] = "ServicePrincipal";
    /** AccessKey */
    KnownConnectionAuthType["AccessKey"] = "AccessKey";
    /** ApiKey */
    KnownConnectionAuthType["ApiKey"] = "ApiKey";
    /** CustomKeys */
    KnownConnectionAuthType["CustomKeys"] = "CustomKeys";
    /** OAuth2 */
    KnownConnectionAuthType["OAuth2"] = "OAuth2";
    /** AAD */
    KnownConnectionAuthType["AAD"] = "AAD";
})(KnownConnectionAuthType || (exports.KnownConnectionAuthType = KnownConnectionAuthType = {}));
/** Known values of {@link ConnectionCategory} that the service accepts. */
var KnownConnectionCategory;
(function (KnownConnectionCategory) {
    /** PythonFeed */
    KnownConnectionCategory["PythonFeed"] = "PythonFeed";
    /** ContainerRegistry */
    KnownConnectionCategory["ContainerRegistry"] = "ContainerRegistry";
    /** Git */
    KnownConnectionCategory["Git"] = "Git";
    /** S3 */
    KnownConnectionCategory["S3"] = "S3";
    /** Snowflake */
    KnownConnectionCategory["Snowflake"] = "Snowflake";
    /** AzureSqlDb */
    KnownConnectionCategory["AzureSqlDb"] = "AzureSqlDb";
    /** AzureSynapseAnalytics */
    KnownConnectionCategory["AzureSynapseAnalytics"] = "AzureSynapseAnalytics";
    /** AzureMySqlDb */
    KnownConnectionCategory["AzureMySqlDb"] = "AzureMySqlDb";
    /** AzurePostgresDb */
    KnownConnectionCategory["AzurePostgresDb"] = "AzurePostgresDb";
    /** AdlsGen2 */
    KnownConnectionCategory["AdlsGen2"] = "ADLSGen2";
    /** Redis */
    KnownConnectionCategory["Redis"] = "Redis";
    /** ApiKey */
    KnownConnectionCategory["ApiKey"] = "ApiKey";
    /** AzureOpenAI */
    KnownConnectionCategory["AzureOpenAI"] = "AzureOpenAI";
    /** AIServices */
    KnownConnectionCategory["AIServices"] = "AIServices";
    /** CognitiveSearch */
    KnownConnectionCategory["CognitiveSearch"] = "CognitiveSearch";
    /** CognitiveService */
    KnownConnectionCategory["CognitiveService"] = "CognitiveService";
    /** CustomKeys */
    KnownConnectionCategory["CustomKeys"] = "CustomKeys";
    /** AzureBlob */
    KnownConnectionCategory["AzureBlob"] = "AzureBlob";
    /** AzureOneLake */
    KnownConnectionCategory["AzureOneLake"] = "AzureOneLake";
    /** CosmosDb */
    KnownConnectionCategory["CosmosDb"] = "CosmosDb";
    /** CosmosDbMongoDbApi */
    KnownConnectionCategory["CosmosDbMongoDbApi"] = "CosmosDbMongoDbApi";
    /** AzureDataExplorer */
    KnownConnectionCategory["AzureDataExplorer"] = "AzureDataExplorer";
    /** AzureMariaDb */
    KnownConnectionCategory["AzureMariaDb"] = "AzureMariaDb";
    /** AzureDatabricksDeltaLake */
    KnownConnectionCategory["AzureDatabricksDeltaLake"] = "AzureDatabricksDeltaLake";
    /** AzureSqlMi */
    KnownConnectionCategory["AzureSqlMi"] = "AzureSqlMi";
    /** AzureTableStorage */
    KnownConnectionCategory["AzureTableStorage"] = "AzureTableStorage";
    /** AmazonRdsForOracle */
    KnownConnectionCategory["AmazonRdsForOracle"] = "AmazonRdsForOracle";
    /** AmazonRdsForSqlServer */
    KnownConnectionCategory["AmazonRdsForSqlServer"] = "AmazonRdsForSqlServer";
    /** AmazonRedshift */
    KnownConnectionCategory["AmazonRedshift"] = "AmazonRedshift";
    /** Db2 */
    KnownConnectionCategory["Db2"] = "Db2";
    /** Drill */
    KnownConnectionCategory["Drill"] = "Drill";
    /** GoogleBigQuery */
    KnownConnectionCategory["GoogleBigQuery"] = "GoogleBigQuery";
    /** Greenplum */
    KnownConnectionCategory["Greenplum"] = "Greenplum";
    /** Hbase */
    KnownConnectionCategory["Hbase"] = "Hbase";
    /** Hive */
    KnownConnectionCategory["Hive"] = "Hive";
    /** Impala */
    KnownConnectionCategory["Impala"] = "Impala";
    /** Informix */
    KnownConnectionCategory["Informix"] = "Informix";
    /** MariaDb */
    KnownConnectionCategory["MariaDb"] = "MariaDb";
    /** MicrosoftAccess */
    KnownConnectionCategory["MicrosoftAccess"] = "MicrosoftAccess";
    /** MySql */
    KnownConnectionCategory["MySql"] = "MySql";
    /** Netezza */
    KnownConnectionCategory["Netezza"] = "Netezza";
    /** Oracle */
    KnownConnectionCategory["Oracle"] = "Oracle";
    /** Phoenix */
    KnownConnectionCategory["Phoenix"] = "Phoenix";
    /** PostgreSql */
    KnownConnectionCategory["PostgreSql"] = "PostgreSql";
    /** Presto */
    KnownConnectionCategory["Presto"] = "Presto";
    /** SapOpenHub */
    KnownConnectionCategory["SapOpenHub"] = "SapOpenHub";
    /** SapBw */
    KnownConnectionCategory["SapBw"] = "SapBw";
    /** SapHana */
    KnownConnectionCategory["SapHana"] = "SapHana";
    /** SapTable */
    KnownConnectionCategory["SapTable"] = "SapTable";
    /** Spark */
    KnownConnectionCategory["Spark"] = "Spark";
    /** SqlServer */
    KnownConnectionCategory["SqlServer"] = "SqlServer";
    /** Sybase */
    KnownConnectionCategory["Sybase"] = "Sybase";
    /** Teradata */
    KnownConnectionCategory["Teradata"] = "Teradata";
    /** Vertica */
    KnownConnectionCategory["Vertica"] = "Vertica";
    /** Pinecone */
    KnownConnectionCategory["Pinecone"] = "Pinecone";
    /** Cassandra */
    KnownConnectionCategory["Cassandra"] = "Cassandra";
    /** Couchbase */
    KnownConnectionCategory["Couchbase"] = "Couchbase";
    /** MongoDbV2 */
    KnownConnectionCategory["MongoDbV2"] = "MongoDbV2";
    /** MongoDbAtlas */
    KnownConnectionCategory["MongoDbAtlas"] = "MongoDbAtlas";
    /** AmazonS3Compatible */
    KnownConnectionCategory["AmazonS3Compatible"] = "AmazonS3Compatible";
    /** FileServer */
    KnownConnectionCategory["FileServer"] = "FileServer";
    /** FtpServer */
    KnownConnectionCategory["FtpServer"] = "FtpServer";
    /** GoogleCloudStorage */
    KnownConnectionCategory["GoogleCloudStorage"] = "GoogleCloudStorage";
    /** Hdfs */
    KnownConnectionCategory["Hdfs"] = "Hdfs";
    /** OracleCloudStorage */
    KnownConnectionCategory["OracleCloudStorage"] = "OracleCloudStorage";
    /** Sftp */
    KnownConnectionCategory["Sftp"] = "Sftp";
    /** GenericHttp */
    KnownConnectionCategory["GenericHttp"] = "GenericHttp";
    /** ODataRest */
    KnownConnectionCategory["ODataRest"] = "ODataRest";
    /** Odbc */
    KnownConnectionCategory["Odbc"] = "Odbc";
    /** GenericRest */
    KnownConnectionCategory["GenericRest"] = "GenericRest";
    /** AmazonMws */
    KnownConnectionCategory["AmazonMws"] = "AmazonMws";
    /** Concur */
    KnownConnectionCategory["Concur"] = "Concur";
    /** Dynamics */
    KnownConnectionCategory["Dynamics"] = "Dynamics";
    /** DynamicsAx */
    KnownConnectionCategory["DynamicsAx"] = "DynamicsAx";
    /** DynamicsCrm */
    KnownConnectionCategory["DynamicsCrm"] = "DynamicsCrm";
    /** GoogleAdWords */
    KnownConnectionCategory["GoogleAdWords"] = "GoogleAdWords";
    /** Hubspot */
    KnownConnectionCategory["Hubspot"] = "Hubspot";
    /** Jira */
    KnownConnectionCategory["Jira"] = "Jira";
    /** Magento */
    KnownConnectionCategory["Magento"] = "Magento";
    /** Marketo */
    KnownConnectionCategory["Marketo"] = "Marketo";
    /** Office365 */
    KnownConnectionCategory["Office365"] = "Office365";
    /** Eloqua */
    KnownConnectionCategory["Eloqua"] = "Eloqua";
    /** Responsys */
    KnownConnectionCategory["Responsys"] = "Responsys";
    /** OracleServiceCloud */
    KnownConnectionCategory["OracleServiceCloud"] = "OracleServiceCloud";
    /** PayPal */
    KnownConnectionCategory["PayPal"] = "PayPal";
    /** QuickBooks */
    KnownConnectionCategory["QuickBooks"] = "QuickBooks";
    /** Salesforce */
    KnownConnectionCategory["Salesforce"] = "Salesforce";
    /** SalesforceServiceCloud */
    KnownConnectionCategory["SalesforceServiceCloud"] = "SalesforceServiceCloud";
    /** SalesforceMarketingCloud */
    KnownConnectionCategory["SalesforceMarketingCloud"] = "SalesforceMarketingCloud";
    /** SapCloudForCustomer */
    KnownConnectionCategory["SapCloudForCustomer"] = "SapCloudForCustomer";
    /** SapEcc */
    KnownConnectionCategory["SapEcc"] = "SapEcc";
    /** ServiceNow */
    KnownConnectionCategory["ServiceNow"] = "ServiceNow";
    /** SharePointOnlineList */
    KnownConnectionCategory["SharePointOnlineList"] = "SharePointOnlineList";
    /** Shopify */
    KnownConnectionCategory["Shopify"] = "Shopify";
    /** Square */
    KnownConnectionCategory["Square"] = "Square";
    /** WebTable */
    KnownConnectionCategory["WebTable"] = "WebTable";
    /** Xero */
    KnownConnectionCategory["Xero"] = "Xero";
    /** Zoho */
    KnownConnectionCategory["Zoho"] = "Zoho";
    /** GenericContainerRegistry */
    KnownConnectionCategory["GenericContainerRegistry"] = "GenericContainerRegistry";
    /** Elasticsearch */
    KnownConnectionCategory["Elasticsearch"] = "Elasticsearch";
    /** OpenAI */
    KnownConnectionCategory["OpenAI"] = "OpenAI";
    /** Serp */
    KnownConnectionCategory["Serp"] = "Serp";
    /** BingLLMSearch */
    KnownConnectionCategory["BingLLMSearch"] = "BingLLMSearch";
    /** Serverless */
    KnownConnectionCategory["Serverless"] = "Serverless";
    /** ManagedOnlineEndpoint */
    KnownConnectionCategory["ManagedOnlineEndpoint"] = "ManagedOnlineEndpoint";
})(KnownConnectionCategory || (exports.KnownConnectionCategory = KnownConnectionCategory = {}));
/** Known values of {@link ConnectionGroup} that the service accepts. */
var KnownConnectionGroup;
(function (KnownConnectionGroup) {
    /** Azure */
    KnownConnectionGroup["Azure"] = "Azure";
    /** AzureAI */
    KnownConnectionGroup["AzureAI"] = "AzureAI";
    /** Database */
    KnownConnectionGroup["Database"] = "Database";
    /** NoSQL */
    KnownConnectionGroup["NoSQL"] = "NoSQL";
    /** File */
    KnownConnectionGroup["File"] = "File";
    /** GenericProtocol */
    KnownConnectionGroup["GenericProtocol"] = "GenericProtocol";
    /** ServicesAndApps */
    KnownConnectionGroup["ServicesAndApps"] = "ServicesAndApps";
})(KnownConnectionGroup || (exports.KnownConnectionGroup = KnownConnectionGroup = {}));
/** Known values of {@link ManagedPERequirement} that the service accepts. */
var KnownManagedPERequirement;
(function (KnownManagedPERequirement) {
    /** Required */
    KnownManagedPERequirement["Required"] = "Required";
    /** NotRequired */
    KnownManagedPERequirement["NotRequired"] = "NotRequired";
    /** NotApplicable */
    KnownManagedPERequirement["NotApplicable"] = "NotApplicable";
})(KnownManagedPERequirement || (exports.KnownManagedPERequirement = KnownManagedPERequirement = {}));
/** Known values of {@link ManagedPEStatus} that the service accepts. */
var KnownManagedPEStatus;
(function (KnownManagedPEStatus) {
    /** Inactive */
    KnownManagedPEStatus["Inactive"] = "Inactive";
    /** Active */
    KnownManagedPEStatus["Active"] = "Active";
    /** NotApplicable */
    KnownManagedPEStatus["NotApplicable"] = "NotApplicable";
})(KnownManagedPEStatus || (exports.KnownManagedPEStatus = KnownManagedPEStatus = {}));
/** Known values of {@link CapabilityHostKind} that the service accepts. */
var KnownCapabilityHostKind;
(function (KnownCapabilityHostKind) {
    /** Agents */
    KnownCapabilityHostKind["Agents"] = "Agents";
})(KnownCapabilityHostKind || (exports.KnownCapabilityHostKind = KnownCapabilityHostKind = {}));
/** Known values of {@link CapabilityHostProvisioningState} that the service accepts. */
var KnownCapabilityHostProvisioningState;
(function (KnownCapabilityHostProvisioningState) {
    /** Succeeded */
    KnownCapabilityHostProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownCapabilityHostProvisioningState["Failed"] = "Failed";
    /** Canceled */
    KnownCapabilityHostProvisioningState["Canceled"] = "Canceled";
    /** Creating */
    KnownCapabilityHostProvisioningState["Creating"] = "Creating";
    /** Updating */
    KnownCapabilityHostProvisioningState["Updating"] = "Updating";
    /** Deleting */
    KnownCapabilityHostProvisioningState["Deleting"] = "Deleting";
})(KnownCapabilityHostProvisioningState || (exports.KnownCapabilityHostProvisioningState = KnownCapabilityHostProvisioningState = {}));
