// @azure/arm-cognitiveservices/dist/commonjs/operations/projectCapabilityHosts.js
"use strict";
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectCapabilityHostsImpl = void 0;
const tslib_1 = require("tslib");
const coreClient = tslib_1.__importStar(require("@azure/core-client"));
const Mappers = tslib_1.__importStar(require("../models/mappers.js"));
const Parameters = tslib_1.__importStar(require("../models/parameters.js"));
const core_lro_1 = require("@azure/core-lro");
const lroImpl_js_1 = require("../lroImpl.js");
/** Class containing ProjectCapabilityHosts operations. */
class ProjectCapabilityHostsImpl {
    client;
    /**
     * Initialize a new instance of the class ProjectCapabilityHosts class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * Delete project capabilityHost.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param accountName The name of Cognitive Services account.
     * @param projectName The name of Cognitive Services account's project.
     * @param capabilityHostName The name of the capability host associated with the Cognitive Services
     *                           Resource
     * @param options The options parameters.
     */
    async beginDelete(resourceGroupName, accountName, projectName, capabilityHostName, options) {
        const directSendOperation = async (args, spec) => {
            return this.client.sendOperationRequest(args, spec);
        };
        const sendOperationFn = async (args, spec) => {
            let currentRawResponse = undefined;
            const providedCallback = args.options?.onResponse;
            const callback = (rawResponse, flatResponse) => {
                currentRawResponse = rawResponse;
                providedCallback?.(rawResponse, flatResponse);
            };
            const updatedArgs = {
                ...args,
                options: {
                    ...args.options,
                    onResponse: callback,
                },
            };
            const flatResponse = await directSendOperation(updatedArgs, spec);
            return {
                flatResponse,
                rawResponse: {
                    statusCode: currentRawResponse.status,
                    body: currentRawResponse.parsedBody,
                    headers: currentRawResponse.headers.toJSON(),
                },
            };
        };
        const lro = (0, lroImpl_js_1.createLroSpec)({
            sendOperationFn,
            args: {
                resourceGroupName,
                accountName,
                projectName,
                capabilityHostName,
                options,
            },
            spec: deleteOperationSpec,
        });
        const poller = await (0, core_lro_1.createHttpPoller)(lro, {
            restoreFrom: options?.resumeFrom,
            intervalInMs: options?.updateIntervalInMs,
            resourceLocationConfig: "location",
        });
        await poller.poll();
        return poller;
    }
    /**
     * Delete project capabilityHost.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param accountName The name of Cognitive Services account.
     * @param projectName The name of Cognitive Services account's project.
     * @param capabilityHostName The name of the capability host associated with the Cognitive Services
     *                           Resource
     * @param options The options parameters.
     */
    async beginDeleteAndWait(resourceGroupName, accountName, projectName, capabilityHostName, options) {
        const poller = await this.beginDelete(resourceGroupName, accountName, projectName, capabilityHostName, options);
        return poller.pollUntilDone();
    }
    /**
     * Get project capabilityHost.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param accountName The name of Cognitive Services account.
     * @param projectName The name of Cognitive Services account's project.
     * @param capabilityHostName The name of the capability host associated with the Cognitive Services
     *                           Resource
     * @param options The options parameters.
     */
    get(resourceGroupName, accountName, projectName, capabilityHostName, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            accountName,
            projectName,
            capabilityHostName,
            options,
        }, getOperationSpec);
    }
    /**
     * Create or update project capabilityHost.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param accountName The name of Cognitive Services account.
     * @param projectName The name of Cognitive Services account's project.
     * @param capabilityHostName The name of the capability host associated with the Cognitive Services
     *                           Resource
     * @param capabilityHost CapabilityHost definition.
     * @param options The options parameters.
     */
    async beginCreateOrUpdate(resourceGroupName, accountName, projectName, capabilityHostName, capabilityHost, options) {
        const directSendOperation = async (args, spec) => {
            return this.client.sendOperationRequest(args, spec);
        };
        const sendOperationFn = async (args, spec) => {
            let currentRawResponse = undefined;
            const providedCallback = args.options?.onResponse;
            const callback = (rawResponse, flatResponse) => {
                currentRawResponse = rawResponse;
                providedCallback?.(rawResponse, flatResponse);
            };
            const updatedArgs = {
                ...args,
                options: {
                    ...args.options,
                    onResponse: callback,
                },
            };
            const flatResponse = await directSendOperation(updatedArgs, spec);
            return {
                flatResponse,
                rawResponse: {
                    statusCode: currentRawResponse.status,
                    body: currentRawResponse.parsedBody,
                    headers: currentRawResponse.headers.toJSON(),
                },
            };
        };
        const lro = (0, lroImpl_js_1.createLroSpec)({
            sendOperationFn,
            args: {
                resourceGroupName,
                accountName,
                projectName,
                capabilityHostName,
                capabilityHost,
                options,
            },
            spec: createOrUpdateOperationSpec,
        });
        const poller = await (0, core_lro_1.createHttpPoller)(lro, {
            restoreFrom: options?.resumeFrom,
            intervalInMs: options?.updateIntervalInMs,
            resourceLocationConfig: "original-uri",
        });
        await poller.poll();
        return poller;
    }
    /**
     * Create or update project capabilityHost.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param accountName The name of Cognitive Services account.
     * @param projectName The name of Cognitive Services account's project.
     * @param capabilityHostName The name of the capability host associated with the Cognitive Services
     *                           Resource
     * @param capabilityHost CapabilityHost definition.
     * @param options The options parameters.
     */
    async beginCreateOrUpdateAndWait(resourceGroupName, accountName, projectName, capabilityHostName, capabilityHost, options) {
        const poller = await this.beginCreateOrUpdate(resourceGroupName, accountName, projectName, capabilityHostName, capabilityHost, options);
        return poller.pollUntilDone();
    }
}
exports.ProjectCapabilityHostsImpl = ProjectCapabilityHostsImpl;
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const deleteOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CognitiveServices/accounts/{accountName}/projects/{projectName}/capabilityHosts/{capabilityHostName}",
    httpMethod: "DELETE",
    responses: {
        200: {
            headersMapper: Mappers.ProjectCapabilityHostsDeleteHeaders,
        },
        201: {
            headersMapper: Mappers.ProjectCapabilityHostsDeleteHeaders,
        },
        202: {
            headersMapper: Mappers.ProjectCapabilityHostsDeleteHeaders,
        },
        204: {
            headersMapper: Mappers.ProjectCapabilityHostsDeleteHeaders,
        },
        default: {
            bodyMapper: Mappers.ErrorResponse,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.accountName,
        Parameters.subscriptionId,
        Parameters.projectName,
        Parameters.capabilityHostName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const getOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CognitiveServices/accounts/{accountName}/projects/{projectName}/capabilityHosts/{capabilityHostName}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.CapabilityHost,
        },
        default: {
            bodyMapper: Mappers.ErrorResponse,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.accountName,
        Parameters.subscriptionId,
        Parameters.projectName,
        Parameters.capabilityHostName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const createOrUpdateOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CognitiveServices/accounts/{accountName}/projects/{projectName}/capabilityHosts/{capabilityHostName}",
    httpMethod: "PUT",
    responses: {
        200: {
            bodyMapper: Mappers.CapabilityHost,
        },
        201: {
            bodyMapper: Mappers.CapabilityHost,
        },
        202: {
            bodyMapper: Mappers.CapabilityHost,
        },
        204: {
            bodyMapper: Mappers.CapabilityHost,
        },
        default: {
            bodyMapper: Mappers.ErrorResponse,
        },
    },
    requestBody: Parameters.capabilityHost,
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.accountName,
        Parameters.subscriptionId,
        Parameters.projectName,
        Parameters.capabilityHostName,
    ],
    headerParameters: [Parameters.contentType, Parameters.accept],
    mediaType: "json",
    serializer,
};
