// codetocloudmcp/out/toolsImplementations/azurePricing/params.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzurePriceType = exports.AzureServiceFamily = void 0;
// Documented serviceFamily values from Azure Retail Prices API
var AzureServiceFamily;
(function (AzureServiceFamily) {
    AzureServiceFamily["Analytics"] = "Analytics";
    AzureServiceFamily["Compute"] = "Compute";
    AzureServiceFamily["Containers"] = "Containers";
    AzureServiceFamily["Databases"] = "Databases";
    AzureServiceFamily["DeveloperTools"] = "Developer Tools";
    AzureServiceFamily["Integration"] = "Integration";
    AzureServiceFamily["InternetOfThings"] = "Internet of Things";
    AzureServiceFamily["ManagementAndGovernance"] = "Management and Governance";
    AzureServiceFamily["Networking"] = "Networking";
    AzureServiceFamily["Security"] = "Security";
    AzureServiceFamily["Storage"] = "Storage";
    AzureServiceFamily["Web"] = "Web";
    AzureServiceFamily["AzureArc"] = "Azure Arc";
    AzureServiceFamily["AzureCommunicationServices"] = "Azure Communication Services";
    AzureServiceFamily["AzureSecurity"] = "Azure Security";
    AzureServiceFamily["AzureStack"] = "Azure Stack";
    AzureServiceFamily["Data"] = "Data";
    AzureServiceFamily["Dynamics"] = "Dynamics";
    AzureServiceFamily["Gaming"] = "Gaming";
    AzureServiceFamily["MicrosoftSyntex"] = "Microsoft Syntex";
    AzureServiceFamily["MixedReality"] = "Mixed Reality";
    AzureServiceFamily["Other"] = "Other";
    AzureServiceFamily["PowerPlatform"] = "Power Platform";
    AzureServiceFamily["QuantumComputing"] = "Quantum Computing";
    AzureServiceFamily["Telecommunications"] = "Telecommunications";
    AzureServiceFamily["WindowsVirtualDesktop"] = "Windows Virtual Desktop";
})(AzureServiceFamily || (exports.AzureServiceFamily = AzureServiceFamily = {}));
// Documented priceType values
var AzurePriceType;
(function (AzurePriceType) {
    AzurePriceType["Consumption"] = "Consumption";
    AzurePriceType["Reservation"] = "Reservation";
    AzurePriceType["DevTestConsumption"] = "DevTestConsumption";
})(AzurePriceType || (exports.AzurePriceType = AzurePriceType = {}));
