// containerization-assist-mcp/dist/src/cli/policy-simulate.js
#!/usr/bin/env node
/**
 * Policy Simulation Tool
 *
 * Simulates policy impact by running tools with and without custom policies,
 * showing the differences in:
 * - Generation configuration
 * - Knowledge filtering
 * - Template injection
 * - Validation results
 *
 * Usage:
 *   npx tsx src/cli/policy-simulate.ts \
 *     --policy policies.user.examples/production-ready/cost-control-by-tier.rego \
 *     --tool generate-dockerfile \
 *     --input '{"language": "node", "environment": "production", "teamTier": "starter"}'
 */
import path from 'node:path';
import { createLogger } from '../lib/logger.js';
import { loadRegoPolicy } from '../config/policy-rego.js';
import generateDockerfileTool from '../tools/generate-dockerfile/tool.js';
import generateK8sManifestsTool from '../tools/generate-k8s-manifests/tool.js';
const logger = createLogger({ level: 'error' });
async function simulatePolicy(options) {
    const { policyPath, tool, input } = options;
    console.log('\n🔬 Policy Simulation\n');
    console.log(`Policy: ${policyPath}`);
    console.log(`Tool: ${tool}`);
    console.log(`Input: ${JSON.stringify(input, null, 2)}\n`);
    // Load the custom policy
    const policyResult = await loadRegoPolicy(path.resolve(policyPath), logger);
    if (!policyResult.ok) {
        throw new Error(`Failed to load policy: ${policyResult.error}`);
    }
    const policy = policyResult.value;
    // Get tool handler
    const toolHandler = tool === 'generate-dockerfile'
        ? generateDockerfileTool
        : generateK8sManifestsTool;
    // Run WITHOUT policy
    console.log('📊 Running WITHOUT custom policy...\n');
    const contextWithout = {
        logger,
        signal: undefined,
        progress: undefined,
        // policy is omitted (no custom policy)
        queryConfig: async () => null,
    };
    const outputWithout = await toolHandler.handler(input, contextWithout);
    // Query policy data (without applying it)
    const configWithout = null;
    const knowledgeWithout = [];
    const templatesWithout = [];
    const validationWithout = [];
    // Run WITH policy
    console.log('📊 Running WITH custom policy...\n');
    const contextWith = {
        logger,
        signal: undefined,
        progress: undefined,
        policy,
        queryConfig: async (packageName, policyInput) => {
            return policy.queryConfig(packageName, policyInput);
        },
    };
    const outputWith = await toolHandler.handler(input, contextWith);
    // Query policy data
    const policyInput = { ...input, tool };
    const configWith = await policy.queryConfig('generation_config', policyInput).catch(() => null);
    // For now, we'll extract policy impact from the output itself
    // The most important thing is showing how the output differs
    const knowledgeWith = [];
    const templatesWith = [];
    const validationWith = [];
    // Calculate differences
    const differences = {
        configChanged: configWith !== null,
        knowledgeFiltered: Array.isArray(knowledgeWith) ? knowledgeWith.length : 0,
        templatesInjected: Array.isArray(templatesWith) ? templatesWith.length : 0,
        validationRulesAdded: Array.isArray(validationWith) ? validationWith.length : 0,
        outputDiffers: JSON.stringify(outputWithout) !== JSON.stringify(outputWith),
    };
    return {
        withoutPolicy: {
            generationConfig: configWithout,
            knowledgeFiltering: knowledgeWithout,
            templates: templatesWithout,
            validationRules: validationWithout,
            output: outputWithout.ok ? outputWithout.value : outputWithout.error,
        },
        withPolicy: {
            generationConfig: configWith,
            knowledgeFiltering: knowledgeWith || [],
            templates: templatesWith || [],
            validationRules: validationWith || [],
            output: outputWith.ok ? outputWith.value : outputWith.error,
        },
        differences,
    };
}
function printSimulationResults(result) {
    console.log(`\n${'='.repeat(80)}`);
    console.log('📈 SIMULATION RESULTS');
    console.log(`${'='.repeat(80)}\n`);
    const { differences, withoutPolicy, withPolicy } = result;
    // Summary
    console.log('📊 Impact Summary:');
    console.log(`  • Generation Config: ${differences.configChanged ? '✅ Modified' : '❌ No change'}`);
    console.log(`  • Knowledge Filtered: ${differences.knowledgeFiltered} rules`);
    console.log(`  • Templates Injected: ${differences.templatesInjected} templates`);
    console.log(`  • Validation Rules: ${differences.validationRulesAdded} rules added`);
    console.log(`  • Output Changed: ${differences.outputDiffers ? '✅ Yes' : '❌ No'}\n`);
    // Generation Config
    if (differences.configChanged) {
        console.log('⚙️  Generation Configuration (Applied):');
        console.log(JSON.stringify(withPolicy.generationConfig, null, 2));
        console.log('');
    }
    // Knowledge Filtering
    if (differences.knowledgeFiltered > 0) {
        console.log(`🔍 Knowledge Filtering (${differences.knowledgeFiltered} rules):`);
        withPolicy.knowledgeFiltering.forEach((filter, i) => {
            console.log(`  ${i + 1}. Action: ${filter.action}`);
            if (filter.pattern)
                console.log(`     Pattern: ${filter.pattern}`);
            if (filter.tags)
                console.log(`     Tags: ${filter.tags.join(', ')}`);
            if (filter.reason)
                console.log(`     Reason: ${filter.reason}`);
            console.log('');
        });
    }
    // Template Injection
    if (differences.templatesInjected > 0) {
        console.log(`📝 Templates Injected (${differences.templatesInjected} templates):`);
        withPolicy.templates.forEach((template, i) => {
            console.log(`  ${i + 1}. ID: ${template.id}`);
            console.log(`     Category: ${template.category}`);
            console.log(`     Recommendation: ${template.recommendation}`);
            if (template.priority)
                console.log(`     Priority: ${template.priority}`);
            console.log('     Code Snippet:');
            console.log(template.code_snippet.split('\n').map((line) => `       ${line}`).join('\n'));
            console.log('');
        });
    }
    // Validation Rules
    if (differences.validationRulesAdded > 0) {
        console.log(`✅ Validation Rules (${differences.validationRulesAdded} rules):`);
        withPolicy.validationRules.forEach((rule, i) => {
            const levelEmoji = rule.level === 'error' ? '❌' : rule.level === 'warning' ? '⚠️' : 'ℹ️';
            console.log(`  ${i + 1}. ${levelEmoji} [${rule.level.toUpperCase()}] ${rule.message}`);
            if (rule.suggestion)
                console.log(`     💡 Suggestion: ${rule.suggestion}`);
            console.log('');
        });
    }
    // Output Comparison
    if (differences.outputDiffers) {
        console.log('📦 Output Comparison:\n');
        console.log('  WITHOUT Policy:');
        console.log(`  ${'-'.repeat(76)}`);
        if (withoutPolicy.output && typeof withoutPolicy.output === 'object' && 'summary' in withoutPolicy.output) {
            console.log(`  Summary: ${withoutPolicy.output.summary}`);
            if ('recommendations' in withoutPolicy.output) {
                const recs = withoutPolicy.output.recommendations;
                const totalRecs = (recs.securityConsiderations?.length || 0) + (recs.bestPractices?.length || 0);
                console.log(`  Recommendations: ${totalRecs} total`);
            }
        }
        else {
            console.log(`  ${JSON.stringify(withoutPolicy.output, null, 2).split('\n').map(line => `  ${line}`).join('\n')}`);
        }
        console.log('\n  WITH Policy:');
        console.log(`  ${'-'.repeat(76)}`);
        if (withPolicy.output && typeof withPolicy.output === 'object' && 'summary' in withPolicy.output) {
            console.log(`  Summary: ${withPolicy.output.summary}`);
            if ('recommendations' in withPolicy.output) {
                const recs = withPolicy.output.recommendations;
                const totalRecs = (recs.securityConsiderations?.length || 0) + (recs.bestPractices?.length || 0);
                console.log(`  Recommendations: ${totalRecs} total`);
                // Show policy-driven recommendations
                const policyDriven = [
                    ...(recs.securityConsiderations || []),
                    ...(recs.bestPractices || []),
                    ...(recs.resourceManagement || []),
                ].filter((r) => r.policyDriven);
                if (policyDriven.length > 0) {
                    console.log(`  Policy-Driven: ${policyDriven.length} recommendations`);
                    policyDriven.forEach((rec) => {
                        console.log(`    • ${rec.id}: ${rec.recommendation}`);
                    });
                }
            }
        }
        else {
            console.log(`  ${JSON.stringify(withPolicy.output, null, 2).split('\n').map(line => `  ${line}`).join('\n')}`);
        }
        console.log('');
    }
    console.log('='.repeat(80));
    console.log('✅ Simulation Complete\n');
}
// CLI entrypoint
async function main() {
    const args = process.argv.slice(2);
    if (args.length === 0 || args.includes('--help')) {
        console.log(`
Policy Simulation Tool

Simulates the impact of a custom policy by comparing tool execution with and without the policy.

Usage:
  npx tsx src/cli/policy-simulate.ts \\
    --policy <path-to-policy.rego> \\
    --tool <tool-name> \\
    --input '<json-input>'

Options:
  --policy   Path to Rego policy file
  --tool     Tool to simulate (generate-dockerfile or generate-k8s-manifests)
  --input    JSON input for the tool (as string)
  --help     Show this help message

Examples:
  # Simulate cost control policy for starter tier
  npx tsx src/cli/policy-simulate.ts \\
    --policy policies.user.examples/production-ready/cost-control-by-tier.rego \\
    --tool generate-dockerfile \\
    --input '{"language": "node", "environment": "production", "teamTier": "starter"}'

  # Simulate security-first policy
  npx tsx src/cli/policy-simulate.ts \\
    --policy policies.user.examples/production-ready/security-first-organization.rego \\
    --tool generate-k8s-manifests \\
    --input '{"name": "secure-app", "language": "java", "environment": "production"}'
`);
        process.exit(0);
    }
    // Parse arguments
    const policyIdx = args.indexOf('--policy');
    const toolIdx = args.indexOf('--tool');
    const inputIdx = args.indexOf('--input');
    if (policyIdx === -1 || toolIdx === -1 || inputIdx === -1) {
        console.error('❌ Error: Missing required arguments');
        console.error('Run with --help for usage information');
        process.exit(1);
    }
    // Validate that required parameters are provided
    const policyPath = args[policyIdx + 1];
    const tool = args[toolIdx + 1];
    const inputStr = args[inputIdx + 1];
    if (!policyPath) {
        console.error('❌ Error: --policy is required');
        process.exit(1);
    }
    if (!tool) {
        console.error('❌ Error: --tool is required');
        process.exit(1);
    }
    if (!inputStr) {
        console.error('❌ Error: --input is required');
        process.exit(1);
    }
    const options = {
        policyPath,
        tool: tool,
        input: JSON.parse(inputStr),
    };
    try {
        const result = await simulatePolicy(options);
        printSimulationResults(result);
    }
    catch (error) {
        console.error('❌ Simulation failed:', error);
        process.exit(1);
    }
}
// Run if executed directly (ESM compatible)
const isMainModule = import.meta.url === `file://${process.argv[1]}`;
if (isMainModule) {
    main().catch(console.error);
}
export { simulatePolicy };
