// codetocloudmcp/out/toolsImplementations/getAppTopology/generateMermaid.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateMermaidChart = generateMermaidChart;
const path = __importStar(require("path"));
const models_1 = require("../../common/models");
const project_1 = require("../../common/project");
var NodeShape;
(function (NodeShape) {
    NodeShape[NodeShape["Rectangle"] = 0] = "Rectangle";
    NodeShape[NodeShape["Circle"] = 1] = "Circle";
    NodeShape[NodeShape["RoundedRectangle"] = 2] = "RoundedRectangle";
    NodeShape[NodeShape["Cylinder"] = 3] = "Cylinder";
    NodeShape[NodeShape["Hexagon"] = 4] = "Hexagon";
})(NodeShape || (NodeShape = {}));
const getProperServiceType = (type) => {
    return (models_1.TargetAppServiceToComputeHost[type] ?? type).replace(/\s/g, '-').toLowerCase();
};
const createInternalName = (type, name) => {
    return `${type}_${name.replace(/\s/g, '-').toLowerCase()}`;
};
const createServiceName = (name, serviceType) => {
    if (serviceType === models_1.TargetAppServiceType.AzureKubernetesService) {
        return name + ' (Containerized Service)';
    }
    if (serviceType === models_1.TargetAppServiceType.AzureContainerApp) {
        return name + ' (Azure Container App)';
    }
    return name + ` (${models_1.TargetAppServiceToComputeHost[serviceType]})`;
};
const createComponentName = (internalName, name, nodeShape) => {
    // dict with key NodeShape and value int array
    const nodeShapeBrackets = {
        [NodeShape.Rectangle]: ["[", "]"],
        [NodeShape.Circle]: ["((", "))"],
        [NodeShape.RoundedRectangle]: ["(", ")"],
        [NodeShape.Cylinder]: ["[(", ")]"],
        [NodeShape.Hexagon]: ["{{", "}}"]
    };
    return `${ensureUrlFriendlyName(internalName)}${nodeShapeBrackets[nodeShape][0]}\"\`${name}\`"${nodeShapeBrackets[nodeShape][1]}`;
};
const aksClusterInternalName = 'akscluster';
const aksClusterName = 'Azure Kubernetes Service (AKS) Cluster';
const containerAppEnvInternalName = 'containerappenv';
const containerAppEnvName = 'Azure Container Apps (ACA) Environment';
var ArrowType;
(function (ArrowType) {
    ArrowType["SOLID"] = "-->";
    ArrowType["OPEN"] = "->";
    ArrowType["DOTTED"] = "-.->";
})(ArrowType || (ArrowType = {}));
const createRelationshipString = (sourceName, targetName, connectionDescription, arrowType) => {
    return `${ensureUrlFriendlyName(sourceName)} ${arrowType} |"${connectionDescription}"| ${ensureUrlFriendlyName(targetName)}`;
};
const ensureUrlFriendlyName = (name) => {
    name = name.replace(/\./g, '_'); // Replace dots with underscores
    return name.replace(/\s+/g, '_').replace(/[^a-zA-Z0-9_]/g, '');
};
function generateMermaidChart(workspaceFolder, requestedServices) {
    console.log(requestedServices);
    // Docs - https://docs.mermaidchart.com/mermaid-oss/syntax/flowchart.html
    let chartComponents = [];
    chartComponents.push('graph TD');
    let services = [];
    let resources = [];
    let dependencyResources = [];
    let relationships = [];
    services.push("%% Services");
    resources.push("%% Resources");
    dependencyResources.push("%% Dependency Resources");
    relationships.push("%% Relationships");
    let aksClusterExists = false;
    let containerAppEnvExists = false;
    let dependencies = [];
    for (const service of requestedServices) {
        let serviceName = [`Name: ${service.name}`];
        const projectRelativePath = path.relative(workspaceFolder, service.path);
        serviceName.push(`Path: ${projectRelativePath}`);
        serviceName.push(`Language: ${service.language}`);
        // check if service is ServiceParameter
        if (isServiceParameter(service)) {
            serviceName.push(`Port: ${service.port}`);
            if (service.dockerSettings && service.azureComputeHost === models_1.TargetAppServiceType.AzureContainerApp) {
                serviceName.push(`DockerFile: ${path.relative(service.path, service.dockerSettings.dockerFilePath) || "./Dockerfile"}`);
                serviceName.push(`Docker Context: ${path.relative(service.path, service.dockerSettings.dockerContext) || "./"}`);
            }
        }
        const serviceInternalName = `svc-${getProperServiceType(service.azureComputeHost)}_${service.name.replace(/\s/g, '').toLowerCase()}`;
        const resourceInternalName = createInternalName(getProperServiceType(service.azureComputeHost), service.name);
        services.push(createComponentName(serviceInternalName, serviceName.join('\n'), NodeShape.Rectangle));
        relationships.push(createRelationshipString(serviceInternalName, resourceInternalName, 'hosted on', ArrowType.SOLID));
        // add resource
        // AKS: add all services to one cluster as microservices
        if (service.azureComputeHost === models_1.TargetAppServiceType.AzureKubernetesService) {
            if (!aksClusterExists) {
                resources.push(`subgraph ${aksClusterInternalName}[\"${aksClusterName}\"]`);
                for (const service of requestedServices) {
                    if (service.azureComputeHost === models_1.TargetAppServiceType.AzureKubernetesService) {
                        resources.push(createComponentName(createInternalName(getProperServiceType(service.azureComputeHost), service.name), createServiceName(service.name, service.azureComputeHost), NodeShape.RoundedRectangle));
                    }
                }
                resources.push('end');
                resources.push(`${aksClusterInternalName}:::cluster`);
                aksClusterExists = true;
            }
        }
        // ACA: add all container apps to an environment
        else if (service.azureComputeHost === models_1.TargetAppServiceType.AzureContainerApp) {
            if (!containerAppEnvExists) {
                resources.push(`subgraph ${containerAppEnvInternalName}[\"${containerAppEnvName}\"]`);
                for (const service of requestedServices) {
                    if (service.azureComputeHost === models_1.TargetAppServiceType.AzureContainerApp) {
                        resources.push(createComponentName(createInternalName(getProperServiceType(service.azureComputeHost), service.name), createServiceName(service.name, service.azureComputeHost), NodeShape.RoundedRectangle));
                    }
                }
                resources.push('end');
                resources.push(`${containerAppEnvInternalName}:::cluster`);
                containerAppEnvExists = true;
            }
        }
        // App service and function app
        else {
            resources.push(createComponentName(createInternalName(getProperServiceType(service.azureComputeHost), service.name), createServiceName(service.name, service.azureComputeHost), NodeShape.RoundedRectangle));
        }
        for (const dependency of service.dependencies) {
            if (isServiceDependencyParameter(dependency)) {
                // compute resources should be added when dealing with service above, only add relationship
                if (Object.values(models_1.TargetAppServiceType).map(x => x.toString()).includes(dependency.serviceType)) {
                    const dependencyInternalName = createInternalName(getProperServiceType(dependency.serviceType.toString()), dependency.name);
                    relationships.push(createRelationshipString(resourceInternalName, dependencyInternalName, (0, project_1.standardizeConnectionType)(dependency.connectionType) ?? dependency.connectionType, ArrowType.DOTTED));
                }
                // add one dependency resource for each type of service
                else if (!dependencies.includes(dependency.serviceType)) {
                    dependencies.push(dependency.serviceType);
                    const dependencyInternalName = createInternalName(dependency.serviceType, dependency.name);
                    dependencyResources.push(createComponentName(dependencyInternalName, dependency.name + ` (${(0, project_1.standardizeResourceType)(dependency.serviceType) ?? dependency.serviceType})`, NodeShape.Rectangle));
                    relationships.push(createRelationshipString(resourceInternalName, dependencyInternalName, (0, project_1.standardizeConnectionType)(dependency.connectionType) ?? dependency.connectionType, ArrowType.DOTTED));
                }
            }
            else {
                // Handle string dependency (from Service type)
                const dependencyName = dependency;
                // Check if this dependency matches any of the requested services by name
                const matchingService = requestedServices.find(s => s.name === dependencyName);
                if (matchingService) {
                    // It's a compute resource dependency - create relationship to the matching service
                    const dependencyInternalName = createInternalName(getProperServiceType(matchingService.azureComputeHost), matchingService.name);
                    relationships.push(createRelationshipString(resourceInternalName, dependencyInternalName, 'depends on', ArrowType.DOTTED));
                }
                else {
                    // It's an external dependency - treat as a generic resource
                    const dependencyInternalName = createInternalName('azure', dependencyName);
                    if (!dependencyResources.some(res => res.includes(dependencyInternalName))) {
                        dependencyResources.push(createComponentName(dependencyInternalName, `${dependencyName}`, NodeShape.Rectangle));
                        relationships.push(createRelationshipString(resourceInternalName, dependencyInternalName, 'depends on', ArrowType.DOTTED));
                    }
                }
            }
        }
    }
    chartComponents.push(...services);
    chartComponents.push("subgraph \"Compute Resources\"");
    chartComponents.push(...resources);
    chartComponents.push("end");
    if (dependencyResources.length > 1) {
        chartComponents.push("subgraph \"Dependency Resources\"");
        chartComponents.push(...dependencyResources);
        chartComponents.push("end");
    }
    chartComponents.push(...relationships);
    const resultingMermaidGraph = chartComponents.join('\n');
    return resultingMermaidGraph;
}
function isServiceParameter(service) {
    return 'port' in service;
}
function isServiceDependencyParameter(serviceDependency) {
    return typeof serviceDependency === 'object' && 'serviceType' in serviceDependency;
}
