// containerization-assist-mcp/dist/src/tools/index.js
import analyzeRepoTool from './analyze-repo/tool.js';
import buildImageContextTool from './build-image-context/tool.js';
import fixDockerfileTool from './fix-dockerfile/tool.js';
import generateDockerfileTool from './generate-dockerfile/tool.js';
import generateK8sManifestsTool from './generate-k8s-manifests/tool.js';
import opsTool from './ops/tool.js';
import prepareClusterTool from './prepare-cluster/tool.js';
import pushImageTool from './push-image/tool.js';
import scanImageTool from './scan-image/tool.js';
import tagImageTool from './tag-image/tool.js';
import verifyDeployTool from './verify-deploy/tool.js';
import { TOOL_NAME } from './shared/toolDefinition.js';
// Ensure proper names on all tools
analyzeRepoTool.name = TOOL_NAME.ANALYZE_REPO;
buildImageContextTool.name = TOOL_NAME.BUILD_IMAGE_CONTEXT;
fixDockerfileTool.name = TOOL_NAME.FIX_DOCKERFILE;
generateDockerfileTool.name = TOOL_NAME.GENERATE_DOCKERFILE;
generateK8sManifestsTool.name = TOOL_NAME.GENERATE_K8S_MANIFESTS;
opsTool.name = TOOL_NAME.OPS;
prepareClusterTool.name = TOOL_NAME.PREPARE_CLUSTER;
pushImageTool.name = TOOL_NAME.PUSH_IMAGE;
scanImageTool.name = TOOL_NAME.SCAN_IMAGE;
tagImageTool.name = TOOL_NAME.TAG_IMAGE;
verifyDeployTool.name = TOOL_NAME.VERIFY_DEPLOY;
// Type-safe tool array using the union type
// All tools are now deterministic plan-based or operational tools
export const ALL_TOOLS = [
    // Plan-based generation tools (use knowledge to create plans)
    analyzeRepoTool,
    fixDockerfileTool,
    generateDockerfileTool,
    generateK8sManifestsTool,
    // Operational/deterministic tools
    buildImageContextTool,
    opsTool,
    prepareClusterTool,
    pushImageTool,
    scanImageTool,
    tagImageTool,
    verifyDeployTool,
];
export { TOOL_NAME, analyzeRepoTool, buildImageContextTool, fixDockerfileTool, generateDockerfileTool, generateK8sManifestsTool, opsTool, prepareClusterTool, pushImageTool, scanImageTool, tagImageTool, verifyDeployTool, };
