// codetocloudmcp/out/common/telemetry/toolTelemetryHelpers/generateArchitectureDiagramTelemetryHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateGenerateArchitectureDiagramTelemetry = updateGenerateArchitectureDiagramTelemetry;
function updateGenerateArchitectureDiagramTelemetry(params, telemetryProperties) {
    const backingServices = new Set(params.services.reduce((acc, service) => {
        return [...acc, ...service.dependencies.map((dep) => dep.serviceType)];
    }, []));
    const appProperties = {
        languages: params.services.map((service) => service.language).join(", "),
        serviceCount: params.services.length,
        backingServiceCount: backingServices.size,
    };
    const infraProperties = {
        computeHostResources: params.services
            .map((service) => service.azureComputeHost)
            .join(", "),
        backingServiceResources: Array.from(backingServices).join(", "),
    };
    const properties = {
        ...telemetryProperties,
        ...appProperties,
        ...infraProperties,
    };
    Object.assign(telemetryProperties, properties);
}
