// codetocloudmcp/out/common/telemetry/toolTelemetryHelpers/getPlanTelemetryHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateGetPlanTelemetry = updateGetPlanTelemetry;
const stringHelpers_1 = require("../../stringHelpers");
const telemetryUtil_1 = require("../telemetryUtil");
function updateGetPlanTelemetry(params, telemetryProperties) {
    const backingServices = new Set(params.services.reduce((acc, service) => {
        return [...acc, ...service.dependencies.map((dep) => dep.serviceType)];
    }, []));
    const appProperties = {
        languages: params.languages.join(", "),
        serviceCount: params.services.length,
        backingServiceCount: backingServices.size,
    };
    const infraProperties = {
        computeHostResources: params.services
            .map((service) => service.azureComputeHost)
            .join(", "),
        backingServiceResources: Array.from(backingServices).join(", "),
    };
    const deploymentProperties = {
        deploymentTool: params.provisioningTool,
        iacType: params.iacOptions ? params.iacOptions : "",
    };
    const additionalProperties = {
        ...appProperties,
        ...infraProperties,
        ...deploymentProperties,
    };
    const properties = {
        ...telemetryProperties,
        ...additionalProperties,
        sourceType: params.sourceType,
        deployOption: params.deployOption,
        subscriptionId: telemetryUtil_1.TelemetryPropertyHelper.prepareSubscriptionIdForTelemetry(params.subscriptionId),
        resourceGroupName: params.resourceGroupName
            ? (0, stringHelpers_1.getHashStr)(params.resourceGroupName, 10)
            : "",
    };
    Object.assign(telemetryProperties, properties);
}
