// codetocloudmcp/out/common/telemetry/toolTelemetryHelpers/summarizeResultTelemetryHelper.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateSummarizeResultTelemetry = updateSummarizeResultTelemetry;
const path_1 = __importDefault(require("path"));
const telemetryUtil_1 = require("../telemetryUtil");
function updateSummarizeResultTelemetry(params, telemetryProperties) {
    const deploymentResult = {
        deploymentStatus: params.deploymentStatus,
        deployOption: params.deployOption,
        createdFiles: params.createdFileNames.map(fileName => path_1.default.basename(fileName)).join(", "),
        deploymentRounds: params.deploymentRounds,
        provisionedResourceTypes: telemetryUtil_1.TelemetryPropertyHelper.prepareResourceTypesForTelemetry(params.provisionedResourceTypes),
    };
    const deploymentProperties = {
        deploymentTool: params.deploymentTool,
        iacType: params.iacOptions ? params.iacOptions : "",
    };
    const properties = {
        ...telemetryProperties,
        ...deploymentResult,
        ...deploymentProperties,
        subscriptionId: telemetryUtil_1.TelemetryPropertyHelper.prepareSubscriptionIdForTelemetry(params.subscriptionId),
    };
    Object.assign(telemetryProperties, properties);
}
