// codetocloudmcp/out/mcpServers/common/environment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResourceBindingEnvKeys = exports.BindingEnvType = exports.DefaultEnvTypes = exports.ServiceUrlDescription = exports.ServiceUrl = exports.EnvTypePrefix = void 0;
const project_1 = require("./project");
exports.EnvTypePrefix = "AZURE_";
exports.ServiceUrl = "SERVICE_URL";
exports.ServiceUrlDescription = "The URL/endpoint of a compute service in the same system. Used to enable communication between two components, such as a back-end and a front-end.";
exports.DefaultEnvTypes = [project_1.EnvType.Endpoint, project_1.EnvType.Config];
exports.BindingEnvType = []; // no common binding anymore
exports.ResourceBindingEnvKeys = {
    [(0, project_1.flattenResourceType)(project_1.BindingResourceType.AZURE_BOT_SERVICE)]: [{
            connection: "bot-connection",
            types: [project_1.EnvType.BotId, project_1.EnvType.BotPassword, project_1.EnvType.BotDomain],
        }],
    [(0, project_1.flattenResourceType)(project_1.BindingResourceType.AZURE_DATABASE_FOR_MYSQL)]: [{
            connection: "secret",
            types: [project_1.EnvType.MysqlHost, project_1.EnvType.MysqlUser, project_1.EnvType.MysqlPassword, project_1.EnvType.MysqlDatabase, project_1.EnvType.MysqlPort, project_1.EnvType.MysqlSsl],
        }],
    [(0, project_1.flattenResourceType)(project_1.BindingResourceType.AZURE_DATABASE_FOR_POSTGRESQL)]: [{
            connection: "secret",
            types: [project_1.EnvType.PostgreSqlHost, project_1.EnvType.PostgreSqlUser, project_1.EnvType.PostgreSqlPassword, project_1.EnvType.PostgreSqlDatabase, project_1.EnvType.PostgreSqlPort],
        }],
    [(0, project_1.flattenResourceType)(project_1.BindingResourceType.AZURE_COSMOS_DB)]: [
        {
            connection: "system-identity",
            types: [project_1.EnvType.CosmosResourceEndpoint],
        },
        {
            connection: "secret",
            types: [project_1.EnvType.CosmosConnectionstring],
        }
    ],
    [(0, project_1.flattenResourceType)(project_1.BindingResourceType.AZURE_KEY_VAULT)]: [{
            connection: "system-identity",
            types: [project_1.EnvType.KeyvaultResourceendpoint],
        }],
    [(0, project_1.flattenResourceType)(project_1.BindingResourceType.AZURE_APPLICATION_INSIGHTS)]: [{
            connection: "secret",
            types: [project_1.EnvType.AppinsightsConnectionString],
        }],
    [(0, project_1.flattenResourceType)(project_1.BindingResourceType.AZURE_OPEN_AI)]: [
        {
            connection: "system-identity",
            types: [project_1.EnvType.OpenaiBase, project_1.EnvType.OpenaiDeployment],
        },
        {
            connection: "secret",
            types: [project_1.EnvType.OpenaiBase, project_1.EnvType.OpenaiKey, project_1.EnvType.OpenaiDeployment],
        }
    ],
    [(0, project_1.flattenResourceType)(project_1.BindingResourceType.AZURE_STORAGE_ACCOUNT)]: [
        {
            connection: "system-identity",
            types: [project_1.EnvType.StorageBlobendpoint, project_1.EnvType.StorageTableendpoint, project_1.EnvType.StorageQueueendpoint, project_1.EnvType.StorageFileendpoint],
        },
        {
            connection: "secret",
            types: [project_1.EnvType.StorageConnectionString],
        }
    ],
    [(0, project_1.flattenResourceType)(project_1.BindingResourceType.AZURE_AI_SEARCH)]: [
        {
            connection: "system-identity",
            types: [project_1.EnvType.AisearchEndpoint],
        },
        {
            connection: "secret",
            types: [project_1.EnvType.AisearchEndpoint, project_1.EnvType.AisearchKey],
        }
    ],
    [(0, project_1.flattenResourceType)(project_1.BindingResourceType.AZURE_AI_SERVICES)]: [
        {
            connection: "system-identity",
            types: [project_1.EnvType.AiservicesOpenaiBase, project_1.EnvType.AiservicesSpeechEndpoint, project_1.EnvType.AiservicesCognitiveservicesEndpoint, project_1.EnvType.AiservicesOpenaiDeployment],
        },
        {
            connection: "secret",
            types: [project_1.EnvType.AiservicesOpenaiBase, project_1.EnvType.AiservicesSpeechEndpoint, project_1.EnvType.AiservicesCognitiveservicesEndpoint, project_1.EnvType.AiservicesKey, project_1.EnvType.AiservicesOpenaiDeployment],
        }
    ],
    [(0, project_1.flattenResourceType)(project_1.BindingResourceType.AZURE_SIGNALR_SERVICE)]: [
        {
            connection: "secret",
            types: [project_1.EnvType.SignalRConnectionString],
        }, {
            connection: "system-identity",
            types: [project_1.EnvType.SignalRConnectionString],
        }
    ],
    [(0, project_1.flattenResourceType)(project_1.BindingResourceType.AZURE_CACHE_FOR_REDIS)]: [{
            connection: "secret",
            types: [project_1.EnvType.RedisConnectionString, project_1.EnvType.RedisHost, project_1.EnvType.RedisPort, project_1.EnvType.RedisPassword, project_1.EnvType.RedisSsl, project_1.EnvType.RedisDatabase, project_1.EnvType.RedisKey],
        }],
    [(0, project_1.flattenResourceType)(project_1.BindingResourceType.AZURE_SERVICE_BUS)]: [
        {
            connection: "system-identity",
            types: [project_1.EnvType.ServiceBusNameSpace],
        },
        {
            connection: "secret",
            types: [project_1.EnvType.ServiceBusConnectionString],
        }
    ],
    [(0, project_1.flattenResourceType)(project_1.BindingResourceType.AZURE_SQL_DATABASE)]: [{
            connection: "secret",
            types: [project_1.EnvType.SqlConnectionString, project_1.EnvType.SqlServer, project_1.EnvType.SqlDatabase, project_1.EnvType.SqlUsername, project_1.EnvType.SqlPassword, project_1.EnvType.SqlPort],
        }],
    [(0, project_1.flattenResourceType)(project_1.BindingResourceType.AZURE_WEB_PUBSUB)]: [
        {
            connection: "system-identity",
            types: [project_1.EnvType.WebPubSubHost],
        }, {
            connection: "secret",
            types: [project_1.EnvType.WebPubSubConnectionString],
        }
    ],
    [(0, project_1.flattenResourceType)(project_1.ComputeResourceType.AZURE_CONTAINER_APP)]: [{
            connection: "http",
            types: [exports.ServiceUrl],
        }],
    [(0, project_1.flattenResourceType)(project_1.ComputeResourceType.AZURE_APP_SERVICE)]: [{
            connection: "http",
            types: [exports.ServiceUrl],
        }],
    [(0, project_1.flattenResourceType)(project_1.ComputeResourceType.AZURE_FUNCTIONS)]: [{
            connection: "http",
            types: [exports.ServiceUrl],
        }],
    [(0, project_1.flattenResourceType)(project_1.ComputeResourceType.AZURE_STATIC_WEB_APPS)]: [{
            connection: "http",
            types: [exports.ServiceUrl],
        }],
};
