// codetocloudmcp/out/mcpServers/common/models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PipelinePlatform = exports.TelemetryRequiredChannels = exports.CustomerChannel = exports.TargetAppServiceToComputeHost = exports.serviceNameToResourceTypeMapping = exports.AzureServiceNames = exports.TargetAppServiceType = exports.DeploymentStatus = exports.DeploymentTool = exports.ResultCode = exports.IaCResourceType = exports.DeployOption = exports.SourceType = exports.IacType = exports.DisplayLanguageType = exports.LanguageType = void 0;
// This should align with AZD supported languages
var LanguageType;
(function (LanguageType) {
    LanguageType["DotNet"] = "dotnet";
    LanguageType["Python"] = "python";
    LanguageType["JavaScript"] = "js";
    LanguageType["TypeScript"] = "ts";
    LanguageType["Java"] = "java";
})(LanguageType || (exports.LanguageType = LanguageType = {}));
var DisplayLanguageType;
(function (DisplayLanguageType) {
    DisplayLanguageType["DotNet"] = ".NET";
    DisplayLanguageType["Python"] = "Python";
    DisplayLanguageType["JavaScript"] = "JavaScript";
    DisplayLanguageType["TypeScript"] = "TypeScript";
    DisplayLanguageType["Java"] = "Java";
})(DisplayLanguageType || (exports.DisplayLanguageType = DisplayLanguageType = {}));
var IacType;
(function (IacType) {
    IacType["Bicep"] = "bicep";
    IacType["Terraform"] = "terraform";
})(IacType || (exports.IacType = IacType = {}));
var SourceType;
(function (SourceType) {
    SourceType["fromProject"] = "from-project";
    SourceType["fromAzure"] = "from-azure";
    SourceType["fromContext"] = "from-context";
})(SourceType || (exports.SourceType = SourceType = {}));
var DeployOption;
(function (DeployOption) {
    DeployOption["deployOnly"] = "deploy-only";
    DeployOption["provisionOnly"] = "provision-only";
    DeployOption["provisionAndDeploy"] = "provision-and-deploy";
})(DeployOption || (exports.DeployOption = DeployOption = {}));
var IaCResourceType;
(function (IaCResourceType) {
    IaCResourceType["AZURE_AI_SEARCH"] = "azureaisearch";
    IaCResourceType["AZURE_AI_SERVICES"] = "azureaiservices";
    IaCResourceType["AZURE_APP_SERVICE"] = "azureappservice";
    IaCResourceType["AZURE_APPLICATION_INSIGHTS"] = "azureapplicationinsights";
    IaCResourceType["AZURE_BOT_SERVICE"] = "azurebotservice";
    IaCResourceType["AZURE_CONTAINER_APP"] = "azurecontainerapp";
    IaCResourceType["AZURE_COSMOS_DB"] = "azurecosmosdb";
    IaCResourceType["AZURE_DIAGNOSTICS_SETTINGS"] = "azurediagnosticssettings";
    IaCResourceType["AZURE_FUNCTION_APP"] = "azurefunctions";
    IaCResourceType["AZURE_KEYVAULT"] = "azurekeyvault";
    IaCResourceType["AZURE_KUBERNETES_SERVICE"] = "azurekubernetesservice";
    IaCResourceType["AZURE_LOG_ANALYTICS"] = "azureloganalytics";
    IaCResourceType["AZURE_MYSQL_DB"] = "azuredatabaseformysql";
    IaCResourceType["AZURE_OPENAI"] = "azureopenai";
    IaCResourceType["AZURE_POSTGRESQL_DB"] = "azuredatabaseforpostgresql";
    IaCResourceType["AZURE_PRIVATE_ENDPOINT"] = "azureprivateendpoint";
    IaCResourceType["AZURE_REDIS_CACHE"] = "azurecacheforredis";
    IaCResourceType["AZURE_SQL_DB"] = "azuresqldatabase";
    IaCResourceType["AZURE_STORAGE_ACCOUNT"] = "azurestorageaccount";
    IaCResourceType["AZURE_STATIC_WEB_APP"] = "azurestaticwebapps";
    IaCResourceType["AZURE_SERVICE_BUS"] = "azureservicebus";
    IaCResourceType["AZURE_SIGNALR"] = "azuresignalrservice";
    IaCResourceType["AZURE_USER_IDENTITY"] = "azureuseridentity";
    IaCResourceType["AZURE_VIRTUAL_NETWORK"] = "azurevirtualnetwork";
    IaCResourceType["AZURE_WEBPUBSUB"] = "azurewebpubsub";
    IaCResourceType["NONE"] = "none";
})(IaCResourceType || (exports.IaCResourceType = IaCResourceType = {}));
exports.ResultCode = {
    "Success": "success",
    "Failure": "failure",
    "Cancelled": "cancelled",
};
var DeploymentTool;
(function (DeploymentTool) {
    DeploymentTool["Azd"] = "azd";
    DeploymentTool["AzCLI"] = "azcli";
})(DeploymentTool || (exports.DeploymentTool = DeploymentTool = {}));
var DeploymentStatus;
(function (DeploymentStatus) {
    DeploymentStatus["Success"] = "success";
    DeploymentStatus["Failure"] = "failure";
})(DeploymentStatus || (exports.DeploymentStatus = DeploymentStatus = {}));
var TargetAppServiceType;
(function (TargetAppServiceType) {
    TargetAppServiceType["AzureAppService"] = "appservice";
    TargetAppServiceType["AzureAppServiceManagedInstance"] = "appservicemanagedinstance";
    TargetAppServiceType["AzureContainerApp"] = "containerapp";
    TargetAppServiceType["AzureFunctionApp"] = "functionapp";
    TargetAppServiceType["AzureKubernetesService"] = "aks";
})(TargetAppServiceType || (exports.TargetAppServiceType = TargetAppServiceType = {}));
var AzureServiceNames;
(function (AzureServiceNames) {
    AzureServiceNames["AzureAISearch"] = "azureaisearch";
    AzureServiceNames["AzureAIServices"] = "azureaiservices";
    AzureServiceNames["AzureAppService"] = "appservice";
    // In azure.yaml, this should still be "appservice". Defined here for clarity.
    AzureServiceNames["AzureAppServiceManagedInstance"] = "appservicemanagedinstance";
    AzureServiceNames["AzureAppInsights"] = "azureapplicationinsights";
    AzureServiceNames["AzureBotService"] = "azurebotservice";
    AzureServiceNames["AzureContainerApp"] = "containerapp";
    AzureServiceNames["AzureCosmosDB"] = "azurecosmosdb";
    AzureServiceNames["AzureFunctionApp"] = "function";
    AzureServiceNames["AzureKeyVault"] = "azurekeyvault";
    AzureServiceNames["AzureKubernetesService"] = "aks";
    AzureServiceNames["AzureMySQL"] = "azuredatabaseformysql";
    AzureServiceNames["AzureOpenAI"] = "azureopenai";
    AzureServiceNames["AzurePostgresSQL"] = "azuredatabaseforpostgresql";
    AzureServiceNames["AzurePrivateEndpoint"] = "azureprivateendpoint";
    AzureServiceNames["AzureRedisCache"] = "azurecacheforredis";
    AzureServiceNames["AzureSQLDB"] = "azuresqldatabase";
    AzureServiceNames["AzureStorageAccount"] = "azurestorageaccount";
    AzureServiceNames["AzureStaticWebApp"] = "staticwebapp";
    AzureServiceNames["AzureServiceBus"] = "azureservicebus";
    AzureServiceNames["AzureSignalRService"] = "azuresignalrservice";
    AzureServiceNames["AzureVirtualNetwork"] = "azurevirtualnetwork";
    AzureServiceNames["AzureWebPubSub"] = "azurewebpubsub";
    AzureServiceNames["AzureContainerRegistry"] = "azurecontainerregistry";
})(AzureServiceNames || (exports.AzureServiceNames = AzureServiceNames = {}));
exports.serviceNameToResourceTypeMapping = {
    [AzureServiceNames.AzureAISearch]: "Microsoft.Search/searchServices",
    [AzureServiceNames.AzureAIServices]: "Microsoft.CognitiveServices/accounts",
    [AzureServiceNames.AzureAppService]: "Microsoft.Web/sites",
    [AzureServiceNames.AzureAppServiceManagedInstance]: "Microsoft.Web/sites",
    [AzureServiceNames.AzureAppInsights]: "Microsoft.Insights/components",
    [AzureServiceNames.AzureBotService]: "Microsoft.BotService/botServices",
    [AzureServiceNames.AzureContainerApp]: "Microsoft.App/containerApps",
    [AzureServiceNames.AzureCosmosDB]: "Microsoft.DocumentDB/databaseAccounts",
    [AzureServiceNames.AzureFunctionApp]: "Microsoft.Web/sites",
    [AzureServiceNames.AzureKeyVault]: "Microsoft.KeyVault/vaults",
    [AzureServiceNames.AzureKubernetesService]: "Microsoft.ContainerService/managedClusters",
    [AzureServiceNames.AzureMySQL]: "Microsoft.DBforMySQL/flexibleServers",
    [AzureServiceNames.AzureOpenAI]: "Microsoft.CognitiveServices/accounts",
    [AzureServiceNames.AzurePostgresSQL]: "Microsoft.DBforPostgreSQL/flexibleServers",
    [AzureServiceNames.AzurePrivateEndpoint]: "Microsoft.Network/privateEndpoints",
    [AzureServiceNames.AzureRedisCache]: "Microsoft.Cache/Redis",
    [AzureServiceNames.AzureSQLDB]: "Microsoft.Sql/servers/databases",
    [AzureServiceNames.AzureStorageAccount]: "Microsoft.Storage/storageAccounts",
    [AzureServiceNames.AzureStaticWebApp]: "Microsoft.Web/staticSites",
    [AzureServiceNames.AzureServiceBus]: "Microsoft.ServiceBus/namespaces",
    [AzureServiceNames.AzureSignalRService]: "Microsoft.SignalRService/signalR",
    [AzureServiceNames.AzureVirtualNetwork]: "Microsoft.Network/virtualNetworks",
    [AzureServiceNames.AzureWebPubSub]: "Microsoft.SignalRService/webPubSub",
    [AzureServiceNames.AzureContainerRegistry]: "Microsoft.ContainerRegistry/registries"
};
exports.TargetAppServiceToComputeHost = {
    [TargetAppServiceType.AzureAppService]: "Azure App Service",
    [TargetAppServiceType.AzureAppServiceManagedInstance]: "Azure App Service Managed Instance",
    [TargetAppServiceType.AzureContainerApp]: "Azure Container Apps",
    [TargetAppServiceType.AzureFunctionApp]: "Azure Functions",
    [TargetAppServiceType.AzureKubernetesService]: "Azure Kubernetes Service"
};
// If a new channel is added, also update TelemetryRequiredChannels
var CustomerChannel;
(function (CustomerChannel) {
    CustomerChannel["JAVA"] = "Java";
    CustomerChannel["DOTNET"] = ".NET";
    CustomerChannel["AZMCP"] = "AzMCP";
    CustomerChannel["CODETOCLOUD"] = "CodeToCloud";
    CustomerChannel["INTELLIJ"] = "IntelliJ";
    CustomerChannel["COPILOTAGENT"] = "CopilotAgent";
    CustomerChannel["COPILOTCLI"] = "CopilotCLI";
})(CustomerChannel || (exports.CustomerChannel = CustomerChannel = {}));
;
exports.TelemetryRequiredChannels = [
    CustomerChannel.JAVA,
    CustomerChannel.DOTNET,
    CustomerChannel.INTELLIJ,
    CustomerChannel.COPILOTAGENT,
    CustomerChannel.COPILOTCLI
];
var PipelinePlatform;
(function (PipelinePlatform) {
    PipelinePlatform["GithubActions"] = "GitHub Actions";
    PipelinePlatform["AzureDevOps"] = "Azure DevOps";
})(PipelinePlatform || (exports.PipelinePlatform = PipelinePlatform = {}));
