// containerization-assist-mcp/dist/src/tools/fix-dockerfile/schema.js
/**
 * Schema definition for fix-dockerfile tool
 */
import { z } from 'zod';
import { environment, platform } from '../shared/schemas.js';
export const fixDockerfileSchema = z
    .object({
    dockerfile: z.string().optional().describe('Dockerfile content to analyze for fixes'),
    path: z.string().optional().describe('Path to Dockerfile file to analyze for fixes'),
    environment: environment.describe('Target environment (production, development, etc.)'),
    targetPlatform: platform.describe('Target platform for deployment (e.g., linux/amd64, linux/arm64). Used for context in recommendations.'),
    strictPlatformValidation: z
        .boolean()
        .optional()
        .default(false)
        .describe('Deprecated: No longer enforces platform flags. Reserved for future use.'),
    policyPath: z
        .string()
        .optional()
        .describe('Optional path to specific policy file to use for organizational policy validation (defaults to all policies in policies/)'),
})
    .refine((data) => data.dockerfile || data.path, {
    message: "Either 'dockerfile' content or 'path' must be provided",
});
