// containerization-assist-mcp/dist-cjs/src/cli/commands/inspect-tools.js
"use strict";
/**
 * Inspect Tools CLI Command
 *
 * Provides command-line interface for tool discovery, capability analysis,
 * and metadata inspection for AI enhancement features.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createInspectToolsCommand = createInspectToolsCommand;
const commander_1 = require("commander");
const tools_1 = require("../../tools");
const tool_metadata_1 = require("../../types/tool-metadata");
const node_fs_1 = require("node:fs");
const node_path_1 = __importDefault(require("node:path"));
const error_formatting_1 = require("../error-formatting");
const render_1 = require("../render");
const types_1 = require("../../types");
/**
 * Discovers and analyzes all available tools and their capabilities
 * Inlined from tool-discovery service to eliminate unnecessary abstraction layer
 */
async function discoverToolCapabilities(options = {}) {
    try {
        const filteredTools = [];
        for (const tool of tools_1.ALL_TOOLS) {
            // Apply filters
            if (options.knowledgeEnhancedOnly && !tool.metadata.knowledgeEnhanced)
                continue;
            filteredTools.push(tool);
        }
        return (0, types_1.Success)(filteredTools.sort((a, b) => a.name.localeCompare(b.name)));
    }
    catch (error) {
        return (0, types_1.Failure)(`Failed to discover tool capabilities: ${error}`);
    }
}
/**
 * Get comprehensive tool statistics
 */
async function getToolStatistics() {
    const result = await discoverToolCapabilities();
    if (!result.ok)
        return result;
    const tools = result.value;
    const stats = {
        total: tools.length,
        knowledgeEnhanced: tools.filter((t) => t.metadata.knowledgeEnhanced).length,
        categories: {},
    };
    // Count categories
    for (const tool of tools) {
        const category = tool.category || 'uncategorized';
        stats.categories[category] = (stats.categories[category] || 0) + 1;
    }
    return (0, types_1.Success)(stats);
}
/**
 * Validate tool metadata using centralized validation
 */
async function validateToolMetadata() {
    const toolsResult = await discoverToolCapabilities();
    if (!toolsResult.ok)
        return toolsResult;
    // Convert to ValidatableTool format for centralized validation
    const validatableTools = toolsResult.value.map((tool) => ({
        name: tool.name,
        metadata: tool.metadata,
    }));
    const validationResult = await (0, tool_metadata_1.validateAllToolMetadata)(validatableTools);
    if (!validationResult.ok)
        return validationResult;
    // Convert back to the expected format
    return (0, types_1.Success)({
        valid: validationResult.value.validTools,
        invalid: validationResult.value.invalidTools.map((invalid) => ({
            name: invalid.name,
            issues: invalid.issues,
        })),
    });
}
/**
 * Generate comprehensive tool capabilities report
 */
async function generateToolCapabilitiesReport() {
    const [capabilitiesResult, statsResult, validationResult] = await Promise.all([
        discoverToolCapabilities({ includeDetails: true }),
        getToolStatistics(),
        validateToolMetadata(),
    ]);
    if (!capabilitiesResult.ok)
        return capabilitiesResult;
    if (!statsResult.ok)
        return statsResult;
    if (!validationResult.ok)
        return validationResult;
    const capabilities = capabilitiesResult.value;
    const stats = statsResult.value;
    const validation = validationResult.value;
    let report = `# Tool Capabilities Report

Generated: ${new Date().toISOString()}

## Summary Statistics

- **Total Tools**: ${stats.total}
- **Knowledge-Enhanced Tools**: ${stats.knowledgeEnhanced} (${Math.round((stats.knowledgeEnhanced / stats.total) * 100)}%)

### Categories
`;
    for (const [category, count] of Object.entries(stats.categories)) {
        report += `- **${category}**: ${count} tools\n`;
    }
    report += `\n## Tool Details\n\n`;
    for (const tool of capabilities) {
        report += `### ${tool.name}
- **Description**: ${tool.description}
- **Category**: ${tool.category || 'uncategorized'}
- **Knowledge-Enhanced**: ${tool.metadata.knowledgeEnhanced ? '✅' : '❌'}

`;
    }
    if (validation.invalid.length > 0) {
        report += `\n## Metadata Issues\n\n`;
        for (const issue of validation.invalid) {
            report += `### ${issue.name}\n`;
            for (const problem of issue.issues) {
                report += `- ⚠️ ${problem}\n`;
            }
            report += '\n';
        }
    }
    return (0, types_1.Success)(report);
}
/**
 * Create inspect-tools CLI command
 */
function createInspectToolsCommand() {
    const cmd = new commander_1.Command('inspect-tools');
    cmd.description('Inspect and analyze tool capabilities and AI enhancement status');
    // List all tools with capabilities
    cmd
        .command('list')
        .description('List all tools with their AI enhancement status')
        .option('--knowledge-enhanced', 'Show only knowledge-enhanced tools')
        .option('--format <format>', 'Output format (table, json, csv)', 'table')
        .option('--detailed', 'Include detailed metadata')
        .action(async (options) => {
        try {
            const discoveryOptions = {
                knowledgeEnhancedOnly: options.knowledgeEnhanced,
                includeDetails: options.detailed,
            };
            const result = await discoverToolCapabilities(discoveryOptions);
            if (!result.ok) {
                (0, error_formatting_1.handleResultError)(result, 'Failed to discover tools');
            }
            const tools = result.value;
            (0, render_1.renderTools)(tools, {
                format: options.format,
                detailed: options.detailed,
            });
        }
        catch (error) {
            (0, error_formatting_1.handleGenericError)('Error during operation', error);
        }
    });
    // Statistics command
    cmd
        .command('stats')
        .description('Show tool statistics and AI enhancement overview')
        .option('--json', 'Output as JSON')
        .action(async (options) => {
        try {
            const result = await getToolStatistics();
            if (!result.ok) {
                (0, error_formatting_1.handleResultError)(result, 'Failed to get statistics');
            }
            const stats = result.value;
            if (options.json) {
                console.info(JSON.stringify(stats, null, 2));
            }
            else {
                console.info('📊 Tool Statistics\n');
                console.info(`Total Tools: ${stats.total}`);
                console.info(`Knowledge-Enhanced: ${stats.knowledgeEnhanced} (${Math.round((stats.knowledgeEnhanced / stats.total) * 100)}%)`);
                console.info('\n📁 Categories:');
                for (const [category, count] of Object.entries(stats.categories)) {
                    console.info(`  ${category}: ${count} tools`);
                }
            }
        }
        catch (error) {
            (0, error_formatting_1.handleGenericError)('Error during operation', error);
        }
    });
    // Knowledge-enhanced tools command
    cmd
        .command('knowledge-enhanced')
        .description('List tools that use knowledge enhancement')
        .option('--count', 'Show only count')
        .action(async (options) => {
        try {
            const result = await discoverToolCapabilities({ knowledgeEnhancedOnly: true });
            if (!result.ok) {
                (0, error_formatting_1.handleResultError)(result, 'Failed to get knowledge-enhanced tools');
            }
            const tools = result.value.map((tool) => tool.name);
            if (options.count) {
                console.info(tools.length);
            }
            else {
                console.info('📚 Knowledge-Enhanced Tools:\n');
                tools.forEach((tool) => console.info(`  • ${tool}`));
                console.info(`\nTotal: ${tools.length} tools`);
            }
        }
        catch (error) {
            (0, error_formatting_1.handleGenericError)('Error during operation', error);
        }
    });
    // Validate command
    cmd
        .command('validate')
        .description('Validate tool metadata consistency and compliance')
        .option('--json', 'Output as JSON')
        .action(async (options) => {
        try {
            const result = await validateToolMetadata();
            if (!result.ok) {
                (0, error_formatting_1.handleResultError)(result, 'Failed to validate metadata');
            }
            const validation = result.value;
            if (options.json) {
                console.info(JSON.stringify(validation, null, 2));
            }
            else {
                console.info('🔍 Metadata Validation Results\n');
                if (validation.valid.length > 0) {
                    console.info(`✅ Valid Tools (${validation.valid.length}):`);
                    validation.valid.forEach((tool) => console.info(`  • ${tool}`));
                    console.info('');
                }
                if (validation.invalid.length > 0) {
                    console.info(`❌ Tools with Issues (${validation.invalid.length}):`);
                    validation.invalid.forEach((item) => {
                        console.info(`  • ${item.name}:`);
                        item.issues.forEach((issue) => console.info(`    - ${issue}`));
                    });
                    console.info('');
                }
                const totalTools = validation.valid.length + validation.invalid.length;
                const validPercentage = Math.round((validation.valid.length / totalTools) * 100);
                console.info(`Overall: ${validation.valid.length}/${totalTools} tools valid (${validPercentage}%)`);
            }
        }
        catch (error) {
            (0, error_formatting_1.handleGenericError)('Error during operation', error);
        }
    });
    // Report command
    cmd
        .command('report')
        .description('Generate comprehensive tool capabilities report')
        .option('--output <file>', 'Output file path (markdown format)')
        .option('--stdout', 'Output to stdout instead of file')
        .action(async (options) => {
        try {
            const result = await generateToolCapabilitiesReport();
            if (!result.ok) {
                (0, error_formatting_1.handleResultError)(result, 'Failed to generate report');
            }
            const report = result.value;
            if (options.stdout) {
                console.info(report);
            }
            else {
                const outputPath = options.output || node_path_1.default.join(process.cwd(), 'tool-capabilities-report.md');
                (0, node_fs_1.writeFileSync)(outputPath, report, 'utf-8');
                console.info(`📄 Report generated: ${outputPath}`);
            }
        }
        catch (error) {
            (0, error_formatting_1.handleGenericError)('Error during operation', error);
        }
    });
    return cmd;
}
