// codetocloudmcp/out/mcpServers/tools/mcpAzureGetRegionsForDeploymentTool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.McpAzureGetRegionsForDeploymentTool = void 0;
const codeToCloudMCPServerTool_1 = require("../codeToCloudMCPServerTool");
const tool_1 = require("../toolsImplementations/getRegionsForDeploymentTool/tool");
const schema_1 = require("./schema");
class McpAzureGetRegionsForDeploymentTool extends codeToCloudMCPServerTool_1.CodeToCloudMCPServerTool {
    tool;
    constructor(stateManager) {
        super(schema_1.AzureGetRegionsForDeploymentSchema, stateManager);
        this.tool = new tool_1.GetRegionsForDeploymentTool(stateManager);
    }
    getCodeToCloudTool() {
        return this.tool;
    }
    updateTelemetryPropertiesWithInput(params, telemetryProperties) {
        super.updateTelemetryPropertiesWithInput(params, telemetryProperties);
        const properties = {
            ...telemetryProperties,
            resourceTypes: params.resourceTypes.map(resourceType => JSON.stringify(resourceType)).join(', '),
            subscriptionId: params.subscriptionId
        };
        Object.assign(telemetryProperties, properties);
    }
}
exports.McpAzureGetRegionsForDeploymentTool = McpAzureGetRegionsForDeploymentTool;
