// codetocloudmcp/out/mcpServers/tools/mcpAzureQuotaCheckTool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.McpAzureQuotaCheckTool = void 0;
const codeToCloudMCPServerTool_1 = require("../codeToCloudMCPServerTool");
const tool_1 = require("../toolsImplementations/azureQuotaCheck/tool");
const schema_1 = require("./schema");
class McpAzureQuotaCheckTool extends codeToCloudMCPServerTool_1.CodeToCloudMCPServerTool {
    tool;
    constructor(stateManager) {
        super(schema_1.AzureQuotaCheckSchema, stateManager);
        this.tool = new tool_1.AzureQuotaCheckTool(stateManager);
    }
    getCodeToCloudTool() {
        return this.tool;
    }
    updateTelemetryPropertiesWithInput(params, telemetryProperties) {
        super.updateTelemetryPropertiesWithInput(params, telemetryProperties);
        const properties = {
            ...telemetryProperties,
            resourceTypes: params.resourceTypes.join(', '),
            region: params.region,
            subscriptionId: params.subscriptionId
        };
        Object.assign(telemetryProperties, properties);
    }
}
exports.McpAzureQuotaCheckTool = McpAzureQuotaCheckTool;
