// codetocloudmcp/out/mcpServers/tools/mcpDiagnoseExistingResourcesTool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.McpDiagnoseExistingResourcesTool = void 0;
const codeToCloudMCPServerTool_1 = require("../codeToCloudMCPServerTool");
const tool_1 = require("../toolsImplementations/diagnoseExistingResources/tool");
const schema_1 = require("./schema");
class McpDiagnoseExistingResourcesTool extends codeToCloudMCPServerTool_1.CodeToCloudMCPServerTool {
    tool;
    constructor(stateManager) {
        super(schema_1.DiagnoseExistingResourcesSchema, stateManager);
        this.tool = new tool_1.DiagnoseExistingResourcesTool(stateManager);
    }
    getCodeToCloudTool() {
        return this.tool;
    }
    updateTelemetryPropertiesWithInput(params, telemetryProperties) {
        super.updateTelemetryPropertiesWithInput(params, telemetryProperties);
        const properties = {
            ...telemetryProperties,
            containerAppIds: params.containerRegistryConnections.containerAppIds,
            kubernetesServiceIds: params.containerRegistryConnections.kubernetesIds,
            containerRegistryId: params.containerRegistryConnections.containerRegistryId,
        };
        Object.assign(telemetryProperties, properties);
    }
}
exports.McpDiagnoseExistingResourcesTool = McpDiagnoseExistingResourcesTool;
