// codetocloudmcp/out/mcpServers/tools/mcpGetAzureResourceRulesTool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.McpGetAzureResourceRulesTool = void 0;
const codeToCloudMCPServerTool_1 = require("../codeToCloudMCPServerTool");
const tool_1 = require("../toolsImplementations/getAzureResourceRules/tool");
const schema_1 = require("./schema");
class McpGetAzureResourceRulesTool extends codeToCloudMCPServerTool_1.CodeToCloudMCPServerTool {
    tool;
    constructor(stateManager) {
        super(schema_1.GetAzureResourceRulesSchema, stateManager);
        this.tool = new tool_1.GetAzureResourceRulesTool(stateManager);
    }
    getCodeToCloudTool() {
        return this.tool;
    }
    updateTelemetryPropertiesWithInput(params, telemetryProperties) {
        super.updateTelemetryPropertiesWithInput(params, telemetryProperties);
        const properties = {
            ...telemetryProperties,
            deploymentTool: params.deploymentTool,
            iacType: params.iacType ? params.iacType : "",
            resourceTypes: params.resourceTypes.join(', ')
        };
        Object.assign(telemetryProperties, properties);
    }
}
exports.McpGetAzureResourceRulesTool = McpGetAzureResourceRulesTool;
