// codetocloudmcp/out/mcpServers/tools/mcpGetContainerizationPlanTool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.McpGetContainerizationPlanTool = void 0;
const codeToCloudMCPServerTool_1 = require("../codeToCloudMCPServerTool");
const tool_1 = require("../toolsImplementations/getContainerizationPlan/tool");
const schema_1 = require("./schema");
class McpGetContainerizationPlanTool extends codeToCloudMCPServerTool_1.CodeToCloudMCPServerTool {
    tool;
    constructor(stateManager) {
        super(schema_1.GetContainerizationPlanSchema, stateManager);
        this.tool = new tool_1.GetContainerizationPlanTool(stateManager);
    }
    getCodeToCloudTool() {
        return this.tool;
    }
    updateTelemetryPropertiesWithInput(params, telemetryProperties) {
        super.updateTelemetryPropertiesWithInput(params, telemetryProperties);
        const properties = {
            ...telemetryProperties,
            projectName: params.workspaceFolder.split(/[\\/]/).pop() || "empty",
            serviceCount: params.servicePaths.length
        };
        Object.assign(telemetryProperties, properties);
    }
}
exports.McpGetContainerizationPlanTool = McpGetContainerizationPlanTool;
