// codetocloudmcp/out/mcpServers/tools/mcpGenerateArchitectureDiagramTool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.McpGenerateArchitectureDiagramTool = void 0;
const codeToCloudMCPServerTool_1 = require("../codeToCloudMCPServerTool");
const tool_1 = require("../toolsImplementations/getAppTopology/tool");
const schema_1 = require("./schema");
class McpGenerateArchitectureDiagramTool extends codeToCloudMCPServerTool_1.CodeToCloudMCPServerTool {
    tool;
    constructor(stateManager) {
        super(schema_1.GenerateArchitectureDiagramSchema, stateManager);
        this.tool = new tool_1.GetAppTopologyTool(stateManager);
    }
    getCodeToCloudTool() {
        return this.tool;
    }
    updateTelemetryPropertiesWithInput(params, telemetryProperties) {
        super.updateTelemetryPropertiesWithInput(params, telemetryProperties);
        const backingServices = new Set(params.services.reduce((acc, service) => { return [...acc, ...(service.dependencies.map(dep => dep.serviceType))]; }, []));
        const appProperties = {
            languages: params.services.map(service => service.language).join(', '),
            serviceCount: params.services.length,
            // the number of the dependency services, the same type of service should be counted only once across all services
            backingServiceCount: backingServices.size,
        };
        const infraProperties = {
            computeHostResources: params.services.map(service => service.azureComputeHost).join(', '),
            backingServiceResources: Array.from(backingServices).join(', '),
        };
        const properties = {
            ...telemetryProperties,
            ...appProperties,
            ...infraProperties,
            mermaidDiagram: 'Pending'
        };
        Object.assign(telemetryProperties, properties);
    }
    updateTelemetryPropertiesWithToolResponse(output, telemetryProperties) {
        super.updateTelemetryPropertiesWithToolResponse(output, telemetryProperties);
        const chartPattern = /```mermaid\s*\n([\s\S]*?)\n```/;
        // Assuming the output is a mermaid diagram string
        const properties = {
            ...telemetryProperties,
            mermaidDiagram: output.match(chartPattern)?.[1] ?? ''
        };
        Object.assign(telemetryProperties, properties);
    }
}
exports.McpGenerateArchitectureDiagramTool = McpGenerateArchitectureDiagramTool;
