// codetocloudmcp/out/mcpServers/tools/mcpGetAppLogsTool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.McpGetAppLogsTool = void 0;
const telemetryUtil_1 = require("../../common/telemetry/telemetryUtil");
const tool_1 = require("../../toolsImplementations/getAppLogs/tool");
const schema_1 = require("../../toolsImplementations/shared/schema");
const codeToCloudMCPServerTool_1 = require("../codeToCloudMCPServerTool");
class McpGetAppLogsTool extends codeToCloudMCPServerTool_1.CodeToCloudMCPServerTool {
    tool;
    constructor(stateManager) {
        super(schema_1.GetAppLogsSchema, stateManager);
        this.tool = new tool_1.GetAppLogsTool(stateManager);
    }
    getCodeToCloudTool() {
        return this.tool;
    }
    updateTelemetryPropertiesWithInput(params, telemetryProperties) {
        super.updateTelemetryPropertiesWithInput(params, telemetryProperties);
        const properties = {
            ...telemetryProperties,
            subscriptionId: telemetryUtil_1.TelemetryPropertyHelper.prepareSubscriptionIdForTelemetry(params.subscriptionId),
            resourceTypes: params.resources.map(r => r.type.toString()).join(",")
        };
        Object.assign(telemetryProperties, properties);
    }
}
exports.McpGetAppLogsTool = McpGetAppLogsTool;
