// codetocloudmcp/out/mcpServers/tools/mcpGetPlanTool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.McpGetPlanTool = void 0;
const codeToCloudMCPServerTool_1 = require("../codeToCloudMCPServerTool");
const stringHelpers_1 = require("../common/stringHelpers");
const tool_1 = require("../toolsImplementations/getPlan/tool");
const schema_1 = require("./schema");
class McpGetPlanTool extends codeToCloudMCPServerTool_1.CodeToCloudMCPServerTool {
    tool;
    constructor(stateManager) {
        super(schema_1.GetPlanSchema, stateManager);
        this.tool = new tool_1.GetPlanTool(stateManager);
    }
    getCodeToCloudTool() {
        return this.tool;
    }
    updateTelemetryPropertiesWithInput(params, telemetryProperties) {
        super.updateTelemetryPropertiesWithInput(params, telemetryProperties);
        // Flatten and stringify the properties, TODO: test if services can be directly used for telemetry
        const backingServices = new Set(params.services.reduce((acc, service) => { return [...acc, ...service.dependencies.map(dep => dep.serviceType)]; }, []));
        const appProperties = {
            languages: params.languages.join(', '),
            serviceCount: params.services.length,
            // the number of the dependency services, the same type of service should be counted only once across all services
            backingServiceCount: backingServices.size,
        };
        const infraProperties = {
            computeHostResources: params.services.map(service => service.azureComputeHost).join(', '),
            backingServiceResources: Array.from(backingServices).join(', '),
        };
        const deploymentProperties = {
            deploymentTool: params.provisioningTool,
            iacType: params.iacOptions ? params.iacOptions : ""
        };
        const additionalProperties = { ...appProperties, ...infraProperties, ...deploymentProperties };
        const properties = {
            ...telemetryProperties,
            ...additionalProperties,
            sourceType: params.sourceType,
            deployOption: params.deployOption,
            subscriptionId: params.subscriptionId ?? "",
            resourceGroupName: params.resourceGroupName ? (0, stringHelpers_1.getHashStr)(params.resourceGroupName, 10) : ""
        };
        Object.assign(telemetryProperties, properties);
    }
}
exports.McpGetPlanTool = McpGetPlanTool;
