// codetocloudmcp/out/mcpServers/tools/mcpSummarizeResultTool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.McpSummarizeResultTool = void 0;
const codeToCloudMCPServerTool_1 = require("../codeToCloudMCPServerTool");
const tool_1 = require("../toolsImplementations/summarizeResult/tool");
const schema_1 = require("./schema");
class McpSummarizeResultTool extends codeToCloudMCPServerTool_1.CodeToCloudMCPServerTool {
    tool;
    constructor(stateManager) {
        super(schema_1.SummarizeResultSchema, stateManager);
        this.tool = new tool_1.SummarizeResultTool(stateManager);
    }
    getCodeToCloudTool() {
        return this.tool;
    }
    updateTelemetryPropertiesWithInput(params, telemetryProperties) {
        super.updateTelemetryPropertiesWithInput(params, telemetryProperties);
        const deploymentResult = {
            deploymentStatus: params.deploymentStatus,
            createdFiles: params.createdFileNames.join(', '),
            deploymentRounds: params.deploymentRounds,
            provisionedResourceTypes: params.provisionedResourceTypes.join(', '),
        };
        const deploymentProperties = {
            deploymentTool: params.deploymentTool,
            iacType: params.iacOptions ? params.iacOptions : ""
        };
        const properties = {
            ...telemetryProperties,
            ...deploymentResult,
            ...deploymentProperties,
            subscriptionId: params.subscriptionId
        };
        Object.assign(telemetryProperties, properties);
    }
}
exports.McpSummarizeResultTool = McpSummarizeResultTool;
