// codetocloudmcp/out/common/planGeneration/planStep.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlanStep = exports.PlanStepBulletType = void 0;
var PlanStepBulletType;
(function (PlanStepBulletType) {
    PlanStepBulletType[PlanStepBulletType["Numbered"] = 0] = "Numbered";
    PlanStepBulletType[PlanStepBulletType["Bulleted"] = 1] = "Bulleted";
})(PlanStepBulletType || (exports.PlanStepBulletType = PlanStepBulletType = {}));
class PlanStep {
    bulletType;
    stepName;
    children;
    constructor(stepName, bulletType) {
        this.stepName = stepName;
        this.children = [];
        this.bulletType = bulletType;
    }
    addStep(planStep) {
        if (typeof planStep === "string") {
            planStep = new PlanStep(planStep, this.bulletType);
        }
        this.children.push(planStep);
        return planStep;
    }
    addSteps(planSteps) {
        for (const step of planSteps) {
            if (typeof step === "string") {
                this.children.push(new PlanStep(step, this.bulletType));
            }
            else {
                this.children.push(step);
            }
        }
    }
    render() {
        const lines = [];
        const hasChildren = this.children.length > 0;
        lines.push(`${this.stepName}${hasChildren && !this.stepName.endsWith(":") ? ":" : ""}`);
        for (let i = 0; i < this.children.length; i++) {
            const child = this.children[i].render();
            lines.push(...child.map((line, index) => {
                if (index === 0) {
                    const heading = this.bulletType === PlanStepBulletType.Numbered ? `${i + 1}.` : `-`;
                    return `${heading} ${line}`; // add a number to the heading line
                }
                return `    ${line}`; // add indentation for child lines
            }));
        }
        return lines;
    }
    renderIntoMarkdown() {
        return this.render().join("\n");
    }
}
exports.PlanStep = PlanStep;
