// containerization-assist-mcp/dist-cjs/src/sdk/index.js
"use strict";
/**
 * Containerization Assist SDK
 *
 * Provides direct access to all 11 containerization tools without requiring
 * the MCP (Model Context Protocol) server infrastructure.
 *
 * This SDK is designed for:
 * - VS Code extension developers integrating with Copilot
 * - Direct programmatic usage without MCP overhead
 * - Lightweight tool execution in Node.js applications
 *
 * @example
 * ```typescript
 * import { analyzeRepo, buildImageContext, scanImage } from 'containerization-assist-mcp/sdk';
 *
 * // Full containerization workflow
 * const analysis = await analyzeRepo({ repositoryPath: './my-app' });
 * const build = await buildImageContext({ path: './my-app', imageName: 'myapp:v1' });
 * const scan = await scanImage({ imageId: 'myapp:v1' });
 * ```
 *
 * @packageDocumentation
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatScanImageResult = exports.formatBuildImageResult = exports.formatFixDockerfileResult = exports.formatGenerateDockerfileResult = exports.formatAnalyzeRepoResult = exports.resultFormatters = exports.opsMetadata = exports.verifyDeployMetadata = exports.prepareClusterMetadata = exports.generateK8sManifestsMetadata = exports.pushImageMetadata = exports.tagImageMetadata = exports.scanImageMetadata = exports.buildImageContextMetadata = exports.fixDockerfileMetadata = exports.generateDockerfileMetadata = exports.analyzeRepoMetadata = exports.requiresDependency = exports.getRequiredDepsString = exports.ExternalDeps = exports.getToolsByCategory = exports.standardWorkflow = exports.toolMetadata = exports.opsJsonSchema = exports.verifyDeployJsonSchema = exports.prepareClusterJsonSchema = exports.generateK8sManifestsJsonSchema = exports.pushImageJsonSchema = exports.tagImageJsonSchema = exports.scanImageJsonSchema = exports.buildImageContextJsonSchema = exports.fixDockerfileJsonSchema = exports.generateDockerfileJsonSchema = exports.analyzeRepoJsonSchema = exports.jsonSchemas = exports.executeTool = exports.tools = exports.ops = exports.verifyDeploy = exports.prepareCluster = exports.generateK8sManifests = exports.pushImage = exports.tagImage = exports.scanImage = exports.buildImageContext = exports.fixDockerfile = exports.generateDockerfile = exports.analyzeRepo = exports.Failure = exports.Success = void 0;
exports.loadKnowledgeData = exports.sanitizeForMarkdown = exports.validateRequiredFields = exports.resolveWorkspacePath = exports.formatErrorForLLM = exports.createAbortSignalFromToken = exports.formatOpsResult = exports.formatVerifyDeployResult = exports.formatPrepareClusterResult = exports.formatGenerateK8sManifestsResult = exports.formatPushImageResult = exports.formatTagImageResult = void 0;
const executor_js_1 = require("./executor.js");
// ===== TOOL AND SCHEMA IMPORTS =====
// Consolidated imports - tools and schemas from each module
// analyze-repo
const tool_1 = __importDefault(require("../tools/analyze-repo/tool"));
// generate-dockerfile
const tool_2 = __importDefault(require("../tools/generate-dockerfile/tool"));
// fix-dockerfile
const tool_3 = __importDefault(require("../tools/fix-dockerfile/tool"));
// build-image-context
const tool_4 = __importDefault(require("../tools/build-image-context/tool"));
// scan-image
const tool_5 = __importDefault(require("../tools/scan-image/tool"));
// tag-image
const tool_6 = __importDefault(require("../tools/tag-image/tool"));
// push-image
const tool_7 = __importDefault(require("../tools/push-image/tool"));
// generate-k8s-manifests
const tool_8 = __importDefault(require("../tools/generate-k8s-manifests/tool"));
// prepare-cluster
const tool_9 = __importDefault(require("../tools/prepare-cluster/tool"));
// verify-deploy
const tool_10 = __importDefault(require("../tools/verify-deploy/tool"));
// ops
const tool_11 = __importDefault(require("../tools/ops/tool"));
var core_1 = require("../types/core");
Object.defineProperty(exports, "Success", { enumerable: true, get: function () { return core_1.Success; } });
Object.defineProperty(exports, "Failure", { enumerable: true, get: function () { return core_1.Failure; } });
// Full type exports available via sdk/types
// import type { ... } from 'containerization-assist-mcp/sdk/types';
// ===== SDK FUNCTION FACTORY =====
/**
 * Create a typed SDK function from a tool.
 *
 * This factory ensures consistent behavior across all SDK functions
 * while preserving full type inference for inputs and outputs.
 *
 * @param tool - The tool to wrap
 * @returns A typed function that executes the tool
 */
function createSDKFunction(tool) {
    return (input, options) => (0, executor_js_1.executeTool)(tool, input, options);
}
// ===== SDK FUNCTION EXPORTS =====
// ----- Analysis Tools -----
/**
 * Analyze a repository to detect language, framework, and dependencies.
 */
exports.analyzeRepo = createSDKFunction(tool_1.default);
// ----- Dockerfile Tools -----
/**
 * Generate Dockerfile recommendations for a repository.
 */
exports.generateDockerfile = createSDKFunction(tool_2.default);
/**
 * Fix and optimize an existing Dockerfile.
 */
exports.fixDockerfile = createSDKFunction(tool_3.default);
// ----- Image Tools -----
/**
 * Prepare Docker build context with security analysis and optimized commands.
 * Returns structured guidance for executing builds - does not execute Docker directly.
 */
exports.buildImageContext = createSDKFunction(tool_4.default);
/**
 * Scan a Docker image for security vulnerabilities.
 * Requires Trivy to be installed for full functionality.
 */
exports.scanImage = createSDKFunction(tool_5.default);
/**
 * Tag a Docker image with additional tags.
 * Requires Docker daemon to be running.
 */
exports.tagImage = createSDKFunction(tool_6.default);
/**
 * Push a Docker image to a registry.
 * Requires Docker daemon and registry authentication.
 */
exports.pushImage = createSDKFunction(tool_7.default);
// ----- Kubernetes Tools -----
/**
 * Generate Kubernetes manifests for deployment.
 */
exports.generateK8sManifests = createSDKFunction(tool_8.default);
/**
 * Prepare a Kubernetes cluster for deployment.
 * Requires kubectl configured with cluster access.
 */
exports.prepareCluster = createSDKFunction(tool_9.default);
/**
 * Verify a Kubernetes deployment status.
 * Requires kubectl configured with cluster access.
 */
exports.verifyDeploy = createSDKFunction(tool_10.default);
// ----- Operational Tools -----
/**
 * Operational utilities (ping, status).
 */
exports.ops = createSDKFunction(tool_11.default);
// ===== ADVANCED: DIRECT TOOL ACCESS =====
/**
 * Direct access to all 11 tool objects for advanced use cases.
 *
 * Use this when you need:
 * - Access to tool schemas for validation
 * - Tool metadata and descriptions
 * - Custom execution patterns
 *
 * Uses `satisfies` to ensure type safety:
 * - TypeScript will error if a tool is missing
 * - TypeScript will error if a tool has the wrong type
 * - TypeScript will error if a key is misspelled
 *
 * @example
 * ```typescript
 * import { tools } from 'containerization-assist-mcp/sdk';
 *
 * // Access tool schema
 * const schema = tools.analyzeRepo.schema;
 *
 * // Access tool metadata
 * console.log(tools.buildImageContext.description);
 * ```
 */
exports.tools = {
    // ===== Analysis =====
    /** Analyze repository structure, detect languages, frameworks, and dependencies */
    analyzeRepo: tool_1.default,
    // ===== Dockerfile =====
    /** Generate optimized Dockerfile with security best practices */
    generateDockerfile: tool_2.default,
    /** Fix and optimize existing Dockerfile issues */
    fixDockerfile: tool_3.default,
    // ===== Image Operations =====
    /** Prepare Docker build context with security analysis (returns build commands) */
    buildImageContext: tool_4.default,
    /** Scan image for security vulnerabilities (requires Trivy) */
    scanImage: tool_5.default,
    /** Tag Docker image with additional tags */
    tagImage: tool_6.default,
    /** Push image to container registry */
    pushImage: tool_7.default,
    // ===== Kubernetes =====
    /** Generate Kubernetes deployment manifests */
    generateK8sManifests: tool_8.default,
    /** Prepare Kubernetes cluster namespace and prerequisites */
    prepareCluster: tool_9.default,
    /** Verify Kubernetes deployment status and health */
    verifyDeploy: tool_10.default,
    // ===== Operations =====
    /** Operational utilities (ping, status checks) */
    ops: tool_11.default,
};
/**
 * Execute any tool directly with full control.
 *
 * Use this when you need to:
 * - Execute tools not exposed via simplified functions
 * - Pass custom options to the executor
 * - Handle tool objects dynamically
 *
 * @example
 * ```typescript
 * import { executeTool, tools } from 'containerization-assist-mcp/sdk';
 *
 * const result = await executeTool(
 *   tools.analyzeRepo,
 *   { repositoryPath: '.' },
 *   { signal: controller.signal }
 * );
 * ```
 */
var executor_js_2 = require("./executor.js");
Object.defineProperty(exports, "executeTool", { enumerable: true, get: function () { return executor_js_2.executeTool; } });
// ===== JSON SCHEMA EXPORTS (for VS Code extension) =====
var schemas_js_1 = require("./schemas.js");
Object.defineProperty(exports, "jsonSchemas", { enumerable: true, get: function () { return schemas_js_1.jsonSchemas; } });
// Individual schemas for selective imports
Object.defineProperty(exports, "analyzeRepoJsonSchema", { enumerable: true, get: function () { return schemas_js_1.analyzeRepoJsonSchema; } });
Object.defineProperty(exports, "generateDockerfileJsonSchema", { enumerable: true, get: function () { return schemas_js_1.generateDockerfileJsonSchema; } });
Object.defineProperty(exports, "fixDockerfileJsonSchema", { enumerable: true, get: function () { return schemas_js_1.fixDockerfileJsonSchema; } });
Object.defineProperty(exports, "buildImageContextJsonSchema", { enumerable: true, get: function () { return schemas_js_1.buildImageContextJsonSchema; } });
Object.defineProperty(exports, "scanImageJsonSchema", { enumerable: true, get: function () { return schemas_js_1.scanImageJsonSchema; } });
Object.defineProperty(exports, "tagImageJsonSchema", { enumerable: true, get: function () { return schemas_js_1.tagImageJsonSchema; } });
Object.defineProperty(exports, "pushImageJsonSchema", { enumerable: true, get: function () { return schemas_js_1.pushImageJsonSchema; } });
Object.defineProperty(exports, "generateK8sManifestsJsonSchema", { enumerable: true, get: function () { return schemas_js_1.generateK8sManifestsJsonSchema; } });
Object.defineProperty(exports, "prepareClusterJsonSchema", { enumerable: true, get: function () { return schemas_js_1.prepareClusterJsonSchema; } });
Object.defineProperty(exports, "verifyDeployJsonSchema", { enumerable: true, get: function () { return schemas_js_1.verifyDeployJsonSchema; } });
Object.defineProperty(exports, "opsJsonSchema", { enumerable: true, get: function () { return schemas_js_1.opsJsonSchema; } });
// ===== TOOL METADATA EXPORTS (for VS Code extension) =====
var metadata_js_1 = require("./metadata.js");
Object.defineProperty(exports, "toolMetadata", { enumerable: true, get: function () { return metadata_js_1.toolMetadata; } });
Object.defineProperty(exports, "standardWorkflow", { enumerable: true, get: function () { return metadata_js_1.standardWorkflow; } });
Object.defineProperty(exports, "getToolsByCategory", { enumerable: true, get: function () { return metadata_js_1.getToolsByCategory; } });
// External dependency types and helpers
Object.defineProperty(exports, "ExternalDeps", { enumerable: true, get: function () { return metadata_js_1.ExternalDeps; } });
Object.defineProperty(exports, "getRequiredDepsString", { enumerable: true, get: function () { return metadata_js_1.getRequiredDepsString; } });
Object.defineProperty(exports, "requiresDependency", { enumerable: true, get: function () { return metadata_js_1.requiresDependency; } });
// Individual metadata for selective imports
Object.defineProperty(exports, "analyzeRepoMetadata", { enumerable: true, get: function () { return metadata_js_1.analyzeRepoMetadata; } });
Object.defineProperty(exports, "generateDockerfileMetadata", { enumerable: true, get: function () { return metadata_js_1.generateDockerfileMetadata; } });
Object.defineProperty(exports, "fixDockerfileMetadata", { enumerable: true, get: function () { return metadata_js_1.fixDockerfileMetadata; } });
Object.defineProperty(exports, "buildImageContextMetadata", { enumerable: true, get: function () { return metadata_js_1.buildImageContextMetadata; } });
Object.defineProperty(exports, "scanImageMetadata", { enumerable: true, get: function () { return metadata_js_1.scanImageMetadata; } });
Object.defineProperty(exports, "tagImageMetadata", { enumerable: true, get: function () { return metadata_js_1.tagImageMetadata; } });
Object.defineProperty(exports, "pushImageMetadata", { enumerable: true, get: function () { return metadata_js_1.pushImageMetadata; } });
Object.defineProperty(exports, "generateK8sManifestsMetadata", { enumerable: true, get: function () { return metadata_js_1.generateK8sManifestsMetadata; } });
Object.defineProperty(exports, "prepareClusterMetadata", { enumerable: true, get: function () { return metadata_js_1.prepareClusterMetadata; } });
Object.defineProperty(exports, "verifyDeployMetadata", { enumerable: true, get: function () { return metadata_js_1.verifyDeployMetadata; } });
Object.defineProperty(exports, "opsMetadata", { enumerable: true, get: function () { return metadata_js_1.opsMetadata; } });
// ===== RESULT FORMATTERS (for VS Code extension) =====
var formatters_js_1 = require("./formatters.js");
Object.defineProperty(exports, "resultFormatters", { enumerable: true, get: function () { return formatters_js_1.resultFormatters; } });
// Individual formatters for selective imports
Object.defineProperty(exports, "formatAnalyzeRepoResult", { enumerable: true, get: function () { return formatters_js_1.formatAnalyzeRepoResult; } });
Object.defineProperty(exports, "formatGenerateDockerfileResult", { enumerable: true, get: function () { return formatters_js_1.formatGenerateDockerfileResult; } });
Object.defineProperty(exports, "formatFixDockerfileResult", { enumerable: true, get: function () { return formatters_js_1.formatFixDockerfileResult; } });
Object.defineProperty(exports, "formatBuildImageResult", { enumerable: true, get: function () { return formatters_js_1.formatBuildImageResult; } });
Object.defineProperty(exports, "formatScanImageResult", { enumerable: true, get: function () { return formatters_js_1.formatScanImageResult; } });
Object.defineProperty(exports, "formatTagImageResult", { enumerable: true, get: function () { return formatters_js_1.formatTagImageResult; } });
Object.defineProperty(exports, "formatPushImageResult", { enumerable: true, get: function () { return formatters_js_1.formatPushImageResult; } });
Object.defineProperty(exports, "formatGenerateK8sManifestsResult", { enumerable: true, get: function () { return formatters_js_1.formatGenerateK8sManifestsResult; } });
Object.defineProperty(exports, "formatPrepareClusterResult", { enumerable: true, get: function () { return formatters_js_1.formatPrepareClusterResult; } });
Object.defineProperty(exports, "formatVerifyDeployResult", { enumerable: true, get: function () { return formatters_js_1.formatVerifyDeployResult; } });
Object.defineProperty(exports, "formatOpsResult", { enumerable: true, get: function () { return formatters_js_1.formatOpsResult; } });
// ===== VS CODE UTILITIES =====
var vscode_utils_js_1 = require("./vscode-utils.js");
Object.defineProperty(exports, "createAbortSignalFromToken", { enumerable: true, get: function () { return vscode_utils_js_1.createAbortSignalFromToken; } });
Object.defineProperty(exports, "formatErrorForLLM", { enumerable: true, get: function () { return vscode_utils_js_1.formatErrorForLLM; } });
Object.defineProperty(exports, "resolveWorkspacePath", { enumerable: true, get: function () { return vscode_utils_js_1.resolveWorkspacePath; } });
Object.defineProperty(exports, "validateRequiredFields", { enumerable: true, get: function () { return vscode_utils_js_1.validateRequiredFields; } });
Object.defineProperty(exports, "sanitizeForMarkdown", { enumerable: true, get: function () { return vscode_utils_js_1.sanitizeForMarkdown; } });
// ===== KNOWLEDGE BASE =====
/**
 * Load the embedded knowledge base and return all entries.
 *
 * Knowledge packs contain best practices for Dockerfile generation, Kubernetes
 * manifests, and security configurations. They are embedded at build time,
 * eliminating runtime filesystem dependencies.
 *
 * @returns Object containing all loaded knowledge entries
 *
 * @example
 * ```typescript
 * import { loadKnowledgeData } from 'containerization-assist-mcp/sdk';
 *
 * const { entries } = loadKnowledgeData();
 * console.log(`Loaded ${entries.length} knowledge entries`);
 * ```
 */
var index_1 = require("../knowledge/index");
Object.defineProperty(exports, "loadKnowledgeData", { enumerable: true, get: function () { return index_1.loadKnowledgeData; } });
