// containerization-assist-mcp/dist-cjs/src/sdk/schemas.js
"use strict";
/**
 * JSON Schema Exports for VS Code Extension Integration
 *
 * Provides JSON Schema versions of all tool input schemas, suitable for
 * use in VS Code package.json languageModelTools definitions.
 *
 * @example
 * ```typescript
 * import { jsonSchemas } from 'containerization-assist-mcp/sdk';
 *
 * // Use in package.json generation or validation
 * console.log(jsonSchemas.analyzeRepo);
 * ```
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.jsonSchemas = exports.opsJsonSchema = exports.verifyDeployJsonSchema = exports.prepareClusterJsonSchema = exports.generateK8sManifestsJsonSchema = exports.pushImageJsonSchema = exports.tagImageJsonSchema = exports.scanImageJsonSchema = exports.buildImageContextJsonSchema = exports.fixDockerfileJsonSchema = exports.generateDockerfileJsonSchema = exports.analyzeRepoJsonSchema = exports.convertZodToJsonSchema = void 0;
const json_schema_converter_js_1 = require("./json-schema-converter.js");
// Re-export for consumers who need to convert custom schemas
var json_schema_converter_js_2 = require("./json-schema-converter.js");
Object.defineProperty(exports, "convertZodToJsonSchema", { enumerable: true, get: function () { return json_schema_converter_js_2.convertZodToJsonSchema; } });
// Import all Zod schemas
const schema_1 = require("../tools/analyze-repo/schema");
const schema_2 = require("../tools/generate-dockerfile/schema");
const schema_3 = require("../tools/fix-dockerfile/schema");
const schema_4 = require("../tools/build-image-context/schema");
const schema_5 = require("../tools/scan-image/schema");
const schema_6 = require("../tools/tag-image/schema");
const schema_7 = require("../tools/push-image/schema");
const schema_8 = require("../tools/generate-k8s-manifests/schema");
const schema_9 = require("../tools/prepare-cluster/schema");
const schema_10 = require("../tools/verify-deploy/schema");
const schema_11 = require("../tools/ops/schema");
// ===== INDIVIDUAL SCHEMA EXPORTS =====
/**
 * JSON Schema for analyze-repo tool input.
 */
exports.analyzeRepoJsonSchema = (0, json_schema_converter_js_1.convertZodToJsonSchema)(schema_1.analyzeRepoSchema);
/**
 * JSON Schema for generate-dockerfile tool input.
 */
exports.generateDockerfileJsonSchema = (0, json_schema_converter_js_1.convertZodToJsonSchema)(schema_2.generateDockerfileSchema);
/**
 * JSON Schema for fix-dockerfile tool input.
 */
exports.fixDockerfileJsonSchema = (0, json_schema_converter_js_1.convertZodToJsonSchema)(schema_3.fixDockerfileSchema);
/**
 * JSON Schema for build-image-context tool input.
 */
exports.buildImageContextJsonSchema = (0, json_schema_converter_js_1.convertZodToJsonSchema)(schema_4.buildImageSchema);
/**
 * JSON Schema for scan-image tool input.
 */
exports.scanImageJsonSchema = (0, json_schema_converter_js_1.convertZodToJsonSchema)(schema_5.scanImageSchema);
/**
 * JSON Schema for tag-image tool input.
 */
exports.tagImageJsonSchema = (0, json_schema_converter_js_1.convertZodToJsonSchema)(schema_6.tagImageSchema);
/**
 * JSON Schema for push-image tool input.
 */
exports.pushImageJsonSchema = (0, json_schema_converter_js_1.convertZodToJsonSchema)(schema_7.pushImageSchema);
/**
 * JSON Schema for generate-k8s-manifests tool input.
 */
exports.generateK8sManifestsJsonSchema = (0, json_schema_converter_js_1.convertZodToJsonSchema)(schema_8.generateK8sManifestsSchema);
/**
 * JSON Schema for prepare-cluster tool input.
 */
exports.prepareClusterJsonSchema = (0, json_schema_converter_js_1.convertZodToJsonSchema)(schema_9.prepareClusterSchema);
/**
 * JSON Schema for verify-deploy tool input.
 */
exports.verifyDeployJsonSchema = (0, json_schema_converter_js_1.convertZodToJsonSchema)(schema_10.verifyDeploySchema);
/**
 * JSON Schema for ops tool input.
 */
exports.opsJsonSchema = (0, json_schema_converter_js_1.convertZodToJsonSchema)(schema_11.opsToolSchema);
// ===== AGGREGATED EXPORT =====
/**
 * All JSON Schemas for VS Code languageModelTools definitions.
 *
 * Use these schemas in your VS Code extension's package.json:
 *
 * @example
 * ```json
 * {
 *   "contributes": {
 *     "languageModelTools": [{
 *       "name": "analyze_repo",
 *       "inputSchema": { ... } // Use jsonSchemas.analyzeRepo
 *     }]
 *   }
 * }
 * ```
 */
exports.jsonSchemas = {
    analyzeRepo: exports.analyzeRepoJsonSchema,
    generateDockerfile: exports.generateDockerfileJsonSchema,
    fixDockerfile: exports.fixDockerfileJsonSchema,
    buildImageContext: exports.buildImageContextJsonSchema,
    scanImage: exports.scanImageJsonSchema,
    tagImage: exports.tagImageJsonSchema,
    pushImage: exports.pushImageJsonSchema,
    generateK8sManifests: exports.generateK8sManifestsJsonSchema,
    prepareCluster: exports.prepareClusterJsonSchema,
    verifyDeploy: exports.verifyDeployJsonSchema,
    ops: exports.opsJsonSchema,
};
