// containerization-assist-mcp/dist/src/tools/shared/knowledge-tool-pattern.js
/**
 * Knowledge-Based Tool Pattern
 *
 * Reusable pattern for tools that query the knowledge base and return
 * structured plans instead of making AI calls. This pattern enables:
 * - Fast, deterministic tool execution (<100ms)
 * - Zero AI calls within tools (client AI generates content using plans)
 * - Testable, consistent behavior
 * - Reduced code duplication (30-50% code reduction per tool)
 *
 * @module tools/shared/knowledge-tool-pattern
 */
import { Success } from '../../types/index.js';
import { getKnowledgeSnippets } from '../../knowledge/matcher.js';
import { getPolicyAwareKnowledgeSnippets } from '../../knowledge/policy-aware-matcher.js';
import { loadKnowledgeData } from '../../knowledge/loader.js';
/**
 * Default confidence calculation.
 * Starts at 0.5 and increases by 0.05 per match, capped at 0.95.
 */
export const defaultConfidenceCalculation = (matchCount) => {
    return matchCount > 0 ? Math.min(0.95, 0.5 + matchCount * 0.05) : 0.5;
};
/**
 * Create a knowledge-based tool runner from configuration.
 *
 * This factory function creates a tool `run` function that:
 * 1. Queries the knowledge base (fast, deterministic)
 * 2. Categorizes knowledge snippets
 * 3. Applies rule-based logic
 * 4. Calculates confidence
 * 5. Builds and returns a structured plan
 *
 * No AI calls are made - the tool returns data for the MCP client AI to use.
 *
 * @param config - Complete tool configuration
 * @returns Tool run function
 *
 * @example
 * ```typescript
 * const run = createKnowledgeTool({
 *   name: 'generate-dockerfile-plan',
 *   query: {
 *     topic: TOPICS.DOCKERFILE,
 *     category: CATEGORY.DOCKERFILE,
 *     maxChars: 8000,
 *     maxSnippets: 20,
 *     extractFilters: (input) => ({
 *       environment: input.environment || 'production',
 *       language: input.language,
 *       framework: input.framework,
 *     }),
 *   },
 *   categorization: {
 *     categoryNames: ['security', 'optimization', 'bestPractices'] as const,
 *     categorize: (snippet) => {
 *       const categories = [];
 *       if (snippet.category === 'security' || snippet.tags?.includes('security')) {
 *         categories.push('security');
 *       }
 *       // ... more categorization logic
 *       return categories;
 *     },
 *   },
 *   rules: {
 *     applyRules: (input, knowledge) => ({
 *       multistage: ['java', 'go', 'rust'].includes(input.language),
 *     }),
 *   },
 *   plan: {
 *     buildPlan: (input, knowledge, rules, confidence) => ({
 *       repositoryInfo: { ... },
 *       recommendations: { ... },
 *       knowledgeMatches: knowledge.all,
 *       confidence,
 *     }),
 *   },
 * });
 * ```
 */
export function createKnowledgeTool(config) {
    return async (input, ctx) => {
        // 1. Query knowledge base
        const filters = config.query.extractFilters(input);
        const topic = typeof config.query.topic === 'function' ? config.query.topic(input) : config.query.topic;
        ctx.logger.info({ topic, filters }, `${config.name}: Querying knowledge base`);
        const detectedDeps = filters.detectedDependencies;
        const hasDetectedDeps = Array.isArray(detectedDeps) && detectedDeps.length > 0;
        const knowledgeOptions = {
            environment: filters.environment || 'production',
            tool: config.name,
            maxChars: config.query.maxChars || 8000,
            maxSnippets: config.query.maxSnippets || 20,
            category: config.query.category,
            ...(filters.language && { language: filters.language }),
            ...(filters.framework && { framework: filters.framework }),
            ...(hasDetectedDeps && { detectedDependencies: detectedDeps }),
        };
        // Use policy-aware knowledge matching if policy is available
        let knowledgeSnippets;
        if (ctx.policy) {
            // Load knowledge data for policy-aware matching
            const knowledgeData = await loadKnowledgeData();
            // Build query for policy-aware matcher
            const query = {
                text: topic,
                category: knowledgeOptions.category,
                environment: knowledgeOptions.environment,
                tool: knowledgeOptions.tool,
                ...(knowledgeOptions.language && { language: knowledgeOptions.language }),
                ...(knowledgeOptions.framework && { framework: knowledgeOptions.framework }),
                tags: [
                    knowledgeOptions.tool,
                    ...(knowledgeOptions.language ? [knowledgeOptions.language] : []),
                    ...(knowledgeOptions.framework ? [knowledgeOptions.framework] : []),
                    ...(hasDetectedDeps ? detectedDeps : []),
                ],
                limit: knowledgeOptions.maxSnippets,
            };
            const { snippets, filterResult } = await getPolicyAwareKnowledgeSnippets(knowledgeData.entries, query, ctx.policy);
            // Log policy filtering results
            if (filterResult.policyApplied) {
                ctx.logger.info({
                    excluded: filterResult.excluded.length,
                    boosted: filterResult.boosted.length,
                    reduced: filterResult.reduced.length,
                    returned: filterResult.totalReturned,
                }, `${config.name}: Policy-aware knowledge filtering applied`);
            }
            knowledgeSnippets = snippets;
        }
        else {
            // Fall back to standard knowledge matching
            knowledgeSnippets = await getKnowledgeSnippets(topic, knowledgeOptions);
        }
        // 2. Categorize knowledge snippets
        // Initialize empty arrays for each category
        const categoriesRecord = {};
        for (const categoryName of config.categorization.categoryNames) {
            categoriesRecord[categoryName] = [];
        }
        // Categorize each snippet (snippets can belong to multiple categories)
        for (const snippet of knowledgeSnippets) {
            const categories = config.categorization.categorize(snippet, input);
            for (const category of categories) {
                const categoryArray = categoriesRecord[category];
                if (categoryArray) {
                    categoryArray.push(snippet);
                }
                else {
                    throw new Error(`categorize() returned undeclared category "${category}" for tool "${config.name}". ` +
                        `Snippet ID: ${snippet.id}, Source: ${snippet.source}. ` +
                        `This likely indicates a configuration error.`);
                }
            }
        }
        const categorized = {
            all: knowledgeSnippets,
            categories: categoriesRecord,
        };
        // 3. Apply rule-based logic
        const rules = await config.rules.applyRules(input, categorized, ctx);
        // 4. Calculate confidence
        const calculateConfidence = config.confidence?.calculateConfidence || defaultConfidenceCalculation;
        const confidence = calculateConfidence(knowledgeSnippets.length);
        // 5. Build plan
        const plan = await config.plan.buildPlan(input, categorized, rules, confidence, ctx);
        // Log completion
        const categoryCounts = Object.entries(categorized.categories).map(([name, snippets]) => `${name}: ${snippets.length}`);
        ctx.logger.info({
            knowledgeMatchCount: knowledgeSnippets.length,
            confidence,
            categories: categoryCounts,
        }, `${config.name}: Planning completed`);
        return Success(plan);
    };
}
/**
 * Helper: Create a simple categorizer based on tags and category fields.
 *
 * This is a common pattern where snippets are categorized by checking
 * if they have specific tags or category values.
 *
 * @param rules - Mapping of category names to filter predicates
 * @returns Categorization function
 *
 * @example
 * ```typescript
 * const categorize = createSimpleCategorizer({
 *   security: (s) => s.category === 'security' || s.tags?.includes('security'),
 *   optimization: (s) => s.tags?.includes('optimization') || s.tags?.includes('caching'),
 *   bestPractices: (s) => true, // catch-all
 * });
 * ```
 */
export function createSimpleCategorizer(rules) {
    return (snippet) => {
        const categories = [];
        for (const [category, predicate] of Object.entries(rules)) {
            if (predicate(snippet)) {
                categories.push(category);
            }
        }
        return categories;
    };
}
