// @azure/arm-keyvault/dist/react-native/operations/keys.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { setContinuationToken } from "../pagingHelper.js";
import * as coreClient from "@azure/core-client";
import * as Mappers from "../models/mappers.js";
import * as Parameters from "../models/parameters.js";
/// <reference lib="esnext.asynciterable" />
/** Class containing Keys operations. */
export class KeysImpl {
    client;
    /**
     * Initialize a new instance of the class Keys class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * Lists the keys in the specified key vault.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param vaultName The name of the vault which contains the key to be retrieved.
     * @param options The options parameters.
     */
    list(resourceGroupName, vaultName, options) {
        const iter = this.listPagingAll(resourceGroupName, vaultName, options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings?.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.listPagingPage(resourceGroupName, vaultName, options, settings);
            },
        };
    }
    async *listPagingPage(resourceGroupName, vaultName, options, settings) {
        let result;
        let continuationToken = settings?.continuationToken;
        if (!continuationToken) {
            result = await this._list(resourceGroupName, vaultName, options);
            const page = result.value || [];
            continuationToken = result.nextLink;
            setContinuationToken(page, continuationToken);
            yield page;
        }
        while (continuationToken) {
            result = await this._listNext(resourceGroupName, vaultName, continuationToken, options);
            continuationToken = result.nextLink;
            const page = result.value || [];
            setContinuationToken(page, continuationToken);
            yield page;
        }
    }
    async *listPagingAll(resourceGroupName, vaultName, options) {
        for await (const page of this.listPagingPage(resourceGroupName, vaultName, options)) {
            yield* page;
        }
    }
    /**
     * Lists the keys in the specified key vault.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param vaultName The name of the vault which contains the key version to be retrieved.
     * @param keyName The name of the key version to be retrieved.
     * @param options The options parameters.
     */
    listVersions(resourceGroupName, vaultName, keyName, options) {
        const iter = this.listVersionsPagingAll(resourceGroupName, vaultName, keyName, options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings?.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.listVersionsPagingPage(resourceGroupName, vaultName, keyName, options, settings);
            },
        };
    }
    async *listVersionsPagingPage(resourceGroupName, vaultName, keyName, options, settings) {
        let result;
        let continuationToken = settings?.continuationToken;
        if (!continuationToken) {
            result = await this._listVersions(resourceGroupName, vaultName, keyName, options);
            const page = result.value || [];
            continuationToken = result.nextLink;
            setContinuationToken(page, continuationToken);
            yield page;
        }
        while (continuationToken) {
            result = await this._listVersionsNext(resourceGroupName, vaultName, keyName, continuationToken, options);
            continuationToken = result.nextLink;
            const page = result.value || [];
            setContinuationToken(page, continuationToken);
            yield page;
        }
    }
    async *listVersionsPagingAll(resourceGroupName, vaultName, keyName, options) {
        for await (const page of this.listVersionsPagingPage(resourceGroupName, vaultName, keyName, options)) {
            yield* page;
        }
    }
    /**
     * Lists the keys in the specified key vault.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param vaultName The name of the vault which contains the key to be retrieved.
     * @param options The options parameters.
     */
    _list(resourceGroupName, vaultName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, vaultName, options }, listOperationSpec);
    }
    /**
     * Gets the current version of the specified key from the specified key vault.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param vaultName The name of the vault which contains the key to be retrieved.
     * @param keyName The name of the key to be retrieved.
     * @param options The options parameters.
     */
    get(resourceGroupName, vaultName, keyName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, vaultName, keyName, options }, getOperationSpec);
    }
    /**
     * Creates the first version of a new key if it does not exist. If it already exists, then the existing
     * key is returned without any write operations being performed. This API does not create subsequent
     * versions, and does not update existing keys.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param vaultName The name of the vault which contains the key to be retrieved.
     * @param keyName The name of the key to be retrieved.
     * @param parameters The parameters used to create the specified key.
     * @param options The options parameters.
     */
    createIfNotExist(resourceGroupName, vaultName, keyName, parameters, options) {
        return this.client.sendOperationRequest({ resourceGroupName, vaultName, keyName, parameters, options }, createIfNotExistOperationSpec);
    }
    /**
     * Lists the keys in the specified key vault.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param vaultName The name of the vault which contains the key version to be retrieved.
     * @param keyName The name of the key version to be retrieved.
     * @param options The options parameters.
     */
    _listVersions(resourceGroupName, vaultName, keyName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, vaultName, keyName, options }, listVersionsOperationSpec);
    }
    /**
     * Gets the specified version of the specified key in the specified key vault.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param vaultName The name of the vault which contains the key version to be retrieved.
     * @param keyName The name of the key version to be retrieved.
     * @param keyVersion The version of the key to be retrieved.
     * @param options The options parameters.
     */
    getVersion(resourceGroupName, vaultName, keyName, keyVersion, options) {
        return this.client.sendOperationRequest({ resourceGroupName, vaultName, keyName, keyVersion, options }, getVersionOperationSpec);
    }
    /**
     * ListNext
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param vaultName The name of the vault which contains the key to be retrieved.
     * @param nextLink The nextLink from the previous successful call to the List method.
     * @param options The options parameters.
     */
    _listNext(resourceGroupName, vaultName, nextLink, options) {
        return this.client.sendOperationRequest({ resourceGroupName, vaultName, nextLink, options }, listNextOperationSpec);
    }
    /**
     * ListVersionsNext
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param vaultName The name of the vault which contains the key version to be retrieved.
     * @param keyName The name of the key version to be retrieved.
     * @param nextLink The nextLink from the previous successful call to the ListVersions method.
     * @param options The options parameters.
     */
    _listVersionsNext(resourceGroupName, vaultName, keyName, nextLink, options) {
        return this.client.sendOperationRequest({ resourceGroupName, vaultName, keyName, nextLink, options }, listVersionsNextOperationSpec);
    }
}
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const listOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.KeyVault/vaults/{vaultName}/keys",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.KeyListResult,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.vaultName2,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const getOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.KeyVault/vaults/{vaultName}/keys/{keyName}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.Key,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.vaultName2,
        Parameters.keyName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const createIfNotExistOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.KeyVault/vaults/{vaultName}/keys/{keyName}",
    httpMethod: "PUT",
    responses: {
        200: {
            bodyMapper: Mappers.Key,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    requestBody: Parameters.parameters5,
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.vaultName2,
        Parameters.keyName,
    ],
    headerParameters: [Parameters.accept, Parameters.contentType],
    mediaType: "json",
    serializer,
};
const listVersionsOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.KeyVault/vaults/{vaultName}/keys/{keyName}/versions",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.KeyListResult,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.vaultName2,
        Parameters.keyName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const getVersionOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.KeyVault/vaults/{vaultName}/keys/{keyName}/versions/{keyVersion}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.Key,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.vaultName2,
        Parameters.keyName,
        Parameters.keyVersion,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listNextOperationSpec = {
    path: "{nextLink}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.KeyListResult,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    urlParameters: [
        Parameters.$host,
        Parameters.nextLink,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.vaultName2,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listVersionsNextOperationSpec = {
    path: "{nextLink}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.KeyListResult,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    urlParameters: [
        Parameters.$host,
        Parameters.nextLink,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.vaultName2,
        Parameters.keyName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
