// @azure/arm-keyvault/dist/esm/operations/mhsmPrivateEndpointConnections.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { setContinuationToken } from "../pagingHelper.js";
import * as coreClient from "@azure/core-client";
import * as Mappers from "../models/mappers.js";
import * as Parameters from "../models/parameters.js";
import { createHttpPoller } from "@azure/core-lro";
import { createLroSpec } from "../lroImpl.js";
/// <reference lib="esnext.asynciterable" />
/** Class containing MhsmPrivateEndpointConnections operations. */
export class MhsmPrivateEndpointConnectionsImpl {
    client;
    /**
     * Initialize a new instance of the class MhsmPrivateEndpointConnections class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * The List operation gets information about the private endpoint connections associated with the
     * managed HSM Pool.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param name The name of the managed HSM Pool.
     * @param options The options parameters.
     */
    listByResource(resourceGroupName, name, options) {
        const iter = this.listByResourcePagingAll(resourceGroupName, name, options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings?.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.listByResourcePagingPage(resourceGroupName, name, options, settings);
            },
        };
    }
    async *listByResourcePagingPage(resourceGroupName, name, options, settings) {
        let result;
        let continuationToken = settings?.continuationToken;
        if (!continuationToken) {
            result = await this._listByResource(resourceGroupName, name, options);
            const page = result.value || [];
            continuationToken = result.nextLink;
            setContinuationToken(page, continuationToken);
            yield page;
        }
        while (continuationToken) {
            result = await this._listByResourceNext(resourceGroupName, name, continuationToken, options);
            continuationToken = result.nextLink;
            const page = result.value || [];
            setContinuationToken(page, continuationToken);
            yield page;
        }
    }
    async *listByResourcePagingAll(resourceGroupName, name, options) {
        for await (const page of this.listByResourcePagingPage(resourceGroupName, name, options)) {
            yield* page;
        }
    }
    /**
     * The List operation gets information about the private endpoint connections associated with the
     * managed HSM Pool.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param name The name of the managed HSM Pool.
     * @param options The options parameters.
     */
    _listByResource(resourceGroupName, name, options) {
        return this.client.sendOperationRequest({ resourceGroupName, name, options }, listByResourceOperationSpec);
    }
    /**
     * Gets the specified private endpoint connection associated with the managed HSM Pool.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param name The name of the managed HSM Pool.
     * @param privateEndpointConnectionName Name of the private endpoint connection associated with the
     *                                      managed hsm pool.
     * @param options The options parameters.
     */
    get(resourceGroupName, name, privateEndpointConnectionName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, name, privateEndpointConnectionName, options }, getOperationSpec);
    }
    /**
     * Updates the specified private endpoint connection associated with the managed hsm pool.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param name The name of the managed HSM Pool.
     * @param privateEndpointConnectionName Name of the private endpoint connection associated with the
     *                                      managed hsm pool.
     * @param properties The intended state of private endpoint connection.
     * @param options The options parameters.
     */
    put(resourceGroupName, name, privateEndpointConnectionName, properties, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            name,
            privateEndpointConnectionName,
            properties,
            options,
        }, putOperationSpec);
    }
    /**
     * Deletes the specified private endpoint connection associated with the managed hsm pool.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param name The name of the managed HSM Pool.
     * @param privateEndpointConnectionName Name of the private endpoint connection associated with the
     *                                      managed hsm pool.
     * @param options The options parameters.
     */
    async beginDelete(resourceGroupName, name, privateEndpointConnectionName, options) {
        const directSendOperation = async (args, spec) => {
            return this.client.sendOperationRequest(args, spec);
        };
        const sendOperationFn = async (args, spec) => {
            let currentRawResponse = undefined;
            const providedCallback = args.options?.onResponse;
            const callback = (rawResponse, flatResponse) => {
                currentRawResponse = rawResponse;
                providedCallback?.(rawResponse, flatResponse);
            };
            const updatedArgs = {
                ...args,
                options: {
                    ...args.options,
                    onResponse: callback,
                },
            };
            const flatResponse = await directSendOperation(updatedArgs, spec);
            return {
                flatResponse,
                rawResponse: {
                    statusCode: currentRawResponse.status,
                    body: currentRawResponse.parsedBody,
                    headers: currentRawResponse.headers.toJSON(),
                },
            };
        };
        const lro = createLroSpec({
            sendOperationFn,
            args: { resourceGroupName, name, privateEndpointConnectionName, options },
            spec: deleteOperationSpec,
        });
        const poller = await createHttpPoller(lro, {
            restoreFrom: options?.resumeFrom,
            intervalInMs: options?.updateIntervalInMs,
            resourceLocationConfig: "location",
        });
        await poller.poll();
        return poller;
    }
    /**
     * Deletes the specified private endpoint connection associated with the managed hsm pool.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param name The name of the managed HSM Pool.
     * @param privateEndpointConnectionName Name of the private endpoint connection associated with the
     *                                      managed hsm pool.
     * @param options The options parameters.
     */
    async beginDeleteAndWait(resourceGroupName, name, privateEndpointConnectionName, options) {
        const poller = await this.beginDelete(resourceGroupName, name, privateEndpointConnectionName, options);
        return poller.pollUntilDone();
    }
    /**
     * ListByResourceNext
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param name The name of the managed HSM Pool.
     * @param nextLink The nextLink from the previous successful call to the ListByResource method.
     * @param options The options parameters.
     */
    _listByResourceNext(resourceGroupName, name, nextLink, options) {
        return this.client.sendOperationRequest({ resourceGroupName, name, nextLink, options }, listByResourceNextOperationSpec);
    }
}
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const listByResourceOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.KeyVault/managedHSMs/{name}/privateEndpointConnections",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.MhsmPrivateEndpointConnectionsListResult,
        },
        default: {
            bodyMapper: Mappers.ManagedHsmError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.name,
        Parameters.resourceGroupName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const getOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.KeyVault/managedHSMs/{name}/privateEndpointConnections/{privateEndpointConnectionName}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.MhsmPrivateEndpointConnection,
        },
        default: {
            bodyMapper: Mappers.ManagedHsmError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.name,
        Parameters.resourceGroupName,
        Parameters.privateEndpointConnectionName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const putOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.KeyVault/managedHSMs/{name}/privateEndpointConnections/{privateEndpointConnectionName}",
    httpMethod: "PUT",
    responses: {
        200: {
            bodyMapper: Mappers.MhsmPrivateEndpointConnection,
            headersMapper: Mappers.MhsmPrivateEndpointConnectionsPutHeaders,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    requestBody: Parameters.properties,
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.name,
        Parameters.resourceGroupName,
        Parameters.privateEndpointConnectionName,
    ],
    headerParameters: [Parameters.accept, Parameters.contentType],
    mediaType: "json",
    serializer,
};
const deleteOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.KeyVault/managedHSMs/{name}/privateEndpointConnections/{privateEndpointConnectionName}",
    httpMethod: "DELETE",
    responses: {
        200: {
            bodyMapper: Mappers.MhsmPrivateEndpointConnection,
        },
        201: {
            bodyMapper: Mappers.MhsmPrivateEndpointConnection,
        },
        202: {
            bodyMapper: Mappers.MhsmPrivateEndpointConnection,
        },
        204: {
            bodyMapper: Mappers.MhsmPrivateEndpointConnection,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.name,
        Parameters.resourceGroupName,
        Parameters.privateEndpointConnectionName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listByResourceNextOperationSpec = {
    path: "{nextLink}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.MhsmPrivateEndpointConnectionsListResult,
        },
        default: {
            bodyMapper: Mappers.ManagedHsmError,
        },
    },
    urlParameters: [
        Parameters.$host,
        Parameters.nextLink,
        Parameters.subscriptionId,
        Parameters.name,
        Parameters.resourceGroupName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
