// containerization-assist-mcp/dist-cjs/src/tools/prepare-cluster/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.prepareClusterSchema = exports.clusterTypeSchema = void 0;
const zod_1 = require("zod");
const constants_1 = require("../../config/constants");
const schemas_1 = require("../shared/schemas");
/**
 * Cluster type determines infrastructure setup behavior.
 * - `kind`: Create/manage a local Kind cluster with local Docker registry
 * - `generic`: Assume an existing cluster (AKS, EKS, GKE, minikube, etc.)
 */
exports.clusterTypeSchema = zod_1.z
    .enum(['kind', 'generic'])
    .optional()
    .describe('Cluster type to prepare. "kind" creates a local Kind cluster with a local Docker registry. "generic" assumes an existing cluster (AKS, EKS, GKE, minikube, etc.). Defaults to inferring from environment if omitted.');
exports.prepareClusterSchema = zod_1.z.object({
    clusterType: exports.clusterTypeSchema,
    environment: constants_1.environmentSchema
        .optional()
        .describe('Target environment for knowledge filtering and policy context. Does not control cluster setup — use clusterType for that.'),
    namespace: zod_1.z.string().optional().describe('Kubernetes namespace'),
    targetPlatform: schemas_1.platform.describe('Target platform for cluster validation. Ensures the cluster can run images built for this platform. Defaults to linux/amd64.'),
    strictPlatformValidation: zod_1.z
        .boolean()
        .optional()
        .default(true)
        .describe('Fail if cluster architecture does not match target platform. When true, prevents deployment to incompatible clusters. Set to false to allow emulation (may have performance impact).'),
});
