// @azure/arm-msi/dist/commonjs/models/index.js
"use strict";
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.KnownIsolationScope = exports.KnownCreatedByType = void 0;
/** Known values of {@link CreatedByType} that the service accepts. */
var KnownCreatedByType;
(function (KnownCreatedByType) {
    /** User */
    KnownCreatedByType["User"] = "User";
    /** Application */
    KnownCreatedByType["Application"] = "Application";
    /** ManagedIdentity */
    KnownCreatedByType["ManagedIdentity"] = "ManagedIdentity";
    /** Key */
    KnownCreatedByType["Key"] = "Key";
})(KnownCreatedByType || (exports.KnownCreatedByType = KnownCreatedByType = {}));
/** Known values of {@link IsolationScope} that the service accepts. */
var KnownIsolationScope;
(function (KnownIsolationScope) {
    /** None */
    KnownIsolationScope["None"] = "None";
    /** Regional */
    KnownIsolationScope["Regional"] = "Regional";
})(KnownIsolationScope || (exports.KnownIsolationScope = KnownIsolationScope = {}));
