// @azure/arm-msi/dist/commonjs/models/mappers.js
"use strict";
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Identity = exports.FederatedIdentityCredential = exports.SystemAssignedIdentity = exports.IdentityUpdate = exports.TrackedResource = exports.ProxyResource = exports.FederatedIdentityCredentialsListResult = exports.UserAssignedIdentitiesListResult = exports.OperationDisplay = exports.Operation = exports.OperationListResult = exports.CloudErrorBody = exports.CloudError = exports.SystemData = exports.Resource = void 0;
exports.Resource = {
    type: {
        name: "Composite",
        className: "Resource",
        modelProperties: {
            id: {
                serializedName: "id",
                readOnly: true,
                type: {
                    name: "String",
                },
            },
            name: {
                serializedName: "name",
                readOnly: true,
                type: {
                    name: "String",
                },
            },
            type: {
                serializedName: "type",
                readOnly: true,
                type: {
                    name: "String",
                },
            },
            systemData: {
                serializedName: "systemData",
                type: {
                    name: "Composite",
                    className: "SystemData",
                },
            },
        },
    },
};
exports.SystemData = {
    type: {
        name: "Composite",
        className: "SystemData",
        modelProperties: {
            createdBy: {
                serializedName: "createdBy",
                type: {
                    name: "String",
                },
            },
            createdByType: {
                serializedName: "createdByType",
                type: {
                    name: "String",
                },
            },
            createdAt: {
                serializedName: "createdAt",
                type: {
                    name: "DateTime",
                },
            },
            lastModifiedBy: {
                serializedName: "lastModifiedBy",
                type: {
                    name: "String",
                },
            },
            lastModifiedByType: {
                serializedName: "lastModifiedByType",
                type: {
                    name: "String",
                },
            },
            lastModifiedAt: {
                serializedName: "lastModifiedAt",
                type: {
                    name: "DateTime",
                },
            },
        },
    },
};
exports.CloudError = {
    type: {
        name: "Composite",
        className: "CloudError",
        modelProperties: {
            error: {
                serializedName: "error",
                type: {
                    name: "Composite",
                    className: "CloudErrorBody",
                },
            },
        },
    },
};
exports.CloudErrorBody = {
    type: {
        name: "Composite",
        className: "CloudErrorBody",
        modelProperties: {
            code: {
                serializedName: "code",
                type: {
                    name: "String",
                },
            },
            message: {
                serializedName: "message",
                type: {
                    name: "String",
                },
            },
            target: {
                serializedName: "target",
                type: {
                    name: "String",
                },
            },
            details: {
                serializedName: "details",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "CloudErrorBody",
                        },
                    },
                },
            },
        },
    },
};
exports.OperationListResult = {
    type: {
        name: "Composite",
        className: "OperationListResult",
        modelProperties: {
            value: {
                serializedName: "value",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "Operation",
                        },
                    },
                },
            },
            nextLink: {
                serializedName: "nextLink",
                type: {
                    name: "String",
                },
            },
        },
    },
};
exports.Operation = {
    type: {
        name: "Composite",
        className: "Operation",
        modelProperties: {
            name: {
                serializedName: "name",
                type: {
                    name: "String",
                },
            },
            display: {
                serializedName: "display",
                type: {
                    name: "Composite",
                    className: "OperationDisplay",
                },
            },
        },
    },
};
exports.OperationDisplay = {
    type: {
        name: "Composite",
        className: "OperationDisplay",
        modelProperties: {
            provider: {
                serializedName: "provider",
                type: {
                    name: "String",
                },
            },
            operation: {
                serializedName: "operation",
                type: {
                    name: "String",
                },
            },
            resource: {
                serializedName: "resource",
                type: {
                    name: "String",
                },
            },
            description: {
                serializedName: "description",
                type: {
                    name: "String",
                },
            },
        },
    },
};
exports.UserAssignedIdentitiesListResult = {
    type: {
        name: "Composite",
        className: "UserAssignedIdentitiesListResult",
        modelProperties: {
            value: {
                serializedName: "value",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "Identity",
                        },
                    },
                },
            },
            nextLink: {
                serializedName: "nextLink",
                type: {
                    name: "String",
                },
            },
        },
    },
};
exports.FederatedIdentityCredentialsListResult = {
    type: {
        name: "Composite",
        className: "FederatedIdentityCredentialsListResult",
        modelProperties: {
            value: {
                serializedName: "value",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "FederatedIdentityCredential",
                        },
                    },
                },
            },
            nextLink: {
                serializedName: "nextLink",
                type: {
                    name: "String",
                },
            },
        },
    },
};
exports.ProxyResource = {
    type: {
        name: "Composite",
        className: "ProxyResource",
        modelProperties: {
            ...exports.Resource.type.modelProperties,
        },
    },
};
exports.TrackedResource = {
    type: {
        name: "Composite",
        className: "TrackedResource",
        modelProperties: {
            ...exports.Resource.type.modelProperties,
            tags: {
                serializedName: "tags",
                type: {
                    name: "Dictionary",
                    value: { type: { name: "String" } },
                },
            },
            location: {
                serializedName: "location",
                required: true,
                type: {
                    name: "String",
                },
            },
        },
    },
};
exports.IdentityUpdate = {
    type: {
        name: "Composite",
        className: "IdentityUpdate",
        modelProperties: {
            ...exports.Resource.type.modelProperties,
            location: {
                serializedName: "location",
                type: {
                    name: "String",
                },
            },
            tags: {
                serializedName: "tags",
                type: {
                    name: "Dictionary",
                    value: { type: { name: "String" } },
                },
            },
            tenantId: {
                serializedName: "properties.tenantId",
                readOnly: true,
                type: {
                    name: "Uuid",
                },
            },
            principalId: {
                serializedName: "properties.principalId",
                readOnly: true,
                type: {
                    name: "Uuid",
                },
            },
            clientId: {
                serializedName: "properties.clientId",
                readOnly: true,
                type: {
                    name: "Uuid",
                },
            },
            isolationScope: {
                serializedName: "properties.isolationScope",
                type: {
                    name: "String",
                },
            },
        },
    },
};
exports.SystemAssignedIdentity = {
    type: {
        name: "Composite",
        className: "SystemAssignedIdentity",
        modelProperties: {
            ...exports.ProxyResource.type.modelProperties,
            location: {
                serializedName: "location",
                required: true,
                type: {
                    name: "String",
                },
            },
            tags: {
                serializedName: "tags",
                type: {
                    name: "Dictionary",
                    value: { type: { name: "String" } },
                },
            },
            tenantId: {
                serializedName: "properties.tenantId",
                readOnly: true,
                type: {
                    name: "Uuid",
                },
            },
            principalId: {
                serializedName: "properties.principalId",
                readOnly: true,
                type: {
                    name: "Uuid",
                },
            },
            clientId: {
                serializedName: "properties.clientId",
                readOnly: true,
                type: {
                    name: "Uuid",
                },
            },
            clientSecretUrl: {
                serializedName: "properties.clientSecretUrl",
                readOnly: true,
                type: {
                    name: "String",
                },
            },
        },
    },
};
exports.FederatedIdentityCredential = {
    type: {
        name: "Composite",
        className: "FederatedIdentityCredential",
        modelProperties: {
            ...exports.ProxyResource.type.modelProperties,
            issuer: {
                serializedName: "properties.issuer",
                type: {
                    name: "String",
                },
            },
            subject: {
                serializedName: "properties.subject",
                type: {
                    name: "String",
                },
            },
            audiences: {
                serializedName: "properties.audiences",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "String",
                        },
                    },
                },
            },
        },
    },
};
exports.Identity = {
    type: {
        name: "Composite",
        className: "Identity",
        modelProperties: {
            ...exports.TrackedResource.type.modelProperties,
            tenantId: {
                serializedName: "properties.tenantId",
                readOnly: true,
                type: {
                    name: "Uuid",
                },
            },
            principalId: {
                serializedName: "properties.principalId",
                readOnly: true,
                type: {
                    name: "Uuid",
                },
            },
            clientId: {
                serializedName: "properties.clientId",
                readOnly: true,
                type: {
                    name: "Uuid",
                },
            },
            isolationScope: {
                serializedName: "properties.isolationScope",
                type: {
                    name: "String",
                },
            },
        },
    },
};
