// containerization-assist-mcp/dist-cjs/src/mcp/context.js
"use strict";
/**
 * MCP Context - Tool execution environment abstraction
 *
 * Invariant: All tools receive consistent context interface
 * Trade-off: Abstraction overhead for tool isolation and testability
 * Design: Factory pattern enables context mocking in tests
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createProgressReporter = exports.extractProgressToken = void 0;
exports.createToolContext = createToolContext;
const context_helpers_js_1 = require("./context-helpers.js");
var context_helpers_js_2 = require("./context-helpers.js");
Object.defineProperty(exports, "extractProgressToken", { enumerable: true, get: function () { return context_helpers_js_2.extractProgressToken; } });
Object.defineProperty(exports, "createProgressReporter", { enumerable: true, get: function () { return context_helpers_js_2.createProgressReporter; } });
/**
 * Create a ToolContext for tool execution
 *
 * @param logger - Logger for debugging and error tracking
 * @param options - Optional configuration
 * @returns Configured ToolContext
 *
 * @example
 * ```typescript
 * const context = createToolContext(logger, {
 *   signal: abortController.signal,
 *   progress: async (msg) => console.log(msg),
 * });
 * ```
 */
function createToolContext(logger, options = {}) {
    const progressReporter = (0, context_helpers_js_1.extractProgressReporter)(options.progress, logger, options.sendNotification);
    return {
        logger,
        signal: options.signal,
        progress: progressReporter,
        ...(options.policy && { policy: options.policy }),
        queryConfig: async (packageName, input) => {
            if (!options.policy) {
                logger.debug({ packageName }, 'No policy configured, returning null for config query');
                return null;
            }
            return options.policy.queryConfig(packageName, input);
        },
    };
}
