// containerization-assist-mcp/dist-cjs/src/mcp/context-helpers.js
"use strict";
/**
 * MCP Context Helper Functions
 *
 * Helper functions for progress handling and context utilities.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractProgressToken = extractProgressToken;
exports.createProgressReporter = createProgressReporter;
exports.extractProgressReporter = extractProgressReporter;
// ===== TYPE GUARDS =====
/**
 * Type guard to check if value is a record object (non-null, non-array object)
 */
function isRecord(value) {
    return typeof value === 'object' && value !== null && !Array.isArray(value);
}
/**
 * Extract progress token from MCP request
 *
 * Per MCP spec, progress token is in params._meta.progressToken
 * @see https://modelcontextprotocol.io/specification#progress
 */
function extractProgressToken(request) {
    if (!isRecord(request)) {
        return undefined;
    }
    const params = request.params;
    if (!isRecord(params)) {
        return undefined;
    }
    const meta = params._meta;
    if (!isRecord(meta)) {
        return undefined;
    }
    const token = meta.progressToken;
    return typeof token === 'string' ? token : undefined;
}
/**
 * Creates a progress reporter that forwards notifications through the MCP protocol
 */
function createProgressReporter(progressToken, logger, sendNotification) {
    if (!progressToken) {
        return undefined;
    }
    return async (message, progress, total, metadata) => {
        try {
            const notification = {
                progressToken: String(progressToken),
                message,
                ...(progress !== undefined && { progress }),
                ...(total !== undefined && { total }),
                ...(metadata && { metadata }),
            };
            await sendProgressNotification(notification, logger, sendNotification);
        }
        catch (error) {
            logger?.warn({
                progressToken,
                message,
                error: error instanceof Error ? error.message : String(error),
            }, 'Failed to send progress notification');
        }
    };
}
/**
 * Sends a progress notification through the MCP server using the proper MCP protocol.
 * Uses sendNotification callback if available (from request handler), otherwise falls back to logging.
 */
async function sendProgressNotification(notification, logger, sendNotification) {
    // If we have a real sendNotification callback from the MCP request handler, use it
    if (sendNotification) {
        try {
            await sendNotification({
                method: 'notifications/progress',
                params: {
                    progressToken: notification.progressToken,
                    progress: notification.progress || 0,
                    ...(notification.total !== undefined && { total: notification.total }),
                    ...(notification.message && { message: notification.message }),
                    ...(notification.metadata && { ...notification.metadata }),
                },
            });
            logger?.debug({
                progressToken: notification.progressToken,
                message: notification.message,
                progress: notification.progress,
                total: notification.total,
                type: 'progress_notification',
            }, 'Progress notification sent via MCP protocol');
        }
        catch (error) {
            logger?.warn({
                error: error instanceof Error ? error.message : String(error),
                progressToken: notification.progressToken,
            }, 'Failed to send MCP progress notification');
        }
    }
    else {
        // Fallback to logging if no notification callback available
        logger?.debug({
            progressToken: notification.progressToken,
            message: notification.message,
            progress: notification.progress,
            total: notification.total,
            metadata: notification.metadata,
            type: 'progress_notification',
        }, 'Progress notification logged - no sendNotification callback available');
    }
}
/**
 * Extract progress reporter from progress token.
 *
 * Takes a simple progress token (string/number) and creates a progress reporter
 * that sends MCP notifications.
 *
 * @param progress - Progress token from MCP protocol
 * @param logger - Logger for creating progress reporter
 * @param sendNotification - MCP notification callback for progress updates
 * @returns ProgressReporter function or undefined
 */
function extractProgressReporter(progress, logger, sendNotification) {
    if (!progress)
        return undefined;
    // Simple progress token (string or number) - wrap it in a reporter
    if (typeof progress === 'string' || typeof progress === 'number') {
        return createProgressReporter(progress, logger, sendNotification);
    }
    return undefined;
}
