// containerization-assist-mcp/dist-cjs/src/mcp/mcp-server.js
"use strict";
/**
 * MCP Server Implementation
 * Register tools against the orchestrator executor and manage transports.
 *
 * @see {@link ../../docs/adr/005-mcp-integration.md ADR-005: MCP Protocol Integration}
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OUTPUTFORMAT = void 0;
exports.createMCPServer = createMCPServer;
exports.registerToolsWithServer = registerToolsWithServer;
exports.formatOutput = formatOutput;
const mcp_js_1 = require("@modelcontextprotocol/sdk/server/mcp.js");
const stdio_js_1 = require("@modelcontextprotocol/sdk/server/stdio.js");
const types_js_1 = require("@modelcontextprotocol/sdk/types.js");
const errors_1 = require("../lib/errors");
const prompts_1 = require("../prompts");
const logger_1 = require("../lib/logger");
const orchestrator_types_1 = require("../app/orchestrator-types");
const natural_language_formatters_1 = require("./formatters/natural-language-formatters");
/**
 * Constants
 */
const RESOURCE_URI = {
    STATUS: 'containerization://status',
};
const ERROR_FORMAT = {
    HINT_PREFIX: '💡',
    RESOLUTION_PREFIX: '🔧',
    DEFAULT_RESOLUTION: 'Check logs for more information',
};
/**
 * Output format options for tool results
 *
 * @property JSON - Full structured JSON output (for APIs and programmatic access)
 * @property TEXT - Concise summary text (for logs and quick display)
 * @property MARKDOWN - Summary with collapsible JSON details (for documentation)
 * @property NATURAL_LANGUAGE - Rich narrative with sections (for user interfaces)
 */
exports.OUTPUTFORMAT = {
    MARKDOWN: 'markdown',
    JSON: 'json',
    TEXT: 'text',
    NATURAL_LANGUAGE: 'natural-language',
};
/**
 * Format error message with guidance for better user experience
 * @param error - The error message
 * @param guidance - Optional error guidance with hints and resolution steps
 * @returns Formatted error message with guidance
 */
function formatErrorWithGuidance(error, guidance) {
    if (!guidance) {
        return error || 'Tool execution failed';
    }
    const parts = [error];
    if (guidance.hint) {
        parts.push(`${ERROR_FORMAT.HINT_PREFIX} ${guidance.hint}`);
    }
    parts.push(`${ERROR_FORMAT.RESOLUTION_PREFIX} Resolution:`, guidance.resolution || ERROR_FORMAT.DEFAULT_RESOLUTION);
    return parts.join('\n\n');
}
/**
 * Create an MCP server that delegates execution to the orchestrator
 * @param tools - Array of MCP tools to register with the server
 * @param options - Server configuration options
 * @param execute - Tool executor function that handles tool execution requests
 * @returns MCPServer interface for managing the server lifecycle
 */
function createMCPServer(tools, options = {}, execute) {
    const logger = options.logger || (0, logger_1.createLogger)({ name: 'mcp-server' });
    const serverOptions = {
        name: options.name || 'containerization-assist',
        version: options.version || '1.0.0',
    };
    const server = new mcp_js_1.McpServer(serverOptions);
    const transportType = options.transport ?? 'stdio';
    const outputFormat = options.outputFormat ?? exports.OUTPUTFORMAT.NATURAL_LANGUAGE;
    const chainHintsMode = options.chainHintsMode ?? orchestrator_types_1.CHAINHINTSMODE.ENABLED;
    let transportInstance = null;
    let isRunning = false;
    registerToolsWithServer({
        outputFormat,
        chainHintsMode,
        server,
        tools,
        logger,
        transport: transportType,
        execute,
    });
    server.resource('status', RESOURCE_URI.STATUS, {
        title: 'Container Status',
        description: 'Current status of the containerization system',
    }, async () => ({
        contents: [
            {
                uri: RESOURCE_URI.STATUS,
                mimeType: 'application/json',
                text: JSON.stringify({
                    running: isRunning,
                    tools: tools.length,
                    transport: transportType,
                    timestamp: new Date().toISOString(),
                }, null, 2),
            },
        ],
    }));
    (0, prompts_1.registerPrompts)(server);
    return {
        async start() {
            if (isRunning) {
                throw new Error('Server is already running');
            }
            if (transportType !== 'stdio') {
                throw new Error(`Only 'stdio' transport is supported. Requested: '${transportType}'`);
            }
            transportInstance = new stdio_js_1.StdioServerTransport();
            await server.connect(transportInstance);
            isRunning = true;
            logger.info({
                version: serverOptions.version,
                transport: transportType,
                toolCount: tools.length,
            }, 'MCP server started');
        },
        async stop() {
            if (!isRunning) {
                return;
            }
            await server.close();
            transportInstance = null;
            isRunning = false;
            logger.info({ transport: transportType }, 'MCP server stopped');
        },
        getServer() {
            return server.server;
        },
        getTools() {
            return tools.map((t) => ({
                name: t.name,
                description: t.description,
            }));
        },
    };
}
/**
 * Create tool handler function with proper typing to avoid deep type instantiation
 */
function getHandler(toolName, transport, outputFormat, chainHintsMode, execute) {
    return async (rawParams, extra) => {
        const params = rawParams ?? {};
        try {
            const { sanitizedParams, metadata } = prepareExecutionPayload(toolName, params, transport, extra);
            const result = await execute({
                toolName,
                params: sanitizedParams,
                metadata,
            });
            if (!result.ok) {
                // Format error with guidance if available
                const errorMessage = formatErrorWithGuidance(result.error, result.guidance);
                throw new types_js_1.McpError(types_js_1.ErrorCode.InternalError, errorMessage);
            }
            return {
                content: [
                    {
                        type: 'text',
                        text: formatOutput(result.value, outputFormat, chainHintsMode),
                    },
                ],
            };
        }
        catch (error) {
            throw error instanceof types_js_1.McpError
                ? error
                : new types_js_1.McpError(types_js_1.ErrorCode.InternalError, (0, errors_1.extractErrorMessage)(error));
        }
    };
}
/**
 * Register tools against an MCP server instance, delegating to the orchestrator executor.
 * Each tool is registered with its name, description, and input schema. Tool execution is
 * delegated to the orchestrator's execute function.
 * @param options - Registration options including server, tools, and executor
 */
function registerToolsWithServer(options) {
    const { server, tools, transport, execute, outputFormat, chainHintsMode = orchestrator_types_1.CHAINHINTSMODE.ENABLED, } = options;
    for (const tool of tools) {
        const handler = getHandler(tool.name, transport, outputFormat, chainHintsMode, execute);
        // Type assertion to avoid deep type instantiation issues with MCP SDK
        // The MCP SDK's complex generic constraints on tool() cause TS2589 errors
        // Runtime safety is preserved by Zod schema validation
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        server.tool(tool.name, tool.description, tool.inputSchema, handler);
    }
}
/**
 * Creates logger context from tool name, transport, and MCP request metadata
 * @param toolName - Name of the tool being executed
 * @param transport - Transport type (e.g., 'stdio')
 * @param requestId - JSON-RPC request ID from MCP SDK
 * @returns Logger context object
 */
function createLoggerContext(toolName, transport, requestId) {
    return {
        transport,
        tool: toolName,
        ...(requestId !== undefined && { requestId: String(requestId) }),
    };
}
/**
 * Creates a type-safe notification adapter that wraps the MCP SDK sendNotification
 * @param mcpSendNotification - The MCP SDK sendNotification function
 * @returns Type-safe notification sender
 */
function createNotificationAdapter(mcpSendNotification) {
    return async (notification) => {
        // Assume the caller provides the correct type, or add runtime validation here if needed
        return mcpSendNotification(notification);
    };
}
/**
 * Creates execution metadata from MCP request context
 * @param toolName - Name of the tool being executed
 * @param transport - Transport type
 * @param extra - Request handler extras from MCP SDK
 * @returns ExecuteMetadata object
 */
function createExecuteMetadata(toolName, transport, extra) {
    return {
        progress: extra._meta?.progressToken,
        loggerContext: createLoggerContext(toolName, transport, extra.requestId),
        ...(extra.sendNotification && {
            sendNotification: createNotificationAdapter(extra.sendNotification),
        }),
    };
}
/**
 * Prepares execution payload by sanitizing params and creating metadata
 * @param toolName - Name of the tool being executed
 * @param params - Raw tool parameters
 * @param transport - Transport type
 * @param extra - Request handler extras from MCP SDK
 * @returns Object containing sanitized params and execution metadata
 */
function prepareExecutionPayload(toolName, params, transport, extra) {
    return {
        sanitizedParams: sanitizeParams(params),
        metadata: createExecuteMetadata(toolName, transport, extra),
    };
}
/**
 * Removes internal metadata fields from parameters
 * @param params - Raw parameters with potential metadata
 * @returns Sanitized parameters without _meta field
 */
function sanitizeParams(params) {
    const entries = Object.entries(params).filter(([key]) => key !== '_meta');
    return Object.fromEntries(entries);
}
/**
 * Format tool output based on requested format
 *
 * @param output - The tool result to format (typically includes a summary field)
 * @param format - Output format (JSON, TEXT, MARKDOWN, or NATURAL_LANGUAGE)
 * @param chainHintsMode - Whether to include "Next Steps" sections in natural language output (default: 'enabled')
 * @returns Formatted string representation of the output
 *
 * @description
 * Transforms tool results into user-friendly formats:
 * - JSON: Full structured data (default, for APIs)
 * - TEXT: Summary field only (for logs/console)
 * - MARKDOWN: Summary + collapsible JSON (for documentation)
 * - NATURAL_LANGUAGE: Rich narrative (for user interfaces)
 *
 * All tool results include a `summary` field for human-readable display.
 * The NATURAL_LANGUAGE format uses type detection to provide tool-specific
 * rich narratives with sections, formatting, and next steps.
 */
function formatOutput(output, format, chainHintsMode = orchestrator_types_1.CHAINHINTSMODE.ENABLED) {
    switch (format) {
        case exports.OUTPUTFORMAT.JSON:
            return JSON.stringify(output, null, 2);
        case exports.OUTPUTFORMAT.NATURAL_LANGUAGE:
            // Rich narrative formatting - delegates to tool-specific formatters
            return formatAsNaturalLanguage(output, chainHintsMode);
        case exports.OUTPUTFORMAT.MARKDOWN:
            // Check if output has a summary field
            if (typeof output === 'object' && output !== null && 'summary' in output) {
                const { summary, ...rest } = output;
                // Display summary prominently, with structured data collapsed
                return `${summary}\n\n<details>\n<summary>View detailed output</summary>\n\n\`\`\`json\n${JSON.stringify(rest, null, 2)}\n\`\`\`\n</details>`;
            }
            // Fallback to JSON code block
            return `\`\`\`json\n${JSON.stringify(output, null, 2)}\n\`\`\``;
        case exports.OUTPUTFORMAT.TEXT:
            // Prioritize summary in plain text mode
            if (typeof output === 'object' && output !== null && 'summary' in output) {
                const { summary } = output;
                return summary;
            }
            // Fallback to JSON
            if (typeof output === 'object' && output !== null) {
                return JSON.stringify(output, null, 2);
            }
            return String(output);
        default:
            return JSON.stringify(output, null, 2);
    }
}
/**
 * Format output as natural language narrative
 *
 * @param output - Tool result to format as narrative
 * @param chainHintsMode - Whether to include "Next Steps" sections (default: 'enabled')
 * @returns Rich narrative with formatting, or summary/JSON fallback
 *
 * @description
 * Delegates to tool-specific formatters for rich narratives with:
 * - Section headers and formatting
 * - Bullet points and structured lists
 * - Context-aware next steps (when chainHintsMode is 'enabled')
 * - Proper handling of optional fields
 *
 * Supported tool types with dedicated formatters:
 * - scan-image: Security scan results with severity breakdown
 * - generate-dockerfile: Planning with base images and recommendations
 * - deploy: Deployment status with endpoints and conditions
 * - build-image-context: Build results with metrics
 * - analyze-repo: Repository analysis with module detection
 *
 * Falls back to summary field or JSON for other tool types.
 */
function formatAsNaturalLanguage(output, chainHintsMode = orchestrator_types_1.CHAINHINTSMODE.ENABLED) {
    if (!output || typeof output !== 'object') {
        return String(output);
    }
    // Type detection and delegation
    // Each tool result type gets its own formatter
    if (isScanImageResult(output)) {
        return (0, natural_language_formatters_1.formatScanImageNarrative)(output, chainHintsMode);
    }
    if (isDockerfilePlan(output)) {
        return (0, natural_language_formatters_1.formatDockerfilePlanNarrative)(output, chainHintsMode);
    }
    if (isBuildImageResult(output)) {
        return (0, natural_language_formatters_1.formatBuildImageNarrative)(output, chainHintsMode);
    }
    if (isAnalyzeRepoResult(output)) {
        return (0, natural_language_formatters_1.formatAnalyzeRepoNarrative)(output, chainHintsMode);
    }
    // NEW FORMATTERS - Add in order of specificity (most specific first)
    if (isVerifyDeployResult(output)) {
        return (0, natural_language_formatters_1.formatVerifyDeployNarrative)(output, chainHintsMode);
    }
    if (isFixDockerfileResult(output)) {
        return (0, natural_language_formatters_1.formatFixDockerfileNarrative)(output, chainHintsMode);
    }
    if (isGenerateK8sManifestsResult(output)) {
        return (0, natural_language_formatters_1.formatGenerateK8sManifestsNarrative)(output, chainHintsMode);
    }
    if (isPushImageResult(output)) {
        return (0, natural_language_formatters_1.formatPushImageNarrative)(output, chainHintsMode);
    }
    if (isTagImageResult(output)) {
        return (0, natural_language_formatters_1.formatTagImageNarrative)(output, chainHintsMode);
    }
    if (isPrepareClusterResult(output)) {
        return (0, natural_language_formatters_1.formatPrepareClusterNarrative)(output, chainHintsMode);
    }
    // Check ops results (check status before ping due to field overlap)
    if (isServerStatusResult(output)) {
        return (0, natural_language_formatters_1.formatOpsStatusNarrative)(output);
    }
    if (isPingResult(output)) {
        return (0, natural_language_formatters_1.formatOpsPingNarrative)(output);
    }
    // Additional tool result types can be added as needed
    // Fallback: use summary if available, otherwise JSON
    if ('summary' in output) {
        const { summary } = output;
        return summary;
    }
    return JSON.stringify(output, null, 2);
}
/**
 * Type guards for result types
 * These enable proper type detection for natural language formatting
 */
function isScanImageResult(output) {
    return 'vulnerabilities' in output && 'scanTime' in output && 'passed' in output;
}
function isDockerfilePlan(output) {
    if (!('recommendations' in output && 'repositoryInfo' in output)) {
        return false;
    }
    if ('manifestType' in output) {
        return false;
    }
    const recommendations = output.recommendations;
    return (typeof recommendations === 'object' &&
        recommendations !== null &&
        'buildStrategy' in recommendations);
}
function isBuildImageResult(output) {
    return 'buildConfig' in output && 'nextAction' in output && 'dockerfileAnalysis' in output;
}
function isAnalyzeRepoResult(output) {
    return 'modules' in output && 'isMonorepo' in output;
}
function isVerifyDeployResult(output) {
    return 'deploymentName' in output && 'pods' in output && 'healthCheck' in output;
}
function isFixDockerfileResult(output) {
    return 'currentIssues' in output && 'fixes' in output && 'validationScore' in output;
}
function isGenerateK8sManifestsResult(output) {
    return 'manifestType' in output && 'recommendations' in output && 'knowledgeMatches' in output;
}
function isTagImageResult(output) {
    return 'tags' in output && 'imageId' in output && Array.isArray(output.tags);
}
function isPushImageResult(output) {
    return 'registry' in output && 'digest' in output && 'pushedTag' in output;
}
function isPrepareClusterResult(output) {
    return 'clusterReady' in output && 'cluster' in output && 'checks' in output;
}
function isPingResult(output) {
    return 'message' in output && 'server' in output && 'capabilities' in output;
}
function isServerStatusResult(output) {
    return 'memory' in output && 'cpu' in output && 'tools' in output && 'version' in output;
}
