// codetocloudmcp/out/mcpServers/toolsImplementations/getAppTopology/tool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetAppTopologyTool = void 0;
const models_1 = require("../../common/models");
const codeToCloudTool_1 = require("../shared/codeToCloudTool");
const toolType_1 = require("../shared/toolType");
const generateMermaid_1 = require("./generateMermaid");
var AzureYamlServiceType;
(function (AzureYamlServiceType) {
    AzureYamlServiceType["AppService"] = "appservice";
    AzureYamlServiceType["FunctionApp"] = "function";
    AzureYamlServiceType["ContainerApp"] = "containerapp";
    AzureYamlServiceType["StaticWebApp"] = "staticwebapp";
    AzureYamlServiceType["KubernetesService"] = "aks";
})(AzureYamlServiceType || (AzureYamlServiceType = {}));
class GetAppTopologyTool extends codeToCloudTool_1.CodeToCloudTool {
    constructor(stateManager) {
        super(toolType_1.ToolType.GetAppTopology, stateManager);
    }
    ;
    async innerInvoke(params) {
        const mermaidChart = (0, generateMermaid_1.generateMermaidChart)(params.workspaceFolder, params.services);
        const copilotResponse = `
Here is the user's mermaid diagram. Write a reminder to the user to install a Mermaid preview extension to be able to render the diagram
Please write this into .azure/architecture.${this.stateManager.getCopilotMarkdownExtension()} WITHOUT additional explanations on the deployment. Explain only the architecture and data flow.
Make changes if these do not fulfill requirements (do not use </br> in strings when generating the diagram):
\`\`\`mermaid
${mermaidChart}
\`\`\`
Ask user if the topology is expected, if not, you should directly update the generated diagram with the user's updated instructions.

Please inform the user that here are the supported hosting technologies:
${Object.values(AzureYamlServiceType).join(", ")}.

Here is the full list of supported binding Azure technologies for the topology:
${Object.values(models_1.AzureServiceNames).filter(type => !Object.values(AzureYamlServiceType).map(x => x.toString()).includes(type.toString())).join(", ")}.
`;
        return copilotResponse;
    }
}
exports.GetAppTopologyTool = GetAppTopologyTool;
