// containerization-assist-mcp/dist-cjs/src/infra/docker/client.js
"use strict";
/**
 * Docker client for containerization operations
 *
 * @see {@link ../../../docs/adr/006-infrastructure-organization.md ADR-006: Infrastructure Layer Organization}
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createDockerClient = void 0;
const dockerode_1 = __importDefault(require("dockerode"));
const tar_fs_1 = __importDefault(require("tar-fs"));
const path_1 = __importDefault(require("path"));
const types_1 = require("../../types");
const errors_1 = require("./errors");
const socket_validation_1 = require("./socket-validation");
const dockerignore_parser_1 = require("../../lib/dockerignore-parser");
/**
 * Generate a digest from an image ID
 * @param imageId - The Docker image ID
 * @param logger - Logger instance
 * @returns A SHA-256 digest string or empty string if invalid
 */
function generateDigestFromImageId(imageId, logger) {
    // If already prefixed, validate the hash portion
    if (imageId.startsWith('sha256:')) {
        const hash = imageId.substring(7);
        if (/^[a-f0-9]{64}$/.test(hash)) {
            return imageId;
        }
    }
    else {
        // If not prefixed, validate and add prefix
        if (/^[a-f0-9]{64}$/.test(imageId)) {
            return `sha256:${imageId}`;
        }
    }
    logger.warn({ imageId }, 'Image ID is not a valid SHA-256 hash, cannot generate digest');
    return '';
}
/**
 * Create base Docker client implementation
 */
function createBaseDockerClient(docker, logger) {
    // Helper function to fetch image info (used by both getImage and inspectImage)
    const fetchImageInfo = async (id) => {
        try {
            const image = docker.getImage(id);
            const inspect = await image.inspect();
            const imageInfo = {
                Id: inspect.Id,
                RepoTags: inspect.RepoTags,
                Size: inspect.Size,
                Created: inspect.Created,
            };
            return (0, types_1.Success)(imageInfo);
        }
        catch (error) {
            const guidance = (0, errors_1.extractDockerErrorGuidance)(error);
            const errorMessage = `Failed to get image: ${guidance.message}`;
            logger.error({
                error: errorMessage,
                hint: guidance.hint,
                resolution: guidance.resolution,
                errorDetails: guidance.details,
                originalError: error,
                imageId: id,
            }, 'Docker get image failed');
            return (0, types_1.Failure)(errorMessage, guidance);
        }
    };
    return {
        async buildImage(options) {
            const buildLogs = [];
            const buildWarnings = [];
            const startTime = Date.now();
            try {
                logger.debug({ options }, 'Starting Docker build');
                const contextPath = options.context || '.';
                const dockerfilePath = options.dockerfile
                    ? path_1.default.resolve(contextPath, options.dockerfile)
                    : undefined;
                const files = await (0, dockerignore_parser_1.getDockerBuildFiles)(contextPath, dockerfilePath);
                const tarStream = tar_fs_1.default.pack(contextPath, {
                    entries: files,
                });
                const stream = await docker.buildImage(tarStream, {
                    t: options.t || options.tags?.[0],
                    dockerfile: options.dockerfile,
                    buildargs: options.buildargs || options.buildArgs,
                    ...(options.platform && { platform: options.platform }),
                    version: '2', // Use BuildKit backend for cross-platform builds
                });
                let buildError = null;
                const result = await new Promise((resolve, reject) => {
                    docker.modem.followProgress(stream, (err, res) => {
                        if (err) {
                            // Log detailed error information before rejecting
                            const guidance = (0, errors_1.extractDockerErrorGuidance)(err);
                            logger.error({
                                error: guidance.message,
                                hint: guidance.hint,
                                resolution: guidance.resolution,
                                errorDetails: guidance.details,
                                originalError: err,
                                options,
                            }, 'Docker build followProgress error');
                            reject(err);
                        }
                        else if (buildError) {
                            // If we detected an error during build progress, treat it as a failure
                            const errorObj = new Error(buildError);
                            logger.error({ buildError, options }, 'Docker build failed with error event');
                            reject(errorObj);
                        }
                        else {
                            resolve(res);
                        }
                    }, (event) => {
                        logger.debug(event, 'Docker build progress');
                        if (event.stream) {
                            const logLine = event.stream.trimEnd();
                            if (logLine) {
                                buildLogs.push(logLine);
                                logger.info(logLine);
                            }
                        }
                        if (event.error || event.errorDetail) {
                            logger.error({ errorEvent: event }, 'Docker build error event received');
                            const errorMsg = event.error || 'Build step failed';
                            const errorLogLine = `ERROR: ${errorMsg}`;
                            buildLogs.push(errorLogLine);
                            logger.error(errorLogLine);
                            // Capture the first error encountered during the build
                            if (!buildError) {
                                buildError =
                                    event.error ||
                                        (event.errorDetail &&
                                            typeof event.errorDetail === 'object' &&
                                            'message' in event.errorDetail
                                            ? String(event.errorDetail.message)
                                            : 'Build step failed');
                            }
                        }
                    });
                });
                const imageId = result[result.length - 1]?.aux?.ID || '';
                const buildTime = Date.now() - startTime;
                // Inspect the image to get size, digest, and layers
                let size = 0;
                let digest = '';
                let layers;
                if (imageId) {
                    try {
                        const image = docker.getImage(imageId);
                        const inspect = await image.inspect();
                        size = inspect.Size || 0;
                        // Use RepoDigests if available, otherwise fallback to image ID if it looks like a valid SHA-256 hash
                        if (inspect.RepoDigests?.[0]) {
                            digest = inspect.RepoDigests[0];
                        }
                        else {
                            digest = generateDigestFromImageId(inspect.Id, logger);
                        }
                        layers = inspect.RootFS?.Layers?.length;
                    }
                    catch (inspectError) {
                        logger.warn({ error: inspectError, imageId }, 'Could not inspect image after build');
                        buildWarnings.push('Could not retrieve complete image metadata');
                        // Use fallback digest from image ID
                        digest = generateDigestFromImageId(imageId, logger);
                    }
                }
                const buildResult = {
                    imageId,
                    digest,
                    size,
                    ...(layers !== undefined && { layers }),
                    buildTime,
                    logs: buildLogs,
                    tags: options.tags || [],
                    warnings: buildWarnings,
                };
                logger.debug({ buildResult }, 'Docker build completed successfully');
                return (0, types_1.Success)(buildResult);
            }
            catch (error) {
                const guidance = (0, errors_1.extractDockerErrorGuidance)(error);
                const errorMessage = `Build failed: ${guidance.message}`;
                logger.error({
                    error: errorMessage,
                    hint: guidance.hint,
                    resolution: guidance.resolution,
                    errorDetails: guidance.details,
                    originalError: error,
                    options,
                    buildLogs,
                }, 'Docker build failed');
                const enhancedGuidance = {
                    ...guidance,
                    details: {
                        ...guidance.details,
                        buildLogs: buildLogs.length > 0 ? buildLogs : ['No build logs captured'],
                        buildTime: Date.now() - startTime,
                    },
                };
                return (0, types_1.Failure)(errorMessage, enhancedGuidance);
            }
        },
        async getImage(id) {
            return fetchImageInfo(id);
        },
        async inspectImage(imageId) {
            // Alias for getImage - delegate to the same implementation
            return fetchImageInfo(imageId);
        },
        async tagImage(imageId, repository, tag) {
            try {
                const image = docker.getImage(imageId);
                await image.tag({ repo: repository, tag });
                logger.info({ imageId, repository, tag }, 'Image tagged successfully');
                return (0, types_1.Success)(undefined);
            }
            catch (error) {
                const guidance = (0, errors_1.extractDockerErrorGuidance)(error);
                const errorMessage = `Failed to tag image: ${guidance.message}`;
                logger.error({
                    error: errorMessage,
                    hint: guidance.hint,
                    resolution: guidance.resolution,
                    errorDetails: guidance.details,
                    originalError: error,
                    imageId,
                    repository,
                    tag,
                }, 'Docker tag image failed');
                return (0, types_1.Failure)(errorMessage, guidance);
            }
        },
        async pushImage(repository, tag, authConfig) {
            try {
                const image = docker.getImage(`${repository}:${tag}`);
                // dockerode's Image.push expects auth config inside the first options object
                // For local registries without auth, provide an empty authconfig object to avoid X-Registry-Auth header issues
                const stream = await image.push({
                    authconfig: authConfig || {},
                });
                let digest = '';
                let size;
                await new Promise((resolve, reject) => {
                    let pushError = null;
                    docker.modem.followProgress(stream, (err) => {
                        if (err) {
                            // Log detailed error information before rejecting
                            const guidance = (0, errors_1.extractDockerErrorGuidance)(err);
                            logger.error({
                                error: guidance.message,
                                hint: guidance.hint,
                                resolution: guidance.resolution,
                                errorDetails: guidance.details,
                                originalError: err,
                                repository,
                                tag,
                            }, 'Docker push followProgress error');
                            reject(err);
                        }
                        else if (pushError) {
                            // Reject if we encountered an error event during the push
                            reject(pushError);
                        }
                        else {
                            resolve();
                        }
                    }, (event) => {
                        logger.debug(event, 'Docker push progress');
                        // Only treat final error events as failures, not intermediate auth challenges
                        if (event.error || event.errorDetail) {
                            logger.debug({ errorEvent: event }, 'Docker push error event (may be intermediate)');
                            // Only set pushError for certain fatal errors, not auth challenges
                            const errorMsg = event.error || event.errorDetail?.message || '';
                            // Don't treat authentication challenges as fatal errors - they're part of the auth handshake
                            if (!errorMsg.toLowerCase().includes('unauthorized') &&
                                !errorMsg.toLowerCase().includes('authentication required')) {
                                pushError = new Error(errorMsg || 'Unknown push error');
                                logger.error({ errorEvent: event }, 'Fatal Docker push error event received');
                            }
                        }
                        if (event.aux?.Digest) {
                            digest = event.aux.Digest;
                        }
                        if (event.aux?.Size) {
                            size = event.aux.Size;
                        }
                    });
                });
                if (!digest) {
                    try {
                        const inspectResult = await image.inspect();
                        digest =
                            inspectResult.RepoDigests?.[0]?.split('@')[1] ||
                                `sha256:${inspectResult.Id.replace('sha256:', '')}`;
                    }
                    catch (inspectError) {
                        logger.warn({ error: inspectError }, 'Could not get digest from image inspection');
                        digest = `sha256:${Date.now().toString(16)}${Math.random().toString(16).substring(2)}`;
                    }
                }
                logger.info({ repository, tag, digest }, 'Image pushed successfully');
                const result = { digest };
                if (size !== undefined) {
                    result.size = size;
                }
                return (0, types_1.Success)(result);
            }
            catch (error) {
                const guidance = (0, errors_1.extractDockerErrorGuidance)(error);
                const errorMessage = `Failed to push image: ${guidance.message}`;
                logger.error({
                    error: errorMessage,
                    hint: guidance.hint,
                    resolution: guidance.resolution,
                    errorDetails: guidance.details,
                    originalError: error,
                    repository,
                    tag,
                }, 'Docker push image failed');
                return (0, types_1.Failure)(errorMessage, guidance);
            }
        },
        async removeImage(imageId, force = false) {
            try {
                logger.debug({ imageId, force }, 'Starting Docker image removal');
                const image = docker.getImage(imageId);
                await image.remove({ force });
                logger.debug({ imageId }, 'Image removed');
                return (0, types_1.Success)(undefined);
            }
            catch (error) {
                const guidance = (0, errors_1.extractDockerErrorGuidance)(error);
                const errorMessage = `Failed to remove image: ${guidance.message}`;
                logger.error({
                    error: errorMessage,
                    hint: guidance.hint,
                    resolution: guidance.resolution,
                    errorDetails: guidance.details,
                    originalError: error,
                    imageId,
                }, 'Docker remove image failed');
                return (0, types_1.Failure)(errorMessage, guidance);
            }
        },
        async removeContainer(containerId, force = false) {
            try {
                logger.debug({ containerId, force }, 'Starting Docker container removal');
                const container = docker.getContainer(containerId);
                await container.remove({ force });
                logger.debug({ containerId }, 'Container removed');
                return (0, types_1.Success)(undefined);
            }
            catch (error) {
                const guidance = (0, errors_1.extractDockerErrorGuidance)(error);
                const errorMessage = `Failed to remove container: ${guidance.message}`;
                logger.error({
                    error: errorMessage,
                    hint: guidance.hint,
                    resolution: guidance.resolution,
                    errorDetails: guidance.details,
                    originalError: error,
                    containerId,
                }, 'Docker remove container failed');
                return (0, types_1.Failure)(errorMessage, guidance);
            }
        },
        async listContainers(options = {}) {
            try {
                logger.debug({ options }, 'Starting Docker container listing');
                const containers = await docker.listContainers(options);
                logger.debug({ containerCount: containers.length }, 'Docker containers listed successfully');
                return (0, types_1.Success)(containers);
            }
            catch (error) {
                const guidance = (0, errors_1.extractDockerErrorGuidance)(error);
                const errorMessage = `Failed to list containers: ${guidance.message}`;
                logger.error({
                    error: errorMessage,
                    hint: guidance.hint,
                    resolution: guidance.resolution,
                    errorDetails: guidance.details,
                    originalError: error,
                    options,
                }, 'Docker list containers failed');
                return (0, types_1.Failure)(errorMessage, guidance);
            }
        },
    };
}
/**
 * Create a Docker client with core operations
 * @param logger - Logger instance for debug output
 * @param config - Optional Docker client configuration
 * @returns DockerClient with build, get, tag, and push operations
 */
const createDockerClient = (logger, config) => {
    // Determine the socket path to use
    let socketPath;
    if (config?.socketPath) {
        socketPath = config.socketPath;
    }
    else {
        socketPath = (0, socket_validation_1.autoDetectDockerSocket)();
        logger.debug({ socketPath }, 'Auto-detected Docker socket');
    }
    // Create Docker client with detected socket path
    const dockerOptions = {};
    if (socketPath.startsWith('tcp://') || socketPath.startsWith('http://')) {
        // TCP connection
        dockerOptions.host = config?.host || 'localhost';
        dockerOptions.port = config?.port || 2375;
    }
    else {
        // Unix socket connection
        dockerOptions.socketPath = socketPath;
    }
    if (config?.timeout) {
        dockerOptions.timeout = config.timeout;
    }
    const docker = new dockerode_1.default(dockerOptions);
    logger.debug({ dockerOptions }, 'Created Docker client');
    // Create and return client
    return createBaseDockerClient(docker, logger);
};
exports.createDockerClient = createDockerClient;
