// codetocloudmcp/out/toolsImplementations/diagnoseExistingResources/keyVault/keyVaultDiagnoseBaseProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyVaultDiagnoseProvider = void 0;
const resourceDiagnoseHandler_1 = require("../resourceDiagnoseHandler");
const acaKeyVaultDiagnosticHandler_1 = require("./aca/acaKeyVaultDiagnosticHandler");
const keyVaultToAksDiagnosticHandler_1 = require("./keyVaultToAksDiagnosticHandler");
class KeyVaultDiagnoseProvider extends resourceDiagnoseHandler_1.ResourceDiagnoseHandler {
    keyVaultId;
    containerAppIds;
    kubernetesIds;
    constructor(params, stateManager) {
        super(params, stateManager);
        this.keyVaultId = params.keyVaultConnections?.keyVaultId ?? "";
        this.containerAppIds = params.keyVaultConnections
            ? [...new Set(params.keyVaultConnections.containerAppIds)]
            : [];
        this.kubernetesIds = params.keyVaultConnections
            ? [...new Set(params.keyVaultConnections.kubernetesIds)]
            : [];
    }
    async diagnose() {
        const headerLines = [];
        const keyVaultRegex = /^\/subscriptions\/([^\/]+)\/resourceGroups\/([^\/]+)\/providers\/Microsoft\.KeyVault\/vaults\/([^\/]+)$/i;
        const keyVaultMatch = this.keyVaultId.match(keyVaultRegex);
        if (!keyVaultMatch) {
            return `Invalid keyVaultId format: ${this.keyVaultId}. Expected format: /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.KeyVault/vaults/{vaultName}`;
        }
        headerLines.push(`\n🔍 Diagnosing Key Vault connections:`);
        headerLines.push(`Key Vault: ${this.keyVaultId}`);
        const [, keyVaultSubscriptionId, keyVaultResourceGroupName, vaultName] = keyVaultMatch;
        // Get Key Vault information
        let keyVault;
        try {
            const azureClient = await this.stateManager.getAzureClient(keyVaultSubscriptionId);
            const vault = await azureClient.getKeyVault(keyVaultResourceGroupName, vaultName);
            keyVault = {
                id: vault.id,
                name: vault.name,
                vaultUri: vault.properties?.vaultUri,
            };
        }
        catch (error) {
            return `Failed to retrieve Key Vault details for ${this.keyVaultId}. Error: ${error.message}`;
        }
        const acaPromises = this.containerAppIds.map((containerAppId) => new acaKeyVaultDiagnosticHandler_1.AcaKeyVaultDiagnosticHandler(this.stateManager).diagnoseResource(containerAppId, keyVault));
        const aksPromises = this.kubernetesIds.map((kubernetesId) => new keyVaultToAksDiagnosticHandler_1.KeyVaultToAksDiagnosticHandler(this.stateManager).diagnoseResource(kubernetesId, keyVault));
        const result = await Promise.all([...acaPromises, ...aksPromises]);
        headerLines.push(this.formatDiagnosticResults(result));
        return headerLines.join("\n");
    }
}
exports.KeyVaultDiagnoseProvider = KeyVaultDiagnoseProvider;
