// containerization-assist-mcp/dist-cjs/src/tools/shared/schemas.js
"use strict";
/**
 * Shared Zod schemas for tool parameters
 * Common building blocks to reduce duplication across tools
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.platform = exports.analysisOptions = exports.port = exports.replicas = exports.buildArgs = exports.tags = exports.imageName = exports.environment = exports.namespaceOptional = exports.repositoryPath = exports.DOCKER_PLATFORMS = void 0;
const zod_1 = require("zod");
const constants_1 = require("../../config/constants");
/**
 * Supported Docker platforms for multi-architecture builds
 * See: https://docs.docker.com/build/building/multi-platform/
 */
exports.DOCKER_PLATFORMS = [
    'linux/amd64',
    'linux/arm64',
    'linux/arm/v7',
    'linux/arm/v6',
    'linux/386',
    'linux/ppc64le',
    'linux/s390x',
    'linux/riscv64',
    'windows/amd64',
];
// Paths
exports.repositoryPath = zod_1.z
    .string()
    .min(1, 'Repository path cannot be empty')
    .describe('Absolute path to the repository. Paths are automatically normalized to forward slashes on all platforms (e.g., /path/to/repo or C:/path/to/repo)');
exports.namespaceOptional = zod_1.z.string().optional().describe('Kubernetes namespace');
// Environment schema
exports.environment = constants_1.environmentSchema.optional();
// Docker image fields
exports.imageName = zod_1.z.string().describe('Name for the Docker image');
exports.tags = zod_1.z.array(zod_1.z.string()).describe('Tags to apply to the image');
exports.buildArgs = zod_1.z.record(zod_1.z.string(), zod_1.z.string()).describe('Build arguments');
// Application basics
exports.replicas = zod_1.z.number().optional().describe('Number of replicas');
exports.port = zod_1.z.number().optional().describe('Application port');
// Service types
// Ingress
// Health checks
// Autoscaling
// Analysis options
exports.analysisOptions = {
    depth: zod_1.z.number().optional().describe('Analysis depth (1-5)'),
    includeTests: zod_1.z.boolean().optional().describe('Include test files in analysis'),
    securityFocus: zod_1.z.boolean().optional().describe('Focus on security aspects'),
    performanceFocus: zod_1.z.boolean().optional().describe('Focus on performance aspects'),
};
// Platform - Required for consistent builds across environments
exports.platform = zod_1.z
    .enum(exports.DOCKER_PLATFORMS)
    .default('linux/amd64')
    .describe('Target platform for the Docker image (e.g., "linux/amd64", "linux/arm64"). Defaults to linux/amd64 for maximum compatibility. Required to ensure consistent builds across environments and cross-platform scenarios (e.g., ARM Mac targeting AMD64 servers).');
