// @azure/arm-appcontainers/dist/commonjs/containerAppsAPIClient.js
"use strict";
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContainerAppsAPIClient = void 0;
const tslib_1 = require("tslib");
const coreClient = tslib_1.__importStar(require("@azure/core-client"));
const coreRestPipeline = tslib_1.__importStar(require("@azure/core-rest-pipeline"));
const index_js_1 = require("./operations/index.js");
const Parameters = tslib_1.__importStar(require("./models/parameters.js"));
const Mappers = tslib_1.__importStar(require("./models/mappers.js"));
class ContainerAppsAPIClient extends coreClient.ServiceClient {
    /**
     * Initializes a new instance of the ContainerAppsAPIClient class.
     * @param credentials Subscription credentials which uniquely identify client subscription.
     * @param subscriptionId The ID of the target subscription.
     * @param options The parameter options
     */
    constructor(credentials, subscriptionId, options) {
        var _a, _b, _c;
        if (credentials === undefined) {
            throw new Error("'credentials' cannot be null");
        }
        if (subscriptionId === undefined) {
            throw new Error("'subscriptionId' cannot be null");
        }
        // Initializing default values for options
        if (!options) {
            options = {};
        }
        const defaults = {
            requestContentType: "application/json; charset=utf-8",
            credential: credentials,
        };
        const packageDetails = `azsdk-js-arm-appcontainers/2.2.0`;
        const userAgentPrefix = options.userAgentOptions && options.userAgentOptions.userAgentPrefix
            ? `${options.userAgentOptions.userAgentPrefix} ${packageDetails}`
            : `${packageDetails}`;
        const optionsWithDefaults = Object.assign(Object.assign(Object.assign({}, defaults), options), { userAgentOptions: {
                userAgentPrefix,
            }, endpoint: (_b = (_a = options.endpoint) !== null && _a !== void 0 ? _a : options.baseUri) !== null && _b !== void 0 ? _b : "https://management.azure.com" });
        super(optionsWithDefaults);
        let bearerTokenAuthenticationPolicyFound = false;
        if ((options === null || options === void 0 ? void 0 : options.pipeline) && options.pipeline.getOrderedPolicies().length > 0) {
            const pipelinePolicies = options.pipeline.getOrderedPolicies();
            bearerTokenAuthenticationPolicyFound = pipelinePolicies.some((pipelinePolicy) => pipelinePolicy.name ===
                coreRestPipeline.bearerTokenAuthenticationPolicyName);
        }
        if (!options ||
            !options.pipeline ||
            options.pipeline.getOrderedPolicies().length == 0 ||
            !bearerTokenAuthenticationPolicyFound) {
            this.pipeline.removePolicy({
                name: coreRestPipeline.bearerTokenAuthenticationPolicyName,
            });
            this.pipeline.addPolicy(coreRestPipeline.bearerTokenAuthenticationPolicy({
                credential: credentials,
                scopes: (_c = optionsWithDefaults.credentialScopes) !== null && _c !== void 0 ? _c : `${optionsWithDefaults.endpoint}/.default`,
                challengeCallbacks: {
                    authorizeRequestOnChallenge: coreClient.authorizeRequestOnClaimChallenge,
                },
            }));
        }
        // Parameter assignments
        this.subscriptionId = subscriptionId;
        // Assigning values to Constant parameters
        this.$host = options.$host || "https://management.azure.com";
        this.apiVersion = options.apiVersion || "2025-01-01";
        this.containerAppsAuthConfigs = new index_js_1.ContainerAppsAuthConfigsImpl(this);
        this.availableWorkloadProfiles = new index_js_1.AvailableWorkloadProfilesImpl(this);
        this.billingMeters = new index_js_1.BillingMetersImpl(this);
        this.connectedEnvironments = new index_js_1.ConnectedEnvironmentsImpl(this);
        this.connectedEnvironmentsCertificates =
            new index_js_1.ConnectedEnvironmentsCertificatesImpl(this);
        this.connectedEnvironmentsDaprComponents =
            new index_js_1.ConnectedEnvironmentsDaprComponentsImpl(this);
        this.connectedEnvironmentsStorages = new index_js_1.ConnectedEnvironmentsStoragesImpl(this);
        this.containerApps = new index_js_1.ContainerAppsImpl(this);
        this.containerAppsRevisions = new index_js_1.ContainerAppsRevisionsImpl(this);
        this.containerAppsRevisionReplicas = new index_js_1.ContainerAppsRevisionReplicasImpl(this);
        this.containerAppsDiagnostics = new index_js_1.ContainerAppsDiagnosticsImpl(this);
        this.managedEnvironmentDiagnostics = new index_js_1.ManagedEnvironmentDiagnosticsImpl(this);
        this.managedEnvironmentsDiagnostics =
            new index_js_1.ManagedEnvironmentsDiagnosticsImpl(this);
        this.jobs = new index_js_1.JobsImpl(this);
        this.operations = new index_js_1.OperationsImpl(this);
        this.javaComponents = new index_js_1.JavaComponentsImpl(this);
        this.jobsExecutions = new index_js_1.JobsExecutionsImpl(this);
        this.managedEnvironments = new index_js_1.ManagedEnvironmentsImpl(this);
        this.certificates = new index_js_1.CertificatesImpl(this);
        this.managedCertificates = new index_js_1.ManagedCertificatesImpl(this);
        this.namespaces = new index_js_1.NamespacesImpl(this);
        this.daprComponents = new index_js_1.DaprComponentsImpl(this);
        this.managedEnvironmentsStorages = new index_js_1.ManagedEnvironmentsStoragesImpl(this);
        this.containerAppsSessionPools = new index_js_1.ContainerAppsSessionPoolsImpl(this);
        this.containerAppsSourceControls = new index_js_1.ContainerAppsSourceControlsImpl(this);
        this.usages = new index_js_1.UsagesImpl(this);
        this.managedEnvironmentUsages = new index_js_1.ManagedEnvironmentUsagesImpl(this);
        this.addCustomApiVersionPolicy(options.apiVersion);
    }
    /** A function that adds a policy that sets the api-version (or equivalent) to reflect the library version. */
    addCustomApiVersionPolicy(apiVersion) {
        if (!apiVersion) {
            return;
        }
        const apiVersionPolicy = {
            name: "CustomApiVersionPolicy",
            async sendRequest(request, next) {
                const param = request.url.split("?");
                if (param.length > 1) {
                    const newParams = param[1].split("&").map((item) => {
                        if (item.indexOf("api-version") > -1) {
                            return "api-version=" + apiVersion;
                        }
                        else {
                            return item;
                        }
                    });
                    request.url = param[0] + "?" + newParams.join("&");
                }
                return next(request);
            },
        };
        this.pipeline.addPolicy(apiVersionPolicy);
    }
    /**
     * Get details of a single job execution
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param jobName Job Name
     * @param jobExecutionName Job execution name.
     * @param options The options parameters.
     */
    jobExecution(resourceGroupName, jobName, jobExecutionName, options) {
        return this.sendOperationRequest({ resourceGroupName, jobName, jobExecutionName, options }, jobExecutionOperationSpec);
    }
    /**
     * Get the verification id of a subscription used for verifying custom domains
     * @param options The options parameters.
     */
    getCustomDomainVerificationId(options) {
        return this.sendOperationRequest({ options }, getCustomDomainVerificationIdOperationSpec);
    }
}
exports.ContainerAppsAPIClient = ContainerAppsAPIClient;
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const jobExecutionOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.App/jobs/{jobName}/executions/{jobExecutionName}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.JobExecution,
        },
        default: {
            bodyMapper: Mappers.DefaultErrorResponse,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.jobName,
        Parameters.jobExecutionName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const getCustomDomainVerificationIdOperationSpec = {
    path: "/subscriptions/{subscriptionId}/providers/Microsoft.App/getCustomDomainVerificationId",
    httpMethod: "POST",
    responses: {
        200: {
            bodyMapper: { type: { name: "String" } },
        },
        default: {
            bodyMapper: Mappers.DefaultErrorResponse,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [Parameters.$host, Parameters.subscriptionId],
    headerParameters: [Parameters.accept],
    serializer,
};
