// codetocloudmcp/out/mcpServers/toolsImplementations/shared/codeToCloudTool.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodeToCloudTool = void 0;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const exceptions_1 = require("../../common/exceptions");
const copilotDefinitions_1 = require("./copilotDefinitions");
class CodeToCloudTool {
    invocationId;
    stateManager;
    toolType;
    constructor(toolType, stateManager) {
        this.invocationId = "NotInvoked";
        this.toolType = toolType;
        this.stateManager = stateManager;
    }
    getToolCopilotDescription() {
        return (0, copilotDefinitions_1.GetToolAgentDescription)(this.toolType, this.stateManager);
    }
    getToolType() {
        return this.toolType;
    }
    getDebugFolder() {
        if (process.env.CODE_TO_CLOUD_DEBUG === 'true' && process.env.PLAYWRIGHT_TEST_LOGGING_ROOT_PATH) {
            let logDirPath = path.join(process.env.PLAYWRIGHT_TEST_LOGGING_ROOT_PATH, 'tool_logs', this.toolType);
            // Ensure the directory exists
            if (!fs.existsSync(logDirPath)) {
                fs.mkdirSync(logDirPath, { recursive: true });
            }
            return logDirPath;
        }
    }
    async getAzureClient(subscriptionId) {
        const azureClient = this.stateManager.getAzureClient(subscriptionId);
        const isValid = await azureClient.validateSubscription();
        if (!isValid) {
            throw new exceptions_1.SubscriptionInvalidException();
        }
        return azureClient;
    }
    logProgress(params, contents) {
        const debugFolder = this.getDebugFolder();
        if (!debugFolder) {
            return;
        }
        const toolLogFileName = `${Date.now()}.progresslog`;
        let logFilePath = path.join(debugFolder, toolLogFileName);
        const results = {
            input: params,
            progress: contents,
            toolType: this.toolType,
            invocationId: this.invocationId
        };
        fs.writeFileSync(logFilePath, JSON.stringify(results), 'utf-8');
    }
    async logResponseWhenDebugging(params, response) {
        const debugFolder = this.getDebugFolder();
        if (!debugFolder) {
            return;
        }
        const toolLogFileName = `${Date.now()}.log`;
        let logFilePath = path.join(debugFolder, toolLogFileName);
        const results = {
            input: params,
            output: response,
            toolType: this.toolType,
            invocationId: this.invocationId
        };
        fs.writeFileSync(logFilePath, JSON.stringify(results), 'utf-8');
    }
    async invokeTool(params) {
        this.invocationId = `${this.toolType}-${Date.now()}`;
        try {
            console.log(`Tool: ${this.toolType} invoked with params: `, params);
            const response = await this.innerInvoke(params);
            await this.logResponseWhenDebugging(params, response);
            console.log('Tool response:\n: ' + response);
            return response;
        }
        catch (error) {
            if (error instanceof exceptions_1.SubscriptionInvalidException) {
                const errorMessage = error.message;
                await this.logResponseWhenDebugging(params, errorMessage);
                return errorMessage;
            }
            console.log('Tool failed with error:\n: ' + error);
            await this.logResponseWhenDebugging(params, error);
            throw error;
        }
    }
}
exports.CodeToCloudTool = CodeToCloudTool;
