// codetocloudmcp/out/toolsImplementations/runInTerminal/impl.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.RunInTerminal = RunInTerminal;
const childProcess = __importStar(require("child_process"));
const params_1 = require("./params");
async function RunInTerminal(workingDir, executable, commandToRun, env, progressLogHandler) {
    return new Promise((resolve, reject) => {
        // Validate azd commands require --no-prompt flag
        const trimmedCommand = commandToRun.trim();
        // Split by semicolons to handle chained commands
        const commands = trimmedCommand.split(';').map(cmd => cmd.trim());
        const allowedCommands = [
            "azd env set",
            "azd env list",
            "azd env get-values",
            "azd env select",
            "azd config unset",
            "azd env get",
            "azd config set"
        ];
        for (const cmd of commands) {
            const lowerCmd = cmd.toLowerCase();
            if (lowerCmd.startsWith('azd ')) {
                // Check if this specific azd command contains --no-prompt flag
                const isAllowedCommand = allowedCommands.some(allowedCmd => lowerCmd.startsWith(allowedCmd.toLowerCase()));
                if (!lowerCmd.includes('--no-prompt') && !isAllowedCommand) {
                    return reject(new Error(`azd commands must include --no-prompt flag to prevent interactive prompts. Found azd command without --no-prompt: "${cmd}". Full command: ${commandToRun}`));
                }
            }
        }
        // Determine the shell executable and arguments
        let shellExecutable;
        let shellArgs;
        switch (executable) {
            case params_1.RunInTerminalExecutable.Powershell:
                shellExecutable = "pwsh.exe";
                shellArgs = ["-NoProfile", "-NonInteractive", "-Command", commandToRun];
                break;
            case params_1.RunInTerminalExecutable.Cmd:
                shellExecutable = "cmd.exe";
                shellArgs = ["/c", commandToRun];
                break;
            default:
                return reject(new Error(`Unsupported executable: ${executable}`));
        }
        // Combine environment variables
        const processEnv = { ...process.env, ...env };
        // Spawn the process
        const child = childProcess.spawn(shellExecutable, shellArgs, {
            cwd: workingDir,
            env: processEnv,
            stdio: ["pipe", "pipe", "pipe"],
            shell: false,
            windowsHide: true
        });
        let stdout = "";
        let stderr = "";
        let processTerminated = false;
        let lastProgressTime = Date.now();
        // Progress logging every 30 seconds
        const progressInterval = setInterval(async () => {
            if (!processTerminated) {
                const elapsed = Math.floor((Date.now() - lastProgressTime) / 1000);
                const progressMessage = `Command still running... (${elapsed}s elapsed)`;
                // Pass the actual stdout and stderr content with progress info
                const stdoutWithProgress = stdout ? `${progressMessage}\n\n${stdout}` : progressMessage;
                await progressLogHandler(stdoutWithProgress, stderr);
            }
        }, 30000); // 30 seconds
        // Clear interval when process completes
        const clearProgressInterval = () => {
            clearInterval(progressInterval);
        };
        // Simple patterns that commonly indicate waiting for input, fill these as they occur
        // Add the pattern "Enter a value" for prompts
        const waitingForInputPatterns = [
            /Enter a value/i
        ];
        const checkForInputPrompt = (text) => {
            const lastFewLines = text.split('\n').slice(-3).join('\n');
            return waitingForInputPatterns.some(pattern => pattern.test(lastFewLines));
        };
        // Collect stdout
        child.stdout?.on("data", (data) => {
            const newData = data.toString();
            stdout += newData;
            if (!processTerminated && checkForInputPrompt(stdout)) {
                processTerminated = true;
                child.kill('SIGTERM');
                resolve("ERROR: Process terminated: appears to be waiting for user input. Do not initiate terminal commands that wait for user input.");
            }
        });
        // Collect stderr  
        child.stderr?.on("data", (data) => {
            stderr += data.toString();
        });
        // Handle process completion
        child.on("close", (code) => {
            clearProgressInterval();
            if (processTerminated) {
                return; // Already handled
            }
            let output = stdout.trim();
            let errorOutput = stderr.trim();
            // the output or errorOutput has ANSI codes: \u001B[31m → Start red text, \u001B[0m → Reset formatting, The ╷ and │ symbols are part of Terraform’s pretty error formatting.
            const ansiRegex = /\\u001B\[[0-9;]*m|╷/g;
            const ansiRegex2 = /\u001B\[[0-9;]*m|╷/g;
            output = output.replace(ansiRegex, '').replace(ansiRegex2, '').replace(/\\r/g, '  ').replace(/\r/g, '  ');
            errorOutput = errorOutput.replace(ansiRegex, '').replace(ansiRegex2, '').replace(/\\r/g, '  ').replace(/\r/g, '  ');
            if (code === 0) {
                // Success - return stdout if available, otherwise indicate success
                resolve(`Command completed with code 0: \nOutput: ${output}\nErrorOutput: ${errorOutput}`);
            }
            else {
                // Failure - include both code and error output
                const errorMessage = errorOutput || `Command failed with exit code ${code}`;
                resolve(`Error (exit code ${code}): ${errorMessage}\nOutput: ${output}`);
            }
        });
        // Handle process errors (e.g., executable not found)
        child.on("error", (error) => {
            clearProgressInterval();
            if (!processTerminated) {
                reject(new Error(`Failed to execute command: ${error.message}`));
            }
        });
        // Set a timeout to prevent hanging (30 seconds)
        const timeout = setTimeout(() => {
            clearProgressInterval();
            child.kill("SIGTERM");
            reject(new Error("Command execution timed out after 40 minutes"));
        }, 2400000);
        // Clear timeout on process completion
        child.on("close", () => {
            clearTimeout(timeout);
        });
    });
}
