// codetocloudmcp/out/mcpServers/toolsImplementations/getAzdAppLogs/util.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzdAppLogRetriever = exports.ResourceType = void 0;
exports.getAzdAppConsoleLogs = getAzdAppConsoleLogs;
const fs = __importStar(require("fs"));
const path_1 = __importDefault(require("path"));
const yaml = __importStar(require("yaml"));
const azureYamlFileName = "azure.yaml";
var ResourceType;
(function (ResourceType) {
    ResourceType["AppService"] = "Microsoft.Web/sites|app";
    ResourceType["ContainerApps"] = "Microsoft.App/containerApps";
    ResourceType["FunctionApp"] = "Microsoft.Web/sites|functionapp";
})(ResourceType || (exports.ResourceType = ResourceType = {}));
const HostToResourceType = {
    "containerapp": ResourceType.ContainerApps,
    "appservice": ResourceType.AppService,
    "function": ResourceType.FunctionApp,
};
function getServicesFromAzureYAML(workspaceFolder) {
    const azureYamlPath = path_1.default.join(workspaceFolder, azureYamlFileName);
    if (!fs.existsSync(azureYamlPath)) {
        throw new Error(`Azure YAML file not found at ${azureYamlPath}`);
    }
    const azureYaml = yaml.parse(fs.readFileSync(azureYamlPath, "utf8"));
    return azureYaml.services;
}
class AzdAppLogRetriever {
    azureClient;
    azdEnvName;
    resourceGroupName = "";
    overrideResourceGroupName;
    limit;
    logAnalyticsWorkspaceIds = [];
    constructor(azureClient, azdEnvName, overrideResourceGroupName, limit) {
        this.azureClient = azureClient;
        this.azdEnvName = azdEnvName;
        this.overrideResourceGroupName = overrideResourceGroupName;
        this.limit = limit || 200;
    }
    async initialize() {
        await this.getResourceGroupName();
        await this.getLogAnalyticsWorkspaceIds();
    }
    async getResourceGroupName() {
        const resourceGroups = await this.azureClient.getResourceGroupsInSubscription();
        for (const rg of resourceGroups) {
            if (this.azdEnvName && rg.tags["azd-env-name"] === this.azdEnvName) {
                this.resourceGroupName = rg.name;
                break;
            }
            if (this.overrideResourceGroupName) {
                if (rg.name === this.overrideResourceGroupName) {
                    this.resourceGroupName = rg.name;
                    break;
                }
            }
        }
        if (!this.resourceGroupName) {
            if (this.azdEnvName) {
                throw new Error(`No resource group with tag {\"azd-env-name\": ${this.azdEnvName}} found. Consider specifying the resource group name directly.`);
            }
            else {
                throw new Error("No resource group with the specified name found. Consider specifying the azd environment name instead.");
            }
        }
    }
    async getLogAnalyticsWorkspaceIds() {
        const filter = `resourceType eq 'Microsoft.OperationalInsights/workspaces'`;
        const resources = await this.azureClient.getResourcesInResourceGroupByFilter(this.resourceGroupName, filter);
        for (const resource of resources) {
            if (resource.id) {
                this.logAnalyticsWorkspaceIds.push(resource.id);
            }
        }
        if (this.logAnalyticsWorkspaceIds.length === 0) {
            throw new Error(`No log analytics workspaces found for resource group ${this.resourceGroupName}. Logs cannot be retrieved using this tool.`);
        }
    }
    // To check whether log tables are ready in the log analytics workspace. The logs might take some time to be ingested.
    availableTablesQuery() {
        return `search * | summarize by $table`;
    }
    getConsoleLogTableName(resourceType) {
        switch (resourceType) {
            case ResourceType.ContainerApps:
                return "ContainerAppConsoleLogs_CL";
            case ResourceType.AppService:
                return "AppServiceConsoleLogs";
            case ResourceType.FunctionApp:
                return "AppTraces";
            default:
                throw new Error(`Unsupported resource type: ${resourceType}`);
        }
    }
    containerAppLogsQuery(containerAppName, limit) {
        return `${this.getConsoleLogTableName(ResourceType.ContainerApps)} |  where ContainerAppName_s == '${containerAppName}' | order by _timestamp_d desc | project TimeGenerated, Log_s | take ${limit}`;
    }
    appServiceLogsQuery(appServiceResourceId, limit) {
        return `${this.getConsoleLogTableName(ResourceType.AppService)} |  where _ResourceId == '${appServiceResourceId.toLocaleLowerCase()}' | order by TimeGenerated desc | project TimeGenerated, ResultDescription | take ${limit}`;
    }
    functionAppLogsQuery(functionAppName, limit) {
        return `${this.getConsoleLogTableName(ResourceType.FunctionApp)} | where AppRoleName == '${functionAppName}' | order by TimeGenerated desc | project TimeGenerated, Message | take ${limit}`;
    }
    async getApp(resourceType, serviceName) {
        const filter = `tagName eq 'azd-service-name' and tagValue eq '${serviceName}'`;
        const apps = [];
        const resources = await this.azureClient.getResourcesInResourceGroupByFilter(this.resourceGroupName, filter);
        for (const resource of resources) {
            const [type, kind] = resourceType.split("|");
            // app service kind: "app", "app,linux,...", function app kind: "functionapp", "functionapp,linux,..."
            if (resource.id && resource.type === type && (!kind || resource.kind?.startsWith(kind))) {
                apps.push(resource);
            }
        }
        if (apps.length === 0) {
            throw new Error(`No resources found for resource type ${resourceType} with tag azd-service-name=${serviceName}`);
        }
        else if (apps.length > 1) {
            throw new Error(`Multiple resources found for resource type ${resourceType} with tag azd-service-name=${serviceName}`);
        }
        return apps[0];
    }
    async queryAppLogs(resourceType, serviceName) {
        // check if logs have been ingested
        const tableCheckResult = await this.azureClient.queryLogsFromWorkspace(this.logAnalyticsWorkspaceIds[0], this.availableTablesQuery(), new Date(new Date().getTime() - 4 * 60 * 60 * 1000), new Date());
        const consoleLogTableName = this.getConsoleLogTableName(resourceType);
        const tableExists = tableCheckResult.status === "Success"
            && tableCheckResult.tables && tableCheckResult.tables.length > 0
            && tableCheckResult.tables[0].rows.map((row) => row[0]).includes(consoleLogTableName);
        if (!tableExists) {
            throw new Error(`Log table ${consoleLogTableName} not found in the log analytics workspace. Logs might not be ingested yet. Try again later.`);
        }
        const app = await this.getApp(resourceType, serviceName);
        let getLogErrors = [];
        let appLogs = "";
        let getLogSuccess = false;
        let logSearchQuery;
        let lastDeploymentTime;
        if (resourceType === ResourceType.ContainerApps) {
            logSearchQuery = this.containerAppLogsQuery(app.name, this.limit);
            const lastRevision = await this.azureClient.getContainerAppsLatestActiveRevision(this.resourceGroupName, app.name);
            if (lastRevision) {
                lastDeploymentTime = lastRevision.createdTime;
            }
        }
        else if (resourceType === ResourceType.AppService) {
            logSearchQuery = this.appServiceLogsQuery(app.id, this.limit);
            const lastDeployment = await this.azureClient.getWebSiteLatestDeployment(this.resourceGroupName, app.name);
            if (lastDeployment) {
                lastDeploymentTime = lastDeployment.startTime;
            }
        }
        else if (resourceType === ResourceType.FunctionApp) {
            logSearchQuery = this.functionAppLogsQuery(app.name, this.limit);
            const lastDeployment = await this.azureClient.getWebSiteLatestDeployment(this.resourceGroupName, app.name);
            if (lastDeployment) {
                lastDeploymentTime = lastDeployment.startTime;
            }
        }
        else {
            throw new Error(`Unsupported resource type: ${resourceType}`);
        }
        const endTime = new Date();
        const startTime = lastDeploymentTime ?? new Date(endTime.getTime() - 4 * 60 * 60 * 1000);
        for (const logAnalyticsId of this.logAnalyticsWorkspaceIds) {
            try {
                const logSearchResult = await this.azureClient.queryLogsFromWorkspace(logAnalyticsId, logSearchQuery, startTime, endTime);
                if (logSearchResult.status === "Success") {
                    getLogSuccess = true;
                    if (!logSearchResult.tables || logSearchResult.tables.length === 0 || logSearchResult.tables[0].rows.length === 0) {
                        continue;
                    }
                    for (const table of logSearchResult.tables) {
                        for (const row of table.rows) {
                            appLogs += `[${row[0]}] ${row[1]}\n`;
                        }
                    }
                    break; // break if logs are found
                }
            }
            catch (error) {
                getLogErrors.push(`Error retrieving logs for ${app.name} from ${logAnalyticsId}: ${error.message}`);
            }
        }
        if (!getLogSuccess) {
            throw new Error(getLogErrors.join("\n"));
        }
        if (appLogs.length === 0) {
            return "No available logs found in the log analytics workspaces. The compute resources might not be configured to send logs.";
        }
        return `Console Logs for ${serviceName} with resource ID ${app.id} from ${startTime} to ${endTime}:\n${appLogs}`;
    }
}
exports.AzdAppLogRetriever = AzdAppLogRetriever;
async function getAzdAppConsoleLogs(azureClient, workspaceFolder, azdEnvName, overrideResourceGroupName, limit) {
    const toolErrorLogs = [];
    const consoleLogs = [];
    try {
        const azdAppLogRetriever = new AzdAppLogRetriever(azureClient, azdEnvName, overrideResourceGroupName, limit);
        await azdAppLogRetriever.initialize();
        const services = getServicesFromAzureYAML(workspaceFolder);
        for (const [key, value] of Object.entries(services)) {
            try {
                const resourceType = HostToResourceType[value.host];
                consoleLogs.push(await azdAppLogRetriever.queryAppLogs(resourceType, key));
            }
            catch (error) {
                toolErrorLogs.push(`Error finding app logs for service ${key}: ${error.message}`);
            }
        }
    }
    catch (error) {
        toolErrorLogs.push(`${error.message}`);
    }
    let finalLogOutput = "";
    if (consoleLogs.length > 0) {
        finalLogOutput += `App Console Logs retrieved: \n ${consoleLogs.join('\n\n')}\n`;
    }
    if (toolErrorLogs.length > 0) {
        finalLogOutput += `Error during retrieval of app logs of azd project:  \n ${toolErrorLogs.join('\n')}`;
    }
    return finalLogOutput;
}
