// containerization-assist-mcp/dist-cjs/src/infra/security/osv-scanner/index.js
"use strict";
// OSV Scanner using Google's OSV API (https://osv.dev)
// Supports Maven packages with rate-limited batch queries
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkOSVAvailability = checkOSVAvailability;
exports.scanImageWithOSV = scanImageWithOSV;
const errors_1 = require("../../../lib/errors");
const types_1 = require("../../../types");
const scanner_common_1 = require("../scanner-common");
const package_extractor_1 = require("./package-extractor");
const osv_api_1 = require("./osv-api");
async function checkOSVAvailability(logger) {
    try {
        await osv_api_1.osvRateLimiter.acquire();
        const response = await fetch('https://api.osv.dev/v1/query', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                package: { name: 'test-availability-check', ecosystem: 'npm' },
                version: '1.0.0',
            }),
        });
        logger.debug({ status: response.status }, 'OSV API availability check');
        const isAvailable = response.ok || (response.status >= 400 && response.status < 500);
        if (isAvailable)
            return (0, types_1.Success)('osv-api-available');
        return (0, types_1.Failure)(`OSV API returned server error: ${response.status}`, {
            message: `OSV API is experiencing issues (HTTP ${response.status})`,
            hint: 'The OSV API server is not responding correctly',
            resolution: 'Wait a few moments and try again, or use an alternative scanner',
            details: { status: response.status, statusText: response.statusText },
        });
    }
    catch (error) {
        return (0, types_1.Failure)('OSV API not accessible', {
            message: 'Cannot reach OSV API',
            hint: 'OSV API requires network connectivity',
            resolution: 'Check your network connection and try again',
            details: { error: (0, errors_1.extractErrorMessage)(error) },
        });
    }
}
async function scanImageWithOSV(docker, imageId, logger) {
    if (!(0, scanner_common_1.validateImageId)(imageId)) {
        return scanner_common_1.ScannerErrors.invalidImageId(imageId);
    }
    const availabilityCheck = await checkOSVAvailability(logger);
    if (!availabilityCheck.ok) {
        return (0, types_1.Failure)(availabilityCheck.error, availabilityCheck.guidance);
    }
    (0, scanner_common_1.logScanStart)(logger, 'OSV', 'api', imageId);
    try {
        const packages = await (0, package_extractor_1.extractPackagesFromImage)(docker, imageId, logger);
        logger.info({
            packageCount: packages.length,
            ecosystems: [...new Set(packages.map((p) => p.ecosystem))],
            samplePackages: packages
                .slice(0, 10)
                .map((p) => ({ name: p.name, version: p.version, ecosystem: p.ecosystem })),
        }, 'Extracted packages for OSV scan');
        if (packages.length === 0) {
            logger.warn({ imageId }, 'No packages found - returning empty scan result');
            return (0, types_1.Success)({
                imageId,
                vulnerabilities: [],
                totalVulnerabilities: 0,
                criticalCount: 0,
                highCount: 0,
                mediumCount: 0,
                lowCount: 0,
                negligibleCount: 0,
                unknownCount: 0,
                scanDate: new Date(),
            });
        }
        const vulnerabilityMap = await (0, osv_api_1.queryOSVBatch)(packages, logger);
        logger.info({ vulnCount: vulnerabilityMap.size }, 'OSV query completed');
        const vulnerabilities = [];
        const counter = new scanner_common_1.SeverityCounter();
        for (const [pkgKey, vulns] of vulnerabilityMap.entries()) {
            const pkg = packages.find((p) => `${p.name}@${p.version}` === pkgKey);
            if (!pkg)
                continue;
            for (const vuln of vulns) {
                const severity = (0, scanner_common_1.normalizeSeverity)((0, osv_api_1.extractSeverity)(vuln));
                counter.increment(severity);
                const vulnEntry = {
                    id: vuln.id,
                    severity,
                    package: pkg.name,
                    version: pkg.version,
                    description: vuln.summary || vuln.details || 'No description available',
                };
                const fixedVersion = (0, osv_api_1.extractFixedVersion)(vuln);
                if (fixedVersion) {
                    vulnEntry.fixedVersion = fixedVersion;
                }
                vulnerabilities.push(vulnEntry);
            }
        }
        const scanResult = {
            imageId,
            vulnerabilities,
            scanDate: new Date(),
            ...counter.getCounts(),
        };
        (0, scanner_common_1.logScanComplete)(logger, 'OSV', imageId, scanResult.totalVulnerabilities, scanResult.criticalCount, scanResult.highCount);
        return (0, types_1.Success)(scanResult);
    }
    catch (error) {
        const errorMessage = (0, errors_1.extractErrorMessage)(error);
        logger.error({ error: errorMessage, imageId }, 'OSV scan failed');
        return scanner_common_1.ScannerErrors.scanExecutionError('OSV', imageId, errorMessage);
    }
}
