// codetocloudmcp/out/common/telemetry/telemetryUtil.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TelemetryPropertyHelper = exports.ResultCode = void 0;
exports.isInternal = isInternal;
exports.getConfig = getConfig;
const fs = __importStar(require("fs"));
const helpers_1 = require("../helpers");
const stringHelpers_1 = require("../stringHelpers");
var ResultCode;
(function (ResultCode) {
    ResultCode["PreInvoked"] = "PreInvoked";
    ResultCode["Invoked"] = "Invoked";
    ResultCode["InProgress"] = "InProgress";
    ResultCode["Success"] = "Success";
    ResultCode["Failure"] = "Failure";
    ResultCode["Cancelled"] = "Cancelled";
})(ResultCode || (exports.ResultCode = ResultCode = {}));
;
function isInternal() {
    const userDomain = process.env.USERDNSDOMAIN;
    return typeof userDomain === 'string' && userDomain.endsWith('.corp.microsoft.com');
}
function getConfig(key, defaultValue) {
    function convertValue(value) {
        if (typeof defaultValue === 'boolean') {
            if (typeof value === 'string') {
                return (value.toLowerCase() === 'true');
            }
            return Boolean(value);
        }
        if (typeof defaultValue === 'number') {
            const num = Number(value);
            return (isNaN(num) ? defaultValue : num);
        }
        return value;
    }
    if (process.env[key] !== undefined) {
        return convertValue(process.env[key]);
    }
    const configPath = (0, helpers_1.findFileIteratively)(__dirname, 'config.json');
    if (configPath === undefined) {
        return defaultValue;
    }
    const config = JSON.parse(fs.readFileSync(configPath, 'utf8'));
    if (config[key] === undefined) {
        return defaultValue;
    }
    return convertValue(config[key]);
}
class TelemetryPropertyHelper {
    static InvalidSubscriptionIdPlaceholder = "[INVALID_SUBSCRIPTION_ID]";
    static InvalidIaCTypePlaceholder = "[INVALID_IAC_TYPE]";
    static InvalidResourceTypePlaceholder = "[INVALID_RESOURCE_TYPE]";
    static InvalidAzureResourceIdPlaceholder = "[INVALID_RESOURCE_ID]";
    static MaskedResourceGroupPlaceholder = "[MASKED_RG]";
    static MaskedResourceNamePlaceholder = "[MASKED_NAME]";
    static prepareSubscriptionIdForTelemetry(subscriptionId) {
        if (!subscriptionId) {
            return "";
        }
        return (0, stringHelpers_1.checkGUIDFormat)(subscriptionId) ? subscriptionId : TelemetryPropertyHelper.InvalidSubscriptionIdPlaceholder;
    }
    static prepareResourceTypesForTelemetry(resourceTypes) {
        if (!resourceTypes || resourceTypes.length === 0) {
            return "";
        }
        if (typeof resourceTypes[0] === "object" && 'type' in resourceTypes[0]) {
            return resourceTypes.map(rt => (0, stringHelpers_1.isValidResourceType)(rt.type) ? JSON.stringify(rt) : TelemetryPropertyHelper.InvalidResourceTypePlaceholder)
                .join(';');
        }
        return resourceTypes.map(rt => (0, stringHelpers_1.isValidResourceType)(rt) ? rt : TelemetryPropertyHelper.InvalidResourceTypePlaceholder).join(';');
    }
    static prepareIaCTypeForTelemetry(iacType) {
        if (!iacType) {
            return "";
        }
        return (0, stringHelpers_1.checkValidIaCType)(iacType) ? iacType : TelemetryPropertyHelper.InvalidIaCTypePlaceholder;
    }
    static prepareAzureResourceIdForTelemetry(resourceId) {
        if (!resourceId) {
            return "";
        }
        const resourceIdRegex = /^\/subscriptions\/([^\/]+)\/resourceGroups\/([^\/]+)\/providers\/([^\/]+)\/([^\/]+)\/([^\/]+)(?:\/.*)?$/i;
        const match = resourceId.match(resourceIdRegex);
        if (!match) {
            return TelemetryPropertyHelper.InvalidAzureResourceIdPlaceholder;
        }
        const [, subscriptionId, , provider, resourceType,] = match;
        if (!(0, stringHelpers_1.checkGUIDFormat)(subscriptionId)) {
            return TelemetryPropertyHelper.InvalidSubscriptionIdPlaceholder;
        }
        return `/subscriptions/${subscriptionId}/resourceGroups/${TelemetryPropertyHelper.MaskedResourceGroupPlaceholder}/providers/${provider}/${resourceType}/${TelemetryPropertyHelper.MaskedResourceNamePlaceholder}`;
    }
    /**
     * Prepares an array of Azure resource IDs for telemetry by masking sensitive information.
     */
    static prepareAzureResourceIdsForTelemetry(resourceIds) {
        if (!resourceIds || resourceIds.length === 0) {
            return "";
        }
        return resourceIds.map(id => TelemetryPropertyHelper.prepareAzureResourceIdForTelemetry(id)).join(';');
    }
}
exports.TelemetryPropertyHelper = TelemetryPropertyHelper;
