// codetocloudmcp/out/common/telemetry/telemetryClient.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TelemetryClient = void 0;
const appInsights = __importStar(require("applicationinsights")); // using version 2.X to write operation_Name
const uuid_1 = require("uuid");
const toolType_1 = require("../../toolsImplementations/shared/toolType");
const helpers_1 = require("../helpers");
const models_1 = require("../models");
const devdeviceid_1 = require("./devdeviceid");
const telemetryUtil_1 = require("./telemetryUtil");
class TelemetryClient {
    client;
    getCustomerChannel;
    sessionId = (0, uuid_1.v4)();
    correlationId = (0, uuid_1.v4)();
    disableTelemetry = false;
    constructor(getCustomerChannel, disableTelemetry) {
        var instrumentationKey;
        this.getCustomerChannel = getCustomerChannel;
        this.disableTelemetry = disableTelemetry;
        if (models_1.TelemetryRequiredChannels.includes(this.getCustomerChannel())) {
            this.sessionId = process.env.APPMOD_CALLER_SESSION_ID || this.sessionId;
            instrumentationKey = 'InstrumentationKey=b73649cf-ca8f-4768-9e0e-b789d8529db5;IngestionEndpoint=https://centralus-2.in.applicationinsights.azure.com/;LiveEndpoint=https://centralus.livediagnostics.monitor.azure.com/;ApplicationId=3e8bbeb8-8ab7-4c6a-aa85-48caa1e858cd';
        }
        else {
            const packageJson = (0, helpers_1.detectPackageJson)(__dirname);
            instrumentationKey = `INSTRUMENTATIONKEY=${packageJson?.aiKey ?? ''}`;
        }
        this.client = new appInsights.TelemetryClient(instrumentationKey);
        this.client.addTelemetryProcessor((envelope) => {
            if (envelope?.tags && envelope?.data?.baseData?.properties) {
                if (envelope.data.baseData.properties.operationName) {
                    envelope.tags[this.client.context.keys.operationName] =
                        envelope.data.baseData.properties.operationName;
                    delete envelope.data.baseData.properties.operationName;
                }
                if (envelope.data.baseData.properties.sessionId) {
                    envelope.tags[this.client.context.keys.sessionId] =
                        envelope.data.baseData.properties.sessionId;
                    delete envelope.data.baseData.properties.sessionId;
                }
                if (envelope.data.baseData.properties.operationId) {
                    envelope.tags[this.client.context.keys.operationId] =
                        envelope.data.baseData.properties.operationId;
                    delete envelope.data.baseData.properties.operationId;
                }
                if (envelope.data.baseData.properties.deviceId) {
                    envelope.tags[this.client.context.keys.deviceId] =
                        envelope.data.baseData.properties.deviceId;
                }
            }
            return true;
        });
    }
    sendEvent(eventName, properties) {
        if (this.disableTelemetry) {
            return;
        }
        this.client.trackTrace({
            message: JSON.stringify({ internal: (0, telemetryUtil_1.isInternal)().toString() }),
            properties: {
                ...(properties || {}),
                operationName: eventName,
                operationId: (0, uuid_1.v4)(),
                sessionId: this.sessionId,
            },
        });
    }
    async initTelemetryProperties(version, toolName, projectId, resultCode = telemetryUtil_1.ResultCode.Invoked) {
        if (toolName === toolType_1.ToolType.GetPlan) {
            this.correlationId = (0, uuid_1.v4)();
        }
        return {
            version: version,
            time: (new Date()).toUTCString(),
            customerChannel: this.getCustomerChannel(),
            responseLatency: 0,
            resultCode: resultCode,
            devDeviceId: await (0, devdeviceid_1.getDeviceId)(),
            correlationId: this.correlationId,
            projectName: projectId,
        };
    }
}
exports.TelemetryClient = TelemetryClient;
