// @azure/arm-appcontainers/dist/commonjs/models/index.js
"use strict";
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.KnownSourceControlOperationState = exports.KnownSessionPoolProvisioningState = exports.KnownSessionNetworkStatus = exports.KnownLifecycleType = exports.KnownPoolManagementType = exports.KnownContainerType = exports.KnownManagedCertificateDomainControlValidation = exports.KnownJobExecutionRunningState = exports.KnownJavaComponentProvisioningState = exports.KnownJavaComponentType = exports.KnownTriggerType = exports.KnownJobProvisioningState = exports.KnownEnvironmentProvisioningState = exports.KnownContainerAppContainerRunningState = exports.KnownContainerAppReplicaRunningState = exports.KnownRevisionRunningState = exports.KnownRevisionProvisioningState = exports.KnownRevisionHealthState = exports.KnownStorageType = exports.KnownType = exports.KnownScheme = exports.KnownIdentitySettingsLifeCycle = exports.KnownLogLevel = exports.KnownAppProtocol = exports.KnownIngressClientCertificateMode = exports.KnownAffinity = exports.KnownAction = exports.KnownBindingType = exports.KnownIngressTransportMethod = exports.KnownActiveRevisionsMode = exports.KnownContainerAppRunningStatus = exports.KnownContainerAppProvisioningState = exports.KnownManagedServiceIdentityType = exports.KnownAccessMode = exports.KnownCertificateProvisioningState = exports.KnownCheckNameAvailabilityReason = exports.KnownConnectedEnvironmentProvisioningState = exports.KnownExtendedLocationTypes = exports.KnownApplicability = exports.KnownCreatedByType = void 0;
/** Known values of {@link CreatedByType} that the service accepts. */
var KnownCreatedByType;
(function (KnownCreatedByType) {
    /** User */
    KnownCreatedByType["User"] = "User";
    /** Application */
    KnownCreatedByType["Application"] = "Application";
    /** ManagedIdentity */
    KnownCreatedByType["ManagedIdentity"] = "ManagedIdentity";
    /** Key */
    KnownCreatedByType["Key"] = "Key";
})(KnownCreatedByType || (exports.KnownCreatedByType = KnownCreatedByType = {}));
/** Known values of {@link Applicability} that the service accepts. */
var KnownApplicability;
(function (KnownApplicability) {
    /** LocationDefault */
    KnownApplicability["LocationDefault"] = "LocationDefault";
    /** Custom */
    KnownApplicability["Custom"] = "Custom";
})(KnownApplicability || (exports.KnownApplicability = KnownApplicability = {}));
/** Known values of {@link ExtendedLocationTypes} that the service accepts. */
var KnownExtendedLocationTypes;
(function (KnownExtendedLocationTypes) {
    /** CustomLocation */
    KnownExtendedLocationTypes["CustomLocation"] = "CustomLocation";
})(KnownExtendedLocationTypes || (exports.KnownExtendedLocationTypes = KnownExtendedLocationTypes = {}));
/** Known values of {@link ConnectedEnvironmentProvisioningState} that the service accepts. */
var KnownConnectedEnvironmentProvisioningState;
(function (KnownConnectedEnvironmentProvisioningState) {
    /** Succeeded */
    KnownConnectedEnvironmentProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownConnectedEnvironmentProvisioningState["Failed"] = "Failed";
    /** Canceled */
    KnownConnectedEnvironmentProvisioningState["Canceled"] = "Canceled";
    /** Waiting */
    KnownConnectedEnvironmentProvisioningState["Waiting"] = "Waiting";
    /** InitializationInProgress */
    KnownConnectedEnvironmentProvisioningState["InitializationInProgress"] = "InitializationInProgress";
    /** InfrastructureSetupInProgress */
    KnownConnectedEnvironmentProvisioningState["InfrastructureSetupInProgress"] = "InfrastructureSetupInProgress";
    /** InfrastructureSetupComplete */
    KnownConnectedEnvironmentProvisioningState["InfrastructureSetupComplete"] = "InfrastructureSetupComplete";
    /** ScheduledForDelete */
    KnownConnectedEnvironmentProvisioningState["ScheduledForDelete"] = "ScheduledForDelete";
})(KnownConnectedEnvironmentProvisioningState || (exports.KnownConnectedEnvironmentProvisioningState = KnownConnectedEnvironmentProvisioningState = {}));
/** Known values of {@link CheckNameAvailabilityReason} that the service accepts. */
var KnownCheckNameAvailabilityReason;
(function (KnownCheckNameAvailabilityReason) {
    /** Invalid */
    KnownCheckNameAvailabilityReason["Invalid"] = "Invalid";
    /** AlreadyExists */
    KnownCheckNameAvailabilityReason["AlreadyExists"] = "AlreadyExists";
})(KnownCheckNameAvailabilityReason || (exports.KnownCheckNameAvailabilityReason = KnownCheckNameAvailabilityReason = {}));
/** Known values of {@link CertificateProvisioningState} that the service accepts. */
var KnownCertificateProvisioningState;
(function (KnownCertificateProvisioningState) {
    /** Succeeded */
    KnownCertificateProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownCertificateProvisioningState["Failed"] = "Failed";
    /** Canceled */
    KnownCertificateProvisioningState["Canceled"] = "Canceled";
    /** DeleteFailed */
    KnownCertificateProvisioningState["DeleteFailed"] = "DeleteFailed";
    /** Pending */
    KnownCertificateProvisioningState["Pending"] = "Pending";
})(KnownCertificateProvisioningState || (exports.KnownCertificateProvisioningState = KnownCertificateProvisioningState = {}));
/** Known values of {@link AccessMode} that the service accepts. */
var KnownAccessMode;
(function (KnownAccessMode) {
    /** ReadOnly */
    KnownAccessMode["ReadOnly"] = "ReadOnly";
    /** ReadWrite */
    KnownAccessMode["ReadWrite"] = "ReadWrite";
})(KnownAccessMode || (exports.KnownAccessMode = KnownAccessMode = {}));
/** Known values of {@link ManagedServiceIdentityType} that the service accepts. */
var KnownManagedServiceIdentityType;
(function (KnownManagedServiceIdentityType) {
    /** None */
    KnownManagedServiceIdentityType["None"] = "None";
    /** SystemAssigned */
    KnownManagedServiceIdentityType["SystemAssigned"] = "SystemAssigned";
    /** UserAssigned */
    KnownManagedServiceIdentityType["UserAssigned"] = "UserAssigned";
    /** SystemAssignedUserAssigned */
    KnownManagedServiceIdentityType["SystemAssignedUserAssigned"] = "SystemAssigned,UserAssigned";
})(KnownManagedServiceIdentityType || (exports.KnownManagedServiceIdentityType = KnownManagedServiceIdentityType = {}));
/** Known values of {@link ContainerAppProvisioningState} that the service accepts. */
var KnownContainerAppProvisioningState;
(function (KnownContainerAppProvisioningState) {
    /** InProgress */
    KnownContainerAppProvisioningState["InProgress"] = "InProgress";
    /** Succeeded */
    KnownContainerAppProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownContainerAppProvisioningState["Failed"] = "Failed";
    /** Canceled */
    KnownContainerAppProvisioningState["Canceled"] = "Canceled";
    /** Deleting */
    KnownContainerAppProvisioningState["Deleting"] = "Deleting";
})(KnownContainerAppProvisioningState || (exports.KnownContainerAppProvisioningState = KnownContainerAppProvisioningState = {}));
/** Known values of {@link ContainerAppRunningStatus} that the service accepts. */
var KnownContainerAppRunningStatus;
(function (KnownContainerAppRunningStatus) {
    /** Container App is transitioning between Stopped and Running states. */
    KnownContainerAppRunningStatus["Progressing"] = "Progressing";
    /** Container App is in Running state. */
    KnownContainerAppRunningStatus["Running"] = "Running";
    /** Container App is in Stopped state. */
    KnownContainerAppRunningStatus["Stopped"] = "Stopped";
    /** Container App Job is in Suspended state. */
    KnownContainerAppRunningStatus["Suspended"] = "Suspended";
    /** Container App Job is in Ready state. */
    KnownContainerAppRunningStatus["Ready"] = "Ready";
})(KnownContainerAppRunningStatus || (exports.KnownContainerAppRunningStatus = KnownContainerAppRunningStatus = {}));
/** Known values of {@link ActiveRevisionsMode} that the service accepts. */
var KnownActiveRevisionsMode;
(function (KnownActiveRevisionsMode) {
    /** Multiple */
    KnownActiveRevisionsMode["Multiple"] = "Multiple";
    /** Single */
    KnownActiveRevisionsMode["Single"] = "Single";
})(KnownActiveRevisionsMode || (exports.KnownActiveRevisionsMode = KnownActiveRevisionsMode = {}));
/** Known values of {@link IngressTransportMethod} that the service accepts. */
var KnownIngressTransportMethod;
(function (KnownIngressTransportMethod) {
    /** Auto */
    KnownIngressTransportMethod["Auto"] = "auto";
    /** Http */
    KnownIngressTransportMethod["Http"] = "http";
    /** Http2 */
    KnownIngressTransportMethod["Http2"] = "http2";
    /** Tcp */
    KnownIngressTransportMethod["Tcp"] = "tcp";
})(KnownIngressTransportMethod || (exports.KnownIngressTransportMethod = KnownIngressTransportMethod = {}));
/** Known values of {@link BindingType} that the service accepts. */
var KnownBindingType;
(function (KnownBindingType) {
    /** Disabled */
    KnownBindingType["Disabled"] = "Disabled";
    /** SniEnabled */
    KnownBindingType["SniEnabled"] = "SniEnabled";
})(KnownBindingType || (exports.KnownBindingType = KnownBindingType = {}));
/** Known values of {@link Action} that the service accepts. */
var KnownAction;
(function (KnownAction) {
    /** Allow */
    KnownAction["Allow"] = "Allow";
    /** Deny */
    KnownAction["Deny"] = "Deny";
})(KnownAction || (exports.KnownAction = KnownAction = {}));
/** Known values of {@link Affinity} that the service accepts. */
var KnownAffinity;
(function (KnownAffinity) {
    /** Sticky */
    KnownAffinity["Sticky"] = "sticky";
    /** None */
    KnownAffinity["None"] = "none";
})(KnownAffinity || (exports.KnownAffinity = KnownAffinity = {}));
/** Known values of {@link IngressClientCertificateMode} that the service accepts. */
var KnownIngressClientCertificateMode;
(function (KnownIngressClientCertificateMode) {
    /** Ignore */
    KnownIngressClientCertificateMode["Ignore"] = "ignore";
    /** Accept */
    KnownIngressClientCertificateMode["Accept"] = "accept";
    /** Require */
    KnownIngressClientCertificateMode["Require"] = "require";
})(KnownIngressClientCertificateMode || (exports.KnownIngressClientCertificateMode = KnownIngressClientCertificateMode = {}));
/** Known values of {@link AppProtocol} that the service accepts. */
var KnownAppProtocol;
(function (KnownAppProtocol) {
    /** Http */
    KnownAppProtocol["Http"] = "http";
    /** Grpc */
    KnownAppProtocol["Grpc"] = "grpc";
})(KnownAppProtocol || (exports.KnownAppProtocol = KnownAppProtocol = {}));
/** Known values of {@link LogLevel} that the service accepts. */
var KnownLogLevel;
(function (KnownLogLevel) {
    /** Info */
    KnownLogLevel["Info"] = "info";
    /** Debug */
    KnownLogLevel["Debug"] = "debug";
    /** Warn */
    KnownLogLevel["Warn"] = "warn";
    /** Error */
    KnownLogLevel["Error"] = "error";
})(KnownLogLevel || (exports.KnownLogLevel = KnownLogLevel = {}));
/** Known values of {@link IdentitySettingsLifeCycle} that the service accepts. */
var KnownIdentitySettingsLifeCycle;
(function (KnownIdentitySettingsLifeCycle) {
    /** Init */
    KnownIdentitySettingsLifeCycle["Init"] = "Init";
    /** Main */
    KnownIdentitySettingsLifeCycle["Main"] = "Main";
    /** None */
    KnownIdentitySettingsLifeCycle["None"] = "None";
    /** All */
    KnownIdentitySettingsLifeCycle["All"] = "All";
})(KnownIdentitySettingsLifeCycle || (exports.KnownIdentitySettingsLifeCycle = KnownIdentitySettingsLifeCycle = {}));
/** Known values of {@link Scheme} that the service accepts. */
var KnownScheme;
(function (KnownScheme) {
    /** Http */
    KnownScheme["Http"] = "HTTP";
    /** Https */
    KnownScheme["Https"] = "HTTPS";
})(KnownScheme || (exports.KnownScheme = KnownScheme = {}));
/** Known values of {@link Type} that the service accepts. */
var KnownType;
(function (KnownType) {
    /** Liveness */
    KnownType["Liveness"] = "Liveness";
    /** Readiness */
    KnownType["Readiness"] = "Readiness";
    /** Startup */
    KnownType["Startup"] = "Startup";
})(KnownType || (exports.KnownType = KnownType = {}));
/** Known values of {@link StorageType} that the service accepts. */
var KnownStorageType;
(function (KnownStorageType) {
    /** AzureFile */
    KnownStorageType["AzureFile"] = "AzureFile";
    /** EmptyDir */
    KnownStorageType["EmptyDir"] = "EmptyDir";
    /** Secret */
    KnownStorageType["Secret"] = "Secret";
    /** NfsAzureFile */
    KnownStorageType["NfsAzureFile"] = "NfsAzureFile";
})(KnownStorageType || (exports.KnownStorageType = KnownStorageType = {}));
/** Known values of {@link RevisionHealthState} that the service accepts. */
var KnownRevisionHealthState;
(function (KnownRevisionHealthState) {
    /** Healthy */
    KnownRevisionHealthState["Healthy"] = "Healthy";
    /** Unhealthy */
    KnownRevisionHealthState["Unhealthy"] = "Unhealthy";
    /** None */
    KnownRevisionHealthState["None"] = "None";
})(KnownRevisionHealthState || (exports.KnownRevisionHealthState = KnownRevisionHealthState = {}));
/** Known values of {@link RevisionProvisioningState} that the service accepts. */
var KnownRevisionProvisioningState;
(function (KnownRevisionProvisioningState) {
    /** Provisioning */
    KnownRevisionProvisioningState["Provisioning"] = "Provisioning";
    /** Provisioned */
    KnownRevisionProvisioningState["Provisioned"] = "Provisioned";
    /** Failed */
    KnownRevisionProvisioningState["Failed"] = "Failed";
    /** Deprovisioning */
    KnownRevisionProvisioningState["Deprovisioning"] = "Deprovisioning";
    /** Deprovisioned */
    KnownRevisionProvisioningState["Deprovisioned"] = "Deprovisioned";
})(KnownRevisionProvisioningState || (exports.KnownRevisionProvisioningState = KnownRevisionProvisioningState = {}));
/** Known values of {@link RevisionRunningState} that the service accepts. */
var KnownRevisionRunningState;
(function (KnownRevisionRunningState) {
    /** Running */
    KnownRevisionRunningState["Running"] = "Running";
    /** Processing */
    KnownRevisionRunningState["Processing"] = "Processing";
    /** Stopped */
    KnownRevisionRunningState["Stopped"] = "Stopped";
    /** Degraded */
    KnownRevisionRunningState["Degraded"] = "Degraded";
    /** Failed */
    KnownRevisionRunningState["Failed"] = "Failed";
    /** Unknown */
    KnownRevisionRunningState["Unknown"] = "Unknown";
})(KnownRevisionRunningState || (exports.KnownRevisionRunningState = KnownRevisionRunningState = {}));
/** Known values of {@link ContainerAppReplicaRunningState} that the service accepts. */
var KnownContainerAppReplicaRunningState;
(function (KnownContainerAppReplicaRunningState) {
    /** Running */
    KnownContainerAppReplicaRunningState["Running"] = "Running";
    /** NotRunning */
    KnownContainerAppReplicaRunningState["NotRunning"] = "NotRunning";
    /** Unknown */
    KnownContainerAppReplicaRunningState["Unknown"] = "Unknown";
})(KnownContainerAppReplicaRunningState || (exports.KnownContainerAppReplicaRunningState = KnownContainerAppReplicaRunningState = {}));
/** Known values of {@link ContainerAppContainerRunningState} that the service accepts. */
var KnownContainerAppContainerRunningState;
(function (KnownContainerAppContainerRunningState) {
    /** Running */
    KnownContainerAppContainerRunningState["Running"] = "Running";
    /** Terminated */
    KnownContainerAppContainerRunningState["Terminated"] = "Terminated";
    /** Waiting */
    KnownContainerAppContainerRunningState["Waiting"] = "Waiting";
})(KnownContainerAppContainerRunningState || (exports.KnownContainerAppContainerRunningState = KnownContainerAppContainerRunningState = {}));
/** Known values of {@link EnvironmentProvisioningState} that the service accepts. */
var KnownEnvironmentProvisioningState;
(function (KnownEnvironmentProvisioningState) {
    /** Succeeded */
    KnownEnvironmentProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownEnvironmentProvisioningState["Failed"] = "Failed";
    /** Canceled */
    KnownEnvironmentProvisioningState["Canceled"] = "Canceled";
    /** Waiting */
    KnownEnvironmentProvisioningState["Waiting"] = "Waiting";
    /** InitializationInProgress */
    KnownEnvironmentProvisioningState["InitializationInProgress"] = "InitializationInProgress";
    /** InfrastructureSetupInProgress */
    KnownEnvironmentProvisioningState["InfrastructureSetupInProgress"] = "InfrastructureSetupInProgress";
    /** InfrastructureSetupComplete */
    KnownEnvironmentProvisioningState["InfrastructureSetupComplete"] = "InfrastructureSetupComplete";
    /** ScheduledForDelete */
    KnownEnvironmentProvisioningState["ScheduledForDelete"] = "ScheduledForDelete";
    /** UpgradeRequested */
    KnownEnvironmentProvisioningState["UpgradeRequested"] = "UpgradeRequested";
    /** UpgradeFailed */
    KnownEnvironmentProvisioningState["UpgradeFailed"] = "UpgradeFailed";
})(KnownEnvironmentProvisioningState || (exports.KnownEnvironmentProvisioningState = KnownEnvironmentProvisioningState = {}));
/** Known values of {@link JobProvisioningState} that the service accepts. */
var KnownJobProvisioningState;
(function (KnownJobProvisioningState) {
    /** InProgress */
    KnownJobProvisioningState["InProgress"] = "InProgress";
    /** Succeeded */
    KnownJobProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownJobProvisioningState["Failed"] = "Failed";
    /** Canceled */
    KnownJobProvisioningState["Canceled"] = "Canceled";
    /** Deleting */
    KnownJobProvisioningState["Deleting"] = "Deleting";
})(KnownJobProvisioningState || (exports.KnownJobProvisioningState = KnownJobProvisioningState = {}));
/** Known values of {@link TriggerType} that the service accepts. */
var KnownTriggerType;
(function (KnownTriggerType) {
    /** Schedule */
    KnownTriggerType["Schedule"] = "Schedule";
    /** Event */
    KnownTriggerType["Event"] = "Event";
    /** Manual */
    KnownTriggerType["Manual"] = "Manual";
})(KnownTriggerType || (exports.KnownTriggerType = KnownTriggerType = {}));
/** Known values of {@link JavaComponentType} that the service accepts. */
var KnownJavaComponentType;
(function (KnownJavaComponentType) {
    /** SpringBootAdmin */
    KnownJavaComponentType["SpringBootAdmin"] = "SpringBootAdmin";
    /** SpringCloudEureka */
    KnownJavaComponentType["SpringCloudEureka"] = "SpringCloudEureka";
    /** SpringCloudConfig */
    KnownJavaComponentType["SpringCloudConfig"] = "SpringCloudConfig";
})(KnownJavaComponentType || (exports.KnownJavaComponentType = KnownJavaComponentType = {}));
/** Known values of {@link JavaComponentProvisioningState} that the service accepts. */
var KnownJavaComponentProvisioningState;
(function (KnownJavaComponentProvisioningState) {
    /** Succeeded */
    KnownJavaComponentProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownJavaComponentProvisioningState["Failed"] = "Failed";
    /** Canceled */
    KnownJavaComponentProvisioningState["Canceled"] = "Canceled";
    /** Deleting */
    KnownJavaComponentProvisioningState["Deleting"] = "Deleting";
    /** InProgress */
    KnownJavaComponentProvisioningState["InProgress"] = "InProgress";
})(KnownJavaComponentProvisioningState || (exports.KnownJavaComponentProvisioningState = KnownJavaComponentProvisioningState = {}));
/** Known values of {@link JobExecutionRunningState} that the service accepts. */
var KnownJobExecutionRunningState;
(function (KnownJobExecutionRunningState) {
    /** Running */
    KnownJobExecutionRunningState["Running"] = "Running";
    /** Processing */
    KnownJobExecutionRunningState["Processing"] = "Processing";
    /** Stopped */
    KnownJobExecutionRunningState["Stopped"] = "Stopped";
    /** Degraded */
    KnownJobExecutionRunningState["Degraded"] = "Degraded";
    /** Failed */
    KnownJobExecutionRunningState["Failed"] = "Failed";
    /** Unknown */
    KnownJobExecutionRunningState["Unknown"] = "Unknown";
    /** Succeeded */
    KnownJobExecutionRunningState["Succeeded"] = "Succeeded";
})(KnownJobExecutionRunningState || (exports.KnownJobExecutionRunningState = KnownJobExecutionRunningState = {}));
/** Known values of {@link ManagedCertificateDomainControlValidation} that the service accepts. */
var KnownManagedCertificateDomainControlValidation;
(function (KnownManagedCertificateDomainControlValidation) {
    /** Cname */
    KnownManagedCertificateDomainControlValidation["Cname"] = "CNAME";
    /** Http */
    KnownManagedCertificateDomainControlValidation["Http"] = "HTTP";
    /** TXT */
    KnownManagedCertificateDomainControlValidation["TXT"] = "TXT";
})(KnownManagedCertificateDomainControlValidation || (exports.KnownManagedCertificateDomainControlValidation = KnownManagedCertificateDomainControlValidation = {}));
/** Known values of {@link ContainerType} that the service accepts. */
var KnownContainerType;
(function (KnownContainerType) {
    /** CustomContainer */
    KnownContainerType["CustomContainer"] = "CustomContainer";
    /** PythonLTS */
    KnownContainerType["PythonLTS"] = "PythonLTS";
})(KnownContainerType || (exports.KnownContainerType = KnownContainerType = {}));
/** Known values of {@link PoolManagementType} that the service accepts. */
var KnownPoolManagementType;
(function (KnownPoolManagementType) {
    /** Manual */
    KnownPoolManagementType["Manual"] = "Manual";
    /** Dynamic */
    KnownPoolManagementType["Dynamic"] = "Dynamic";
})(KnownPoolManagementType || (exports.KnownPoolManagementType = KnownPoolManagementType = {}));
/** Known values of {@link LifecycleType} that the service accepts. */
var KnownLifecycleType;
(function (KnownLifecycleType) {
    /** Timed */
    KnownLifecycleType["Timed"] = "Timed";
    /** OnContainerExit */
    KnownLifecycleType["OnContainerExit"] = "OnContainerExit";
})(KnownLifecycleType || (exports.KnownLifecycleType = KnownLifecycleType = {}));
/** Known values of {@link SessionNetworkStatus} that the service accepts. */
var KnownSessionNetworkStatus;
(function (KnownSessionNetworkStatus) {
    /** EgressEnabled */
    KnownSessionNetworkStatus["EgressEnabled"] = "EgressEnabled";
    /** EgressDisabled */
    KnownSessionNetworkStatus["EgressDisabled"] = "EgressDisabled";
})(KnownSessionNetworkStatus || (exports.KnownSessionNetworkStatus = KnownSessionNetworkStatus = {}));
/** Known values of {@link SessionPoolProvisioningState} that the service accepts. */
var KnownSessionPoolProvisioningState;
(function (KnownSessionPoolProvisioningState) {
    /** InProgress */
    KnownSessionPoolProvisioningState["InProgress"] = "InProgress";
    /** Succeeded */
    KnownSessionPoolProvisioningState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownSessionPoolProvisioningState["Failed"] = "Failed";
    /** Canceled */
    KnownSessionPoolProvisioningState["Canceled"] = "Canceled";
    /** Deleting */
    KnownSessionPoolProvisioningState["Deleting"] = "Deleting";
})(KnownSessionPoolProvisioningState || (exports.KnownSessionPoolProvisioningState = KnownSessionPoolProvisioningState = {}));
/** Known values of {@link SourceControlOperationState} that the service accepts. */
var KnownSourceControlOperationState;
(function (KnownSourceControlOperationState) {
    /** InProgress */
    KnownSourceControlOperationState["InProgress"] = "InProgress";
    /** Succeeded */
    KnownSourceControlOperationState["Succeeded"] = "Succeeded";
    /** Failed */
    KnownSourceControlOperationState["Failed"] = "Failed";
    /** Canceled */
    KnownSourceControlOperationState["Canceled"] = "Canceled";
})(KnownSourceControlOperationState || (exports.KnownSourceControlOperationState = KnownSourceControlOperationState = {}));
