// @azure/arm-appcontainers/dist/commonjs/models/parameters.js
"use strict";
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.location1 = exports.sourceControlEnvelope = exports.sourceControlName = exports.sessionPoolEnvelope1 = exports.sessionPoolEnvelope = exports.sessionPoolName = exports.storageEnvelope1 = exports.managedCertificateEnvelope1 = exports.managedCertificateEnvelope = exports.managedCertificateName = exports.environmentEnvelope1 = exports.javaComponentEnvelope = exports.name = exports.environmentName1 = exports.jobExecutionName = exports.template = exports.jobEnvelope1 = exports.jobEnvelope = exports.apiName = exports.detectorName1 = exports.jobName = exports.environmentName = exports.detectorName = exports.replicaName = exports.revisionName = exports.filter = exports.containerAppName1 = exports.customHostname = exports.containerAppEnvelope = exports.storageEnvelope = exports.storageName = exports.daprComponentEnvelope = exports.componentName = exports.certificateEnvelope1 = exports.certificateEnvelope = exports.certificateName = exports.checkNameAvailabilityRequest = exports.environmentEnvelope = exports.connectedEnvironmentName = exports.location = exports.nextLink = exports.authConfigEnvelope = exports.contentType = exports.authConfigName = exports.apiVersion = exports.containerAppName = exports.resourceGroupName = exports.subscriptionId = exports.$host = exports.accept = void 0;
exports.environmentName2 = void 0;
const mappers_js_1 = require("../models/mappers.js");
exports.accept = {
    parameterPath: "accept",
    mapper: {
        defaultValue: "application/json",
        isConstant: true,
        serializedName: "Accept",
        type: {
            name: "String",
        },
    },
};
exports.$host = {
    parameterPath: "$host",
    mapper: {
        serializedName: "$host",
        required: true,
        type: {
            name: "String",
        },
    },
    skipEncoding: true,
};
exports.subscriptionId = {
    parameterPath: "subscriptionId",
    mapper: {
        constraints: {
            MinLength: 1,
        },
        serializedName: "subscriptionId",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.resourceGroupName = {
    parameterPath: "resourceGroupName",
    mapper: {
        constraints: {
            MaxLength: 90,
            MinLength: 1,
        },
        serializedName: "resourceGroupName",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.containerAppName = {
    parameterPath: "containerAppName",
    mapper: {
        serializedName: "containerAppName",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.apiVersion = {
    parameterPath: "apiVersion",
    mapper: {
        defaultValue: "2025-01-01",
        isConstant: true,
        serializedName: "api-version",
        type: {
            name: "String",
        },
    },
};
exports.authConfigName = {
    parameterPath: "authConfigName",
    mapper: {
        serializedName: "authConfigName",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.contentType = {
    parameterPath: ["options", "contentType"],
    mapper: {
        defaultValue: "application/json",
        isConstant: true,
        serializedName: "Content-Type",
        type: {
            name: "String",
        },
    },
};
exports.authConfigEnvelope = {
    parameterPath: "authConfigEnvelope",
    mapper: mappers_js_1.AuthConfig,
};
exports.nextLink = {
    parameterPath: "nextLink",
    mapper: {
        serializedName: "nextLink",
        required: true,
        type: {
            name: "String",
        },
    },
    skipEncoding: true,
};
exports.location = {
    parameterPath: "location",
    mapper: {
        constraints: {
            MinLength: 1,
        },
        serializedName: "location",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.connectedEnvironmentName = {
    parameterPath: "connectedEnvironmentName",
    mapper: {
        serializedName: "connectedEnvironmentName",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.environmentEnvelope = {
    parameterPath: "environmentEnvelope",
    mapper: mappers_js_1.ConnectedEnvironment,
};
exports.checkNameAvailabilityRequest = {
    parameterPath: "checkNameAvailabilityRequest",
    mapper: mappers_js_1.CheckNameAvailabilityRequest,
};
exports.certificateName = {
    parameterPath: "certificateName",
    mapper: {
        serializedName: "certificateName",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.certificateEnvelope = {
    parameterPath: ["options", "certificateEnvelope"],
    mapper: mappers_js_1.Certificate,
};
exports.certificateEnvelope1 = {
    parameterPath: "certificateEnvelope",
    mapper: mappers_js_1.CertificatePatch,
};
exports.componentName = {
    parameterPath: "componentName",
    mapper: {
        serializedName: "componentName",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.daprComponentEnvelope = {
    parameterPath: "daprComponentEnvelope",
    mapper: mappers_js_1.DaprComponent,
};
exports.storageName = {
    parameterPath: "storageName",
    mapper: {
        serializedName: "storageName",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.storageEnvelope = {
    parameterPath: "storageEnvelope",
    mapper: mappers_js_1.ConnectedEnvironmentStorage,
};
exports.containerAppEnvelope = {
    parameterPath: "containerAppEnvelope",
    mapper: mappers_js_1.ContainerApp,
};
exports.customHostname = {
    parameterPath: ["options", "customHostname"],
    mapper: {
        serializedName: "customHostname",
        type: {
            name: "String",
        },
    },
};
exports.containerAppName1 = {
    parameterPath: "containerAppName",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[-\\w\\._\\(\\)]+$"),
        },
        serializedName: "containerAppName",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.filter = {
    parameterPath: ["options", "filter"],
    mapper: {
        serializedName: "$filter",
        type: {
            name: "String",
        },
    },
};
exports.revisionName = {
    parameterPath: "revisionName",
    mapper: {
        serializedName: "revisionName",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.replicaName = {
    parameterPath: "replicaName",
    mapper: {
        serializedName: "replicaName",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.detectorName = {
    parameterPath: "detectorName",
    mapper: {
        serializedName: "detectorName",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.environmentName = {
    parameterPath: "environmentName",
    mapper: {
        serializedName: "environmentName",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.jobName = {
    parameterPath: "jobName",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[-\\w\\._\\(\\)]+$"),
        },
        serializedName: "jobName",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.detectorName1 = {
    parameterPath: "detectorName",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[-\\w\\._\\(\\)]+$"),
        },
        serializedName: "detectorName",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.apiName = {
    parameterPath: "apiName",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[-\\w\\._]+$"),
        },
        serializedName: "apiName",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.jobEnvelope = {
    parameterPath: "jobEnvelope",
    mapper: mappers_js_1.Job,
};
exports.jobEnvelope1 = {
    parameterPath: "jobEnvelope",
    mapper: mappers_js_1.JobPatchProperties,
};
exports.template = {
    parameterPath: ["options", "template"],
    mapper: mappers_js_1.JobExecutionTemplate,
};
exports.jobExecutionName = {
    parameterPath: "jobExecutionName",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[-\\w\\._\\(\\)]+$"),
        },
        serializedName: "jobExecutionName",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.environmentName1 = {
    parameterPath: "environmentName",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[-\\w\\._\\(\\)]+$"),
        },
        serializedName: "environmentName",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.name = {
    parameterPath: "name",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[-\\w\\._\\(\\)]+$"),
        },
        serializedName: "name",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.javaComponentEnvelope = {
    parameterPath: "javaComponentEnvelope",
    mapper: mappers_js_1.JavaComponent,
};
exports.environmentEnvelope1 = {
    parameterPath: "environmentEnvelope",
    mapper: mappers_js_1.ManagedEnvironment,
};
exports.managedCertificateName = {
    parameterPath: "managedCertificateName",
    mapper: {
        serializedName: "managedCertificateName",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.managedCertificateEnvelope = {
    parameterPath: ["options", "managedCertificateEnvelope"],
    mapper: mappers_js_1.ManagedCertificate,
};
exports.managedCertificateEnvelope1 = {
    parameterPath: "managedCertificateEnvelope",
    mapper: mappers_js_1.ManagedCertificatePatch,
};
exports.storageEnvelope1 = {
    parameterPath: "storageEnvelope",
    mapper: mappers_js_1.ManagedEnvironmentStorage,
};
exports.sessionPoolName = {
    parameterPath: "sessionPoolName",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[a-z][a-z0-9]*$"),
            MaxLength: 63,
            MinLength: 3,
        },
        serializedName: "sessionPoolName",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.sessionPoolEnvelope = {
    parameterPath: "sessionPoolEnvelope",
    mapper: mappers_js_1.SessionPool,
};
exports.sessionPoolEnvelope1 = {
    parameterPath: "sessionPoolEnvelope",
    mapper: mappers_js_1.SessionPoolUpdatableProperties,
};
exports.sourceControlName = {
    parameterPath: "sourceControlName",
    mapper: {
        serializedName: "sourceControlName",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.sourceControlEnvelope = {
    parameterPath: "sourceControlEnvelope",
    mapper: mappers_js_1.SourceControl,
};
exports.location1 = {
    parameterPath: "location",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[-\\w\\._]+$"),
        },
        serializedName: "location",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.environmentName2 = {
    parameterPath: "environmentName",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[-\\w\\._]+$"),
        },
        serializedName: "environmentName",
        required: true,
        type: {
            name: "String",
        },
    },
};
