// containerization-assist-mcp/dist/src/sdk/index.js
/**
 * Containerization Assist SDK
 *
 * Provides direct access to all 11 containerization tools without requiring
 * the MCP (Model Context Protocol) server infrastructure.
 *
 * This SDK is designed for:
 * - VS Code extension developers integrating with Copilot
 * - Direct programmatic usage without MCP overhead
 * - Lightweight tool execution in Node.js applications
 *
 * @example
 * ```typescript
 * import { analyzeRepo, buildImageContext, scanImage } from 'containerization-assist-mcp/sdk';
 *
 * // Full containerization workflow
 * const analysis = await analyzeRepo({ repositoryPath: './my-app' });
 * const build = await buildImageContext({ path: './my-app', imageName: 'myapp:v1' });
 * const scan = await scanImage({ imageId: 'myapp:v1' });
 * ```
 *
 * @packageDocumentation
 */
import { executeTool } from './executor.js';
// ===== TOOL AND SCHEMA IMPORTS =====
// Consolidated imports - tools and schemas from each module
// analyze-repo
import analyzeRepoTool from '../tools/analyze-repo/tool.js';
// generate-dockerfile
import generateDockerfileTool from '../tools/generate-dockerfile/tool.js';
// fix-dockerfile
import fixDockerfileTool from '../tools/fix-dockerfile/tool.js';
// build-image-context
import buildImageContextTool from '../tools/build-image-context/tool.js';
// scan-image
import scanImageTool from '../tools/scan-image/tool.js';
// tag-image
import tagImageTool from '../tools/tag-image/tool.js';
// push-image
import pushImageTool from '../tools/push-image/tool.js';
// generate-k8s-manifests
import generateK8sManifestsTool from '../tools/generate-k8s-manifests/tool.js';
// prepare-cluster
import prepareClusterTool from '../tools/prepare-cluster/tool.js';
// verify-deploy
import verifyDeployTool from '../tools/verify-deploy/tool.js';
// ops
import opsTool from '../tools/ops/tool.js';
export { Success, Failure } from '../types/core.js';
// Full type exports available via sdk/types
// import type { ... } from 'containerization-assist-mcp/sdk/types';
// ===== SDK FUNCTION FACTORY =====
/**
 * Create a typed SDK function from a tool.
 *
 * This factory ensures consistent behavior across all SDK functions
 * while preserving full type inference for inputs and outputs.
 *
 * @param tool - The tool to wrap
 * @returns A typed function that executes the tool
 */
function createSDKFunction(tool) {
    return (input, options) => executeTool(tool, input, options);
}
// ===== SDK FUNCTION EXPORTS =====
// ----- Analysis Tools -----
/**
 * Analyze a repository to detect language, framework, and dependencies.
 */
export const analyzeRepo = createSDKFunction(analyzeRepoTool);
// ----- Dockerfile Tools -----
/**
 * Generate Dockerfile recommendations for a repository.
 */
export const generateDockerfile = createSDKFunction(generateDockerfileTool);
/**
 * Fix and optimize an existing Dockerfile.
 */
export const fixDockerfile = createSDKFunction(fixDockerfileTool);
// ----- Image Tools -----
/**
 * Prepare Docker build context with security analysis and optimized commands.
 * Returns structured guidance for executing builds - does not execute Docker directly.
 */
export const buildImageContext = createSDKFunction(buildImageContextTool);
/**
 * Scan a Docker image for security vulnerabilities.
 * Requires Trivy to be installed for full functionality.
 */
export const scanImage = createSDKFunction(scanImageTool);
/**
 * Tag a Docker image with additional tags.
 * Requires Docker daemon to be running.
 */
export const tagImage = createSDKFunction(tagImageTool);
/**
 * Push a Docker image to a registry.
 * Requires Docker daemon and registry authentication.
 */
export const pushImage = createSDKFunction(pushImageTool);
// ----- Kubernetes Tools -----
/**
 * Generate Kubernetes manifests for deployment.
 */
export const generateK8sManifests = createSDKFunction(generateK8sManifestsTool);
/**
 * Prepare a Kubernetes cluster for deployment.
 * Requires kubectl configured with cluster access.
 */
export const prepareCluster = createSDKFunction(prepareClusterTool);
/**
 * Verify a Kubernetes deployment status.
 * Requires kubectl configured with cluster access.
 */
export const verifyDeploy = createSDKFunction(verifyDeployTool);
// ----- Operational Tools -----
/**
 * Operational utilities (ping, status).
 */
export const ops = createSDKFunction(opsTool);
// ===== ADVANCED: DIRECT TOOL ACCESS =====
/**
 * Direct access to all 11 tool objects for advanced use cases.
 *
 * Use this when you need:
 * - Access to tool schemas for validation
 * - Tool metadata and descriptions
 * - Custom execution patterns
 *
 * Uses `satisfies` to ensure type safety:
 * - TypeScript will error if a tool is missing
 * - TypeScript will error if a tool has the wrong type
 * - TypeScript will error if a key is misspelled
 *
 * @example
 * ```typescript
 * import { tools } from 'containerization-assist-mcp/sdk';
 *
 * // Access tool schema
 * const schema = tools.analyzeRepo.schema;
 *
 * // Access tool metadata
 * console.log(tools.buildImageContext.description);
 * ```
 */
export const tools = {
    // ===== Analysis =====
    /** Analyze repository structure, detect languages, frameworks, and dependencies */
    analyzeRepo: analyzeRepoTool,
    // ===== Dockerfile =====
    /** Generate optimized Dockerfile with security best practices */
    generateDockerfile: generateDockerfileTool,
    /** Fix and optimize existing Dockerfile issues */
    fixDockerfile: fixDockerfileTool,
    // ===== Image Operations =====
    /** Prepare Docker build context with security analysis (returns build commands) */
    buildImageContext: buildImageContextTool,
    /** Scan image for security vulnerabilities (requires Trivy) */
    scanImage: scanImageTool,
    /** Tag Docker image with additional tags */
    tagImage: tagImageTool,
    /** Push image to container registry */
    pushImage: pushImageTool,
    // ===== Kubernetes =====
    /** Generate Kubernetes deployment manifests */
    generateK8sManifests: generateK8sManifestsTool,
    /** Prepare Kubernetes cluster namespace and prerequisites */
    prepareCluster: prepareClusterTool,
    /** Verify Kubernetes deployment status and health */
    verifyDeploy: verifyDeployTool,
    // ===== Operations =====
    /** Operational utilities (ping, status checks) */
    ops: opsTool,
};
/**
 * Execute any tool directly with full control.
 *
 * Use this when you need to:
 * - Execute tools not exposed via simplified functions
 * - Pass custom options to the executor
 * - Handle tool objects dynamically
 *
 * @example
 * ```typescript
 * import { executeTool, tools } from 'containerization-assist-mcp/sdk';
 *
 * const result = await executeTool(
 *   tools.analyzeRepo,
 *   { repositoryPath: '.' },
 *   { signal: controller.signal }
 * );
 * ```
 */
export { executeTool } from './executor.js';
// ===== JSON SCHEMA EXPORTS (for VS Code extension) =====
export { jsonSchemas, 
// Individual schemas for selective imports
analyzeRepoJsonSchema, generateDockerfileJsonSchema, fixDockerfileJsonSchema, buildImageContextJsonSchema, scanImageJsonSchema, tagImageJsonSchema, pushImageJsonSchema, generateK8sManifestsJsonSchema, prepareClusterJsonSchema, verifyDeployJsonSchema, opsJsonSchema, } from './schemas.js';
// ===== TOOL METADATA EXPORTS (for VS Code extension) =====
export { toolMetadata, standardWorkflow, getToolsByCategory, 
// External dependency types and helpers
ExternalDeps, getRequiredDepsString, requiresDependency, 
// Individual metadata for selective imports
analyzeRepoMetadata, generateDockerfileMetadata, fixDockerfileMetadata, buildImageContextMetadata, scanImageMetadata, tagImageMetadata, pushImageMetadata, generateK8sManifestsMetadata, prepareClusterMetadata, verifyDeployMetadata, opsMetadata, } from './metadata.js';
// ===== RESULT FORMATTERS (for VS Code extension) =====
export { resultFormatters, 
// Individual formatters for selective imports
formatAnalyzeRepoResult, formatGenerateDockerfileResult, formatFixDockerfileResult, formatBuildImageResult, formatScanImageResult, formatTagImageResult, formatPushImageResult, formatGenerateK8sManifestsResult, formatPrepareClusterResult, formatVerifyDeployResult, formatOpsResult, } from './formatters.js';
// ===== VS CODE UTILITIES =====
export { createAbortSignalFromToken, formatErrorForLLM, resolveWorkspacePath, validateRequiredFields, sanitizeForMarkdown, } from './vscode-utils.js';
// ===== KNOWLEDGE BASE =====
/**
 * Load the embedded knowledge base and return all entries.
 *
 * Knowledge packs contain best practices for Dockerfile generation, Kubernetes
 * manifests, and security configurations. They are embedded at build time,
 * eliminating runtime filesystem dependencies.
 *
 * @returns Object containing all loaded knowledge entries
 *
 * @example
 * ```typescript
 * import { loadKnowledgeData } from 'containerization-assist-mcp/sdk';
 *
 * const { entries } = loadKnowledgeData();
 * console.log(`Loaded ${entries.length} knowledge entries`);
 * ```
 */
export { loadKnowledgeData } from '../knowledge/index.js';
