// @azure/arm-machinelearning/dist-esm/src/models/parameters.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
import { Workspace as WorkspaceMapper, WorkspaceUpdateParameters as WorkspaceUpdateParametersMapper, DiagnoseWorkspaceParameters as DiagnoseWorkspaceParametersMapper, QuotaUpdateParameters as QuotaUpdateParametersMapper, ComputeResource as ComputeResourceMapper, ClusterUpdateParameters as ClusterUpdateParametersMapper, PrivateEndpointConnection as PrivateEndpointConnectionMapper, WorkspaceConnectionPropertiesV2BasicResource as WorkspaceConnectionPropertiesV2BasicResourceMapper, OutboundRuleBasicResource as OutboundRuleBasicResourceMapper, ManagedNetworkProvisionOptions as ManagedNetworkProvisionOptionsMapper, CodeContainer as CodeContainerMapper, CodeVersion as CodeVersionMapper, PendingUploadRequestDto as PendingUploadRequestDtoMapper, ComponentContainer as ComponentContainerMapper, ComponentVersion as ComponentVersionMapper, DataContainer as DataContainerMapper, DataVersionBase as DataVersionBaseMapper, GetBlobReferenceSASRequestDto as GetBlobReferenceSASRequestDtoMapper, EnvironmentContainer as EnvironmentContainerMapper, EnvironmentVersion as EnvironmentVersionMapper, ModelContainer as ModelContainerMapper, ModelVersion as ModelVersionMapper, PartialMinimalTrackedResourceWithIdentity as PartialMinimalTrackedResourceWithIdentityMapper, BatchEndpoint as BatchEndpointMapper, PartialBatchDeploymentPartialMinimalTrackedResourceWithProperties as PartialBatchDeploymentPartialMinimalTrackedResourceWithPropertiesMapper, BatchDeployment as BatchDeploymentMapper, DestinationAsset as DestinationAssetMapper, Datastore as DatastoreMapper, FeaturesetContainer as FeaturesetContainerMapper, FeaturesetVersion as FeaturesetVersionMapper, FeaturesetVersionBackfillRequest as FeaturesetVersionBackfillRequestMapper, FeaturestoreEntityContainer as FeaturestoreEntityContainerMapper, FeaturestoreEntityVersion as FeaturestoreEntityVersionMapper, JobBase as JobBaseMapper, MarketplaceSubscription as MarketplaceSubscriptionMapper, OnlineEndpoint as OnlineEndpointMapper, RegenerateEndpointKeysRequest as RegenerateEndpointKeysRequestMapper, PartialMinimalTrackedResourceWithSku as PartialMinimalTrackedResourceWithSkuMapper, OnlineDeployment as OnlineDeploymentMapper, DeploymentLogsRequest as DeploymentLogsRequestMapper, Schedule as ScheduleMapper, PartialMinimalTrackedResourceWithSkuAndIdentity as PartialMinimalTrackedResourceWithSkuAndIdentityMapper, ServerlessEndpoint as ServerlessEndpointMapper, PartialRegistryPartialTrackedResource as PartialRegistryPartialTrackedResourceMapper, Registry as RegistryMapper, } from "../models/mappers";
export const accept = {
    parameterPath: "accept",
    mapper: {
        defaultValue: "application/json",
        isConstant: true,
        serializedName: "Accept",
        type: {
            name: "String",
        },
    },
};
export const $host = {
    parameterPath: "$host",
    mapper: {
        serializedName: "$host",
        required: true,
        type: {
            name: "String",
        },
    },
    skipEncoding: true,
};
export const apiVersion = {
    parameterPath: "apiVersion",
    mapper: {
        defaultValue: "2024-04-01",
        isConstant: true,
        serializedName: "api-version",
        type: {
            name: "String",
        },
    },
};
export const subscriptionId = {
    parameterPath: "subscriptionId",
    mapper: {
        constraints: {
            MinLength: 1,
        },
        serializedName: "subscriptionId",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const resourceGroupName = {
    parameterPath: "resourceGroupName",
    mapper: {
        constraints: {
            MaxLength: 90,
            MinLength: 1,
        },
        serializedName: "resourceGroupName",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const workspaceName = {
    parameterPath: "workspaceName",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[a-zA-Z0-9][a-zA-Z0-9_-]{2,32}$"),
        },
        serializedName: "workspaceName",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const contentType = {
    parameterPath: ["options", "contentType"],
    mapper: {
        defaultValue: "application/json",
        isConstant: true,
        serializedName: "Content-Type",
        type: {
            name: "String",
        },
    },
};
export const parameters = {
    parameterPath: "parameters",
    mapper: WorkspaceMapper,
};
export const forceToPurge = {
    parameterPath: ["options", "forceToPurge"],
    mapper: {
        defaultValue: false,
        serializedName: "forceToPurge",
        type: {
            name: "Boolean",
        },
    },
};
export const parameters1 = {
    parameterPath: "parameters",
    mapper: WorkspaceUpdateParametersMapper,
};
export const skip = {
    parameterPath: ["options", "skip"],
    mapper: {
        serializedName: "$skip",
        type: {
            name: "String",
        },
    },
};
export const parameters2 = {
    parameterPath: ["options", "parameters"],
    mapper: DiagnoseWorkspaceParametersMapper,
};
export const nextLink = {
    parameterPath: "nextLink",
    mapper: {
        serializedName: "nextLink",
        required: true,
        type: {
            name: "String",
        },
    },
    skipEncoding: true,
};
export const location = {
    parameterPath: "location",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[-\\w\\._]+$"),
        },
        serializedName: "location",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const parameters3 = {
    parameterPath: "parameters",
    mapper: QuotaUpdateParametersMapper,
};
export const computeName = {
    parameterPath: "computeName",
    mapper: {
        serializedName: "computeName",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const parameters4 = {
    parameterPath: "parameters",
    mapper: ComputeResourceMapper,
};
export const parameters5 = {
    parameterPath: "parameters",
    mapper: ClusterUpdateParametersMapper,
};
export const underlyingResourceAction = {
    parameterPath: "underlyingResourceAction",
    mapper: {
        serializedName: "underlyingResourceAction",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const privateEndpointConnectionName = {
    parameterPath: "privateEndpointConnectionName",
    mapper: {
        serializedName: "privateEndpointConnectionName",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const properties = {
    parameterPath: "properties",
    mapper: PrivateEndpointConnectionMapper,
};
export const parameters6 = {
    parameterPath: "parameters",
    mapper: WorkspaceConnectionPropertiesV2BasicResourceMapper,
};
export const connectionName = {
    parameterPath: "connectionName",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[a-zA-Z0-9][a-zA-Z0-9_-]{2,32}$"),
        },
        serializedName: "connectionName",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const target = {
    parameterPath: ["options", "target"],
    mapper: {
        serializedName: "target",
        type: {
            name: "String",
        },
    },
};
export const category = {
    parameterPath: ["options", "category"],
    mapper: {
        serializedName: "category",
        type: {
            name: "String",
        },
    },
};
export const ruleName = {
    parameterPath: "ruleName",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[a-zA-Z0-9][a-zA-Z0-9_-]{2,32}$"),
        },
        serializedName: "ruleName",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const body = {
    parameterPath: "body",
    mapper: OutboundRuleBasicResourceMapper,
};
export const body1 = {
    parameterPath: ["options", "body"],
    mapper: ManagedNetworkProvisionOptionsMapper,
};
export const registryName = {
    parameterPath: "registryName",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[a-zA-Z0-9][a-zA-Z0-9\\-_]{2,32}$"),
        },
        serializedName: "registryName",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const codeName = {
    parameterPath: "codeName",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[a-zA-Z0-9][a-zA-Z0-9\\-_]{0,254}$"),
        },
        serializedName: "codeName",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const body2 = {
    parameterPath: "body",
    mapper: CodeContainerMapper,
};
export const orderBy = {
    parameterPath: ["options", "orderBy"],
    mapper: {
        serializedName: "$orderBy",
        type: {
            name: "String",
        },
    },
};
export const top = {
    parameterPath: ["options", "top"],
    mapper: {
        serializedName: "$top",
        type: {
            name: "Number",
        },
    },
};
export const version = {
    parameterPath: "version",
    mapper: {
        serializedName: "version",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const body3 = {
    parameterPath: "body",
    mapper: CodeVersionMapper,
};
export const body4 = {
    parameterPath: "body",
    mapper: PendingUploadRequestDtoMapper,
};
export const componentName = {
    parameterPath: "componentName",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[a-zA-Z0-9][a-zA-Z0-9\\-_]{0,254}$"),
        },
        serializedName: "componentName",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const body5 = {
    parameterPath: "body",
    mapper: ComponentContainerMapper,
};
export const body6 = {
    parameterPath: "body",
    mapper: ComponentVersionMapper,
};
export const listViewType = {
    parameterPath: ["options", "listViewType"],
    mapper: {
        serializedName: "listViewType",
        type: {
            name: "String",
        },
    },
};
export const name = {
    parameterPath: "name",
    mapper: {
        serializedName: "name",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const body7 = {
    parameterPath: "body",
    mapper: DataContainerMapper,
};
export const name1 = {
    parameterPath: "name",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[a-zA-Z0-9][a-zA-Z0-9\\-_]{0,254}$"),
        },
        serializedName: "name",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const tags = {
    parameterPath: ["options", "tags"],
    mapper: {
        serializedName: "$tags",
        type: {
            name: "String",
        },
    },
};
export const body8 = {
    parameterPath: "body",
    mapper: DataVersionBaseMapper,
};
export const body9 = {
    parameterPath: "body",
    mapper: GetBlobReferenceSASRequestDtoMapper,
};
export const environmentName = {
    parameterPath: "environmentName",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[a-zA-Z0-9][a-zA-Z0-9\\-_]{0,254}$"),
        },
        serializedName: "environmentName",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const body10 = {
    parameterPath: "body",
    mapper: EnvironmentContainerMapper,
};
export const body11 = {
    parameterPath: "body",
    mapper: EnvironmentVersionMapper,
};
export const modelName = {
    parameterPath: "modelName",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[a-zA-Z0-9][a-zA-Z0-9\\-_]{0,254}$"),
        },
        serializedName: "modelName",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const body12 = {
    parameterPath: "body",
    mapper: ModelContainerMapper,
};
export const version1 = {
    parameterPath: ["options", "version"],
    mapper: {
        serializedName: "version",
        type: {
            name: "String",
        },
    },
};
export const description = {
    parameterPath: ["options", "description"],
    mapper: {
        serializedName: "description",
        type: {
            name: "String",
        },
    },
};
export const tags1 = {
    parameterPath: ["options", "tags"],
    mapper: {
        serializedName: "tags",
        type: {
            name: "String",
        },
    },
};
export const properties1 = {
    parameterPath: ["options", "properties"],
    mapper: {
        serializedName: "properties",
        type: {
            name: "String",
        },
    },
};
export const body13 = {
    parameterPath: "body",
    mapper: ModelVersionMapper,
};
export const count = {
    parameterPath: ["options", "count"],
    mapper: {
        serializedName: "count",
        type: {
            name: "Number",
        },
    },
};
export const endpointName = {
    parameterPath: "endpointName",
    mapper: {
        serializedName: "endpointName",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const body14 = {
    parameterPath: "body",
    mapper: PartialMinimalTrackedResourceWithIdentityMapper,
};
export const endpointName1 = {
    parameterPath: "endpointName",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[a-zA-Z0-9][a-zA-Z0-9\\-_]{0,254}$"),
        },
        serializedName: "endpointName",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const body15 = {
    parameterPath: "body",
    mapper: BatchEndpointMapper,
};
export const deploymentName = {
    parameterPath: "deploymentName",
    mapper: {
        serializedName: "deploymentName",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const body16 = {
    parameterPath: "body",
    mapper: PartialBatchDeploymentPartialMinimalTrackedResourceWithPropertiesMapper,
};
export const deploymentName1 = {
    parameterPath: "deploymentName",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[a-zA-Z0-9][a-zA-Z0-9\\-_]{0,254}$"),
        },
        serializedName: "deploymentName",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const body17 = {
    parameterPath: "body",
    mapper: BatchDeploymentMapper,
};
export const hash = {
    parameterPath: ["options", "hash"],
    mapper: {
        serializedName: "hash",
        type: {
            name: "String",
        },
    },
};
export const hashVersion = {
    parameterPath: ["options", "hashVersion"],
    mapper: {
        serializedName: "hashVersion",
        type: {
            name: "String",
        },
    },
};
export const body18 = {
    parameterPath: "body",
    mapper: DestinationAssetMapper,
};
export const count1 = {
    parameterPath: ["options", "count"],
    mapper: {
        defaultValue: 30,
        serializedName: "count",
        type: {
            name: "Number",
        },
    },
};
export const isDefault = {
    parameterPath: ["options", "isDefault"],
    mapper: {
        serializedName: "isDefault",
        type: {
            name: "Boolean",
        },
    },
};
export const names = {
    parameterPath: ["options", "names"],
    mapper: {
        serializedName: "names",
        type: {
            name: "Sequence",
            element: {
                type: {
                    name: "String",
                },
            },
        },
    },
    collectionFormat: "CSV",
};
export const searchText = {
    parameterPath: ["options", "searchText"],
    mapper: {
        serializedName: "searchText",
        type: {
            name: "String",
        },
    },
};
export const orderBy1 = {
    parameterPath: ["options", "orderBy"],
    mapper: {
        serializedName: "orderBy",
        type: {
            name: "String",
        },
    },
};
export const orderByAsc = {
    parameterPath: ["options", "orderByAsc"],
    mapper: {
        defaultValue: false,
        serializedName: "orderByAsc",
        type: {
            name: "Boolean",
        },
    },
};
export const body19 = {
    parameterPath: "body",
    mapper: DatastoreMapper,
};
export const skipValidation = {
    parameterPath: ["options", "skipValidation"],
    mapper: {
        defaultValue: false,
        serializedName: "skipValidation",
        type: {
            name: "Boolean",
        },
    },
};
export const pageSize = {
    parameterPath: ["options", "pageSize"],
    mapper: {
        defaultValue: 20,
        serializedName: "pageSize",
        type: {
            name: "Number",
        },
    },
};
export const name2 = {
    parameterPath: ["options", "name"],
    mapper: {
        serializedName: "name",
        type: {
            name: "String",
        },
    },
};
export const createdBy = {
    parameterPath: ["options", "createdBy"],
    mapper: {
        serializedName: "createdBy",
        type: {
            name: "String",
        },
    },
};
export const body20 = {
    parameterPath: "body",
    mapper: FeaturesetContainerMapper,
};
export const featuresetName = {
    parameterPath: "featuresetName",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[a-zA-Z0-9][a-zA-Z0-9\\-_]{0,254}$"),
        },
        serializedName: "featuresetName",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const featuresetVersion = {
    parameterPath: "featuresetVersion",
    mapper: {
        serializedName: "featuresetVersion",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const featureName = {
    parameterPath: ["options", "featureName"],
    mapper: {
        serializedName: "featureName",
        type: {
            name: "String",
        },
    },
};
export const pageSize1 = {
    parameterPath: ["options", "pageSize"],
    mapper: {
        defaultValue: 1000,
        serializedName: "pageSize",
        type: {
            name: "Number",
        },
    },
};
export const featureName1 = {
    parameterPath: "featureName",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[a-zA-Z0-9][a-zA-Z0-9\\-_]{0,254}$"),
        },
        serializedName: "featureName",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const versionName = {
    parameterPath: ["options", "versionName"],
    mapper: {
        serializedName: "versionName",
        type: {
            name: "String",
        },
    },
};
export const stage = {
    parameterPath: ["options", "stage"],
    mapper: {
        serializedName: "stage",
        type: {
            name: "String",
        },
    },
};
export const body21 = {
    parameterPath: "body",
    mapper: FeaturesetVersionMapper,
};
export const body22 = {
    parameterPath: "body",
    mapper: FeaturesetVersionBackfillRequestMapper,
};
export const body23 = {
    parameterPath: "body",
    mapper: FeaturestoreEntityContainerMapper,
};
export const body24 = {
    parameterPath: "body",
    mapper: FeaturestoreEntityVersionMapper,
};
export const jobType = {
    parameterPath: ["options", "jobType"],
    mapper: {
        serializedName: "jobType",
        type: {
            name: "String",
        },
    },
};
export const tag = {
    parameterPath: ["options", "tag"],
    mapper: {
        serializedName: "tag",
        type: {
            name: "String",
        },
    },
};
export const id = {
    parameterPath: "id",
    mapper: {
        serializedName: "id",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const body25 = {
    parameterPath: "body",
    mapper: JobBaseMapper,
};
export const id1 = {
    parameterPath: "id",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[a-zA-Z0-9][a-zA-Z0-9\\-_]{0,254}$"),
        },
        serializedName: "id",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const body26 = {
    parameterPath: "body",
    mapper: MarketplaceSubscriptionMapper,
};
export const offset = {
    parameterPath: ["options", "offset"],
    mapper: {
        serializedName: "offset",
        type: {
            name: "Number",
        },
    },
};
export const feed = {
    parameterPath: ["options", "feed"],
    mapper: {
        serializedName: "feed",
        type: {
            name: "String",
        },
    },
};
export const computeType = {
    parameterPath: ["options", "computeType"],
    mapper: {
        serializedName: "computeType",
        type: {
            name: "String",
        },
    },
};
export const orderBy2 = {
    parameterPath: ["options", "orderBy"],
    mapper: {
        serializedName: "orderBy",
        type: {
            name: "String",
        },
    },
};
export const body27 = {
    parameterPath: "body",
    mapper: OnlineEndpointMapper,
};
export const body28 = {
    parameterPath: "body",
    mapper: RegenerateEndpointKeysRequestMapper,
};
export const body29 = {
    parameterPath: "body",
    mapper: PartialMinimalTrackedResourceWithSkuMapper,
};
export const body30 = {
    parameterPath: "body",
    mapper: OnlineDeploymentMapper,
};
export const body31 = {
    parameterPath: "body",
    mapper: DeploymentLogsRequestMapper,
};
export const listViewType1 = {
    parameterPath: ["options", "listViewType"],
    mapper: {
        serializedName: "listViewType",
        type: {
            name: "String",
        },
    },
};
export const body32 = {
    parameterPath: "body",
    mapper: ScheduleMapper,
};
export const body33 = {
    parameterPath: "body",
    mapper: PartialMinimalTrackedResourceWithSkuAndIdentityMapper,
};
export const body34 = {
    parameterPath: "body",
    mapper: ServerlessEndpointMapper,
};
export const name3 = {
    parameterPath: "name",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[a-zA-Z][a-zA-Z0-9-]{0,51}$"),
        },
        serializedName: "name",
        required: true,
        type: {
            name: "String",
        },
    },
};
export const body35 = {
    parameterPath: "body",
    mapper: PartialRegistryPartialTrackedResourceMapper,
};
export const body36 = {
    parameterPath: "body",
    mapper: RegistryMapper,
};
