// codetocloudmcp/out/mcpServers/toolsImplementations/azureRegionCheck/azureRegionChecker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAllRegions = getAllRegions;
exports.getAvailableRegionsByResourceTypes = getAvailableRegionsByResourceTypes;
exports.getAvailableRegionsForResourceTypes = getAvailableRegionsForResourceTypes;
// Abstract base class for checking Azure regions
class AzureRegionChecker {
    azureClient;
    constructor(azureClient) {
        this.azureClient = azureClient;
    }
    async getAzureManagementResultByUrl(requestAPI) {
        try {
            const requestUrl = `https://management.azure.com${requestAPI}`;
            return await this.azureClient.queryManagementPlane(requestUrl);
        }
        catch (error) {
            console.error(`Error fetching region status directly: ${error}`);
            return null;
        }
    }
}
// Default Region Checker Implementation
class DefaultRegionChecker extends AzureRegionChecker {
    constructor(azureClient) {
        super(azureClient);
    }
    async getAvailableRegions(resourceType, assignedRegions) {
        try {
            const providerNamespace = resourceType.split('/')[0];
            const resourceTypeName = resourceType.split('/')[1];
            const providerResourceTypes = await this.azureClient.getProviderResourceTypesByNamespace(providerNamespace);
            const resourceTypeInfo = providerResourceTypes.find(rt => rt.resourceType === resourceTypeName);
            if (!resourceTypeInfo || !resourceTypeInfo.locations) {
                return [];
            }
            return resourceTypeInfo.locations
                .map(location => {
                const normalizedLocation = location.replace(/\s+/g, '').toLowerCase();
                if (assignedRegions && assignedRegions.length > 0) {
                    if (assignedRegions.includes(normalizedLocation)) {
                        return normalizedLocation;
                    }
                    else {
                        return null;
                    }
                }
                else {
                    return normalizedLocation;
                }
            })
                .filter((location) => location !== null);
        }
        catch (error) {
            console.error(`Error fetching regions for resource type ${resourceType}:`, error);
            return [];
        }
    }
}
// Cognitive Services Region Checker Implementation
class CognitiveServicesRegionChecker extends AzureRegionChecker {
    skuName;
    apiVersion;
    modelName;
    constructor(azureClient, skuName, apiVersion, modelName) {
        super(azureClient);
        this.skuName = skuName;
        this.apiVersion = apiVersion;
        this.modelName = modelName;
    }
    async getAvailableRegions(resourceType, assignedRegions) {
        const providerNamespace = resourceType.split('/')[0];
        const resourceTypeName = resourceType.split('/')[1];
        const providerResourceTypes = await this.azureClient.getProviderResourceTypesByNamespace(providerNamespace);
        let regions = (providerResourceTypes.find(rt => rt.resourceType === resourceTypeName)?.locations || []).map(location => location.replace(/\s+/g, '').toLowerCase());
        if (assignedRegions && assignedRegions.length > 0) {
            regions = assignedRegions.filter(region => regions.includes(region));
        }
        // Process regions in parallel for better performance
        const regionCheckPromises = regions.map(async (region) => {
            try {
                const quotas = await this.azureClient.listCognitiveServicesModels(region);
                let hasMatchingModel = false;
                for await (const item of quotas) {
                    const nameMatch = !this.modelName || item.name === this.modelName;
                    const versionMatch = !this.apiVersion || item.version === this.apiVersion;
                    const skuMatch = !this.skuName || (item.skuNames && item.skuNames.some((sku) => sku === this.skuName));
                    if (nameMatch && versionMatch && skuMatch) {
                        hasMatchingModel = true;
                        break;
                    }
                }
                return hasMatchingModel ? region : null;
            }
            catch (error) {
                console.error(`Error checking cognitive services models for region ${region}:`, error);
                return null;
            }
        });
        // Wait for all region checks to complete and filter out null results
        const results = await Promise.all(regionCheckPromises);
        return results.filter((region) => region !== null);
    }
}
class SQLChecker extends AzureRegionChecker {
    constructor(azureClient) {
        super(azureClient);
    }
    async getAvailableRegions(resourceType, assignedRegions) {
        const providerNamespace = resourceType.split('/')[0];
        const resourceTypeName = resourceType.split('/')[1];
        const providerResourceTypes = await this.azureClient.getProviderResourceTypesByNamespace(providerNamespace);
        let regions = (providerResourceTypes.find(rt => rt.resourceType === resourceTypeName)?.locations || []).map(location => location.replace(/\s+/g, '').toLowerCase());
        if (assignedRegions && assignedRegions.length > 0) {
            regions = assignedRegions.filter(region => regions.includes(region));
        }
        const apiVersion = '2023-05-01-preview';
        const regionCheckPromises = regions.map(async (region) => {
            console.log(`Checking for region ${region}`);
            try {
                const url = `/subscriptions/${this.azureClient.getCurrentSubscriptionId()}/providers/Microsoft.Sql/locations/${region}/capabilities/?api-version=${apiVersion}`;
                const rawResponse = await this.getAzureManagementResultByUrl(url);
                if (!rawResponse || rawResponse.reason) {
                    console.log(`No SQL capabilities found for region ${region}, ${rawResponse?.reason}`);
                    return null;
                }
                return region;
            }
            catch (error) {
                console.error(`Error checking SQL capabilities for region ${region}:`, error);
                return null;
            }
        });
        // Wait for all region checks to complete and filter out null results
        const results = await Promise.all(regionCheckPromises);
        return results.filter((region) => region !== null);
    }
}
class PostgresSQLRegionChecker extends AzureRegionChecker {
    // TODO: could't find a SDK for checking PostgreSQL regions in typescript, using raw API calls temporarily
    constructor(azureClient) {
        super(azureClient);
    }
    async getAvailableRegions(resourceType, assignedRegions) {
        const providerNamespace = resourceType.split('/')[0];
        const resourceTypeName = resourceType.split('/')[1];
        const providerResourceTypes = await this.azureClient.getProviderResourceTypesByNamespace(providerNamespace);
        let regions = (providerResourceTypes.find(rt => rt.resourceType === resourceTypeName)?.locations || []).map(location => location.replace(/\s+/g, '').toLowerCase());
        if (assignedRegions && assignedRegions.length > 0) {
            regions = assignedRegions.filter(region => regions.includes(region));
        }
        const apiVersion = '2024-11-01-preview';
        // Process regions in parallel for better performance
        const regionCheckPromises = regions.map(async (region) => {
            try {
                const url = `/subscriptions/${this.azureClient.getCurrentSubscriptionId()}/providers/Microsoft.DBforPostgreSQL/locations/${region}/capabilities?api-version=${apiVersion}`;
                const rawResponse = await this.getAzureManagementResultByUrl(url);
                if (!rawResponse || !rawResponse.value || rawResponse.value[0].reason) {
                    return null;
                }
                return region;
            }
            catch (error) {
                console.error(`Error checking PostgreSQL capabilities for region ${region}:`, error);
                return null;
            }
        });
        // Wait for all region checks to complete and filter out null results
        const results = await Promise.all(regionCheckPromises);
        return results.filter((region) => region !== null);
    }
}
class MysqlRegionChecker extends AzureRegionChecker {
    constructor(azureClient) {
        super(azureClient);
    }
    async getAvailableRegions(resourceType, assignedRegions) {
        const providerNamespace = resourceType.split('/')[0];
        const resourceTypeName = resourceType.split('/')[1];
        const providerResourceTypes = await this.azureClient.getProviderResourceTypesByNamespace(providerNamespace);
        let regions = (providerResourceTypes.find(rt => rt.resourceType === resourceTypeName)?.locations || []).map(location => location.replace(/\s+/g, '').toLowerCase());
        if (assignedRegions && assignedRegions.length > 0) {
            regions = assignedRegions.filter(region => regions.includes(region));
        }
        const apiVersion = '2023-06-30';
        const regionCheckPromises = regions.map(async (region) => {
            try {
                const url = `/subscriptions/${this.azureClient.getCurrentSubscriptionId()}/providers/Microsoft.DBforMySQL/locations/${region}/capabilities?api-version=${apiVersion}`;
                const rawResponse = await this.getAzureManagementResultByUrl(url);
                if (!rawResponse || !rawResponse.value || rawResponse.value[0].reason) {
                    return null;
                }
                return region;
            }
            catch (error) {
                console.error(`Error checking Mysql capabilities for region ${region}:`, error);
                return null;
            }
        });
        const results = await Promise.all(regionCheckPromises);
        return results.filter((region) => region !== null);
    }
}
class DocumentDBRegionChecker extends AzureRegionChecker {
    constructor(azureClient) {
        super(azureClient);
    }
    async getAvailableRegions(resourceType, assignedRegions) {
        const apiVersion = '2023-04-15';
        const url = `/subscriptions/${this.azureClient.getCurrentSubscriptionId()}/providers/Microsoft.DocumentDB/locations?api-version=${apiVersion}`;
        const rawResponse = await this.getAzureManagementResultByUrl(url);
        return rawResponse?.value?.filter((location) => location.properties?.isSubscriptionRegionAccessAllowedForRegular === true &&
            (!assignedRegions || assignedRegions.includes(location.name.toLowerCase().replace(' ', '')))).map((location) => location.name.toLowerCase().replace(' ', '')) || [];
    }
}
class ComputeChecker extends AzureRegionChecker {
    constructor(azureClient) {
        super(azureClient);
    }
    async getAvailableRegions(resourceType, assignedRegions) {
        try {
            const capabilities = await this.azureClient.getComputeSkus();
            const availableRegions = new Set();
            // Filter SKUs that match the resource type and have no restrictions
            for (const capability of capabilities) {
                capability.locations.forEach(location => {
                    const normalizedLocation = location.replace(/\s+/g, '').toLowerCase();
                    if (!assignedRegions || assignedRegions.includes(normalizedLocation)) {
                        availableRegions.add(normalizedLocation);
                    }
                });
            }
            // Return distinct regions as an array
            return Array.from(availableRegions).sort();
        }
        catch (error) {
            console.error(`Error fetching compute regions for resource type ${resourceType}:`, error);
            return [];
        }
    }
}
class AppServiceManagedInstanceRegionChecker extends AzureRegionChecker {
    constructor(azureClient) {
        super(azureClient);
    }
    async getAvailableRegions(resourceType, assignedRegions) {
        // Limited support on East Asia, West Central US, North Europe, East US
        return ['eastus', 'eastasia', 'westcentralus', 'northeurope'];
    }
}
// Factory function to create region checkers
function createRegionChecker(azureClient, resourceType, properties) {
    const provider = resourceType.split('/')[0].toLowerCase();
    switch (provider) {
        case "microsoft.cognitiveservices":
            return new CognitiveServicesRegionChecker(azureClient, properties?.deploymentSkuName, properties?.modelVersion, properties?.modelName);
        case "microsoft.dbforpostgresql":
            return new PostgresSQLRegionChecker(azureClient);
        case "microsoft.sql":
            return new SQLChecker(azureClient);
        case "microsoft.dbformysql":
            return new MysqlRegionChecker(azureClient);
        case "microsoft.documentdb":
            return new DocumentDBRegionChecker(azureClient);
        case "microsoft.compute":
            return new ComputeChecker(azureClient);
        case "appservicemanagedinstance":
            return new AppServiceManagedInstanceRegionChecker(azureClient);
        default:
            return new DefaultRegionChecker(azureClient);
    }
}
// Main function to get available regions for resource types
async function getAvailableRegionsForResourceTypes(resourceTypes, azureClient, cognitiveServiceProperties, assignedRegions) {
    const regionCheckPromises = resourceTypes.map(async (resourceType) => {
        try {
            const checker = createRegionChecker(azureClient, resourceType, cognitiveServiceProperties);
            const regions = await checker.getAvailableRegions(resourceType, assignedRegions);
            return regions;
        }
        catch (error) {
            return [];
        }
    });
    const availableRegions = await Promise.all(regionCheckPromises);
    const allRegions = availableRegions.filter(regions => regions.length > 0);
    // List regions that are available for all resource types
    let validRegions = allRegions[0];
    for (const availableRegionSet of allRegions) {
        validRegions = validRegions.filter(x => availableRegionSet.includes(x));
    }
    return validRegions;
}
// Enhanced function to get both common regions and per-resource-type breakdown
async function getAvailableRegionsByResourceTypes(resourceTypes, azureClient, cognitiveServiceProperties, assignedRegions) {
    const regionCheckPromises = resourceTypes.map(async (resourceType) => {
        try {
            const checker = createRegionChecker(azureClient, resourceType, cognitiveServiceProperties);
            const regions = await checker.getAvailableRegions(resourceType, assignedRegions);
            return { resourceType, regions };
        }
        catch (error) {
            return { resourceType, regions: [] };
        }
    });
    const resourceRegionResults = await Promise.all(regionCheckPromises);
    // Create the per-resource-type breakdown
    const regionsByResourceType = {};
    resourceRegionResults.forEach(result => {
        regionsByResourceType[result.resourceType] = result.regions;
    });
    // Calculate common regions (intersection of all resource types)
    const allRegions = resourceRegionResults.map(result => result.regions).filter(regions => regions.length > 0);
    let commonRegions = allRegions[0] || [];
    for (const availableRegionSet of allRegions) {
        commonRegions = commonRegions.filter(x => availableRegionSet.includes(x));
    }
    return { commonRegions, regionsByResourceType };
}
async function getAllRegions(azureClient) {
    const requestUrl = `https://management.azure.com/subscriptions/${azureClient.getCurrentSubscriptionId()}/locations?api-version=2022-12-01`;
    const data = await azureClient.queryManagementPlane(requestUrl);
    if (data && data.value) {
        return data.value.map((location) => location.name);
    }
    return [];
}
