// containerization-assist-mcp/dist-cjs/src/tools/tag-image/tool.js
"use strict";
/**
 * Tag Image Tool - Modernized Implementation
 *
 * Tags Docker images with version and registry information
 * Follows the new Tool interface pattern
 *
 * This is a deterministic operational tool with no AI calls.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tool_context_helpers_1 = require("../../lib/tool-context-helpers");
const errors_1 = require("../../lib/errors");
const client_1 = require("../../infra/docker/client");
const validation_helpers_1 = require("../../lib/validation-helpers");
const types_1 = require("../../types");
const tool_1 = require("../../types/tool");
const schema_1 = require("./schema");
const summary_helpers_1 = require("../../lib/summary-helpers");
/**
 * Tag image handler
 */
async function handleTagImage(input, ctx) {
    const { logger, timer } = (0, tool_context_helpers_1.setupToolContext)(ctx, 'tag-image');
    const { tag } = input;
    if (!tag) {
        return (0, types_1.Failure)('Tag parameter is required', {
            message: 'Missing required parameter: tag',
            hint: 'Tag name must be specified for the image',
            resolution: 'Add tag parameter with the desired image tag (e.g., myapp:v1.0)',
        });
    }
    // Parse and validate image name
    const parsedImage = (0, validation_helpers_1.parseImageName)(tag);
    if (!parsedImage.ok) {
        return parsedImage;
    }
    try {
        const dockerClient = (0, client_1.createDockerClient)(logger);
        const source = input.imageId;
        if (!source) {
            return (0, types_1.Failure)('No image specified. Provide imageId parameter.', {
                message: 'Missing required parameter: imageId',
                hint: 'Source image ID or name must be specified to tag',
                resolution: 'Add imageId parameter with the Docker image ID or name to tag',
            });
        }
        // Extract repository and tag from parsed image
        const { repository, tag: tagName } = parsedImage.value;
        // For Docker tag operation, repository includes registry if present
        const fullRepository = parsedImage.value.registry
            ? `${parsedImage.value.registry}/${repository}`
            : repository;
        const tagResult = await dockerClient.tagImage(source, fullRepository, tagName);
        if (!tagResult.ok) {
            return (0, types_1.Failure)(`Failed to tag image: ${tagResult.error ?? 'Unknown error'}`, tagResult.guidance);
        }
        const tags = [tag];
        // Generate summary
        const tagsList = (0, summary_helpers_1.summarizeList)(tags);
        const summary = `✅ Tagged image. Applied ${tags.length === 1 ? 'tag' : `${tags.length} tags`}: ${tagsList}. Ready to push.`;
        const result = {
            summary,
            success: true,
            tags,
            imageId: source,
        };
        timer.end({ tags });
        return (0, types_1.Success)(result);
    }
    catch (error) {
        timer.error(error);
        return (0, types_1.Failure)((0, errors_1.extractErrorMessage)(error), {
            message: (0, errors_1.extractErrorMessage)(error),
            hint: 'An unexpected error occurred while tagging the image',
            resolution: 'Verify that Docker is running, the source image exists, and the tag format is valid',
        });
    }
}
/**
 * Tag image tool conforming to Tool interface
 */
exports.default = (0, tool_1.tool)({
    name: 'tag-image',
    description: 'Tag Docker images with version and registry information',
    category: 'docker',
    version: '2.0.0',
    schema: schema_1.tagImageSchema,
    metadata: {
        knowledgeEnhanced: false,
    },
    handler: handleTagImage,
});
