// containerization-assist-mcp/dist-cjs/src/tools/fix-dockerfile/schema.js
"use strict";
/**
 * Schema definition for fix-dockerfile tool
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.fixDockerfileSchema = void 0;
const zod_1 = require("zod");
const schemas_1 = require("../shared/schemas");
exports.fixDockerfileSchema = zod_1.z
    .object({
    dockerfile: zod_1.z.string().optional().describe('Dockerfile content to analyze for fixes'),
    path: zod_1.z.string().optional().describe('Path to Dockerfile file to analyze for fixes'),
    environment: schemas_1.environment.describe('Target environment (production, development, etc.)'),
    targetPlatform: schemas_1.platform.describe('Target platform for deployment (e.g., linux/amd64, linux/arm64). Used for context in recommendations.'),
    strictPlatformValidation: zod_1.z
        .boolean()
        .optional()
        .default(false)
        .describe('Deprecated: No longer enforces platform flags. Reserved for future use.'),
    policyPath: zod_1.z
        .string()
        .optional()
        .describe('Optional path to specific policy file to use for organizational policy validation (defaults to all policies in policies/)'),
})
    .refine((data) => data.dockerfile || data.path, {
    message: "Either 'dockerfile' content or 'path' must be provided",
});
