// codetocloudmcp/out/toolsImplementations/shared/azureClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureClient = void 0;
const arm_appcontainers_1 = require("@azure/arm-appcontainers");
const arm_appservice_1 = require("@azure/arm-appservice");
const arm_authorization_1 = require("@azure/arm-authorization");
const arm_cognitiveservices_1 = require("@azure/arm-cognitiveservices");
const arm_compute_1 = require("@azure/arm-compute");
const arm_containerregistry_1 = require("@azure/arm-containerregistry");
const arm_containerservice_1 = require("@azure/arm-containerservice");
const arm_keyvault_1 = require("@azure/arm-keyvault");
const arm_machinelearning_1 = require("@azure/arm-machinelearning");
const arm_network_1 = require("@azure/arm-network");
const arm_resources_1 = require("@azure/arm-resources");
const arm_storage_1 = require("@azure/arm-storage");
const monitor_query_1 = require("@azure/monitor-query");
const stringHelpers_1 = require("../../common/stringHelpers");
class AzureClient {
    subscriptionId;
    credential;
    computeClient;
    cognitiveClient;
    mlClient;
    networkClient;
    containerAppClient;
    webSiteClient;
    storageClient;
    logsQueryClient;
    containerServiceClient;
    resourceClient;
    containerRegistryClient;
    authorizationClient;
    keyVaultClient;
    constructor(subscriptionId, credential) {
        this.subscriptionId = subscriptionId;
        this.credential = credential;
        this.computeClient = new arm_compute_1.ComputeManagementClient(this.credential, subscriptionId);
        this.cognitiveClient = new arm_cognitiveservices_1.CognitiveServicesManagementClient(this.credential, subscriptionId);
        this.mlClient = new arm_machinelearning_1.AzureMachineLearningServicesManagementClient(this.credential, subscriptionId);
        this.networkClient = new arm_network_1.NetworkManagementClient(this.credential, subscriptionId);
        this.containerAppClient = new arm_appcontainers_1.ContainerAppsAPIClient(this.credential, subscriptionId);
        this.webSiteClient = new arm_appservice_1.WebSiteManagementClient(this.credential, subscriptionId, { apiVersion: '2025-03-01' });
        this.computeClient = new arm_compute_1.ComputeManagementClient(this.credential, subscriptionId);
        this.cognitiveClient = new arm_cognitiveservices_1.CognitiveServicesManagementClient(this.credential, subscriptionId);
        this.storageClient = new arm_storage_1.StorageManagementClient(this.credential, subscriptionId, { apiVersion: '2025-01-01' });
        this.logsQueryClient = new monitor_query_1.LogsQueryClient(this.credential);
        this.resourceClient = new arm_resources_1.ResourceManagementClient(this.credential, subscriptionId, { apiVersion: '2025-04-01' });
        this.containerServiceClient = new arm_containerservice_1.ContainerServiceClient(this.credential, subscriptionId);
        this.containerRegistryClient = new arm_containerregistry_1.ContainerRegistryManagementClient(this.credential, subscriptionId);
        this.authorizationClient = new arm_authorization_1.AuthorizationManagementClient(this.credential, subscriptionId);
        this.keyVaultClient = new arm_keyvault_1.KeyVaultManagementClient(this.credential, subscriptionId);
    }
    async validateSubscription() {
        if (!(0, stringHelpers_1.checkGUIDFormat)(this.subscriptionId)) {
            return false;
        }
        try {
            const query = await this.queryManagementPlane(`https://management.azure.com/subscriptions/${this.subscriptionId}?api-version=2022-12-01`);
            if (!query || !query.subscriptionId) {
                return false;
            }
            return true;
        }
        catch (error) {
            return false;
        }
    }
    async queryManagementPlane(uri) {
        try {
            const token = await this.credential.getToken("https://management.azure.com/.default");
            if (!token) {
                throw new Error(`Error! Your Azure credentials are invalid.`);
            }
            const response = await fetch(uri, {
                headers: {
                    'Authorization': `Bearer ${token.token}`,
                    'Content-Type': 'application/json'
                }
            });
            if (!response.ok) {
                throw new Error(`HTTP error! status: ${response.status}`);
            }
            return await response.json();
        }
        catch (error) {
            console.error(`Error fetching quotas for ${uri} directly: ${error}`);
            return null;
        }
    }
    getCurrentSubscriptionId() {
        return this.subscriptionId;
    }
    async getAMLUsageForLocation(location) {
        const usages = this.mlClient.usages.list(location);
        const amlUsages = [];
        for await (const usage of usages) {
            amlUsages.push({
                name: usage.name?.value || '',
                limit: usage.limit || 0,
                used: usage.currentValue || 0,
                unit: usage.unit || ''
            });
        }
        return amlUsages;
    }
    async getNetworkUsageForLocation(location) {
        const usages = this.networkClient.usages.list(location);
        const result = [];
        for await (const item of usages) {
            result.push({
                name: item.name?.value || '',
                limit: item.limit || 0,
                used: item.currentValue || 0,
                unit: item.unit
            });
        }
        return result;
    }
    async getAppServiceManagedInstanceUsageForLocation(location) {
        const quotas = this.webSiteClient.getUsagesInLocation.list(location);
        const result = [];
        // Managed Instance quota names are now limited to Pv4 and Pmv4, i.e., "P0v4", "P1mv4", etc.
        // "Total Regional VMs" limit 0: 
        // 1. if only total regional VMS is returned, usually indicates sufficient quota
        // 2. if returned with other quotas, usually indicates insufficient quota
        // Reference: https://techcommunity.microsoft.com/blog/appsonazureblog/announcing-the-public-preview-of-the-new-app-service-quota-self-service-experien/4450415
        for await (const item of quotas) {
            const quotaName = item.name?.value || '';
            if (quotaName === '*' || /^P[0-5]m?v4$/.test(quotaName)) {
                result.push({
                    name: quotaName,
                    limit: item.limit || 0,
                    used: item.currentValue || 0,
                    unit: item.unit
                });
            }
        }
        const totalRegionalVMQuota = result.find(q => q.name === '*');
        if (result.length === 1 && totalRegionalVMQuota?.limit === 0 && totalRegionalVMQuota?.used === 0) {
            return [{
                    name: 'Use any of the Pv4 or Pmv4 SKUs, total regional VMs quota 0 of 0 indicates sufficient quota. Only verifiable during real deployment.',
                    limit: 0,
                    used: 0,
                    unit: 'Instances'
                }];
        }
        else if (result.length > 1 && totalRegionalVMQuota && totalRegionalVMQuota.limit - totalRegionalVMQuota.used <= 0) {
            return [{
                    name: 'No Total Regional VMs quota left.',
                    limit: totalRegionalVMQuota.limit,
                    used: totalRegionalVMQuota.used,
                    unit: 'Instances'
                }];
        }
        return result.filter(q => q.name !== '*');
    }
    async getContainerAppUsageForLocation(location) {
        const quotas = this.containerAppClient.usages.list(location);
        const result = [];
        for await (const page of quotas.byPage()) {
            for (const item of page) {
                result.push({
                    name: item.name?.value || '',
                    limit: item.limit || 0,
                    used: item.currentValue || 0,
                    unit: item.unit
                });
            }
        }
        return result;
    }
    async getComputeUsageForLocation(location) {
        const usages = this.computeClient.usageOperations.list(location);
        const computeUsages = [];
        for await (const usage of usages) {
            computeUsages.push({
                name: usage.name?.value || '',
                limit: usage.limit || 0,
                used: usage.currentValue || 0,
                unit: usage.unit || ''
            });
        }
        return computeUsages;
    }
    async getComputeSkus(location) {
        const skus = location ? this.computeClient.resourceSkus.list({ filter: `location eq '${location}'` }) : this.computeClient.resourceSkus.list();
        const computeSkus = [];
        for await (const sku of skus) {
            if (sku.resourceType?.toLowerCase() === "virtualmachines") {
                // Filter out locations that are restricted for this SKU
                const availableLocations = (sku.locations || []).filter(loc => {
                    if (!sku.restrictions || sku.restrictions.length === 0) {
                        return true;
                    }
                    // Check if this location is restricted by any restriction
                    return !sku.restrictions.some(restriction => restriction.restrictionInfo?.locations?.some(restrictedLoc => restrictedLoc.toLowerCase() === loc.toLowerCase()));
                });
                if (availableLocations.length > 0) {
                    computeSkus.push({
                        family: sku.family,
                        name: sku.name,
                        capabilities: sku.capabilities || [],
                        locations: availableLocations
                    });
                }
            }
        }
        return computeSkus;
    }
    async getCognitiveServicesUsageForLocation(location) {
        const quotas = this.cognitiveClient.usages.list(location);
        const result = [];
        for await (const item of quotas) {
            result.push({
                name: item.name?.localizedValue || item.name?.value || '',
                limit: item.limit || 0,
                used: item.currentValue || 0,
                unit: item.unit
            });
        }
        return result;
    }
    async listCognitiveServicesModels(location) {
        const quotas = this.cognitiveClient.models.list(location);
        const result = [];
        for await (const item of quotas) {
            result.push({
                name: item.model?.name || '',
                version: item.model?.version || '',
                skuNames: item.model?.skus?.map(sku => sku.name || "").filter(name => name !== "") || []
            });
        }
        return result;
    }
    async getStorageUsageForLocation(location) {
        const usages = this.storageClient.usages.listByLocation(location);
        const result = [];
        for await (const item of usages) {
            result.push({
                name: item.name?.value || '',
                limit: item.limit || 0,
                used: item.currentValue || 0,
                unit: item.unit
            });
        }
        return result;
    }
    async getProviderResourceTypesByNamespace(namespace) {
        const provider = await this.resourceClient.providers.get(namespace);
        if (!provider.resourceTypes) {
            return [];
        }
        return provider.resourceTypes.map(rt => {
            return {
                resourceType: rt.resourceType ?? '',
                locations: rt.locations || [],
                capabilities: rt.capabilities || ''
            };
        });
    }
    async getResourceGroupsInSubscription() {
        const resourceGroups = this.resourceClient.resourceGroups.list();
        const result = [];
        for await (const rg of resourceGroups) {
            result.push({
                name: rg.name || '',
                tags: rg.tags || {}
            });
        }
        return result;
    }
    async getResourcesInResourceGroupByFilter(resourceGroupName, filter) {
        const result = [];
        for await (const resource of this.resourceClient.resources.listByResourceGroup(resourceGroupName, { filter })) {
            result.push({
                id: resource.id,
                name: resource.name,
                type: resource.type,
                kind: resource.kind,
                location: resource.location,
                tags: resource.tags
            });
        }
        return result;
    }
    async queryLogsFromWorkspace(workspaceId, query, startTime, endTime) {
        const result = await this.logsQueryClient.queryResource(workspaceId, query, { startTime, endTime });
        return {
            status: result.status,
            tables: 'tables' in result ? result.tables.map(table => ({ name: table.name, rows: table.rows })) : []
        };
    }
    async getContainerAppsLatestActiveRevision(resourceGroupName, containerAppName) {
        for await (const revision of this.containerAppClient.containerAppsRevisions.listRevisions(resourceGroupName, containerAppName)) {
            if (revision.active) {
                return {
                    active: revision.active,
                    createdTime: revision.createdTime
                };
            }
        }
        return undefined;
    }
    // This function works for both app service and function app
    async getWebSiteLatestDeployment(resourceGroupName, appName) {
        for await (const deployment of this.webSiteClient.webApps.listDeployments(resourceGroupName, appName)) {
            if (deployment.active) {
                return {
                    active: deployment.active,
                    startTime: deployment.startTime
                };
            }
        }
        return undefined;
    }
    async getAKSCluster(resourceGroupName, clusterName) {
        const aksCluster = await this.containerServiceClient.managedClusters.get(resourceGroupName, clusterName);
        return {
            kubeletIdentityObjectId: aksCluster.identityProfile?.kubeletidentity?.objectId
        };
    }
    async getContainerRegistryDetails(resourceGroupName, registryName) {
        const registry = await this.containerRegistryClient.registries.get(resourceGroupName, registryName);
        return {
            id: registry.id,
            name: registry.name,
            loginServer: registry.loginServer,
            adminUserEnabled: registry.adminUserEnabled
        };
    }
    async listRoleAssignmentsForScope(scope, filter) {
        let additionalArgs = undefined;
        if (filter) {
            additionalArgs = {
                filter: filter
            };
        }
        const roleAssignments = this.authorizationClient.roleAssignments.listForScope(scope, additionalArgs);
        const result = [];
        for await (const roleAssignment of roleAssignments) {
            result.push({
                roleDefinitionId: roleAssignment.roleDefinitionId
            });
        }
        return result;
    }
    async getContainerApp(resourceGroupName, containerAppName) {
        const containerApp = await this.containerAppClient.containerApps.get(resourceGroupName, containerAppName);
        const registryCredentials = [];
        for (let registry of (containerApp.configuration?.registries ?? [])) {
            registryCredentials.push({
                server: registry.server,
                username: registry.username,
                passwordSecretRef: registry.passwordSecretRef,
                identity: registry.identity
            });
        }
        const secrets = [];
        for (let secret of (containerApp.configuration?.secrets ?? [])) {
            secrets.push({
                name: secret.name,
                value: secret.value,
                identity: secret.identity,
                keyVaultUrl: secret.keyVaultUrl
            });
        }
        let identity = undefined;
        if (containerApp.identity) {
            const userAssignedIdentities = {};
            if (containerApp.identity.userAssignedIdentities) {
                for (const key of Object.keys(containerApp.identity.userAssignedIdentities)) {
                    const uai = containerApp.identity.userAssignedIdentities[key];
                    userAssignedIdentities[key] = {
                        principalId: uai.principalId,
                        clientId: uai.clientId
                    };
                }
            }
            identity = {
                principalId: containerApp.identity.principalId,
                tenantId: containerApp.identity.tenantId,
                type: containerApp.identity.type,
                userAssignedIdentities: userAssignedIdentities
            };
        }
        return {
            name: containerApp.name,
            registryCredentials: registryCredentials,
            identity: identity,
            secrets: secrets,
            outboundIpAddresses: containerApp.outboundIpAddresses
        };
    }
    async getKeyVault(resourceGroupName, vaultName) {
        return this.keyVaultClient.vaults.get(resourceGroupName, vaultName);
    }
}
exports.AzureClient = AzureClient;
