// codetocloudmcp/out/toolsImplementations/shared/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetWAFRulesSchema = exports.GetALZPlanSchema = exports.DiagnoseExistingResourcesSchema = exports.RunInTerminalSchema = exports.SummarizeResultSchema = exports.GetPlanSchema = exports.GetContainerizationPlanSchema = exports.GetAzureResourceRulesSchema = exports.GetAppLogsSchema = exports.GenerateArchitectureDiagramSchema = exports.DeployPipelineSchema = exports.DebugInBrowserSchema = exports.AzureGetRegionsForDeploymentSchema = exports.AzureRegionCheckSchema = exports.AzureQuotaCheckSchema = exports.AzurePricingSchema = void 0;
const zod_1 = require("zod");
const models_1 = require("../../common/getAppLogs/models");
const models_2 = require("../../common/models");
const project_1 = require("../../common/project");
const params_1 = require("../azureRegionCheck/params");
const params_2 = require("../debugInBrowser/params");
const params_3 = require("../getAzureResourceRules/params");
const params_4 = require("../runInTerminal/params");
const params_5 = require("../azurePricing/params");
exports.AzurePricingSchema = zod_1.z.object({
    queries: zod_1.z.array(zod_1.z.object({
        serviceName: zod_1.z.string().optional()
            .describe("The Azure service name to query pricing for. Case-sensitive. Common values: 'Virtual Machines', 'Azure App Service', 'Azure Cosmos DB', 'Azure Kubernetes Service', 'SQL Database', 'Storage', 'Redis Cache', 'Container Instances', 'Azure Container Apps', 'Azure Functions'. Note: must match exact API service name."),
        serviceFamily: zod_1.z.nativeEnum(params_5.AzureServiceFamily).optional()
            .describe("The Azure service family to query pricing for. Use this as an alternative or alongside serviceName for broader queries."),
        skuName: zod_1.z.string().optional()
            .describe("Optional SKU name to filter pricing. E.g. 'B1', 'S1', 'P1 v3' (note: spaces matter). This is the commercial SKU name shown in pricing pages."),
        armSkuName: zod_1.z.string().optional()
            .describe("Optional ARM SKU name for VM-style resources. E.g. 'Standard_D2s_v3', 'Standard_B1s'. Use this instead of skuName for Virtual Machines."),
        armRegionName: zod_1.z.string().optional()
            .describe("Optional Azure region. E.g. 'eastus', 'westus2', 'westeurope', 'southeastasia', etc."),
        priceType: zod_1.z.nativeEnum(params_5.AzurePriceType).optional()
            .describe("Filter by price type. Defaults to all types if not specified.")
    }))
        .describe("An array of pricing queries. Each query specifies filters for the Azure Retail Prices API. At least one of serviceName or serviceFamily should be provided."),
    currencyCode: zod_1.z.string().optional()
        .describe("The currency code for pricing. Defaults to 'USD'. The API prices all Azure services in USD; other currencies (e.g. 'EUR', 'GBP', 'JPY', 'AUD', 'CAD') are reference estimates."),
    topN: zod_1.z.number().optional()
        .describe("Maximum number of pricing results to return per query. Defaults to 20."),
    workspaceFolder: zod_1.z.string()
        .describe("The full path of the workspace folder. Use '/' as path separator.")
});
exports.AzureQuotaCheckSchema = zod_1.z.object({
    subscriptionId: zod_1.z.string()
        .describe("The Azure subscription ID where the resources will be deployed. Retrieve the subscription ID from the context, or by prompting the user to provide it."),
    region: zod_1.z.string()
        .describe("The valid Azure region where the resources will be deployed. E.g. 'eastus2', 'westus', 'westeurope', etc."),
    resourceTypes: zod_1.z.array(zod_1.z.string())
        .describe("The valid Azure resource types that are going to be deployed. E.g. 'Microsoft.App/containerApps', 'Microsoft.Web/sites', 'Microsoft.CognitiveServices/accounts', etc, or 'appservicemanagedinstance' ONLY if App Service Managed Instance is used."),
    workspaceFolder: zod_1.z.string()
        .describe("The full path of the workspace folder. Use '/' as path separator.")
});
exports.AzureRegionCheckSchema = zod_1.z.object({
    subscriptionId: zod_1.z.string()
        .describe("The user's Azure subscription ID (in GUID format). Retrieve the subscription ID from the context, or by prompting the user to provide it."),
    resourceTypes: zod_1.z.array(zod_1.z.string())
        .describe("The valid Azure resource types. E.g. 'Microsoft.App/containerApps', 'Microsoft.Web/sites', 'Microsoft.CognitiveServices/accounts', etc, or 'appservicemanagedinstance' ONLY if App Service Managed Instance is used."),
    cognitiveServiceProperties: zod_1.z.object({
        modelName: zod_1.z.string().optional()
            .describe("The name of the model."),
        modelVersion: zod_1.z.string().optional()
            .describe("The version of the model."),
        deploymentSkuName: zod_1.z.nativeEnum(params_1.SkuType)
            .optional()
            .describe("The SKU name of the model deployment.")
    })
        .describe("Optional configuration for model-specific checks. This is only needed when Microsoft.CognitiveServices is included in resourceTypes above.")
        .optional(),
    workspaceFolder: zod_1.z.string()
        .describe("The full path of the workspace folder. Use '/' as path separator.")
});
exports.AzureGetRegionsForDeploymentSchema = zod_1.z.object({
    subscriptionId: zod_1.z.string()
        .describe("The user's Azure subscription ID (in GUID format). DO NOT FORCE an azure login to use azcli context."),
    preferredRegions: zod_1.z.array(zod_1.z.string())
        .describe("The regions where user prefers to deploy, e.g. eastus2, australiaeast. If no specific preference, leave it blank")
        .optional(),
    resourceTypes: zod_1.z.array(zod_1.z.object({
        type: zod_1.z.string()
            .describe("The valid Azure resource types. E.g. 'Microsoft.App/containerApps', 'Microsoft.Web/sites', 'Microsoft.CognitiveServices/accounts', etc, or 'appservicemanagedinstance' ONLY if App Service Managed Instance is used."),
        quota: zod_1.z.number().min(1).default(1)
            .describe("The quota required for the resource type. Default to 1.")
    }))
        .describe("The valid Azure resource types and quota each resource needs."),
    cognitiveServiceProperties: zod_1.z.object({
        modelName: zod_1.z.string().optional()
            .describe("The name of the model."),
        modelVersion: zod_1.z.string().optional()
            .describe("The version of the model."),
        deploymentSkuName: zod_1.z.nativeEnum(params_1.SkuType)
            .optional()
            .describe("The SKU name of the model deployment.")
    })
        .describe("Optional configuration for model-specific checks. This is only needed when Microsoft.CognitiveServices is included in resourceTypes above.")
        .optional(),
    workspaceFolder: zod_1.z.string()
        .describe("The full path of the workspace folder. Use '/' as path separator.")
});
exports.DebugInBrowserSchema = zod_1.z.object({
    workspaceFolder: zod_1.z.string()
        .describe("The workspace folder of the current project. Use '/' as path separator."),
    webUrl: zod_1.z.string()
        .describe("The URL of the website to open."),
    actions: zod_1.z.array(zod_1.z.object({
        actionType: zod_1.z.nativeEnum(params_2.DebugActionType)
            .describe("The type of action to perform."),
        selector: zod_1.z.string()
            .describe("The CSS selector for the element to interact with."),
        actionValue: zod_1.z.string()
            .describe("The value associated with the action. Required for 'input' (text, not to be used for files), 'wait' (time in milliseconds), 'validateElementText' (expected text), and 'uploadFile' (file path).")
    }))
        .describe("An array of actions to perform on the webpage. Each action is an object with a type and optional parameters.")
});
exports.DeployPipelineSchema = zod_1.z.object({
    isAZDProject: zod_1.z.boolean()
        .describe("Whether to use azd tool in the deployment pipeline. Set to true ONLY if azure.yaml is provided or the context suggests AZD tools, using AZCLI by default."),
    pipelinePlatform: zod_1.z.nativeEnum(models_2.PipelinePlatform)
        .describe("The platform for the deployment pipeline."),
    deployOption: zod_1.z.nativeEnum(models_2.DeployOption)
        .optional()
        .describe("Default to deploy-only. Set to 'provision-and-deploy' ONLY WHEN user explicitly wants infra provisioning pipeline using local provisioning scripts."),
    subscriptionId: zod_1.z.string()
        .optional()
        .describe("The Azure subscription ID to deploy to."),
    workspaceFolder: zod_1.z.string()
        .describe("The full path of the workspace folder. Use '/' as path separator."),
    isAKS: zod_1.z.boolean()
        .describe("Whether the pipeline will be deploying to AKS (Azure Kubernetes Service). Infer from user's input or project context, or set to false by default.")
});
exports.GenerateArchitectureDiagramSchema = zod_1.z.object({
    workspaceFolder: zod_1.z.string()
        .describe("The full path of the workspace folder. Use '/' as path separator."),
    projectName: zod_1.z.string()
        .describe("The name of the project."),
    services: zod_1.z.array(zod_1.z.object({
        name: zod_1.z.string()
            .describe("The name of the service."),
        path: zod_1.z.string()
            .describe("The absolute path of the service folder. This folder should contain an entrypoint file such as main.py, index.js, Program.cs, App.java, etc. Use '/' as path separator."),
        language: zod_1.z.nativeEnum(models_2.LanguageType)
            .describe("The programming language of the service."),
        port: zod_1.z.string()
            .describe("The port number the service uses. Get this from Dockerfile or Kubernetes manifest. If not available, default to '80'."),
        azureComputeHost: zod_1.z.nativeEnum(models_2.TargetAppServiceType)
            .describe("The azure service to host this service. Keep consistent with plan tool."),
        dockerSettings: zod_1.z.object({
            dockerFilePath: zod_1.z.string()
                .describe("The absolute path to the Dockerfile for the service. Leave blank if the service is not containerized or Dockerfile is not yet created. Use '/' as path separator."),
            dockerContext: zod_1.z.string()
                .describe("The absolute path to the Docker build context for the service. Leave blank if the service is not containerized or Docker build context is not yet created. Use '/' as path separator."),
        }).describe("Docker settings for the service. Leave blank if the service is not containerized or Docker settings are not yet created.")
            .optional(),
        dependencies: zod_1.z.array(zod_1.z.object({
            name: zod_1.z.string()
                .describe("The name of the dependent service. Can be arbitary, or must reference another service in the services array if referencing azureappservice, azurecontainerapp, or azurefunctions. If existing azure resources are provided, try to match the dependency name to the existing resource name with the same type."),
            serviceType: zod_1.z.nativeEnum(models_2.AzureServiceNames)
                .describe("The type of the azure service that can be used for this dependent service. If existing azure resources are provided, only fill in the serviceType that represents the existing resource type."),
            connectionType: zod_1.z.nativeEnum(project_1.ConnectionType)
                .describe("The connection authentication type of the dependency."),
        })).describe("An array of dependent services. A compute service may have a dependency on another compute service."),
    })).describe("An array of services. Each service is a separately deployed compute service. A service can be a container app, aks, web app, function app, etc. Each service can have dependencies."),
});
exports.GetAppLogsSchema = zod_1.z.object({
    workspaceFolder: zod_1.z.string()
        .describe("The full path of the workspace folder. Use '/' as path separator."),
    subscriptionId: zod_1.z.string()
        .describe("The Azure subscription ID (AZURE_SUBSCRIPTION_ID). If not provided in context, try to find it in the .azure/**/env file."),
    limit: zod_1.z.number().min(1).optional()
        .describe("The maximum number of log rows to retrieve. Default is 200."),
    resourceGroupName: zod_1.z.string()
        .describe("The name of the Azure resource group where the resources are deployed. Used as the '--resource-group' argument in azcli commands (e.g. 'az webapp log tail --resource-group <name>', 'az containerapp logs show --resource-group <name>'). If not provided in context, try to find it in the project's IaC files (e.g. Bicep, ARM, Terraform) or ask the user."),
    resources: zod_1.z.array(zod_1.z.object({
        name: zod_1.z.string()
            .describe("The Azure resource name (the 'name' property, i.e., the last segment of the resource ID path)."),
        type: zod_1.z.nativeEnum(models_1.LLMSupportedResourceType)
            .describe("Resource type. Use exactly one of: 'containerapp', 'appservice', 'functionapp'. Do NOT use ARM-style types like 'Microsoft.App/containerApps'.")
    })).describe("List of Azure resources to query logs for. Each entry must have 'name' (resource name) and 'type'. Example: [{\"name\": \"myapp\", \"type\": \"containerapp\"}]")
});
exports.GetAzureResourceRulesSchema = zod_1.z.object({
    deploymentTool: zod_1.z.nativeEnum(models_2.DeploymentTool)
        .describe("Deployment tool used. Can be 'azd' or 'azcli'. 'azcli' is recommended."),
    iacType: zod_1.z.union([zod_1.z.nativeEnum(models_2.IacType), zod_1.z.literal("")])
        .describe("The type of IaC file used for deployment. Keep consistent with get-plan tool input. Leave empty if user wants to use azcli command script.")
        .optional(),
    resourceTypes: zod_1.z.array(zod_1.z.string())
        .describe("List of Azure resource types to generate rules for. Get the value from context and use the same resources defined in plan. Valid value: " + Object.values(params_3.RulesSupportedResourceType).join(',')),
    workspaceFolder: zod_1.z.string()
        .describe("The full path of the workspace folder. Use '/' as path separator.")
});
exports.GetContainerizationPlanSchema = zod_1.z.object({
    workspaceFolder: zod_1.z.string()
        .describe("The workspace folder of the current project. Use '/' as path separator."),
    servicePaths: zod_1.z.array(zod_1.z.string())
        .describe("The absolute paths of each service to deploy. Each service is a separately deployed compute service. Use '/' as path separator."),
    existingContainerRegistryResourceId: zod_1.z.string()
        .describe("The existing container registry to perform remote docker builds. Leave blank if not available.")
        .optional(),
});
exports.GetPlanSchema = zod_1.z.object({
    workspaceFolder: zod_1.z.string()
        .describe("The full path of the workspace folder. Use '/' as path separator."),
    projectName: zod_1.z.string()
        .describe("The name of the project to generate the deployment plan for. If not provided, will be inferred from the workspace."),
    services: zod_1.z.array(zod_1.z.object({
        name: zod_1.z.string()
            .describe("The name of the service."),
        path: zod_1.z.string()
            .describe("The absolute path of the service folder. This folder should contain an entrypoint file such as main.py, index.js, Program.cs, App.java, etc. Use '/' as path separator."),
        language: zod_1.z.nativeEnum(models_2.LanguageType)
            .describe("The programming language of the service."),
        azureComputeHost: zod_1.z.nativeEnum(models_2.TargetAppServiceType)
            .describe("The compute host Azure resource type for the service. If the project is written in .NET framework, use AzureAppServiceManagedInstance as default."),
        dependencies: zod_1.z.array(zod_1.z.object({
            name: zod_1.z.string()
                .describe("The name of the dependent service. Can be arbitrary, or must reference another service in the services array if referencing azureappservice, azurecontainerapp, or azurefunctions. If existing azure resources are provided, try to match the dependency name to the existing resource name with the same type."),
            serviceType: zod_1.z.nativeEnum(models_2.AzureServiceNames)
                .describe("The type of the azure service that can be used for this dependent service. If existing azure resources are provided, only fill in the serviceType that represents the existing resource type."),
            connectionType: zod_1.z.nativeEnum(project_1.ConnectionType)
                .describe("The connection authentication type of the dependency. Infer this from the user application."),
        })).describe("An array of dependent services. A compute service may have a dependency on another compute service."),
    }))
        .describe("The list of services in the project. Each service is a separately deployed compute service."),
    languages: zod_1.z.array(zod_1.z.nativeEnum(models_2.LanguageType))
        .describe("The list of programming languages used in the project for each service."),
    provisioningTool: zod_1.z.nativeEnum(models_2.DeploymentTool)
        .describe("The tool to use for provisioning Azure resources. Use azcli by default unless the context suggests using azd."),
    iacOptions: zod_1.z.union([zod_1.z.nativeEnum(models_2.IacType), zod_1.z.literal("")])
        .describe("The Infrastructure as Code option. Leave empty if user wants to use azcli command script.")
        .optional(),
    sourceType: zod_1.z.nativeEnum(models_2.SourceType)
        .describe("The source of the plan to generate from. If user doesn't have existing resources, set 'from-project' and generating deploy plan based on the project files in the workspace. If user mentions Azure resources exist, set 'from-azure' and ask for existing Azure resources details to generate plan. If the user have no existing resource but declare the expected Azure resources, use 'from-context' and the deploy plan should be based on the user's input."),
    deployOption: zod_1.z.nativeEnum(models_2.DeployOption)
        .describe("Set the value based on project and user's input. Use 'deploy-only' if user mentions they want to deploy to existing Azure resources or Iac files already exist in project, get Azure resource group from project files or from user. Use 'provision-only' if user only wants to provision Azure resource. Use 'provision-and-deploy' if user wants to deploy application and doesn't have existing infrastructure resources, or are starting from an empty resource group."),
    subscriptionId: zod_1.z.string()
        .describe("The Azure subscription ID to use for the deployment. Prompt the user to provide. Leave this empty if agent is not sure.")
        .optional(),
    resourceGroupName: zod_1.z.string()
        .describe("The Azure resource group name to use for the deployment. If user has existing resources and is deploying to existing resources, agent must use existing resource's resource group, ask for user's confirmation before using its resource group name.")
        .optional(),
});
exports.SummarizeResultSchema = zod_1.z.object({
    workspaceFolder: zod_1.z.string()
        .describe("The full path of the workspace folder. Use '/' as path separator."),
    projectName: zod_1.z.string()
        .describe("The name of the project."),
    deploymentStatus: zod_1.z.nativeEnum(models_2.DeploymentStatus)
        .describe("The final status of the deployment."),
    createdFileNames: zod_1.z.array(zod_1.z.string())
        .describe("List of files' name created during the deployment."),
    deployOption: zod_1.z.nativeEnum(models_2.DeployOption)
        .describe("Keep this consistent with the deploy option or infer from the deployment plan. 'deploy-only' if user deployed to existing Azure resources. Use 'provision-only' if user only provisioned Azure resource. Use 'provision-and-deploy' if user created infra from scratch and deployed application."),
    deploymentRounds: zod_1.z.number()
        .describe("Number of rounds the deployment was executed until a successful deployment."),
    provisionedResourceTypes: zod_1.z.array(zod_1.z.string())
        .describe("Types of Azure resources provisioned during the deployment."),
    deploymentTool: zod_1.z.nativeEnum(models_2.DeploymentTool)
        .describe("The tool to use for provisioning Azure resources."),
    iacOptions: zod_1.z.union([zod_1.z.nativeEnum(models_2.IacType), zod_1.z.literal("")])
        .describe("The Infrastructure as Code option. Leave empty if user wants to use azcli command script.")
        .optional(),
    subscriptionId: zod_1.z.string()
        .describe("The Azure subscription ID where the resources were provisioned and apps were deployed.")
});
exports.RunInTerminalSchema = zod_1.z.object({
    workingDir: zod_1.z.string()
        .describe("The working directory to run the command in."),
    executable: zod_1.z.nativeEnum(params_4.RunInTerminalExecutable)
        .describe("The type of shell to run. If not provided, powershell will be used."),
    commandToRun: zod_1.z.string()
        .describe("The command to run in the terminal."),
    env: zod_1.z.array(zod_1.z.object({
        key: zod_1.z.string(),
        value: zod_1.z.string()
    }))
        .describe("The environment variables to set for the command.")
});
exports.DiagnoseExistingResourcesSchema = zod_1.z.object({
    containerRegistryConnections: zod_1.z.object({
        containerRegistryId: zod_1.z.string()
            .describe("ID of the existing Azure Container Registry resource. Format: /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ContainerRegistry/registries/{registryName}"),
        containerAppIds: zod_1.z.array(zod_1.z.string())
            .describe("List of existing Azure Container App resource IDs which should access the Container Registry. Format: /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.App/containerApps/{containerAppName}"),
        kubernetesIds: zod_1.z.array(zod_1.z.string())
            .describe("List of existing Azure Kubernetes Service resource IDs which should access the Container Registry. Format: /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ContainerService/managedClusters/{clusterName}")
    })
        .describe("Container Registry configuration with associated container apps or Kubernetes services.")
        .optional(),
    keyVaultConnections: zod_1.z.object({
        keyVaultId: zod_1.z.string()
            .describe("ID of the existing Azure Key Vault resource. Format: /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.KeyVault/vaults/{vaultName}"),
        containerAppIds: zod_1.z.array(zod_1.z.string())
            .describe("List of existing Azure Container App resource IDs which should access the Key Vault. Format: /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.App/containerApps/{containerAppName}"),
        kubernetesIds: zod_1.z.array(zod_1.z.string())
            .describe("List of existing Azure Kubernetes Service resource IDs which should access the Key Vault. Format: /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ContainerService/managedClusters/{clusterName}")
    })
        .describe("Key Vault configuration with associated container apps or Kubernetes services.")
        .optional()
});
exports.GetALZPlanSchema = zod_1.z.object({
    iacType: zod_1.z.nativeEnum(models_2.IacType)
        .describe("The type of IaC files to generate for the customized Azure Landing Zone. Default to Bicep if not specified."),
});
exports.GetWAFRulesSchema = zod_1.z.object({
    resourceTypes: zod_1.z.nativeEnum(models_2.WafResourceType).array()
        .describe("The types of Azure resources to get WAF rules for.")
});
