// codetocloudmcp/out/toolsImplementations/shared/skuParser.js
"use strict";
// VM SKU parsing and sorting utilities
Object.defineProperty(exports, "__esModule", { value: true });
exports.collectVmSkusBySeries = collectVmSkusBySeries;
exports.sortAndSelectTopSkus = sortAndSelectTopSkus;
const SUPPORTED_VM_SERIES = ['D', 'B', 'E'];
const MAX_SKUS_PER_SERIES = 5;
/**
 * Parse version number from VM SKU name (e.g., v6 from Standard_D2s_v6)
 */
function parseVmSkuVersion(skuName) {
    const match = skuName.match(/_v(\d+)$/i);
    return match ? parseInt(match[1], 10) : 0;
}
/**
 * Get VM series prefix (D, B, or E) from SKU name
 * Returns null if not a supported series
 */
function getVmSeriesPrefix(skuName) {
    const match = skuName.match(/^Standard_([DBE])/i);
    if (match) {
        const prefix = match[1].toUpperCase();
        return SUPPORTED_VM_SERIES.includes(prefix) ? prefix : null;
    }
    return null;
}
/**
 * Parse VM SKU info from name (e.g., Standard_D2s_v6)
 */
function parseVmSkuInfo(skuName, family, vCPUs) {
    const namePart = skuName.replace(/^Standard_/i, '').split('_')[0].toLowerCase();
    return {
        name: skuName,
        family: family,
        vCPUs: vCPUs,
        version: parseVmSkuVersion(skuName),
        hasSSD: namePart.includes('s') && !namePart.includes('ls'),
        hasLocalDisk: namePart.includes('d'),
        isAmd: namePart.includes('a'),
    };
}
/**
 * Sort D-series VMs: prioritize newest version -> smaller vCPU -> SSD -> local disk
 * Example priority: D2s_v6 > D2ds_v4 > D2as_v6 > D2_v5 > D2lds_v5
 */
function sortVmDSeries(a, b) {
    if (b.version !== a.version) {
        return b.version - a.version;
    }
    if (a.vCPUs !== b.vCPUs) {
        return a.vCPUs - b.vCPUs;
    }
    if (a.hasSSD !== b.hasSSD) {
        return a.hasSSD ? -1 : 1;
    }
    if (a.hasLocalDisk !== b.hasLocalDisk) {
        return a.hasLocalDisk ? -1 : 1;
    }
    return 0;
}
/**
 * Sort B-series VMs: prioritize 2-4 vCPU burstable instances
 * Example priority: B2s_v2 > B2als_v2 > B2ls_v2 > B4als_v2 > B2pls_v2
 */
function sortVmBSeries(a, b) {
    const aInRange = a.vCPUs >= 2 && a.vCPUs <= 4;
    const bInRange = b.vCPUs >= 2 && b.vCPUs <= 4;
    if (aInRange !== bInRange) {
        return aInRange ? -1 : 1;
    }
    if (b.version !== a.version) {
        return b.version - a.version;
    }
    if (a.vCPUs !== b.vCPUs) {
        return a.vCPUs - b.vCPUs;
    }
    if (a.hasSSD !== b.hasSSD) {
        return a.hasSSD ? -1 : 1;
    }
    return 0;
}
/**
 * Sort E-series VMs: prioritize 2-4 vCPU memory-optimized instances
 * Example priority: E2s_v6 > E2ds_v4 > E2as_v6 > E2ads_v6 > E2_v5
 */
function sortVmESeries(a, b) {
    const aInRange = a.vCPUs >= 2 && a.vCPUs <= 4;
    const bInRange = b.vCPUs >= 2 && b.vCPUs <= 4;
    if (aInRange !== bInRange) {
        return aInRange ? -1 : 1;
    }
    if (b.version !== a.version) {
        return b.version - a.version;
    }
    if (a.vCPUs !== b.vCPUs) {
        return a.vCPUs - b.vCPUs;
    }
    if (a.hasSSD !== b.hasSSD) {
        return a.hasSSD ? -1 : 1;
    }
    return 0;
}
function getVmSeriesSortFunction(series) {
    switch (series) {
        case 'D':
            return sortVmDSeries;
        case 'B':
            return sortVmBSeries;
        case 'E':
            return sortVmESeries;
    }
}
/**
 * Collect and categorize VM SKUs by series (D, B, E only)
 */
function collectVmSkusBySeries(skus, usageMap, availableCores) {
    const skusBySeries = new Map([
        ['D', []],
        ['B', []],
        ['E', []]
    ]);
    for (const sku of skus) {
        if (!sku.family || !sku.name) {
            continue;
        }
        const seriesPrefix = getVmSeriesPrefix(sku.name);
        if (!seriesPrefix) {
            continue;
        }
        const normalizedFamily = sku.family.replace(/\s+/g, '').toLowerCase();
        const usageInfo = usageMap.get(normalizedFamily);
        if (!usageInfo) {
            continue;
        }
        const vCPUsCapability = sku.capabilities?.find(cap => cap.name === 'vCPUs');
        const requiredCores = vCPUsCapability ? parseInt(vCPUsCapability.value || '0', 10) : 0;
        if (requiredCores > 0 && requiredCores <= availableCores) {
            const skuInfo = parseVmSkuInfo(sku.name, normalizedFamily, requiredCores);
            skusBySeries.get(seriesPrefix).push({ ...skuInfo, usageInfo });
        }
    }
    return skusBySeries;
}
/**
 * Sort SKUs by series priority and select top N from each series
 */
function sortAndSelectTopSkus(skusBySeries) {
    const result = [];
    for (const [series, skuList] of skusBySeries) {
        const sortFn = getVmSeriesSortFunction(series);
        skuList.sort(sortFn);
        const topSkus = skuList.slice(0, MAX_SKUS_PER_SERIES);
        for (const sku of topSkus) {
            result.push({
                name: sku.name,
                limit: sku.usageInfo.limit,
                used: sku.usageInfo.used
            });
            console.log(`${series}-series: ${sku.name}, vCPUs: ${sku.vCPUs}, version: v${sku.version}, SSD: ${sku.hasSSD}`);
        }
    }
    return result;
}
