// codetocloudmcp/out/toolsImplementations/shared/toolType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToolType = void 0;
// Used as identifier in telemetry. DO NOT change the values here.
var ToolType;
(function (ToolType) {
    // c2c tools
    ToolType["GetPlan"] = "appmod-get-plan";
    ToolType["GetContainerizationPlan"] = "appmod-get-containerization-plan";
    ToolType["GetAppTopology"] = "appmod-generate-architecture-diagram";
    ToolType["GetAzureResourceRules"] = "appmod-get-iac-rules";
    ToolType["GetAppLogs"] = "appmod-get-azd-app-logs";
    ToolType["DeployPipeline"] = "appmod-get-cicd-pipeline-guidance";
    ToolType["GetRegionsForDeployment"] = "appmod-get-available-region-sku";
    ToolType["SummarizeResult"] = "appmod-summarize-result";
    ToolType["AzureRegionCheck"] = "appmod-get-available-region";
    ToolType["AzureQuotaCheck"] = "appmod-check-quota";
    ToolType["AzurePricing"] = "appmod-get-azure-pricing";
    // container assist tools
    ToolType["AnalyzeRepository"] = "appmod-analyze-repository";
    ToolType["PlanGenerateDockerfile"] = "appmod-plan-generate-dockerfile";
    ToolType["BuildDockerImage"] = "appmod-build-docker-image";
    ToolType["GenerateKubernetesManifest"] = "appmod-generate-k8s-manifest";
    ToolType["RunSubagent"] = "RunSubagent";
    // Test or deprecated tools
    ToolType["RunInTerminal"] = "RunInTerminal";
    ToolType["DebugInBrowser"] = "DebugInBrowser";
    ToolType["AzureImportResource"] = "AzureImportResource";
    ToolType["AzdDeploy"] = "AzdDeploy";
    ToolType["DiagnoseExistingResources"] = "DiagnoseExistingResources";
    ToolType["GetALZPlan"] = "appmod-get-azure-landing-zone-plan";
    ToolType["GetWAFRules"] = "appmod-get-waf-rules";
})(ToolType || (exports.ToolType = ToolType = {}));
