// containerization-assist-mcp/dist-cjs/src/infra/security/grype-scanner.js
"use strict";
/**
 * Grype Security Scanner Implementation
 *
 * Integrates with Grype CLI for container image vulnerability scanning.
 * Grype is an open-source vulnerability scanner maintained by Anchore.
 *
 * @see https://github.com/anchore/grype
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkGrypeAvailability = checkGrypeAvailability;
exports.scanImageWithGrype = scanImageWithGrype;
const node_child_process_1 = require("node:child_process");
const node_util_1 = require("node:util");
const errors_1 = require("../../lib/errors");
const types_1 = require("../../types");
const constants_1 = require("../../config/constants");
const scanner_common_1 = require("./scanner-common");
const execFileAsync = (0, node_util_1.promisify)(node_child_process_1.execFile);
/**
 * Parse Grype JSON output to our BasicScanResult format
 */
function parseGrypeOutput(grypeOutput, imageId) {
    const vulnerabilities = [];
    const counter = new scanner_common_1.SeverityCounter();
    // Grype uses matches array with nested vulnerability objects
    for (const match of grypeOutput.matches || []) {
        const vuln = match.vulnerability;
        const artifact = match.artifact;
        const severity = (0, scanner_common_1.normalizeSeverity)(vuln.severity);
        counter.increment(severity);
        // Build vulnerability entry
        const vulnEntry = {
            id: vuln.id,
            severity,
            package: artifact.name,
            version: artifact.version,
            description: vuln.description || 'No description available',
        };
        // Add fixedVersion if available (from fix.versions array)
        if (vuln.fix?.versions && vuln.fix.versions.length > 0 && vuln.fix.state === 'fixed') {
            const fixedVersion = vuln.fix.versions[0];
            if (fixedVersion) {
                vulnEntry.fixedVersion = fixedVersion;
            }
        }
        vulnerabilities.push(vulnEntry);
    }
    return {
        imageId,
        vulnerabilities,
        scanDate: new Date(),
        ...counter.getCounts(),
    };
}
/**
 * Get Grype version
 * @throws Error if Grype is not installed or execution fails
 */
async function getGrypeVersion(logger) {
    try {
        const { stdout } = await execFileAsync('grype', ['version'], { timeout: 5000 });
        // Grype version output format: multi-line with "Version: X.Y.Z"
        const version = (0, scanner_common_1.parseVersion)(stdout, /Version:\s*([^\s\n]+)/);
        if (!version) {
            logger.debug({ stdout }, 'Could not parse Grype version from output');
            return scanner_common_1.ScannerErrors.versionParseError('Grype', 'grype version');
        }
        return (0, types_1.Success)(version);
    }
    catch (error) {
        const err = error;
        if (err?.code === 'ETIMEDOUT') {
            logger.error({ error }, 'Grype version check timed out');
            return scanner_common_1.ScannerErrors.versionCheckTimeout('Grype', 'grype version');
        }
        throw error;
    }
}
/**
 * Check if Grype is installed and accessible
 */
async function checkGrypeAvailability(logger) {
    try {
        const versionResult = await getGrypeVersion(logger);
        if (!versionResult.ok) {
            return versionResult;
        }
        return (0, types_1.Success)(versionResult.value);
    }
    catch {
        return scanner_common_1.ScannerErrors.scannerNotInstalled('Grype', 'brew install grype or https://github.com/anchore/grype#installation');
    }
}
/**
 * Scan a Docker image using Grype
 */
async function scanImageWithGrype(imageId, logger) {
    // Validate imageId to prevent command injection
    if (!(0, scanner_common_1.validateImageId)(imageId)) {
        return scanner_common_1.ScannerErrors.invalidImageId(imageId);
    }
    // Check if Grype is available
    const availabilityCheck = await checkGrypeAvailability(logger);
    if (!availabilityCheck.ok) {
        return (0, types_1.Failure)(availabilityCheck.error, availabilityCheck.guidance);
    }
    const grypeVersion = availabilityCheck.value;
    (0, scanner_common_1.logScanStart)(logger, 'Grype', grypeVersion, imageId);
    try {
        // Run Grype scan with JSON output
        // -o json: output in JSON format
        // -q: quiet mode (suppress non-essential output)
        const args = [imageId, '-o', 'json', '-q'];
        logger.debug({ args }, 'Executing Grype command');
        const { stdout, stderr } = await execFileAsync('grype', args, {
            maxBuffer: constants_1.LIMITS.MAX_SCAN_BUFFER,
        });
        // Log any warnings from stderr
        if (stderr) {
            logger.debug({ stderr }, 'Grype stderr output');
        }
        // Validate output size before parsing
        if (stdout.length === 0) {
            return scanner_common_1.ScannerErrors.emptyOutput('Grype', imageId);
        }
        // Parse JSON output
        let grypeOutput;
        try {
            grypeOutput = JSON.parse(stdout);
        }
        catch (parseError) {
            return scanner_common_1.ScannerErrors.jsonParseError('Grype', (0, errors_1.extractErrorMessage)(parseError), stdout.substring(0, 200));
        }
        // Parse the Grype output into our format
        const scanResult = parseGrypeOutput(grypeOutput, imageId);
        (0, scanner_common_1.logScanComplete)(logger, 'Grype', imageId, scanResult.totalVulnerabilities, scanResult.criticalCount, scanResult.highCount);
        return (0, types_1.Success)(scanResult);
    }
    catch (error) {
        const errorMessage = (0, errors_1.extractErrorMessage)(error);
        logger.error({ error: errorMessage, imageId }, 'Grype scan failed');
        return scanner_common_1.ScannerErrors.scanExecutionError('Grype', imageId, errorMessage);
    }
}
