// containerization-assist-mcp/dist-cjs/src/infra/security/scanner-common.js
"use strict";
/**
 * Common Scanner Utilities
 *
 * Shared functionality across security scanner implementations
 * to reduce code duplication and centralize security-critical logic.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SeverityCounter = exports.ScannerErrors = void 0;
exports.validateImageId = validateImageId;
exports.normalizeSeverity = normalizeSeverity;
exports.parseVersion = parseVersion;
exports.logScanStart = logScanStart;
exports.logScanComplete = logScanComplete;
const types_1 = require("../../types");
/**
 * Safely quote a value for use in shell commands
 * Uses POSIX single-quote escaping: wraps in '' and escapes internal quotes as '\''
 * This is the standard approach used by shell-quote and similar libraries
 *
 * @example
 * shellQuote("hello") => "'hello'"
 * shellQuote("it's") => "'it'\\''s'"
 * shellQuote("a;rm -rf /") => "'a;rm -rf /'"
 */
function shellQuote(value) {
    // POSIX single-quote escaping: replace ' with '\''
    // This closes the quote, adds escaped quote, then reopens quote
    return `'${value.replace(/'/g, "'\\''")}'`;
}
/**
 * Common error messages and guidance for scanner implementations
 */
exports.ScannerErrors = {
    /**
     * Create invalid imageId error
     */
    invalidImageId: (imageId) => (0, types_1.Failure)('Invalid imageId format', {
        message: 'ImageId contains invalid characters',
        hint: 'ImageId must contain only alphanumeric characters, dots, colons, slashes, at-signs, underscores, and hyphens',
        resolution: 'Verify the imageId is a valid Docker image identifier',
        details: { imageId },
    }),
    /**
     * Create scanner not installed error
     */
    scannerNotInstalled: (scannerName, installUrl) => (0, types_1.Failure)(`${scannerName} not installed or not in PATH`, {
        message: `${scannerName} CLI not found`,
        hint: `${scannerName} CLI is required for security scanning`,
        resolution: `Install ${scannerName}: ${installUrl}`,
    }),
    /**
     * Create JSON parse error
     */
    jsonParseError: (scannerName, parseError, outputPreview) => (0, types_1.Failure)(`Failed to parse ${scannerName} output`, {
        message: `${scannerName} output parsing failed`,
        hint: `${scannerName} may have returned invalid JSON`,
        resolution: `Try running ${scannerName} manually to verify`,
        details: {
            parseError,
            outputPreview,
        },
    }),
    /**
     * Create empty output error
     */
    emptyOutput: (scannerName, imageId) => (0, types_1.Failure)(`${scannerName} returned empty output`, {
        message: 'No scan results received',
        hint: `${scannerName} may not have found the image or encountered an error`,
        resolution: `Verify image exists: docker image inspect ${shellQuote(imageId)}`,
    }),
    /**
     * Create scan execution error
     */
    scanExecutionError: (scannerName, imageId, errorMessage) => (0, types_1.Failure)(`${scannerName} scan failed: ${errorMessage}`, {
        message: 'Security scan execution failed',
        hint: `${scannerName} encountered an error while scanning the image`,
        resolution: `Check image exists and is accessible: docker image ls | grep ${shellQuote(imageId)}`,
        details: { error: errorMessage },
    }),
    /**
     * Create version check timeout error
     */
    versionCheckTimeout: (scannerName, command) => (0, types_1.Failure)(`${scannerName} version check timed out`, {
        message: 'Command execution timeout',
        hint: `${scannerName} CLI took too long to respond`,
        resolution: `Check if ${scannerName} is functioning correctly: ${command}`,
    }),
    /**
     * Create version parse error
     */
    versionParseError: (scannerName, command) => (0, types_1.Failure)(`${scannerName} version could not be parsed`, {
        message: `${scannerName} version check failed`,
        hint: `${scannerName} CLI may not be properly configured`,
        resolution: `Try running: ${command}`,
    }),
};
/**
 * Validate imageId against allowlist pattern to prevent command injection
 * Allows: alphanumeric, dots, colons, slashes, at-signs, underscores, and hyphens
 */
function validateImageId(imageId) {
    const allowedPattern = /^[a-zA-Z0-9._:/@-]+$/;
    return allowedPattern.test(imageId);
}
/**
 * Normalize severity string to our standard format
 * Handles various case formats (UPPERCASE, lowercase, TitleCase)
 */
function normalizeSeverity(severity) {
    const normalized = severity.toUpperCase();
    switch (normalized) {
        case 'CRITICAL':
            return 'CRITICAL';
        case 'HIGH':
            return 'HIGH';
        case 'MEDIUM':
            return 'MEDIUM';
        case 'LOW':
            return 'LOW';
        case 'NEGLIGIBLE':
            return 'NEGLIGIBLE';
        case 'UNKNOWN':
            return 'UNKNOWN';
        default:
            return 'UNKNOWN';
    }
}
/**
 * Severity counter helper
 */
class SeverityCounter {
    critical = 0;
    high = 0;
    medium = 0;
    low = 0;
    negligible = 0;
    unknown = 0;
    increment(severity) {
        switch (severity) {
            case 'CRITICAL':
                this.critical++;
                break;
            case 'HIGH':
                this.high++;
                break;
            case 'MEDIUM':
                this.medium++;
                break;
            case 'LOW':
                this.low++;
                break;
            case 'NEGLIGIBLE':
                this.negligible++;
                break;
            case 'UNKNOWN':
                this.unknown++;
                break;
        }
    }
    get total() {
        return this.critical + this.high + this.medium + this.low + this.negligible + this.unknown;
    }
    getCounts() {
        return {
            criticalCount: this.critical,
            highCount: this.high,
            mediumCount: this.medium,
            lowCount: this.low,
            negligibleCount: this.negligible,
            unknownCount: this.unknown,
            totalVulnerabilities: this.total,
        };
    }
}
exports.SeverityCounter = SeverityCounter;
/**
 * Parse version from scanner output using a regex pattern
 */
function parseVersion(output, pattern) {
    const match = output.match(pattern);
    return match?.[1]?.trim();
}
/**
 * Log scanner information
 */
function logScanStart(logger, scanner, version, imageId) {
    logger.info({ scanner, version, imageId }, `Starting ${scanner} scan`);
}
/**
 * Log scan completion
 */
function logScanComplete(logger, scanner, imageId, totalVulnerabilities, criticalCount, highCount) {
    logger.info({
        scanner,
        imageId,
        totalVulnerabilities,
        criticalCount,
        highCount,
    }, `${scanner} scan completed successfully`);
}
