// containerization-assist-mcp/dist-cjs/src/infra/security/scanner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSecurityScanner = void 0;
const types_1 = require("../../types");
const errors_1 = require("../../lib/errors");
const trivy_scanner_1 = require("./trivy-scanner");
/**
 * Create a Trivy-based security scanner
 */
function createTrivyScanner(logger) {
    return {
        async scanImage(imageId) {
            return (0, trivy_scanner_1.scanImageWithTrivy)(imageId, logger);
        },
        async ping() {
            const result = await (0, trivy_scanner_1.checkTrivyAvailability)(logger);
            if (result.ok) {
                logger.debug({ version: result.value }, 'Trivy scanner available');
                return (0, types_1.Success)(true);
            }
            return (0, types_1.Failure)(result.error, result.guidance);
        },
    };
}
/**
 * Create a stub scanner that returns empty results
 * Used when no real scanner is configured
 */
function createStubScanner(logger) {
    return {
        async scanImage(imageId) {
            try {
                logger.info({ imageId, scanner: 'stub' }, 'Starting stub security scan');
                const result = {
                    imageId,
                    vulnerabilities: [],
                    totalVulnerabilities: 0,
                    criticalCount: 0,
                    highCount: 0,
                    mediumCount: 0,
                    lowCount: 0,
                    negligibleCount: 0,
                    unknownCount: 0,
                    scanDate: new Date(),
                };
                logger.warn({ imageId }, 'Stub scanner returns empty results - no actual scanning performed');
                return (0, types_1.Success)(result);
            }
            catch (error) {
                const errorMessage = (0, errors_1.extractErrorMessage)(error);
                logger.error({ error: errorMessage, imageId }, 'Security scan failed');
                return (0, types_1.Failure)(errorMessage);
            }
        },
        async ping() {
            logger.debug('Checking stub scanner availability');
            return (0, types_1.Success)(true);
        },
    };
}
/**
 * Create a security scanner based on the specified type
 *
 * @param logger - Logger instance
 * @param scannerType - Type of scanner to create ('trivy', 'stub', or undefined for 'trivy')
 * @returns SecurityScanner instance
 */
const createSecurityScanner = (logger, scannerType) => {
    const type = (scannerType || 'trivy').toLowerCase();
    switch (type) {
        case 'trivy':
            return createTrivyScanner(logger);
        case 'stub':
            return createStubScanner(logger);
        default:
            logger.warn({ scannerType: type }, 'Unknown scanner type, falling back to Trivy');
            return createTrivyScanner(logger);
    }
};
exports.createSecurityScanner = createSecurityScanner;
