// containerization-assist-mcp/dist-cjs/src/infra/security/scanner.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSecurityScanner = void 0;
const dockerode_1 = __importDefault(require("dockerode"));
const types_1 = require("../../types");
const errors_1 = require("../../lib/errors");
const trivy_scanner_1 = require("./trivy-scanner");
const snyk_scanner_1 = require("./snyk-scanner");
const grype_scanner_1 = require("./grype-scanner");
const index_1 = require("./osv-scanner/index");
const socket_validation_1 = require("../docker/socket-validation");
/**
 * Create a Trivy-based security scanner
 */
function createTrivyScanner(logger) {
    return {
        async scanImage(imageId) {
            return (0, trivy_scanner_1.scanImageWithTrivy)(imageId, logger);
        },
        async ping() {
            const result = await (0, trivy_scanner_1.checkTrivyAvailability)(logger);
            if (result.ok) {
                logger.debug({ version: result.value }, 'Trivy scanner available');
                return (0, types_1.Success)(true);
            }
            return (0, types_1.Failure)(result.error, result.guidance);
        },
    };
}
/**
 * Create a Snyk-based security scanner
 */
function createSnykScanner(logger) {
    return {
        async scanImage(imageId) {
            return (0, snyk_scanner_1.scanImageWithSnyk)(imageId, logger);
        },
        async ping() {
            const result = await (0, snyk_scanner_1.checkSnykAvailability)(logger);
            if (result.ok) {
                logger.debug({ version: result.value }, 'Snyk scanner available');
                return (0, types_1.Success)(true);
            }
            return (0, types_1.Failure)(result.error, result.guidance);
        },
    };
}
/**
 * Create a Grype-based security scanner
 */
function createGrypeScanner(logger) {
    return {
        async scanImage(imageId) {
            return (0, grype_scanner_1.scanImageWithGrype)(imageId, logger);
        },
        async ping() {
            const result = await (0, grype_scanner_1.checkGrypeAvailability)(logger);
            if (result.ok) {
                logger.debug({ version: result.value }, 'Grype scanner available');
                return (0, types_1.Success)(true);
            }
            return (0, types_1.Failure)(result.error, result.guidance);
        },
    };
}
/**
 * Create an OSV-based security scanner
 * Uses OSV API (no external CLI required)
 */
function createOSVScanner(logger) {
    // Create Docker client for image inspection
    const socketPath = (0, socket_validation_1.autoDetectDockerSocket)();
    const docker = new dockerode_1.default({ socketPath });
    return {
        async scanImage(imageId) {
            return (0, index_1.scanImageWithOSV)(docker, imageId, logger);
        },
        async ping() {
            const result = await (0, index_1.checkOSVAvailability)(logger);
            if (result.ok) {
                logger.debug('OSV scanner available');
                return (0, types_1.Success)(true);
            }
            return (0, types_1.Failure)(result.error, result.guidance);
        },
    };
}
/**
 * Create a stub scanner that returns empty results
 * Used when no real scanner is configured
 */
function createStubScanner(logger) {
    return {
        async scanImage(imageId) {
            try {
                logger.info({ imageId, scanner: 'stub' }, 'Starting stub security scan');
                const result = {
                    imageId,
                    vulnerabilities: [],
                    totalVulnerabilities: 0,
                    criticalCount: 0,
                    highCount: 0,
                    mediumCount: 0,
                    lowCount: 0,
                    negligibleCount: 0,
                    unknownCount: 0,
                    scanDate: new Date(),
                };
                logger.warn({ imageId }, 'Stub scanner returns empty results - no actual scanning performed');
                return (0, types_1.Success)(result);
            }
            catch (error) {
                const errorMessage = (0, errors_1.extractErrorMessage)(error);
                logger.error({ error: errorMessage, imageId }, 'Security scan failed');
                return (0, types_1.Failure)(errorMessage);
            }
        },
        async ping() {
            logger.debug('Checking stub scanner availability');
            return (0, types_1.Success)(true);
        },
    };
}
/**
 * Create a security scanner based on the specified type
 *
 * @param logger - Logger instance
 * @param scannerType - Type of scanner to create ('osv', 'trivy', 'snyk', 'grype', 'stub', or undefined for 'osv')
 * @returns SecurityScanner instance
 */
const createSecurityScanner = (logger, scannerType) => {
    const type = (scannerType || 'osv').toLowerCase();
    switch (type) {
        case 'osv':
            return createOSVScanner(logger);
        case 'trivy':
            return createTrivyScanner(logger);
        case 'snyk':
            return createSnykScanner(logger);
        case 'grype':
            return createGrypeScanner(logger);
        case 'stub':
            return createStubScanner(logger);
        default:
            logger.warn({ scannerType: type }, 'Unknown scanner type, falling back to OSV');
            return createOSVScanner(logger);
    }
};
exports.createSecurityScanner = createSecurityScanner;
