// containerization-assist-mcp/dist-cjs/src/infra/security/snyk-scanner.js
"use strict";
/**
 * Snyk Security Scanner Implementation
 *
 * Integrates with Snyk CLI for container image vulnerability scanning.
 * Snyk is a developer security platform by Snyk Ltd.
 *
 * @see https://docs.snyk.io/snyk-cli
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkSnykAvailability = checkSnykAvailability;
exports.scanImageWithSnyk = scanImageWithSnyk;
const node_child_process_1 = require("node:child_process");
const node_util_1 = require("node:util");
const errors_1 = require("../../lib/errors");
const types_1 = require("../../types");
const constants_1 = require("../../config/constants");
const scanner_common_1 = require("./scanner-common");
const execFileAsync = (0, node_util_1.promisify)(node_child_process_1.execFile);
/**
 * Parse Snyk JSON output to our BasicScanResult format
 */
function parseSnykOutput(snykOutput, imageId) {
    const vulnerabilities = [];
    const counter = new scanner_common_1.SeverityCounter();
    // Snyk uses flat vulnerabilities array
    for (const vuln of snykOutput.vulnerabilities || []) {
        const severity = (0, scanner_common_1.normalizeSeverity)(vuln.severity);
        counter.increment(severity);
        // Build vulnerability entry
        const vulnEntry = {
            id: vuln.id,
            severity,
            package: vuln.packageName,
            version: vuln.version,
            description: vuln.title || 'No description available',
        };
        // Add fixedVersion if available (using Snyk's nearestFixedInVersion)
        if (vuln.nearestFixedInVersion !== undefined) {
            vulnEntry.fixedVersion = vuln.nearestFixedInVersion;
        }
        vulnerabilities.push(vulnEntry);
    }
    return {
        imageId,
        vulnerabilities,
        scanDate: new Date(),
        ...counter.getCounts(),
    };
}
/**
 * Get Snyk version
 * @throws Error if Snyk is not installed or execution fails
 */
async function getSnykVersion(logger) {
    try {
        const { stdout } = await execFileAsync('snyk', ['--version'], { timeout: 5000 });
        // Snyk version output format: just the version number (e.g., "1.1230.0")
        const version = stdout.trim();
        if (!version) {
            return scanner_common_1.ScannerErrors.versionParseError('Snyk', 'snyk --version');
        }
        return (0, types_1.Success)(version);
    }
    catch (error) {
        const err = error;
        if (err?.code === 'ETIMEDOUT') {
            logger.error({ error }, 'Snyk version check timed out');
            return scanner_common_1.ScannerErrors.versionCheckTimeout('Snyk', 'snyk --version');
        }
        throw error;
    }
}
/**
 * Check if Snyk is installed and accessible
 */
async function checkSnykAvailability(logger) {
    try {
        const versionResult = await getSnykVersion(logger);
        if (!versionResult.ok) {
            return versionResult;
        }
        return (0, types_1.Success)(versionResult.value);
    }
    catch {
        return scanner_common_1.ScannerErrors.scannerNotInstalled('Snyk', 'npm install -g snyk or https://docs.snyk.io/snyk-cli/install-the-snyk-cli');
    }
}
/**
 * Check if Snyk is authenticated
 */
async function checkSnykAuth(logger) {
    // Check for SNYK_TOKEN environment variable
    if (process.env.SNYK_TOKEN) {
        logger.debug('SNYK_TOKEN environment variable found');
        return (0, types_1.Success)(true);
    }
    // Try to verify auth status by running a quick command
    try {
        await execFileAsync('snyk', ['config', 'get', 'api'], { timeout: 5000 });
        return (0, types_1.Success)(true);
    }
    catch (error) {
        return (0, types_1.Failure)('Snyk authentication required', {
            message: 'Snyk CLI requires authentication',
            hint: 'Snyk needs an API token to scan images',
            resolution: 'Set SNYK_TOKEN environment variable or run: snyk auth\nGet a token from: https://app.snyk.io/account',
            details: { error: (0, errors_1.extractErrorMessage)(error) },
        });
    }
}
/**
 * Scan a Docker image using Snyk
 */
async function scanImageWithSnyk(imageId, logger) {
    // Validate imageId to prevent command injection
    if (!(0, scanner_common_1.validateImageId)(imageId)) {
        return scanner_common_1.ScannerErrors.invalidImageId(imageId);
    }
    // Check if Snyk is available
    const availabilityCheck = await checkSnykAvailability(logger);
    if (!availabilityCheck.ok) {
        return (0, types_1.Failure)(availabilityCheck.error, availabilityCheck.guidance);
    }
    // Check if Snyk is authenticated
    const authCheck = await checkSnykAuth(logger);
    if (!authCheck.ok) {
        return (0, types_1.Failure)(authCheck.error, authCheck.guidance);
    }
    const snykVersion = availabilityCheck.value;
    (0, scanner_common_1.logScanStart)(logger, 'Snyk', snykVersion, imageId);
    try {
        // Run Snyk scan with JSON output
        // container test: scan container image
        // --json: output in JSON format
        const args = ['container', 'test', imageId, '--json'];
        logger.debug({ args }, 'Executing Snyk command');
        const { stdout, stderr } = await execFileAsync('snyk', args, {
            maxBuffer: constants_1.LIMITS.MAX_SCAN_BUFFER,
            env: {
                ...process.env,
                // Ensure token is passed if set
                ...(process.env.SNYK_TOKEN && { SNYK_TOKEN: process.env.SNYK_TOKEN }),
            },
        });
        // Log any warnings from stderr
        if (stderr) {
            logger.debug({ stderr }, 'Snyk stderr output');
        }
        // Validate output size before parsing
        if (stdout.length === 0) {
            return scanner_common_1.ScannerErrors.emptyOutput('Snyk', imageId);
        }
        // Parse JSON output
        let snykOutput;
        try {
            snykOutput = JSON.parse(stdout);
        }
        catch (parseError) {
            return scanner_common_1.ScannerErrors.jsonParseError('Snyk', (0, errors_1.extractErrorMessage)(parseError), stdout.substring(0, 200));
        }
        // Check for errors in the output
        if (snykOutput.error) {
            return (0, types_1.Failure)(`Snyk scan error: ${snykOutput.error}`, {
                message: 'Snyk encountered an error',
                hint: snykOutput.error,
                resolution: 'Check Snyk documentation or run command manually for more details',
            });
        }
        // Parse the Snyk output into our format
        const scanResult = parseSnykOutput(snykOutput, imageId);
        (0, scanner_common_1.logScanComplete)(logger, 'Snyk', imageId, scanResult.totalVulnerabilities, scanResult.criticalCount, scanResult.highCount);
        return (0, types_1.Success)(scanResult);
    }
    catch (error) {
        // When Snyk exits with non-zero code, try to parse stdout for error details
        const hasStdout = (err) => {
            return typeof err === 'object' && err !== null && 'stdout' in err;
        };
        if (hasStdout(error)) {
            const stdout = String(error.stdout || '').trim();
            if (stdout) {
                try {
                    const snykOutput = JSON.parse(stdout);
                    if (snykOutput.error) {
                        const isAuthError = snykOutput.error.includes('auth') || snykOutput.error.includes('authenticate');
                        return (0, types_1.Failure)(`Snyk ${isAuthError ? 'authentication failed' : 'scan error'}: ${snykOutput.error}`, {
                            message: isAuthError ? 'Authentication required' : 'Snyk encountered an error',
                            hint: snykOutput.error,
                            resolution: isAuthError
                                ? 'Option 1: Set environment variable: export SNYK_TOKEN=<your-token>\nOption 2: Run: snyk auth\n\nGet a token from: https://app.snyk.io/account'
                                : 'Check Snyk documentation or run command manually for more details',
                            details: { error: snykOutput.error },
                        });
                    }
                }
                catch {
                    // Fall through if JSON parsing fails
                }
            }
        }
        // Generic error handling
        const errorMessage = (0, errors_1.extractErrorMessage)(error);
        logger.error({ error: errorMessage, imageId }, 'Snyk scan failed');
        return scanner_common_1.ScannerErrors.scanExecutionError('Snyk', imageId, errorMessage);
    }
}
