// containerization-assist-mcp/dist-cjs/src/infra/security/trivy-scanner.js
"use strict";
/**
 * Trivy Security Scanner Implementation
 *
 * Integrates with Trivy CLI for container image vulnerability scanning.
 * Trivy is an industry-standard security scanner maintained by Aqua Security.
 *
 * @see https://aquasecurity.github.io/trivy/
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkTrivyAvailability = checkTrivyAvailability;
exports.scanImageWithTrivy = scanImageWithTrivy;
const node_child_process_1 = require("node:child_process");
const node_util_1 = require("node:util");
const errors_1 = require("../../lib/errors");
const types_1 = require("../../types");
const constants_1 = require("../../config/constants");
const scanner_common_1 = require("./scanner-common");
const execAsync = (0, node_util_1.promisify)(node_child_process_1.exec);
const execFileAsync = (0, node_util_1.promisify)(node_child_process_1.execFile);
/**
 * Parse Trivy JSON output to our BasicScanResult format
 */
function parseTrivyOutput(trivyOutput, imageId) {
    const vulnerabilities = [];
    const counter = new scanner_common_1.SeverityCounter();
    // Iterate through all results and their vulnerabilities
    for (const result of trivyOutput.Results || []) {
        for (const vuln of result.Vulnerabilities || []) {
            const severity = (0, scanner_common_1.normalizeSeverity)(vuln.Severity);
            counter.increment(severity);
            // Build vulnerability entry
            const vulnEntry = {
                id: vuln.VulnerabilityID,
                severity,
                package: vuln.PkgName,
                version: vuln.InstalledVersion,
                description: vuln.Title || vuln.Description || 'No description available',
            };
            // Only add fixedVersion if it exists (exactOptionalPropertyTypes compliance)
            if (vuln.FixedVersion !== undefined) {
                vulnEntry.fixedVersion = vuln.FixedVersion;
            }
            vulnerabilities.push(vulnEntry);
        }
    }
    return {
        imageId,
        vulnerabilities,
        scanDate: new Date(),
        ...counter.getCounts(),
    };
}
/**
 * Get Trivy version
 * @throws Error if Trivy is not installed or execution fails
 */
async function getTrivyVersion(logger) {
    try {
        const { stdout } = await execAsync('trivy --version', { timeout: constants_1.DEFAULT_TIMEOUTS.trivyVersionCheck });
        // Trivy version output format: "Version: X.Y.Z"
        const version = (0, scanner_common_1.parseVersion)(stdout, /Version:\s*([^\s\n]+)/);
        if (!version) {
            logger.debug({ stdout }, 'Could not parse Trivy version from output');
            return scanner_common_1.ScannerErrors.versionParseError('Trivy', 'trivy --version');
        }
        return (0, types_1.Success)(version);
    }
    catch (error) {
        const err = error;
        if (err?.code === 'ETIMEDOUT') {
            logger.error({ error }, 'Trivy version check timed out');
            return scanner_common_1.ScannerErrors.versionCheckTimeout('Trivy', 'trivy --version');
        }
        throw error;
    }
}
/**
 * Check if Trivy is installed and accessible
 */
async function checkTrivyAvailability(logger) {
    try {
        const versionResult = await getTrivyVersion(logger);
        if (!versionResult.ok) {
            return versionResult;
        }
        return (0, types_1.Success)(versionResult.value);
    }
    catch {
        return scanner_common_1.ScannerErrors.scannerNotInstalled('Trivy', 'https://aquasecurity.github.io/trivy/latest/getting-started/installation/');
    }
}
/**
 * Scan a Docker image using Trivy
 */
async function scanImageWithTrivy(imageId, logger) {
    // Validate imageId to prevent command injection
    if (!(0, scanner_common_1.validateImageId)(imageId)) {
        return scanner_common_1.ScannerErrors.invalidImageId(imageId);
    }
    // Check if Trivy is available
    const availabilityCheck = await checkTrivyAvailability(logger);
    if (!availabilityCheck.ok) {
        return (0, types_1.Failure)(availabilityCheck.error, availabilityCheck.guidance);
    }
    const trivyVersion = availabilityCheck.value;
    (0, scanner_common_1.logScanStart)(logger, 'Trivy', trivyVersion, imageId);
    try {
        // Run Trivy scan with JSON output using execFile to prevent command injection
        // --format json: output in JSON format
        // --quiet: suppress progress output
        // --timeout 5m: set timeout to 5 minutes
        const args = ['image', '--format', 'json', '--quiet', '--timeout', '5m', imageId];
        logger.debug({ args }, 'Executing Trivy command');
        const { stdout, stderr } = await execFileAsync('trivy', args, {
            maxBuffer: constants_1.LIMITS.MAX_SCAN_BUFFER, // 10MB buffer for large scan results
        });
        // Log any warnings from stderr
        if (stderr) {
            logger.debug({ stderr }, 'Trivy stderr output');
        }
        // Validate output size before parsing
        if (stdout.length === 0) {
            return scanner_common_1.ScannerErrors.emptyOutput('Trivy', imageId);
        }
        // Parse JSON output
        let trivyOutput;
        try {
            trivyOutput = JSON.parse(stdout);
        }
        catch (parseError) {
            return scanner_common_1.ScannerErrors.jsonParseError('Trivy', (0, errors_1.extractErrorMessage)(parseError), stdout.substring(0, 200));
        }
        // Parse the Trivy output into our format
        const scanResult = parseTrivyOutput(trivyOutput, imageId);
        (0, scanner_common_1.logScanComplete)(logger, 'Trivy', imageId, scanResult.totalVulnerabilities, scanResult.criticalCount, scanResult.highCount);
        return (0, types_1.Success)(scanResult);
    }
    catch (error) {
        const errorMessage = (0, errors_1.extractErrorMessage)(error);
        logger.error({ error: errorMessage, imageId }, 'Trivy scan failed');
        return scanner_common_1.ScannerErrors.scanExecutionError('Trivy', imageId, errorMessage);
    }
}
