// codetocloudmcp/out/toolsImplementations/debugInBrowser/browserTester.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.openAndDebugWebpage = openAndDebugWebpage;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const playwright_1 = require("playwright");
const params_1 = require("./params");
async function getPageForUrl(url) {
    const browser = await playwright_1.chromium.launch({
        headless: true,
        args: ['--start-maximized'] // Start browser maximized
    });
    try {
        const context = await browser.newContext({
            viewport: null
        });
        const page = await context.newPage();
        await page.goto(url);
        return page;
    }
    catch (error) {
        console.error(`Failed to open webpage: ${url}`, error);
        await browser.close();
        throw new Error(`Failed to open webpage: ${url}. Error: ${error instanceof Error ? error.message : String(error)}`);
    }
}
async function getPageContentSafely(page) {
    try {
        // Wait for the page to reach a stable state before getting content
        await page.waitForLoadState('domcontentloaded', { timeout: 5000 }).catch(() => { });
        return await page.content();
    }
    catch (error) {
        // If content retrieval fails (e.g., during navigation), return empty string
        return '';
    }
}
async function screenshotPageAndGetMarkdown(page, markdownFilePath, screenshotsFolder, actionIndex) {
    const screenshotPath = path.join(screenshotsFolder, `screenshot-${actionIndex + 1}.png`);
    await page.screenshot({ path: screenshotPath });
    return `![Screenshot ${actionIndex + 1}](${path.relative(path.dirname(markdownFilePath), screenshotPath)})\n`;
}
function addToMarkdownResponse(response, progressMarkdownFile, payload) {
    response.push(payload);
    // write markdownResponsePieces into progressMarkdownFile but not before deleting it first
    fs.writeFileSync(progressMarkdownFile, response.join('\n'));
}
async function openAndDebugWebpage(workspaceFolder, url, actions, screenshotsFolder, progressMarkdownFile) {
    const page = await getPageForUrl(url);
    let markdownResponsePieces = [];
    let pageLastSeenHtml = undefined;
    for (var i = 0; i < actions.length; i++) {
        const action = actions[i];
        addToMarkdownResponse(markdownResponsePieces, progressMarkdownFile, `# Action ${i + 1}: ${action.actionType} on ${action.selector}${action.actionValue ? ` - ${action.actionValue}` : ""}\n`);
        const screenshotMarkdown = await screenshotPageAndGetMarkdown(page, progressMarkdownFile, screenshotsFolder, i);
        pageLastSeenHtml = await getPageContentSafely(page);
        addToMarkdownResponse(markdownResponsePieces, progressMarkdownFile, screenshotMarkdown);
        try {
            switch (action.actionType) {
                case params_1.DebugActionType.Click:
                    await page.click(action.selector);
                    break;
                case params_1.DebugActionType.Input:
                    await page.fill(action.selector, action.actionValue || "");
                    break;
                case params_1.DebugActionType.Hover:
                    await page.hover(action.selector);
                    break;
                case params_1.DebugActionType.Wait:
                    await page.waitForTimeout(Number(action.actionValue) || 1000);
                    break;
                case params_1.DebugActionType.ValidateElementVisible:
                    const isVisible = await page.isVisible(action.selector);
                    if (!isVisible) {
                        throw new Error(`Element not visible: ${action.selector}`);
                    }
                    break;
                case params_1.DebugActionType.ValidateElementText:
                    const elementText = await page.textContent(action.selector);
                    if (elementText !== action.actionValue) {
                        throw new Error(`Element text does not match: ${action.selector}`);
                    }
                    break;
                case params_1.DebugActionType.UploadFile:
                    const filePath = action.actionValue;
                    if (!filePath || !fs.existsSync(filePath)) {
                        throw new Error(`File does not exist: ${filePath}`);
                    }
                    if (!page.isVisible(action.selector)) {
                        throw new Error(`Selector not visible for file selection: ${action.selector}`);
                    }
                    const fileChooserPromise = page.waitForEvent('filechooser');
                    await page.click(action.selector);
                    const fileChooser = await fileChooserPromise;
                    await fileChooser.setFiles(filePath);
                    /* For true drag and drop
                    const buffer = fs.readFileSync(filePath);
                    const fileExtension = path.extname(filePath).substring(1); // Get file extension without dot
          
                    const dataTransfer = await page.evaluateHandle(
                      ({ bufferData, localFileName, localFileType }) => {
                        const dt = new DataTransfer();
                        const blobData = new Blob([bufferData], { type: localFileType });
                        const file = new File([blobData], localFileName, { type: localFileType });
                        dt.items.add(file);
                        return dt;
                      },
                      { bufferData: buffer, localFileName: path.basename(filePath), localFileType: `image/${fileExtension}` }
                    );
          
                    await page.dispatchEvent(action.selector, 'drop', { dataTransfer });
                    */
                    // For file upload using file chooser
                    break;
                default:
                    throw new Error(`Unknown action type: ${action.actionType}`);
            }
            addToMarkdownResponse(markdownResponsePieces, progressMarkdownFile, `**Success:** Action ${i + 1} executed successfully.\n`);
            await page.waitForTimeout(1000);
        }
        catch (error) {
            const errorMessage = `Error executing action ${i + 1} (${action.actionType}): ${error instanceof Error ? error.message : String(error)}`;
            addToMarkdownResponse(markdownResponsePieces, progressMarkdownFile, `**Error:** ${errorMessage}\n`);
            console.error(errorMessage);
            break;
        }
    }
    // if page is still open
    if (!page.isClosed()) {
        pageLastSeenHtml = await getPageContentSafely(page);
        await page.close();
        await page.context()?.browser()?.close();
    }
    // write last seen html as an html file in screenshots folder
    const lastSeenHtmlPath = path.join(screenshotsFolder, 'lastSeen.html');
    fs.writeFileSync(lastSeenHtmlPath, pageLastSeenHtml || "");
    markdownResponsePieces.push(`# Last Seen HTML File: \`${lastSeenHtmlPath}\`. Please open this file to observe what is in the browser contents.`);
    return markdownResponsePieces.join('\n');
}
