// codetocloudmcp/out/mcpServers/toolsImplementations/deployPipeline/prompts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeployPipelineSetupStepProvider = void 0;
const models_1 = require("../../common/models");
const planStep_1 = require("../../common/planGeneration/planStep");
const toolType_1 = require("../shared/toolType");
class DeployPipelineSetupStepProvider {
    pipelinePlatform;
    deploymentTool;
    deployOption;
    stateManager;
    pipelineSetupStepsAndRules;
    addPrerequisiteChecksPrompts() {
        const prerequisiteCheckStepsAndRules = new planStep_1.PlanStep('Prerequisite Checks', planStep_1.PlanStepBulletType.Bulleted);
        if (this.deployOption === models_1.DeployOption.deployOnly) {
            prerequisiteCheckStepsAndRules.addStep('When user confirms that Azure resources are ready for deployment, you need to know at least two things: the resource groups (with environments, e.g., dev, prod) and the hosting service TYPE(e.g., AKS, Azure Container Apps, App Service).');
        }
        if (this.deployOption === models_1.DeployOption.provisionAndDeploy) {
            prerequisiteCheckStepsAndRules.addStep(`When user wants to include provisioning, check if there are available first run ${this.stateManager.getToolName(toolType_1.ToolType.GetAzureResourceRules)} to create infra-provisioning files.`);
        }
        if (this.deploymentTool === models_1.DeploymentTool.Azd) {
            prerequisiteCheckStepsAndRules.addStep(`AZD IaC check: if Bicep is using resource group scope, resource group should be created in advance.`);
        }
        this.pipelineSetupStepsAndRules.addStep(prerequisiteCheckStepsAndRules);
    }
    addPipelineFilePrompts() {
        const pipelineFileGenerationStepsAndRules = new planStep_1.PlanStep('Pipeline File Generation Guidance', planStep_1.PlanStepBulletType.Bulleted);
        pipelineFileGenerationStepsAndRules.addSteps([
            `Generate a CI/CD pipeline file using ${this.deploymentTool} for ${this.pipelinePlatform}.`,
            `The CI part should include build (and test if applicable). CI should NOT include Azure-related steps, for example, pushing images to ACR.`,
            `The CD part should include multi-environment deployment steps. Images, if applicable, should be pushed to ACR in different environments.`,
            `**'pull_request' should not start CD(deployment).**`,
            `Use different environments according to user-provided Azure resources. Default to use 'dev', 'staging', and 'production' IF NOT SPECIFIED by user. *If user states specific environments, use those.*`,
            `Pay attention to the dependency order among build and multi-environment deployments.`
        ]);
        if (this.pipelinePlatform === models_1.PipelinePlatform.GithubActions) {
            pipelineFileGenerationStepsAndRules.addSteps([
                `Use User-assigned Managed Identity with OIDC for login to Azure in the pipeline.`,
                `Use azure/login@v2 action to set up OIDC authentication in Github Actions. Add 'id-token: write' permission.`,
                `Use variables and secrets for specific values. Do NOT hardcode any values like service names.`
            ]);
        }
        else if (this.pipelinePlatform === models_1.PipelinePlatform.AzureDevOps) {
            pipelineFileGenerationStepsAndRules.addStep(`Use Service Principal(app registration) with workflow identity federation to login to Azure in the pipeline.`);
        }
        if (this.deploymentTool === models_1.DeploymentTool.Azd) {
            pipelineFileGenerationStepsAndRules.addStep(`Use 'azd deploy --no-prompt' to skip provisioning in CD pipeline.`);
        }
        if (this.deployOption === models_1.DeployOption.provisionAndDeploy) {
            pipelineFileGenerationStepsAndRules.addStep(`Add a SEPARATE infra-deploy pipeline file to provision Azure resources.`);
        }
        pipelineFileGenerationStepsAndRules.addStep(`**You should ONLY care about deployment-related variables like hosting services' names. Ignore application-level variables like connection strings.**`);
        this.pipelineSetupStepsAndRules.addStep(pipelineFileGenerationStepsAndRules);
    }
    addAzureAuthConfigurationPrompts() {
        // Copilot knows best practices for Azure/Github integration, such as OIDC and RBAC, as well as ADO pipeline set-ups like service connections.
        // No detailed instructions needed yet.
        const azureAuthConfigStepsAndRules = new planStep_1.PlanStep('Azure Authentication Configuration Guidance', planStep_1.PlanStepBulletType.Bulleted);
        if (this.pipelinePlatform === models_1.PipelinePlatform.GithubActions) {
            azureAuthConfigStepsAndRules.addSteps([
                `Create a new **User-assigned Managed Identity** in a SEPARATE resource group.`,
                `This Managed Identity works for the pipeline. DO NOT CONFUSE it with any existing Managed Identity used by the application.`,
                `Set up Managed Identity with federated credentials and RBAC (e.g. contributor to resource groups and AcrPull) to authenticate (azure login with OIDC), push images and deploy applications to Azure in the pipeline for different environments.`,
                `Federated credentials should be set with proper subject **to different environments** (instead of branch), issuer, and audience. Description is NOT a valid property.`
            ]);
        }
        else {
            azureAuthConfigStepsAndRules.addSteps([
                `Set up Service Connection in Azure DevOps using app registration with workflow identity federation.`,
                `Use RBAC (e.g. contributor to resource groups and AcrPull) to the app registration to authenticate, push images and deploy applications to Azure in the pipeline for different environments.`
            ]);
        }
        this.pipelineSetupStepsAndRules.addStep(azureAuthConfigStepsAndRules);
    }
    addSetUpMethodInstructions() {
        const toolSetupInstructions = new planStep_1.PlanStep('How you do configurations', planStep_1.PlanStepBulletType.Bulleted);
        if (this.stateManager.getCustomerChannelValue() === models_1.CustomerChannel.COPILOTAGENT) {
            toolSetupInstructions.addStep(`Copilot Agent do not have permission to configure Azure authentication for the pipeline. Please inform the user to follow the instructions below.`);
        }
        else {
            toolSetupInstructions.addStep(`Configure Azure authentication and environment setup for the pipeline as per the instructions below.`);
        }
        toolSetupInstructions.addStep(`You should use a .azure/pipeline-setup.md file to outline the steps. The file should contain the steps and explain the tasks to be done by the user.`);
        if (this.pipelinePlatform === models_1.PipelinePlatform.GithubActions) {
            toolSetupInstructions.addSteps([
                `Create a setup-azure-auth-for-pipeline.sh or setup-azure-auth-for-pipeline.ps1 script to automate the auth configuration. Include detailed commands or UI instructions according to the pipeline platform.`
            ]);
        }
        this.pipelineSetupStepsAndRules.addStep(toolSetupInstructions);
    }
    addEnvironmentSetupPrompts() {
        const environmentSetupStepsAndRules = new planStep_1.PlanStep('Pipeline Environment Setup Guidance', planStep_1.PlanStepBulletType.Bulleted);
        if (this.pipelinePlatform === models_1.PipelinePlatform.GithubActions) {
            environmentSetupStepsAndRules.addSteps([
                `Create Github environments and set up approval checks in ALL environments.`,
                `Configure the Github Actions variables and secrets needed for the deployment pipeline for *different environments*.`,
                `Refer to the managed identity created in the 'Azure Authentication Configuration Guidance' section for OIDC setup.`,
                `You can use Github CLI and Github CLI or guide user to use UI to finish the setup.`
            ]);
        }
        else {
            environmentSetupStepsAndRules.addSteps([
                `Configure the Azure DevOps pipeline variables and service connections needed for the deployment pipeline.`,
                `Set up the ADO environment and set up deployment approvals and checks.`
            ]);
        }
        if (this.deploymentTool === models_1.DeploymentTool.Azd) {
            environmentSetupStepsAndRules.addStep(`AZD IaC check: if Bicep is using resource group scope, AZURE_RESOURCE_GROUP variable should be set to the environment.`);
        }
        this.pipelineSetupStepsAndRules.addStep(environmentSetupStepsAndRules);
    }
    constructor(stateManager, pipelinePlatform, deploymentTool, deployOption) {
        this.stateManager = stateManager;
        this.pipelinePlatform = pipelinePlatform;
        this.deploymentTool = deploymentTool;
        this.deployOption = deployOption ?? models_1.DeployOption.deployOnly;
        this.pipelineSetupStepsAndRules = new planStep_1.PlanStep('Follow ALL of the following rules to provide CI/CD pipeline files and configuration setup instructions', planStep_1.PlanStepBulletType.Bulleted);
    }
    getPipelineSetupPrompt() {
        this.addPrerequisiteChecksPrompts();
        this.addPipelineFilePrompts();
        this.addSetUpMethodInstructions();
        this.addAzureAuthConfigurationPrompts();
        this.addEnvironmentSetupPrompts();
        return this.pipelineSetupStepsAndRules.renderIntoMarkdown();
    }
}
exports.DeployPipelineSetupStepProvider = DeployPipelineSetupStepProvider;
