// containerization-assist-mcp/dist/src/tools/analyze-repo/schema.js
/**
 * Schema definition for analyze-repo tool
 */
import { z } from 'zod';
import { repositoryPath, analysisOptions } from '../shared/schemas.js';
const moduleInfo = z.object({
    name: z.string().describe('The name of the module'),
    modulePath: z
        .string()
        .describe('Absolute path to module root. Paths are automatically normalized to forward slashes on all platforms.'),
    dockerfilePath: z.string().optional().describe('Path where the Dockerfile should be generated'),
    language: z
        .enum(['java', 'dotnet', 'javascript', 'typescript', 'python', 'rust', 'go', 'other'])
        .optional()
        .describe('Primary programming language used in the module'),
    frameworks: z
        .array(z.object({
        name: z
            .string()
            .describe('Frameworks used in project like Java Spring, SpringBoot, Hibernate etc.'),
        version: z.string().optional(),
    }))
        .optional(),
    buildSystems: z
        .array(z.object({
        type: z.string(),
        languageVersion: z.string().optional(),
    }))
        .optional()
        .describe('All detected build systems with their language versions'),
    dependencies: z
        .array(z.string())
        .optional()
        .describe('List of module dependencies including database drivers and system libraries'),
    ports: z.array(z.number()).optional(),
    entryPoint: z.string().optional(),
});
export const analyzeRepoSchema = z.object({
    repositoryPath,
    ...analysisOptions,
    modules: z
        .array(moduleInfo)
        .optional()
        .describe('Optional pre-analyzed modules. If not provided, AI will analyze the repository.'),
});
