// codetocloudmcp/out/mcpServers/common/telemetry/telemetryUtil.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResultCode = void 0;
exports.isInternal = isInternal;
exports.getConfig = getConfig;
const fs = __importStar(require("fs"));
const helpers_1 = require("../helpers");
var ResultCode;
(function (ResultCode) {
    ResultCode["PreInvoked"] = "PreInvoked";
    ResultCode["Invoked"] = "Invoked";
    ResultCode["InProgress"] = "InProgress";
    ResultCode["Success"] = "Success";
    ResultCode["Failure"] = "Failure";
    ResultCode["Cancelled"] = "Cancelled";
})(ResultCode || (exports.ResultCode = ResultCode = {}));
;
function isInternal() {
    const userDomain = process.env.USERDNSDOMAIN;
    return typeof userDomain === 'string' && userDomain.endsWith('.corp.microsoft.com');
}
function getConfig(key, defaultValue) {
    function convertValue(value) {
        if (typeof defaultValue === 'boolean') {
            if (typeof value === 'string') {
                return (value.toLowerCase() === 'true');
            }
            return Boolean(value);
        }
        if (typeof defaultValue === 'number') {
            const num = Number(value);
            return (isNaN(num) ? defaultValue : num);
        }
        return value;
    }
    if (process.env[key] !== undefined) {
        return convertValue(process.env[key]);
    }
    const configPath = (0, helpers_1.findFileIteratively)(__dirname, 'config.json');
    if (configPath === undefined) {
        return defaultValue;
    }
    const config = JSON.parse(fs.readFileSync(configPath, 'utf8'));
    if (config[key] === undefined) {
        return defaultValue;
    }
    return convertValue(config[key]);
}
