// codetocloudmcp/out/mcpServers/toolsImplementations/getPlan/executionSteps/azdExecutionSteps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzdExecutionStepGenerator = void 0;
const models_1 = require("../../../common/models");
const planStep_1 = require("../../../common/planGeneration/planStep");
const toolType_1 = require("../../shared/toolType");
const baseExecutionSteps_1 = require("./baseExecutionSteps");
class AzdExecutionStepGenerator extends baseExecutionSteps_1.BaseExecutionSteps {
    constructor(parameters, stateManager) {
        super(parameters, stateManager);
    }
    // For AZD, there is a dependency chain: iac files -> azd env -> azd provision preview -> azd provision/azd up
    // the env setup is supposed to take place after iac files are generated
    generateExecutionSteps() {
        this.requiredToolsInPlan = [];
        var planSteps = new planStep_1.PlanStep("Execution Steps", planStep_1.PlanStepBulletType.Numbered);
        if (this.shouldIncludeContainerization()) {
            planSteps.addStep(this.getContainerizationStep());
        }
        if (this.shouldIncludeProvision()) {
            planSteps.addStep(this.getCreateIaCFilesStep());
        }
        planSteps.addStep(this.getEnvironmentSetupStep());
        if (this.shouldIncludeDeployment()) {
            planSteps.addStep(this.getDeploymentStep());
        }
        else if (this.shouldIncludeProvision()) {
            planSteps.addStep(this.getProvisionStep());
        }
        planSteps.addStep(this.getSummarizationStep());
        return {
            planSteps: planSteps,
            requiredTools: this.requiredToolsInPlan
        };
    }
    getCreateIaCFilesStep() {
        let createIaCSteps = new planStep_1.PlanStep(`Create Azure Infrastructure Files for AZD`, planStep_1.PlanStepBulletType.Numbered);
        var expectedFiles = this.iaCPromptProvider?.getOutputFileName();
        if (this.targetAppServices.some(service => service.azureComputeHost === models_1.TargetAppServiceType.AzureAppServiceManagedInstance) && this.iacOptions === models_1.IacType.Bicep) {
            expectedFiles += ", storage.bicep, setupstorage.ps1(.sh)";
        }
        createIaCSteps.addStep(`Provisioning tool: AZD. Expected files: ${expectedFiles}`);
        createIaCSteps.addStep(this.getRegionsPrompt(this.targetAppServices.some(service => service.azureComputeHost === models_1.TargetAppServiceType.AzureKubernetesService)));
        var filesExistSteps = new planStep_1.PlanStep(`If expected files exist, please follow the below steps`, planStep_1.PlanStepBulletType.Bulleted);
        filesExistSteps.addSteps([
            `(Fill in with existing file paths) Existing file paths: <filePaths>`,
            `Check if the Azure resources in the existing files match what we required in the plan. If not, add steps to update the existing files with missing Azure resource.`,
            `Generating Files: No`
        ]);
        var filesNotExistSteps = new planStep_1.PlanStep(`If expected files do not exist, please follow the below steps`, planStep_1.PlanStepBulletType.Bulleted);
        const getResourceRulesToolName = this.stateManager.getToolName(toolType_1.ToolType.GetAzureResourceRules);
        this.requiredToolsInPlan.push(getResourceRulesToolName);
        filesNotExistSteps.addSteps([
            `Generate the missing expected files, call tool \`${getResourceRulesToolName}\` to get the rules for file generation.`,
            `Call \`get_errors\` on the generated files and iterate on any errors.`,
            `Generating Files: ${this.iaCPromptProvider?.getOutputFileName()}`
        ]);
        createIaCSteps.addSteps([filesExistSteps, filesNotExistSteps]);
        createIaCSteps.addStep(`Fix the ${this.iacOptions} files and retry if there's any error in the provisioning files. Use \`get_errors\` tool if available to help look for errors.`);
        return createIaCSteps;
    }
    getProvisionStep() {
        if (!this.shouldIncludeProvision()) {
            return '';
        }
        const provisionCommand = this.serviceProviders.map(provider => provider.getProvisionPrompts(this.deployOption, models_1.DeploymentTool.Azd)).flat();
        let provisionSteps = new planStep_1.PlanStep(`Provisioning with AZD`, planStep_1.PlanStepBulletType.Numbered);
        provisionSteps.addStep(Array.from(new Set(provisionCommand)).join(" "));
        return provisionSteps;
    }
    getEnvironmentSetupStep() {
        var envSetupStep = new planStep_1.PlanStep(`Env setup for ${this.deploymentTool} {COPY the steps for 'Env set up' directly.}`, planStep_1.PlanStepBulletType.Numbered);
        envSetupStep.addSteps([`Install AZ CLI and AZD if not installed.`,
            `Run 'azd env new <envName> --no-prompt' to create a new environment, generating a name for the user. If there's existing environment, check if the environment matches the requirements.`,
            `Review ${this.iacOptions === models_1.IacType.Bicep ? "infra/main.bicep and infra/main.parameters.bicep" : "infra/main.tf, infra/variables.tf, infra/terraform.tfvars.json"} for environment variables. Set these for the user using azd env set <key> <value>. Use default values (from AZ CLI) if not provided by the user. Subscription ID is always required.`,
            `Set the subscription: ${this.subscriptionId ? `azd env set AZURE_SUBSCRIPTION_ID ${this.subscriptionId}` : "Use default subscription"}`,
            `Set AZURE_RESOURCE_GROUP environment variable. The region of the resource group does not have to match its inner resources.${this.iacOptions === models_1.IacType.Bicep ? ` Check if Bicep is deployed with resource group scope. If so, CREATE a resource group with AZ CLI.` : ''}`
        ]);
        return envSetupStep;
    }
    // For AZD, this step handles the deployment by azd up for E2E scenario, and handles the actual provisioning for provision-only scenario.
    getDeploymentStep() {
        if (!this.shouldIncludeDeployment()) {
            return '';
        }
        let deploymentSteps = new planStep_1.PlanStep('Deployment', planStep_1.PlanStepBulletType.Numbered);
        // provision preview can help copilot isolate iac errors from build and deployment errors
        const provisionPreviewCommand = this.serviceProviders.map(provider => provider.getProvisionPrompts(this.deployOption, models_1.DeploymentTool.Azd)).flat();
        deploymentSteps.addStep(Array.from(new Set(provisionPreviewCommand)).join(" "));
        let deploySet = {};
        this.serviceProviders.forEach(provider => {
            const providerKey = provider.constructor.name;
            if (deploySet[providerKey]) {
                return;
            }
            deploySet[providerKey] = true;
            deploymentSteps.addStep(provider.getDeploymentStep(this.deployOption, models_1.DeploymentTool.Azd));
        });
        deploymentSteps.addStep(this.getValidationStep());
        return deploymentSteps;
    }
    getValidationStep() {
        if (!this.shouldIncludeValidation()) {
            return undefined;
        }
        let validationStep = new planStep_1.PlanStep("Deployment Validation", planStep_1.PlanStepBulletType.Numbered);
        this.serviceProviders.forEach(provider => {
            validationStep = provider.getValidationStep(this.stateManager);
        });
        return validationStep;
    }
}
exports.AzdExecutionStepGenerator = AzdExecutionStepGenerator;
