// codetocloudmcp/out/mcpServers/toolsImplementations/getPlan/executionSteps/baseExecutionSteps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseExecutionSteps = void 0;
const models_1 = require("../../../common/models");
const planStep_1 = require("../../../common/planGeneration/planStep");
const toolType_1 = require("../../shared/toolType");
const IaCPromptProviderFactory_1 = require("./IaCPromptProviders/IaCPromptProviderFactory");
const ServiceStepProviderFactory_1 = require("./serviceStepProviders/ServiceStepProviderFactory");
class BaseExecutionSteps {
    deployOption;
    deploymentTool;
    targetAppServices;
    subscriptionId;
    resourceGroupName;
    iacOptions;
    serviceProviders;
    iaCPromptProvider;
    stateManager;
    requiredToolsInPlan = [];
    projectName = "";
    constructor(parameters, stateManager) {
        this.projectName = parameters.projectName;
        this.deployOption = parameters.deployOption;
        this.targetAppServices = parameters.services;
        this.subscriptionId = parameters.subscriptionId;
        this.resourceGroupName = parameters.resourceGroupName;
        this.deploymentTool = parameters.provisioningTool;
        this.iacOptions = parameters.iacOptions || '';
        this.serviceProviders = ServiceStepProviderFactory_1.ServiceStepProviderFactory.createProviders(parameters.projectName, parameters.services, parameters.deployOption, this.iacOptions);
        this.iaCPromptProvider = IaCPromptProviderFactory_1.IaCPromptProviderFactory.createIaCProviders(this.iacOptions);
        this.stateManager = stateManager;
    }
    // Template method with enhanced logic
    // azd execution step is implemented in subclass
    generateExecutionSteps() {
        this.requiredToolsInPlan = [];
        var planSteps = new planStep_1.PlanStep("Execution Steps", planStep_1.PlanStepBulletType.Numbered);
        if (this.shouldIncludeContainerization()) {
            planSteps.addStep(this.getContainerizationStep());
        }
        planSteps.addStep(this.getEnvironmentSetupStep());
        if (this.shouldIncludeProvision()) {
            planSteps.addStep(this.getProvisionStep());
        }
        if (this.deployOption === models_1.DeployOption.deployOnly) {
            planSteps.addStep(this.getExistingResourceStep());
        }
        if (this.shouldIncludeDeployment()) {
            planSteps.addStep(this.getDeploymentStep());
        }
        planSteps.addStep(this.getSummarizationStep());
        return {
            planSteps: planSteps,
            requiredTools: this.requiredToolsInPlan
        };
    }
    // Common methods with default implementations
    shouldIncludeProvision() {
        return this.deployOption !== models_1.DeployOption.deployOnly;
    }
    shouldIncludeContainerization() {
        return this.deployOption !== models_1.DeployOption.provisionOnly &&
            (this.targetAppServices.some(i => i.azureComputeHost === models_1.TargetAppServiceType.AzureKubernetesService || i.azureComputeHost === models_1.TargetAppServiceType.AzureContainerApp));
    }
    shouldIncludeDeployment() {
        return this.deployOption !== models_1.DeployOption.provisionOnly;
    }
    shouldIncludeValidation() {
        return this.deployOption !== models_1.DeployOption.provisionOnly;
    }
    getExistingResourceStep() {
        if (this.deployOption !== models_1.DeployOption.deployOnly) {
            return undefined;
        }
        let steps = new planStep_1.PlanStep("Check Azure resources existence", planStep_1.PlanStepBulletType.Numbered);
        this.serviceProviders.forEach(provider => {
            steps.addSteps(provider.getExistingResourcesSteps());
        });
        if (this.shouldIncludeContainerization() && !this.targetAppServices.some(service => service.dependencies.map(dep => dep.serviceType).includes(models_1.AzureServiceNames.AzureContainerRegistry))) {
            const acrCheckingStep = new planStep_1.PlanStep("Azure Container Registry", planStep_1.PlanStepBulletType.Bulleted);
            acrCheckingStep.addStep("login server: <>. Check with 'az acr show -o json'");
            steps.addStep(acrCheckingStep);
        }
        let creatingMissingResourcesStep = new planStep_1.PlanStep("Create missing resources", planStep_1.PlanStepBulletType.Bulleted);
        creatingMissingResourcesStep.addSteps([
            "If any resource is missing, ask user to provide the resource id or create a new one, then get the resource information with Az CLI command",
            "If user want to create new resources, generate a script to do so using Azure CLI command. Run the script and confirms all resources are ready."
        ]);
        steps.addStep(creatingMissingResourcesStep);
        return steps;
    }
    getContainerizationStep() {
        if (this.shouldIncludeContainerization()) {
            var containerizationSteps = new planStep_1.PlanStep("Containerization", planStep_1.PlanStepBulletType.Bulleted);
            containerizationSteps.addSteps([
                "If Dockerfiles exist, list their paths."
            ]);
            if (this.stateManager.toolTypeExists(toolType_1.ToolType.AnalyzeRepository) && this.stateManager.toolTypeExists(toolType_1.ToolType.PlanGenerateDockerfile)) {
                const analyzeRepositoryToolName = this.stateManager.getToolName(toolType_1.ToolType.AnalyzeRepository);
                const planGenerateDockerfileToolName = this.stateManager.getToolName(toolType_1.ToolType.PlanGenerateDockerfile);
                if (this.stateManager.toolTypeExists(toolType_1.ToolType.RunSubagent)) {
                    const subagentToolName = this.stateManager.getToolName(toolType_1.ToolType.RunSubagent);
                    this.requiredToolsInPlan.push(`${subagentToolName} for containerization`);
                    containerizationSteps.addSteps([
                        `If missing: call tool #${subagentToolName} with the following prompt:`,
                        `<prompt>Analyze my repo using tool ${analyzeRepositoryToolName}. Then, generate Dockerfiles based on the analysis with ${planGenerateDockerfileToolName}. Ensure the dockerfiles can be built. Return each dockerfile created, and its required docker build context path.</prompt>`,
                        `Agent must create Dockerfile first before deployment!`
                    ]);
                }
                else {
                    this.requiredToolsInPlan.push(analyzeRepositoryToolName);
                    this.requiredToolsInPlan.push(planGenerateDockerfileToolName);
                    containerizationSteps.addSteps([
                        `If missing: Ensure the repo is analyzed using ${analyzeRepositoryToolName}.`,
                        `Based on that response, use ${planGenerateDockerfileToolName} to generate the required Dockerfiles. Agent must create Dockerfile first before deployment!`
                    ]);
                }
            }
            else {
                containerizationSteps.addStep("If missing: Generate required Dockerfiles and build scripts, as well as compose.yaml if required. Agent must create Dockerfile first before deployment!");
            }
            containerizationSteps.addStep("Output: Docker artifacts");
            return containerizationSteps;
        }
    }
    getSummarizationStep() {
        const summarizeToolName = this.stateManager.getToolName(toolType_1.ToolType.SummarizeResult);
        this.requiredToolsInPlan.push(summarizeToolName);
        let summarizeSteps = new planStep_1.PlanStep("Summarize Result", planStep_1.PlanStepBulletType.Numbered);
        var summary = `Use \`${summarizeToolName}\` tool to summarize the deployment result.
    2. Generating files: .azure/summary.${this.stateManager.getCopilotMarkdownExtension()}`;
        summarizeSteps.addStep(summary);
        return summarizeSteps;
    }
    hasAKS() {
        return this.targetAppServices.some(service => service.azureComputeHost === models_1.TargetAppServiceType.AzureKubernetesService);
    }
    getRegionsPrompt(isAKS) {
        const getRegionsToolName = this.stateManager.getToolName(toolType_1.ToolType.GetRegionsForDeployment);
        let prompt = '';
        if (this.stateManager.toolTypeExists(toolType_1.ToolType.RunSubagent)) {
            const subagentToolName = this.stateManager.getToolName(toolType_1.ToolType.RunSubagent);
            this.requiredToolsInPlan.push(`${subagentToolName} for region check`);
            prompt = `Use tool #${subagentToolName} with the following prompt. Populate <REQUIRED_RESOURCES> based on your context. <prompt>Grab current subscriptionID using 'az account show --query id -o tsv', call tool \`${getRegionsToolName}\` to get available region and SKUs for all needed Azure resource types. Check for <REQUIRED_RESOURCES>. Return with ONLY the available region and SKUs available in that region.</prompt>`;
        }
        else {
            this.requiredToolsInPlan.push(getRegionsToolName);
            prompt = `Grab current subscriptionID, call tool \`${getRegionsToolName}\` to get available region and SKUs for all needed Azure resource types.`;
            prompt += this.serviceProviders.map(provider => provider.getRegionCheckPrompts()).join(" ");
        }
        return prompt;
    }
}
exports.BaseExecutionSteps = BaseExecutionSteps;
