// codetocloudmcp/out/mcpServers/toolsImplementations/getAzureResourceRules/params.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RuleSupportedResourceTypeToNameMapping = exports.RulesSupportedResourceType = void 0;
exports.convertResourceTypeToNames = convertResourceTypeToNames;
var RulesSupportedResourceType;
(function (RulesSupportedResourceType) {
    RulesSupportedResourceType["AzureAppService"] = "appservice";
    RulesSupportedResourceType["AzureAppServiceManagedInstance"] = "appservicemanagedinstance";
    RulesSupportedResourceType["AzureContainerApp"] = "containerapp";
    RulesSupportedResourceType["AzureFunctionApp"] = "function";
    RulesSupportedResourceType["AzureKubernetesService"] = "aks";
    RulesSupportedResourceType["PostgreSqlDatabase"] = "azuredatabaseforpostgresql";
    RulesSupportedResourceType["MySqlServer"] = "azuredatabaseformysql";
    RulesSupportedResourceType["CosmosDB"] = "azurecosmosdb";
    RulesSupportedResourceType["StorageAccount"] = "azurestorageaccount";
    RulesSupportedResourceType["KeyVault"] = "azurekeyvault";
    RulesSupportedResourceType["ContainerRegistry"] = "azurecontainerregistry";
})(RulesSupportedResourceType || (exports.RulesSupportedResourceType = RulesSupportedResourceType = {}));
exports.RuleSupportedResourceTypeToNameMapping = {
    'Microsoft.Web/sites': RulesSupportedResourceType.AzureAppService,
    'Microsoft.App/containerApps': RulesSupportedResourceType.AzureContainerApp,
    'Microsoft.ContainerService/managedClusters': RulesSupportedResourceType.AzureKubernetesService,
    'Microsoft.DBforPostgreSQL/flexibleServers': RulesSupportedResourceType.PostgreSqlDatabase,
    'Microsoft.DBforMySQL/flexibleServers': RulesSupportedResourceType.MySqlServer,
    'Microsoft.DocumentDB/databaseAccounts': RulesSupportedResourceType.CosmosDB,
    'Microsoft.Storage/storageAccounts': RulesSupportedResourceType.StorageAccount,
    'Microsoft.KeyVault/vaults': RulesSupportedResourceType.KeyVault,
    'Microsoft.ContainerRegistry/registries': RulesSupportedResourceType.ContainerRegistry
};
function convertResourceTypeToNames(i) {
    if (Object.values(RulesSupportedResourceType).includes(i)) {
        return i;
    }
    for (const [resourceType, name] of Object.entries(exports.RuleSupportedResourceTypeToNameMapping)) {
        const provider = resourceType.split('/')[0];
        if (resourceType.toLocaleLowerCase() === i.toLocaleLowerCase() || provider.toLocaleLowerCase() === i.toLocaleLowerCase()) {
            return name;
        }
    }
    return i;
}
