// codetocloudmcp/out/mcpServers/toolsImplementations/getAzureResourceRules/prompts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.populateLLMResponse = populateLLMResponse;
const models_1 = require("../../common/models");
const planStep_1 = require("../../common/planGeneration/planStep");
const toolType_1 = require("../shared/toolType");
const params_1 = require("./params");
// General AZD rules
function populateAZDPrompts(iacType, resourceTypes, llmResponse) {
    llmResponse.push(`- Ensure an User-Assigned Managed Identity (UAMI) exists.`);
    llmResponse.push(`- Resource Group resource (if exists) must have tag "azd-env-name" = environmentName. Apply this tag to resource group resource ONLY.`);
    llmResponse.push(`- Expected parameters in ${iacType} parameters: environmentName='\${AZURE_ENV_NAME}', location='\${AZURE_LOCATION}'. For subscription-scoped deployments, set resourceGroupName='rg-\${AZURE_ENV_NAME}'. For resource group scoped deployments, set env var 'AZURE_RESOURCE_GROUP'. If resource group is not created, you must create one before provisioning.`);
    llmResponse.push(`- All container apps, app services, function apps, static web apps (and nothing else) must have tag "azd-service-name" matching the service name in azure.yaml.`);
    llmResponse.push(`- Add code comment at the top of the file - "# yaml-language-server: $schema=https://raw.githubusercontent.com/Azure/azure-dev/refs/heads/main/schemas/v1.0/azure.yaml.json"`);
    if (resourceTypes.includes(params_1.RulesSupportedResourceType.AzureContainerApp)) {
        llmResponse.push(`For Container Apps, azure.yaml example:
name: yourApp-aca
metadata:
  template: yourApp-aca@0.0.1-beta
services:
  api:
    project: ./src/api    # Path to the service source code directory
    language: js
    host: containerapp
    docker:
      path: ./Dockerfile  # path to Dockerfile, relative to project path. Dockerfile must exist in ./src/api/Dockerfile in this case.
      context: ../        # docker context to execute docker build, relative to project path, it will be './src/api/../' in this case. Check the docker file content to set the context properly.
  web:
    project: ./src/web
    language: js
    host: containerapp
    docker:
      remoteBuild: true
`);
    }
    const outputsFileName = iacType === 'bicep' ? 'main.bicep' : 'outputs.tf';
    llmResponse.push(`- Expected output in ${outputsFileName}: RESOURCE_GROUP_ID.`);
    if (resourceTypes.includes(params_1.RulesSupportedResourceType.AzureContainerApp)) {
        llmResponse.push(`- Expected output in ${outputsFileName}: AZURE_CONTAINER_REGISTRY_ENDPOINT to be the loginServer of the container registry, with format 'acrname.azurecr.io'. Don't add 'http' prefix.`);
    }
    return;
}
// General AZ CLI rules
function populateAZCLIPrompts(llmResponse, resourceTypes, iacType) {
    llmResponse.push(`- Use .ps1 extension for PowerShell scripts and .sh extension for Bash scripts.`);
    llmResponse.push(`- Ensure all the steps have been executed successfully. Fix the script and RERUN if ANY step fails. Use the SAME resourceToken params as the failed run when rerunning the script.`);
    if (!iacType) {
        populateAZCLIScriptPrompts(llmResponse, resourceTypes);
    }
    if (resourceTypes.includes(params_1.RulesSupportedResourceType.AzureKubernetesService)) {
        llmResponse.push(`- Kubernetes (K8s) YAML naming: only Lowercase letters (a-z), digits (0-9), hyphens (-) is allowed. Must start and end with a letter or digit. Less than 20 characters.`);
    }
}
function populateAZCLIScriptPrompts(llmResponse, resourceTypes) {
    llmResponse.push(`- The script should be idempotent, i.e. it should check for resource existence first and only create resources that don't exist, making it safe to rerun. Check resource existence using '$<resource>exist = az <resource> show 2>$null' for ps1. Use '$ErrorActionPreference = "Continue"' to avoid breaking existence check.`);
    llmResponse.push(`- Set up an optional input *resourceToken* for the script. This is a random alphanumeric string of <= 5 characters with lowercase letters only. E.g. \`abcde\`.`);
    llmResponse.push(`- Azure resource Naming: All resources should be named like {resourcePrefix}{*resourceToken*}{instance}. Alphanumeric only! Don't use special characters! resourcePrefix is a prefix for the resource (ex. 'kv' for key vault) and <= 3 characters. instance is a number for multiple resource with the same type. E.g. rgabcde(resource group), kvabcde (key vault), sqlabcde(sql server), appabcde1(container app 1), appabcde2 (container app 2).`);
    llmResponse.push(`- Use '--yes' or equivalent flags to avoid interactive prompts.`);
}
// General IaC rules
function populateBicepPrompts(llmResponse) {
    llmResponse.push("- Expected files: main.bicep, main.parameters.json (with parameters from main.bicep).");
    llmResponse.push("- Resource token format: 'uniqueString(subscription().id, resourceGroup().id, location, environmentName)' (scope = resourceGroup) or 'uniqueString(subscription().id, location, environmentName)' (scope = subscription).");
    llmResponse.push("- All resources should be named like az{resourcePrefix}{resourceToken}, where resourcePrefix is a prefix for the resource (ex. 'kv' for key vault) and <= 3 characters. Alphanumeric only. ResourceToken is the string generated by uniqueString as per earlier. Apply this to ALL resources as well as newly written resources.");
}
;
function PopulateTerraformPrompts(llmResponse) {
    llmResponse.push("- Expected files: main.tf, main.tfvars.json (with the minimally required parameters), outputs.tf.");
    llmResponse.push("- Resource names should use Azure CAF naming convention. This is required for deployments. Add aztfmod/azurecaf in the required provider. DO NOT use random_length. NO suffixes needed.");
}
;
// Role assignment prompt helper
function AddPromptForRoleAssignment(roleId, roleName, iacType, additionalInstructions) {
    const roleAssignmentResourceName = iacType === models_1.IacType.Bicep ? "Microsoft.Authorization/roleAssignments" : "azurerm_role_assignment";
    let returnString = `- MANDATORY: Add a ${roleAssignmentResourceName} resource to assign the ${roleName} (${roleId}) role to the user-assigned managed identity`;
    if (additionalInstructions) {
        returnString += ` (${additionalInstructions})`;
    }
    return returnString;
}
// App Service rules
function populateAppServiceGeneralPrompts(llmResponse) {
    llmResponse.push("- Must attach user-assigned managed identity.");
}
function populateAppServiceIaCPrompts(llmResponse, iacType) {
    if (iacType === models_1.IacType.Bicep) {
        llmResponse.push("- Enable application insights via environment variable APPLICATIONINSIGHTS_CONNECTION_STRING.");
        llmResponse.push("- Enable CORS in SiteConfig.cors.");
        llmResponse.push("- Must define diagnostic settings (type: \"Microsoft.Insights/diagnosticSettings\").");
        llmResponse.push("- On the App Service Plan (Microsoft.Web/serverfarms), properties.reserved must be 'true' for Linux, and 'false' for Windows.");
    }
    else if (iacType === models_1.IacType.Terraform) {
        llmResponse.push("- Use \"azurerm_monitor_diagnostic_setting\" resource to send app service logs to Log Analytics Workspace.");
        llmResponse.push("- On the App Service Plan, reserved must be 'true' for Linux, and 'false' for Windows.");
    }
}
function populateAppServicePrompts(llmResponse, deploymentTool, iacType) {
    llmResponse.push("=== Additional requirements for App Service:");
    populateAppServiceGeneralPrompts(llmResponse);
    if (iacType) {
        populateAppServiceIaCPrompts(llmResponse, iacType);
    }
    llmResponse.push("===");
}
// App Service Managed Instance rules
function populateAppServiceManagedInstanceGeneralPrompts(llmResponse) {
    // context explanation
    llmResponse.push("- Managed Instance on Azure App Service is a plan-scoped hosting option for Windows web apps. ");
    const appServiceManagedInstanceSetupSteps = new planStep_1.PlanStep("Set up a Managed Instance with TWO major STAGES:", planStep_1.PlanStepBulletType.Bulleted);
    const storageAccountSetupSteps = new planStep_1.PlanStep("STAGE 1: Storage Account and Blob Container Setup with a **SEPARATE** step and file.", planStep_1.PlanStepBulletType.Numbered);
    storageAccountSetupSteps.addSteps([
        "Create a user-assigned managed identity.",
        "Create an Azure Storage Account.",
        "Assign the managed identity to the storage container.",
        "Grant Storage-Blob-Data-Contributor(ba92f5b4-2d11-453d-a403-e96b0029c9fe) access on the storage container to the managed identity and *current user(for uploading scripts)*.",
        "Compress the plan-level persistent customization into a scripts.zip.",
        "Upload scripts.zip to the storage container."
    ]);
    const appServiceSetupSteps = new planStep_1.PlanStep("STAGE 2: App Service Plan and Web App setup", planStep_1.PlanStepBulletType.Numbered);
    appServiceSetupSteps.addSteps([
        "Create the app service plan and Windows web app",
        "Attach the user-assigned managed identity to both the app service plan and web app."
    ]);
    appServiceManagedInstanceSetupSteps.addSteps([storageAccountSetupSteps, appServiceSetupSteps]);
    llmResponse.push(appServiceManagedInstanceSetupSteps.renderIntoMarkdown());
}
function populateAppServiceManagedInstanceCLIPrompts(llmResponse) {
    llmResponse.push("- Create the plan using \`az appservice plan create\` with parameter \`--is-managed-instance\`, \`--install-script name=<name> source-uri=<uri> type=\"RemoteAzureBlob\"\`, and `\--mi-user-assigned <resource-id>`\.");
}
function populateAppServiceManagedInstanceIaCPrompts(llmResponse, iacType) {
    if (iacType === models_1.IacType.Bicep) {
        llmResponse.push("- Use *SEPARATE* files (storage.bicep and setupstorage.ps1(.sh)) to prepare storage account and scripts FIRST. Create web app in main.bicep using the created storage account and script URIs.");
        llmResponse.push("- Create app service plan using API version \`2025-03-01\` with \`isCustomMode=true\` and \`installScripts\` filled. Enable user-assigned managed identity.");
        llmResponse.push("- The property for installScripts is an array of {name: '<name>'\nsource: {\nsourceUri: '<scriptUri>'\ntype: 'RemoteAzureBlob'\n}.");
        llmResponse.push("- **Managed instance is available for Windows web apps and limited to Pv4 and Pmv4.**");
    }
    else if (iacType === models_1.IacType.Terraform) {
        llmResponse.push("- Terraform not supported for App Service Managed Instance yet.");
    }
}
function populateAppServiceManagedInstanceAZDPrompts(llmResponse) {
    llmResponse.push("- In azure.yaml, the host should still be set to 'appservice'. The managed instance will be created based on the app service plan settings.");
}
function populateAppServiceManagedInstancePrompts(llmResponse, deploymentTool, iacType) {
    llmResponse.push(`=== Additional requirements for App Service Managed Instance:`);
    populateAppServiceManagedInstanceGeneralPrompts(llmResponse);
    if (deploymentTool === models_1.DeploymentTool.AzCLI) {
        populateAppServiceManagedInstanceCLIPrompts(llmResponse);
    }
    else if (deploymentTool === models_1.DeploymentTool.Azd) {
        populateAppServiceManagedInstanceAZDPrompts(llmResponse);
    }
    if (iacType) {
        populateAppServiceManagedInstanceIaCPrompts(llmResponse, iacType);
    }
    llmResponse.push("===");
}
// Container App rules
function populateContainerAppGeneralPrompts(llmResponse) {
    llmResponse.push("- Attach User-Assigned Managed Identity.");
}
function populateContainerAppIaCPrompts(llmResponse, iacType) {
    llmResponse.push(AddPromptForRoleAssignment('7f951dda-4ed3-4680-a7ca-43fe172d538d', 'AcrPull', iacType, "Only one instance is required per-container registry. Define this BEFORE any container apps."));
    llmResponse.push(`- Use this identity (NOT system) to connect to the container registry. A registry connection needs to be created even if we are using a template base image.`);
    const image = "mcr.microsoft.com/azuredocs/containerapps-helloworld:latest";
    const imageProperty = iacType === models_1.IacType.Bicep ? "properties.template.containers.image" : "azurerm_container_app.template.container.image";
    llmResponse.push(`- Container Apps MUST use base container image ${image}. The property is set via ${imageProperty}.`);
    if (iacType === models_1.IacType.Bicep) {
        llmResponse.push("- Use 'properties.configuration.registries' to set the container registry connection.");
        llmResponse.push("- Enable CORS via properties.configuration.ingress.corsPolicy.");
    }
    else if (iacType === models_1.IacType.Terraform) {
        llmResponse.push("- Create an ***azapi_resource_action*** resource using :type `Microsoft.App/containerApps`, method `PATCH`, and body `properties.configuration.ingress.corsPolicy` property to enable CORS for all origins, headers, and methods. Use 'azure/azapi' provider version *2.0*. DO NOT use jsonencode() for the body.");
    }
    llmResponse.push("- Define all used secrets; Use Key Vault if possible.");
    llmResponse.push("- MANDATORY: You MUST add Key Vault secrets and their related role assignment in explicit dependencies to use these secrets.");
    llmResponse.push(`- IMPORTANT NOTES: If you meet deployment error: "Field 'configuration.secrets' is invalid with details: Invalid value: <secret-name>: unable to get value using managed identity". DO NOT use key vault references first, use connection string directly in container app environment variable and deploy.
  After deployment is successful, get IP from the container app, switch to use key vault references in secrets, and 1) add the IP to key vault network rules 2) set defaultAction: 'Deny'. This is a workaround for the issue that managed identity cannot access key vault when key vault has network rules enabled.`);
    if (iacType === models_1.IacType.Bicep) {
        llmResponse.push("- Container App Environment must be connected to Log Analytics Workspace. Use logAnalyticsConfiguration -> customerId=logAnalytics.properties.customerId and sharedKey=logAnalytics.listKeys().primarySharedKey.");
    }
    else {
        llmResponse.push("- Container App Environment must be connected to Log Analytics Workspace. Use logs_destination=\"log-analytics\" azurerm_container_app_environment.log_analytics_workspace_id = azurerm_log_analytics_workspace.<workspaceName>.id.");
    }
}
function populateContainerAppPrompts(llmResponse, deploymentTool, iacType) {
    llmResponse.push(`=== Additional requirements for Container Apps:`);
    populateContainerAppGeneralPrompts(llmResponse);
    if (iacType) {
        populateContainerAppIaCPrompts(llmResponse, iacType);
    }
    llmResponse.push("===");
}
;
// Function App rules
function populateFunctionAppGeneralPrompts(llmResponse) {
    llmResponse.push("- Attach User-Assigned Managed Identity.");
}
function populateFunctionAppIaCPrompts(llmResponse, iacType) {
    const requiredRoles = [
        { roleId: 'b7e6dc6d-f1e8-4753-8033-0f276bb0955b', name: 'Storage Blob Data Owner' },
        { roleId: 'ba92f5b4-2d11-453d-a403-e96b0029c9fe', name: 'Storage Blob Data Contributor' },
        { roleId: '974c5e8b-45b9-4653-ba55-5f855dd0fb88', name: 'Storage Queue Data Contributor' },
        { roleId: '0a9a7e1f-b9d0-4cc4-a60d-0319b160aaa3', name: 'Storage Table Data Contributor' },
        { roleId: '3913510d-42f4-4e42-8a64-420c390055eb', name: 'Monitoring Metrics Publisher' }
    ];
    for (const role of requiredRoles) {
        llmResponse.push(AddPromptForRoleAssignment(role.roleId, role.name, iacType));
    }
    llmResponse.push("- Create a storage account and connect to the function app.");
    llmResponse.push("- Always use a V1 storage account to create file shares, V2 storage accounts are not supported.");
    const diagnosticSettingsResourceType = iacType === models_1.IacType.Bicep ? "Microsoft.Insights/diagnosticSettings" : "azurerm_monitor_diagnostic_setting";
    llmResponse.push(`- Define diagnostic settings to save logs. The resource type is ${diagnosticSettingsResourceType}.`);
    llmResponse.push("- Do not add `node_modules` to `.funcignore` when using the Flex Consumption plan. This must not be ignored.");
    llmResponse.push("- Include `function.json` files. This is part of the older paradigm, but Copilot may not yet support the newer approach fully.");
}
function populateFunctionAppPrompts(llmResponse, deploymentTool, iacType) {
    llmResponse.push("=== Additional requirements for Function Apps:");
    populateFunctionAppGeneralPrompts(llmResponse);
    if (iacType) {
        populateFunctionAppIaCPrompts(llmResponse, iacType);
    }
    llmResponse.push("===");
}
;
// AKS rules
function populateAKSCLIPrompts(llmResponse, stateManager) {
    llmResponse.push("- Attach ACR to AKS cluster for pulling images using 'az aks update --attach-acr'.");
    llmResponse.push("- Set up AKS workload identity with user-assigned managed identity by creating aks with '--enable-oidc-issuer --enable-workload-identity' AND CREATE a federated identity credential to link the managed identity with the Kubernetes service account.");
    llmResponse.push("- Use existing workspace resource needs '--enable-addons monitoring'.");
}
function populateAKSIaCPrompts(llmResponse, iacType) {
    llmResponse.push("- Create a kubelet identity for AKS Cluster, and add 'Managed Identity Operator' role assignment to the identity.");
    llmResponse.push("- Assign AcrPull role to the kubelet identity for pulling images from ACR.");
}
function populateAzureKubernetesPrompts(llmResponse, deploymentTool, stateManager, iacType) {
    llmResponse.push("=== Additional requirements for Azure Kubernetes Service (AKS):");
    llmResponse.push("- **CRITICAL** PREREQUISITE: You MUST use a stable kubernetesVersion, get by `az aks get-versions --query 'values[?isDefault].version' --location <> -o tsv`.");
    llmResponse.push(`- For SKU, set based on ${stateManager.getToolName(toolType_1.ToolType.GetRegionsForDeployment)} output to get available VM size in the region, e.g. Standard_D2ds_v5.`);
    llmResponse.push("- Use Kubernetes Secret for environment variables, don't get it from Key Vault.");
    if (deploymentTool === models_1.DeploymentTool.AzCLI && !iacType) {
        populateAKSCLIPrompts(llmResponse, stateManager);
    }
    if (iacType) {
        populateAKSIaCPrompts(llmResponse, iacType);
    }
    llmResponse.push("===");
}
// Database common rules
function DatabaseCommonPrompts(llmResponse) {
    llmResponse.push("- Add firewall rules to allow traffic from Azure Services (allow IP 0.0.0.0).");
    llmResponse.push("- If username and password are required, you MUST leave them as params.");
    llmResponse.push("- Create secrets in Key Vault to store the connection string or credentials, and assign `Key Vault Secrets User` role to the user-assigned managed identity.");
}
// PostgreSQL rules
function populatePostgreSqlGeneralPrompts(llmResponse, stateManager, iacType) {
    if (iacType === models_1.IacType.Terraform) {
        llmResponse.push(`- PostgreSQL SKU name format: B_Standard_B1ms(Burstable tier), GP_Standard_D2s_v3(GeneralPurpose), MO_Standard_E4s_v3(MemoryOptimized)`);
        llmResponse.push("- For version, prefer to use '16'.");
    }
    else {
        llmResponse.push(`- Recommended PostgreSQL SKU: GeneralPurpose tier, at least 2 vCores, format: Standard_D2ds_v5`);
        llmResponse.push("- For version, use '17' or higher.");
    }
    llmResponse.push("- Don't create a database naming postgres, it's a built-in db.");
    llmResponse.push(`- If you have error provisioning a PostgreSQL flexible server: 'An unexpected error while processing the request'. Check quota for Microsoft.DBforPostgreSQL/flexibleServers in the region with tool. If current region is not available, 1)choose an available region and 2)add an index to the postgresql name suffix and retry.`);
}
function populatePostgreSqlCLIPrompts(llmResponse) {
    llmResponse.push("- If PostgreSQL server uses Azure AD authentication, use \'--microsoft-entra-auth Enabled\' when creating.");
    llmResponse.push("- Azure CLI uses parameters \'--name <server-name> --rule-name <rule-name>\' for firewall rules creation.");
    // Tried to use Service Connector but Copilot is not aware of it fully.
    llmResponse.push("- IMPORTANT: **If using Azure AD authentication, you MUST ADD a database USER for the managed identity and GRANT all privileges to make the connection work.** Use 'az postgres flexible-server execute' command to run SQL commands to create the user and grant privileges.");
    // Comment out for now to reduce complexity
    // llmResponse.push("- You might update PostgreSQL server to disable password authentication after enabling Azure AD authentication using '--password-auth Disabled'.");
}
function populatePostgreSqlPrompts(llmResponse, deploymentTool, stateManager, iacType) {
    llmResponse.push("=== Additional requirements for PostgreSQL:");
    populatePostgreSqlGeneralPrompts(llmResponse, stateManager, iacType);
    DatabaseCommonPrompts(llmResponse);
    if (deploymentTool === models_1.DeploymentTool.AzCLI && !iacType) {
        populatePostgreSqlCLIPrompts(llmResponse);
    }
    llmResponse.push("===");
}
;
// MySQL rules
function populateMySqlPrompts(llmResponse, deploymentTool, stateManager, iacType) {
    llmResponse.push("=== Additional requirements for MySQL:");
    llmResponse.push(`- Choose model based on ${stateManager.getToolName(toolType_1.ToolType.GetRegionsForDeployment)} output to make sure the model is available in the region. For DTU Model (Basic/Standard/Premium), check ServerQuota only. For vCore Model (GP/BC/HS), you must check both ServerQuota and vCore. For FreeDatabase, make sure SubscriptionFreeDatabaseCount quota is sufficient. `);
    DatabaseCommonPrompts(llmResponse);
    llmResponse.push("===");
}
;
// CosmosDB rules
function populateCosmosDBPrompts(llmResponse, deploymentTool, iacType) {
    llmResponse.push("=== Additional requirements for CosmosDB:");
    llmResponse.push("- Add firewall rules to allow traffic from Azure Services (allow IP 0.0.0.0).");
    llmResponse.push("===");
}
;
// Storage Account rules
function populateStorageAccountGeneralPrompts(llmResponse) {
    llmResponse.push("- By default, disable storage account local auth (key access).");
    llmResponse.push("- By default, disable storage account anonymous blob access.");
}
function populateStorageAccountIaCPrompts(llmResponse, iacType) {
    if (iacType === models_1.IacType.Terraform) {
        // work with shared_access_key_enabled = false
        llmResponse.push("- Add `storage_use_azuread = true` in azurerm provider.");
    }
}
function populateStorageAccountPrompts(llmResponse, deploymentTool, iacType) {
    llmResponse.push("=== Additional requirements for Storage Accounts:");
    populateStorageAccountGeneralPrompts(llmResponse);
    if (iacType) {
        populateStorageAccountIaCPrompts(llmResponse, iacType);
    }
    llmResponse.push("===");
}
;
// Key Vault rules
function populateKeyVaultGeneralPrompts(llmResponse, iacType) {
    if (iacType === models_1.IacType.Terraform) {
        llmResponse.push("- Set `rbac_authorization_enabled = true`."); // avoid model hallucination to use deprecated property
    }
    else {
        llmResponse.push("- Use RBAC authentication.");
    }
    llmResponse.push("- Assign role 'Key Vault Secrets Officer (b86a8fe4-44ce-4948-aee5-eccb2c155cd7)' to managed identity. Add dependencies to ensure role assignment is finished before application accesses secrets.");
    llmResponse.push("- Allow public access from all networks (set publicNetworkAccess = Enabled).");
}
function populateKeyVaultCLIPrompts(llmResponse) {
    llmResponse.push("- IMPORTANT: Assign Key Vault Secrets Officer to current user. Add delay after RBAC role assignment to allow propagation before creating secrets.");
}
function populateKeyVaultPrompts(llmResponse, deploymentTool, iacType) {
    llmResponse.push("=== Additional requirements for Key Vault:");
    populateKeyVaultGeneralPrompts(llmResponse, iacType);
    if (deploymentTool === models_1.DeploymentTool.AzCLI && !iacType) {
        populateKeyVaultCLIPrompts(llmResponse);
    }
    llmResponse.push("===");
}
function populateResourcePrompts(resourceType, llmResponse, deploymentTool, stateManager, iacType) {
    if (resourceType === params_1.RulesSupportedResourceType.AzureContainerApp) {
        populateContainerAppPrompts(llmResponse, deploymentTool, iacType);
    }
    else if (resourceType === params_1.RulesSupportedResourceType.AzureAppService) {
        populateAppServicePrompts(llmResponse, deploymentTool, iacType);
    }
    else if (resourceType === params_1.RulesSupportedResourceType.AzureAppServiceManagedInstance) {
        populateAppServiceManagedInstancePrompts(llmResponse, deploymentTool, iacType);
    }
    else if (resourceType === params_1.RulesSupportedResourceType.AzureFunctionApp) {
        populateFunctionAppPrompts(llmResponse, deploymentTool, iacType);
    }
    else if (resourceType === params_1.RulesSupportedResourceType.AzureKubernetesService) {
        populateAzureKubernetesPrompts(llmResponse, deploymentTool, stateManager, iacType);
    }
    else if (resourceType === params_1.RulesSupportedResourceType.PostgreSqlDatabase) {
        populatePostgreSqlPrompts(llmResponse, deploymentTool, stateManager, iacType);
    }
    else if (resourceType === params_1.RulesSupportedResourceType.MySqlServer) {
        populateMySqlPrompts(llmResponse, deploymentTool, stateManager, iacType);
    }
    else if (resourceType === params_1.RulesSupportedResourceType.CosmosDB) {
        populateCosmosDBPrompts(llmResponse, deploymentTool, iacType);
    }
    else if (resourceType === params_1.RulesSupportedResourceType.StorageAccount) {
        populateStorageAccountPrompts(llmResponse, deploymentTool, iacType);
    }
    else if (resourceType === params_1.RulesSupportedResourceType.KeyVault) {
        populateKeyVaultPrompts(llmResponse, deploymentTool, iacType);
    }
    else {
        llmResponse.push(`No additional IaC rules for resource type: ${resourceType}`);
    }
    llmResponse.push("\n");
}
function populateLLMResponse(deploymentTool, iacTypeInput, resourceTypes, llmResponse, stateManager) {
    let iacType = undefined;
    if (typeof iacTypeInput === "string") {
        if (Object.values(models_1.IacType).includes(iacTypeInput)) {
            iacType = iacTypeInput;
        }
    }
    else if (iacTypeInput !== undefined) {
        iacType = iacTypeInput;
    }
    llmResponse.push(`Mandatory rules for deployment. You must implement every rule exactly as stated, with no exceptions or omissions, even if it is not a common pattern or seems redundant. Do not use your own judgment to simplify, skip, or modify any rule. If a rule is present, it must be enforced in the code, regardless of context.` + (iacType ? ` Adjust ${iacType} files to align with these rules.` : ''));
    llmResponse.push("\n");
    llmResponse.push(`- [IMPORTANT] Must call ${stateManager.getToolName(toolType_1.ToolType.GetRegionsForDeployment)} first to get available regions and SKU for each resources. Resource region can be different with resource group, set each resource location accordingly!`);
    llmResponse.push(`- [IMPORTANT] If current region for a resource is not available, 1)choose an available region and 2)add an index to the resource name suffix to avoid naming conflict. Don't change other resources' region in the resource group.`);
    llmResponse.push("\n");
    llmResponse.push(`## Deployment Tool ${deploymentTool} rules:`);
    if (deploymentTool === models_1.DeploymentTool.Azd) {
        if (!iacType) {
            iacType = models_1.IacType.Bicep; // Default to Bicep if not specified
        }
        populateAZDPrompts(iacType, resourceTypes, llmResponse);
    }
    else if (deploymentTool === models_1.DeploymentTool.AzCLI) {
        populateAZCLIPrompts(llmResponse, resourceTypes, iacType);
    }
    if (iacType) {
        llmResponse.push("\n");
        llmResponse.push(`## IaC Type: ${iacType} rules:`);
        if (iacType === models_1.IacType.Bicep) {
            populateBicepPrompts(llmResponse);
        }
        else if (iacType === models_1.IacType.Terraform) {
            PopulateTerraformPrompts(llmResponse);
        }
        ;
    }
    // Resource-specific rules
    llmResponse.push("\n");
    llmResponse.push(`## Resources: ${resourceTypes.join(', ')}`);
    for (const resourceType of resourceTypes) {
        populateResourcePrompts(resourceType, llmResponse, deploymentTool, stateManager, iacType);
    }
    if (!resourceTypes.includes(params_1.RulesSupportedResourceType.KeyVault)) {
        populateResourcePrompts(params_1.RulesSupportedResourceType.KeyVault, llmResponse, deploymentTool, stateManager, iacType);
    }
    llmResponse.push(`Call get_errors every time you make code changes, otherwise your deployment will fail. You must follow ALL of the previously mentioned rules. DO NOT IGNORE ANY RULES. Call ${stateManager.getToolName(toolType_1.ToolType.GetAzureResourceRules)} again if need to get the rules again. Show the user a report line-by-line of each rule that was applied. Only skip a rule if there is no corresponding resource (e.g. no function app). Do not stop at error-free code, you must apply all the rules.`);
    llmResponse.push("\n");
    let necessaryTools = ["az cli (az --version)"];
    if (deploymentTool === models_1.DeploymentTool.Azd) {
        necessaryTools.push("azd (azd version)");
    }
    ;
    if (resourceTypes.includes(params_1.RulesSupportedResourceType.AzureContainerApp)) {
        necessaryTools.push("docker (docker --version)");
    }
    ;
    llmResponse.push(`## Tools needed: ${necessaryTools.join(',')}.`);
    if (iacType === models_1.IacType.Terraform && deploymentTool === models_1.DeploymentTool.Azd) {
        llmResponse.push("\n");
        llmResponse.push(`Note: Do not use Terraform CLI.`);
    }
}
