// containerization-assist-mcp/dist/src/infra/security/scanner.js
import { Success, Failure } from '../../types/index.js';
import { extractErrorMessage } from '../../lib/errors.js';
import { scanImageWithTrivy, checkTrivyAvailability } from './trivy-scanner.js';
/**
 * Create a Trivy-based security scanner
 */
function createTrivyScanner(logger) {
    return {
        async scanImage(imageId) {
            return scanImageWithTrivy(imageId, logger);
        },
        async ping() {
            const result = await checkTrivyAvailability(logger);
            if (result.ok) {
                logger.debug({ version: result.value }, 'Trivy scanner available');
                return Success(true);
            }
            return Failure(result.error, result.guidance);
        },
    };
}
/**
 * Create a stub scanner that returns empty results
 * Used when no real scanner is configured
 */
function createStubScanner(logger) {
    return {
        async scanImage(imageId) {
            try {
                logger.info({ imageId, scanner: 'stub' }, 'Starting stub security scan');
                const result = {
                    imageId,
                    vulnerabilities: [],
                    totalVulnerabilities: 0,
                    criticalCount: 0,
                    highCount: 0,
                    mediumCount: 0,
                    lowCount: 0,
                    negligibleCount: 0,
                    unknownCount: 0,
                    scanDate: new Date(),
                };
                logger.warn({ imageId }, 'Stub scanner returns empty results - no actual scanning performed');
                return Success(result);
            }
            catch (error) {
                const errorMessage = extractErrorMessage(error);
                logger.error({ error: errorMessage, imageId }, 'Security scan failed');
                return Failure(errorMessage);
            }
        },
        async ping() {
            logger.debug('Checking stub scanner availability');
            return Success(true);
        },
    };
}
/**
 * Create a security scanner based on the specified type
 *
 * @param logger - Logger instance
 * @param scannerType - Type of scanner to create ('trivy', 'stub', or undefined for 'trivy')
 * @returns SecurityScanner instance
 */
export const createSecurityScanner = (logger, scannerType) => {
    const type = (scannerType || 'trivy').toLowerCase();
    switch (type) {
        case 'trivy':
            return createTrivyScanner(logger);
        case 'stub':
            return createStubScanner(logger);
        default:
            logger.warn({ scannerType: type }, 'Unknown scanner type, falling back to Trivy');
            return createTrivyScanner(logger);
    }
};
