// containerization-assist-mcp/dist-cjs/src/tools/generate-k8s-manifests/schema.js
"use strict";
/**
 * Schema definition for generate-k8s-manifests tool
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateK8sManifestsSchema = void 0;
const zod_1 = require("zod");
const schemas_1 = require("../shared/schemas");
exports.generateK8sManifestsSchema = zod_1.z
    .object({
    // Module info fields - required for repository mode, optional for ACA mode
    name: zod_1.z
        .string()
        .optional()
        .describe('Module name. Required when generating from repository analysis.'),
    modulePath: zod_1.z
        .string()
        .optional()
        .describe('Absolute path to module root. Required when generating from repository analysis. Paths are automatically normalized to forward slashes on all platforms.'),
    dockerfilePath: zod_1.z.string().optional().describe('Path where the Dockerfile should be generated'),
    language: zod_1.z
        .enum(['java', 'dotnet', 'javascript', 'typescript', 'python', 'rust', 'go', 'other'])
        .optional()
        .describe('Primary programming language used in the module'),
    languageVersion: zod_1.z.string().optional(),
    frameworks: zod_1.z
        .array(zod_1.z.object({
        name: zod_1.z.string().describe('Framework name (e.g., Spring Boot, Express, Flask)'),
        version: zod_1.z.string().optional(),
    }))
        .optional(),
    buildSystem: zod_1.z
        .object({
        type: zod_1.z.string().optional(),
        configFile: zod_1.z.string().optional(),
    })
        .optional()
        .describe('Build system information'),
    dependencies: zod_1.z
        .array(zod_1.z.string())
        .optional()
        .describe('List of module dependencies including database drivers and system libraries'),
    ports: zod_1.z.array(zod_1.z.number()).optional(),
    entryPoint: zod_1.z.string().optional(),
    // ACA conversion field
    acaManifest: zod_1.z
        .string()
        .optional()
        .describe('Azure Container Apps manifest content to convert (YAML or JSON). Required when converting from ACA manifest; omit when generating from repository analysis.'),
    // Common fields
    manifestType: zod_1.z
        .enum(['kubernetes', 'helm', 'aca', 'kustomize'])
        .describe('Type of manifest to generate'),
    environment: schemas_1.environment.describe('Target environment (production, development, etc.)'),
    detectedDependencies: zod_1.z
        .array(zod_1.z.string())
        .optional()
        .describe('Detected libraries/frameworks/features from repository analysis (e.g., ["redis", "ef-core", "signalr", "mongodb", "health-checks"]). This helps match relevant knowledge entries.'),
    includeComments: zod_1.z
        .boolean()
        .optional()
        .default(true)
        .describe('Add helpful comments in the output (primarily for ACA conversions)'),
    namespace: zod_1.z.string().optional().describe('Target Kubernetes namespace'),
    trafficLevel: zod_1.z
        .enum(['high', 'medium', 'low'])
        .optional()
        .describe('Expected traffic level for dynamic defaults calculation (affects replica counts and scaling).'),
    criticalityTier: zod_1.z
        .enum(['tier-1', 'tier-2', 'tier-3'])
        .optional()
        .describe('Criticality tier for dynamic defaults calculation (tier-1=mission-critical, tier-3=low-priority).'),
})
    .superRefine((data, ctx) => {
    const hasAcaManifest = !!data.acaManifest;
    const hasModuleInfo = !!data.name && !!data.modulePath;
    // Require exactly one mode: ACA conversion OR repository analysis
    if (!hasAcaManifest && !hasModuleInfo) {
        ctx.addIssue({
            code: zod_1.z.ZodIssueCode.custom,
            message: 'Either provide acaManifest (for ACA conversion) or name+modulePath (for repository analysis)',
            path: ['acaManifest'],
        });
        ctx.addIssue({
            code: zod_1.z.ZodIssueCode.custom,
            message: 'Either provide acaManifest (for ACA conversion) or name+modulePath (for repository analysis)',
            path: ['name'],
        });
    }
    // Repository mode requires both name and modulePath
    if (!hasAcaManifest) {
        if (!data.name) {
            ctx.addIssue({
                code: zod_1.z.ZodIssueCode.custom,
                message: 'name is required when not using acaManifest',
                path: ['name'],
            });
        }
        if (!data.modulePath) {
            ctx.addIssue({
                code: zod_1.z.ZodIssueCode.custom,
                message: 'modulePath is required when not using acaManifest',
                path: ['modulePath'],
            });
        }
    }
});
