// containerization-assist-mcp/dist/src/lib/param-defaults.js
/**
 * Parameter defaulting utilities to reduce duplication of params.X || 'default' patterns
 */
/**
 * Applies default values to tool parameters.
 * Type-safe parameter defaulting with explicit schema.
 * Consolidates the pattern: params.field || defaultValue
 *
 * Note: The `Record<string, unknown>` constraint provides type safety while supporting
 * flexible tool parameter objects with mixed types (strings, numbers, booleans,
 * nested objects, arrays, etc.). The actual type safety comes from the generic
 * parameter `T` which should be the concrete interface of your tool parameters.
 *
 * @param params - Partial parameters from user input
 * @param defaults - Complete default values
 * @returns Merged parameters with defaults applied
 *
 * @example
 * ```typescript
 * interface K8sParams {
 *   namespace: string;
 *   replicas: number;
 *   serviceType: 'ClusterIP' | 'LoadBalancer';
 * }
 *
 * const params: Partial<K8sParams> = { replicas: 3 };
 * const result = withDefaults(params, {
 *   namespace: 'default',
 *   replicas: 1,
 *   serviceType: 'ClusterIP'
 * }); // result: K8sParams with replicas: 3, other defaults applied
 * ```
 */
export function withDefaults(params, defaults) {
    const result = { ...defaults };
    // Only override defaults with truthy values from params
    for (const key in params) {
        const value = params[key];
        if (value !== undefined && value !== null) {
            result[key] = value;
        }
    }
    return result;
}
/**
 * Common Kubernetes parameter defaults
 */
export const K8S_DEFAULTS = {
    namespace: 'default',
    serviceType: 'ClusterIP',
    replicas: 1,
    port: 8080,
    targetPort: 8080,
};
/**
 * Common container/Docker parameter defaults
 */
export const CONTAINER_DEFAULTS = {
    environment: 'production',
    registry: 'docker.io',
    cpu: '0.5',
    memory: '1Gi',
    tag: 'latest',
};
/**
 * Azure Container Apps defaults
 */
export const ACA_DEFAULTS = {
    environment: 'production',
    location: 'eastus',
    cpu: '0.5',
    memory: '1Gi',
    minReplicas: 0,
    maxReplicas: 10,
};
/**
 * Build-related defaults
 */
export const BUILD_DEFAULTS = {
    platform: 'linux/amd64',
    nocache: false,
    push: false,
};
/**
 * Helper to get defaults for a specific tool category
 */
export function getToolDefaults(category) {
    switch (category) {
        case 'k8s':
            return K8S_DEFAULTS;
        case 'container':
            return CONTAINER_DEFAULTS;
        case 'aca':
            return ACA_DEFAULTS;
        case 'build':
            return BUILD_DEFAULTS;
        default:
            return {};
    }
}
