// containerization-assist-mcp/dist/src/lib/tool-logger.js
import { writeFileSync, mkdirSync } from 'fs';
import { appendFile } from 'fs/promises';
import { join } from 'path';
import { config } from '../config/index.js';
export function createToolLogEntry(toolName, input) {
    const entry = {
        timestamp: Date.now().toString(),
        toolName,
        input,
        params: input,
        output: undefined,
        success: false,
    };
    return entry;
}
let logFileName = null;
function isToolLoggingEnabled() {
    return config.toolLogging.enabled;
}
export function getLogFilePath() {
    const dirPath = config.toolLogging.dirPath;
    if (!dirPath) {
        return '';
    }
    if (logFileName) {
        return join(dirPath, logFileName);
    }
    const timestamp = Date.now();
    logFileName = `ca-tool-logs-${timestamp}.jsonl`;
    return join(dirPath, logFileName);
}
export function createToolLoggerFile(logger) {
    if (!isToolLoggingEnabled()) {
        return;
    }
    logger?.info({ message: 'Starting tool logger file creation' });
    const dirPath = config.toolLogging.dirPath;
    if (!dirPath) {
        logger?.warn('Tool logging directory path is not configured');
        return;
    }
    try {
        // Ensure directory exists
        mkdirSync(dirPath, { recursive: true });
        const logFilePath = getLogFilePath();
        writeFileSync(logFilePath, '', 'utf-8');
        logger?.info({ path: logFilePath }, 'Tool logging file created');
    }
    catch (error) {
        const errorMsg = `Tool logging directory '${dirPath}' is not writable: ${error.message}`;
        logger?.warn({ error, path: dirPath }, errorMsg);
    }
}
export async function logToolExecution(entry, logger) {
    if (!isToolLoggingEnabled()) {
        return;
    }
    try {
        const filepath = getLogFilePath();
        const logLine = `${JSON.stringify(entry)}\n`;
        await appendFile(filepath, logLine, 'utf-8');
        logger?.debug({ filepath, toolName: entry.toolName }, 'Tool execution logged to file');
    }
    catch (error) {
        logger?.warn({ error, toolName: entry.toolName }, 'Failed to write tool execution log');
    }
}
