// containerization-assist-mcp/dist-cjs/src/infra/security/osv-scanner/maven/pom-parser.js
"use strict";
// Maven POM parser with XML bomb protection (max 10MB)
// Local property resolution only - no remote parent POM fetching
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseMavenPackages = parseMavenPackages;
const tar = __importStar(require("tar"));
const xml2js_1 = require("xml2js");
const errors_1 = require("../../../../lib/errors");
const property_resolver_1 = require("./property-resolver");
const MAX_XML_SIZE_BYTES = 10 * 1024 * 1024;
async function parseMavenPackages(buffer, path, logger) {
    try {
        const pomContent = await extractFileFromTar(buffer, path);
        if (!pomContent) {
            logger.debug({ path }, 'pom.xml not found in tar');
            return [];
        }
        const packages = await parsePomXml(pomContent, logger);
        logger.debug({ path, packageCount: packages.length }, 'Parsed Maven dependencies');
        return packages;
    }
    catch (err) {
        logger.warn({ path, error: (0, errors_1.extractErrorMessage)(err) }, 'Failed to parse pom.xml');
        return [];
    }
}
async function extractFileFromTar(buffer, targetPath) {
    const { Readable } = await Promise.resolve().then(() => __importStar(require('stream')));
    const { pipeline } = await Promise.resolve().then(() => __importStar(require('stream/promises')));
    let content = null;
    let sizeError = null;
    const fileName = targetPath.split('/').pop() || targetPath;
    const stream = Readable.from(buffer);
    try {
        await pipeline(stream, tar.list({
            onentry: (entry) => {
                if (entry.path.includes('..')) {
                    entry.resume();
                    return;
                }
                const entryFileName = entry.path.split('/').pop() || entry.path;
                if (entryFileName === fileName) {
                    if (entry.size > MAX_XML_SIZE_BYTES) {
                        sizeError = new Error(`pom.xml file too large (${entry.size} bytes, max ${MAX_XML_SIZE_BYTES})`);
                        entry.resume();
                        return;
                    }
                    const chunks = [];
                    let totalSize = 0;
                    entry.on('data', (chunk) => {
                        totalSize += chunk.length;
                        if (totalSize > MAX_XML_SIZE_BYTES) {
                            sizeError = new Error(`pom.xml content exceeded size limit (${totalSize} bytes, max ${MAX_XML_SIZE_BYTES})`);
                            entry.resume();
                            return;
                        }
                        chunks.push(chunk);
                    });
                    entry.on('end', () => {
                        if (!sizeError)
                            content = Buffer.concat(chunks).toString('utf-8');
                    });
                }
                else {
                    entry.resume();
                }
            },
        }));
    }
    catch {
        if (sizeError)
            throw sizeError;
    }
    if (sizeError)
        throw sizeError;
    return content;
}
async function parsePomXml(xmlContent, logger) {
    const contentSizeBytes = Buffer.byteLength(xmlContent, 'utf-8');
    if (contentSizeBytes > MAX_XML_SIZE_BYTES) {
        throw new Error(`pom.xml content too large (${contentSizeBytes} bytes, max ${MAX_XML_SIZE_BYTES})`);
    }
    return new Promise((resolve, reject) => {
        (0, xml2js_1.parseString)(xmlContent, {
            trim: true,
            explicitArray: false,
            strict: true,
            xmlns: false,
            explicitCharkey: false,
        }, (err, result) => {
            if (err) {
                reject(err);
                return;
            }
            try {
                const project = result.project;
                if (!project) {
                    logger.debug('No project element in pom.xml');
                    resolve([]);
                    return;
                }
                const properties = (0, property_resolver_1.extractLocalProperties)(project, logger);
                const dependencyManagement = (0, property_resolver_1.extractLocalDependencyManagement)(project);
                const packages = [];
                let skippedDeps = 0;
                if (project.dependencies?.dependency) {
                    const deps = Array.isArray(project.dependencies.dependency)
                        ? project.dependencies.dependency
                        : [project.dependencies.dependency];
                    for (const dep of deps) {
                        if (dep.groupId && dep.artifactId) {
                            let version = dep.version;
                            if (!version || version.includes('${')) {
                                const key = `${dep.groupId}:${dep.artifactId}`;
                                version = dependencyManagement.get(key);
                            }
                            if (version?.includes('${')) {
                                version = (0, property_resolver_1.resolvePropertyValue)(version, properties, project);
                            }
                            if (version && !version.includes('${')) {
                                packages.push({
                                    name: `${dep.groupId}:${dep.artifactId}`,
                                    version,
                                    ecosystem: 'Maven',
                                });
                            }
                            else {
                                skippedDeps++;
                            }
                        }
                    }
                }
                if (project.dependencyManagement?.dependencies?.dependency) {
                    const deps = Array.isArray(project.dependencyManagement.dependencies.dependency)
                        ? project.dependencyManagement.dependencies.dependency
                        : [project.dependencyManagement.dependencies.dependency];
                    for (const dep of deps) {
                        if (dep.groupId && dep.artifactId && dep.version) {
                            let version = dep.version;
                            if (version.includes('${')) {
                                version = (0, property_resolver_1.resolvePropertyValue)(version, properties, project);
                            }
                            if (!version.includes('${')) {
                                packages.push({
                                    name: `${dep.groupId}:${dep.artifactId}`,
                                    version,
                                    ecosystem: 'Maven',
                                });
                            }
                            else {
                                skippedDeps++;
                            }
                        }
                    }
                }
                if (skippedDeps > 0) {
                    logger.info({ skippedCount: skippedDeps }, 'Skipped dependencies with unresolved versions');
                }
                resolve(packages);
            }
            catch (parseErr) {
                reject(parseErr);
            }
        });
    });
}
