// containerization-assist-mcp/dist-cjs/src/tools/analyze-repo/schema.js
"use strict";
/**
 * Schema definition for analyze-repo tool
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.analyzeRepoSchema = void 0;
const zod_1 = require("zod");
const schemas_1 = require("../shared/schemas");
const moduleInfo = zod_1.z.object({
    name: zod_1.z.string().describe('The name of the module'),
    modulePath: zod_1.z
        .string()
        .describe('Absolute path to module root. Paths are automatically normalized to forward slashes on all platforms.'),
    dockerfilePath: zod_1.z.string().optional().describe('Path where the Dockerfile should be generated'),
    language: zod_1.z
        .enum(['java', 'dotnet', 'javascript', 'typescript', 'python', 'rust', 'go', 'other'])
        .optional()
        .describe('Primary programming language used in the module'),
    frameworks: zod_1.z
        .array(zod_1.z.object({
        name: zod_1.z
            .string()
            .describe('Frameworks used in project like Java Spring, SpringBoot, Hibernate etc.'),
        version: zod_1.z.string().optional(),
    }))
        .optional(),
    buildSystems: zod_1.z
        .array(zod_1.z.object({
        type: zod_1.z.string(),
        languageVersion: zod_1.z.string().optional(),
    }))
        .optional()
        .describe('All detected build systems with their language versions'),
    dependencies: zod_1.z
        .array(zod_1.z.string())
        .optional()
        .describe('List of module dependencies including database drivers and system libraries'),
    ports: zod_1.z.array(zod_1.z.number()).optional(),
    entryPoint: zod_1.z.string().optional(),
});
exports.analyzeRepoSchema = zod_1.z.object({
    repositoryPath: schemas_1.repositoryPath,
    ...schemas_1.analysisOptions,
    modules: zod_1.z
        .array(moduleInfo)
        .optional()
        .describe('Optional pre-analyzed modules. If not provided, AI will analyze the repository.'),
});
