// @azure/arm-keyvault/dist/commonjs/models/parameters.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.parameters7 = exports.parameters6 = exports.secretName = exports.properties1 = exports.parameters5 = exports.properties = exports.privateEndpointConnectionName = exports.keyVersion = exports.parameters4 = exports.keyName = exports.name1 = exports.apiVersion1 = exports.filter = exports.operationKind = exports.vaultName2 = exports.parameters3 = exports.parameters2 = exports.parameters1 = exports.vaultName1 = exports.vaultName = exports.parameters = exports.resourceGroupName = exports.top = exports.location = exports.name = exports.subscriptionId = exports.mhsmName = exports.contentType = exports.nextLink = exports.apiVersion = exports.$host = exports.accept = void 0;
const mappers_js_1 = require("../models/mappers.js");
exports.accept = {
    parameterPath: "accept",
    mapper: {
        defaultValue: "application/json",
        isConstant: true,
        serializedName: "Accept",
        type: {
            name: "String",
        },
    },
};
exports.$host = {
    parameterPath: "$host",
    mapper: {
        serializedName: "$host",
        required: true,
        type: {
            name: "String",
        },
    },
    skipEncoding: true,
};
exports.apiVersion = {
    parameterPath: "apiVersion",
    mapper: {
        defaultValue: "2025-05-01",
        isConstant: true,
        serializedName: "api-version",
        type: {
            name: "String",
        },
    },
};
exports.nextLink = {
    parameterPath: "nextLink",
    mapper: {
        serializedName: "nextLink",
        required: true,
        type: {
            name: "String",
        },
    },
    skipEncoding: true,
};
exports.contentType = {
    parameterPath: ["options", "contentType"],
    mapper: {
        defaultValue: "application/json",
        isConstant: true,
        serializedName: "Content-Type",
        type: {
            name: "String",
        },
    },
};
exports.mhsmName = {
    parameterPath: "mhsmName",
    mapper: mappers_js_1.CheckMhsmNameAvailabilityParameters,
};
exports.subscriptionId = {
    parameterPath: "subscriptionId",
    mapper: {
        serializedName: "subscriptionId",
        required: true,
        type: {
            name: "Uuid",
        },
    },
};
exports.name = {
    parameterPath: "name",
    mapper: {
        serializedName: "name",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.location = {
    parameterPath: "location",
    mapper: {
        constraints: {
            MinLength: 1,
        },
        serializedName: "location",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.top = {
    parameterPath: ["options", "top"],
    mapper: {
        serializedName: "$top",
        type: {
            name: "Number",
        },
    },
};
exports.resourceGroupName = {
    parameterPath: "resourceGroupName",
    mapper: {
        constraints: {
            MaxLength: 90,
            MinLength: 1,
        },
        serializedName: "resourceGroupName",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.parameters = {
    parameterPath: "parameters",
    mapper: mappers_js_1.ManagedHsm,
};
exports.vaultName = {
    parameterPath: "vaultName",
    mapper: mappers_js_1.VaultCheckNameAvailabilityParameters,
};
exports.vaultName1 = {
    parameterPath: "vaultName",
    mapper: {
        serializedName: "vaultName",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.parameters1 = {
    parameterPath: "parameters",
    mapper: mappers_js_1.VaultCreateOrUpdateParameters,
};
exports.parameters2 = {
    parameterPath: "parameters",
    mapper: mappers_js_1.VaultPatchParameters,
};
exports.parameters3 = {
    parameterPath: "parameters",
    mapper: mappers_js_1.VaultAccessPolicyParameters,
};
exports.vaultName2 = {
    parameterPath: "vaultName",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[a-zA-Z0-9-]{3,24}$"),
        },
        serializedName: "vaultName",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.operationKind = {
    parameterPath: "operationKind",
    mapper: {
        serializedName: "operationKind",
        required: true,
        type: {
            name: "Enum",
            allowedValues: ["add", "replace", "remove"],
        },
    },
};
exports.filter = {
    parameterPath: "filter",
    mapper: {
        defaultValue: "resourceType eq 'Microsoft.KeyVault/vaults'",
        isConstant: true,
        serializedName: "$filter",
        type: {
            name: "String",
        },
    },
};
exports.apiVersion1 = {
    parameterPath: "apiVersion",
    mapper: {
        defaultValue: "2015-11-01",
        isConstant: true,
        serializedName: "api-version",
        type: {
            name: "String",
        },
    },
};
exports.name1 = {
    parameterPath: "name",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[A-Za-z]([A-Za-z0-9]|\\-[A-Za-z0-9])+$"),
            MaxLength: 24,
            MinLength: 3,
        },
        serializedName: "name",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.keyName = {
    parameterPath: "keyName",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[a-zA-Z0-9-]{1,127}$"),
        },
        serializedName: "keyName",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.parameters4 = {
    parameterPath: "parameters",
    mapper: mappers_js_1.ManagedHsmKeyCreateParameters,
};
exports.keyVersion = {
    parameterPath: "keyVersion",
    mapper: {
        constraints: {
            Pattern: new RegExp("^[a-fA-F0-9]{32}$"),
        },
        serializedName: "keyVersion",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.privateEndpointConnectionName = {
    parameterPath: "privateEndpointConnectionName",
    mapper: {
        serializedName: "privateEndpointConnectionName",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.properties = {
    parameterPath: "properties",
    mapper: mappers_js_1.MhsmPrivateEndpointConnection,
};
exports.parameters5 = {
    parameterPath: "parameters",
    mapper: mappers_js_1.KeyCreateParameters,
};
exports.properties1 = {
    parameterPath: "properties",
    mapper: mappers_js_1.PrivateEndpointConnection,
};
exports.secretName = {
    parameterPath: "secretName",
    mapper: {
        serializedName: "secretName",
        required: true,
        type: {
            name: "String",
        },
    },
};
exports.parameters6 = {
    parameterPath: "parameters",
    mapper: mappers_js_1.SecretCreateOrUpdateParameters,
};
exports.parameters7 = {
    parameterPath: "parameters",
    mapper: mappers_js_1.SecretPatchParameters,
};
