// codetocloudmcp/out/mcpServers/toolsImplementations/diagnoseExistingResources/containerRegistryToAcaDiagnosticHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContainerRegistryToAcaDiagnosticHandler = void 0;
class ContainerRegistryToAcaDiagnosticHandler {
    stateManager;
    constructor(stateManager) {
        this.stateManager = stateManager;
    }
    async diagnoseResource(containerAppId, registry) {
        let containerApp;
        // Parse container app ID
        const appRegex = /^\/subscriptions\/([^\/]+)\/resourceGroups\/([^\/]+)\/providers\/Microsoft\.App\/containerApps\/([^\/]+)$/i;
        const appMatch = containerAppId.match(appRegex);
        if (!appMatch) {
            return {
                success: false,
                message: `Invalid containerAppId format: ${containerAppId}. Expected format: /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.App/containerApps/{containerAppName}`
            };
        }
        const [, subscriptionId, resourceGroupName, containerAppName] = appMatch;
        try {
            // Get container app
            containerApp = await this.stateManager.getAzureClient(subscriptionId).getContainerApp(resourceGroupName, containerAppName);
        }
        catch (error) {
            return {
                success: false,
                message: `Fail to access Container App ${containerAppId}: ${error}`
            };
        }
        // Check registry configuration
        const acaRegistryInfo = containerApp.registryCredentials.find(r => r.server?.toLowerCase() === registry.loginServer?.toLowerCase());
        if (!acaRegistryInfo) {
            return {
                success: false,
                message: `Container App ${containerApp.name} does not have the registry ${registry.name} configured.`,
                recommendations: [
                    "Use user-assigned managed identity to access the registry.",
                    `Configure the registry in Container App: az containerapp registry set -g ${resourceGroupName} -n ${containerAppName} --server ${registry.loginServer} --identity <userAssignedIdentityResourceId>`
                ]
            };
        }
        // Check authentication configuration
        return await this.checkAuthenticationConfiguration(containerApp, acaRegistryInfo, registry.id, subscriptionId, resourceGroupName, containerAppName, registry.name);
    }
    async checkAuthenticationConfiguration(containerApp, acaRegistryInfo, containerRegistryId, subscriptionId, resourceGroupName, containerAppName, registryName) {
        const recommendations = [];
        if (acaRegistryInfo.identity === "system") {
            return await this.checkSystemAssignedIdentity(containerApp, acaRegistryInfo, containerRegistryId, subscriptionId, resourceGroupName, containerAppName, registryName);
        }
        else if (acaRegistryInfo.identity) {
            return await this.checkUserAssignedIdentity(containerApp, acaRegistryInfo, containerRegistryId, subscriptionId, resourceGroupName, containerAppName, registryName);
        }
        else if (acaRegistryInfo.passwordSecretRef) {
            return this.checkPasswordSecret(containerApp, acaRegistryInfo, resourceGroupName, containerAppName, registryName);
        }
        else {
            return {
                success: false,
                message: `Container App ${containerAppName} is using unknown identity type to access the registry ${registryName}.`,
                recommendations: [
                    "Recommend to use user managed identity.",
                    "Prepare a user assigned identity: create a new one or find an existing one.",
                    `Assign a user assigned identity to the Container App: az containerapp identity assign -g ${resourceGroupName} -n ${containerAppName} --user-assigned <userAssignedIdentityResourceId>`,
                    `Assign the acrpull role to the identity: az role assignment create --assignee-object-id <identityInfo.principalId> --role acrpull --scope ${containerRegistryId}`
                ]
            };
        }
    }
    async checkSystemAssignedIdentity(containerApp, acaRegistryInfo, containerRegistryId, subscriptionId, resourceGroupName, containerAppName, registryName) {
        if (!containerApp.identity?.principalId) {
            return {
                success: false,
                message: `Container App ${containerAppName} is using system assigned identity to access the registry ${registryName}, but the identity is not enabled.`,
                recommendations: [
                    `Enable the system assigned identity: az containerapp identity assign -g ${resourceGroupName} -n ${containerAppName} --system-assigned`,
                    `Assign the acrpull role: az role assignment create --assignee-object-id <identity.principalId> --role acrpull --scope ${containerRegistryId}`
                ]
            };
        }
        try {
            const roles = await this.stateManager.getAzureClient(subscriptionId).listRoleAssignmentsForScope(containerRegistryId, `principalId eq '${containerApp.identity.principalId}'`);
            let hasAcrPull = false;
            for await (const role of roles) {
                if (role.roleDefinitionId?.toLowerCase().endsWith("/roledefinitions/7f951dda-4ed3-4680-a7ca-43fe172d538d")) {
                    hasAcrPull = true;
                    break;
                }
            }
            if (!hasAcrPull) {
                return {
                    success: false,
                    message: `System assigned identity does not have acrpull role assignment to the registry ${registryName}.`,
                    recommendations: [
                        `Assign the acrpull role: az role assignment create --assignee-object-id ${containerApp.identity.principalId} --role acrpull --scope ${containerRegistryId}`
                    ]
                };
            }
        }
        catch (error) {
            return {
                success: false,
                message: `Failed to check role assignments for the system assigned identity of Container App ${containerAppName}: ${error}`,
                recommendations: []
            };
        }
        return {
            success: true,
            message: `Container App ${containerAppName} is correctly configured to access registry ${registryName} using system assigned identity.`
        };
    }
    async checkUserAssignedIdentity(containerApp, acaRegistryInfo, containerRegistryId, subscriptionId, resourceGroupName, containerAppName, registryName) {
        // Parse user assigned identity ID
        const identityRegex = /^\/subscriptions\/([^\/]+)\/resourceGroups\/([^\/]+)\/providers\/Microsoft\.ManagedIdentity\/userAssignedIdentities\/([^\/]+)$/i;
        const identityMatch = acaRegistryInfo.identity?.match(identityRegex);
        const acaRegistryIdentityId = acaRegistryInfo?.identity;
        if (!identityMatch) {
            return {
                success: false,
                message: `Invalid user assigned identity format: ${acaRegistryInfo.identity}. Expected format: /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ManagedIdentity/userAssignedIdentities/{identityName}`
            };
        }
        const [, identitySubscriptionId, , identityName] = identityMatch;
        const existingUserMi = containerApp.identity?.userAssignedIdentities;
        // Find the identity with case-insensitive comparison since Azure resource IDs can have different casing
        let identityInfo;
        if (existingUserMi) {
            const normalizedRegistryIdentityId = acaRegistryIdentityId.toLowerCase();
            for (const [key, value] of Object.entries(existingUserMi)) {
                if (key.toLowerCase() === normalizedRegistryIdentityId) {
                    identityInfo = value;
                    break;
                }
            }
        }
        if (!identityInfo) {
            return {
                success: false,
                message: `Container App ${containerAppName} is using user assigned identity ${identityName} to access the registry ${registryName}, but the identity is not enabled.`,
                recommendations: [
                    `Enable the user assigned identity: az containerapp identity assign -g ${resourceGroupName} -n ${containerAppName} --user-assigned ${acaRegistryInfo.identity}`,
                    `Assign the acrpull role: az role assignment create --assignee-object-id <identityInfo.principalId> --role acrpull --scope ${containerRegistryId}`
                ]
            };
        }
        try {
            const roles = await this.stateManager.getAzureClient(subscriptionId).listRoleAssignmentsForScope(containerRegistryId, `principalId eq '${identityInfo.principalId}'`);
            let hasAcrPull = false;
            for await (const role of roles) {
                if (role.roleDefinitionId?.toLowerCase().endsWith("/roledefinitions/7f951dda-4ed3-4680-a7ca-43fe172d538d")) {
                    hasAcrPull = true;
                    break;
                }
            }
            if (!hasAcrPull) {
                return {
                    success: false,
                    message: `User assigned identity does not have acrpull role assignment to the registry ${registryName}.`,
                    recommendations: [
                        `Assign the acrpull role: az role assignment create --assignee-object-id ${identityInfo.principalId} --role acrpull --scope ${containerRegistryId}`
                    ]
                };
            }
        }
        catch (error) {
            return {
                success: false,
                message: `Failed to check role assignments for the user assigned identity ${identityName} of Container App ${containerAppName}: ${error}`,
                recommendations: []
            };
        }
        return {
            success: true,
            message: `Container App ${containerAppName} is correctly configured to access registry ${registryName} using user assigned identity ${identityName}.`
        };
    }
    checkPasswordSecret(containerApp, registryInfo, resourceGroupName, containerAppName, registryName) {
        const hasSecret = containerApp.secrets.some((s) => s.name === registryInfo.passwordSecretRef);
        if (!hasSecret) {
            return {
                success: false,
                message: `Container App ${containerAppName} is using secret reference ${registryInfo.passwordSecretRef} for the registry ${registryName}, but the secret is not found in the Container App.`,
                recommendations: [
                    `Add the secret to the Container App: az containerapp secret set -g ${resourceGroupName} --name ${containerAppName} --secrets ${registryInfo.passwordSecretRef}=<secretValue>`
                ]
            };
        }
        return {
            success: true,
            message: `Container App ${containerAppName} is correctly configured to access registry ${registryName} using password secret.`
        };
    }
}
exports.ContainerRegistryToAcaDiagnosticHandler = ContainerRegistryToAcaDiagnosticHandler;
