// codetocloudmcp/out/mcpServers/toolsImplementations/diagnoseExistingResources/containerRegistryToAksDiagnosticHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContainerRegistryToAksDiagnosticHandler = void 0;
class ContainerRegistryToAksDiagnosticHandler {
    stateManager;
    constructor(stateManager) {
        this.stateManager = stateManager;
    }
    async diagnoseResource(aksClusterId, registry) {
        try {
            // Parse AKS cluster ID
            const aksRegex = /^\/subscriptions\/([^\/]+)\/resourceGroups\/([^\/]+)\/providers\/Microsoft\.ContainerService\/managedClusters\/([^\/]+)$/i;
            const aksMatch = aksClusterId.match(aksRegex);
            if (!aksMatch) {
                return {
                    success: false,
                    message: `Invalid AKS cluster ID format: ${aksClusterId}. Expected format: /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ContainerService/managedClusters/{clusterName}`
                };
            }
            const [, subscriptionId, resourceGroupName, clusterName] = aksMatch;
            // Get AKS cluster
            const aksCluster = await this.stateManager.getAzureClient(subscriptionId).getAKSCluster(resourceGroupName, clusterName);
            return await this.checkAuthenticationConfiguration(aksCluster, registry, subscriptionId, resourceGroupName, clusterName);
        }
        catch (error) {
            return {
                success: false,
                message: `Error accessing AKS cluster ${aksClusterId}: ${error}`
            };
        }
    }
    async checkAuthenticationConfiguration(aksCluster, registry, subscriptionId, resourceGroupName, clusterName) {
        // Check if ACR admin user is enabled (manual authentication)
        if (registry.adminUserEnabled) {
            return this.checkImagePullSecrets(aksCluster, registry, subscriptionId, resourceGroupName, clusterName);
        }
        else {
            if (!aksCluster.kubeletIdentityObjectId) {
                return {
                    success: false,
                    message: `AKS cluster ${clusterName} does not have a kubelet managed identity configured for ACR access.`,
                    recommendations: [
                        "Kubelet identity is required for AKS to pull images from ACR.",
                        `Recommend to use attach acr to automatically configure kubelet identity: az aks update --name ${clusterName} --resource-group ${resourceGroupName} --attach-acr ${registry.name}`
                    ]
                };
            }
            // Check kubelet identity for ACR access (this is what actually pulls images)
            return this.checkKubeletIdentity(aksCluster, registry, subscriptionId, resourceGroupName, clusterName);
        }
    }
    async checkKubeletIdentity(aksCluster, registry, subscriptionId, resourceGroupName, clusterName) {
        if (!aksCluster.kubeletIdentityObjectId) {
            return {
                success: false,
                message: `AKS cluster ${clusterName} kubelet identity is not properly configured.`,
                recommendations: [
                    `Attach ACR to AKS cluster: az aks update --name ${clusterName} --resource-group ${resourceGroupName} --attach-acr ${registry.name}`
                ]
            };
        }
        try {
            const roles = await this.stateManager.getAzureClient(subscriptionId).listRoleAssignmentsForScope(registry.id, `principalId eq '${aksCluster.kubeletIdentityObjectId}'`);
            let hasAcrPull = false;
            for await (const role of roles) {
                if (role.roleDefinitionId?.toLowerCase().endsWith("/roledefinitions/7f951dda-4ed3-4680-a7ca-43fe172d538d")) {
                    hasAcrPull = true;
                    break;
                }
            }
            if (!hasAcrPull) {
                return {
                    success: false,
                    message: `AKS cluster ${clusterName} kubelet identity does not have AcrPull role to access registry ${registry.name}.`,
                    recommendations: [
                        `Attach ACR to AKS cluster: az aks update --name ${clusterName} --resource-group ${resourceGroupName} --attach-acr ${registry.name}`,
                        `Or manually assign AcrPull role: az role assignment create --assignee ${aksCluster.kubeletIdentityObjectId} --role acrpull --scope ${registry.id}`
                    ]
                };
            }
        }
        catch (error) {
            return {
                success: false,
                message: `Failed to check kubelet identity role assignments for AKS cluster ${clusterName}: ${error}`,
            };
        }
        return {
            success: true,
            message: `AKS cluster ${clusterName} is properly configured to access ACR ${registry.name} using kubelet identity.`
        };
    }
    async checkImagePullSecrets(aksCluster, registry, subscriptionId, resourceGroupName, clusterName) {
        return {
            success: true,
            message: `
      ACR ${registry.name} admin user is enabled for imagePullSecrets authentication, but need more manual verification and setup. 
      To complete the setup: 1) Get ACR credentials: 'az acr credential show --name ${registry.name}' 
      2) Create Kubernetes secret: 'kubectl create secret docker-registry acr-secret --docker-server=${registry.loginServer} --docker-username=<admin-username> --docker-password=<admin-password>' 
      3) Reference the secret in your Pod/Deployment spec under 'imagePullSecrets'. Verify existing secrets with: 'kubectl get secrets --all-namespaces --field-selector type=kubernetes.io/dockerconfigjson'
      `
        };
    }
}
exports.ContainerRegistryToAksDiagnosticHandler = ContainerRegistryToAksDiagnosticHandler;
