// containerization-assist-mcp/dist-cjs/src/infra/kubernetes/idempotent-apply.js
"use strict";
/**
 * Idempotent Kubernetes resource application using create/patch strategy
 *
 * Note: This function should be called sequentially per resource.
 * Resources are created if they don't exist, or patched if they already exist.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.createIdempotentApply = createIdempotentApply;
exports.parseManifests = parseManifests;
const k8s = __importStar(require("@kubernetes/client-node"));
const yaml = __importStar(require("js-yaml"));
const resource_operations_1 = require("./resource-operations");
/**
 * Creates an idempotent K8s apply function using the consolidated resource-operations module
 *
 * Note: This function uses the consolidated resource-operations module which handles
 * idempotent apply using create/patch strategy. Resources are created if they don't exist,
 * or patched if they already exist.
 */
function createIdempotentApply(logger, kubeconfig) {
    const kc = new k8s.KubeConfig();
    if (kubeconfig) {
        kc.loadFromString(kubeconfig);
    }
    else {
        kc.loadFromDefault();
    }
    /**
     * Apply K8s resource using the consolidated resource-operations module
     */
    return async function applyResource(resource, _options = {}) {
        // Use the consolidated resource operations module
        return (0, resource_operations_1.applyResource)(kc, resource, logger);
    };
}
/**
 * Parse YAML manifests into K8s resources
 */
function parseManifests(yamlContent) {
    try {
        // In js-yaml v4, loadAll is safe by default (no code execution)
        const docs = yaml.loadAll(yamlContent);
        return docs.filter((doc) => {
            const resource = doc;
            return resource?.kind && resource.apiVersion;
        });
    }
    catch {
        return [];
    }
}
