// containerization-assist-mcp/dist-cjs/src/infra/kubernetes/kubeconfig-discovery.js
"use strict";
/**
 * Kubeconfig discovery and validation utilities
 *
 * Simplifies kubeconfig detection with fast-fail error handling
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.discoverKubeconfigPath = discoverKubeconfigPath;
exports.validateKubeconfig = validateKubeconfig;
exports.discoverAndValidateKubeconfig = discoverAndValidateKubeconfig;
exports.isInCluster = isInCluster;
const fs = __importStar(require("node:fs"));
const os = __importStar(require("node:os"));
const path = __importStar(require("node:path"));
const k8s = __importStar(require("@kubernetes/client-node"));
const types_1 = require("../../types");
/**
 * Discover kubeconfig location following kubectl's logic:
 * 1. KUBECONFIG environment variable
 * 2. ~/.kube/config
 * 3. In-cluster config (not applicable for single-user scenarios)
 */
function discoverKubeconfigPath() {
    // Check KUBECONFIG environment variable
    const kubeconfigEnv = process.env.KUBECONFIG;
    if (kubeconfigEnv) {
        // Handle multiple paths separated by colon (kubectl behavior)
        const paths = kubeconfigEnv.split(path.delimiter);
        for (const configPath of paths) {
            if (fs.existsSync(configPath)) {
                return (0, types_1.Success)(configPath);
            }
        }
        return (0, types_1.Failure)(`KUBECONFIG environment variable is set but file not found: ${kubeconfigEnv}`, {
            message: 'Kubeconfig file not found',
            hint: 'KUBECONFIG environment variable points to a non-existent file',
            resolution: 'Verify the path in KUBECONFIG is correct, or unset it to use the default ~/.kube/config location.',
            details: { kubeconfigEnv },
        });
    }
    // Check default location
    const defaultPath = path.join(os.homedir(), '.kube', 'config');
    if (fs.existsSync(defaultPath)) {
        return (0, types_1.Success)(defaultPath);
    }
    return (0, types_1.Failure)('Kubeconfig not found', {
        message: 'No kubeconfig file found',
        hint: 'Neither KUBECONFIG environment variable nor ~/.kube/config exists',
        resolution: 'Set up kubectl configuration: run `kubectl config view` to check current config, or configure access to your cluster using cloud provider CLI (e.g., `aws eks update-kubeconfig`, `gcloud container clusters get-credentials`, or `az aks get-credentials`).',
        details: { defaultPath },
    });
}
/**
 * Validate that a kubeconfig file is readable and parseable
 */
function validateKubeconfig(configPath) {
    try {
        // Check file exists
        if (!fs.existsSync(configPath)) {
            return (0, types_1.Failure)(`Kubeconfig file does not exist: ${configPath}`, {
                message: 'Kubeconfig file not found',
                hint: `File does not exist at path: ${configPath}`,
                resolution: 'Verify the kubeconfig path is correct and the file exists.',
                details: { configPath },
            });
        }
        // Check file is readable
        try {
            fs.accessSync(configPath, fs.constants.R_OK);
        }
        catch (error) {
            return (0, types_1.Failure)(`Kubeconfig file is not readable: ${configPath}`, {
                message: 'Cannot read kubeconfig file',
                hint: 'Insufficient permissions to read the kubeconfig file',
                resolution: `Check file permissions: \`chmod 600 ${configPath}\` or verify file ownership.`,
                details: { configPath, error: String(error) },
            });
        }
        // Parse the kubeconfig
        const kc = new k8s.KubeConfig();
        try {
            kc.loadFromFile(configPath);
        }
        catch (error) {
            return (0, types_1.Failure)(`Failed to parse kubeconfig: ${configPath}`, {
                message: 'Invalid kubeconfig format',
                hint: 'The kubeconfig file could not be parsed',
                resolution: 'Verify the file is valid YAML. Run `kubectl config view` to check for syntax errors.',
                details: { configPath, error: String(error) },
            });
        }
        // Get current context info
        const currentContext = kc.getCurrentContext();
        if (!currentContext) {
            return (0, types_1.Failure)('No current context set in kubeconfig', {
                message: 'No active Kubernetes context',
                hint: 'The kubeconfig file has no current context configured',
                resolution: 'Set a context: `kubectl config use-context <context-name>` or `kubectl config get-contexts` to list available contexts.',
                details: { configPath },
            });
        }
        const cluster = kc.getCurrentCluster();
        const user = kc.getCurrentUser();
        return (0, types_1.Success)({
            path: configPath,
            contextName: currentContext,
            clusterName: cluster?.name || 'unknown',
            user: user?.name || 'unknown',
        });
    }
    catch (error) {
        return (0, types_1.Failure)(`Kubeconfig validation failed: ${String(error)}`, {
            message: 'Failed to validate kubeconfig',
            hint: 'An unexpected error occurred while validating the kubeconfig',
            resolution: 'Check the kubeconfig file format and permissions.',
            details: { configPath, error: String(error) },
        });
    }
}
/**
 * Discover and validate kubeconfig in one step
 */
function discoverAndValidateKubeconfig() {
    const pathResult = discoverKubeconfigPath();
    if (!pathResult.ok) {
        return pathResult;
    }
    return validateKubeconfig(pathResult.value);
}
/**
 * Check if running in-cluster (Kubernetes pod)
 * Not applicable for single-user scenarios but included for completeness
 */
function isInCluster() {
    return (fs.existsSync('/var/run/secrets/kubernetes.io/serviceaccount/token') &&
        fs.existsSync('/var/run/secrets/kubernetes.io/serviceaccount/ca.crt'));
}
