// containerization-assist-mcp/dist/src/mcp/context.js
/**
 * MCP Context - Tool execution environment abstraction
 *
 * Invariant: All tools receive consistent context interface
 * Trade-off: Abstraction overhead for tool isolation and testability
 * Design: Factory pattern enables context mocking in tests
 */
import { extractProgressReporter } from './context-helpers.js';
export { extractProgressToken, createProgressReporter } from './context-helpers.js';
/**
 * Create a ToolContext for tool execution
 *
 * @param logger - Logger for debugging and error tracking
 * @param options - Optional configuration
 * @returns Configured ToolContext
 *
 * @example
 * ```typescript
 * const context = createToolContext(logger, {
 *   signal: abortController.signal,
 *   progress: async (msg) => console.log(msg),
 * });
 * ```
 */
export function createToolContext(logger, options = {}) {
    const progressReporter = extractProgressReporter(options.progress, logger, options.sendNotification);
    return {
        logger,
        signal: options.signal,
        progress: progressReporter,
        ...(options.policy && { policy: options.policy }),
        queryConfig: async (packageName, input) => {
            if (!options.policy) {
                logger.debug({ packageName }, 'No policy configured, returning null for config query');
                return null;
            }
            return options.policy.queryConfig(packageName, input);
        },
    };
}
