// containerization-assist-mcp/dist-cjs/src/types/tool-metadata.js
"use strict";
/**
 * Tool Metadata Contract & Validation
 *
 * Provides Zod-based validation for tool metadata.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateAllToolMetadata = validateAllToolMetadata;
const zod_1 = require("zod");
const core_1 = require("./core");
/**
 * Tool Metadata Schema
 *
 * All tools must specify whether they use knowledge enhancement.
 * @internal - Only used internally for validation
 */
const ToolMetadataSchema = zod_1.z.object({
    /** Whether this tool uses knowledge enhancement (required) */
    knowledgeEnhanced: zod_1.z.boolean(),
});
/**
 * Validates tool metadata against the schema
 *
 * @param metadata - Unknown metadata object to validate
 * @returns Result containing validated metadata or validation errors
 * @internal - Only used by validateAllToolMetadata
 */
function validateToolMetadata(metadata) {
    try {
        const validatedMetadata = ToolMetadataSchema.parse(metadata);
        return (0, core_1.Success)(validatedMetadata);
    }
    catch (error) {
        if (error instanceof zod_1.z.ZodError) {
            const issues = error.issues
                .map((issue) => `${issue.path.join('.')}: ${issue.message}`)
                .join(', ');
            return (0, core_1.Failure)(`Tool metadata validation failed: ${issues}`);
        }
        return (0, core_1.Failure)(`Tool metadata validation failed: ${error}`);
    }
}
/**
 * Suggestion mapping for common validation issues
 */
const VALIDATION_SUGGESTIONS = {
    'Invalid metadata schema': 'Fix metadata schema validation errors',
};
/**
 * Validates all tool metadata and returns comprehensive report
 * Centralized validation logic for use by CLI commands
 */
async function validateAllToolMetadata(tools) {
    try {
        const validTools = [];
        const invalidTools = [];
        const metadataErrors = [];
        for (const tool of tools) {
            // Validate metadata schema compliance
            const metadataValidation = validateToolMetadata(tool.metadata);
            if (!metadataValidation.ok) {
                metadataErrors.push({
                    name: tool.name,
                    error: metadataValidation.error,
                });
                invalidTools.push({
                    name: tool.name,
                    issues: ['Invalid metadata schema'],
                    suggestions: [
                        VALIDATION_SUGGESTIONS['Invalid metadata schema'] ??
                            'Fix metadata schema validation errors',
                    ],
                });
            }
            else {
                validTools.push(tool.name);
            }
        }
        const totalTools = tools.length;
        const validCount = validTools.length;
        const compliancePercentage = Math.round((validCount / totalTools) * 100);
        return (0, core_1.Success)({
            summary: {
                totalTools,
                validTools: validCount,
                invalidTools: totalTools - validCount,
                compliancePercentage,
            },
            validTools,
            invalidTools,
            metadataErrors,
        });
    }
    catch (error) {
        return (0, core_1.Failure)(`Validation failed: ${error instanceof Error ? error.message : String(error)}`);
    }
}
