// containerization-assist-mcp/dist-cjs/src/tools/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifyDeployTool = exports.tagImageTool = exports.scanImageTool = exports.pushImageTool = exports.prepareClusterTool = exports.opsTool = exports.generateK8sManifestsTool = exports.generateDockerfileTool = exports.fixDockerfileTool = exports.buildImageTool = exports.analyzeRepoTool = exports.TOOL_NAME = exports.ALL_TOOLS = void 0;
const tool_1 = __importDefault(require("./analyze-repo/tool"));
exports.analyzeRepoTool = tool_1.default;
const tool_2 = __importDefault(require("./build-image/tool"));
exports.buildImageTool = tool_2.default;
const tool_3 = __importDefault(require("./fix-dockerfile/tool"));
exports.fixDockerfileTool = tool_3.default;
const tool_4 = __importDefault(require("./generate-dockerfile/tool"));
exports.generateDockerfileTool = tool_4.default;
const tool_5 = __importDefault(require("./generate-k8s-manifests/tool"));
exports.generateK8sManifestsTool = tool_5.default;
const tool_6 = __importDefault(require("./ops/tool"));
exports.opsTool = tool_6.default;
const tool_7 = __importDefault(require("./prepare-cluster/tool"));
exports.prepareClusterTool = tool_7.default;
const tool_8 = __importDefault(require("./push-image/tool"));
exports.pushImageTool = tool_8.default;
const tool_9 = __importDefault(require("./scan-image/tool"));
exports.scanImageTool = tool_9.default;
const tool_10 = __importDefault(require("./tag-image/tool"));
exports.tagImageTool = tool_10.default;
const tool_11 = __importDefault(require("./verify-deploy/tool"));
exports.verifyDeployTool = tool_11.default;
const TOOL_NAME = {
    ANALYZE_REPO: 'analyze-repo',
    BUILD_IMAGE: 'build-image',
    FIX_DOCKERFILE: 'fix-dockerfile',
    GENERATE_DOCKERFILE: 'generate-dockerfile',
    GENERATE_K8S_MANIFESTS: 'generate-k8s-manifests',
    OPS: 'ops',
    PREPARE_CLUSTER: 'prepare-cluster',
    PUSH_IMAGE: 'push-image',
    SCAN_IMAGE: 'scan-image',
    TAG_IMAGE: 'tag-image',
    VERIFY_DEPLOY: 'verify-deploy',
};
exports.TOOL_NAME = TOOL_NAME;
// Ensure proper names on all tools
tool_1.default.name = TOOL_NAME.ANALYZE_REPO;
tool_2.default.name = TOOL_NAME.BUILD_IMAGE;
tool_3.default.name = TOOL_NAME.FIX_DOCKERFILE;
tool_4.default.name = TOOL_NAME.GENERATE_DOCKERFILE;
tool_5.default.name = TOOL_NAME.GENERATE_K8S_MANIFESTS;
tool_6.default.name = TOOL_NAME.OPS;
tool_7.default.name = TOOL_NAME.PREPARE_CLUSTER;
tool_8.default.name = TOOL_NAME.PUSH_IMAGE;
tool_9.default.name = TOOL_NAME.SCAN_IMAGE;
tool_10.default.name = TOOL_NAME.TAG_IMAGE;
tool_11.default.name = TOOL_NAME.VERIFY_DEPLOY;
// Type-safe tool array using the union type
// All tools are now deterministic plan-based or operational tools
exports.ALL_TOOLS = [
    // Plan-based generation tools (use knowledge to create plans)
    tool_1.default,
    tool_3.default,
    tool_4.default,
    tool_5.default,
    // Operational/deterministic tools
    tool_2.default,
    tool_6.default,
    tool_7.default,
    tool_8.default,
    tool_9.default,
    tool_10.default,
    tool_11.default,
];
