// containerization-assist-mcp/dist-cjs/src/tools/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ALL_TOOL_DEFINITIONS = exports.TOOL_NAME = exports.verifyDeployToolDefinition = exports.tagImageToolDefinition = exports.scanImageToolDefinition = exports.pushImageToolDefinition = exports.prepareClusterToolDefinition = exports.opsToolDefinition = exports.generateK8sManifestsToolDefinition = exports.generateDockerfileToolDefinition = exports.fixDockerfileToolDefinition = exports.buildImageContextToolDefinition = exports.analyzeRepoToolDefinition = void 0;
const types_1 = require("./analyze-repo/types");
Object.defineProperty(exports, "analyzeRepoToolDefinition", { enumerable: true, get: function () { return types_1.analyzeRepoToolDefinition; } });
const types_2 = require("./build-image-context/types");
Object.defineProperty(exports, "buildImageContextToolDefinition", { enumerable: true, get: function () { return types_2.buildImageContextToolDefinition; } });
const types_3 = require("./fix-dockerfile/types");
Object.defineProperty(exports, "fixDockerfileToolDefinition", { enumerable: true, get: function () { return types_3.fixDockerfileToolDefinition; } });
const types_4 = require("./generate-dockerfile/types");
Object.defineProperty(exports, "generateDockerfileToolDefinition", { enumerable: true, get: function () { return types_4.generateDockerfileToolDefinition; } });
const types_5 = require("./generate-k8s-manifests/types");
Object.defineProperty(exports, "generateK8sManifestsToolDefinition", { enumerable: true, get: function () { return types_5.generateK8sManifestsToolDefinition; } });
const types_6 = require("./ops/types");
Object.defineProperty(exports, "opsToolDefinition", { enumerable: true, get: function () { return types_6.opsToolDefinition; } });
const types_7 = require("./prepare-cluster/types");
Object.defineProperty(exports, "prepareClusterToolDefinition", { enumerable: true, get: function () { return types_7.prepareClusterToolDefinition; } });
const types_8 = require("./push-image/types");
Object.defineProperty(exports, "pushImageToolDefinition", { enumerable: true, get: function () { return types_8.pushImageToolDefinition; } });
const types_9 = require("./scan-image/types");
Object.defineProperty(exports, "scanImageToolDefinition", { enumerable: true, get: function () { return types_9.scanImageToolDefinition; } });
const types_10 = require("./tag-image/types");
Object.defineProperty(exports, "tagImageToolDefinition", { enumerable: true, get: function () { return types_10.tagImageToolDefinition; } });
const types_11 = require("./verify-deploy/types");
Object.defineProperty(exports, "verifyDeployToolDefinition", { enumerable: true, get: function () { return types_11.verifyDeployToolDefinition; } });
var toolDefinition_1 = require("./shared/toolDefinition");
Object.defineProperty(exports, "TOOL_NAME", { enumerable: true, get: function () { return toolDefinition_1.TOOL_NAME; } });
exports.ALL_TOOL_DEFINITIONS = [
    types_1.analyzeRepoToolDefinition,
    types_2.buildImageContextToolDefinition,
    types_3.fixDockerfileToolDefinition,
    types_4.generateDockerfileToolDefinition,
    types_5.generateK8sManifestsToolDefinition,
    types_6.opsToolDefinition,
    types_7.prepareClusterToolDefinition,
    types_8.pushImageToolDefinition,
    types_9.scanImageToolDefinition,
    types_10.tagImageToolDefinition,
    types_11.verifyDeployToolDefinition,
];
