// codetocloudmcp/out/toolsImplementations/getAzureResourceRules/sampleCode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KvCsiYamlExample = exports.AksKvRoleSample = exports.AksTfExample = void 0;
exports.AksTfExample = `
\`\`\`
resource "azurerm_kubernetes_cluster" "aks" {
  name                = azurecaf_name.aks.result
  location            = azurerm_resource_group.rg.location
  resource_group_name = azurerm_resource_group.rg.name
  dns_prefix          = var.project_name
  kubernetes_version  = var.kubernetes_version    // Use a stable version from \`az aks get-versions --query 'values[?isDefault].version' --location <> -o tsv\`

  oidc_issuer_enabled       = true
  workload_identity_enabled = true

  default_node_pool {
    name                = "default"
    node_count          = 2
    vm_size             = "Standard_D2s_v6"  # MUST use a vm_size from appmod-get-available-region-sku output. Do NOT hardcode.
    os_disk_size_gb     = 30
    vnet_subnet_id      = null
  }

  identity {
    type         = "SystemAssigned"  // Use SystemAssigned identity
  }

  network_profile {
    network_plugin    = "kubenet"
    load_balancer_sku = "standard"
  }

  key_vault_secrets_provider {  // Add if using Key Vault
    secret_rotation_enabled = true
  }

  tags = {
    Environment = var.environment
  }

  depends_on = [
  ]
}

# AcrPull role assignment for AKS identity
resource "azurerm_role_assignment" "aks_acr_pull" {
  principal_id                     = azurerm_kubernetes_cluster.aks.kubelet_identity[0].object_id
  role_definition_id               = "/subscriptions/{subscription_id}/providers/Microsoft.Authorization/roleDefinitions/7f951dda-4ed3-4680-a7ca-43fe172d538d"
  scope                            = azurerm_container_registry.acr.id
  skip_service_principal_aad_check = true
  depends_on = [ azurerm_kubernetes_cluster.aks ]  // Important
}
\`\`\`
`;
exports.AksKvRoleSample = `
\`\`\`
# Key Vault Secrets User role assignment for AKS managed identity
resource "azurerm_role_assignment" "secret_user" {
  scope                = azurerm_key_vault.keyvault.id
  role_definition_name = "Key Vault Secrets User"
  principal_id         = azurerm_kubernetes_cluster.aks.key_vault_secrets_provider[0].secret_identity[0].object_id
  depends_on = [ azurerm_kubernetes_cluster.aks ]
}
\`\`\`
`;
exports.KvCsiYamlExample = `
1) Create a SecretProviderClass. Example:
\`\`\`
apiVersion: secrets-store.csi.x-k8s.io/v1   # MUST use x-k8s.io (this is the CRD group name registered in AKS)
kind: SecretProviderClass
metadata:
  name: azure-kvname-user-msi
spec:
  provider: azure
  parameters:
    usePodIdentity: "false"
    useVMManagedIdentity: "true"            # Set to true for using managed identity
    userAssignedIdentityID: \${AKS_KV_IDENTITY_CLIENT_ID}   # The clientID of the AKS Key Vault Secrets Provider identity, it can be got after AKS creation
    keyvaultName: \${KEY_VAULT_NAME}          # The name of the Key Vault
    objects:  |
      array:
        - |
          objectName: postgres-connection-string
          objectType: secret              # object types: secret, key, or cert
    tenantId: \${TENANT_ID}                   # The tenant ID of the key vault
  secretObjects:                          
    - secretName: test-secrets-obj1       # it will create a Kubernetes secret in cluster with the name 'test-secrets-obj1'
      type: Opaque
      data:
      - objectName: postgres-connection-string    # must use the same objectName as above
        key: postgres-connection-string
\`\`\`
2) In the deployment YAML file, mount the secret to pod. Example:
⚠️ IMPORTANT: The CSI driver name in volumes.csi.driver MUST be "secrets-store.csi.k8s.io" (without x- prefix).
    This is different from the SecretProviderClass apiVersion. AKS registers the CRD as x-k8s.io but the CSI driver 
as k8s.io.
\`\`\`
kind: Pod
apiVersion: v1
metadata:
  name: busybox-secrets-store-inline-user-msi
spec:
  containers:
    - name: busybox
      image: registry.k8s.io/e2e-test-images/busybox:1.29-4
      command:
        - "/bin/sleep"
        - "10000"
      env:
      - name: SPRING_DATASOURCE_URL     # Important: set the env variable to use the secret
        valueFrom:
          secretKeyRef:
            name: test-secrets-obj1    # use the secretName defined in secretObjects above
            key: postgres-connection-string  # use the objectName defined in secretObjects above
      volumeMounts:                    # Important: mount the volume
      - name: secrets-store01-inline
        mountPath: "/mnt/secrets-store"
        readOnly: true
  volumes:                            # Important: define the volume
    - name: secrets-store01-inline  
      csi:
        driver: secrets-store.csi.k8s.io  # MUST use k8s.io here (NOT x-k8s.io) - this is the CSI driver
        readOnly: true
        volumeAttributes:
          secretProviderClass: "azure-kvname-user-msi"  # Use the SecretProviderClass name defined above
\`\`\`
`;
