// codetocloudmcp/out/mcpServers/toolsImplementations/getRegionsForDeploymentTool/getRegionsForDeploymentHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRegionsForDeployment = getRegionsForDeployment;
const azureQuotaChecker_1 = require("../azureQuotaCheck/azureQuotaChecker");
const azureRegionChecker_1 = require("../azureRegionCheck/azureRegionChecker");
const util_1 = require("./util");
async function checkQuotaForRegion(region, resourceTypes, azureClient) {
    const resourceProviders = resourceTypes.map(resourceType => resourceType.type.split('/')[0]);
    try {
        const filteredResourceProviders = resourceProviders.filter((provider) => Object.values(azureQuotaChecker_1.ResourceProvider).includes(provider));
        const regionQuotaInfo = await (0, azureQuotaChecker_1.GetAzureQuota)(filteredResourceProviders, azureClient, region);
        const regionQuotaDetails = {};
        let allSupport = true;
        for (const resourceType of resourceTypes) {
            const quotaInfoArray = regionQuotaInfo[resourceType.type.split('/')[0]];
            if (!quotaInfoArray || quotaInfoArray.length === 0) {
                continue;
            }
            const satisfyingQuotas = quotaInfoArray
                .filter(quotaInfo => (quotaInfo.limit - quotaInfo.used) >= resourceType.quota)
                .map(quotaInfo => quotaInfo.name);
            if (satisfyingQuotas.length === 0) {
                // regionQuotaDetails[resourceType.type] = null;
                allSupport = false;
                continue;
            }
            regionQuotaDetails[resourceType.type] = satisfyingQuotas;
        }
        return { region, allSupport, quotaDetails: regionQuotaDetails };
    }
    catch (error) {
        return null;
    }
}
async function getRegionsForDeployment(params, azureClient) {
    const regionStartTime = performance.now();
    const availableRegionsForSubscriptionPromise = (0, azureRegionChecker_1.getAllRegions)(azureClient);
    params = (0, util_1.getDeploymentRegionParams)(params);
    let assignedRegions = params.preferredRegions ? [...new Set(params.preferredRegions.concat(util_1.DefaultPreferredRegions))] : util_1.DefaultPreferredRegions;
    let regionToCheck = Math.min(10, assignedRegions.length);
    assignedRegions = assignedRegions.slice(0, regionToCheck);
    const resourceTypeNames = params.resourceTypes.map(rt => rt.type);
    // Get both common regions and per-resource-type regions in a single efficient call
    const regionsDataPromise = (0, azureRegionChecker_1.getAvailableRegionsByResourceTypes)(resourceTypeNames, azureClient, params.cognitiveServiceProperties, assignedRegions);
    const [availableRegionsForSubscription, regionsData] = await Promise.all([
        availableRegionsForSubscriptionPromise,
        regionsDataPromise
    ]);
    console.log(` [PERF] Total Check Regions innerInvoke execution time: ${(performance.now() - regionStartTime).toFixed(2)}ms`);
    if (!availableRegionsForSubscription) {
        return {
            notAvailable: [],
            available: [],
            notChecked: [],
        };
    }
    // Filter regions by subscription availability
    const availableByResourceType = {};
    for (const [resourceType, regions] of Object.entries(regionsData.regionsByResourceType)) {
        availableByResourceType[resourceType] = regions.filter(region => availableRegionsForSubscription.includes(region));
    }
    let availableRegions = regionsData.commonRegions.filter(region => availableRegionsForSubscription.includes(region));
    let availablePreferredRegions = availableRegions.filter(region => (params.preferredRegions || []).includes(region));
    let defaultPreferredRegionsList = util_1.DefaultPreferredRegions.filter(region => !availablePreferredRegions.includes(region) && availableRegions.includes(region));
    let availableRemainingRegions = availableRegions.filter(region => !availablePreferredRegions.includes(region) && !defaultPreferredRegionsList.includes(region));
    let regionsToCheckForQuota = [...availablePreferredRegions, ...defaultPreferredRegionsList, ...availableRemainingRegions];
    let amountChecked = Math.min(5, regionsToCheckForQuota.length);
    const allCheckedRegions = regionsToCheckForQuota.slice(0, amountChecked);
    const quotaCheckPromises = allCheckedRegions
        .map(region => checkQuotaForRegion(region, params.resourceTypes, azureClient));
    const quotaCheckResults = await Promise.all(quotaCheckPromises);
    let regionsWithSufficientQuota = quotaCheckResults.filter(regionInfo => regionInfo?.allSupport);
    let notChecked = [];
    if (amountChecked < regionsToCheckForQuota.length) {
        notChecked = regionsToCheckForQuota.slice(amountChecked);
    }
    // If no common regions available, check quota for individual resource type regions
    let quotaDetailsByResourceType = {};
    if (regionsWithSufficientQuota.length === 0) {
        for (const [resourceType, regions] of Object.entries(availableByResourceType)) {
            let unCheckedRegions = regions;
            allCheckedRegions.forEach((r, idx) => {
                if (regions.includes(r) && quotaCheckResults[idx]?.quotaDetails && quotaCheckResults[idx]?.quotaDetails[resourceType]) {
                    quotaDetailsByResourceType[resourceType][r] = quotaCheckResults[idx]?.quotaDetails[resourceType];
                    unCheckedRegions = unCheckedRegions.filter(region => region !== r);
                }
            });
            quotaDetailsByResourceType[resourceType] = {};
            const resourceTypeData = params.resourceTypes.find(rt => rt.type === resourceType);
            if (resourceTypeData && unCheckedRegions.length > 0) {
                // Check quota for up to 3 regions per resource type to avoid too many API calls
                const regionsToCheck = unCheckedRegions.slice(0, 3);
                const individualQuotaPromises = regionsToCheck.map(region => checkQuotaForRegion(region, [resourceTypeData], azureClient));
                const individualQuotaResults = await Promise.all(individualQuotaPromises);
                individualQuotaResults.forEach((result, index) => {
                    const region = regionsToCheck[index];
                    if (result && result.quotaDetails && result.quotaDetails[resourceType]) {
                        quotaDetailsByResourceType[resourceType][region] = result.quotaDetails[resourceType];
                    }
                });
            }
        }
    }
    let regionsForDeploymentInfo = {
        notAvailable: availableRegionsForSubscription.filter(region => !regionsWithSufficientQuota.map(info => info?.region).includes(region) &&
            !notChecked.includes(region)),
        available: regionsWithSufficientQuota.filter(info => info !== null),
        notChecked: notChecked,
        availableByResourceType: availableByResourceType,
        quotaDetailsByResourceType: quotaDetailsByResourceType
    };
    console.log(` [PERF] Total innerInvoke execution time: ${(performance.now() - regionStartTime).toFixed(2)}ms`);
    return regionsForDeploymentInfo;
}
