// containerization-assist-mcp/dist-cjs/src/tools/generate-dockerfile/schema.js
"use strict";
/**
 * Schema definition for generate-dockerfile tool
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateDockerfileSchema = exports.DOCKER_PLATFORMS = void 0;
const zod_1 = require("zod");
const schemas_1 = require("../shared/schemas");
Object.defineProperty(exports, "DOCKER_PLATFORMS", { enumerable: true, get: function () { return schemas_1.DOCKER_PLATFORMS; } });
exports.generateDockerfileSchema = zod_1.z.object({
    repositoryPath: schemas_1.repositoryPath.describe('Repository path (automatically normalized to forward slashes on all platforms).'),
    modulePath: zod_1.z
        .string()
        .optional()
        .describe('Module path for monorepo/multi-module projects to locate where the Dockerfile should be generated (automatically normalized to forward slashes).'),
    language: zod_1.z.string().optional().describe('Primary programming language (e.g., "java", "python")'),
    languageVersion: zod_1.z
        .string()
        .optional()
        .describe('Language version (e.g., "17", "3.11", "20"). For Java, this is the java.version from pom.xml or sourceCompatibility from build.gradle.'),
    framework: zod_1.z.string().optional().describe('Framework used (e.g., "spring", "django")'),
    environment: schemas_1.environment.describe('Target environment (production, development, etc.)'),
    detectedDependencies: zod_1.z
        .array(zod_1.z.string())
        .optional()
        .describe('Detected libraries/frameworks/features from repository analysis (e.g., ["redis", "ef-core", "signalr", "mongodb", "health-checks"]). This helps match relevant knowledge entries.'),
    targetPlatform: schemas_1.platform.describe('Target platform for the Docker image (e.g., "linux/amd64", "linux/arm64"). Defaults to linux/amd64 for maximum compatibility. Use this to cross-compile for different architectures (e.g., ARM Mac targeting AMD64 servers).'),
    trafficLevel: zod_1.z
        .enum(['high', 'medium', 'low'])
        .optional()
        .describe('Expected traffic level for dynamic defaults calculation (affects replica counts and scaling).'),
    criticalityTier: zod_1.z
        .enum(['tier-1', 'tier-2', 'tier-3'])
        .optional()
        .describe('Criticality tier for dynamic defaults calculation (tier-1=mission-critical, tier-3=low-priority).'),
});
