// containerization-assist-mcp/dist-cjs/src/tools/generate-dockerfile/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateDockerfileToolDefinition = void 0;
const schema_1 = require("./schema");
const toolDefinition_1 = require("../shared/toolDefinition");
exports.generateDockerfileToolDefinition = {
    name: toolDefinition_1.TOOL_NAME.GENERATE_DOCKERFILE,
    description: 'Gather insights from knowledgebase and return requirements for Dockerfile creation or enhancement. Automatically detects existing Dockerfiles and provides detailed analysis and guidance.',
    category: 'docker',
    version: '2.0.0',
    schema: schema_1.generateDockerfileSchema,
    metadata: {
        knowledgeEnhanced: true,
    },
    chainHints: {
        success: 'Dockerfile plan generated successfully and passed policy validation. Next: Use fix-dockerfile to validate the actual Dockerfile content before building.',
        failure: 'Failed to generate Dockerfile plan or plan violates policies. Review repository analysis and policy violations.',
    },
};
