// containerization-assist-mcp/dist-cjs/src/lib/errors.js
"use strict";
/**
 * Error handling utilities and message templates
 *
 * Consolidates error utilities and centralized error messages
 * for consistent error handling across the application.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ERROR_MESSAGES = void 0;
exports.extractErrorMessage = extractErrorMessage;
exports.createErrorGuidance = createErrorGuidance;
// ============================================================================
// Error Message Templates
// ============================================================================
/**
 * Centralized Error Messages
 *
 * Provides consistent error message templates across the application.
 * Uses template functions for parameterized messages.
 */
exports.ERROR_MESSAGES = {
    // Tool-related errors
    TOOL_NOT_FOUND: (name) => `Tool not found: ${name}`,
    VALIDATION_FAILED: (issues) => `Validation failed: ${issues}`,
    // Policy-related errors
    POLICY_BLOCKED: (rules) => `Blocked by policy rules: ${rules.join(', ')}\n` +
        `Tip: Review policy configuration or adjust enforcement level. See https://github.com/Azure/containerization-assist/blob/main/docs/policy-guide.md`,
    POLICY_VALIDATION_FAILED: (issues) => `Policy validation failed: ${issues}\n` +
        `Tip: Check policy file syntax against schema. Available policies in policies/ directory.`,
    POLICY_LOAD_FAILED: (error) => `Failed to load policy: ${error}\n` +
        `Tip: Verify policy file exists and is valid YAML. See https://github.com/Azure/containerization-assist/blob/main/docs/policy-guide.md for format.`,
    // Infrastructure-related errors
    DOCKER_OPERATION_FAILED: (operation, error) => `Docker ${operation} failed: ${error}`,
    K8S_OPERATION_FAILED: (operation, error) => `Kubernetes ${operation} failed: ${error}`,
    K8S_APPLY_FAILED: (kind, name, error) => `Failed to apply ${kind}/${name}: ${error}`,
    // Execution errors
    // Generic templates
    OPERATION_FAILED: (operation, error) => `${operation} failed: ${error}`,
    RESOURCE_NOT_FOUND: (type, id) => `${type} not found: ${id}`,
};
// ============================================================================
// Error Utilities
// ============================================================================
/**
 * Safely extracts error message from unknown error types.
 * Invariant: Always returns a string message
 */
function extractErrorMessage(error) {
    if (error instanceof Error) {
        return error.message;
    }
    return String(error);
}
/**
 * Create error guidance with context
 */
function createErrorGuidance(message, hint, resolution, details) {
    const guidance = { message };
    if (hint !== undefined)
        guidance.hint = hint;
    if (resolution !== undefined)
        guidance.resolution = resolution;
    if (details !== undefined)
        guidance.details = details;
    return guidance;
}
