// containerization-assist-mcp/dist-cjs/src/lib/file-utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.downloadFile = downloadFile;
exports.makeExecutable = makeExecutable;
exports.createTempFile = createTempFile;
exports.deleteTempFile = deleteTempFile;
exports.readDockerfile = readDockerfile;
const node_fs_1 = require("node:fs");
const node_os_1 = require("node:os");
const node_path_1 = __importDefault(require("node:path"));
const node_https_1 = __importDefault(require("node:https"));
const node_http_1 = __importDefault(require("node:http"));
const node_url_1 = require("node:url");
const node_crypto_1 = __importDefault(require("node:crypto"));
const types_1 = require("../types");
const errors_1 = require("./errors");
const validation_helpers_1 = require("./validation-helpers");
/**
 * Download a file from URL to destination path
 */
async function downloadFile(url, dest) {
    return new Promise((resolve, reject) => {
        const parsedUrl = new node_url_1.URL(url);
        const client = parsedUrl.protocol === 'https:' ? node_https_1.default : node_http_1.default;
        const file = (0, node_fs_1.createWriteStream)(dest);
        const request = client.get(url, (response) => {
            response.pipe(file);
            file.on('finish', () => {
                file.close();
                resolve();
            });
        });
        request.on('error', (err) => {
            node_fs_1.promises.unlink(dest).catch(() => { }); // Delete the file on error
            reject(err);
        });
    });
}
/**
 * Make a file executable (Unix-like systems only)
 */
async function makeExecutable(filePath) {
    await node_fs_1.promises.chmod(filePath, 0o755);
}
/**
 * Create a temporary file with content
 */
async function createTempFile(content, extension = '') {
    const tempPath = node_path_1.default.join((0, node_os_1.tmpdir)(), `temp-${node_crypto_1.default.randomUUID()}${extension}`);
    // Create file with permissions (readable/writable only by owner)
    await node_fs_1.promises.writeFile(tempPath, content, { encoding: 'utf8', mode: 0o600 });
    return tempPath;
}
/**
 * Delete a temporary file (ignores errors)
 */
async function deleteTempFile(filePath) {
    try {
        await node_fs_1.promises.unlink(filePath);
    }
    catch {
        // Ignore errors when deleting temp files
    }
}
/**
 * Read Dockerfile from path or use provided content
 *
 * Consolidates duplicate Dockerfile reading logic across tools.
 * Handles both direct content and file path inputs.
 *
 * @param options - Either path to Dockerfile or content directly
 * @returns Result containing Dockerfile content
 *
 * @example
 * ```typescript
 * // Read from file path
 * const result = await readDockerfile({ path: './Dockerfile' });
 * if (!result.ok) return result;
 * const content = result.value;
 *
 * // Use provided content
 * const result2 = await readDockerfile({ content: 'FROM node:20\n...' });
 * ```
 */
async function readDockerfile(options) {
    // Check if content was explicitly provided (even if empty)
    if (options.content !== undefined) {
        if (options.content.trim().length === 0) {
            return (0, types_1.Failure)('Dockerfile content is empty', {
                message: 'Dockerfile content is empty',
                hint: 'Dockerfile must contain at least one instruction',
                resolution: 'Provide valid Dockerfile content or path',
            });
        }
        return (0, types_1.Success)(options.content);
    }
    if (options.path) {
        const dockerfilePath = node_path_1.default.resolve(options.path);
        // Validate file exists and is readable
        const validation = await (0, validation_helpers_1.validatePathOrFail)(dockerfilePath, {
            mustExist: true,
            mustBeFile: true,
            readable: true,
        });
        if (!validation.ok) {
            return validation;
        }
        try {
            const content = await node_fs_1.promises.readFile(dockerfilePath, 'utf-8');
            if (content.trim().length === 0) {
                return (0, types_1.Failure)(`Dockerfile at ${dockerfilePath} is empty`, {
                    message: 'Dockerfile is empty',
                    hint: 'Dockerfile must contain at least one instruction',
                    resolution: 'Add FROM instruction and other build steps',
                });
            }
            return (0, types_1.Success)(content);
        }
        catch (error) {
            return (0, types_1.Failure)(`Failed to read Dockerfile at ${dockerfilePath}: ${(0, errors_1.extractErrorMessage)(error)}`, {
                message: 'Failed to read Dockerfile',
                hint: 'Check file permissions and path',
                resolution: `Ensure file is readable: ls -la ${dockerfilePath}`,
            });
        }
    }
    return (0, types_1.Failure)("Either 'path' or 'content' must be provided", {
        message: 'Missing Dockerfile input',
        hint: 'Provide dockerfile content or path to Dockerfile',
        resolution: 'Add either path or content parameter',
    });
}
