// containerization-assist-mcp/dist-cjs/src/lib/logger.js
"use strict";
/**
 * Standardized Logger Utility
 *
 * Wrapper around Pino logger with helper functions
 * Avoids creating yet another logger interface - just use Pino!
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createLogger = createLogger;
exports.createTimer = createTimer;
const pino_1 = __importDefault(require("pino"));
const errors_1 = require("./errors");
/**
 * Create a Pino logger with sensible defaults for containerization assist
 */
function createLogger(options = {}) {
    // When running as MCP server, output to stderr to avoid interfering with JSON-RPC protocol
    // Skip transport in test environment to avoid worker thread leaks
    const isMCPMode = process.env.MCP_MODE === 'true' || process.argv.includes('--mcp');
    const isTestEnv = process.env.NODE_ENV === 'test' || process.env.JEST_WORKER_ID !== undefined;
    const transport = isMCPMode && !isTestEnv
        ? pino_1.default.transport({
            target: 'pino/file',
            options: { destination: 2 }, // 2 is stderr
        })
        : undefined;
    return (0, pino_1.default)({
        name: 'containerization-assist',
        level: process.env.LOG_LEVEL ?? (process.env.NODE_ENV === 'development' ? 'debug' : 'info'),
        redact: {
            paths: [
                // Docker/Registry credentials
                'errorDetails.json.auth',
                'errorDetails.json.token',
                'errorDetails.json.password',
                'errorDetails.json.credential',
                'errorDetails.json.bearer',
                'errorDetails.json.*.auth',
                'errorDetails.json.*.token',
                'errorDetails.json.*.password',
                'errorDetails.json.*.credential',
                'errorDetails.json.*.bearer',
                // General sensitive patterns
                '*.auth',
                '*.token',
                '*.password',
                '*.secret',
                '*.key',
                '*.credential',
                '*.bearer',
            ],
            censor: '[REDACTED]',
        },
        ...options,
    }, transport);
}
/**
 * Create a performance timer for an operation - functional approach
 */
function createTimer(logger, operation, context = {}) {
    const startTime = Date.now();
    // Log start of operation
    logger.debug({ operation, ...context }, `Starting ${operation}`);
    return {
        end(additionalContext = {}) {
            const duration = Date.now() - startTime;
            logger.info({
                operation,
                duration_ms: duration,
                ...context,
                ...additionalContext,
            }, `Completed ${operation} in ${duration}ms`);
        },
        error(error, additionalContext = {}) {
            const duration = Date.now() - startTime;
            logger.error({
                operation,
                duration_ms: duration,
                error: (0, errors_1.extractErrorMessage)(error),
                stack: error instanceof Error ? error.stack : undefined,
                ...context,
                ...additionalContext,
            }, `Failed ${operation} after ${duration}ms`);
        },
        checkpoint(label, additionalContext = {}) {
            const elapsed = Date.now() - startTime;
            logger.debug({
                operation,
                checkpoint: label,
                elapsed_ms: elapsed,
                ...context,
                ...additionalContext,
            }, `${operation} checkpoint: ${label} at ${elapsed}ms`);
            return elapsed;
        },
    };
}
