// containerization-assist-mcp/dist-cjs/src/lib/platform.js
"use strict";
/**
 * Platform and path utilities
 *
 * Consolidates cross-platform system detection and path normalization
 * for consistent behavior across different operating systems.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSystemInfo = getSystemInfo;
exports.getDownloadOS = getDownloadOS;
exports.getDownloadArch = getDownloadArch;
exports.mapNodeArchToPlatform = mapNodeArchToPlatform;
exports.isPlatformCompatible = isPlatformCompatible;
exports.normalizePath = normalizePath;
const path = __importStar(require("path"));
/**
 * Get system information for cross-platform logic
 */
function getSystemInfo() {
    return {
        isWindows: process.platform === 'win32',
        isMac: process.platform === 'darwin',
        isLinux: process.platform === 'linux',
    };
}
/**
 * Get OS string for download URLs
 */
function getDownloadOS() {
    const system = getSystemInfo();
    if (system.isWindows)
        return 'windows';
    if (system.isMac)
        return 'darwin';
    return 'linux';
}
/**
 * Get architecture string for download URLs
 */
function getDownloadArch() {
    switch (process.arch) {
        case 'x64':
            return 'amd64';
        case 'arm64':
            return 'arm64';
        default:
            return 'amd64';
    }
}
/**
 * Map Kubernetes node architecture to Docker platform format
 */
function mapNodeArchToPlatform(arch, os = 'linux') {
    const normalized = arch.toLowerCase();
    // Map common Kubernetes arch values to Docker platform format
    const archMap = {
        'amd64': 'amd64',
        'x86_64': 'amd64',
        'arm64': 'arm64',
        'aarch64': 'arm64',
        'armv7l': 'arm/v7',
        'armv6l': 'arm/v6',
        '386': '386',
        'i386': '386',
        'i686': '386',
        'ppc64le': 'ppc64le',
        's390x': 's390x',
        'riscv64': 'riscv64',
    };
    const mappedArch = archMap[normalized];
    if (!mappedArch) {
        return null;
    }
    const platform = `${os}/${mappedArch}`;
    // Validate against known platforms
    const validPlatforms = [
        'linux/amd64',
        'linux/arm64',
        'linux/arm/v7',
        'linux/arm/v6',
        'linux/386',
        'linux/ppc64le',
        'linux/s390x',
        'linux/riscv64',
        'windows/amd64',
    ];
    return validPlatforms.includes(platform) ? platform : null;
}
/**
 * Check if a target platform is compatible with a cluster platform.
 * Returns true if images built for targetPlatform can run on clusterPlatform.
 *
 * @param targetPlatform - Platform images are built for (e.g., "linux/amd64")
 * @param clusterPlatform - Platform cluster nodes support (e.g., "linux/amd64")
 * @returns true if compatible, false otherwise
 */
function isPlatformCompatible(targetPlatform, clusterPlatform) {
    // Exact match is always compatible
    if (targetPlatform === clusterPlatform) {
        return true;
    }
    // ARM64 can run ARM/v7 and ARM/v6 binaries
    if (clusterPlatform === 'linux/arm64') {
        return targetPlatform === 'linux/arm/v7' || targetPlatform === 'linux/arm/v6';
    }
    // AMD64 can run 386 binaries
    if (clusterPlatform === 'linux/amd64' && targetPlatform === 'linux/386') {
        return true;
    }
    // Different platforms are generally not compatible without emulation
    return false;
}
function normalizePath(inputPath) {
    if (inputPath == null)
        return inputPath; // handles both null and undefined
    if (inputPath === '')
        return inputPath;
    // Convert all backslashes to forward slashes, then normalize
    return path.posix.normalize(inputPath.replace(/\\/g, '/'));
}
