// containerization-assist-mcp/dist-cjs/src/lib/policy-helpers.js
"use strict";
/**
 * Shared Policy Validation Utilities
 *
 * Common functions and types for policy validation across tools.
 * This module provides shared utilities for Rego policy evaluation and violation mapping.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.mapRegoPolicyViolation = mapRegoPolicyViolation;
exports.validateContentAgainstPolicy = validateContentAgainstPolicy;
const policy_eval_1 = require("../config/policy-eval");
/**
 * Convert RegoPolicyViolation to PolicyViolation format
 *
 * @param regoViolation - Violation from Rego evaluation
 * @returns Standardized PolicyViolation
 */
function mapRegoPolicyViolation(regoViolation) {
    const violation = {
        ruleId: regoViolation.rule,
        category: regoViolation.category,
        message: regoViolation.message,
        severity: regoViolation.severity,
    };
    // Only add optional fields if they have values
    if (regoViolation.priority !== undefined) {
        violation.priority = regoViolation.priority;
    }
    if (regoViolation.description !== undefined) {
        violation.description = regoViolation.description;
    }
    return violation;
}
/**
 * Validate content against Rego policy
 *
 * Generic utility for evaluating content against a Rego policy evaluator.
 * Used by tools like generate-dockerfile, fix-dockerfile, and generate-k8s-manifests.
 *
 * @param content - Content to validate (Dockerfile text, K8s manifest, etc.)
 * @param policyEvaluator - Rego policy evaluator
 * @param logger - Logger instance for diagnostics
 * @param contentType - Description of content type for logging (e.g., "Dockerfile", "K8s manifest")
 * @returns PolicyValidationResult
 *
 * @example
 * ```typescript
 * const result = await validateContentAgainstPolicy(
 *   dockerfileContent,
 *   policyEvaluator,
 *   logger,
 *   'Dockerfile'
 * );
 * if (!result.passed) {
 *   console.log('Violations:', result.violations);
 * }
 * ```
 */
async function validateContentAgainstPolicy(content, policyEvaluator, logger, contentType = 'content') {
    logger.debug({ contentType, contentLength: content.length }, 'Validating content against policy');
    // Apply Rego policy to the content
    const policyResult = await (0, policy_eval_1.applyPolicy)(policyEvaluator, content);
    // Map Rego violations to PolicyViolation format
    const violations = policyResult.violations.map(mapRegoPolicyViolation);
    const warnings = policyResult.warnings.map(mapRegoPolicyViolation);
    const suggestions = policyResult.suggestions.map(mapRegoPolicyViolation);
    const passed = violations.length === 0;
    const totalRules = (policyResult.summary?.total_violations || 0) +
        (policyResult.summary?.total_warnings || 0) +
        (policyResult.summary?.total_suggestions || 0);
    logger.info({
        contentType,
        totalRules,
        matchedRules: totalRules,
        passed,
        violations: violations.length,
        warnings: warnings.length,
        suggestions: suggestions.length,
    }, `Policy validation completed for ${contentType}`);
    return {
        passed,
        violations,
        warnings,
        suggestions,
        summary: {
            totalRules,
            matchedRules: totalRules,
            blockingViolations: violations.length,
            warnings: warnings.length,
            suggestions: suggestions.length,
        },
    };
}
