// containerization-assist-mcp/dist-cjs/src/lib/regex-patterns.js
"use strict";
/**
 * Centralized regex patterns for the codebase
 *
 * Simple, flat structure - no nested objects or abstractions
 * Each pattern is a direct export for easy import and use
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TOKEN_PATTERN = exports.SECRET_PATTERN = exports.API_KEY_PATTERN = exports.PASSWORD_PATTERN = exports.PACKAGE_FILES = exports.SUDO_INSTALL = exports.LATEST_TAG = exports.AS_CLAUSE = void 0;
// Multi-stage patterns
exports.AS_CLAUSE = /\s+AS\s+/i;
// Version and tag patterns
exports.LATEST_TAG = /:latest(?:\s|$)/;
// Package manager patterns
exports.SUDO_INSTALL = /install.*sudo|apt-get.*sudo|yum.*sudo|apk.*sudo/;
exports.PACKAGE_FILES = /package.*\.json|requirements\.txt|go\.mod|pom\.xml/;
// Secret detection patterns (for validation, not extraction)
exports.PASSWORD_PATTERN = /\w*password\w*\s*=\s*["']?.+["']?/i;
exports.API_KEY_PATTERN = /\w*api[_-]?key\w*\s*=\s*["']?.+["']?/i;
exports.SECRET_PATTERN = /\w*secret\w*\s*=\s*["']?.+["']?/i;
exports.TOKEN_PATTERN = /\w*token\w*\s*=\s*["']?.+["']?/i;
