// containerization-assist-mcp/dist-cjs/src/lib/telemetry-utils.js
"use strict";
/**
 * Telemetry Sanitization Utilities
 *
 * Provides utilities to safely hash/obfuscate customer data for telemetry.
 *
 * Security Principle: "When in doubt, hash it out"
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.hashValue = hashValue;
exports.sanitizePath = sanitizePath;
exports.sanitizeToolInput = sanitizeToolInput;
exports.extractSafeTelemetryMetrics = extractSafeTelemetryMetrics;
exports.createSafeTelemetryEvent = createSafeTelemetryEvent;
const node_crypto_1 = require("node:crypto");
/**
 * Hash a string using SHA256 and return first 16 characters
 * Provides enough uniqueness for grouping while preventing reverse lookup
 */
function hashValue(value) {
    return (0, node_crypto_1.createHash)('sha256').update(value).digest('hex').substring(0, 16);
}
/**
 * Sanitize a file path for telemetry
 * Replaces path segments with hashed values
 *
 * Security strategy:
 * - Relative paths (./ or ../): Keep as-is (no customer data)
 * - Short paths (<=4 segments): Hash all (too risky to preserve any)
 * - Long paths (5+ segments): Hash all except last 2 (preserve structure like src/index.ts)
 *
 * @example
 * sanitizePath('/home/user/myapp/src/index.ts')
 * // Returns: '<hash1>/<hash2>/<hash3>/src/index.ts'
 *
 * sanitizePath('/home/user/myapp/src')
 * // Returns: '<hash1>/<hash2>/<hash3>/<hash4>' (all hashed - too short, 4 segments)
 */
function sanitizePath(path) {
    // For relative paths starting with ./ or ../, preserve as-is (no customer data)
    if (path.startsWith('./') || path.startsWith('../')) {
        return path;
    }
    const parts = path.split('/');
    // Filter out empty parts (from leading /) and count real segments
    const nonEmptyParts = parts.filter((p) => p !== '' && p !== '.' && p !== '..');
    // For paths with 4 or fewer segments, hash everything (security-first approach)
    // This includes most user/project paths like /home/user/project or /etc/config
    if (nonEmptyParts.length <= 4) {
        return parts
            .map((part) => {
            if (part === '' || part === '.' || part === '..') {
                return part;
            }
            return `<${hashValue(part)}>`;
        })
            .join('/');
    }
    // For longer paths (5+ segments), hash all but the last 2 real segments
    // This preserves structural context like "src/index.ts" while hiding customer identifiers
    let realSegmentCount = 0;
    const sanitizedParts = parts.map((part) => {
        // Keep empty parts, dots, and parent refs as-is
        if (part === '' || part === '.' || part === '..') {
            return part;
        }
        realSegmentCount++;
        // Hash all segments except the last 2 real ones
        if (realSegmentCount > nonEmptyParts.length - 2) {
            return part; // Preserve last 2 real segments
        }
        return `<${hashValue(part)}>`;
    });
    return sanitizedParts.join('/');
}
/**
 * Sanitize tool input parameters for telemetry
 * Removes or hashes sensitive fields
 */
function sanitizeToolInput(params) {
    const sanitized = {};
    // Fields containing file paths - always hash
    const pathFields = ['path', 'dockerfile', 'manifestPath', 'contextPath'];
    // Fields containing customer identifiers - always hash
    const sensitiveFields = [
        'imageName',
        'registryUrl',
        'deploymentName',
        'namespace',
        'serviceName',
        'clusterName',
    ];
    // Known safe string fields (enums, constants, flags)
    // These can be logged as-is because they don't contain customer data
    const safeStringFields = [
        'severity',
        'platform',
        'pullPolicy',
        'restartPolicy',
        'strategy',
        'protocol',
        'vendor',
        'format',
        'outputFormat',
        'chainHintsMode',
    ];
    for (const [key, value] of Object.entries(params)) {
        if (pathFields.includes(key) && typeof value === 'string') {
            // Hash file paths
            sanitized[key] = sanitizePath(value);
        }
        else if (sensitiveFields.includes(key) && typeof value === 'string') {
            // Hash sensitive identifiers
            sanitized[key] = hashValue(value);
        }
        else if (typeof value === 'boolean' || typeof value === 'number') {
            // Booleans and numbers are safe (flags, counts, sizes)
            sanitized[key] = value;
        }
        else if (typeof value === 'string' && safeStringFields.includes(key)) {
            // Known safe string fields (enums, constants)
            sanitized[key] = value;
        }
        else if (typeof value === 'string') {
            // Unknown string fields - hash to be safe
            sanitized[key] = hashValue(value);
        }
        else {
            // Complex objects - mark as present but don't log
            sanitized[key] = '<omitted>';
        }
    }
    return sanitized;
}
/**
 * Extract safe telemetry metrics from tool results
 * Returns only aggregate metrics, no sensitive data
 */
function extractSafeTelemetryMetrics(toolName, result) {
    const metrics = {};
    // Tool-specific safe metrics
    switch (toolName) {
        case 'analyze-repo':
            // Safe: Framework types, languages (enums)
            if ('framework' in result)
                metrics.framework = result.framework;
            if ('language' in result)
                metrics.language = result.language;
            if ('moduleCount' in result)
                metrics.moduleCount = result.moduleCount;
            break;
        case 'build-image-context':
            // Safe: Security analysis metrics, BuildKit recommendations (aggregates)
            if ('securityAnalysis' in result && result.securityAnalysis) {
                const security = result.securityAnalysis;
                metrics.securityWarningCount = security.warnings?.length || 0;
                metrics.riskLevel = security.riskLevel;
            }
            if ('buildKitAnalysis' in result && result.buildKitAnalysis) {
                const buildKit = result.buildKitAnalysis;
                metrics.buildKitRecommended = buildKit.recommended;
            }
            if ('dockerfileAnalysis' in result && result.dockerfileAnalysis) {
                const dockerfile = result.dockerfileAnalysis;
                metrics.layerCount = dockerfile.layerCount;
                metrics.hasHealthcheck = dockerfile.hasHealthcheck;
            }
            if ('context' in result && result.context) {
                const context = result.context;
                metrics.hasDockerignore = context.hasDockerignore;
            }
            // Unsafe: paths, image names, tags (contain customer identifiers)
            break;
        case 'scan-image':
            // Safe: Vulnerability counts (aggregates)
            if ('summary' in result && result.summary) {
                const summary = result.summary;
                metrics.criticalVulns = summary.critical || 0;
                metrics.highVulns = summary.high || 0;
                metrics.mediumVulns = summary.medium || 0;
                metrics.lowVulns = summary.low || 0;
            }
            break;
        case 'verify-deploy':
            // Safe: Replica counts, status (aggregates)
            if ('replicas' in result)
                metrics.replicas = result.replicas;
            if ('readyReplicas' in result)
                metrics.readyReplicas = result.readyReplicas;
            if ('status' in result)
                metrics.status = result.status;
            // Unsafe: namespace, deploymentName (customer identifiers)
            break;
        default:
            // For unknown tools, only log success/failure
            metrics.hasResult = true;
    }
    return metrics;
}
/**
 * Create a safe telemetry event from tool execution
 */
function createSafeTelemetryEvent(toolName, params, result, durationMs) {
    const event = {
        toolName,
        success: result.ok,
        sanitizedInput: sanitizeToolInput(params),
        metrics: result.ok && result.value ? extractSafeTelemetryMetrics(toolName, result.value) : {},
    };
    // Only include optional properties if they have values
    if (durationMs !== undefined) {
        event.durationMs = durationMs;
    }
    if (result.error) {
        event.errorType = 'Error';
    }
    return event;
}
