// containerization-assist-mcp/dist-cjs/src/lib/tool-helpers.js
"use strict";
/**
 * Common helper utilities for MCP tools to reduce code duplication
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getToolLogger = getToolLogger;
exports.createToolTimer = createToolTimer;
exports.createStandardizedToolTracker = createStandardizedToolTracker;
const logger_js_1 = require("./logger.js");
const runtime_logging_js_1 = require("./runtime-logging.js");
/**
 * Gets or creates a logger for a tool.
 * Consolidates the pattern: context.logger || createLogger({ name: 'tool-name' })
 * Invariant: Always returns a valid logger instance
 *
 * @param context - The tool context that may contain a logger
 * @param toolName - Name of the tool for logger creation
 * @returns Logger instance from context or newly created
 */
function getToolLogger(context, toolName) {
    return context.logger || (0, logger_js_1.createLogger)({ name: toolName });
}
/**
 * Creates a timer for measuring tool execution time with safeguards.
 * Consolidates timer creation pattern for consistent usage across tools.
 * Wraps the timer to prevent double end/error calls which can skew metrics.
 *
 * @param logger - Logger instance for timer output
 * @param toolName - Name of the tool for timer identification
 * @returns Timer instance with double-call protection
 */
function createToolTimer(logger, toolName) {
    const timer = (0, logger_js_1.createTimer)(logger, toolName);
    let completed = false;
    return {
        end(additionalContext) {
            if (completed)
                return;
            completed = true;
            timer.end(additionalContext);
        },
        error(error, additionalContext) {
            if (completed)
                return;
            completed = true;
            timer.error(error, additionalContext);
        },
        checkpoint(label, additionalContext) {
            // Checkpoints are allowed even after completion for debugging
            return timer.checkpoint(label, additionalContext);
        },
    };
}
/**
 * Create a standardized tool execution tracker with automatic start/complete logging.
 * Uses consistent "Starting X" / "Completed X" format for Copilot transcripts.
 *
 * @param toolName - Name of the tool (e.g., 'build-image', 'generate-k8s-manifests')
 * @param params - Key parameters to log at start
 * @param logger - Logger instance
 * @returns Object with complete() and fail() methods for structured logging
 *
 * @example
 * ```typescript
 * const tracker = createStandardizedToolTracker('build-image', { path: './app', tags }, logger);
 * try {
 *   const result = await performBuild();
 *   tracker.complete({ imageId: result.imageId });
 *   return Success(result);
 * } catch (error) {
 *   tracker.fail(error);
 *   return Failure(error.message);
 * }
 * ```
 */
function createStandardizedToolTracker(toolName, params, logger) {
    const startTime = Date.now();
    (0, runtime_logging_js_1.logToolStart)(toolName, params, logger);
    return {
        complete: (result) => {
            const duration = Date.now() - startTime;
            (0, runtime_logging_js_1.logToolComplete)(toolName, result, logger, duration);
        },
        fail: (error, context) => {
            (0, runtime_logging_js_1.logToolFailure)(toolName, error, logger, context);
        },
    };
}
