// containerization-assist-mcp/dist-cjs/src/lib/tool-logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createToolLogEntry = createToolLogEntry;
exports.getLogFilePath = getLogFilePath;
exports.createToolLoggerFile = createToolLoggerFile;
exports.logToolExecution = logToolExecution;
const fs_1 = require("fs");
const promises_1 = require("fs/promises");
const path_1 = require("path");
const config_1 = require("../config");
function createToolLogEntry(toolName, input) {
    const entry = {
        timestamp: Date.now().toString(),
        toolName,
        input,
        params: input,
        output: undefined,
        success: false,
    };
    return entry;
}
let logFileName = null;
function isToolLoggingEnabled() {
    return config_1.config.toolLogging.enabled;
}
function getLogFilePath() {
    const dirPath = config_1.config.toolLogging.dirPath;
    if (!dirPath) {
        return '';
    }
    if (logFileName) {
        return (0, path_1.join)(dirPath, logFileName);
    }
    const timestamp = Date.now();
    logFileName = `ca-tool-logs-${timestamp}.jsonl`;
    return (0, path_1.join)(dirPath, logFileName);
}
function createToolLoggerFile(logger) {
    if (!isToolLoggingEnabled()) {
        return;
    }
    logger?.info({ message: 'Starting tool logger file creation' });
    const dirPath = config_1.config.toolLogging.dirPath;
    if (!dirPath) {
        logger?.warn('Tool logging directory path is not configured');
        return;
    }
    try {
        // Ensure directory exists
        (0, fs_1.mkdirSync)(dirPath, { recursive: true });
        const logFilePath = getLogFilePath();
        (0, fs_1.writeFileSync)(logFilePath, '', 'utf-8');
        logger?.info({ path: logFilePath }, 'Tool logging file created');
    }
    catch (error) {
        const errorMsg = `Tool logging directory '${dirPath}' is not writable: ${error.message}`;
        logger?.warn({ error, path: dirPath }, errorMsg);
    }
}
async function logToolExecution(entry, logger) {
    if (!isToolLoggingEnabled()) {
        return;
    }
    try {
        const filepath = getLogFilePath();
        const logLine = `${JSON.stringify(entry)}\n`;
        await (0, promises_1.appendFile)(filepath, logLine, 'utf-8');
        logger?.debug({ filepath, toolName: entry.toolName }, 'Tool execution logged to file');
    }
    catch (error) {
        logger?.warn({ error, toolName: entry.toolName }, 'Failed to write tool execution log');
    }
}
