// containerization-assist-mcp/dist-cjs/src/lib/tool-registration.js
"use strict";
/**
 * Tool Registration Helpers
 *
 * Utilities for registering Container Assist tools with MCP servers
 * while maintaining full control over telemetry and error handling.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createToolHandler = createToolHandler;
exports.registerTool = registerTool;
exports.registerTools = registerTools;
const types_js_1 = require("@modelcontextprotocol/sdk/types.js");
const errors_1 = require("./errors");
const mcp_server_1 = require("../mcp/mcp-server");
const orchestrator_types_1 = require("../app/orchestrator-types");
/**
 * Creates a tool handler function that delegates to app.execute()
 *
 * This allows you to use server.tool() for full control over telemetry
 * while letting Container Assist handle all context creation internally.
 *
 * When called with a literal tool name (e.g., 'build-image'), TypeScript
 * automatically infers the specific type, giving you fully-typed callbacks.
 *
 * @template TName - The tool name type (automatically inferred from literal strings)
 * @param app - AppRuntime instance (from createApp())
 * @param toolName - Name of the tool to execute
 * @param options - Optional configuration for the handler
 * @returns Tool handler function compatible with MCP SDK
 *
 * @example
 * ```typescript
 * import { createApp, ALL_TOOLS, createToolHandler } from 'containerization-assist';
 *
 * const app = createApp();
 *
 * // Fully typed - TypeScript infers TName = 'build-image'
 * server.tool(
 *   'build-image',
 *   buildImageTool.description,
 *   buildImageTool.inputSchema,
 *   createToolHandler(app, 'build-image', {
 *     transport: 'my-integration',
 *     outputFormat: 'markdown', // Use markdown format instead of JSON
 *     onSuccess: (result, toolName, params) => {
 *       // result: BuildImageResult (fully typed!)
 *       // params: BuildImageInput (fully typed!)
 *       console.log(params.imageName, result.imageId);
 *       myTelemetry.trackToolExecution(toolName, true);
 *     },
 *     onError: (error, toolName, params) => {
 *       // params: BuildImageInput (fully typed!)
 *       console.log(params.dockerfile);
 *       myTelemetry.trackToolExecution(toolName, false);
 *     },
 *   })
 * );
 * ```
 */
function createToolHandler(app, toolName, options = {}) {
    const { transport = 'external', outputFormat = app.config?.outputFormat ?? mcp_server_1.OUTPUTFORMAT.NATURAL_LANGUAGE, chainHintsMode = app.config?.chainHintsMode ?? orchestrator_types_1.CHAINHINTSMODE.ENABLED, onError, onSuccess, } = options;
    return async (rawParams, extra) => {
        const params = rawParams ?? {};
        try {
            // Extract _meta if present
            const meta = params._meta && typeof params._meta === 'object'
                ? params._meta
                : {};
            // Extract metadata from MCP request
            const metadata = {
                transport,
                signal: extra.signal,
                ...meta,
            };
            // Add sendNotification only if available
            if (extra.sendNotification) {
                const sendNotification = extra.sendNotification;
                metadata.sendNotification = async (notification) => {
                    await sendNotification(notification);
                };
            }
            // Remove _meta from params before execution
            const { _meta, ...sanitizedParams } = params;
            // Execute via app.execute() - this handles all context creation
            // Type assertions are safe here because:
            // 1. The orchestrator validates toolName exists in the registry
            // 2. Each tool's Zod schema validates params at runtime
            // 3. Invalid toolName or params will return a proper Result.Failure
            const result = await app.execute(toolName, sanitizedParams, metadata);
            if (!result.ok) {
                // Call error handler if provided
                if (onError) {
                    // Type assertion is safe because:
                    // 1. toolName is validated by the orchestrator at runtime
                    // 2. params are validated by the tool's Zod schema
                    // 3. The generic TName ensures type consistency between handler and callbacks
                    onError(result.error, toolName, sanitizedParams);
                }
                // Format error with guidance
                let errorMessage = result.error || 'Tool execution failed';
                if (result.guidance) {
                    const parts = [errorMessage];
                    if (result.guidance.hint) {
                        parts.push(`💡 ${result.guidance.hint}`);
                    }
                    parts.push(`🔧 Resolution:`, result.guidance.resolution || 'Check logs for more information');
                    errorMessage = parts.join('\n\n');
                }
                throw new types_js_1.McpError(types_js_1.ErrorCode.InternalError, errorMessage);
            }
            // Call success handler if provided
            if (onSuccess) {
                // Type assertion is safe for the same reasons as above
                onSuccess(result.value, toolName, sanitizedParams);
            }
            // Return formatted result
            return {
                content: [
                    {
                        type: 'text',
                        text: (0, mcp_server_1.formatOutput)(result.value, outputFormat, chainHintsMode),
                    },
                ],
            };
        }
        catch (error) {
            if (error instanceof types_js_1.McpError) {
                throw error;
            }
            if (onError) {
                // Type assertion is safe (params include _meta but that's acceptable for error reporting)
                onError(error, toolName, params);
            }
            throw new types_js_1.McpError(types_js_1.ErrorCode.InternalError, (0, errors_1.extractErrorMessage)(error));
        }
    };
}
/**
 * Register a single Container Assist tool with an MCP server
 *
 * Convenience wrapper around createToolHandler() for registering individual tools.
 * For full type safety, use createToolHandler() directly with literal tool names instead.
 *
 * @param server - MCP server instance
 * @param app - AppRuntime instance
 * @param tool - Tool definition
 * @param options - Optional handler configuration (params/results will be union types)
 *
 * @example
 * ```typescript
 * import { createApp, analyzeRepoTool, registerTool } from 'containerization-assist';
 *
 * const app = createApp();
 *
 * // Simple registration (params/result are union types)
 * registerTool(server, app, analyzeRepoTool);
 *
 * // For full type safety, use createToolHandler instead:
 * server.tool(
 *   'analyze-repo',
 *   analyzeRepoTool.description,
 *   analyzeRepoTool.inputSchema,
 *   createToolHandler(app, 'analyze-repo', {
 *     onSuccess: (result, toolName, params) => {
 *       // Fully typed as AnalyzeRepoResult and AnalyzeRepoInput!
 *       myTelemetry.track('analyze-repo-success', params.path);
 *     },
 *   })
 * );
 * ```
 */
function registerTool(server, app, tool, options) {
    // Type assertion to avoid deep type instantiation issues with MCP SDK
    // The MCP SDK's complex generic constraints on tool() cause TS2589 errors
    // Runtime safety is preserved by Zod schema validation
    server.tool(tool.name, tool.description, tool.inputSchema, createToolHandler(app, tool.name, options));
}
/**
 * Register all Container Assist tools with an MCP server
 *
 * Convenience function to register multiple tools at once. For full type safety,
 * register tools individually using createToolHandler() with literal tool names.
 *
 * @param server - MCP server instance
 * @param app - AppRuntime instance
 * @param tools - Array of tools to register
 * @param options - Optional handler configuration (applied to all tools with union types)
 *
 * @example
 * ```typescript
 * import { createApp, ALL_TOOLS, registerTools } from 'containerization-assist';
 *
 * const app = createApp();
 *
 * // Bulk registration - convenient but broader types
 * registerTools(server, app, ALL_TOOLS, {
 *   transport: 'my-integration',
 *   onSuccess: (result, toolName, params) => {
 *     // result and params are unions of all tool types
 *     myTelemetry.trackToolExecution(toolName, true);
 *   },
 * });
 *
 * // For type-safe registration, use createToolHandler directly:
 * server.tool(
 *   'build-image',
 *   buildImageTool.description,
 *   buildImageTool.inputSchema,
 *   createToolHandler(app, 'build-image', {
 *     onSuccess: (result, toolName, params) => {
 *       // Fully typed! result: BuildImageResult, params: BuildImageInput
 *     }
 *   })
 * );
 * ```
 */
function registerTools(server, app, tools, options) {
    for (const tool of tools) {
        // Type assertion to avoid deep type instantiation issues with MCP SDK
        // The MCP SDK's complex generic constraints on tool() cause TS2589 errors
        // Runtime safety is preserved by Zod schema validation
        server.tool(tool.name, tool.description, tool.inputSchema, createToolHandler(app, tool.name, options));
    }
}
