// containerization-assist-mcp/dist-cjs/src/knowledge/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPolicyAwareKnowledgeSnippets = exports.findPolicyAwareKnowledgeMatches = exports.loadKnowledgeData = void 0;
exports.getKnowledgeForCategory = getKnowledgeForCategory;
const matcher_1 = require("./matcher");
const loader_1 = require("./loader");
Object.defineProperty(exports, "loadKnowledgeData", { enumerable: true, get: function () { return loader_1.loadKnowledgeData; } });
const policy_aware_matcher_1 = require("./policy-aware-matcher");
Object.defineProperty(exports, "findPolicyAwareKnowledgeMatches", { enumerable: true, get: function () { return policy_aware_matcher_1.findPolicyAwareKnowledgeMatches; } });
Object.defineProperty(exports, "getPolicyAwareKnowledgeSnippets", { enumerable: true, get: function () { return policy_aware_matcher_1.getPolicyAwareKnowledgeSnippets; } });
// Internal helper - only used by getKnowledgeForCategory
function getKnowledgeRecommendations(query) {
    if (!(0, loader_1.isKnowledgeLoaded)())
        (0, loader_1.loadKnowledgeBase)();
    return (0, matcher_1.findKnowledgeMatches)((0, loader_1.getAllEntries)(), query);
}
function getKnowledgeForCategory(category, text, context) {
    const query = {
        category,
        ...(text && { text }),
        ...(context?.language && { language: context.language }),
        ...(context?.framework && { framework: context.framework }),
        ...(context?.environment && { environment: context.environment }),
        limit: 5,
    };
    return getKnowledgeRecommendations(query);
}
