// containerization-assist-mcp/dist-cjs/src/knowledge/loader.js
"use strict";
/**
 * Knowledge Pack Loader
 * Loads and manages static knowledge packs for AI enhancement
 *
 * Knowledge packs are embedded at build time via static JSON imports,
 * eliminating the need for filesystem access at runtime. This allows
 * the SDK to work correctly when bundled by consumer applications.
 *
 * @see {@link ../../docs/adr/003-knowledge-enhancement.md ADR-003: Knowledge Enhancement System}
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadKnowledgeBase = loadKnowledgeBase;
exports.getAllEntries = getAllEntries;
exports.isKnowledgeLoaded = isKnowledgeLoaded;
exports.loadKnowledgeData = loadKnowledgeData;
exports.resetKnowledgeState = resetKnowledgeState;
const logger_1 = require("../lib/logger");
const schemas_1 = require("./schemas");
const embedded_packs_1 = require("./embedded-packs");
const zod_1 = require("zod");
// ===== Constants =====
const logger = (0, logger_1.createLogger)({ name: 'knowledge-loader' });
// ===== State =====
const knowledgeState = {
    entries: new Map(),
    byCategory: new Map(),
    byTag: new Map(),
    loaded: false,
};
// ===== Validation =====
/**
 * Format Zod errors for logging
 */
function formatZodErrors(errors) {
    return errors.map((e) => ({
        path: e.path.join('.'),
        message: e.message,
    }));
}
/**
 * Validate and normalize pack structure
 * Handles both array and object-wrapped pack formats
 */
function validateAndNormalizePack(packName, data) {
    try {
        const validated = schemas_1.KnowledgePackSchema.parse(data);
        // Extract entries based on format
        const entries = Array.isArray(validated)
            ? validated // Format 1: Flat array
            : validated.rules; // Format 2: Object with rules
        return { success: true, entries };
    }
    catch (error) {
        if (error instanceof zod_1.z.ZodError) {
            logger.warn({
                pack: packName,
                errors: formatZodErrors(error.issues.slice(0, 5)),
                totalErrors: error.issues.length,
            }, 'Pack validation failed');
        }
        return { success: false, error: String(error) };
    }
}
/**
 * Validate a single knowledge entry
 */
function validateEntry(entry) {
    try {
        schemas_1.KnowledgeEntrySchema.parse(entry);
        return true;
    }
    catch (error) {
        if (error instanceof zod_1.z.ZodError) {
            logger.warn({
                entryId: entry?.id || 'unknown',
                errors: formatZodErrors(error.issues),
            }, 'Entry validation failed');
        }
        return false;
    }
}
// ===== State Management =====
/**
 * Add an entry to the knowledge state
 */
function addEntry(entry) {
    knowledgeState.entries.set(entry.id, entry);
}
/**
 * Helper to add entry to a map with array values
 */
function addToMapArray(map, key, value) {
    const existing = map.get(key);
    if (existing) {
        existing.push(value);
    }
    else {
        map.set(key, [value]);
    }
}
/**
 * Build category and tag indices for fast lookup
 */
function buildIndices() {
    knowledgeState.byCategory.clear();
    knowledgeState.byTag.clear();
    for (const entry of knowledgeState.entries.values()) {
        // Index by category
        addToMapArray(knowledgeState.byCategory, entry.category, entry);
        // Index by tags
        if (entry.tags) {
            for (const tag of entry.tags) {
                addToMapArray(knowledgeState.byTag, tag, entry);
            }
        }
    }
}
/**
 * Get top N most common tags with their counts
 */
function getTopTags(limit) {
    const tagCounts = new Map();
    for (const entry of knowledgeState.entries.values()) {
        if (entry.tags) {
            for (const tag of entry.tags) {
                tagCounts.set(tag, (tagCounts.get(tag) ?? 0) + 1);
            }
        }
    }
    return Array.from(tagCounts.entries())
        .sort((a, b) => b[1] - a[1])
        .slice(0, limit)
        .map(([tag, count]) => ({ tag, count }));
}
/**
 * Load a single embedded knowledge pack
 */
function loadEmbeddedPack(packName, packData, stats) {
    try {
        // Validate and normalize pack structure
        const result = validateAndNormalizePack(packName, packData);
        if (!result.success) {
            const error = 'Pack validation failed (see previous log)';
            stats.packsFailed++;
            stats.failures.push({ name: packName, error });
            throw new Error(`Failed to load embedded knowledge pack ${packName}: ${error}`);
        }
        logger.debug({ pack: packName, count: result.entries.length }, 'Loading knowledge pack');
        // Validate and add individual entries
        for (const entry of result.entries) {
            if (validateEntry(entry)) {
                addEntry(entry);
                stats.entriesValid++;
            }
            else {
                stats.entriesInvalid++;
            }
        }
        stats.packsLoaded++;
    }
    catch (error) {
        stats.packsFailed++;
        const errorMessage = String(error);
        stats.failures.push({ name: packName, error: errorMessage });
        logger.error({ pack: packName, error }, 'Failed to load knowledge pack');
        throw new Error(`Failed to load embedded knowledge pack ${packName}: ${errorMessage}`);
    }
}
/**
 * Load knowledge entries from embedded knowledge packs
 * Throws an error if any embedded pack fails to load
 */
function loadKnowledgeBase() {
    if (knowledgeState.loaded) {
        return;
    }
    const stats = {
        packsAttempted: embedded_packs_1.EMBEDDED_PACKS.length,
        packsLoaded: 0,
        packsFailed: 0,
        entriesValid: 0,
        entriesInvalid: 0,
        failures: [],
    };
    if (embedded_packs_1.EMBEDDED_PACKS.length === 0) {
        const error = new Error(`No knowledge packs embedded - server cannot start without knowledge base.\n` +
            `\n` +
            `This is likely a build issue. The embedded-packs.ts file should contain\n` +
            `static imports of all knowledge pack JSON files.\n` +
            `\n` +
            `Resolution:\n` +
            `  - Ensure knowledge/packs/*.json files exist\n` +
            `  - Rebuild the package with: npm run build`);
        logger.error({ error }, 'No knowledge packs embedded');
        throw error;
    }
    logger.info({ totalPacks: embedded_packs_1.EMBEDDED_PACKS.length }, 'Loading embedded knowledge packs');
    // Load each embedded pack
    for (const pack of embedded_packs_1.EMBEDDED_PACKS) {
        loadEmbeddedPack(pack.name, pack.data, stats);
    }
    buildIndices();
    knowledgeState.loaded = true;
    logger.info({
        packsAttempted: stats.packsAttempted,
        packsLoaded: stats.packsLoaded,
        packsFailed: stats.packsFailed,
        entriesValid: stats.entriesValid,
        entriesInvalid: stats.entriesInvalid,
        totalEntries: knowledgeState.entries.size,
        categories: Array.from(knowledgeState.byCategory.keys()),
        topTags: getTopTags(5),
    }, 'Knowledge base loaded');
}
// ===== Public API =====
/**
 * Get all loaded knowledge entries
 */
function getAllEntries() {
    return Array.from(knowledgeState.entries.values());
}
/**
 * Check if knowledge base is loaded
 */
function isKnowledgeLoaded() {
    return knowledgeState.loaded;
}
/**
 * Load knowledge data and return entries.
 * Used by prompt engine for knowledge selection.
 */
function loadKnowledgeData() {
    if (!isKnowledgeLoaded()) {
        loadKnowledgeBase();
    }
    return {
        entries: getAllEntries(),
    };
}
/**
 * Reset knowledge state (for testing purposes)
 */
function resetKnowledgeState() {
    knowledgeState.entries.clear();
    knowledgeState.byCategory.clear();
    knowledgeState.byTag.clear();
    knowledgeState.loaded = false;
}
