// containerization-assist-mcp/dist-cjs/src/knowledge/schemas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KnowledgePackSchema = exports.KnowledgeEntrySchema = void 0;
const zod_1 = require("zod");
/**
 * Knowledge category schema - matches CATEGORY constants in types.ts
 */
const KnowledgeCategorySchema = zod_1.z.enum([
    'api',
    'architecture',
    'build',
    'caching',
    'configuration',
    'dockerfile',
    'features',
    'generic',
    'kubernetes',
    'optimization',
    'reliability',
    'resilience',
    'security',
    'streaming',
    'validation',
]);
/**
 * Schema for individual knowledge entry
 */
exports.KnowledgeEntrySchema = zod_1.z.object({
    id: zod_1.z.string().min(1),
    category: KnowledgeCategorySchema,
    pattern: zod_1.z.string().min(1),
    recommendation: zod_1.z.string().min(1),
    example: zod_1.z.string().optional(),
    severity: zod_1.z.enum(['required', 'high', 'medium', 'low']).optional(),
    tags: zod_1.z.array(zod_1.z.string()).optional(),
    description: zod_1.z.string().optional(),
});
/**
 * Schema for knowledge pack metadata wrapper
 * Some packs are wrapped in an object with metadata
 */
const KnowledgePackMetadataSchema = zod_1.z.object({
    name: zod_1.z.string(),
    version: zod_1.z.string(),
    description: zod_1.z.string().optional(),
    triggers: zod_1.z.array(zod_1.z.string()).optional(),
    rules: zod_1.z.array(exports.KnowledgeEntrySchema).min(1),
});
/**
 * Schema for knowledge pack file structure
 * Supports both formats:
 * 1. Array of entries (flat format)
 * 2. Object with metadata and rules array (wrapped format)
 */
exports.KnowledgePackSchema = zod_1.z.union([
    zod_1.z.array(exports.KnowledgeEntrySchema).min(1),
    KnowledgePackMetadataSchema,
]);
