// containerization-assist-mcp/dist-cjs/src/validation/dockerfilelint-adapter.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.lintWithDockerfilelint = lintWithDockerfilelint;
const node_child_process_1 = require("node:child_process");
const fs = __importStar(require("node:fs"));
const path = __importStar(require("node:path"));
const os = __importStar(require("node:os"));
const crypto = __importStar(require("node:crypto"));
const logger_1 = require("../lib/logger");
const logger = (0, logger_1.createLogger)({ name: 'dockerfilelint-adapter' });
/**
 * Maps dockerfilelint severity levels to our ValidationSeverity
 */
function mapSeverity(level) {
    switch (level) {
        case 'error':
            return 'error';
        case 'warn':
            return 'warning';
        case 'info':
        case 'style':
        default:
            return 'info';
    }
}
/**
 * Converts dockerfilelint issues to our ValidationResult format
 */
function toValidationResults(issues) {
    return issues.map((issue) => ({
        isValid: false,
        passed: false,
        errors: issue.level === 'error' ? [issue.message] : [],
        warnings: issue.level === 'warn' ? [issue.message] : [],
        ruleId: issue.rule ? `dockerfilelint-${issue.rule}` : 'dockerfilelint',
        message: issue.message,
        metadata: {
            severity: mapSeverity(issue.level),
            ...(issue.line && { location: `line ${issue.line}` }),
        },
    }));
}
/**
 * Cached module instance - load once at module initialization
 * This avoids repeatedly trying to load the module on every linting call
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
let cachedModule = undefined;
/**
 * Track whether CLI is available - cache the result to avoid repeated npx attempts
 */
let cliAvailable = undefined;
/**
 * Try to load dockerfilelint module programmatically
 * Uses caching to avoid repeated load attempts
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
async function tryLoadModule() {
    // Return cached result if already attempted
    if (cachedModule !== undefined) {
        return cachedModule;
    }
    try {
        // Try to dynamically import dockerfilelint - using require syntax for optional dependency
        cachedModule = eval('require("dockerfilelint")');
        logger.debug('Loaded dockerfilelint module programmatically');
        return cachedModule;
    }
    catch {
        logger.debug('Could not load dockerfilelint module, will use CLI fallback');
        cachedModule = null;
        return null;
    }
}
/**
 * Run dockerfilelint programmatically if module is available
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
async function runProgrammatic(mod, dockerfileContent) {
    try {
        // Try to use the module API if available
        if (mod.lint || mod.default?.lint) {
            const lintFn = mod.lint || mod.default.lint;
            const result = await lintFn(dockerfileContent);
            // Handle different possible result formats
            if (Array.isArray(result)) {
                return result;
            }
            else if (result?.issues) {
                return result.issues;
            }
            else if (result?.problems) {
                return result.problems;
            }
        }
    }
    catch (e) {
        logger.debug({ error: e }, 'Failed to run dockerfilelint programmatically');
    }
    throw new Error('Programmatic API not available');
}
/**
 * Lint a Dockerfile using dockerfilelint
 * Falls back to CLI if programmatic API is not available
 */
async function lintWithDockerfilelint(dockerfileContent) {
    // Try programmatic API first
    const mod = await tryLoadModule();
    if (mod) {
        try {
            const issues = await runProgrammatic(mod, dockerfileContent);
            return toReport(issues);
        }
        catch {
            // Fall through to CLI
            logger.debug('Falling back to CLI approach');
        }
    }
    // If we've already determined CLI is not available, skip the expensive npx call
    if (cliAvailable === false) {
        return createEmptyReport();
    }
    // Skip CLI fallback in test environment to avoid slow npx calls
    if (process.env.NODE_ENV === 'test' || process.env.SKIP_DOCKERFILELINT_CLI === 'true') {
        cliAvailable = false;
        return createEmptyReport();
    }
    // CLI fallback using npx (no external binary needed)
    try {
        // Secure temp file creation using Node.js built-ins
        const tmpDir = os.tmpdir();
        const randomSuffix = crypto.randomBytes(8).toString('hex');
        const tmpFile = path.join(tmpDir, `dockerfile-${randomSuffix}`);
        // Create file with secure permissions (0o600 = rw-------)
        const fd = fs.openSync(tmpFile, 'w', 0o600);
        fs.writeSync(fd, dockerfileContent);
        fs.closeSync(fd);
        try {
            const cli = (0, node_child_process_1.spawnSync)(process.platform === 'win32' ? 'npx.cmd' : 'npx', ['-y', 'dockerfilelint', '--json', tmpFile], { encoding: 'utf8' });
            // Clean up temp file
            try {
                fs.unlinkSync(tmpFile);
            }
            catch {
                // Ignore cleanup errors
            }
            if (cli.error) {
                logger.warn({ error: cli.error }, 'dockerfilelint CLI error');
                cliAvailable = false;
                return createEmptyReport();
            }
            // Parse JSON output
            if (cli.stdout) {
                try {
                    const output = JSON.parse(cli.stdout);
                    if (output.issues) {
                        cliAvailable = true;
                        return toReport(output.issues);
                    }
                }
                catch {
                    logger.debug({ output: cli.stdout }, 'Failed to parse dockerfilelint output');
                }
            }
            // If stderr has content, it might be non-JSON output
            if (cli.stderr) {
                logger.debug({ stderr: cli.stderr }, 'dockerfilelint stderr output');
            }
            // CLI ran but returned no valid output - mark as available but return empty
            cliAvailable = true;
            return createEmptyReport();
        }
        finally {
            // Ensure temp file is cleaned up
            try {
                if (fs.existsSync(tmpFile)) {
                    fs.unlinkSync(tmpFile);
                }
            }
            catch {
                // Ignore cleanup errors
            }
        }
    }
    catch (error) {
        logger.error({ error }, 'Failed to run dockerfilelint');
        cliAvailable = false;
        return createEmptyReport();
    }
}
/**
 * Convert dockerfilelint issues to a ValidationReport
 */
function toReport(issues) {
    const results = toValidationResults(issues);
    // Count by severity
    let errors = 0;
    let warnings = 0;
    let info = 0;
    for (const result of results) {
        const severity = result.metadata?.severity;
        if (severity === 'error')
            errors++;
        else if (severity === 'warning')
            warnings++;
        else
            info++;
    }
    // Calculate score (simple linear deduction)
    const score = Math.max(0, 100 - errors * 10 - warnings * 3 - info * 1);
    // Calculate grade
    let grade;
    if (score >= 90)
        grade = 'A';
    else if (score >= 80)
        grade = 'B';
    else if (score >= 70)
        grade = 'C';
    else if (score >= 60)
        grade = 'D';
    else
        grade = 'F';
    return {
        results,
        score,
        grade,
        passed: results.filter((r) => r.passed).length,
        failed: results.filter((r) => !r.passed).length,
        errors,
        warnings,
        info,
        timestamp: new Date().toISOString(),
    };
}
/**
 * Create an empty validation report when dockerfilelint is not available
 */
function createEmptyReport() {
    return {
        results: [],
        score: 100,
        grade: 'A',
        passed: 0,
        failed: 0,
        errors: 0,
        warnings: 0,
        info: 0,
        timestamp: new Date().toISOString(),
    };
}
