// containerization-assist-mcp/dist-cjs/src/validation/dockerfile-validator.js
"use strict";
/**
 * Dockerfile Validator - Security and best practice enforcement (Functional)
 *
 * Invariant: All validation rules maintain backwards compatibility
 * Trade-off: Runtime parsing cost over build-time validation for flexibility
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateDockerfileContent = void 0;
const dockerParser = __importStar(require("docker-file-parser"));
const validate_dockerfile_1 = __importDefault(require("validate-dockerfile"));
const types_1 = require("../types");
const errors_1 = require("../lib/errors");
const logger_1 = require("../lib/logger");
const core_types_1 = require("./core-types");
const dockerfilelint_adapter_1 = require("./dockerfilelint-adapter");
const merge_reports_1 = require("./merge-reports");
const regex_patterns_1 = require("../lib/regex-patterns");
/**
 * Get argument value from docker command
 */
const getArgValue = (command) => {
    if (typeof command.args === 'string') {
        return command.args;
    }
    if (Array.isArray(command.args)) {
        return command.args.join(' ');
    }
    if (typeof command.args === 'object' && command.args !== null) {
        // ENV supports both space-separated and = formats
        return Object.entries(command.args)
            .map(([k, v]) => `${k}=${v}`)
            .join(' ');
    }
    return '';
};
/**
 * Validation rules for Dockerfile analysis
 */
const DOCKERFILE_RULES = [
    {
        id: 'no-root-user',
        name: 'Non-root user required',
        description: 'Container should run as non-root user for security',
        check: (commands) => {
            const userCommands = commands.filter((cmd) => cmd.name === 'USER');
            if (userCommands.length === 0)
                return false;
            const lastUser = userCommands[userCommands.length - 1];
            if (!lastUser)
                return false;
            const userArg = getArgValue(lastUser);
            return userArg !== 'root' && userArg !== '0';
        },
        message: 'Container should run as non-root user',
        severity: core_types_1.ValidationSeverity.ERROR,
        fix: 'Add USER directive with non-root user (e.g., USER node)',
        category: core_types_1.ValidationCategory.SECURITY,
    },
    {
        id: 'no-sudo-install',
        name: 'No sudo in containers',
        description: 'Avoid installing sudo in containers for security',
        check: (commands) => {
            const runCommands = commands.filter((cmd) => cmd.name === 'RUN');
            return !runCommands.some((cmd) => {
                const args = getArgValue(cmd);
                return regex_patterns_1.SUDO_INSTALL.test(args);
            });
        },
        message: 'Avoid installing sudo in containers',
        severity: core_types_1.ValidationSeverity.WARNING,
        fix: 'Remove sudo installation, use specific user permissions instead',
        category: core_types_1.ValidationCategory.SECURITY,
    },
    {
        id: 'specific-base-image',
        name: 'Use specific version tags',
        description: 'Base images should use specific versions for reproducibility',
        check: (commands) => {
            const fromCommands = commands.filter((cmd) => cmd.name === 'FROM');
            return fromCommands.every((cmd) => {
                const image = getArgValue(cmd);
                // Multi-stage builds use 'FROM image AS stage' syntax
                const cleanImage = image?.split(regex_patterns_1.AS_CLAUSE)?.[0];
                return cleanImage && cleanImage.includes(':') && !regex_patterns_1.LATEST_TAG.test(cleanImage);
            });
        },
        message: 'Use specific version tags instead of latest',
        severity: core_types_1.ValidationSeverity.WARNING,
        fix: 'Replace :latest with specific version (e.g., node:18-alpine)',
        category: core_types_1.ValidationCategory.BEST_PRACTICE,
    },
    {
        id: 'has-healthcheck',
        name: 'Health check defined',
        description: 'Containers should define health checks for monitoring',
        check: (commands) => {
            return commands.some((cmd) => cmd.name === 'HEALTHCHECK');
        },
        message: 'Add HEALTHCHECK for container monitoring',
        severity: core_types_1.ValidationSeverity.INFO,
        fix: 'Add HEALTHCHECK CMD curl -f http://localhost/health || exit 1',
        category: core_types_1.ValidationCategory.BEST_PRACTICE,
    },
    {
        id: 'layer-caching-optimization',
        name: 'Optimize layer caching',
        description: 'Dependencies should be copied before application code',
        check: (commands) => {
            const copyCommands = commands.filter((cmd) => cmd.name === 'COPY');
            let packageCopyIndex = -1;
            let sourceCopyIndex = -1;
            copyCommands.forEach((cmd, index) => {
                const args = getArgValue(cmd);
                if (regex_patterns_1.PACKAGE_FILES.test(args)) {
                    if (packageCopyIndex === -1)
                        packageCopyIndex = index;
                }
                if ((args.includes(' . ') || args.endsWith(' .')) && !args.includes('*.')) {
                    if (sourceCopyIndex === -1)
                        sourceCopyIndex = index;
                }
            });
            // Optimal layer caching requires dependencies before source code
            if (sourceCopyIndex === -1)
                return true;
            // Layer invalidation occurs when source changes before dependency install
            if (packageCopyIndex === -1)
                return false;
            return packageCopyIndex < sourceCopyIndex;
        },
        message: 'Copy dependency files before source code for better caching',
        severity: core_types_1.ValidationSeverity.INFO,
        fix: 'COPY package*.json ./ before COPY . .',
        category: core_types_1.ValidationCategory.OPTIMIZATION,
    },
    {
        id: 'no-secrets',
        name: 'No hardcoded secrets',
        description: 'Secrets should not be hardcoded in Dockerfile',
        check: (commands) => {
            const suspicious = [regex_patterns_1.PASSWORD_PATTERN, regex_patterns_1.API_KEY_PATTERN, regex_patterns_1.SECRET_PATTERN, regex_patterns_1.TOKEN_PATTERN];
            return !commands.some((cmd) => {
                if (cmd.name === 'ENV' || cmd.name === 'ARG') {
                    const value = getArgValue(cmd);
                    return suspicious.some((pattern) => pattern.test(value));
                }
                return false;
            });
        },
        message: 'Do not hardcode secrets in Dockerfile',
        severity: core_types_1.ValidationSeverity.ERROR,
        fix: 'Use build arguments or runtime environment variables',
        category: core_types_1.ValidationCategory.SECURITY,
    },
    {
        id: 'multi-stage-optimization',
        name: 'Multi-stage builds for production',
        description: 'Complex applications should use multi-stage builds',
        check: (commands) => {
            const fromCount = commands.filter((cmd) => cmd.name === 'FROM').length;
            const totalLines = commands.length;
            // Multi-stage builds reduce final image size significantly
            if (totalLines < 10)
                return true;
            // Complex builds benefit from separate build and runtime stages
            return fromCount > 1;
        },
        message: 'Consider multi-stage builds for smaller images',
        severity: core_types_1.ValidationSeverity.INFO,
        fix: 'Use FROM ... AS build pattern for build dependencies',
        category: core_types_1.ValidationCategory.OPTIMIZATION,
    },
    {
        id: 'has-expose',
        name: 'Port exposure documented',
        description: 'Applications should document exposed ports',
        check: (commands) => {
            const hasCmd = commands.some((cmd) => cmd.name === 'CMD' || cmd.name === 'ENTRYPOINT');
            const hasExpose = commands.some((cmd) => cmd.name === 'EXPOSE');
            if (!hasCmd)
                return true;
            return hasExpose;
        },
        message: 'Document exposed ports with EXPOSE instruction',
        severity: core_types_1.ValidationSeverity.INFO,
        fix: 'Add EXPOSE <port> for application ports',
        category: core_types_1.ValidationCategory.BEST_PRACTICE,
    },
    {
        id: 'workdir-set',
        name: 'Working directory set',
        description: 'Set a working directory for better organization',
        check: (commands) => {
            return commands.some((cmd) => cmd.name === 'WORKDIR');
        },
        message: 'Set WORKDIR for better file organization',
        severity: core_types_1.ValidationSeverity.INFO,
        fix: 'Add WORKDIR /app or appropriate directory',
        category: core_types_1.ValidationCategory.BEST_PRACTICE,
    },
];
/**
 * Parse Dockerfile content into commands
 */
const parseDockerfile = (content) => {
    try {
        const commands = dockerParser.parse(content);
        return (0, types_1.Success)(commands);
    }
    catch (error) {
        return (0, types_1.Failure)(`Failed to parse Dockerfile: ${(0, errors_1.extractErrorMessage)(error)}`);
    }
};
/**
 * Check if a specific line contains a HEALTHCHECK instruction
 * @param lines Pre-split lines array for efficient lookup
 * @param lineNumber 1-based line number
 */
const isHealthCheckLine = (lines, lineNumber) => {
    const line = lines[lineNumber - 1]; // lineNumber is 1-based
    return line?.trim().toUpperCase().startsWith('HEALTHCHECK') ?? false;
};
/**
 * Validate Dockerfile syntax using external validator
 */
const validateSyntax = (content) => {
    const basicValidation = (0, validate_dockerfile_1.default)(content);
    if (!basicValidation.valid) {
        // Split content once for efficient line lookups
        const lines = content.split('\n');
        // Check if there are true syntax/parse errors (priority 0)
        // Exclude HEALTHCHECK instruction which validate-dockerfile incorrectly flags as invalid
        const syntaxErrors = basicValidation.errors?.filter((err) => err.priority === 0 &&
            (err.message.includes('Missing or misplaced FROM') ||
                (err.message.includes('Invalid instruction') && !isHealthCheckLine(lines, err.line)))) || [];
        if (syntaxErrors.length > 0) {
            return (0, types_1.Failure)(`Failed to parse Dockerfile: ${syntaxErrors[0]?.message || 'Unknown error'}`);
        }
    }
    return (0, types_1.Success)(true);
};
/**
 * Security-critical rules that cap the grade
 */
const SECURITY_RULES = ['no-root-user', 'no-sudo-install', 'no-secrets'];
/**
 * Severity weights for scoring
 */
const SEVERITY_WEIGHTS = {
    [core_types_1.ValidationSeverity.ERROR]: 15,
    [core_types_1.ValidationSeverity.WARNING]: 4,
    [core_types_1.ValidationSeverity.INFO]: 1,
};
/**
 * Calculate validation grade from score
 */
const calculateGrade = (score, hasCriticalSecurity) => {
    let grade;
    if (score >= 90)
        grade = 'A';
    else if (score >= 80)
        grade = 'B';
    else if (score >= 70)
        grade = 'C';
    else if (score >= 60)
        grade = 'D';
    else
        grade = 'F';
    // Cap at C if critical security issues
    if (hasCriticalSecurity && (grade === 'A' || grade === 'B')) {
        grade = 'C';
    }
    return grade;
};
/**
 * Create validation report from results
 */
const createReport = (results) => {
    const errors = results.filter((r) => !r.passed && r.metadata?.severity === core_types_1.ValidationSeverity.ERROR).length;
    const warnings = results.filter((r) => !r.passed && r.metadata?.severity === core_types_1.ValidationSeverity.WARNING).length;
    const info = results.filter((r) => !r.passed && r.metadata?.severity === core_types_1.ValidationSeverity.INFO).length;
    const passed = results.filter((r) => r.passed).length;
    const total = results.length;
    // Calculate weighted score
    const deductions = errors * SEVERITY_WEIGHTS[core_types_1.ValidationSeverity.ERROR] +
        warnings * SEVERITY_WEIGHTS[core_types_1.ValidationSeverity.WARNING] +
        info * SEVERITY_WEIGHTS[core_types_1.ValidationSeverity.INFO];
    const score = Math.max(0, 100 - deductions);
    // Check for critical security failures
    const hasCriticalSecurity = results.some((r) => SECURITY_RULES.includes(r.ruleId || '') &&
        r.metadata?.severity === core_types_1.ValidationSeverity.ERROR &&
        !r.passed);
    const grade = calculateGrade(score, hasCriticalSecurity);
    return {
        results,
        score,
        grade,
        passed,
        failed: total - passed,
        errors,
        warnings,
        info,
        timestamp: new Date().toISOString(),
    };
};
/**
 * Detect BuildKit features in Dockerfile content
 */
function detectBuildKitFeatures(content) {
    const lines = content.split('\n');
    const syntaxLine = lines.find((l) => l.startsWith('# syntax='));
    const syntaxMatch = syntaxLine?.replace('# syntax=', '').trim();
    return {
        ...(syntaxMatch && { syntax: syntaxMatch }),
        hasHeredocs: /RUN\s+<</.test(content),
        hasMounts: /RUN\s+--mount/.test(content),
        hasSecrets: /RUN\s+--mount=type=secret/.test(content),
    };
}
/**
 * Validate BuildKit Dockerfile
 * Used when standard parser fails due to BuildKit syntax
 */
function validateBuildKitDockerfile(content) {
    const results = [];
    const lines = content.split('\n');
    lines.forEach((line, idx) => {
        const lineNumber = idx + 1;
        const trimmedLine = line.trim();
        // Skip comments and empty lines
        if (!trimmedLine || trimmedLine.startsWith('#')) {
            return;
        }
        // Check for :latest tags
        if (line.includes('latest') && line.startsWith('FROM')) {
            results.push({
                ruleId: 'specific-base-image',
                isValid: false,
                passed: false,
                errors: [`Line ${lineNumber}: Avoid using :latest tag`],
                warnings: [],
                message: `✗ Use specific version tags: Line ${lineNumber}`,
                suggestions: ['Replace :latest with specific version (e.g., node:20-alpine)'],
                metadata: {
                    severity: core_types_1.ValidationSeverity.WARNING,
                    location: `line ${lineNumber}`,
                    aiEnhanced: false,
                },
            });
        }
        // Check for potential secrets
        if (/password|api_key|secret|token/i.test(line) && !line.startsWith('#')) {
            // Skip if it's in a comment or mount directive (secrets are OK in BuildKit mounts)
            if (!line.includes('--mount=type=secret')) {
                // Extract the variable name from ENV/ARG statements
                const envMatch = line.match(/^\s*(?:ENV|ARG)\s+([A-Z_][A-Z0-9_]*)\s*=/i);
                const variableName = envMatch ? envMatch[1] : 'secret';
                results.push({
                    ruleId: 'no-secrets',
                    isValid: false,
                    passed: false,
                    errors: [`Line ${lineNumber}: Potential secret exposed`],
                    warnings: [],
                    message: `✗ No hardcoded secrets: Do not hardcode secrets in Dockerfile (found: ${variableName})`,
                    suggestions: ['Use build arguments or runtime environment variables'],
                    metadata: {
                        severity: core_types_1.ValidationSeverity.ERROR,
                        location: `line ${lineNumber}`,
                        aiEnhanced: false,
                    },
                });
            }
        }
        // Check for root user
        if (line.startsWith('USER root') || line.startsWith('USER 0')) {
            results.push({
                ruleId: 'no-root-user',
                isValid: false,
                passed: false,
                errors: [`Line ${lineNumber}: Container runs as root user`],
                warnings: [],
                message: `✗ Non-root user required: Line ${lineNumber}`,
                suggestions: ['Add USER directive with non-root user (e.g., USER node)'],
                metadata: {
                    severity: core_types_1.ValidationSeverity.ERROR,
                    location: `line ${lineNumber}`,
                    aiEnhanced: false,
                },
            });
        }
        // Check for package manager optimizations
        if (line.includes('apt-get install') && !line.includes('--no-install-recommends')) {
            results.push({
                ruleId: 'optimize-package-install',
                isValid: false,
                passed: false,
                errors: [`Line ${lineNumber}: Missing --no-install-recommends flag`],
                warnings: [],
                message: `✗ Optimize package install: Line ${lineNumber}`,
                suggestions: ['Add --no-install-recommends to apt-get install commands'],
                metadata: {
                    severity: core_types_1.ValidationSeverity.WARNING,
                    location: `line ${lineNumber}`,
                    aiEnhanced: false,
                },
            });
        }
    });
    // Add positive results for detected BuildKit features
    const buildKit = detectBuildKitFeatures(content);
    if (buildKit.syntax) {
        results.push({
            ruleId: 'buildkit-syntax',
            isValid: true,
            passed: true,
            errors: [],
            warnings: [],
            message: `✓ Uses BuildKit syntax: ${buildKit.syntax}`,
            suggestions: [],
            metadata: {
                severity: core_types_1.ValidationSeverity.INFO,
                location: 'BuildKit syntax',
                aiEnhanced: false,
            },
        });
    }
    if (buildKit.hasMounts) {
        results.push({
            ruleId: 'buildkit-mounts',
            isValid: true,
            passed: true,
            errors: [],
            warnings: [],
            message: '✓ Uses BuildKit mount optimizations',
            suggestions: [],
            metadata: {
                severity: core_types_1.ValidationSeverity.INFO,
                location: 'BuildKit mounts',
                aiEnhanced: false,
            },
        });
    }
    return createReport(results);
}
/**
 * Validate Dockerfile content using functional pipeline
 */
const validateDockerfileContent = async (dockerfileContent, options) => {
    // Detect BuildKit features first
    const buildKit = detectBuildKitFeatures(dockerfileContent);
    // Prepare validation rules
    const validationRules = [...DOCKERFILE_RULES];
    if (buildKit.syntax || buildKit.hasHeredocs || buildKit.hasMounts) {
        // Log BuildKit features detected
        const logger = (0, logger_1.createLogger)({ name: 'dockerfile-validator' });
        logger.info({ buildKit }, 'BuildKit features detected');
        // Try standard parsing first
        const parseResult = parseDockerfile(dockerfileContent);
        if (parseResult.ok) {
            // Standard validation path works despite BuildKit features
            const commands = parseResult.value;
            // Run all validation rules directly
            const results = [];
            for (const rule of validationRules) {
                const passed = rule.check(commands);
                // Special handling for no-secrets rule to include the specific secret name
                let message = passed ? `✓ ${rule.name}` : `✗ ${rule.name}: ${rule.message}`;
                if (!passed && rule.id === 'no-secrets') {
                    // Extract the secret variable name from the ENV/ARG commands
                    const secretVariable = commands.find((cmd) => {
                        if (cmd.name === 'ENV' || cmd.name === 'ARG') {
                            const value = getArgValue(cmd);
                            const suspicious = [regex_patterns_1.PASSWORD_PATTERN, regex_patterns_1.API_KEY_PATTERN, regex_patterns_1.SECRET_PATTERN, regex_patterns_1.TOKEN_PATTERN];
                            return suspicious.some((pattern) => pattern.test(value));
                        }
                        return false;
                    });
                    if (secretVariable) {
                        const value = getArgValue(secretVariable);
                        const variableMatch = value.match(/^([A-Z_][A-Z0-9_]*)\s*=/i);
                        const variableName = variableMatch ? variableMatch[1] : 'secret';
                        message = `✗ ${rule.name}: ${rule.message} (found: ${variableName})`;
                    }
                }
                results.push({
                    ruleId: rule.id,
                    isValid: passed,
                    passed,
                    errors: passed ? [] : [`${rule.name}: ${rule.message}`],
                    warnings: [],
                    message,
                    suggestions: !passed && rule.fix ? [rule.fix] : [],
                    metadata: {
                        severity: rule.severity,
                    },
                });
            }
            const internalReport = createReport(results);
            // Add external linter if enabled
            if (options?.enableExternalLinter !== false) {
                try {
                    const externalReport = await (0, dockerfilelint_adapter_1.lintWithDockerfilelint)(dockerfileContent);
                    return (0, merge_reports_1.mergeReports)(internalReport, externalReport);
                }
                catch (error) {
                    console.warn('External linter failed, using internal validation only:', error);
                    return internalReport;
                }
            }
            return internalReport;
        }
        // Fall back to BuildKit-specific validation
        const buildKitReport = validateBuildKitDockerfile(dockerfileContent);
        // Add external linter if enabled (it might handle BuildKit better)
        if (options?.enableExternalLinter !== false) {
            try {
                const externalReport = await (0, dockerfilelint_adapter_1.lintWithDockerfilelint)(dockerfileContent);
                return (0, merge_reports_1.mergeReports)(buildKitReport, externalReport);
            }
            catch (error) {
                console.warn('External linter failed with BuildKit file, using BuildKit-specific validation:', error);
                return buildKitReport;
            }
        }
        return buildKitReport;
    }
    // Standard validation path for non-BuildKit files
    // First check syntax
    const syntaxResult = validateSyntax(dockerfileContent);
    if (!syntaxResult.ok) {
        return {
            results: [
                {
                    ruleId: 'parse-error',
                    isValid: false,
                    passed: false,
                    errors: [syntaxResult.error],
                    warnings: [],
                    message: syntaxResult.error,
                    metadata: {
                        severity: core_types_1.ValidationSeverity.ERROR,
                    },
                },
            ],
            score: 0,
            grade: 'F',
            passed: 0,
            failed: 1,
            errors: 1,
            warnings: 0,
            info: 0,
            timestamp: new Date().toISOString(),
        };
    }
    // Parse Dockerfile
    const parseResult = parseDockerfile(dockerfileContent);
    if (!parseResult.ok) {
        return {
            results: [
                {
                    ruleId: 'parse-error',
                    isValid: false,
                    passed: false,
                    errors: [parseResult.error],
                    warnings: [],
                    message: parseResult.error,
                    metadata: {
                        severity: core_types_1.ValidationSeverity.ERROR,
                    },
                },
            ],
            score: 0,
            grade: 'F',
            passed: 0,
            failed: 1,
            errors: 1,
            warnings: 0,
            info: 0,
            timestamp: new Date().toISOString(),
        };
    }
    const commands = parseResult.value;
    // Run all validation rules directly
    const results = [];
    for (const rule of validationRules) {
        const passed = rule.check(commands);
        // Special handling for no-secrets rule to include the specific secret name
        let message = passed ? `✓ ${rule.name}` : `✗ ${rule.name}: ${rule.message}`;
        if (!passed && rule.id === 'no-secrets') {
            // Extract the secret variable name from the ENV/ARG commands
            const secretVariable = commands.find((cmd) => {
                if (cmd.name === 'ENV' || cmd.name === 'ARG') {
                    const value = getArgValue(cmd);
                    const suspicious = [regex_patterns_1.PASSWORD_PATTERN, regex_patterns_1.API_KEY_PATTERN, regex_patterns_1.SECRET_PATTERN, regex_patterns_1.TOKEN_PATTERN];
                    return suspicious.some((pattern) => pattern.test(value));
                }
                return false;
            });
            if (secretVariable) {
                const value = getArgValue(secretVariable);
                const variableMatch = value.match(/^([A-Z_][A-Z0-9_]*)\s*=/i);
                const variableName = variableMatch ? variableMatch[1] : 'secret';
                message = `✗ ${rule.name}: ${rule.message} (found: ${variableName})`;
            }
        }
        results.push({
            ruleId: rule.id,
            isValid: passed,
            passed,
            errors: passed ? [] : [`${rule.name}: ${rule.message}`],
            warnings: [],
            message,
            suggestions: !passed && rule.fix ? [rule.fix] : [],
            metadata: {
                severity: rule.severity,
            },
        });
    }
    const internalReport = createReport(results);
    // Add external linter if enabled (default: true)
    if (options?.enableExternalLinter !== false) {
        try {
            const externalReport = await (0, dockerfilelint_adapter_1.lintWithDockerfilelint)(dockerfileContent);
            return (0, merge_reports_1.mergeReports)(internalReport, externalReport);
        }
        catch (error) {
            // If external linter fails, just return internal results
            console.warn('External linter failed, using internal validation only:', error);
            return internalReport;
        }
    }
    return internalReport;
};
exports.validateDockerfileContent = validateDockerfileContent;
