// containerization-assist-mcp/dist-cjs/src/validation/kubernetes-validator.js
"use strict";
/**
 * Kubernetes validation using YAML parser (Functional)
 *
 * Invariant: All validation rules maintain backwards compatibility
 * Trade-off: Runtime YAML parsing cost over build-time validation for flexibility
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createKubernetesValidator = void 0;
const yaml_1 = require("yaml");
const errors_1 = require("../lib/errors");
const core_types_1 = require("./core-types");
/**
 * Check if manifest is a workload type
 */
const isWorkload = (manifest) => {
    const workloadKinds = ['Deployment', 'StatefulSet', 'DaemonSet', 'Job', 'CronJob', 'Pod'];
    return manifest.kind ? workloadKinds.includes(manifest.kind) : false;
};
/**
 * Get pod spec from different workload types
 */
const getPodSpec = (manifest) => {
    if (manifest.kind === 'Pod') {
        return manifest.spec;
    }
    if (manifest.kind === 'Job' || manifest.kind === 'CronJob') {
        const spec = manifest.spec;
        return spec?.jobTemplate?.spec?.template?.spec || spec?.template?.spec;
    }
    return manifest.spec?.template?.spec;
};
/**
 * Get all containers from manifest
 */
const getContainers = (manifest) => {
    const podSpec = getPodSpec(manifest);
    return [...(podSpec?.containers || []), ...(podSpec?.initContainers || [])];
};
/**
 * Kubernetes validation rules
 */
const KUBERNETES_RULES = [
    {
        id: 'has-resource-limits',
        name: 'Resource limits defined',
        description: 'Containers must have CPU and memory limits defined',
        check: (manifest) => {
            if (!isWorkload(manifest))
                return true;
            const containers = getContainers(manifest);
            return containers.every((container) => {
                const resources = container.resources;
                return resources?.limits?.cpu && resources?.limits?.memory;
            });
        },
        message: 'Define CPU and memory limits for all containers',
        severity: core_types_1.ValidationSeverity.ERROR,
        fix: 'Add resources.limits.cpu and resources.limits.memory',
        category: core_types_1.ValidationCategory.BEST_PRACTICE,
    },
    {
        id: 'has-resource-requests',
        name: 'Resource requests defined',
        description: 'Containers should have resource requests for proper scheduling',
        check: (manifest) => {
            if (!isWorkload(manifest))
                return true;
            const containers = getContainers(manifest);
            return containers.every((container) => {
                const resources = container.resources;
                return resources?.requests?.cpu && resources?.requests?.memory;
            });
        },
        message: 'Define CPU and memory requests for proper scheduling',
        severity: core_types_1.ValidationSeverity.WARNING,
        fix: 'Add resources.requests.cpu and resources.requests.memory',
        category: core_types_1.ValidationCategory.BEST_PRACTICE,
    },
    {
        id: 'has-readiness-probe',
        name: 'Readiness probe configured',
        description: 'Containers should have readiness probes for traffic management',
        check: (manifest) => {
            if (!isWorkload(manifest))
                return true;
            const containers = getContainers(manifest);
            return containers.every((container) => container.readinessProbe);
        },
        message: 'Add readiness probe for traffic management',
        severity: core_types_1.ValidationSeverity.WARNING,
        fix: 'Add readinessProbe with httpGet, tcpSocket, or exec',
        category: core_types_1.ValidationCategory.BEST_PRACTICE,
    },
    {
        id: 'has-liveness-probe',
        name: 'Liveness probe configured',
        description: 'Containers should have liveness probes for auto-restart',
        check: (manifest) => {
            if (!isWorkload(manifest))
                return true;
            const containers = getContainers(manifest);
            // Liveness probes prevent zombie containers in production workloads
            return containers.some((container) => container.livenessProbe);
        },
        message: 'Consider adding liveness probe for auto-restart',
        severity: core_types_1.ValidationSeverity.INFO,
        fix: 'Add livenessProbe with httpGet, tcpSocket, or exec',
        category: core_types_1.ValidationCategory.BEST_PRACTICE,
    },
    {
        id: 'no-privileged-containers',
        name: 'No privileged containers',
        description: 'Containers should not run in privileged mode',
        check: (manifest) => {
            if (!isWorkload(manifest))
                return true;
            const containers = getContainers(manifest);
            return containers.every((container) => {
                const securityContext = container.securityContext;
                return !securityContext?.privileged;
            });
        },
        message: 'Containers should not run in privileged mode',
        severity: core_types_1.ValidationSeverity.ERROR,
        fix: 'Remove privileged: true or set to false',
        category: core_types_1.ValidationCategory.SECURITY,
    },
    {
        id: 'security-context-defined',
        name: 'Security context configured',
        description: 'Pods should define appropriate security context',
        check: (manifest) => {
            if (!isWorkload(manifest))
                return true;
            const podSpec = getPodSpec(manifest);
            const securityContext = podSpec?.securityContext;
            return !!(securityContext?.runAsNonRoot ||
                securityContext?.runAsUser ||
                securityContext?.fsGroup);
        },
        message: 'Define pod security context for better security',
        severity: core_types_1.ValidationSeverity.WARNING,
        fix: 'Add securityContext with runAsNonRoot: true and runAsUser',
        category: core_types_1.ValidationCategory.SECURITY,
    },
    {
        id: 'has-labels',
        name: 'Proper labeling',
        description: 'Resources should have meaningful labels',
        check: (manifest) => {
            const labels = manifest.metadata?.labels;
            const hasAppLabel = labels?.app || labels?.['app.kubernetes.io/name'];
            return !!hasAppLabel;
        },
        message: 'Add meaningful labels for resource organization',
        severity: core_types_1.ValidationSeverity.INFO,
        fix: 'Add labels like app, version, component under metadata.labels',
        category: core_types_1.ValidationCategory.BEST_PRACTICE,
    },
    {
        id: 'image-pull-policy',
        name: 'Appropriate image pull policy',
        description: 'Image pull policy should match image tag strategy',
        check: (manifest) => {
            if (!isWorkload(manifest))
                return true;
            const containers = getContainers(manifest);
            return containers.every((container) => {
                const c = container;
                // :latest tags change content, requiring Always pull policy
                if (c.image?.includes(':latest')) {
                    return c.imagePullPolicy === 'Always';
                }
                // Immutable tags don't need Always, reducing network overhead
                return true;
            });
        },
        message: 'Set appropriate imagePullPolicy for image tag strategy',
        severity: core_types_1.ValidationSeverity.INFO,
        fix: 'Use imagePullPolicy: Always for :latest, IfNotPresent for specific tags',
        category: core_types_1.ValidationCategory.BEST_PRACTICE,
    },
    {
        id: 'no-host-network',
        name: 'Avoid host networking',
        description: 'Pods should not use host networking unless necessary',
        check: (manifest) => {
            if (!isWorkload(manifest))
                return true;
            const podSpec = getPodSpec(manifest);
            return !podSpec?.hostNetwork;
        },
        message: 'Avoid hostNetwork unless absolutely necessary',
        severity: core_types_1.ValidationSeverity.WARNING,
        fix: 'Remove hostNetwork: true or use proper service exposure',
        category: core_types_1.ValidationCategory.SECURITY,
    },
    {
        id: 'no-host-path-volumes',
        name: 'Avoid hostPath volumes',
        description: 'Avoid mounting host directories unless necessary',
        check: (manifest) => {
            if (!isWorkload(manifest))
                return true;
            const podSpec = getPodSpec(manifest);
            const volumes = podSpec?.volumes || [];
            return !volumes.some((volume) => volume.hostPath);
        },
        message: 'Avoid hostPath volumes for better security',
        severity: core_types_1.ValidationSeverity.WARNING,
        fix: 'Use configMaps, secrets, or persistent volumes instead',
        category: core_types_1.ValidationCategory.SECURITY,
    },
    {
        id: 'service-has-selector',
        name: 'Service has proper selector',
        description: 'Services should have selectors to target pods',
        check: (manifest) => {
            if (manifest.kind !== 'Service')
                return true;
            return !!(manifest.spec?.selector && Object.keys(manifest.spec.selector).length > 0);
        },
        message: 'Service should have selector to target pods',
        severity: core_types_1.ValidationSeverity.ERROR,
        fix: 'Add spec.selector with labels matching your pods',
        category: core_types_1.ValidationCategory.BEST_PRACTICE,
    },
    {
        id: 'deployment-has-strategy',
        name: 'Deployment strategy defined',
        description: 'Deployments should define update strategy',
        check: (manifest) => {
            if (manifest.kind !== 'Deployment')
                return true;
            return !!manifest.spec?.strategy?.type;
        },
        message: 'Define deployment strategy for updates',
        severity: core_types_1.ValidationSeverity.INFO,
        fix: 'Add strategy.type (RollingUpdate or Recreate)',
        category: core_types_1.ValidationCategory.BEST_PRACTICE,
    },
];
/**
 * Parse YAML documents from content
 */
const parseDocuments = (yamlContent) => {
    // YAML allows multiple K8s resources separated by ---
    const parts = yamlContent.split(/^---\s*$/m);
    const documents = [];
    for (const part of parts) {
        const trimmed = part.trim();
        if (trimmed) {
            try {
                const doc = (0, yaml_1.parse)(trimmed);
                if (doc && typeof doc === 'object') {
                    documents.push(doc);
                }
            }
            catch {
                // Ignore parsing errors for individual documents
            }
        }
    }
    return documents;
};
/**
 * Calculate validation grade from score
 */
const calculateGrade = (score) => {
    if (score >= 90)
        return 'A';
    if (score >= 80)
        return 'B';
    if (score >= 70)
        return 'C';
    if (score >= 60)
        return 'D';
    return 'F';
};
/**
 * Create validation report from results
 */
const createReport = (results) => {
    const errors = results.filter((r) => !r.passed && r.metadata?.severity === core_types_1.ValidationSeverity.ERROR).length;
    const warnings = results.filter((r) => !r.passed && r.metadata?.severity === core_types_1.ValidationSeverity.WARNING).length;
    const info = results.filter((r) => !r.passed && r.metadata?.severity === core_types_1.ValidationSeverity.INFO).length;
    const passed = results.filter((r) => r.passed).length;
    const total = results.length;
    // Weighted scoring prioritizes security and correctness over style
    const score = Math.max(0, 100 - errors * 15 - warnings * 5 - info * 2);
    const grade = calculateGrade(score);
    return {
        results,
        score,
        grade,
        passed,
        failed: total - passed,
        errors,
        warnings,
        info,
        timestamp: new Date().toISOString(),
    };
};
/**
 * Validate Kubernetes YAML content
 */
const validateKubernetesContent = (yamlContent) => {
    try {
        try {
            (0, yaml_1.parse)(yamlContent);
        }
        catch (parseError) {
            return {
                results: [
                    {
                        ruleId: 'parse-error',
                        isValid: false,
                        passed: false,
                        errors: [`Failed to parse YAML: ${(0, errors_1.extractErrorMessage)(parseError)}`],
                        warnings: [],
                        message: `Failed to parse YAML: ${(0, errors_1.extractErrorMessage)(parseError)}`,
                        metadata: {
                            severity: core_types_1.ValidationSeverity.ERROR,
                        },
                    },
                ],
                score: 0,
                grade: 'F',
                passed: 0,
                failed: 1,
                errors: 1,
                warnings: 0,
                info: 0,
                timestamp: new Date().toISOString(),
            };
        }
        // Extract all K8s resources from potentially multi-document YAML
        const documents = parseDocuments(yamlContent);
        if (documents.length === 0) {
            return {
                results: [
                    {
                        ruleId: 'no-documents',
                        isValid: false,
                        passed: false,
                        errors: ['No valid Kubernetes documents found'],
                        warnings: [],
                        message: 'No valid Kubernetes documents found',
                        metadata: {
                            severity: core_types_1.ValidationSeverity.ERROR,
                        },
                    },
                ],
                score: 0,
                grade: 'F',
                passed: 0,
                failed: 1,
                errors: 1,
                warnings: 0,
                info: 0,
                timestamp: new Date().toISOString(),
            };
        }
        const allResults = [];
        let validDocumentCount = 0;
        for (const doc of documents) {
            if (!doc.apiVersion || !doc.kind)
                continue;
            validDocumentCount++;
            for (const rule of KUBERNETES_RULES) {
                const passed = rule.check(doc);
                const resourceName = doc.metadata?.name || doc.kind;
                allResults.push({
                    ruleId: `${resourceName}-${rule.id}`,
                    isValid: passed,
                    passed,
                    errors: passed ? [] : [`[${resourceName}] ${rule.name}: ${rule.message}`],
                    warnings: [],
                    message: passed
                        ? `✓ [${resourceName}] ${rule.name}`
                        : `✗ [${resourceName}] ${rule.name}: ${rule.message}`,
                    suggestions: !passed && rule.fix ? [rule.fix] : [],
                    metadata: {
                        severity: rule.severity,
                        location: `${doc.kind}/${resourceName}`,
                    },
                });
            }
        }
        // Fail-fast if content parses as YAML but contains no K8s resources
        if (validDocumentCount === 0) {
            return {
                results: [
                    {
                        ruleId: 'no-documents',
                        isValid: false,
                        passed: false,
                        errors: ['No valid Kubernetes documents found'],
                        warnings: [],
                        message: 'No valid Kubernetes documents found',
                        metadata: {
                            severity: core_types_1.ValidationSeverity.ERROR,
                        },
                    },
                ],
                score: 0,
                grade: 'F',
                passed: 0,
                failed: 1,
                errors: 1,
                warnings: 0,
                info: 0,
                timestamp: new Date().toISOString(),
            };
        }
        return createReport(allResults);
    }
    catch (error) {
        return {
            results: [
                {
                    ruleId: 'parse-error',
                    isValid: false,
                    passed: false,
                    errors: [`Failed to parse YAML: ${(0, errors_1.extractErrorMessage)(error)}`],
                    warnings: [],
                    message: `Failed to parse YAML: ${(0, errors_1.extractErrorMessage)(error)}`,
                    metadata: {
                        severity: core_types_1.ValidationSeverity.ERROR,
                    },
                },
            ],
            score: 0,
            grade: 'F',
            passed: 0,
            failed: 1,
            errors: 1,
            warnings: 0,
            info: 0,
            timestamp: new Date().toISOString(),
        };
    }
};
/**
 * Get all Kubernetes validation rules
 */
const getKubernetesRules = () => {
    return [...KUBERNETES_RULES];
};
/**
 * Get Kubernetes rules by category
 */
const getKubernetesRulesByCategory = (category) => {
    return KUBERNETES_RULES.filter((rule) => rule.category === category);
};
/**
 * Create a Kubernetes validator factory
 */
const createKubernetesValidator = () => {
    return {
        validate: validateKubernetesContent,
        getRules: getKubernetesRules,
        getCategory: getKubernetesRulesByCategory,
    };
};
exports.createKubernetesValidator = createKubernetesValidator;
/**
 * Standalone validation function for simple use cases
 */
