// containerization-assist-mcp/dist/src/tools/generate-dockerfile/schema.js
/**
 * Schema definition for generate-dockerfile tool
 */
import { z } from 'zod';
import { environment, repositoryPath, platform, DOCKER_PLATFORMS, } from '../shared/schemas.js';
// Re-export for backward compatibility
export { DOCKER_PLATFORMS };
export const generateDockerfileSchema = z.object({
    repositoryPath: repositoryPath.describe('Repository path (automatically normalized to forward slashes on all platforms).'),
    modulePath: z
        .string()
        .optional()
        .describe('Module path for monorepo/multi-module projects to locate where the Dockerfile should be generated (automatically normalized to forward slashes).'),
    language: z.string().optional().describe('Primary programming language (e.g., "java", "python")'),
    languageVersion: z
        .string()
        .optional()
        .describe('Language version (e.g., "17", "3.11", "20"). For Java, this is the java.version from pom.xml or sourceCompatibility from build.gradle.'),
    framework: z.string().optional().describe('Framework used (e.g., "spring", "django")'),
    environment: environment.describe('Target environment (production, development, etc.)'),
    detectedDependencies: z
        .array(z.string())
        .optional()
        .describe('Detected libraries/frameworks/features from repository analysis (e.g., ["redis", "ef-core", "signalr", "mongodb", "health-checks"]). This helps match relevant knowledge entries.'),
    targetPlatform: platform.describe('Target platform for the Docker image (e.g., "linux/amd64", "linux/arm64"). Defaults to linux/amd64 for maximum compatibility. Use this to cross-compile for different architectures (e.g., ARM Mac targeting AMD64 servers).'),
    trafficLevel: z
        .enum(['high', 'medium', 'low'])
        .optional()
        .describe('Expected traffic level for dynamic defaults calculation (affects replica counts and scaling).'),
    criticalityTier: z
        .enum(['tier-1', 'tier-2', 'tier-3'])
        .optional()
        .describe('Criticality tier for dynamic defaults calculation (tier-1=mission-critical, tier-3=low-priority).'),
});
