// containerization-assist-mcp/dist/src/app/index.js
/**
 * Application Entry Point - AppRuntime Implementation
 * Provides type-safe runtime with dependency injection support
 */
import { createLogger } from '../lib/logger.js';
import { ALL_TOOLS } from '../tools/index.js';
import { createMCPServer, OUTPUTFORMAT, registerToolsWithServer, } from '../mcp/mcp-server.js';
import { createOrchestrator } from './orchestrator.js';
import { CHAINHINTSMODE, } from './orchestrator-types.js';
import { createToolLoggerFile, getLogFilePath } from '../lib/tool-logger.js';
import { checkDockerHealth, checkKubernetesHealth } from '../infra/health/checks.js';
import { DEFAULT_CHAIN_HINTS } from './chain-hints.js';
import { loadKnowledgeBase } from '../knowledge/loader.js';
/**
 * Apply tool aliases to create renamed versions of tools
 * Returns both the aliased tools and a reverse mapping (alias -> original)
 */
function applyToolAliases(tools, aliases) {
    if (!aliases) {
        return { aliasedTools: [...tools], aliasToOriginalMap: {} };
    }
    const aliasToOriginalMap = {};
    const aliasedTools = tools.map((tool) => {
        const alias = aliases[tool.name];
        if (!alias)
            return tool;
        // Store reverse mapping (alias -> original)
        aliasToOriginalMap[alias] = tool.name;
        // Create a new tool object with the alias name
        // Type assertion: aliases are treated as valid ToolNames for registration
        return { ...tool, name: alias };
    });
    return { aliasedTools, aliasToOriginalMap };
}
/**
 * Create the containerization assist application with AppRuntime interface
 */
export function createApp(config = {}) {
    const logger = config.logger || createLogger({ name: 'containerization-assist' });
    // Initialize tool logging file at startup
    createToolLoggerFile(logger);
    const tools = config.tools || ALL_TOOLS;
    const { aliasedTools, aliasToOriginalMap } = applyToolAliases(tools, config.toolAliases);
    // Erase per-tool generics for runtime registration; validation still re-parses inputs per schema
    const registryTools = aliasedTools.map((tool) => tool);
    const toolsMap = new Map();
    for (const tool of registryTools) {
        toolsMap.set(tool.name, tool);
    }
    const chainHintsMode = config.chainHintsMode || CHAINHINTSMODE.ENABLED;
    const outputFormat = config.outputFormat || OUTPUTFORMAT.NATURAL_LANGUAGE;
    const orchestratorConfig = {
        chainHintsMode,
        chainHints: DEFAULT_CHAIN_HINTS,
        aliasToOriginalMap,
    };
    const toolList = Array.from(toolsMap.values());
    let activeServer = null;
    let activeMcpServer = null;
    let orchestrator = buildOrchestrator();
    let orchestratorClosed = false;
    function buildOrchestrator() {
        return createOrchestrator({
            registry: toolsMap,
            logger,
            config: orchestratorConfig,
            ...(activeServer && { server: activeServer }),
        });
    }
    function ensureOrchestrator() {
        if (orchestratorClosed) {
            orchestrator = buildOrchestrator();
            orchestratorClosed = false;
        }
        return orchestrator;
    }
    const orchestratedExecute = (request) => {
        logger.info({ toolName: request.toolName }, 'orchestratedExecute called in app/index.ts');
        const orch = ensureOrchestrator();
        logger.info('About to call orch.execute');
        return orch.execute(request);
    };
    return {
        /**
         * Configuration values from createApp
         */
        config: {
            chainHintsMode,
            outputFormat,
        },
        /**
         * Execute a tool with type-safe parameters and results
         */
        execute: async (toolName, params, metadata) => orchestratedExecute({
            toolName: toolName,
            params,
            metadata: {
                ...(metadata?.signal && { signal: metadata.signal }),
                ...(metadata?.progress !== undefined && { progress: metadata.progress }),
                ...(metadata?.sendNotification && { sendNotification: metadata.sendNotification }),
                loggerContext: {
                    transport: metadata?.transport || 'programmatic',
                    requestId: metadata?.requestId,
                    ...metadata,
                },
            },
        }),
        /**
         * Start MCP server with the specified transport
         */
        startServer: async (transport) => {
            if (activeMcpServer) {
                throw new Error('MCP server is already running');
            }
            // Load knowledge base before starting server
            // This will throw if any built-in packs fail to load
            try {
                loadKnowledgeBase();
            }
            catch (error) {
                logger.error({ error }, 'Failed to load knowledge base during server startup');
                throw new Error(`Server startup failed: ${error instanceof Error ? error.message : String(error)}`);
            }
            ensureOrchestrator();
            const serverOptions = {
                logger,
                transport: transport.transport,
                name: 'containerization-assist',
                version: '1.0.0',
                outputFormat,
                chainHintsMode,
            };
            const mcpServer = createMCPServer(toolList, serverOptions, orchestratedExecute);
            activeServer = mcpServer.getServer();
            try {
                await mcpServer.start();
                activeMcpServer = mcpServer;
                return mcpServer;
            }
            catch (error) {
                activeServer = null;
                throw error;
            }
        },
        /**
         * Bind to existing MCP server
         */
        bindToMCP: (server, transportLabel = 'external') => {
            ensureOrchestrator();
            // Extract the underlying SDK Server from McpServer
            const sdkServer = server.server;
            activeServer = sdkServer;
            registerToolsWithServer({
                outputFormat,
                chainHintsMode,
                server,
                tools: toolList,
                logger,
                transport: transportLabel,
                execute: orchestratedExecute,
            });
        },
        /**
         * List all available tools with their metadata
         */
        listTools: () => toolList.map((t) => ({
            name: t.name,
            description: t.description,
            ...(t.version && { version: t.version }),
            ...(t.category && { category: t.category }),
        })),
        /**
         * Perform health check
         */
        healthCheck: async () => {
            const toolCount = toolsMap.size;
            // Check Docker and Kubernetes connectivity in parallel
            const [dockerStatus, k8sStatus] = await Promise.all([
                checkDockerHealth(logger),
                checkKubernetesHealth(logger),
            ]);
            const hasIssues = !dockerStatus.available || !k8sStatus.available;
            const status = hasIssues ? 'unhealthy' : 'healthy';
            return {
                status,
                tools: toolCount,
                message: hasIssues
                    ? `${toolCount} tools loaded, but some dependencies are unavailable`
                    : `${toolCount} tools loaded`,
                dependencies: {
                    docker: dockerStatus,
                    kubernetes: k8sStatus,
                },
            };
        },
        /**
         * Stop the server and orchestrator if running
         */
        stop: async () => {
            if (activeMcpServer) {
                await activeMcpServer.stop();
                activeMcpServer = null;
            }
            activeServer = null;
            if (!orchestratorClosed) {
                orchestrator.close();
                orchestratorClosed = true;
            }
        },
        /**
         * Get the current log file path (if tool logging is enabled)
         */
        getLogFilePath: () => {
            return getLogFilePath();
        },
    };
}
