// containerization-assist-mcp/dist/src/infra/kubernetes/idempotent-apply.js
/**
 * Idempotent Kubernetes resource application using create/patch strategy
 *
 * Note: This function should be called sequentially per resource.
 * Resources are created if they don't exist, or patched if they already exist.
 */
import * as k8s from '@kubernetes/client-node';
import * as yaml from 'js-yaml';
import { applyResource as applyK8sResource } from './resource-operations.js';
/**
 * Creates an idempotent K8s apply function using the consolidated resource-operations module
 *
 * Note: This function uses the consolidated resource-operations module which handles
 * idempotent apply using create/patch strategy. Resources are created if they don't exist,
 * or patched if they already exist.
 */
export function createIdempotentApply(logger, kubeconfig) {
    const kc = new k8s.KubeConfig();
    if (kubeconfig) {
        kc.loadFromString(kubeconfig);
    }
    else {
        kc.loadFromDefault();
    }
    /**
     * Apply K8s resource using the consolidated resource-operations module
     */
    return async function applyResource(resource, _options = {}) {
        // Use the consolidated resource operations module
        return applyK8sResource(kc, resource, logger);
    };
}
/**
 * Parse YAML manifests into K8s resources
 */
export function parseManifests(yamlContent) {
    try {
        // In js-yaml v4, loadAll is safe by default (no code execution)
        const docs = yaml.loadAll(yamlContent);
        return docs.filter((doc) => {
            const resource = doc;
            return resource?.kind && resource.apiVersion;
        });
    }
    catch {
        return [];
    }
}
