// containerization-assist-mcp/dist/src/infra/kubernetes/resource-operations.js
/**
 * Kubernetes Resource Operations - Consolidated K8s resource management
 *
 * Provides operations for creating, updating, deleting, and reading K8s resources.
 * Uses create/patch strategy for idempotent operations.
 */
import * as k8s from '@kubernetes/client-node';
import { Success, Failure } from '../../types/index.js';
import { extractK8sErrorGuidance } from './errors.js';
import { getResourceConfig, getApiMethod } from './types.js';
/**
 * Helper function to check if an error is a 409 Conflict (AlreadyExists) error
 */
function isConflictError(error) {
    return error.response?.statusCode === 409;
}
/**
 * Handle K8s errors with actionable guidance
 */
function handleK8sError(error, operation, logger) {
    const guidance = extractK8sErrorGuidance(error, operation);
    logger.error({ error: guidance.message, operation }, 'K8s operation failed');
    return Failure(guidance.message, guidance);
}
/**
 * Helper function to call create method with appropriate arguments based on whether resource is namespaced
 */
async function callCreateMethod(createMethod, isNamespaced, namespace, resource) {
    if (typeof createMethod !== 'function') {
        throw new TypeError('createMethod must be a function');
    }
    return isNamespaced
        ? await createMethod({ namespace, body: resource })
        : await createMethod({ body: resource });
}
/**
 * Helper function to call patch method with appropriate arguments based on whether resource is namespaced
 */
async function callPatchMethod(patchMethod, isNamespaced, name, namespace, resource) {
    if (typeof patchMethod !== 'function') {
        throw new TypeError('patchMethod must be a function');
    }
    return isNamespaced
        ? await patchMethod({ name, namespace, body: resource })
        : await patchMethod({ name, body: resource });
}
/**
 * Apply a Kubernetes resource using create/patch strategy.
 * Idempotent - safe to call multiple times.
 *
 * Note: This function should be called sequentially per resource.
 * Creates the resource if it doesn't exist, or patches it if it already exists.
 *
 * @param kc - Kubernetes config
 * @param resource - Resource to apply
 * @param logger - Logger instance
 * @returns Success with applied resource, or Failure with error guidance
 */
export async function applyResource(kc, resource, logger) {
    try {
        const namespace = resource.metadata.namespace || 'default';
        const name = resource.metadata.name;
        const kind = resource.kind;
        logger.debug({ kind, name, namespace }, 'Applying K8s resource');
        // Validate manifest has required metadata
        if (!name || name.trim() === '') {
            const errorMessage = 'Resource is missing required metadata.name';
            logger.error({ kind, metadata: resource.metadata }, errorMessage);
            return Failure(errorMessage);
        }
        const config = getResourceConfig(kc, kind);
        // Try to create the resource first (optimistic approach)
        if (config?.api) {
            const createMethod = getApiMethod(config.api, config.createMethod);
            if (typeof createMethod !== 'function') {
                return Failure(`Method ${config.createMethod} not found on API client`);
            }
            try {
                const result = await callCreateMethod(createMethod, config.namespaced, namespace, resource);
                const resourceBody = result.body || result;
                logger.info({ kind, name, namespace }, 'Resource created successfully');
                return Success(resourceBody);
            }
            catch (createError) {
                // If resource already exists, try to update it with patch
                if (isConflictError(createError)) {
                    logger.debug({ kind, name, namespace }, 'Resource exists, updating with patch');
                    const patchMethod = getApiMethod(config.api, config.patchMethod);
                    if (typeof patchMethod !== 'function') {
                        return Failure(`Method ${config.patchMethod} not found on API client`);
                    }
                    const patchResult = await callPatchMethod(patchMethod, config.namespaced, name, namespace, resource);
                    const patchBody = patchResult.body || patchResult;
                    logger.info({ kind, name, namespace }, 'Resource updated successfully');
                    return Success(patchBody);
                }
                // Other errors - propagate
                throw createError;
            }
        }
        else {
            // For unsupported resource types, use the generic KubernetesObjectApi
            const objectApi = k8s.KubernetesObjectApi.makeApiClient(kc);
            try {
                const result = await objectApi.create(resource);
                logger.info({ kind, name, namespace }, 'Resource created successfully (generic API)');
                // KubernetesObjectApi returns the resource directly
                return Success(result);
            }
            catch (createError) {
                // If resource already exists, try to patch it
                if (isConflictError(createError)) {
                    logger.debug({ kind, name, namespace }, 'Resource exists, updating with patch (generic API)');
                    const patchResult = await objectApi.patch(resource);
                    logger.info({ kind, name, namespace }, 'Resource updated successfully (generic API)');
                    // KubernetesObjectApi returns the resource directly
                    return Success(patchResult);
                }
                // Other errors - propagate
                throw createError;
            }
        }
    }
    catch (error) {
        return handleK8sError(error, 'apply resource', logger);
    }
}
