// @azure/arm-msi/dist/commonjs/operations/systemAssignedIdentities.js
"use strict";
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SystemAssignedIdentitiesImpl = void 0;
const tslib_1 = require("tslib");
const coreClient = tslib_1.__importStar(require("@azure/core-client"));
const Mappers = tslib_1.__importStar(require("../models/mappers.js"));
const Parameters = tslib_1.__importStar(require("../models/parameters.js"));
/** Class containing SystemAssignedIdentities operations. */
class SystemAssignedIdentitiesImpl {
    client;
    /**
     * Initialize a new instance of the class SystemAssignedIdentities class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * Gets the systemAssignedIdentity available under the specified RP scope.
     * @param scope The resource provider scope of the resource. Parent resource being extended by Managed
     *              Identities.
     * @param options The options parameters.
     */
    getByScope(scope, options) {
        return this.client.sendOperationRequest({ scope, options }, getByScopeOperationSpec);
    }
}
exports.SystemAssignedIdentitiesImpl = SystemAssignedIdentitiesImpl;
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const getByScopeOperationSpec = {
    path: "/{scope}/providers/Microsoft.ManagedIdentity/identities/default",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.SystemAssignedIdentity,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [Parameters.$host, Parameters.scope],
    headerParameters: [Parameters.accept],
    serializer,
};
