// containerization-assist-mcp/dist-cjs/src/index.js
"use strict";
/**
 * Clean, idiomatic TypeScript API for Container Assist
 * Provides strongly-typed runtime with dependency injection support
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractSchemaShape = exports.verifyDeployTool = exports.tagImageTool = exports.scanImageTool = exports.pushImageTool = exports.prepareClusterTool = exports.opsTool = exports.generateK8sManifestsTool = exports.generateDockerfileTool = exports.fixDockerfileTool = exports.buildImageContextTool = exports.analyzeRepoTool = exports.ALL_TOOLS = exports.tool = exports.getToolDefaults = exports.BUILD_DEFAULTS = exports.ACA_DEFAULTS = exports.CONTAINER_DEFAULTS = exports.K8S_DEFAULTS = exports.withDefaults = exports.createSafeTelemetryEvent = exports.extractSafeTelemetryMetrics = exports.sanitizeToolInput = exports.sanitizePath = exports.hashValue = exports.registerTools = exports.registerTool = exports.createToolHandler = exports.createStandardizedToolTracker = exports.createToolTimer = exports.getToolLogger = exports.createApp = void 0;
// Primary API - AppRuntime with type safety and dependency injection
/**
 * Creates a configured application runtime instance with MCP server integration.
 *
 * This is the primary entry point for creating a containerization-assist application.
 * It sets up dependency injection, tool registration, and MCP transport configuration.
 *
 * @returns AppRuntime instance with all tools registered and ready to execute
 *
 * @example
 * ```typescript
 * import { createApp } from 'containerization-assist';
 *
 * const app = await createApp();
 *
 * // Use MCP server
 * await app.mcp.start();
 *
 * // Or execute tools directly
 * const result = await app.executeTool('analyze-repo', { path: './my-app' });
 * if (result.ok) {
 *   console.log('Analysis:', result.value);
 * }
 * ```
 *
 * @public
 */
var index_js_1 = require("./app/index.js");
Object.defineProperty(exports, "createApp", { enumerable: true, get: function () { return index_js_1.createApp; } });
/**
 * Tool helper utilities for creating consistent tool implementations.
 *
 * - `getToolLogger`: Create a scoped logger for a specific tool
 * - `createToolTimer`: Create a timer for tracking tool execution duration
 * - `createStandardizedToolTracker`: Create a combined logger + timer tracker
 *
 * @example
 * ```typescript
 * import { getToolLogger, createToolTimer } from 'containerization-assist';
 *
 * const logger = getToolLogger('my-tool');
 * const timer = createToolTimer(logger);
 *
 * logger.info('Starting operation');
 * // ... do work ...
 * timer.end({ success: true });
 * ```
 *
 * @public
 */
var tool_helpers_js_1 = require("./lib/tool-helpers.js");
Object.defineProperty(exports, "getToolLogger", { enumerable: true, get: function () { return tool_helpers_js_1.getToolLogger; } });
Object.defineProperty(exports, "createToolTimer", { enumerable: true, get: function () { return tool_helpers_js_1.createToolTimer; } });
Object.defineProperty(exports, "createStandardizedToolTracker", { enumerable: true, get: function () { return tool_helpers_js_1.createStandardizedToolTracker; } });
/**
 * Tool registration helpers for integrating with MCP servers.
 *
 * These utilities allow you to register Container Assist tools with MCP servers
 * while maintaining full control over telemetry, error handling, and lifecycle hooks.
 * All context creation (logger, policy, etc.) is handled automatically.
 *
 * - `createToolHandler`: Create a handler function for use with server.tool()
 * - `registerTool`: Register a single tool with an MCP server
 * - `registerTools`: Register multiple tools at once
 *
 * @example
 * ```typescript
 * import { createApp, ALL_TOOLS, createToolHandler } from 'containerization-assist';
 *
 * const app = createApp();
 *
 * // Option 1: Use createToolHandler for full control
 * for (const tool of ALL_TOOLS) {
 *   server.tool(
 *     tool.name,
 *     tool.description,
 *     tool.inputSchema,
 *     createToolHandler(app, tool.name, {
 *       onSuccess: (result, toolName) => {
 *         myTelemetry.trackSuccess(toolName);
 *       },
 *     })
 *   );
 * }
 *
 * // Option 2: Use registerTools for convenience
 * import { registerTools } from 'containerization-assist';
 * registerTools(server, app, ALL_TOOLS);
 * ```
 *
 * @public
 */
var tool_registration_js_1 = require("./lib/tool-registration.js");
Object.defineProperty(exports, "createToolHandler", { enumerable: true, get: function () { return tool_registration_js_1.createToolHandler; } });
Object.defineProperty(exports, "registerTool", { enumerable: true, get: function () { return tool_registration_js_1.registerTool; } });
Object.defineProperty(exports, "registerTools", { enumerable: true, get: function () { return tool_registration_js_1.registerTools; } });
/**
 * Telemetry sanitization utilities for safe logging.
 *
 * Provides functions to hash, obfuscate, and extract safe metrics from tool inputs
 * and outputs. Protects customer workspace information (file paths, code snippets,
 * identifiers) while preserving aggregate metrics for observability.
 *
 * Security Principle: "When in doubt, hash it out"
 *
 * - `hashValue`: Hash a string using SHA256 (first 16 chars)
 * - `sanitizePath`: Replace path segments with hashed values
 * - `sanitizeToolInput`: Remove or hash sensitive fields from tool inputs
 * - `extractSafeTelemetryMetrics`: Extract only aggregate metrics from tool results
 * - `createSafeTelemetryEvent`: Create a complete safe telemetry event
 * - `SafeTelemetryEvent`: Type for safe telemetry events
 *
 * @example
 * ```typescript
 * import { createSafeTelemetryEvent } from 'containerization-assist';
 *
 * // In your telemetry callback
 * onSuccess: (result, toolName, params) => {
 *   const event = createSafeTelemetryEvent(
 *     toolName,
 *     params,
 *     { ok: true, value: result }
 *   );
 *   // Event contains only hashed paths and aggregate metrics
 *   telemetry.track(event);
 * }
 * ```
 *
 * @see docs/TELEMETRY_GUIDELINES.md for detailed security guidelines
 * @public
 */
var telemetry_utils_js_1 = require("./lib/telemetry-utils.js");
Object.defineProperty(exports, "hashValue", { enumerable: true, get: function () { return telemetry_utils_js_1.hashValue; } });
Object.defineProperty(exports, "sanitizePath", { enumerable: true, get: function () { return telemetry_utils_js_1.sanitizePath; } });
Object.defineProperty(exports, "sanitizeToolInput", { enumerable: true, get: function () { return telemetry_utils_js_1.sanitizeToolInput; } });
Object.defineProperty(exports, "extractSafeTelemetryMetrics", { enumerable: true, get: function () { return telemetry_utils_js_1.extractSafeTelemetryMetrics; } });
Object.defineProperty(exports, "createSafeTelemetryEvent", { enumerable: true, get: function () { return telemetry_utils_js_1.createSafeTelemetryEvent; } });
/**
 * Parameter defaulting utilities for tool inputs.
 *
 * Provides default values for common tool parameters across different categories:
 * - `K8S_DEFAULTS`: Kubernetes deployment defaults (namespace, replicas, etc.)
 * - `CONTAINER_DEFAULTS`: Container image defaults (registry, platform, etc.)
 * - `ACA_DEFAULTS`: Azure Container Apps defaults
 * - `BUILD_DEFAULTS`: Docker build defaults (context, dockerfile path, etc.)
 * - `withDefaults`: Merge user params with defaults
 * - `getToolDefaults`: Get defaults for a specific tool by name
 *
 * @public
 */
var param_defaults_js_1 = require("./lib/param-defaults.js");
Object.defineProperty(exports, "withDefaults", { enumerable: true, get: function () { return param_defaults_js_1.withDefaults; } });
Object.defineProperty(exports, "K8S_DEFAULTS", { enumerable: true, get: function () { return param_defaults_js_1.K8S_DEFAULTS; } });
Object.defineProperty(exports, "CONTAINER_DEFAULTS", { enumerable: true, get: function () { return param_defaults_js_1.CONTAINER_DEFAULTS; } });
Object.defineProperty(exports, "ACA_DEFAULTS", { enumerable: true, get: function () { return param_defaults_js_1.ACA_DEFAULTS; } });
Object.defineProperty(exports, "BUILD_DEFAULTS", { enumerable: true, get: function () { return param_defaults_js_1.BUILD_DEFAULTS; } });
Object.defineProperty(exports, "getToolDefaults", { enumerable: true, get: function () { return param_defaults_js_1.getToolDefaults; } });
/**
 * Helper function to create a tool definition with the unified Tool interface.
 *
 * Ensures tools follow the standard structure with proper typing and schema validation.
 *
 * @example
 * ```typescript
 * import { tool } from 'containerization-assist';
 * import { z } from 'zod';
 *
 * const mySchema = z.object({ input: z.string() });
 *
 * export default tool({
 *   name: 'my-tool',
 *   description: 'My custom tool',
 *   version: '1.0.0',
 *   schema: mySchema,
 *   handler: async (input, ctx) => {
 *     return Success({ result: input.input.toUpperCase() });
 *   }
 * });
 * ```
 *
 * @public
 */
var tool_js_1 = require("./types/tool.js");
Object.defineProperty(exports, "tool", { enumerable: true, get: function () { return tool_js_1.tool; } });
/**
 * All available MCP tools for containerization workflows.
 *
 * Tools are organized by workflow stage:
 * 1. Analysis: `analyzeRepoTool` - Detect language, framework, and dependencies
 * 2. Dockerfile: `generateDockerfileTool`, `fixDockerfileTool`
 * 3. Build: `buildImageContextTool`, `scanImageTool`, `tagImageTool`, `pushImageTool`
 * 4. Deploy: `generateK8sManifestsTool`, `prepareClusterTool`, `verifyDeployTool`
 * 5. Operations: `opsTool` - Operational utilities
 *
 * @public
 */
var index_js_2 = require("./tools/index.js");
Object.defineProperty(exports, "ALL_TOOLS", { enumerable: true, get: function () { return index_js_2.ALL_TOOLS; } });
Object.defineProperty(exports, "analyzeRepoTool", { enumerable: true, get: function () { return index_js_2.analyzeRepoTool; } });
Object.defineProperty(exports, "buildImageContextTool", { enumerable: true, get: function () { return index_js_2.buildImageContextTool; } });
Object.defineProperty(exports, "fixDockerfileTool", { enumerable: true, get: function () { return index_js_2.fixDockerfileTool; } });
Object.defineProperty(exports, "generateDockerfileTool", { enumerable: true, get: function () { return index_js_2.generateDockerfileTool; } });
Object.defineProperty(exports, "generateK8sManifestsTool", { enumerable: true, get: function () { return index_js_2.generateK8sManifestsTool; } });
Object.defineProperty(exports, "opsTool", { enumerable: true, get: function () { return index_js_2.opsTool; } });
Object.defineProperty(exports, "prepareClusterTool", { enumerable: true, get: function () { return index_js_2.prepareClusterTool; } });
Object.defineProperty(exports, "pushImageTool", { enumerable: true, get: function () { return index_js_2.pushImageTool; } });
Object.defineProperty(exports, "scanImageTool", { enumerable: true, get: function () { return index_js_2.scanImageTool; } });
Object.defineProperty(exports, "tagImageTool", { enumerable: true, get: function () { return index_js_2.tagImageTool; } });
Object.defineProperty(exports, "verifyDeployTool", { enumerable: true, get: function () { return index_js_2.verifyDeployTool; } });
/**
 * Utility to extract the shape of a Zod schema for telemetry and type introspection.
 *
 * @public
 */
var zod_utils_js_1 = require("./lib/zod-utils.js");
Object.defineProperty(exports, "extractSchemaShape", { enumerable: true, get: function () { return zod_utils_js_1.extractSchemaShape; } });
